// @project-chip/matter.js/dist/esm/model/models/ClusterModel.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterElement, Globals } from "../elements/index.js";
import { AttributeModel } from "./AttributeModel.js";
import { CommandModel } from "./CommandModel.js";
import { DatatypeModel } from "./DatatypeModel.js";
import { EventModel } from "./EventModel.js";
import { Model } from "./Model.js";
class ClusterModel extends Model {
  constructor(definition) {
    super(definition);
    this.tag = ClusterElement.Tag;
    this.isTypeScope = true;
  }
  get attributes() {
    return this.all(AttributeModel);
  }
  get commands() {
    return this.all(CommandModel);
  }
  get events() {
    return this.all(EventModel);
  }
  get datatypes() {
    return this.all(DatatypeModel);
  }
  get revision() {
    let revision = 1;
    const revisionAttr = this.get(AttributeModel, Globals.ClusterRevision.id);
    if (typeof revisionAttr?.default === "number") {
      revision = revisionAttr.default;
    }
    return revision;
  }
  get features() {
    return this.featureMap?.children ?? [];
  }
  get featureMap() {
    return this.get(AttributeModel, Globals.FeatureMap.id);
  }
  get children() {
    return super.children;
  }
  set children(children) {
    super.children = children;
  }
  static {
    Model.constructors[ClusterElement.Tag] = this;
  }
}
export {
  ClusterModel
};
