// @project-chip/matter.js/dist/esm/model/models/CommandModel.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { CommandElement } from "../elements/index.js";
import { ModelTraversal } from "../logic/ModelTraversal.js";
import { Model } from "./Model.js";
import { ValueModel } from "./ValueModel.js";
class CommandModel extends ValueModel {
  constructor(definition) {
    super(definition);
    this.tag = CommandElement.Tag;
  }
  get isRequest() {
    return this.direction === CommandElement.Direction.Request;
  }
  get isResponse() {
    return this.direction === CommandElement.Direction.Response;
  }
  get responseModel() {
    return new ModelTraversal().findResponse(this);
  }
  /**
   * Commands may re-use the ID for request and response so identification
   * requires the ID in conjunction with the direction.
   */
  get key() {
    return `${super.key}:${this.direction}`;
  }
  static {
    Model.constructors[CommandElement.Tag] = this;
  }
  static {
    this.Tag = CommandElement.Tag;
  }
}
export {
  CommandModel
};
