// @project-chip/matter.js/dist/esm/model/models/MatterModel.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Globals, MatterElement } from "../elements/index.js";
import { Matter } from "../standard/index.js";
import { AttributeModel } from "./AttributeModel.js";
import { ClusterModel } from "./ClusterModel.js";
import { DatatypeModel } from "./DatatypeModel.js";
import { DeviceTypeModel } from "./DeviceTypeModel.js";
import { FabricModel } from "./FabricModel.js";
import { Model } from "./Model.js";
class MatterModel extends Model {
  constructor(definition = Matter, globals = Object.values(Globals)) {
    const children = [...globals, ...definition.children || []];
    super({ ...definition, children });
    this.tag = MatterElement.Tag;
    this.isTypeScope = true;
  }
  /**
   * Clusters.
   */
  get clusters() {
    return this.all(ClusterModel);
  }
  /**
   * Device types.
   */
  get deviceTypes() {
    return this.all(DeviceTypeModel);
  }
  /**
   * Global datatypes.
   */
  get datatypes() {
    return this.all(DatatypeModel);
  }
  /**
   * Global attributes.
   */
  get attributes() {
    return this.all(AttributeModel);
  }
  /**
   * Fabrics.
   */
  get fabrics() {
    return this.all(FabricModel);
  }
  get children() {
    return super.children;
  }
  set children(children) {
    super.children = children;
  }
  static {
    Model.constructors[MatterElement.Tag] = this;
  }
}
export {
  MatterModel
};
