// @project-chip/matter.js/dist/esm/model/models/Model.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { InternalError } from "../../common/MatterError.js";
import { ElementTag } from "../definitions/index.js";
import { ModelTraversal } from "../logic/ModelTraversal.js";
const CHILDREN = Symbol("children");
const PARENT = Symbol("parent");
class Model {
  static {
    CHILDREN, PARENT;
  }
  /**
   * Did validation find errors?
   */
  get valid() {
    return !this.errors;
  }
  /**
   * The full path ("." delimited) in the Matter tree.
   */
  get path() {
    if (this.parent && this.parent.tag !== ElementTag.Matter) {
      return `${this.parent.path}.${this.name}`;
    } else {
      return this.name;
    }
  }
  /**
   * The structural parent.  This is the model for the element that contains
   * this element's definition.
   */
  get parent() {
    return this[PARENT];
  }
  set parent(parent) {
    if (this.parent === parent) {
      return;
    }
    if (this.parent) {
      const index = this.parent.children.indexOf(this);
      if (index !== -1) {
        this.parent.children.splice(index, 1);
      }
    }
    if (!parent) {
      delete this[PARENT];
    } else {
      this[PARENT] = parent;
    }
    if (parent) {
      parent.children.push(this);
    }
  }
  /**
   * Element view of children.  For TypeScript this allows children to be
   * added as elements.  For JavaScript this is identical to children().
   */
  get elements() {
    if (!this[CHILDREN]) {
      this.children = [];
    }
    return this[CHILDREN];
  }
  /**
   * Allows subclasses to pull a working ID from an alternate source.
   */
  get effectiveId() {
    return this.id;
  }
  /**
   * Get a string that uniquely identifies this model.  This is normally
   * the effective ID but some models require a generated identifier.
   */
  get key() {
    return this.effectiveId?.toString();
  }
  /**
   * Children of models are always models.
   */
  get children() {
    if (!this[CHILDREN]) {
      this.children = [];
    }
    return this[CHILDREN];
  }
  /**
   * Children can be added as models or elements.
   */
  set children(children) {
    this[CHILDREN] = new Proxy(Array(), {
      get: (target, p, receiver) => {
        let result = Reflect.get(target, p, receiver);
        if (!(result instanceof Model) && typeof p === "string" && p.match(/^[0-9]+$/)) {
          result = Model.create(result);
          result[PARENT] = this;
          Reflect.set(target, p, result, receiver);
        }
        return result;
      },
      set: (target, p, newValue, receiver) => {
        if (typeof p === "string" && p.match(/^[0-9]+$/)) {
          if (typeof newValue !== "object" || newValue === null || !newValue.tag) {
            throw new InternalError("Child must be Model or AnyElement");
          }
        }
        const result = Reflect.set(target, p, newValue, receiver);
        if (newValue instanceof Model) {
          if (newValue[PARENT] !== this) {
            if (newValue[PARENT]) {
              newValue.parent = void 0;
            }
            newValue[PARENT] = this;
          }
        }
        return result;
      },
      deleteProperty: (target, p) => {
        let child;
        if (typeof p === "string" && p.match(/^[0-9]+$/)) {
          child = target[Number.parseInt(p)];
        }
        if (Reflect.deleteProperty(target, p) && child) {
          if (child[PARENT] === this && this.children.indexOf(child) === -1) {
            child[PARENT] = void 0;
          }
          return true;
        }
        return false;
      }
    });
    children = [...children];
    this[CHILDREN].push(...children);
  }
  static {
    /**
     * Factory support.  Populated by derivatives upon definition.
     */
    this.constructors = {};
  }
  /**
   * In some circumstances the base type can be inferred.  This inference
   * happens here.
   *
   * Does not recurse so only returns the direct base type.
   */
  get effectiveType() {
    return this.type;
  }
  /**
   * Get a Model for my base type, if any.
   */
  get base() {
    return new ModelTraversal().findBase(this);
  }
  /**
   * Get shadow model, if any.  A "shadow" is an element in my parent's
   * inheritance hierarchy that I override.
   */
  get shadow() {
    return new ModelTraversal().findShadow(this);
  }
  /**
   * Get the first global base type.  This may have semantic meaning more
   * specific than the base primitive type.
   */
  get globalBase() {
    return new ModelTraversal().findGlobalBase(this);
  }
  /**
   * A local or parent xref.
   */
  get effectiveXref() {
    return new ModelTraversal().findXref(this);
  }
  /**
   * The set of tags from which this model may derive.
   */
  get allowedBaseTags() {
    return [this.tag];
  }
  /**
   * Add a child.  children.push works too but only accepts models.
   */
  add(...children) {
    this.children.push(...children);
  }
  /**
   * Create a model for an element.
   */
  static create(definition) {
    if (typeof definition !== "object") {
      throw new InternalError(`Model definition must be object, not ${typeof definition}`);
    }
    const t = definition["tag"];
    const constructor = Model.constructors[t];
    if (!constructor) {
      throw new InternalError(`Unknown element tag "${t}"`);
    }
    return new constructor(definition);
  }
  /**
   * Retrieve all models of a specific element type from local scope.
   *
   * @param constructor model class or a predicate object
   */
  all(constructor) {
    return this.children.filter((c) => c instanceof constructor);
  }
  /**
   * Retrieve a specific model by ID or name.
   */
  get(constructor, key) {
    return this.children.find(
      (c) => c instanceof constructor && typeof key === "number" ? c.effectiveId === key : c.name === key
    );
  }
  /**
   * Retrieve a model of a specific type from the ownership hierarchy.
   */
  owner(constructor) {
    return new ModelTraversal().findOwner(constructor, this.parent);
  }
  /**
   * Check identity of element by name or ID.
   */
  is(key) {
    if (typeof key === "number") {
      return this.id === key;
    } else if (typeof key === "function") {
      return key(this);
    }
    return this.name === key;
  }
  /**
   * Record a validation error for this model.
   */
  error(code, message) {
    if (!this.errors) {
      this.errors = [];
    }
    this.errors.push({
      code,
      source: this.path,
      message,
      xref: this.effectiveXref?.toString()
    });
  }
  /**
   * Convert model to JSON.
   */
  toJSON() {
    return this.valueOf();
  }
  /**
   * Convert to non-class structure.
   */
  valueOf() {
    const result = {};
    for (const key in this) {
      switch (key) {
        case "parent":
        case "errors":
        case "isTypeScope":
        case "isType":
          continue;
        default:
          result[key] = this[key];
      }
    }
    return result;
  }
  /**
   * Apply a function to all tree elements.
   */
  visit(visitor) {
    return new ModelTraversal().visit(this, visitor);
  }
  /**
   * Find all children that reference a specific type.
   */
  references(type) {
    return new ModelTraversal().findReferences(this, type);
  }
  /**
   * Search the inheritance chain for a child property.
   */
  member(key, allowedTags = [ElementTag.Datatype, ElementTag.Attribute]) {
    return new ModelTraversal().findMember(this, key, allowedTags);
  }
  /**
   * Does this model derive from another?
   */
  instanceOf(other) {
    return new ModelTraversal().instanceOf(this, other);
  }
  constructor(definition) {
    for (const [k, v] of Object.entries(definition)) {
      if (v !== void 0) {
        this[k] = v;
      }
    }
    if (this.xref) {
      this.xref = Model.CrossReference.get(this.xref);
    }
  }
}
((Model2) => {
  class CrossReference {
    static {
      this.instances = {};
    }
    constructor({ document, section }) {
      this.document = document;
      this.section = section;
    }
    toString() {
      return `${this.document}\xA7${this.section}`;
    }
    static get(xref) {
      const key = `${xref.document}:${xref.section}`;
      const canonical = this.instances[key];
      if (canonical) {
        return canonical;
      }
      return this.instances[key] = new CrossReference(xref);
    }
  }
  Model2.CrossReference = CrossReference;
})(Model || (Model = {}));
export {
  Model
};
