// @project-chip/matter.js/dist/cjs/net/fake/NetworkFake.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NetworkFake_exports = {};
__export(NetworkFake_exports, {
  NetworkFake: () => NetworkFake
});
module.exports = __toCommonJS(NetworkFake_exports);
var import_Network = require("../Network.js");
var import_SimulatedNetwork = require("./SimulatedNetwork.js");
var import_UdpChannelFake = require("./UdpChannelFake.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class NetworkFake extends import_Network.Network {
  constructor(mac, ips) {
    super();
    this.mac = mac;
    this.ips = ips;
  }
  getNetInterfaces() {
    return [import_SimulatedNetwork.FAKE_INTERFACE_NAME];
  }
  getIpMac(_netInterface) {
    return { mac: this.mac, ips: this.ips };
  }
  createUdpChannel(options) {
    return import_UdpChannelFake.UdpChannelFake.create(this, options);
  }
}
