// @project-chip/matter.js/dist/cjs/net/fake/UdpChannelFake.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var UdpChannelFake_exports = {};
__export(UdpChannelFake_exports, {
  UdpChannelFake: () => UdpChannelFake
});
module.exports = __toCommonJS(UdpChannelFake_exports);
var import_Ip = require("../../util/Ip.js");
var import_Network = require("../Network.js");
var import_SimulatedNetwork = require("./SimulatedNetwork.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class UdpChannelFake {
  constructor(localAddress, listeningAddress, listeningPort) {
    this.localAddress = localAddress;
    this.listeningAddress = listeningAddress;
    this.netListeners = new Array();
    this.simulatedNetwork = import_SimulatedNetwork.SimulatedNetwork.get();
    this.listeningPort = listeningPort ?? 1024 + Math.floor(Math.random() * 64511);
  }
  static async create(network, { listeningAddress, listeningPort, netInterface, type }) {
    const { ips } = network.getIpMac(netInterface ?? import_SimulatedNetwork.FAKE_INTERFACE_NAME);
    const ipv4 = type === "udp4";
    const localAddress = ips.filter((ip) => (0, import_Ip.isIPv4)(ip) || !ipv4)[0];
    if (localAddress === void 0)
      throw new import_Network.NetworkError("No matching IP on the specified interface");
    return new UdpChannelFake(localAddress, listeningAddress, listeningPort);
  }
  onData(listener) {
    const netListener = this.simulatedNetwork.onUdpData(this.listeningAddress, this.listeningPort, listener);
    this.netListeners.push(netListener);
    return netListener;
  }
  async send(host, port, data) {
    this.simulatedNetwork.sendUdp(this.localAddress, this.listeningPort, host, port, data);
  }
  async close() {
    for (const netListener of this.netListeners) {
      await netListener.close();
    }
    this.netListeners.length = 0;
  }
}
