// @project-chip/matter.js/dist/esm/time/Time.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { NoProviderError } from "../common/MatterError.js";
class Time {
  static {
    this.get = () => DefaultTime;
  }
  static {
    this.now = () => Time.get().now();
  }
  static {
    this.nowMs = () => Time.get().nowMs();
  }
  static {
    this.getTimer = (durationMs, callback) => Time.get().getTimer(durationMs, callback);
  }
  static {
    this.getPeriodicTimer = (intervalMs, callback) => Time.get().getPeriodicTimer(intervalMs, callback);
  }
  static {
    this.sleep = async (durationMs) => new Promise((resolve) => Time.get().getTimer(durationMs, resolve).start());
  }
}
const DefaultTime = new class extends Time {
  now() {
    return /* @__PURE__ */ new Date();
  }
  nowMs() {
    return this.now().getTime();
  }
  getTimer() {
    throw new NoProviderError("Default time provider does not implement timers");
  }
  getPeriodicTimer() {
    return this.getTimer();
  }
}();
if (typeof MatterHooks !== "undefined") {
  MatterHooks.timeSetup?.(Time);
}
export {
  Time
};
