// @project-chip/matter.js/dist/cjs/model/definitions/FieldValue.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FieldValue_exports = {};
__export(FieldValue_exports, {
  FieldValue: () => FieldValue
});
module.exports = __toCommonJS(FieldValue_exports);
var import_MatterError = require("../../common/MatterError.js");
var import_String = require("../../util/String.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var FieldValue;
((FieldValue2) => {
  FieldValue2.percent = "percent";
  FieldValue2.celsius = "celsius";
  FieldValue2.reference = "reference";
  FieldValue2.properties = "properties";
  function is(value, type) {
    return value && value.type === type;
  }
  FieldValue2.is = is;
  FieldValue2.Invalid = Symbol("invalid");
  function Reference(name) {
    return { type: FieldValue2.reference, name };
  }
  FieldValue2.Reference = Reference;
  function Celsius(value) {
    return { type: FieldValue2.celsius, value };
  }
  FieldValue2.Celsius = Celsius;
  function Percent(value) {
    return { type: FieldValue2.percent, value };
  }
  FieldValue2.Percent = Percent;
  function serialize(value) {
    if (value === null) {
      return "null";
    }
    if (is(value, FieldValue2.reference)) {
      return value.name;
    }
    if (is(value, FieldValue2.celsius)) {
      return `${value.value}\xB0C`;
    }
    if (is(value, FieldValue2.percent)) {
      return `${value.value}%';`;
    }
    if (is(value, FieldValue2.properties)) {
      return (0, import_String.serialize)(value.properties);
    }
    return value.toString();
  }
  FieldValue2.serialize = serialize;
  function numericValue(value, typeName) {
    if (typeof value === "boolean") {
      return value ? 1 : 0;
    }
    if (typeof value === "number") {
      return value;
    }
    if (is(value, FieldValue2.celsius)) {
      switch (typeName) {
        case "temperature":
        case "temp-diff":
          return value.value * 100;
        case "temp-u8":
        case "temp-s8":
          return value.value * 10;
      }
      return;
    }
    if (is(value, FieldValue2.percent)) {
      switch (typeName) {
        case "percent100ths":
          return value.value * 100;
        default:
          return value.value;
      }
    }
  }
  FieldValue2.numericValue = numericValue;
  function objectValue(value) {
    if (is(value, FieldValue2.properties)) {
      return value.properties;
    }
  }
  FieldValue2.objectValue = objectValue;
  function unwrap(value, typeName) {
    if (value === null || typeof value !== "object" || Array.isArray(value) || value instanceof Date) {
      return value;
    }
    const type = value.type;
    switch (type) {
      case "properties":
        return objectValue(value);
      case "reference":
        return;
      case "percent":
      case "celsius":
        return numericValue(value, typeName);
      case void 0:
        throw new import_MatterError.UnexpectedDataError(`Field value objects must be wrappers with "type" field`);
      default:
        throw new import_MatterError.UnexpectedDataError(`Unsupported wrapped object type "${type}"`);
    }
  }
  FieldValue2.unwrap = unwrap;
  function referenced(value) {
    if (is(value, FieldValue2.reference)) {
      return value.name;
    }
  }
  FieldValue2.referenced = referenced;
})(FieldValue || (FieldValue = {}));
