// @project-chip/matter.js/dist/cjs/model/definitions/Metatype.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Metatype_exports = {};
__export(Metatype_exports, {
  Metatype: () => Metatype
});
module.exports = __toCommonJS(Metatype_exports);
var import_ByteArray = require("../../util/ByteArray.js");
var import_FieldValue = require("./FieldValue.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Metatype = /* @__PURE__ */ ((Metatype2) => {
  Metatype2["any"] = "any";
  Metatype2["boolean"] = "boolean";
  Metatype2["bitmap"] = "bitmap";
  Metatype2["enum"] = "enum";
  Metatype2["integer"] = "integer";
  Metatype2["float"] = "float";
  Metatype2["bytes"] = "bytes";
  Metatype2["array"] = "array";
  Metatype2["object"] = "object";
  Metatype2["string"] = "string";
  Metatype2["date"] = "date";
  return Metatype2;
})(Metatype || {});
((Metatype2) => {
  function hasChildren(type) {
    switch (type) {
      case "enum" /* enum */:
      case "bitmap" /* bitmap */:
      case "object" /* object */:
        return true;
      default:
        return false;
    }
  }
  Metatype2.hasChildren = hasChildren;
  function native(type) {
    switch (type) {
      case "boolean" /* boolean */:
        return Boolean;
      case "integer" /* integer */:
        return BigInt;
      case "bitmap" /* bitmap */:
      case "enum" /* enum */:
      case "float" /* float */:
        return Number;
      case "bytes" /* bytes */:
        return import_ByteArray.ByteArray;
      case "array" /* array */:
        return Array;
      case "object" /* object */:
        return Object;
      case "string" /* string */:
        return String;
      case "date" /* date */:
        return Date;
    }
  }
  Metatype2.native = native;
  function cast(type, value) {
    if (value === void 0 || value === null || type === "any" /* any */) {
      return value;
    }
    if (value === "null") {
      return null;
    }
    if (value === "") {
      if (type === "string" /* string */) {
        return "";
      }
      return void 0;
    }
    if (import_FieldValue.FieldValue.is(value, import_FieldValue.FieldValue.reference)) {
      return value;
    }
    switch (type) {
      case "string" /* string */:
        return value.toString();
      case "boolean" /* boolean */:
        if (typeof value === "string") {
          value = value.trim().toLowerCase();
        }
        return value === "false" || value === "no" || !!value;
      case "bitmap" /* bitmap */:
      case "enum" /* enum */:
        const id = Number(value);
        if (Number.isNaN(id)) {
          return `${value}`;
        }
        return id;
      case "integer" /* integer */:
        if (typeof value === "string") {
          let type2;
          if (value.endsWith("\xB0C")) {
            type2 = import_FieldValue.FieldValue.celsius;
          } else if (value.endsWith("%")) {
            type2 = import_FieldValue.FieldValue.percent;
          }
          if (type2) {
            value = Number.parseInt(value);
            if (Number.isNaN(value)) {
              return import_FieldValue.FieldValue.Invalid;
            }
            return { type: type2, value };
          }
          const match = value.match(/^(0x[0-9a-f]+|\d+)/);
          if (match) {
            value = match[1];
          }
        }
        try {
          switch (typeof value) {
            case "string":
            case "number":
            case "bigint":
            case "boolean":
              break;
            default:
              if (import_FieldValue.FieldValue.is(value, import_FieldValue.FieldValue.celsius) || import_FieldValue.FieldValue.is(value, import_FieldValue.FieldValue.percent)) {
                return value;
              }
              return import_FieldValue.FieldValue.Invalid;
          }
          const i = BigInt(value);
          const n = Number(i);
          if (BigInt(n) === i) {
            return n;
          }
          return i;
        } catch (e) {
          if (e instanceof SyntaxError) {
            return import_FieldValue.FieldValue.Invalid;
          }
          throw e;
        }
      case "float" /* float */:
        const float = Number(value);
        if (Number.isNaN(float)) {
          return import_FieldValue.FieldValue.Invalid;
        }
        return float.valueOf();
      case "date" /* date */:
        if (value instanceof Date) {
          return value;
        }
        if (typeof value !== "string") {
          return import_FieldValue.FieldValue.Invalid;
        }
        value = new Date(value);
        if (Number.isNaN(value.valueOf())) {
          return import_FieldValue.FieldValue.Invalid;
        }
        return value;
      case "object" /* object */:
        if (value === "null") {
          return null;
        }
        if (import_FieldValue.FieldValue.is(value, import_FieldValue.FieldValue.properties)) {
          return value;
        }
        break;
      case "bytes" /* bytes */:
        if (value === "empty") {
          return void 0;
        }
        if (!(value instanceof Uint8Array)) {
          return import_FieldValue.FieldValue.Invalid;
        }
        return value;
      case "array" /* array */:
        if (value === "0" || value === "{0,0}") {
          return;
        }
        if (value === "empty" || value === "[]" || value === "{}") {
          return [];
        }
        if (Array.isArray(value) && !value.length) {
          return value;
        }
        return import_FieldValue.FieldValue.Invalid;
    }
    return import_FieldValue.FieldValue.Invalid;
  }
  Metatype2.cast = cast;
})(Metatype || (Metatype = {}));
