// @project-chip/matter.js/dist/cjs/model/definitions/StatusCode.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var StatusCode_exports = {};
__export(StatusCode_exports, {
  StatusCode: () => StatusCode
});
module.exports = __toCommonJS(StatusCode_exports);
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var StatusCode = /* @__PURE__ */ ((StatusCode2) => {
  StatusCode2[StatusCode2["Success"] = 0] = "Success";
  StatusCode2[StatusCode2["Failure"] = 1] = "Failure";
  StatusCode2[StatusCode2["InvalidSubscription"] = 125] = "InvalidSubscription";
  StatusCode2[StatusCode2["UnsupportedAccess"] = 126] = "UnsupportedAccess";
  StatusCode2[StatusCode2["UnsupportedEndpoint"] = 127] = "UnsupportedEndpoint";
  StatusCode2[StatusCode2["InvalidAction"] = 128] = "InvalidAction";
  StatusCode2[StatusCode2["UnsupportedCommand"] = 129] = "UnsupportedCommand";
  StatusCode2[StatusCode2["InvalidCommand"] = 133] = "InvalidCommand";
  StatusCode2[StatusCode2["UnsupportedAttribute"] = 134] = "UnsupportedAttribute";
  StatusCode2[StatusCode2["ConstraintError"] = 135] = "ConstraintError";
  StatusCode2[StatusCode2["UnsupportedWrite"] = 136] = "UnsupportedWrite";
  StatusCode2[StatusCode2["ResourceExhausted"] = 137] = "ResourceExhausted";
  StatusCode2[StatusCode2["NotFound"] = 139] = "NotFound";
  StatusCode2[StatusCode2["UnreportableAttribute"] = 140] = "UnreportableAttribute";
  StatusCode2[StatusCode2["InvalidDataType"] = 141] = "InvalidDataType";
  StatusCode2[StatusCode2["UnsupportedRead"] = 143] = "UnsupportedRead";
  StatusCode2[StatusCode2["DataVersionMismatch"] = 146] = "DataVersionMismatch";
  StatusCode2[StatusCode2["UnsupportedNode"] = 155] = "UnsupportedNode";
  StatusCode2[StatusCode2["Busy"] = 156] = "Busy";
  StatusCode2[StatusCode2["UnsupportedCluster"] = 195] = "UnsupportedCluster";
  StatusCode2[StatusCode2["NoUpstreamSubscription"] = 197] = "NoUpstreamSubscription";
  StatusCode2[StatusCode2["NeedsTimedInteraction"] = 198] = "NeedsTimedInteraction";
  StatusCode2[StatusCode2["UnsupportedEvent"] = 199] = "UnsupportedEvent";
  StatusCode2[StatusCode2["PathsExhausted"] = 200] = "PathsExhausted";
  StatusCode2[StatusCode2["TimedRequestMismatch"] = 201] = "TimedRequestMismatch";
  StatusCode2[StatusCode2["FailsafeRequired"] = 202] = "FailsafeRequired";
  return StatusCode2;
})(StatusCode || {});
