// @project-chip/matter.js/dist/cjs/protocol/interaction/EventHandler.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EventHandler_exports = {};
__export(EventHandler_exports, {
  EventHandler: () => EventHandler
});
module.exports = __toCommonJS(EventHandler_exports);
var import_Cluster = require("../../cluster/Cluster.js");
var import_ClusterHelper = require("../../cluster/ClusterHelper.js");
var import_Logger = require("../../log/Logger.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("EventHandler");
const MAX_EVENTS = 1e4;
class EventHandler {
  constructor(storage) {
    this.eventNumber = 0;
    this.storedEventCount = 0;
    this.events = {
      [import_Cluster.EventPriority.Critical]: new Array(),
      [import_Cluster.EventPriority.Info]: new Array(),
      [import_Cluster.EventPriority.Debug]: new Array()
    };
    this.eventStorage = storage.createContext("EventHandler");
    this.eventNumber = this.eventStorage.get("lastEventNumber", this.eventNumber);
    logger.debug(`Set/Restore last event number: ${this.eventNumber}`);
  }
  getEvents(eventPath, filters) {
    const eventFilter = filters !== void 0 && filters.length > 0 ? (event) => filters.some((filter) => filter.eventMin !== void 0 && event.eventNumber >= filter.eventMin) : () => true;
    const events = new Array();
    const { endpointId, clusterId, eventId } = eventPath;
    for (const priority of [import_Cluster.EventPriority.Critical, import_Cluster.EventPriority.Info, import_Cluster.EventPriority.Debug]) {
      const eventsToCheck = this.events[priority];
      for (const event of eventsToCheck) {
        if (eventFilter(event) && endpointId === event.endpointId && clusterId === event.clusterId && eventId === event.eventId) {
          events.push(event);
        }
      }
    }
    logger.debug(
      `Got ${events.length} events for ${(0, import_ClusterHelper.resolveEventName)(eventPath)} with filters: ${import_Logger.Logger.toJSON(filters)}`
    );
    return events;
  }
  pushEvent(event) {
    const eventData = {
      eventNumber: ++this.eventNumber,
      ...event
    };
    logger.debug(`Received event: ${JSON.stringify(eventData)}`);
    this.events[event.priority].push(eventData);
    this.storedEventCount++;
    this.eventStorage.set("lastEventNumber", this.eventNumber);
    this.cleanUpEvents();
    return eventData;
  }
  cleanUpEvents() {
    if (this.storedEventCount < MAX_EVENTS)
      return;
    const eventsToDelete = this.storedEventCount - MAX_EVENTS;
    for (const priority of [import_Cluster.EventPriority.Debug, import_Cluster.EventPriority.Info, import_Cluster.EventPriority.Critical]) {
      const events = this.events[priority];
      if (events.length > 0) {
        const removedEvents = events.splice(0, events.length - eventsToDelete);
        logger.debug(`Removed ${removedEvents.length} events from priority ${priority}`);
        return;
      }
    }
  }
}
