// @project-chip/matter.js/dist/cjs/protocol/interaction/InteractionServer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InteractionServer_exports = {};
__export(InteractionServer_exports, {
  INTERACTION_MODEL_REVISION: () => INTERACTION_MODEL_REVISION,
  INTERACTION_PROTOCOL_ID: () => INTERACTION_PROTOCOL_ID,
  InteractionServer: () => InteractionServer,
  attributePathToId: () => attributePathToId,
  clusterPathToId: () => clusterPathToId,
  commandPathToId: () => commandPathToId,
  eventPathToId: () => eventPathToId,
  genericElementPathToId: () => genericElementPathToId
});
module.exports = __toCommonJS(InteractionServer_exports);
var import_Cluster = require("../../cluster/Cluster.js");
var import_AttributeServer = require("../../cluster/server/AttributeServer.js");
var import_ClusterServerTypes = require("../../cluster/server/ClusterServerTypes.js");
var import_MessageCodec = require("../../codec/MessageCodec.js");
var import_MatterError = require("../../common/MatterError.js");
var import_TryCatchHandler = require("../../common/TryCatchHandler.js");
var import_ValidationError = require("../../common/ValidationError.js");
var import_Crypto = require("../../crypto/Crypto.js");
var import_Logger = require("../../log/Logger.js");
var import_EventHandler = require("../../protocol/interaction/EventHandler.js");
var import_SecureSession = require("../../session/SecureSession.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
var import_Number = require("../../util/Number.js");
var import_AttributeDataDecoder = require("./AttributeDataDecoder.js");
var import_InteractionEndpointStructure = require("./InteractionEndpointStructure.js");
var import_InteractionMessenger = require("./InteractionMessenger.js");
var import_InteractionProtocol = require("./InteractionProtocol.js");
var import_StatusCode = require("./StatusCode.js");
var import_SubscriptionHandler = require("./SubscriptionHandler.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const INTERACTION_PROTOCOL_ID = 1;
const INTERACTION_MODEL_REVISION = 11;
const logger = import_Logger.Logger.get("InteractionServer");
function genericElementPathToId(endpointId, clusterId, elementId) {
  return `${endpointId}/${clusterId}/${elementId}`;
}
function commandPathToId({ endpointId, clusterId, commandId }) {
  return genericElementPathToId(endpointId, clusterId, commandId);
}
function attributePathToId({ endpointId, clusterId, attributeId }) {
  return genericElementPathToId(endpointId, clusterId, attributeId);
}
function eventPathToId({ endpointId, clusterId, eventId }) {
  return genericElementPathToId(endpointId, clusterId, eventId);
}
function clusterPathToId({ nodeId, endpointId, clusterId }) {
  return `${nodeId}/${endpointId}/${clusterId}`;
}
class InteractionServer {
  constructor(storage, options) {
    this.storage = storage;
    this.options = options;
    this.endpointStructure = new import_InteractionEndpointStructure.InteractionEndpointStructure();
    this.nextSubscriptionId = import_Crypto.Crypto.getRandomUInt32();
    this.eventHandler = new import_EventHandler.EventHandler(this.storage);
    this.subscriptionMap = /* @__PURE__ */ new Map();
    this.isClosing = false;
  }
  getId() {
    return INTERACTION_PROTOCOL_ID;
  }
  setRootEndpoint(endpoint) {
    this.endpointStructure.initializeFromEndpoint(endpoint);
    for (const endpoint2 of this.endpointStructure.endpoints.values()) {
      for (const cluster of endpoint2.getAllClusterServers()) {
        const clusterServer = (0, import_ClusterServerTypes.asClusterServerInternal)(cluster);
        clusterServer._setStorage(this.storage.createContext(`Cluster-${endpoint2.id}-${cluster.id}`));
        clusterServer._registerEventHandler(this.eventHandler);
      }
    }
    for (const subscription of this.subscriptionMap.values()) {
      subscription.updateSubscription();
    }
  }
  async onNewExchange(exchange) {
    if (this.isClosing)
      return;
    await new import_InteractionMessenger.InteractionServerMessenger(exchange).handleRequest(
      (readRequest) => this.handleReadRequest(exchange, readRequest),
      (writeRequest, message) => this.handleWriteRequest(exchange, writeRequest, message),
      (subscribeRequest, messenger) => this.handleSubscribeRequest(exchange, subscribeRequest, messenger),
      (invokeRequest, message) => this.handleInvokeRequest(exchange, invokeRequest, message),
      (timedRequest) => this.handleTimedRequest(exchange, timedRequest)
    );
  }
  handleReadRequest(exchange, {
    attributeRequests,
    dataVersionFilters,
    eventRequests,
    eventFilters,
    isFabricFiltered,
    interactionModelRevision
  }) {
    logger.debug(
      `Received read request from ${exchange.channel.name}: attributes:${attributeRequests?.map((path) => this.endpointStructure.resolveAttributeName(path)).join(", ") ?? "none"}, events:${eventRequests?.map((path) => this.endpointStructure.resolveEventName(path)).join(", ") ?? "none"} isFabricFiltered=${isFabricFiltered}`
    );
    if (interactionModelRevision > INTERACTION_MODEL_REVISION) {
      logger.debug(
        `Interaction model revision of sender ${interactionModelRevision} is higher than supported ${INTERACTION_MODEL_REVISION}.`
      );
    }
    if (attributeRequests === void 0 && eventRequests === void 0) {
      throw new import_StatusCode.StatusResponseError(
        "Only Read requests with attributeRequests or eventRequests are supported right now",
        import_StatusCode.StatusCode.UnsupportedRead
      );
    }
    const dataVersionFilterMap = new Map(
      dataVersionFilters?.map(({ path, dataVersion }) => [clusterPathToId(path), dataVersion]) ?? []
    );
    if (dataVersionFilterMap.size > 0) {
      logger.debug(
        `DataVersionFilters: ${Array.from(dataVersionFilterMap.entries()).map(([path, version]) => `${path}=${version}`).join(", ")}`
      );
    }
    const attributeReportsPayload = attributeRequests?.flatMap(
      (path) => {
        const attributes = this.endpointStructure.getAttributes([path]);
        if (attributes.length === 0) {
          const { endpointId, clusterId, attributeId } = path;
          if (endpointId === void 0 || clusterId === void 0 || attributeId === void 0) {
            logger.debug(
              `Read attribute from ${exchange.channel.name}: ${this.endpointStructure.resolveAttributeName(path)}: ignore non-existing attribute`
            );
          } else {
            return (0, import_TryCatchHandler.tryCatch)(
              () => {
                this.endpointStructure.validateConcreteAttributePath(
                  endpointId,
                  clusterId,
                  attributeId
                );
                throw new import_MatterError.InternalError(
                  "validateConcreteAttributePath should throw StatusResponseError but did not."
                );
              },
              import_StatusCode.StatusResponseError,
              (error) => {
                logger.debug(
                  `Read attribute from ${exchange.channel.name}: ${this.endpointStructure.resolveAttributeName(path)}: unsupported path: Status=${error.code}`
                );
                return [{ attributeStatus: { path, status: { status: error.code } } }];
              }
            );
          }
        }
        return attributes.flatMap(({ path: path2, attribute }) => {
          const { nodeId, endpointId, clusterId } = path2;
          const { value, version } = (0, import_TryCatchHandler.tryCatch)(
            () => attribute.getWithVersion(exchange.session, isFabricFiltered),
            import_SecureSession.NoAssociatedFabricError,
            () => {
              if (endpointId === void 0 || clusterId === void 0) {
                throw new import_MatterError.MatterFlowError("Should never happen");
              }
              const cluster = this.endpointStructure.getClusterServer(endpointId, clusterId);
              if (cluster === void 0) {
                throw new import_MatterError.MatterFlowError("Should never happen");
              }
              return {
                version: cluster.clusterDataVersion,
                value: []
              };
            }
          );
          const versionFilterValue = endpointId !== void 0 && clusterId !== void 0 ? dataVersionFilterMap.get(clusterPathToId({ nodeId, endpointId, clusterId })) : void 0;
          if (versionFilterValue !== void 0 && versionFilterValue === version) {
            logger.debug(
              `Read attribute from ${exchange.channel.name}: ${this.endpointStructure.resolveAttributeName(path2)}=${import_Logger.Logger.toJSON(
                value
              )} (version=${version}) ignored because of dataVersionFilter`
            );
            return [];
          }
          logger.debug(
            `Read attribute from ${exchange.channel.name}: ${this.endpointStructure.resolveAttributeName(
              path2
            )}=${import_Logger.Logger.toJSON(value)} (version=${version})`
          );
          const { schema } = attribute;
          return [{ attributeData: { path: path2, dataVersion: version, payload: value, schema } }];
        });
      }
    );
    const eventReportsPayload = eventRequests?.flatMap(
      (path) => {
        const events = this.endpointStructure.getEvents([path]);
        if (events.length === 0) {
          const { endpointId, clusterId, eventId } = path;
          if (endpointId === void 0 || clusterId === void 0 || eventId === void 0) {
            logger.debug(
              `Read event from ${exchange.channel.name}: ${this.endpointStructure.resolveEventName(
                path
              )}: ignore non-existing event`
            );
          } else {
            return (0, import_TryCatchHandler.tryCatch)(
              () => {
                this.endpointStructure.validateConcreteEventPath(endpointId, clusterId, eventId);
                throw new import_MatterError.InternalError(
                  "validateConcreteEventPath should throw StatusResponseError but did not."
                );
              },
              import_StatusCode.StatusResponseError,
              (error) => {
                logger.debug(
                  `Read event from ${exchange.channel.name}: ${this.endpointStructure.resolveEventName(path)}: unsupported path: Status=${error.code}`
                );
                return [{ eventStatus: { path, status: { status: error.code } } }];
              }
            );
          }
        }
        return events.flatMap(({ path: path2, event }) => {
          const matchingEvents = this.eventHandler.getEvents(path2, eventFilters);
          logger.debug(
            `Read event from ${exchange.channel.name}:  ${this.endpointStructure.resolveEventName(
              path2
            )}=${import_Logger.Logger.toJSON(matchingEvents)}`
          );
          const { schema } = event;
          return matchingEvents.map(({ eventNumber, priority, epochTimestamp, data }) => ({
            eventData: {
              path: path2,
              eventNumber,
              priority,
              epochTimestamp,
              payload: data,
              schema
            }
          }));
        }).sort((a, b) => a.eventData.eventNumber - b.eventData.eventNumber);
      }
    );
    return {
      interactionModelRevision: INTERACTION_MODEL_REVISION,
      suppressResponse: false,
      attributeReportsPayload,
      // TODO Return compressed response once https://github.com/project-chip/connectedhomeip/issues/29359 is solved
      eventReportsPayload
    };
  }
  handleWriteRequest(exchange, { suppressResponse, timedRequest, writeRequests, interactionModelRevision }, { packetHeader: { sessionType } }) {
    logger.debug(
      `Received write request from ${exchange.channel.name}: ${writeRequests.map((req) => this.endpointStructure.resolveAttributeName(req.path)).join(", ")}, suppressResponse=${suppressResponse}`
    );
    if (interactionModelRevision > INTERACTION_MODEL_REVISION) {
      logger.debug(
        `Interaction model revision of sender ${interactionModelRevision} is higher than supported ${INTERACTION_MODEL_REVISION}.`
      );
    }
    const receivedWithinTimedInteraction = exchange.hasActiveTimedInteraction();
    if (exchange.hasExpiredTimedInteraction()) {
      exchange.clearTimedInteraction();
      throw new import_StatusCode.StatusResponseError(`Timed request window expired. Decline write request.`, import_StatusCode.StatusCode.Timeout);
    }
    if (timedRequest !== exchange.hasTimedInteraction()) {
      throw new import_StatusCode.StatusResponseError(
        `timedRequest flag of write interaction (${timedRequest}) mismatch with expected timed interaction (${receivedWithinTimedInteraction}).`,
        import_StatusCode.StatusCode.TimedRequestMismatch
      );
    }
    if (receivedWithinTimedInteraction) {
      logger.debug(`Write request from ${exchange.channel.name} received while timed interaction is running.`);
      exchange.clearTimedInteraction();
      if (sessionType !== import_MessageCodec.SessionType.Unicast) {
        throw new import_StatusCode.StatusResponseError(
          "Write requests are only allowed on unicast sessions when a timed interaction is running.",
          import_StatusCode.StatusCode.InvalidAction
        );
      }
    }
    if (sessionType === import_MessageCodec.SessionType.Group && !suppressResponse) {
      throw new import_StatusCode.StatusResponseError(
        "Write requests are only allowed as group casts when suppressResponse=true.",
        import_StatusCode.StatusCode.InvalidAction
      );
    }
    const writeData = (0, import_AttributeDataDecoder.normalizeAttributeData)(writeRequests, true);
    const writeResults = writeData.flatMap(
      (values) => {
        const { path, dataVersion } = values[0];
        const attributes = this.endpointStructure.getAttributes([path], true);
        const { endpointId, clusterId, attributeId } = path;
        if (attributes.length === 0) {
          if (endpointId === void 0 || clusterId === void 0 || attributeId === void 0) {
            logger.debug(
              `Write from ${exchange.channel.name}: ${this.endpointStructure.resolveAttributeName(
                path
              )}: ignore non-existing (wildcard) attribute`
            );
          } else {
            return (0, import_TryCatchHandler.tryCatch)(
              () => {
                if (this.endpointStructure.validateConcreteAttributePath(
                  endpointId,
                  clusterId,
                  attributeId
                )) {
                  throw new import_StatusCode.StatusResponseError(
                    `Attribute ${attributeId} is not writable.`,
                    import_StatusCode.StatusCode.UnsupportedWrite
                  );
                }
                throw new import_MatterError.InternalError(
                  "validateConcreteAttributePath check should throw StatusResponseError but did not."
                );
              },
              import_StatusCode.StatusResponseError,
              (error) => {
                logger.debug(
                  `Write from ${exchange.channel.name}: ${this.endpointStructure.resolveAttributeName(
                    path
                  )} not allowed: Status=${error.code}`
                );
                return [{ path, statusCode: error.code }];
              }
            );
          }
        } else if (attributes.length === 1 && endpointId !== void 0 && clusterId !== void 0 && attributeId !== void 0) {
          const { attribute } = attributes[0];
          if (attribute.requiresTimedInteraction && !receivedWithinTimedInteraction) {
            logger.debug(`This write requires a timed interaction which is not initialized.`);
            return [{ path, statusCode: import_StatusCode.StatusCode.NeedsTimedInteraction }];
          }
          if (attribute instanceof import_AttributeServer.FabricScopedAttributeServer && (!exchange.session.isSecure() || !exchange.session.getFabric())) {
            logger.debug(`This write requires a secure session with a fabric assigned which is missing.`);
            return [{ path, statusCode: import_StatusCode.StatusCode.UnsupportedAccess }];
          }
          if (dataVersion !== void 0) {
            const cluster = this.endpointStructure.getClusterServer(endpointId, clusterId);
            if (cluster !== void 0 && dataVersion !== cluster.clusterDataVersion) {
              logger.debug(
                `This write requires a specific data version (${dataVersion}) which do not match the current cluster data version (${cluster.clusterDataVersion}).`
              );
              return [{ path, statusCode: import_StatusCode.StatusCode.DataVersionMismatch }];
            }
          }
        }
        return attributes.map(({ path: path2, attribute }) => {
          const { schema, defaultValue } = attribute;
          try {
            const value = (0, import_AttributeDataDecoder.decodeAttributeValueWithSchema)(schema, values, defaultValue);
            logger.debug(
              `Handle write request from ${exchange.channel.name} resolved to: ${this.endpointStructure.resolveAttributeName(path2)}=${import_Logger.Logger.toJSON(
                value
              )} (Version=${dataVersion})`
            );
            if (!(attribute instanceof import_AttributeServer.AttributeServer) && !(attribute instanceof import_AttributeServer.FabricScopedAttributeServer)) {
              throw new import_StatusCode.StatusResponseError(
                "Fixed attributes cannot be written",
                import_StatusCode.StatusCode.UnsupportedWrite
              );
            }
            if (attribute.requiresTimedInteraction && !receivedWithinTimedInteraction) {
              logger.debug(`This write requires a timed interaction which is not initialized.`);
              throw new import_StatusCode.StatusResponseError(
                "This write requires a timed interaction which is not initialized.",
                import_StatusCode.StatusCode.NeedsTimedInteraction
              );
            }
            attribute.set(value, exchange.session);
          } catch (error) {
            if (attributes.length === 1) {
              logger.error(
                `Error while handling write request from ${exchange.channel.name} to ${this.endpointStructure.resolveAttributeName(path2)}: ${error.message}`
              );
              if (error instanceof import_StatusCode.StatusResponseError) {
                return { path: path2, statusCode: error.code, clusterStatusCode: error.clusterCode };
              }
              return { path: path2, statusCode: import_StatusCode.StatusCode.ConstraintError };
            } else {
              logger.debug(
                `While handling write request from ${exchange.channel.name} to ${this.endpointStructure.resolveAttributeName(path2)} ignored: ${error.message}`
              );
            }
          }
          return { path: path2, statusCode: import_StatusCode.StatusCode.Success };
        });
      }
    );
    const errorResults = writeResults.filter(({ statusCode }) => statusCode !== import_StatusCode.StatusCode.Success);
    logger.debug(
      `Write request from ${exchange.channel.name} done ${errorResults.length ? `with following errors: ${errorResults.map(
        ({ path, statusCode }) => `${this.endpointStructure.resolveAttributeName(path)}=${import_Logger.Logger.toJSON(statusCode)}`
      ).join(", ")}` : "without errors"}`
    );
    return {
      interactionModelRevision: INTERACTION_MODEL_REVISION,
      writeResponses: writeResults.map(({ path, statusCode, clusterStatusCode }) => ({
        path,
        status: { status: statusCode, clusterStatus: clusterStatusCode }
      }))
    };
  }
  async handleSubscribeRequest(exchange, {
    minIntervalFloorSeconds,
    maxIntervalCeilingSeconds,
    attributeRequests,
    dataVersionFilters,
    eventRequests,
    eventFilters,
    keepSubscriptions,
    isFabricFiltered,
    interactionModelRevision
  }, messenger) {
    logger.debug(
      `Received subscribe request from ${exchange.channel.name} (keepSubscriptions=${keepSubscriptions}, isFabricFiltered=${isFabricFiltered})`
    );
    if (interactionModelRevision > INTERACTION_MODEL_REVISION) {
      logger.debug(
        `Interaction model revision of sender ${interactionModelRevision} is higher than supported ${INTERACTION_MODEL_REVISION}.`
      );
    }
    (0, import_SecureSession.assertSecureSession)(exchange.session, "Subscriptions are only implemented on secure sessions");
    const session = exchange.session;
    const fabric = session.getFabric();
    if (fabric === void 0)
      throw new import_StatusCode.StatusResponseError(
        "Subscriptions are only implemented after a fabric has been assigned",
        import_StatusCode.StatusCode.InvalidAction
      );
    if ((!Array.isArray(attributeRequests) || attributeRequests.length === 0) && (!Array.isArray(eventRequests) || eventRequests.length === 0)) {
      throw new import_StatusCode.StatusResponseError("No attributes or events requested", import_StatusCode.StatusCode.InvalidAction);
    }
    logger.debug(
      `Subscribe to attributes:${attributeRequests?.map((path) => this.endpointStructure.resolveAttributeName(path)).join(", ") ?? "none"}, events:${eventRequests?.map((path) => this.endpointStructure.resolveEventName(path)).join(", ") ?? "none"}`
    );
    if (dataVersionFilters !== void 0 && dataVersionFilters.length > 0) {
      logger.debug(
        `DataVersionFilters: ${dataVersionFilters.map(
          ({ path: { nodeId, endpointId, clusterId }, dataVersion }) => `${clusterPathToId({ nodeId, endpointId, clusterId })}=${dataVersion}`
        ).join(", ")}`
      );
    }
    if (eventFilters !== void 0 && eventFilters.length > 0)
      logger.debug(
        `Event filters: ${eventFilters.map((filter) => `${filter.nodeId}/${filter.eventMin}`).join(", ")}`
      );
    if (minIntervalFloorSeconds < 0) {
      throw new import_StatusCode.StatusResponseError(
        "minIntervalFloorSeconds should be greater or equal to 0",
        import_StatusCode.StatusCode.InvalidAction
      );
    }
    if (maxIntervalCeilingSeconds < 0) {
      throw new import_StatusCode.StatusResponseError(
        "maxIntervalCeilingSeconds should be greater or equal to 1",
        import_StatusCode.StatusCode.InvalidAction
      );
    }
    if (maxIntervalCeilingSeconds < minIntervalFloorSeconds) {
      throw new import_StatusCode.StatusResponseError(
        "maxIntervalCeilingSeconds should be greater or equal to minIntervalFloorSeconds",
        import_StatusCode.StatusCode.InvalidAction
      );
    }
    if (this.nextSubscriptionId === 4294967295)
      this.nextSubscriptionId = 0;
    const subscriptionId = this.nextSubscriptionId++;
    const subscriptionHandler = new import_SubscriptionHandler.SubscriptionHandler(
      subscriptionId,
      session,
      this.endpointStructure,
      attributeRequests,
      dataVersionFilters,
      eventRequests,
      eventFilters,
      this.eventHandler,
      isFabricFiltered,
      minIntervalFloorSeconds,
      maxIntervalCeilingSeconds,
      this.options?.subscriptionMaxIntervalSeconds,
      this.options?.subscriptionMinIntervalSeconds,
      this.options?.subscriptionRandomizationWindowSeconds,
      () => this.subscriptionMap.delete(subscriptionId)
    );
    try {
      await subscriptionHandler.sendInitialReport(messenger);
    } catch (error) {
      logger.error(
        `Subscription ${subscriptionId} for Session ${session.getId()}: Error while sending initial data reports`,
        error
      );
      await subscriptionHandler.cancel();
      if (error instanceof import_StatusCode.StatusResponseError) {
        logger.info(`Sending status response ${error.code} for interaction error: ${error.message}`);
        await messenger.sendStatus(error.code);
      }
      await messenger.close();
      return;
    }
    if (!keepSubscriptions) {
      logger.debug(`Clear subscriptions for Session ${session.name} because keepSubscriptions=false`);
      await session.clearSubscriptions(true);
    }
    const maxInterval = subscriptionHandler.getMaxInterval();
    logger.info(
      `Successfully created subscription ${subscriptionId} for Session ${session.getId()}. Updates: ${minIntervalFloorSeconds} - ${maxIntervalCeilingSeconds} => ${maxInterval} seconds (sendInterval = ${subscriptionHandler.getSendInterval()} seconds)`
    );
    await messenger.send(
      import_InteractionMessenger.MessageType.SubscribeResponse,
      import_InteractionProtocol.TlvSubscribeResponse.encode({
        subscriptionId,
        maxInterval,
        interactionModelRevision: INTERACTION_MODEL_REVISION
      })
    );
    this.subscriptionMap.set(subscriptionId, subscriptionHandler);
    session.addSubscription(subscriptionHandler);
    subscriptionHandler.activateSendingUpdates();
  }
  async handleInvokeRequest(exchange, { invokeRequests, timedRequest, suppressResponse, interactionModelRevision }, message) {
    logger.debug(
      `Received invoke request from ${exchange.channel.name}: ${invokeRequests.map(
        ({ commandPath: { endpointId, clusterId, commandId } }) => this.endpointStructure.resolveCommandName({ endpointId, clusterId, commandId })
      ).join(", ")}, suppressResponse=${suppressResponse}`
    );
    if (interactionModelRevision > INTERACTION_MODEL_REVISION) {
      logger.debug(
        `Interaction model revision of sender ${interactionModelRevision} is higher than supported ${INTERACTION_MODEL_REVISION}.`
      );
    }
    const {
      packetHeader: { sessionType }
    } = message;
    const receivedWithinTimedInteraction = exchange.hasActiveTimedInteraction();
    if (exchange.hasExpiredTimedInteraction()) {
      exchange.clearTimedInteraction();
      throw new import_StatusCode.StatusResponseError(`Timed request window expired. Decline invoke request.`, import_StatusCode.StatusCode.Timeout);
    }
    if (timedRequest !== exchange.hasTimedInteraction()) {
      throw new import_StatusCode.StatusResponseError(
        `timedRequest flag of invoke interaction (${timedRequest}) mismatch with expected timed interaction (${receivedWithinTimedInteraction}).`,
        import_StatusCode.StatusCode.TimedRequestMismatch
      );
    }
    if (receivedWithinTimedInteraction) {
      logger.debug(`Invoke request from ${exchange.channel.name} received while timed interaction is running.`);
      exchange.clearTimedInteraction();
      if (sessionType !== import_MessageCodec.SessionType.Unicast) {
        throw new import_StatusCode.StatusResponseError(
          "Invoke requests are only allowed on unicast sessions when a timed interaction is running.",
          import_StatusCode.StatusCode.InvalidAction
        );
      }
    }
    const invokeResponses = [];
    await Promise.all(
      invokeRequests.flatMap(async ({ commandPath, commandFields }) => {
        const commands = this.endpointStructure.getCommands([commandPath]);
        if (commands.length === 0) {
          const { endpointId, clusterId, commandId } = commandPath;
          if (endpointId === void 0 || clusterId === void 0 || commandId === void 0) {
            logger.debug(
              `Invoke from ${exchange.channel.name}: ${this.endpointStructure.resolveCommandName(
                commandPath
              )} ignore non-existing attribute`
            );
          } else {
            invokeResponses.push(
              (0, import_TryCatchHandler.tryCatch)(
                () => {
                  this.endpointStructure.validateConcreteCommandPath(
                    endpointId,
                    clusterId,
                    commandId
                  );
                  throw new import_MatterError.InternalError(
                    "validateConcreteCommandPath should throw StatusResponseError but did not."
                  );
                },
                import_StatusCode.StatusResponseError,
                (error) => {
                  logger.debug(
                    `Invoke from ${exchange.channel.name}: ${this.endpointStructure.resolveCommandName(
                      commandPath
                    )} unsupported path: Status=${error.code}`
                  );
                  return { status: { commandPath, status: { status: error.code } } };
                }
              )
            );
          }
          return;
        }
        for (const { command, path } of commands) {
          const { endpointId } = path;
          if (endpointId === void 0) {
            logger.error(
              `Invoke from ${exchange.channel.name}: ${this.endpointStructure.resolveCommandName(
                path
              )} invalid path because empty endpoint!`
            );
            invokeResponses.push({
              status: { commandPath: path, status: { status: import_StatusCode.StatusCode.UnsupportedEndpoint } }
            });
            continue;
          }
          const endpoint = this.endpointStructure.getEndpoint(endpointId);
          if (endpoint === void 0) {
            logger.error(
              `Invoke from ${exchange.channel.name}: ${this.endpointStructure.resolveCommandName(
                path
              )} invalid path because endpoint not found!`
            );
            invokeResponses.push({
              status: { commandPath: path, status: { status: import_StatusCode.StatusCode.UnsupportedEndpoint } }
            });
            continue;
          }
          if (command.requiresTimedInteraction && !receivedWithinTimedInteraction) {
            logger.debug(`This invoke requires a timed interaction which is not initialized.`);
            invokeResponses.push({
              status: { commandPath: path, status: { status: import_StatusCode.StatusCode.NeedsTimedInteraction } }
            });
            continue;
          }
          const result = await (0, import_TryCatchHandler.tryCatchAsync)(
            async () => await command.invoke(
              exchange.session,
              commandFields ?? import_TlvNoArguments.TlvNoArguments.encodeTlv(commandFields),
              message,
              endpoint
            ),
            import_StatusCode.StatusResponseError,
            async (error) => {
              const errorLogText = `Error ${(0, import_Number.toHexString)(error.code)}${error.clusterCode !== void 0 ? `/${(0, import_Number.toHexString)(error.clusterCode)}` : ""} while invoking command: ${error.message}`;
              if (error instanceof import_ValidationError.ValidationError) {
                logger.info(
                  `Validation-${errorLogText}${error.fieldName !== void 0 ? ` in field ${error.fieldName}` : ""}`
                );
              } else {
                logger.info(errorLogText);
              }
              return {
                code: error.code,
                clusterCode: error.clusterCode,
                responseId: command.responseId,
                response: import_Cluster.TlvNoResponse.encodeTlv()
              };
            }
          );
          const { code, clusterCode, responseId, response } = result;
          if (response.length === 0) {
            invokeResponses.push({
              status: { commandPath: path, status: { status: code, clusterStatus: clusterCode } }
            });
          } else {
            invokeResponses.push({
              command: {
                commandPath: { ...path, commandId: responseId },
                commandFields: response
              }
            });
          }
        }
      })
    );
    return {
      suppressResponse: false,
      interactionModelRevision: INTERACTION_MODEL_REVISION,
      invokeResponses
    };
  }
  handleTimedRequest(exchange, { timeout, interactionModelRevision }) {
    logger.debug(`Received timed request (${timeout}ms) from ${exchange.channel.name}`);
    if (interactionModelRevision > INTERACTION_MODEL_REVISION) {
      logger.debug(
        `Interaction model revision of sender ${interactionModelRevision} is higher than supported ${INTERACTION_MODEL_REVISION}.`
      );
    }
    exchange.startTimedInteraction(timeout);
  }
  async close() {
    this.isClosing = true;
    this.endpointStructure.destroy();
    for (const subscription of this.subscriptionMap.values()) {
      await subscription.cancel(true);
    }
  }
}
