// @project-chip/matter.js/dist/cjs/protocol/interaction/InteractionProtocol.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InteractionProtocol_exports = {};
__export(InteractionProtocol_exports, {
  TlvAttributeData: () => TlvAttributeData,
  TlvAttributePath: () => TlvAttributePath,
  TlvAttributeReport: () => TlvAttributeReport,
  TlvAttributeReportData: () => TlvAttributeReportData,
  TlvAttributeStatus: () => TlvAttributeStatus,
  TlvClusterPath: () => TlvClusterPath,
  TlvCommandData: () => TlvCommandData,
  TlvCommandPath: () => TlvCommandPath,
  TlvCommandStatus: () => TlvCommandStatus,
  TlvDataReport: () => TlvDataReport,
  TlvDataReportForSend: () => TlvDataReportForSend,
  TlvDataVersionFilter: () => TlvDataVersionFilter,
  TlvEventData: () => TlvEventData,
  TlvEventFilter: () => TlvEventFilter,
  TlvEventPath: () => TlvEventPath,
  TlvEventReport: () => TlvEventReport,
  TlvEventStatus: () => TlvEventStatus,
  TlvInvokeRequest: () => TlvInvokeRequest,
  TlvInvokeResponse: () => TlvInvokeResponse,
  TlvInvokeResponseData: () => TlvInvokeResponseData,
  TlvReadRequest: () => TlvReadRequest,
  TlvStatus: () => TlvStatus,
  TlvStatusResponse: () => TlvStatusResponse,
  TlvSubscribeRequest: () => TlvSubscribeRequest,
  TlvSubscribeResponse: () => TlvSubscribeResponse,
  TlvTimedRequest: () => TlvTimedRequest,
  TlvWriteRequest: () => TlvWriteRequest,
  TlvWriteResponse: () => TlvWriteResponse
});
module.exports = __toCommonJS(InteractionProtocol_exports);
var import_AttributeId = require("../../datatype/AttributeId.js");
var import_ClusterId = require("../../datatype/ClusterId.js");
var import_CommandId = require("../../datatype/CommandId.js");
var import_EndpointNumber = require("../../datatype/EndpointNumber.js");
var import_EventId = require("../../datatype/EventId.js");
var import_NodeId = require("../../datatype/NodeId.js");
var import_TlvAny = require("../../tlv/TlvAny.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TlvAttributePath = (0, import_TlvObject.TlvTaggedList)({
  // AttributePathIB
  enableTagCompression: (0, import_TlvObject.TlvOptionalField)(0, import_TlvBoolean.TlvBoolean),
  nodeId: (0, import_TlvObject.TlvOptionalField)(1, import_NodeId.TlvNodeId),
  endpointId: (0, import_TlvObject.TlvOptionalField)(2, import_EndpointNumber.TlvEndpointNumber),
  clusterId: (0, import_TlvObject.TlvOptionalField)(3, import_ClusterId.TlvClusterId),
  attributeId: (0, import_TlvObject.TlvOptionalField)(4, import_AttributeId.TlvAttributeId),
  listIndex: (0, import_TlvObject.TlvOptionalField)(5, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16))
});
const TlvEventPath = (0, import_TlvObject.TlvTaggedList)({
  // EventPathIB
  nodeId: (0, import_TlvObject.TlvOptionalField)(0, import_NodeId.TlvNodeId),
  endpointId: (0, import_TlvObject.TlvOptionalField)(1, import_EndpointNumber.TlvEndpointNumber),
  clusterId: (0, import_TlvObject.TlvOptionalField)(2, import_ClusterId.TlvClusterId),
  eventId: (0, import_TlvObject.TlvOptionalField)(3, import_EventId.TlvEventId),
  isUrgent: (0, import_TlvObject.TlvOptionalField)(4, import_TlvBoolean.TlvBoolean)
});
const TlvEventData = (0, import_TlvObject.TlvObject)({
  // EventDataIB
  path: (0, import_TlvObject.TlvField)(0, TlvEventPath),
  eventNumber: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt64),
  priority: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvEnum)()),
  epochTimestamp: (0, import_TlvObject.TlvOptionalField)(3, import_TlvNumber.TlvPosixMs),
  systemTimestamp: (0, import_TlvObject.TlvOptionalField)(4, import_TlvNumber.TlvSysTimeMS),
  deltaEpochTimestamp: (0, import_TlvObject.TlvOptionalField)(5, import_TlvNumber.TlvPosixMs),
  deltaSystemTimestamp: (0, import_TlvObject.TlvOptionalField)(6, import_TlvNumber.TlvSysTimeMS),
  data: (0, import_TlvObject.TlvOptionalField)(7, import_TlvAny.TlvAny)
});
const TlvEventFilter = (0, import_TlvObject.TlvObject)({
  // EventFilterIB
  nodeId: (0, import_TlvObject.TlvOptionalField)(0, import_NodeId.TlvNodeId),
  eventMin: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt64)
});
const TlvClusterPath = (0, import_TlvObject.TlvTaggedList)({
  // ClusterPathIB
  nodeId: (0, import_TlvObject.TlvOptionalField)(0, import_NodeId.TlvNodeId),
  endpointId: (0, import_TlvObject.TlvField)(1, import_EndpointNumber.TlvEndpointNumber),
  clusterId: (0, import_TlvObject.TlvField)(2, import_ClusterId.TlvClusterId)
});
const TlvDataVersionFilter = (0, import_TlvObject.TlvObject)({
  // DataVersionFilterIB
  path: (0, import_TlvObject.TlvField)(0, TlvClusterPath),
  dataVersion: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt32)
});
const TlvStatus = (0, import_TlvObject.TlvObject)({
  // StatusIB
  status: (0, import_TlvObject.TlvOptionalField)(0, (0, import_TlvNumber.TlvEnum)()),
  clusterStatus: (0, import_TlvObject.TlvOptionalField)(1, (0, import_TlvNumber.TlvEnum)())
});
const TlvAttributeStatus = (0, import_TlvObject.TlvObject)({
  // AttributeStatusIB
  path: (0, import_TlvObject.TlvField)(0, TlvAttributePath),
  status: (0, import_TlvObject.TlvField)(1, TlvStatus)
});
const TlvAttributeData = (0, import_TlvObject.TlvObject)({
  // AttributeDataIB version for Non-Reports
  dataVersion: (0, import_TlvObject.TlvOptionalField)(0, import_TlvNumber.TlvUInt32),
  path: (0, import_TlvObject.TlvField)(1, TlvAttributePath),
  data: (0, import_TlvObject.TlvField)(2, import_TlvAny.TlvAny)
});
const TlvAttributeReportData = (0, import_TlvObject.TlvObject)({
  // AttributeDataIB version for Reports
  dataVersion: (0, import_TlvObject.TlvOptionalField)(0, import_TlvNumber.TlvUInt32),
  path: (0, import_TlvObject.TlvField)(1, TlvAttributePath),
  data: (0, import_TlvObject.TlvField)(2, import_TlvAny.TlvAny)
});
const TlvAttributeReport = (0, import_TlvObject.TlvObject)({
  // AttributeReportIB
  attributeStatus: (0, import_TlvObject.TlvOptionalField)(0, TlvAttributeStatus),
  attributeData: (0, import_TlvObject.TlvOptionalField)(1, TlvAttributeReportData)
});
const TlvEventStatus = (0, import_TlvObject.TlvObject)({
  // EventStatusIB
  path: (0, import_TlvObject.TlvField)(0, TlvEventPath),
  status: (0, import_TlvObject.TlvField)(1, TlvStatus)
});
const TlvEventReport = (0, import_TlvObject.TlvObject)({
  // EventReportIB
  eventStatus: (0, import_TlvObject.TlvOptionalField)(0, TlvEventStatus),
  eventData: (0, import_TlvObject.TlvOptionalField)(1, TlvEventData)
});
const TlvCommandPath = (0, import_TlvObject.TlvTaggedList)({
  // CommandPathIB
  endpointId: (0, import_TlvObject.TlvOptionalField)(0, import_EndpointNumber.TlvEndpointNumber),
  clusterId: (0, import_TlvObject.TlvField)(1, import_ClusterId.TlvClusterId),
  commandId: (0, import_TlvObject.TlvField)(2, import_CommandId.TlvCommandId)
});
const TlvCommandData = (0, import_TlvObject.TlvObject)({
  // CommandDataIB
  commandPath: (0, import_TlvObject.TlvField)(0, TlvCommandPath),
  commandFields: (0, import_TlvObject.TlvOptionalField)(1, import_TlvAny.TlvAny)
});
const TlvCommandStatus = (0, import_TlvObject.TlvObject)({
  // CommandStatusIB
  commandPath: (0, import_TlvObject.TlvField)(0, TlvCommandPath),
  status: (0, import_TlvObject.TlvField)(1, TlvStatus)
});
const TlvInvokeResponseData = (0, import_TlvObject.TlvObject)({
  // InvokeResponseIB
  command: (0, import_TlvObject.TlvOptionalField)(0, TlvCommandData),
  status: (0, import_TlvObject.TlvOptionalField)(1, TlvCommandStatus)
});
const TlvStatusResponse = (0, import_TlvObject.TlvObject)({
  status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
const TlvReadRequest = (0, import_TlvObject.TlvObject)({
  attributeRequests: (0, import_TlvObject.TlvOptionalField)(0, (0, import_TlvArray.TlvArray)(TlvAttributePath)),
  eventRequests: (0, import_TlvObject.TlvOptionalField)(1, (0, import_TlvArray.TlvArray)(TlvEventPath)),
  eventFilters: (0, import_TlvObject.TlvOptionalField)(2, (0, import_TlvArray.TlvArray)(TlvEventFilter)),
  isFabricFiltered: (0, import_TlvObject.TlvField)(3, import_TlvBoolean.TlvBoolean),
  dataVersionFilters: (0, import_TlvObject.TlvOptionalField)(4, (0, import_TlvArray.TlvArray)(TlvDataVersionFilter)),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
const TlvDataReport = (0, import_TlvObject.TlvObject)({
  subscriptionId: (0, import_TlvObject.TlvOptionalField)(0, import_TlvNumber.TlvUInt32),
  attributeReports: (0, import_TlvObject.TlvOptionalField)(1, (0, import_TlvArray.TlvArray)(TlvAttributeReport)),
  eventReports: (0, import_TlvObject.TlvOptionalField)(2, (0, import_TlvArray.TlvArray)(TlvEventReport)),
  moreChunkedMessages: (0, import_TlvObject.TlvOptionalField)(3, import_TlvBoolean.TlvBoolean),
  suppressResponse: (0, import_TlvObject.TlvOptionalField)(4, import_TlvBoolean.TlvBoolean),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
const TlvDataReportForSend = (0, import_TlvObject.TlvObject)({
  subscriptionId: (0, import_TlvObject.TlvOptionalField)(0, import_TlvNumber.TlvUInt32),
  attributeReports: (0, import_TlvObject.TlvOptionalField)(1, (0, import_TlvArray.TlvArray)(import_TlvAny.TlvAny)),
  eventReports: (0, import_TlvObject.TlvOptionalField)(2, (0, import_TlvArray.TlvArray)(import_TlvAny.TlvAny)),
  moreChunkedMessages: (0, import_TlvObject.TlvOptionalField)(3, import_TlvBoolean.TlvBoolean),
  suppressResponse: (0, import_TlvObject.TlvOptionalField)(4, import_TlvBoolean.TlvBoolean),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
const TlvSubscribeRequest = (0, import_TlvObject.TlvObject)({
  keepSubscriptions: (0, import_TlvObject.TlvField)(0, import_TlvBoolean.TlvBoolean),
  minIntervalFloorSeconds: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16),
  maxIntervalCeilingSeconds: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt16),
  attributeRequests: (0, import_TlvObject.TlvOptionalField)(3, (0, import_TlvArray.TlvArray)(TlvAttributePath)),
  eventRequests: (0, import_TlvObject.TlvOptionalField)(4, (0, import_TlvArray.TlvArray)(TlvEventPath)),
  eventFilters: (0, import_TlvObject.TlvOptionalField)(5, (0, import_TlvArray.TlvArray)(TlvEventFilter)),
  isFabricFiltered: (0, import_TlvObject.TlvField)(7, import_TlvBoolean.TlvBoolean),
  dataVersionFilters: (0, import_TlvObject.TlvOptionalField)(8, (0, import_TlvArray.TlvArray)(TlvDataVersionFilter)),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
const TlvSubscribeResponse = (0, import_TlvObject.TlvObject)({
  subscriptionId: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt32),
  maxInterval: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt16),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
const TlvInvokeRequest = (0, import_TlvObject.TlvObject)({
  suppressResponse: (0, import_TlvObject.TlvField)(0, import_TlvBoolean.TlvBoolean),
  timedRequest: (0, import_TlvObject.TlvField)(1, import_TlvBoolean.TlvBoolean),
  invokeRequests: (0, import_TlvObject.TlvField)(2, (0, import_TlvArray.TlvArray)(TlvCommandData)),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
const TlvInvokeResponse = (0, import_TlvObject.TlvObject)({
  suppressResponse: (0, import_TlvObject.TlvField)(0, import_TlvBoolean.TlvBoolean),
  invokeResponses: (0, import_TlvObject.TlvField)(1, (0, import_TlvArray.TlvArray)(TlvInvokeResponseData)),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
const TlvTimedRequest = (0, import_TlvObject.TlvObject)({
  timeout: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
const TlvWriteRequest = (0, import_TlvObject.TlvObject)({
  suppressResponse: (0, import_TlvObject.TlvOptionalField)(0, import_TlvBoolean.TlvBoolean),
  timedRequest: (0, import_TlvObject.TlvField)(1, import_TlvBoolean.TlvBoolean),
  writeRequests: (0, import_TlvObject.TlvField)(2, (0, import_TlvArray.TlvArray)(TlvAttributeData)),
  moreChunkedMessages: (0, import_TlvObject.TlvOptionalField)(3, import_TlvBoolean.TlvBoolean),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
const TlvWriteResponse = (0, import_TlvObject.TlvObject)({
  writeResponses: (0, import_TlvObject.TlvField)(0, (0, import_TlvArray.TlvArray)(TlvAttributeStatus)),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
