// @project-chip/matter.js/dist/cjs/protocol/interaction/InteractionEndpointStructure.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InteractionEndpointStructure_exports = {};
__export(InteractionEndpointStructure_exports, {
  InteractionEndpointStructure: () => InteractionEndpointStructure
});
module.exports = __toCommonJS(InteractionEndpointStructure_exports);
var import_ClusterServerTypes = require("../../cluster/server/ClusterServerTypes.js");
var import_MatterError = require("../../common/MatterError.js");
var import_InteractionServer = require("./InteractionServer.js");
var import_StatusCode = require("./StatusCode.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class InteractionEndpointStructure {
  constructor() {
    this.endpoints = /* @__PURE__ */ new Map();
    this.attributes = /* @__PURE__ */ new Map();
    this.attributePaths = new Array();
    this.events = /* @__PURE__ */ new Map();
    this.eventPaths = new Array();
    this.commands = /* @__PURE__ */ new Map();
    this.commandPaths = new Array();
  }
  clear() {
    this.endpoints.clear();
    this.attributes.clear();
    this.attributePaths.length = 0;
    this.events.clear();
    this.eventPaths.length = 0;
    this.commands.clear();
    this.commandPaths.length = 0;
  }
  destroy() {
    for (const endpoint of this.endpoints.values()) {
      endpoint.destroy();
    }
  }
  initializeFromEndpoint(endpoint) {
    this.clear();
    this.verifyAndInitializeStructureElementsFromEndpoint(endpoint);
    this.initializeStructureFromEndpoints(endpoint);
  }
  initializeStructureFromEndpoints(endpoint) {
    const endpoints = endpoint.getChildEndpoints();
    for (let endpointIndex = 0; endpointIndex < endpoints.length; endpointIndex++) {
      this.verifyAndInitializeStructureElementsFromEndpoint(endpoints[endpointIndex]);
      this.initializeStructureFromEndpoints(endpoints[endpointIndex]);
    }
  }
  verifyAndInitializeStructureElementsFromEndpoint(endpoint) {
    if (endpoint.id === void 0) {
      throw new import_MatterError.InternalError(`Endpoint ID is undefined. It needs to be initialized first!`);
    }
    endpoint.verifyRequiredClusters();
    for (const cluster of endpoint.getAllClusterServers()) {
      const {
        id: clusterId,
        attributes: clusterAttributes,
        _events: clusterEvents,
        _commands: clusterCommands
      } = (0, import_ClusterServerTypes.asClusterServerInternal)(cluster);
      for (const name in clusterAttributes) {
        const attribute = clusterAttributes[name];
        const path = { endpointId: endpoint.id, clusterId, attributeId: attribute.id };
        this.attributes.set((0, import_InteractionServer.attributePathToId)(path), attribute);
        this.attributePaths.push(path);
      }
      for (const name in clusterEvents) {
        const event = clusterEvents[name];
        const path = { endpointId: endpoint.id, clusterId, eventId: event.id };
        this.events.set((0, import_InteractionServer.eventPathToId)(path), event);
        this.eventPaths.push(path);
      }
      for (const name in clusterCommands) {
        const command = clusterCommands[name];
        const path = { endpointId: endpoint.id, clusterId, commandId: command.invokeId };
        this.commands.set((0, import_InteractionServer.commandPathToId)(path), command);
        this.commandPaths.push(path);
      }
    }
    if (this.endpoints.has(endpoint.id))
      throw new import_MatterError.ImplementationError(`Endpoint ID ${endpoint.id} exists twice`);
    this.endpoints.set(endpoint.id, endpoint);
  }
  toHex(value) {
    return value === void 0 ? "*" : `0x${value.toString(16)}`;
  }
  resolveGenericElementName(nodeId, endpointId, clusterId, elementId, elementMap) {
    const nodeIdPrefix = nodeId === void 0 ? "" : `${this.toHex(nodeId)}/`;
    if (endpointId === void 0) {
      return `${nodeIdPrefix}*/${this.toHex(clusterId)}/${this.toHex(elementId)}`;
    }
    const endpoint = this.endpoints.get(endpointId);
    if (endpoint === void 0) {
      return `${nodeIdPrefix}unknown(${this.toHex(endpointId)})/${this.toHex(clusterId)}/${this.toHex(
        elementId
      )}`;
    }
    const endpointName = `${endpoint.name}(${this.toHex(endpointId)})`;
    if (clusterId === void 0) {
      return `${nodeIdPrefix}${endpointName}/*/${this.toHex(elementId)}`;
    }
    const cluster = endpoint.getClusterServerById(clusterId);
    if (cluster === void 0) {
      return `${nodeIdPrefix}${endpointName}/unknown(${this.toHex(clusterId)})/${this.toHex(elementId)}`;
    }
    const clusterName = `${cluster.name}(${this.toHex(clusterId)})`;
    if (elementId === void 0) {
      return `${nodeIdPrefix}${endpointName}/${clusterName}/*`;
    }
    const element = elementMap.get((0, import_InteractionServer.genericElementPathToId)(endpointId, clusterId, elementId));
    const elementName = `${element?.name ?? "unknown"}(${this.toHex(elementId)})`;
    return `${nodeIdPrefix}${endpointName}/${clusterName}/${elementName}`;
  }
  resolveAttributeName({ nodeId, endpointId, clusterId, attributeId }) {
    return this.resolveGenericElementName(nodeId, endpointId, clusterId, attributeId, this.attributes);
  }
  resolveEventName({ nodeId, endpointId, clusterId, eventId, isUrgent }) {
    return `${isUrgent ? "!" : ""}${this.resolveGenericElementName(
      nodeId,
      endpointId,
      clusterId,
      eventId,
      this.events
    )}`;
  }
  resolveCommandName({ endpointId, clusterId, commandId }) {
    return this.resolveGenericElementName(void 0, endpointId, clusterId, commandId, this.commands);
  }
  getEndpoint(endpointId) {
    return this.endpoints.get(endpointId);
  }
  hasEndpoint(endpointId) {
    return this.endpoints.has(endpointId);
  }
  getClusterServer(endpointId, clusterId) {
    return this.endpoints.get(endpointId)?.getClusterServerById(clusterId);
  }
  hasClusterServer(endpointId, clusterId) {
    return !!this.getClusterServer(endpointId, clusterId);
  }
  getAttribute(endpointId, clusterId, attributeId) {
    return this.attributes.get((0, import_InteractionServer.attributePathToId)({ endpointId, clusterId, attributeId }));
  }
  hasAttribute(endpointId, clusterId, attributeId) {
    return !!this.getAttribute(endpointId, clusterId, attributeId);
  }
  validateConcreteAttributePath(endpointId, clusterId, attributeId) {
    if (!this.hasEndpoint(endpointId)) {
      throw new import_StatusCode.StatusResponseError(`Endpoint ${endpointId} does not exist.`, import_StatusCode.StatusCode.UnsupportedEndpoint);
    }
    if (!this.hasClusterServer(endpointId, clusterId)) {
      throw new import_StatusCode.StatusResponseError(`Cluster ${clusterId} does not exist.`, import_StatusCode.StatusCode.UnsupportedCluster);
    }
    if (this.hasAttribute(endpointId, clusterId, attributeId))
      return true;
    throw new import_StatusCode.StatusResponseError(`Attribute ${attributeId} does not exist`, import_StatusCode.StatusCode.UnsupportedAttribute);
  }
  getEvent(endpointId, clusterId, eventId) {
    return this.events.get((0, import_InteractionServer.eventPathToId)({ endpointId, clusterId, eventId }));
  }
  hasEvent(endpointId, clusterId, eventId) {
    return !!this.getEvent(endpointId, clusterId, eventId);
  }
  validateConcreteEventPath(endpointId, clusterId, eventId) {
    if (!this.hasEndpoint(endpointId)) {
      throw new import_StatusCode.StatusResponseError(`Endpoint ${endpointId} does not exist.`, import_StatusCode.StatusCode.UnsupportedEndpoint);
    }
    if (!this.hasClusterServer(endpointId, clusterId)) {
      throw new import_StatusCode.StatusResponseError(`Cluster ${clusterId} does not exist.`, import_StatusCode.StatusCode.UnsupportedCluster);
    }
    if (this.hasEvent(endpointId, clusterId, eventId))
      return true;
    throw new import_StatusCode.StatusResponseError(`Event ${eventId} does not exist`, import_StatusCode.StatusCode.UnsupportedEvent);
  }
  getCommand(endpointId, clusterId, commandId) {
    return this.commands.get((0, import_InteractionServer.commandPathToId)({ endpointId, clusterId, commandId }));
  }
  hasCommand(endpointId, clusterId, commandId) {
    return !!this.getCommand(endpointId, clusterId, commandId);
  }
  validateConcreteCommandPath(endpointId, clusterId, commandId) {
    if (!this.hasEndpoint(endpointId)) {
      throw new import_StatusCode.StatusResponseError(`Endpoint ${endpointId} does not exist.`, import_StatusCode.StatusCode.UnsupportedEndpoint);
    }
    if (!this.hasClusterServer(endpointId, clusterId)) {
      throw new import_StatusCode.StatusResponseError(`Cluster ${clusterId} does not exist.`, import_StatusCode.StatusCode.UnsupportedCluster);
    }
    if (this.hasCommand(endpointId, clusterId, commandId))
      return true;
    throw new import_StatusCode.StatusResponseError(`Command ${commandId} does not exist`, import_StatusCode.StatusCode.UnsupportedCommand);
  }
  getAttributes(filters, onlyWritable = false) {
    const result = new Array();
    filters.forEach(({ endpointId, clusterId, attributeId }) => {
      if (endpointId !== void 0 && clusterId !== void 0 && attributeId !== void 0) {
        const path = { endpointId, clusterId, attributeId };
        const attribute = this.attributes.get((0, import_InteractionServer.attributePathToId)(path));
        if (attribute === void 0)
          return;
        if (onlyWritable && !attribute.isWritable)
          return;
        result.push({ path, attribute });
      } else {
        this.attributePaths.filter(
          (path) => (endpointId === void 0 || endpointId === path.endpointId) && (clusterId === void 0 || clusterId === path.clusterId) && (attributeId === void 0 || attributeId === path.attributeId)
        ).forEach((path) => {
          const attribute = this.attributes.get((0, import_InteractionServer.attributePathToId)(path));
          if (attribute === void 0)
            return;
          if (onlyWritable && !attribute.isWritable)
            return;
          result.push({ path, attribute });
        });
      }
    });
    return result;
  }
  getEvents(filters) {
    const result = new Array();
    filters.forEach(({ endpointId, clusterId, eventId, isUrgent }) => {
      if (endpointId !== void 0 && clusterId !== void 0 && eventId !== void 0) {
        const path = { endpointId, clusterId, eventId, isUrgent };
        const event = this.events.get((0, import_InteractionServer.eventPathToId)(path));
        if (event === void 0)
          return;
        result.push({ path, event });
      } else {
        this.eventPaths.filter(
          (path) => (endpointId === void 0 || endpointId === path.endpointId) && (clusterId === void 0 || clusterId === path.clusterId) && (eventId === void 0 || eventId === path.eventId)
        ).forEach(({ endpointId: endpointId2, clusterId: clusterId2, eventId: eventId2 }) => {
          const path = { endpointId: endpointId2, clusterId: clusterId2, eventId: eventId2, isUrgent };
          const event = this.events.get((0, import_InteractionServer.eventPathToId)(path));
          if (event === void 0)
            return;
          result.push({ path, event });
        });
      }
    });
    return result;
  }
  getCommands(filters) {
    const result = new Array();
    filters.forEach(({ endpointId, clusterId, commandId }) => {
      if (endpointId !== void 0 && clusterId !== void 0 && commandId !== void 0) {
        const path = { endpointId, clusterId, commandId };
        const command = this.commands.get((0, import_InteractionServer.commandPathToId)(path));
        if (command === void 0)
          return;
        result.push({ path, command });
      } else {
        this.commandPaths.filter(
          (path) => (endpointId === void 0 || endpointId === path.endpointId) && (clusterId === void 0 || clusterId === path.clusterId) && (commandId === void 0 || commandId === path.commandId)
        ).forEach((path) => {
          const command = this.commands.get((0, import_InteractionServer.commandPathToId)(path));
          if (command === void 0)
            return;
          result.push({ path, command });
        });
      }
    });
    return result;
  }
}
