// @project-chip/matter.js/dist/cjs/cluster/definitions/AudioOutputCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AudioOutputCluster_exports = {};
__export(AudioOutputCluster_exports, {
  AudioOutput: () => AudioOutput,
  AudioOutputCluster: () => AudioOutputCluster
});
module.exports = __toCommonJS(AudioOutputCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvString = require("../../tlv/TlvString.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var AudioOutput;
((AudioOutput2) => {
  let OutputType;
  ((OutputType2) => {
    OutputType2[OutputType2["Hdmi"] = 0] = "Hdmi";
    OutputType2[OutputType2["Bt"] = 1] = "Bt";
    OutputType2[OutputType2["Optical"] = 2] = "Optical";
    OutputType2[OutputType2["Headphone"] = 3] = "Headphone";
    OutputType2[OutputType2["Internal"] = 4] = "Internal";
    OutputType2[OutputType2["Other"] = 5] = "Other";
  })(OutputType = AudioOutput2.OutputType || (AudioOutput2.OutputType = {}));
  AudioOutput2.TlvOutputInfoStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the unique index into the list of outputs.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.5.1.1
     */
    index: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    /**
     * This shall indicate the type of output
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.5.1.2
     */
    outputType: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The device defined and user editable output name, such as “Soundbar”, “Speakers”. This field may be blank,
     * but SHOULD be provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.5.1.3
     */
    name: (0, import_TlvObject.TlvField)(2, import_TlvString.TlvString.bound({ maxLength: 32 }))
  });
  AudioOutput2.TlvSelectOutputRequest = (0, import_TlvObject.TlvObject)({ index: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8) });
  AudioOutput2.TlvRenameOutputRequest = (0, import_TlvObject.TlvObject)({ index: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8), name: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvString) });
  let Feature;
  ((Feature2) => {
    Feature2["NameUpdates"] = "NameUpdates";
  })(Feature = AudioOutput2.Feature || (AudioOutput2.Feature = {}));
  AudioOutput2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 1291,
    name: "AudioOutput",
    revision: 1,
    features: {
      /**
       * NameUpdates
       *
       * Supports updates to output names
       */
      nameUpdates: (0, import_BitmapSchema.BitFlag)(0)
    },
    attributes: {
      /**
       * This list provides the outputs supported by the device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.3.1
       */
      outputList: (0, import_Cluster.Attribute)(0, (0, import_TlvArray.TlvArray)(AudioOutput2.TlvOutputInfoStruct)),
      /**
       * This field contains the value of the index field of the currently selected OutputInfoStruct.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.3.2
       */
      currentOutput: (0, import_Cluster.Attribute)(1, import_TlvNumber.TlvUInt8, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.4
       */
      selectOutput: (0, import_Cluster.Command)(0, AudioOutput2.TlvSelectOutputRequest, 0, import_Cluster.TlvNoResponse)
    }
  });
  AudioOutput2.NameUpdatesComponent = import_ClusterFactory.ClusterFactory.Component({
    commands: {
      /**
       * Upon receipt, this shall rename the output at a specific index in the Output List.
       *
       * Updates to the output name shall appear in the device’s settings menus. Name updates may automatically
       * be sent to the actual device to which the output connects.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.4.2
       */
      renameOutput: (0, import_Cluster.Command)(1, AudioOutput2.TlvRenameOutputRequest, 1, import_Cluster.TlvNoResponse, { invokeAcl: import_Cluster.AccessLevel.Manage })
    }
  });
  AudioOutput2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    AudioOutput2.Base,
    /**
     * Use this factory method to create an AudioOutput cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns an AudioOutput cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...AudioOutput2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(AudioOutput2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, AudioOutput2.NameUpdatesComponent, { nameUpdates: true });
      return cluster;
    }
  );
  const NU = { nameUpdates: true };
  AudioOutput2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: AudioOutput2.Cluster.id,
    name: AudioOutput2.Cluster.name,
    revision: AudioOutput2.Cluster.revision,
    features: AudioOutput2.Cluster.features,
    attributes: AudioOutput2.Cluster.attributes,
    commands: {
      ...AudioOutput2.Cluster.commands,
      renameOutput: import_ClusterFactory.ClusterFactory.AsConditional(
        AudioOutput2.NameUpdatesComponent.commands.renameOutput,
        { mandatoryIf: [NU] }
      )
    }
  });
})(AudioOutput || (AudioOutput = {}));
const AudioOutputCluster = AudioOutput.Cluster;
