// @project-chip/matter.js/dist/cjs/cluster/definitions/BarrierControlCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BarrierControlCluster_exports = {};
__export(BarrierControlCluster_exports, {
  BarrierControl: () => BarrierControl,
  BarrierControlCluster: () => BarrierControlCluster
});
module.exports = __toCommonJS(BarrierControlCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var BarrierControl;
((BarrierControl2) => {
  BarrierControl2.TlvBarrierControlGoToPercentRequest = (0, import_TlvObject.TlvObject)({ percentOpen: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8) });
  BarrierControl2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 259,
    name: "BarrierControl",
    revision: 1,
    attributes: {
      barrierMovingState: (0, import_Cluster.Attribute)(1, import_TlvNumber.TlvUInt8),
      barrierSafetyStatus: (0, import_Cluster.Attribute)(2, import_TlvNumber.TlvUInt16),
      barrierCapabilities: (0, import_Cluster.Attribute)(3, import_TlvNumber.TlvUInt8),
      barrierOpenEvents: (0, import_Cluster.OptionalWritableAttribute)(4, import_TlvNumber.TlvUInt16.bound({ max: 65534 }), { default: 0 }),
      barrierCloseEvents: (0, import_Cluster.OptionalWritableAttribute)(5, import_TlvNumber.TlvUInt16.bound({ max: 65534 }), { default: 0 }),
      barrierCommandOpenEvents: (0, import_Cluster.OptionalWritableAttribute)(6, import_TlvNumber.TlvUInt16.bound({ max: 65534 }), { default: 0 }),
      barrierCommandCloseEvents: (0, import_Cluster.OptionalWritableAttribute)(7, import_TlvNumber.TlvUInt16.bound({ max: 65534 }), { default: 0 }),
      barrierOpenPeriod: (0, import_Cluster.OptionalWritableAttribute)(8, import_TlvNumber.TlvUInt16.bound({ max: 65534 })),
      barrierClosePeriod: (0, import_Cluster.OptionalWritableAttribute)(9, import_TlvNumber.TlvUInt16.bound({ max: 65534 })),
      barrierPosition: (0, import_Cluster.Attribute)(10, import_TlvNumber.TlvUInt8)
    },
    commands: {
      barrierControlGoToPercent: (0, import_Cluster.Command)(0, BarrierControl2.TlvBarrierControlGoToPercentRequest, 0, import_Cluster.TlvNoResponse),
      barrierControlStop: (0, import_Cluster.Command)(1, import_TlvNoArguments.TlvNoArguments, 1, import_Cluster.TlvNoResponse)
    }
  });
})(BarrierControl || (BarrierControl = {}));
const BarrierControlCluster = BarrierControl.Cluster;
