// @project-chip/matter.js/dist/cjs/cluster/definitions/BinaryInputBasicCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BinaryInputBasicCluster_exports = {};
__export(BinaryInputBasicCluster_exports, {
  BinaryInputBasic: () => BinaryInputBasic,
  BinaryInputBasicCluster: () => BinaryInputBasicCluster
});
module.exports = __toCommonJS(BinaryInputBasicCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var BinaryInputBasic;
((BinaryInputBasic2) => {
  BinaryInputBasic2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 15,
    name: "BinaryInputBasic",
    revision: 1,
    attributes: {
      activeText: (0, import_Cluster.OptionalWritableAttribute)(4, import_TlvString.TlvString.bound({ maxLength: 16 }), { default: "" }),
      description: (0, import_Cluster.OptionalWritableAttribute)(28, import_TlvString.TlvString.bound({ maxLength: 16 }), { default: "" }),
      inactiveText: (0, import_Cluster.OptionalWritableAttribute)(46, import_TlvString.TlvString.bound({ maxLength: 16 }), { default: "" }),
      outOfService: (0, import_Cluster.WritableAttribute)(81, import_TlvBoolean.TlvBoolean, { default: true }),
      polarity: (0, import_Cluster.OptionalAttribute)(84, import_TlvNumber.TlvUInt8),
      presentValue: (0, import_Cluster.WritableAttribute)(85, import_TlvBoolean.TlvBoolean),
      reliability: (0, import_Cluster.OptionalWritableAttribute)(103, import_TlvNumber.TlvUInt8),
      statusFlags: (0, import_Cluster.Attribute)(111, import_TlvNumber.TlvUInt8),
      applicationType: (0, import_Cluster.OptionalAttribute)(256, import_TlvNumber.TlvUInt32)
    }
  });
})(BinaryInputBasic || (BinaryInputBasic = {}));
const BinaryInputBasicCluster = BinaryInputBasic.Cluster;
