// @project-chip/matter.js/dist/cjs/cluster/definitions/BooleanStateCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BooleanStateCluster_exports = {};
__export(BooleanStateCluster_exports, {
  BooleanState: () => BooleanState,
  BooleanStateCluster: () => BooleanStateCluster
});
module.exports = __toCommonJS(BooleanStateCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var BooleanState;
((BooleanState2) => {
  BooleanState2.TlvStateChangeEvent = (0, import_TlvObject.TlvObject)({ stateValue: (0, import_TlvObject.TlvField)(0, import_TlvBoolean.TlvBoolean) });
  BooleanState2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 69,
    name: "BooleanState",
    revision: 1,
    attributes: {
      /**
       * This represents a Boolean state.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.7.4.1
       */
      stateValue: (0, import_Cluster.Attribute)(0, import_TlvBoolean.TlvBoolean)
    },
    events: {
      /**
       * This event shall be generated when the StateValue attribute changes.
       *
       * The StateValue field shall indicate the new value of the StateValue attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.7.5.1
       */
      stateChange: (0, import_Cluster.OptionalEvent)(0, import_Cluster.EventPriority.Info, BooleanState2.TlvStateChangeEvent)
    }
  });
})(BooleanState || (BooleanState = {}));
const BooleanStateCluster = BooleanState.Cluster;
