// @project-chip/matter.js/dist/cjs/cluster/definitions/BridgedDeviceBasicInformationCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BridgedDeviceBasicInformationCluster_exports = {};
__export(BridgedDeviceBasicInformationCluster_exports, {
  BridgedDeviceBasicInformation: () => BridgedDeviceBasicInformation,
  BridgedDeviceBasicInformationCluster: () => BridgedDeviceBasicInformationCluster
});
module.exports = __toCommonJS(BridgedDeviceBasicInformationCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_VendorId = require("../../datatype/VendorId.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var BridgedDeviceBasicInformation;
((BridgedDeviceBasicInformation2) => {
  let ProductFinish;
  ((ProductFinish2) => {
    ProductFinish2[ProductFinish2["Other"] = 0] = "Other";
    ProductFinish2[ProductFinish2["Matte"] = 1] = "Matte";
    ProductFinish2[ProductFinish2["Satin"] = 2] = "Satin";
    ProductFinish2[ProductFinish2["Polished"] = 3] = "Polished";
    ProductFinish2[ProductFinish2["Rugged"] = 4] = "Rugged";
    ProductFinish2[ProductFinish2["Fabric"] = 5] = "Fabric";
  })(ProductFinish = BridgedDeviceBasicInformation2.ProductFinish || (BridgedDeviceBasicInformation2.ProductFinish = {}));
  let Color;
  ((Color2) => {
    Color2[Color2["Black"] = 0] = "Black";
    Color2[Color2["Navy"] = 1] = "Navy";
    Color2[Color2["Green"] = 2] = "Green";
    Color2[Color2["Teal"] = 3] = "Teal";
    Color2[Color2["Maroon"] = 4] = "Maroon";
    Color2[Color2["Purple"] = 5] = "Purple";
    Color2[Color2["Olive"] = 6] = "Olive";
    Color2[Color2["Gray"] = 7] = "Gray";
    Color2[Color2["Blue"] = 8] = "Blue";
    Color2[Color2["Lime"] = 9] = "Lime";
    Color2[Color2["Aqua"] = 10] = "Aqua";
    Color2[Color2["Red"] = 11] = "Red";
    Color2[Color2["Fuchsia"] = 12] = "Fuchsia";
    Color2[Color2["Yellow"] = 13] = "Yellow";
    Color2[Color2["White"] = 14] = "White";
    Color2[Color2["Nickel"] = 15] = "Nickel";
    Color2[Color2["Chrome"] = 16] = "Chrome";
    Color2[Color2["Brass"] = 17] = "Brass";
    Color2[Color2["Copper"] = 18] = "Copper";
    Color2[Color2["Silver"] = 19] = "Silver";
    Color2[Color2["Gold"] = 20] = "Gold";
  })(Color = BridgedDeviceBasicInformation2.Color || (BridgedDeviceBasicInformation2.Color = {}));
  BridgedDeviceBasicInformation2.TlvProductAppearanceStruct = (0, import_TlvObject.TlvObject)({
    finish: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    primaryColor: (0, import_TlvObject.TlvField)(1, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)()))
  });
  BridgedDeviceBasicInformation2.TlvStartUpEvent = (0, import_TlvObject.TlvObject)({ softwareVersion: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt32) });
  BridgedDeviceBasicInformation2.TlvReachableChangedEvent = (0, import_TlvObject.TlvObject)({ reachableNewValue: (0, import_TlvObject.TlvField)(0, import_TlvBoolean.TlvBoolean) });
  BridgedDeviceBasicInformation2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 57,
    name: "BridgedDeviceBasicInformation",
    revision: 2,
    attributes: {
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      vendorName: (0, import_Cluster.OptionalFixedAttribute)(1, import_TlvString.TlvString.bound({ maxLength: 32 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      vendorId: (0, import_Cluster.OptionalFixedAttribute)(2, import_VendorId.TlvVendorId),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      productName: (0, import_Cluster.OptionalFixedAttribute)(3, import_TlvString.TlvString.bound({ maxLength: 32 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      nodeLabel: (0, import_Cluster.OptionalWritableAttribute)(
        5,
        import_TlvString.TlvString.bound({ maxLength: 32 }),
        { persistent: true, default: "", writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      hardwareVersion: (0, import_Cluster.OptionalFixedAttribute)(7, import_TlvNumber.TlvUInt16, { default: 0 }),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      hardwareVersionString: (0, import_Cluster.OptionalFixedAttribute)(8, import_TlvString.TlvString.bound({ minLength: 1, maxLength: 64 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      softwareVersion: (0, import_Cluster.OptionalFixedAttribute)(9, import_TlvNumber.TlvUInt32, { default: 0 }),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      softwareVersionString: (0, import_Cluster.OptionalFixedAttribute)(10, import_TlvString.TlvString.bound({ minLength: 1, maxLength: 64 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      manufacturingDate: (0, import_Cluster.OptionalFixedAttribute)(11, import_TlvString.TlvString.bound({ minLength: 8, maxLength: 16 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      partNumber: (0, import_Cluster.OptionalFixedAttribute)(12, import_TlvString.TlvString.bound({ maxLength: 32 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      productUrl: (0, import_Cluster.OptionalFixedAttribute)(13, import_TlvString.TlvString.bound({ maxLength: 256 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      productLabel: (0, import_Cluster.OptionalFixedAttribute)(14, import_TlvString.TlvString.bound({ maxLength: 64 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      serialNumber: (0, import_Cluster.OptionalFixedAttribute)(15, import_TlvString.TlvString.bound({ maxLength: 32 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      reachable: (0, import_Cluster.Attribute)(17, import_TlvBoolean.TlvBoolean, { default: true }),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      uniqueId: (0, import_Cluster.OptionalFixedAttribute)(18, import_TlvString.TlvString.bound({ maxLength: 32 })),
      productAppearance: (0, import_Cluster.OptionalFixedAttribute)(20, BridgedDeviceBasicInformation2.TlvProductAppearanceStruct)
    },
    events: {
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.5
       */
      startUp: (0, import_Cluster.OptionalEvent)(0, import_Cluster.EventPriority.Critical, BridgedDeviceBasicInformation2.TlvStartUpEvent),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.5
       */
      shutDown: (0, import_Cluster.OptionalEvent)(1, import_Cluster.EventPriority.Critical, import_TlvNoArguments.TlvNoArguments),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.5
       */
      leave: (0, import_Cluster.OptionalEvent)(2, import_Cluster.EventPriority.Info, import_TlvNoArguments.TlvNoArguments),
      /**
       * This event shall be generated when there is a change in the Reachable attribute. Its purpose is to
       * provide an indication towards interested parties that the reachability of a bridged device (over the
       * non-Matter network) has changed, so they may take appropriate action.
       *
       * After (re)start of a bridge this event may be generated.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.5.1
       */
      reachableChanged: (0, import_Cluster.Event)(3, import_Cluster.EventPriority.Info, BridgedDeviceBasicInformation2.TlvReachableChangedEvent)
    }
  });
})(BridgedDeviceBasicInformation || (BridgedDeviceBasicInformation = {}));
const BridgedDeviceBasicInformationCluster = BridgedDeviceBasicInformation.Cluster;
