// @project-chip/matter.js/dist/cjs/cluster/definitions/ChannelCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChannelCluster_exports = {};
__export(ChannelCluster_exports, {
  Channel: () => Channel,
  ChannelCluster: () => ChannelCluster
});
module.exports = __toCommonJS(ChannelCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Channel;
((Channel2) => {
  Channel2.TlvChannelInfoStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the channel major number value (for example, using ATSC format). When the channel number
     * is expressed as a string, such as "13.1" or "256", the major number would be 13 or 256, respectively.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.1.1
     */
    majorNumber: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16),
    /**
     * This shall indicate the channel minor number value (for example, using ATSC format). When the channel number
     * is expressed as a string, such as "13.1" or "256", the minor number would be 1 or 0, respectively.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.1.2
     */
    minorNumber: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16),
    /**
     * This shall indicate the marketing name for the channel, such as “The CW" or "Comedy Central". This field is
     * optional, but SHOULD be provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.1.3
     */
    name: (0, import_TlvObject.TlvOptionalField)(2, import_TlvString.TlvString),
    /**
     * This shall indicate the call sign of the channel, such as "PBS". This field is optional, but SHOULD be
     * provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.1.4
     */
    callSign: (0, import_TlvObject.TlvOptionalField)(3, import_TlvString.TlvString),
    /**
     * This shall indicate the local affiliate call sign, such as "KCTS". This field is optional, but SHOULD be
     * provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.1.5
     */
    affiliateCallSign: (0, import_TlvObject.TlvOptionalField)(4, import_TlvString.TlvString)
  });
  Channel2.TlvChangeChannelByNumberRequest = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the channel major number value (ATSC format) to which the channel should change.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.3.1
     */
    majorNumber: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16),
    /**
     * This shall indicate the channel minor number value (ATSC format) to which the channel should change.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.3.2
     */
    minorNumber: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16)
  });
  Channel2.TlvSkipChannelRequest = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the number of steps to increase (Count is positive) or decrease (Count is negative) the
     * current channel.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.4.1
     */
    count: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvInt16)
  });
  let LineupInfoType;
  ((LineupInfoType2) => {
    LineupInfoType2[LineupInfoType2["Mso"] = 0] = "Mso";
  })(LineupInfoType = Channel2.LineupInfoType || (Channel2.LineupInfoType = {}));
  Channel2.TlvLineupInfoStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the name of the operator, for example “Comcast”.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.2.1
     */
    operatorName: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvString),
    lineupName: (0, import_TlvObject.TlvOptionalField)(1, import_TlvString.TlvString),
    postalCode: (0, import_TlvObject.TlvOptionalField)(2, import_TlvString.TlvString),
    /**
     * This shall indicate the type of lineup. This field is optional, but SHOULD be provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.2.4
     */
    lineupInfoType: (0, import_TlvObject.TlvField)(3, (0, import_TlvNumber.TlvEnum)())
  });
  Channel2.TlvChangeChannelRequest = (0, import_TlvObject.TlvObject)({
    /**
     * This shall contain a user-input string to match in order to identify the target channel.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.1.1
     */
    match: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvString)
  });
  let Status;
  ((Status2) => {
    Status2[Status2["Success"] = 0] = "Success";
    Status2[Status2["MultipleMatches"] = 1] = "MultipleMatches";
    Status2[Status2["NoMatches"] = 2] = "NoMatches";
  })(Status = Channel2.Status || (Channel2.Status = {}));
  Channel2.TlvChangeChannelResponse = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the status of the command which resulted in this response.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.2.1
     */
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.2.2
     */
    data: (0, import_TlvObject.TlvOptionalField)(1, import_TlvString.TlvByteString)
  });
  let Feature;
  ((Feature2) => {
    Feature2["ChannelList"] = "ChannelList";
    Feature2["LineupInfo"] = "LineupInfo";
  })(Feature = Channel2.Feature || (Channel2.Feature = {}));
  Channel2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 1284,
    name: "Channel",
    revision: 1,
    features: {
      /**
       * ChannelList
       *
       * Provides list of available channels.
       */
      channelList: (0, import_BitmapSchema.BitFlag)(0),
      /**
       * LineupInfo
       *
       * Provides lineup info, which is a reference to an external source of lineup information.
       */
      lineupInfo: (0, import_BitmapSchema.BitFlag)(1)
    },
    attributes: {
      /**
       * This optional field contains the current channel. When supported but a channel is not currently tuned to
       * (if a content application is in foreground), the value of the field shall be null.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.3.3
       */
      currentChannel: (0, import_Cluster.OptionalAttribute)(2, (0, import_TlvNullable.TlvNullable)(Channel2.TlvChannelInfoStruct), { default: null })
    },
    commands: {
      /**
       * Change the channel to the channel with the given Number in the ChannelList attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.3
       */
      changeChannelByNumber: (0, import_Cluster.Command)(2, Channel2.TlvChangeChannelByNumberRequest, 2, import_Cluster.TlvNoResponse),
      /**
       * This command provides channel up and channel down functionality, but allows channel index jumps of size
       * Count.
       *
       * When the value of the increase or decrease is larger than the number of channels remaining in the given
       * direction, then the behavior shall be to return to the beginning (or end) of the channel list and
       * continue. For example, if the current channel is at index 0 and count value of -1 is given, then the
       * current channel should change to the last channel.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.4
       */
      skipChannel: (0, import_Cluster.Command)(3, Channel2.TlvSkipChannelRequest, 3, import_Cluster.TlvNoResponse)
    }
  });
  Channel2.ChannelListComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This optional list provides the channels supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.3.1
       */
      channelList: (0, import_Cluster.Attribute)(0, (0, import_TlvArray.TlvArray)(Channel2.TlvChannelInfoStruct), { default: [] })
    }
  });
  Channel2.LineupInfoComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This optional field identifies the channel lineup using external data sources.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.3.2
       */
      lineup: (0, import_Cluster.Attribute)(1, (0, import_TlvNullable.TlvNullable)(Channel2.TlvLineupInfoStruct), { default: null })
    }
  });
  Channel2.ChannelListOrLineupInfoComponent = import_ClusterFactory.ClusterFactory.Component({
    commands: {
      /**
       * Change the channel to the channel case-insensitive exact matching the value passed as an argument.
       *
       * The match priority order shall be: AffiliateCallSign ("KCTS"), CallSign ("PBS"), Name ("Comedy
       * Central"), Number ("13.1")
       *
       * Upon receipt, this shall generate a ChangeChannelResponse command.
       *
       * Upon success, the CurrentChannel attribute, if supported, shall be updated to reflect the change.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.1
       */
      changeChannel: (0, import_Cluster.Command)(0, Channel2.TlvChangeChannelRequest, 1, Channel2.TlvChangeChannelResponse)
    }
  });
  Channel2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    Channel2.Base,
    /**
     * Use this factory method to create a Channel cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a Channel cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...Channel2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(Channel2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, Channel2.ChannelListComponent, { channelList: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Channel2.LineupInfoComponent, { lineupInfo: true });
      import_ClusterFactory.ClusterFactory.extend(
        cluster,
        Channel2.ChannelListOrLineupInfoComponent,
        { channelList: true },
        { lineupInfo: true }
      );
      return cluster;
    }
  );
  const CL = { channelList: true };
  const LI = { lineupInfo: true };
  Channel2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: Channel2.Cluster.id,
    name: Channel2.Cluster.name,
    revision: Channel2.Cluster.revision,
    features: Channel2.Cluster.features,
    attributes: {
      ...Channel2.Cluster.attributes,
      channelList: import_ClusterFactory.ClusterFactory.AsConditional(
        Channel2.ChannelListComponent.attributes.channelList,
        { mandatoryIf: [CL] }
      ),
      lineup: import_ClusterFactory.ClusterFactory.AsConditional(Channel2.LineupInfoComponent.attributes.lineup, { mandatoryIf: [LI] })
    },
    commands: {
      ...Channel2.Cluster.commands,
      changeChannel: import_ClusterFactory.ClusterFactory.AsConditional(
        Channel2.ChannelListOrLineupInfoComponent.commands.changeChannel,
        { mandatoryIf: [CL, LI] }
      )
    }
  });
})(Channel || (Channel = {}));
const ChannelCluster = Channel.Cluster;
