// @project-chip/matter.js/dist/cjs/cluster/definitions/ClientMonitoringCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClientMonitoringCluster_exports = {};
__export(ClientMonitoringCluster_exports, {
  ClientMonitoring: () => ClientMonitoring,
  ClientMonitoringCluster: () => ClientMonitoringCluster
});
module.exports = __toCommonJS(ClientMonitoringCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_NodeId = require("../../datatype/NodeId.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ClientMonitoring;
((ClientMonitoring2) => {
  ClientMonitoring2.TlvMonitoringRegistration = (0, import_TlvObject.TlvObject)({
    clientNodeId: (0, import_TlvObject.TlvField)(1, import_NodeId.TlvNodeId),
    iCid: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt64)
  });
  ClientMonitoring2.TlvRegisterClientMonitoringRequest = (0, import_TlvObject.TlvObject)({
    clientNodeId: (0, import_TlvObject.TlvField)(0, import_NodeId.TlvNodeId),
    iCid: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt64)
  });
  ClientMonitoring2.TlvUnregisterClientMonitoringRequest = (0, import_TlvObject.TlvObject)({
    clientNodeId: (0, import_TlvObject.TlvField)(0, import_NodeId.TlvNodeId),
    iCid: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt64)
  });
  ClientMonitoring2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 4166,
    name: "ClientMonitoring",
    revision: 1,
    attributes: {
      idleModeInterval: (0, import_Cluster.Attribute)(0, import_TlvNumber.TlvUInt32.bound({ min: 300, max: 864e5 }), { default: 18 }),
      activeModeInterval: (0, import_Cluster.Attribute)(1, import_TlvNumber.TlvUInt32.bound({ min: 300, max: 864e5 }), { default: 18 }),
      activeModeThreshold: (0, import_Cluster.Attribute)(2, import_TlvNumber.TlvUInt16.bound({ min: 300, max: 6e4 }), { default: 0 }),
      expectedClients: (0, import_Cluster.Attribute)(3, (0, import_TlvArray.TlvArray)(ClientMonitoring2.TlvMonitoringRegistration, { maxLength: 4 }), { default: [] })
    },
    commands: {
      registerClientMonitoring: (0, import_Cluster.Command)(
        0,
        ClientMonitoring2.TlvRegisterClientMonitoringRequest,
        0,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      ),
      unregisterClientMonitoring: (0, import_Cluster.Command)(
        1,
        ClientMonitoring2.TlvUnregisterClientMonitoringRequest,
        1,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      ),
      stayAwakeRequest: (0, import_Cluster.OptionalCommand)(
        2,
        import_TlvNoArguments.TlvNoArguments,
        2,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      )
    }
  });
})(ClientMonitoring || (ClientMonitoring = {}));
const ClientMonitoringCluster = ClientMonitoring.Cluster;
