// @project-chip/matter.js/dist/cjs/cluster/definitions/ContentLauncherCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ContentLauncherCluster_exports = {};
__export(ContentLauncherCluster_exports, {
  ContentLauncher: () => ContentLauncher,
  ContentLauncherCluster: () => ContentLauncherCluster
});
module.exports = __toCommonJS(ContentLauncherCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ContentLauncher;
((ContentLauncher2) => {
  ContentLauncher2.SupportedStreamingProtocols = {
    /**
     * Device supports Dynamic Adaptive Streaming over HTTP (DASH)
     */
    dash: (0, import_BitmapSchema.BitFlag)(0),
    /**
     * Device supports HTTP Live Streaming (HLS)
     */
    hls: (0, import_BitmapSchema.BitFlag)(1)
  };
  let MetricType;
  ((MetricType2) => {
    MetricType2[MetricType2["Pixels"] = 0] = "Pixels";
    MetricType2[MetricType2["Percentage"] = 1] = "Percentage";
  })(MetricType = ContentLauncher2.MetricType || (ContentLauncher2.MetricType = {}));
  ContentLauncher2.TlvDimensionStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This indicates the width using the metric defined in Metric
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.8.1
     */
    width: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvDouble),
    /**
     * This indicates the height using the metric defined in Metric
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.8.2
     */
    height: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvDouble),
    /**
     * This shall indicate metric used for defining Height/Width.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.8.3
     */
    metric: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvEnum)())
  });
  ContentLauncher2.TlvStyleInformationStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the URL of image used for Styling different Video Player sections like Logo, Watermark
     * etc.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.7.1
     */
    imageUrl: (0, import_TlvObject.TlvOptionalField)(0, import_TlvString.TlvString.bound({ maxLength: 8192 })),
    /**
     * This shall indicate the color, in RGB or RGBA, used for styling different Video Player sections like Logo,
     * Watermark, etc. The value shall conform to the 6-digit or 8-digit format defined for CSS sRGB hexadecimal
     * color notation [https://www.w3.org/TR/css-color-4/#hex-notation]. Examples:
     *
     *   • #76DE19 for R=0x76, G=0xDE, B=0x19, A absent
     *
     *   • #76DE1980 for R=0x76, G=0xDE, B=0x19, A=0x80
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.7.2
     */
    color: (0, import_TlvObject.TlvOptionalField)(1, import_TlvString.TlvString),
    /**
     * This shall indicate the size of the image used for Styling different Video Player sections like Logo,
     * Watermark etc.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.7.3
     */
    size: (0, import_TlvObject.TlvOptionalField)(2, ContentLauncher2.TlvDimensionStruct)
  });
  ContentLauncher2.TlvBrandingInformationStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate name of of the provider for the given content.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.1
     */
    providerName: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvString.bound({ maxLength: 256 })),
    /**
     * This shall indicate background of the Video Player while content launch request is being processed by it.
     * This background information may also be used by the Video Player when it is in idle state.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.2
     */
    background: (0, import_TlvObject.TlvOptionalField)(1, ContentLauncher2.TlvStyleInformationStruct),
    /**
     * This shall indicate the logo shown when the Video Player is launching. This is also used when the Video
     * Player is in the idle state and Splash field is not available.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.3
     */
    logo: (0, import_TlvObject.TlvOptionalField)(2, ContentLauncher2.TlvStyleInformationStruct),
    /**
     * This shall indicate the style of progress bar for media playback.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.4
     */
    progressBar: (0, import_TlvObject.TlvOptionalField)(3, ContentLauncher2.TlvStyleInformationStruct),
    /**
     * This shall indicate the screen shown when the Video Player is in an idle state. If this property is not
     * populated, the Video Player shall default to logo or the provider name.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.5
     */
    splash: (0, import_TlvObject.TlvOptionalField)(4, ContentLauncher2.TlvStyleInformationStruct),
    /**
     * This shall indicate watermark shown when the media is playing.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.6
     */
    waterMark: (0, import_TlvObject.TlvOptionalField)(5, ContentLauncher2.TlvStyleInformationStruct)
  });
  ContentLauncher2.TlvLaunchUrlRequest = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the URL of content to launch.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.2.1
     */
    contentUrl: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvString),
    /**
     * This field, if present, shall provide a string that may be used to describe the content being accessed at
     * the given URL.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.2.2
     */
    displayString: (0, import_TlvObject.TlvOptionalField)(1, import_TlvString.TlvString),
    /**
     * This field, if present, shall indicate the branding information that may be displayed when playing back the
     * given content.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.2.3
     */
    brandingInformation: (0, import_TlvObject.TlvOptionalField)(2, ContentLauncher2.TlvBrandingInformationStruct)
  });
  let Status;
  ((Status2) => {
    Status2[Status2["Success"] = 0] = "Success";
    Status2[Status2["UrlNotAvailable"] = 1] = "UrlNotAvailable";
    Status2[Status2["AuthFailed"] = 2] = "AuthFailed";
  })(Status = ContentLauncher2.Status || (ContentLauncher2.Status = {}));
  ContentLauncher2.TlvLauncherResponse = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the status of the command which resulted in this response.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.3.1
     */
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.3.2
     */
    data: (0, import_TlvObject.TlvOptionalField)(1, import_TlvString.TlvByteString)
  });
  let Parameter;
  ((Parameter2) => {
    Parameter2[Parameter2["Actor"] = 0] = "Actor";
    Parameter2[Parameter2["Channel"] = 1] = "Channel";
    Parameter2[Parameter2["Character"] = 2] = "Character";
    Parameter2[Parameter2["Director"] = 3] = "Director";
    Parameter2[Parameter2["Event"] = 4] = "Event";
    Parameter2[Parameter2["Franchise"] = 5] = "Franchise";
    Parameter2[Parameter2["Genre"] = 6] = "Genre";
    Parameter2[Parameter2["League"] = 7] = "League";
    Parameter2[Parameter2["Popularity"] = 8] = "Popularity";
    Parameter2[Parameter2["Provider"] = 9] = "Provider";
    Parameter2[Parameter2["Sport"] = 10] = "Sport";
    Parameter2[Parameter2["SportsTeam"] = 11] = "SportsTeam";
    Parameter2[Parameter2["Type"] = 12] = "Type";
    Parameter2[Parameter2["Video"] = 13] = "Video";
  })(Parameter = ContentLauncher2.Parameter || (ContentLauncher2.Parameter = {}));
  ContentLauncher2.TlvAdditionalInfoStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the name of external id, ex. "musicbrainz".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.5.1
     */
    name: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvString.bound({ maxLength: 256 })),
    /**
     * This shall indicate the value for external id, ex. "ST0000000666661".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.5.2
     */
    value: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvString.bound({ maxLength: 8192 }))
  });
  ContentLauncher2.TlvParameterStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the entity type.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.3.1
     */
    type: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * This shall indicate the entity value, which is a search string, ex. “Manchester by the Sea”.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.3.2
     */
    value: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvString.bound({ maxLength: 1024 })),
    /**
     * This shall indicate the list of additional external content identifiers.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.3.3
     */
    externalIdList: (0, import_TlvObject.TlvOptionalField)(2, (0, import_TlvArray.TlvArray)(ContentLauncher2.TlvAdditionalInfoStruct))
  });
  ContentLauncher2.TlvContentSearchStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the list of parameters comprising the search. If multiple parameters are provided, the
     * search parameters shall be joined with 'AND' logic. e.g. action movies with Tom Cruise will be represented
     * as [{Actor: 'Tom Cruise'}, {Type: 'Movie'}, {Genre: 'Action'}]
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.2.1
     */
    parameterList: (0, import_TlvObject.TlvField)(0, (0, import_TlvArray.TlvArray)(ContentLauncher2.TlvParameterStruct))
  });
  ContentLauncher2.TlvLaunchContentRequest = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the content to launch.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.1.1
     */
    search: (0, import_TlvObject.TlvField)(0, ContentLauncher2.TlvContentSearchStruct),
    /**
     * This shall indicate whether to automatically start playing content, where: * TRUE means best match should
     * start playing automatically. * FALSE means matches should be displayed on screen for user selection.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.1.2
     */
    autoPlay: (0, import_TlvObject.TlvField)(1, import_TlvBoolean.TlvBoolean),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.1.3
     */
    data: (0, import_TlvObject.TlvOptionalField)(2, import_TlvString.TlvByteString)
  });
  let Feature;
  ((Feature2) => {
    Feature2["ContentSearch"] = "ContentSearch";
    Feature2["UrlPlayback"] = "UrlPlayback";
  })(Feature = ContentLauncher2.Feature || (ContentLauncher2.Feature = {}));
  ContentLauncher2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 1290,
    name: "ContentLauncher",
    revision: 1,
    features: {
      /**
       * ContentSearch
       *
       * Device supports content search (non-app specific)
       */
      contentSearch: (0, import_BitmapSchema.BitFlag)(0),
      /**
       * UrlPlayback
       *
       * Device supports basic URL-based file playback
       */
      urlPlayback: (0, import_BitmapSchema.BitFlag)(1)
    }
  });
  ContentLauncher2.UrlPlaybackComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This list provides list of content types supported by the Video Player or Content App in the form of
       * entries in the HTTP "Accept" request header.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.3.1
       */
      acceptHeader: (0, import_Cluster.Attribute)(0, (0, import_TlvArray.TlvArray)(import_TlvString.TlvString, { maxLength: 100 }), { persistent: true, default: [] }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.3.2.1
       */
      supportedStreamingProtocols: (0, import_Cluster.Attribute)(
        1,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt32, ContentLauncher2.SupportedStreamingProtocols),
        { persistent: true }
      )
    },
    commands: {
      /**
       * Upon receipt, this shall launch content from the specified URL.
       *
       * The content types supported include those identified in the AcceptHeader and SupportedStreamingProtocols
       * attributes.
       *
       * A check shall be made to ensure the URL is secure (uses HTTPS). This command returns a Launch Response.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.2
       */
      launchUrl: (0, import_Cluster.Command)(1, ContentLauncher2.TlvLaunchUrlRequest, 2, ContentLauncher2.TlvLauncherResponse)
    }
  });
  ContentLauncher2.ContentSearchComponent = import_ClusterFactory.ClusterFactory.Component({
    commands: {
      /**
       * Upon receipt, this shall launch the specified content with optional search criteria. This command
       * returns a Launch Response.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.1
       */
      launchContent: (0, import_Cluster.Command)(0, ContentLauncher2.TlvLaunchContentRequest, 2, ContentLauncher2.TlvLauncherResponse)
    }
  });
  ContentLauncher2.ContentSearchOrUrlPlaybackComponent = import_ClusterFactory.ClusterFactory.Component({});
  ContentLauncher2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    ContentLauncher2.Base,
    /**
     * Use this factory method to create a ContentLauncher cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a ContentLauncher cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...ContentLauncher2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(ContentLauncher2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, ContentLauncher2.UrlPlaybackComponent, { urlPlayback: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, ContentLauncher2.ContentSearchComponent, { contentSearch: true });
      import_ClusterFactory.ClusterFactory.extend(
        cluster,
        ContentLauncher2.ContentSearchOrUrlPlaybackComponent,
        { contentSearch: true },
        { urlPlayback: true }
      );
      return cluster;
    }
  );
  const UP = { urlPlayback: true };
  const CS = { contentSearch: true };
  ContentLauncher2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: ContentLauncher2.Cluster.id,
    name: ContentLauncher2.Cluster.name,
    revision: ContentLauncher2.Cluster.revision,
    features: ContentLauncher2.Cluster.features,
    attributes: {
      acceptHeader: import_ClusterFactory.ClusterFactory.AsConditional(
        ContentLauncher2.UrlPlaybackComponent.attributes.acceptHeader,
        { mandatoryIf: [UP] }
      ),
      supportedStreamingProtocols: import_ClusterFactory.ClusterFactory.AsConditional(
        ContentLauncher2.UrlPlaybackComponent.attributes.supportedStreamingProtocols,
        { mandatoryIf: [UP] }
      )
    },
    commands: {
      launchContent: import_ClusterFactory.ClusterFactory.AsConditional(
        ContentLauncher2.ContentSearchComponent.commands.launchContent,
        { mandatoryIf: [CS] }
      ),
      launchUrl: import_ClusterFactory.ClusterFactory.AsConditional(ContentLauncher2.UrlPlaybackComponent.commands.launchUrl, { mandatoryIf: [UP] })
    }
  });
})(ContentLauncher || (ContentLauncher = {}));
const ContentLauncherCluster = ContentLauncher.Cluster;
