// @project-chip/matter.js/dist/cjs/cluster/definitions/DescriptorCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DescriptorCluster_exports = {};
__export(DescriptorCluster_exports, {
  Descriptor: () => Descriptor,
  DescriptorCluster: () => DescriptorCluster
});
module.exports = __toCommonJS(DescriptorCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_DeviceTypeId = require("../../datatype/DeviceTypeId.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_ClusterId = require("../../datatype/ClusterId.js");
var import_EndpointNumber = require("../../datatype/EndpointNumber.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Descriptor;
((Descriptor2) => {
  Descriptor2.TlvDeviceTypeStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the device type definition. The endpoint shall conform to the device type definition and
     * cluster specifications required by the device type.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 9.5.4.1.1
     */
    deviceType: (0, import_TlvObject.TlvField)(0, import_DeviceTypeId.TlvDeviceTypeId),
    /**
     * This is the implemented revision of the device type definition. The endpoint shall conform to this revision
     * of the device type.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 9.5.4.1.2
     */
    revision: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16.bound({ min: 1 }))
  });
  Descriptor2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 29,
    name: "Descriptor",
    revision: 1,
    attributes: {
      /**
       * This is a list of device types and corresponding revisions declaring endpoint conformance (see
       * DeviceTypeStruct). At least one device type entry shall be present.
       *
       * An endpoint shall conform to all device types listed in the DeviceTypeList. A cluster instance that is
       * in common for more than one device type in the DeviceTypeList shall be supported as a shared cluster
       * instance on the endpoint.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.5.5.1
       */
      deviceTypeList: (0, import_Cluster.FixedAttribute)(0, (0, import_TlvArray.TlvArray)(Descriptor2.TlvDeviceTypeStruct, { minLength: 1 })),
      /**
       * This attribute shall list each cluster ID for the server clusters present on the endpoint instance.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.5.5.2
       */
      serverList: (0, import_Cluster.FixedAttribute)(1, (0, import_TlvArray.TlvArray)(import_ClusterId.TlvClusterId), { default: [] }),
      /**
       * This attribute shall list each cluster ID for the client clusters present on the endpoint instance.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.5.5.3
       */
      clientList: (0, import_Cluster.FixedAttribute)(2, (0, import_TlvArray.TlvArray)(import_ClusterId.TlvClusterId), { default: [] }),
      /**
       * This attribute indicates composition of the device type instance. Device type instance composition shall
       * include the endpoints in this list. See Endpoint Composition for more information which endpoints to
       * include in this list.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.5.5.4
       */
      partsList: (0, import_Cluster.Attribute)(3, (0, import_TlvArray.TlvArray)(import_EndpointNumber.TlvEndpointNumber), { default: [] })
    }
  });
})(Descriptor || (Descriptor = {}));
const DescriptorCluster = Descriptor.Cluster;
