// @project-chip/matter.js/dist/cjs/cluster/definitions/DoorLockCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DoorLockCluster_exports = {};
__export(DoorLockCluster_exports, {
  DoorLock: () => DoorLock,
  DoorLockCluster: () => DoorLockCluster
});
module.exports = __toCommonJS(DoorLockCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_FabricIndex = require("../../datatype/FabricIndex.js");
var import_NodeId = require("../../datatype/NodeId.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DoorLock;
((DoorLock2) => {
  let LockState;
  ((LockState2) => {
    LockState2[LockState2["NotFullyLocked"] = 0] = "NotFullyLocked";
    LockState2[LockState2["Locked"] = 1] = "Locked";
    LockState2[LockState2["Unlocked"] = 2] = "Unlocked";
  })(LockState = DoorLock2.LockState || (DoorLock2.LockState = {}));
  let LockType;
  ((LockType2) => {
    LockType2[LockType2["Deadbolt"] = 0] = "Deadbolt";
    LockType2[LockType2["Magnetic"] = 1] = "Magnetic";
    LockType2[LockType2["Other"] = 2] = "Other";
    LockType2[LockType2["Mortise"] = 3] = "Mortise";
    LockType2[LockType2["Rim"] = 4] = "Rim";
    LockType2[LockType2["LatchBolt"] = 5] = "LatchBolt";
    LockType2[LockType2["CylindricalLock"] = 6] = "CylindricalLock";
    LockType2[LockType2["TubularLock"] = 7] = "TubularLock";
    LockType2[LockType2["InterconnectedLock"] = 8] = "InterconnectedLock";
    LockType2[LockType2["DeadLatch"] = 9] = "DeadLatch";
    LockType2[LockType2["DoorFurniture"] = 10] = "DoorFurniture";
  })(LockType = DoorLock2.LockType || (DoorLock2.LockType = {}));
  let OperatingMode;
  ((OperatingMode2) => {
    OperatingMode2[OperatingMode2["Normal"] = 0] = "Normal";
    OperatingMode2[OperatingMode2["Vacation"] = 1] = "Vacation";
    OperatingMode2[OperatingMode2["Privacy"] = 2] = "Privacy";
    OperatingMode2[OperatingMode2["NoRemoteLockUnlock"] = 3] = "NoRemoteLockUnlock";
    OperatingMode2[OperatingMode2["Passage"] = 4] = "Passage";
  })(OperatingMode = DoorLock2.OperatingMode || (DoorLock2.OperatingMode = {}));
  DoorLock2.SupportedOperatingModes = {
    normal: (0, import_BitmapSchema.BitFlag)(0),
    vacation: (0, import_BitmapSchema.BitFlag)(1),
    privacy: (0, import_BitmapSchema.BitFlag)(2),
    noRemoteLockUnlock: (0, import_BitmapSchema.BitFlag)(3),
    passage: (0, import_BitmapSchema.BitFlag)(4)
  };
  DoorLock2.DefaultConfigurationRegister = {
    enableLocalProgrammingEnabled: (0, import_BitmapSchema.BitFlag)(0),
    keypadInterfaceDefaultAccessEnabled: (0, import_BitmapSchema.BitFlag)(1),
    remoteInterfaceDefaultAccessIsEnabled: (0, import_BitmapSchema.BitFlag)(2),
    soundEnabled: (0, import_BitmapSchema.BitFlag)(5),
    autoRelockTimeSet: (0, import_BitmapSchema.BitFlag)(6),
    ledSettingsSet: (0, import_BitmapSchema.BitFlag)(7)
  };
  DoorLock2.LocalProgrammingFeatures = {
    addUsersCredentialsSchedulesLocally: (0, import_BitmapSchema.BitFlag)(0),
    modifyUsersCredentialsSchedulesLocally: (0, import_BitmapSchema.BitFlag)(1),
    clearUsersCredentialsSchedulesLocally: (0, import_BitmapSchema.BitFlag)(2),
    adjustLockSettingsLocally: (0, import_BitmapSchema.BitFlag)(3)
  };
  DoorLock2.AlarmMask = {
    lockingMechanismJammed: (0, import_BitmapSchema.BitFlag)(0),
    lockResetToFactoryDefaults: (0, import_BitmapSchema.BitFlag)(1),
    reserved: (0, import_BitmapSchema.BitFlag)(2),
    rfModulePowerCycled: (0, import_BitmapSchema.BitFlag)(3),
    tamperAlarmWrongCodeEntryLimit: (0, import_BitmapSchema.BitFlag)(4),
    tamperAlarmFrontEscutcheonRemovedFromMain: (0, import_BitmapSchema.BitFlag)(5),
    forcedDoorOpenUnderDoorLockedCondition: (0, import_BitmapSchema.BitFlag)(6)
  };
  DoorLock2.TlvLockDoorRequest = (0, import_TlvObject.TlvObject)({ pinCode: (0, import_TlvObject.TlvOptionalField)(0, import_TlvString.TlvByteString) });
  DoorLock2.TlvUnlockDoorRequest = (0, import_TlvObject.TlvObject)({ pinCode: (0, import_TlvObject.TlvOptionalField)(0, import_TlvString.TlvByteString) });
  DoorLock2.TlvUnlockWithTimeoutRequest = (0, import_TlvObject.TlvObject)({
    timeout: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16),
    pinCode: (0, import_TlvObject.TlvOptionalField)(1, import_TlvString.TlvByteString)
  });
  let AlarmCode;
  ((AlarmCode2) => {
    AlarmCode2[AlarmCode2["LockJammed"] = 0] = "LockJammed";
    AlarmCode2[AlarmCode2["LockFactoryReset"] = 1] = "LockFactoryReset";
    AlarmCode2[AlarmCode2["LockRadioPowerCycled"] = 3] = "LockRadioPowerCycled";
    AlarmCode2[AlarmCode2["WrongCodeEntryLimit"] = 4] = "WrongCodeEntryLimit";
    AlarmCode2[AlarmCode2["FrontEsceutcheonRemoved"] = 5] = "FrontEsceutcheonRemoved";
    AlarmCode2[AlarmCode2["DoorForcedOpen"] = 6] = "DoorForcedOpen";
    AlarmCode2[AlarmCode2["DoorAjar"] = 7] = "DoorAjar";
    AlarmCode2[AlarmCode2["ForcedUser"] = 8] = "ForcedUser";
  })(AlarmCode = DoorLock2.AlarmCode || (DoorLock2.AlarmCode = {}));
  DoorLock2.TlvDoorLockAlarmEvent = (0, import_TlvObject.TlvObject)({
    /**
     * The alarm code of the event that has happened.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.1.1
     */
    alarmCode: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)())
  });
  let LockOperationType;
  ((LockOperationType2) => {
    LockOperationType2[LockOperationType2["Lock"] = 0] = "Lock";
    LockOperationType2[LockOperationType2["Unlock"] = 1] = "Unlock";
    LockOperationType2[LockOperationType2["NonAccessUserEvent"] = 2] = "NonAccessUserEvent";
    LockOperationType2[LockOperationType2["ForcedUserEvent"] = 3] = "ForcedUserEvent";
  })(LockOperationType = DoorLock2.LockOperationType || (DoorLock2.LockOperationType = {}));
  let OperationSource;
  ((OperationSource2) => {
    OperationSource2[OperationSource2["Unspecified"] = 0] = "Unspecified";
    OperationSource2[OperationSource2["Manual"] = 1] = "Manual";
    OperationSource2[OperationSource2["ProprietaryRemote"] = 2] = "ProprietaryRemote";
    OperationSource2[OperationSource2["Keypad"] = 3] = "Keypad";
    OperationSource2[OperationSource2["Auto"] = 4] = "Auto";
    OperationSource2[OperationSource2["Button"] = 5] = "Button";
    OperationSource2[OperationSource2["Schedule"] = 6] = "Schedule";
    OperationSource2[OperationSource2["Remote"] = 7] = "Remote";
    OperationSource2[OperationSource2["Rfid"] = 8] = "Rfid";
    OperationSource2[OperationSource2["Biometric"] = 9] = "Biometric";
  })(OperationSource = DoorLock2.OperationSource || (DoorLock2.OperationSource = {}));
  let CredentialType;
  ((CredentialType2) => {
    CredentialType2[CredentialType2["ProgrammingPin"] = 0] = "ProgrammingPin";
    CredentialType2[CredentialType2["Pin"] = 1] = "Pin";
    CredentialType2[CredentialType2["Rfid"] = 2] = "Rfid";
    CredentialType2[CredentialType2["Fingerprint"] = 3] = "Fingerprint";
    CredentialType2[CredentialType2["FingerVein"] = 4] = "FingerVein";
    CredentialType2[CredentialType2["Face"] = 5] = "Face";
  })(CredentialType = DoorLock2.CredentialType || (DoorLock2.CredentialType = {}));
  DoorLock2.TlvCredentialStruct = (0, import_TlvObject.TlvObject)({
    /**
     * The credential type used to authorize the lock operation.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.6.3.1
     */
    credentialType: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * This is the index of the specific credential used to authorize the lock operation in the list of credentials
     * identified by CredentialType (e.g. schedule, PIN, RFID, etc.). This shall be set to 0 if CredentialType is
     * ProgrammingPIN or does not correspond to a list that can be indexed into.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.6.3.2
     */
    credentialIndex: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16)
  });
  DoorLock2.TlvLockOperationEvent = (0, import_TlvObject.TlvObject)({
    /**
     * The type of the lock operation that was performed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.1
     */
    lockOperationType: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The source of the lock operation that was performed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.2
     */
    operationSource: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The lock UserIndex who performed the lock operation. This shall be null if there is no user index that can
     * be determined for the given operation source. This shall NOT be null if a user index can be determined. In
     * particular, this shall NOT be null if the operation was associated with a valid credential.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.3
     */
    userIndex: (0, import_TlvObject.TlvField)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16)),
    /**
     * The fabric index of the fabric that performed the lock operation. This shall be null if there is no fabric
     * that can be determined for the given operation source. This shall NOT be null if the operation source is
     * "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.4
     */
    fabricIndex: (0, import_TlvObject.TlvField)(3, (0, import_TlvNullable.TlvNullable)(import_FabricIndex.TlvFabricIndex)),
    /**
     * The Node ID of the node that performed the lock operation. This shall be null if there is no Node associated
     * with the given operation source. This shall NOT be null if the operation source is "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.5
     */
    sourceNode: (0, import_TlvObject.TlvField)(4, (0, import_TlvNullable.TlvNullable)(import_NodeId.TlvNodeId)),
    /**
     * The list of credentials used in performing the lock operation. This shall be null if no credentials were
     * involved.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.6
     */
    credentials: (0, import_TlvObject.TlvOptionalField)(5, (0, import_TlvNullable.TlvNullable)((0, import_TlvArray.TlvArray)(DoorLock2.TlvCredentialStruct, { minLength: 1 })))
  });
  let OperationError;
  ((OperationError2) => {
    OperationError2[OperationError2["Unspecified"] = 0] = "Unspecified";
    OperationError2[OperationError2["InvalidCredential"] = 1] = "InvalidCredential";
    OperationError2[OperationError2["DisabledUserDenied"] = 2] = "DisabledUserDenied";
    OperationError2[OperationError2["Restricted"] = 3] = "Restricted";
    OperationError2[OperationError2["InsufficientBattery"] = 4] = "InsufficientBattery";
  })(OperationError = DoorLock2.OperationError || (DoorLock2.OperationError = {}));
  DoorLock2.TlvLockOperationErrorEvent = (0, import_TlvObject.TlvObject)({
    /**
     * The type of the lock operation that was performed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.1
     */
    lockOperationType: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The source of the lock operation that was performed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.2
     */
    operationSource: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The lock operation error triggered when the operation was performed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.3
     */
    operationError: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The lock UserIndex who performed the lock operation. This shall be null if there is no user id that can be
     * determined for the given operation source.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.4
     */
    userIndex: (0, import_TlvObject.TlvField)(3, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16)),
    /**
     * The fabric index of the fabric that performed the lock operation. This shall be null if there is no fabric
     * that can be determined for the given operation source. This shall NOT be null if the operation source is
     * "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.5
     */
    fabricIndex: (0, import_TlvObject.TlvField)(4, (0, import_TlvNullable.TlvNullable)(import_FabricIndex.TlvFabricIndex)),
    /**
     * The Node ID of the node that performed the lock operation. This shall be null if there is no Node associated
     * with the given operation source. This shall NOT be null if the operation source is
     *
     * "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.6
     */
    sourceNode: (0, import_TlvObject.TlvField)(5, (0, import_TlvNullable.TlvNullable)(import_NodeId.TlvNodeId)),
    /**
     * The list of credentials used in performing the lock operation. This shall be null if no credentials were
     * involved.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.7
     */
    credentials: (0, import_TlvObject.TlvOptionalField)(6, (0, import_TlvNullable.TlvNullable)((0, import_TlvArray.TlvArray)(DoorLock2.TlvCredentialStruct, { minLength: 1 })))
  });
  let DoorState;
  ((DoorState2) => {
    DoorState2[DoorState2["DoorOpen"] = 0] = "DoorOpen";
    DoorState2[DoorState2["DoorClosed"] = 1] = "DoorClosed";
    DoorState2[DoorState2["DoorJammed"] = 2] = "DoorJammed";
    DoorState2[DoorState2["DoorForcedOpen"] = 3] = "DoorForcedOpen";
    DoorState2[DoorState2["DoorUnspecifiedError"] = 4] = "DoorUnspecifiedError";
    DoorState2[DoorState2["DoorAjar"] = 5] = "DoorAjar";
  })(DoorState = DoorLock2.DoorState || (DoorLock2.DoorState = {}));
  DoorLock2.TlvDoorStateChangeEvent = (0, import_TlvObject.TlvObject)({
    /**
     * The new door state for this door event.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.2.1
     */
    doorState: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)())
  });
  DoorLock2.CredentialRulesSupport = { single: (0, import_BitmapSchema.BitFlag)(0), dual: (0, import_BitmapSchema.BitFlag)(1), tri: (0, import_BitmapSchema.BitFlag)(2) };
  let DataOperationType;
  ((DataOperationType2) => {
    DataOperationType2[DataOperationType2["Add"] = 0] = "Add";
    DataOperationType2[DataOperationType2["Clear"] = 1] = "Clear";
    DataOperationType2[DataOperationType2["Modify"] = 2] = "Modify";
  })(DataOperationType = DoorLock2.DataOperationType || (DoorLock2.DataOperationType = {}));
  let UserStatus;
  ((UserStatus2) => {
    UserStatus2[UserStatus2["Available"] = 0] = "Available";
    UserStatus2[UserStatus2["OccupiedEnabled"] = 1] = "OccupiedEnabled";
    UserStatus2[UserStatus2["OccupiedDisabled"] = 3] = "OccupiedDisabled";
  })(UserStatus = DoorLock2.UserStatus || (DoorLock2.UserStatus = {}));
  let UserType;
  ((UserType2) => {
    UserType2[UserType2["UnrestrictedUser"] = 0] = "UnrestrictedUser";
    UserType2[UserType2["YearDayScheduleUser"] = 1] = "YearDayScheduleUser";
    UserType2[UserType2["WeekDayScheduleUser"] = 2] = "WeekDayScheduleUser";
    UserType2[UserType2["ProgrammingUser"] = 3] = "ProgrammingUser";
    UserType2[UserType2["NonAccessUser"] = 4] = "NonAccessUser";
    UserType2[UserType2["ForcedUser"] = 5] = "ForcedUser";
    UserType2[UserType2["DisposableUser"] = 6] = "DisposableUser";
    UserType2[UserType2["ExpiringUser"] = 7] = "ExpiringUser";
    UserType2[UserType2["ScheduleRestrictedUser"] = 8] = "ScheduleRestrictedUser";
    UserType2[UserType2["RemoteOnlyUser"] = 9] = "RemoteOnlyUser";
  })(UserType = DoorLock2.UserType || (DoorLock2.UserType = {}));
  let CredentialRule;
  ((CredentialRule2) => {
    CredentialRule2[CredentialRule2["Single"] = 0] = "Single";
    CredentialRule2[CredentialRule2["Dual"] = 1] = "Dual";
    CredentialRule2[CredentialRule2["Tri"] = 2] = "Tri";
  })(CredentialRule = DoorLock2.CredentialRule || (DoorLock2.CredentialRule = {}));
  DoorLock2.TlvSetUserRequest = (0, import_TlvObject.TlvObject)({
    operationType: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    userIndex: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16),
    userName: (0, import_TlvObject.TlvField)(2, (0, import_TlvNullable.TlvNullable)(import_TlvString.TlvString)),
    userUniqueId: (0, import_TlvObject.TlvField)(3, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt32)),
    userStatus: (0, import_TlvObject.TlvField)(4, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)())),
    userType: (0, import_TlvObject.TlvField)(5, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)())),
    credentialRule: (0, import_TlvObject.TlvField)(6, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)()))
  });
  DoorLock2.TlvGetUserRequest = (0, import_TlvObject.TlvObject)({ userIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16) });
  DoorLock2.TlvGetUserResponse = (0, import_TlvObject.TlvObject)({
    userIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16),
    userName: (0, import_TlvObject.TlvField)(1, (0, import_TlvNullable.TlvNullable)(import_TlvString.TlvString)),
    userUniqueId: (0, import_TlvObject.TlvField)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt32)),
    userStatus: (0, import_TlvObject.TlvField)(3, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)())),
    userType: (0, import_TlvObject.TlvField)(4, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)())),
    credentialRule: (0, import_TlvObject.TlvField)(5, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)())),
    credentials: (0, import_TlvObject.TlvField)(6, (0, import_TlvNullable.TlvNullable)((0, import_TlvArray.TlvArray)(DoorLock2.TlvCredentialStruct))),
    creatorFabricIndex: (0, import_TlvObject.TlvField)(7, (0, import_TlvNullable.TlvNullable)(import_FabricIndex.TlvFabricIndex)),
    lastModifiedFabricIndex: (0, import_TlvObject.TlvField)(8, (0, import_TlvNullable.TlvNullable)(import_FabricIndex.TlvFabricIndex)),
    nextUserIndex: (0, import_TlvObject.TlvField)(9, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16))
  });
  DoorLock2.TlvClearUserRequest = (0, import_TlvObject.TlvObject)({ userIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16) });
  DoorLock2.TlvSetCredentialRequest = (0, import_TlvObject.TlvObject)({
    operationType: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    credential: (0, import_TlvObject.TlvField)(1, DoorLock2.TlvCredentialStruct),
    credentialData: (0, import_TlvObject.TlvField)(2, import_TlvString.TlvByteString),
    userIndex: (0, import_TlvObject.TlvField)(3, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16)),
    userStatus: (0, import_TlvObject.TlvField)(4, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)())),
    userType: (0, import_TlvObject.TlvField)(5, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)()))
  });
  let DlStatus;
  ((DlStatus2) => {
    DlStatus2[DlStatus2["Success"] = 0] = "Success";
    DlStatus2[DlStatus2["Failure"] = 1] = "Failure";
    DlStatus2[DlStatus2["Duplicate"] = 2] = "Duplicate";
    DlStatus2[DlStatus2["Occupied"] = 3] = "Occupied";
    DlStatus2[DlStatus2["InvalidField"] = 133] = "InvalidField";
    DlStatus2[DlStatus2["ResourceExhausted"] = 137] = "ResourceExhausted";
    DlStatus2[DlStatus2["NotFound"] = 139] = "NotFound";
  })(DlStatus = DoorLock2.DlStatus || (DoorLock2.DlStatus = {}));
  DoorLock2.TlvSetCredentialResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    userIndex: (0, import_TlvObject.TlvField)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16)),
    nextCredentialIndex: (0, import_TlvObject.TlvField)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16))
  });
  DoorLock2.TlvGetCredentialStatusRequest = (0, import_TlvObject.TlvObject)({ credential: (0, import_TlvObject.TlvField)(0, DoorLock2.TlvCredentialStruct) });
  DoorLock2.TlvGetCredentialStatusResponse = (0, import_TlvObject.TlvObject)({
    credentialExists: (0, import_TlvObject.TlvField)(0, import_TlvBoolean.TlvBoolean),
    userIndex: (0, import_TlvObject.TlvField)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16)),
    creatorFabricIndex: (0, import_TlvObject.TlvField)(2, (0, import_TlvNullable.TlvNullable)(import_FabricIndex.TlvFabricIndex)),
    lastModifiedFabricIndex: (0, import_TlvObject.TlvField)(3, (0, import_TlvNullable.TlvNullable)(import_FabricIndex.TlvFabricIndex)),
    nextCredentialIndex: (0, import_TlvObject.TlvField)(4, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16))
  });
  DoorLock2.TlvClearCredentialRequest = (0, import_TlvObject.TlvObject)({ credential: (0, import_TlvObject.TlvField)(0, (0, import_TlvNullable.TlvNullable)(DoorLock2.TlvCredentialStruct)) });
  let LockDataType;
  ((LockDataType2) => {
    LockDataType2[LockDataType2["Unspecified"] = 0] = "Unspecified";
    LockDataType2[LockDataType2["ProgrammingCode"] = 1] = "ProgrammingCode";
    LockDataType2[LockDataType2["UserIndex"] = 2] = "UserIndex";
    LockDataType2[LockDataType2["WeekDaySchedule"] = 3] = "WeekDaySchedule";
    LockDataType2[LockDataType2["YearDaySchedule"] = 4] = "YearDaySchedule";
    LockDataType2[LockDataType2["HolidaySchedule"] = 5] = "HolidaySchedule";
    LockDataType2[LockDataType2["Pin"] = 6] = "Pin";
    LockDataType2[LockDataType2["Rfid"] = 7] = "Rfid";
    LockDataType2[LockDataType2["Fingerprint"] = 8] = "Fingerprint";
    LockDataType2[LockDataType2["FingerVein"] = 9] = "FingerVein";
    LockDataType2[LockDataType2["Face"] = 10] = "Face";
  })(LockDataType = DoorLock2.LockDataType || (DoorLock2.LockDataType = {}));
  DoorLock2.TlvLockUserChangeEvent = (0, import_TlvObject.TlvObject)({
    /**
     * The lock data type that was changed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.1
     */
    lockDataType: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The data operation performed on the lock data type changed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.2
     */
    dataOperationType: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The source of the user data change.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.3
     */
    operationSource: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The lock UserIndex associated with the change (if any). This shall be null if there is no specific user
     * associated with the data operation. This shall be 0xFFFE if all users are affected (e.g. Clear Users).
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.4
     */
    userIndex: (0, import_TlvObject.TlvField)(3, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16)),
    /**
     * The fabric index of the fabric that performed the change (if any). This shall be null if there is no fabric
     * that can be determined to have caused the change. This shall NOT be null if the operation source is "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.5
     */
    fabricIndex: (0, import_TlvObject.TlvField)(4, (0, import_TlvNullable.TlvNullable)(import_FabricIndex.TlvFabricIndex)),
    /**
     * The Node ID that that performed the change (if any). The Node ID of the node that performed the change. This
     * shall be null if there was no Node involved in the change. This shall NOT be null if the operation source is
     * "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.6
     */
    sourceNode: (0, import_TlvObject.TlvField)(5, (0, import_TlvNullable.TlvNullable)(import_NodeId.TlvNodeId)),
    /**
     * This is the index of the specific item that was changed (e.g. schedule, PIN, RFID, etc.) in the list of
     * items identified by LockDataType. This shall be null if the LockDataType does not correspond to a list that
     * can be indexed into (e.g. ProgrammingUser). This shall be 0xFFFE if all indices are affected (e.g. Clear PIN
     * Code, Clear RFID Code, Clear Week Day Schedule, Clear Year Day Schedule, etc.).
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.7
     */
    dataIndex: (0, import_TlvObject.TlvField)(6, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16))
  });
  DoorLock2.DaysMaskMap = {
    sunday: (0, import_BitmapSchema.BitFlag)(0),
    monday: (0, import_BitmapSchema.BitFlag)(1),
    tuesday: (0, import_BitmapSchema.BitFlag)(2),
    wednesday: (0, import_BitmapSchema.BitFlag)(3),
    thursday: (0, import_BitmapSchema.BitFlag)(4),
    friday: (0, import_BitmapSchema.BitFlag)(5),
    saturday: (0, import_BitmapSchema.BitFlag)(6)
  };
  DoorLock2.TlvSetWeekDayScheduleRequest = (0, import_TlvObject.TlvObject)({
    weekDayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    userIndex: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16),
    daysMask: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, DoorLock2.DaysMaskMap)),
    startHour: (0, import_TlvObject.TlvField)(3, import_TlvNumber.TlvUInt8),
    startMinute: (0, import_TlvObject.TlvField)(4, import_TlvNumber.TlvUInt8),
    endHour: (0, import_TlvObject.TlvField)(5, import_TlvNumber.TlvUInt8),
    endMinute: (0, import_TlvObject.TlvField)(6, import_TlvNumber.TlvUInt8)
  });
  DoorLock2.TlvGetWeekDayScheduleRequest = (0, import_TlvObject.TlvObject)({
    weekDayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    userIndex: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16)
  });
  DoorLock2.TlvGetWeekDayScheduleResponse = (0, import_TlvObject.TlvObject)({
    weekDayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    userIndex: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16),
    status: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvEnum)()),
    daysMask: (0, import_TlvObject.TlvOptionalField)(3, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, DoorLock2.DaysMaskMap)),
    startHour: (0, import_TlvObject.TlvOptionalField)(4, import_TlvNumber.TlvUInt8),
    startMinute: (0, import_TlvObject.TlvOptionalField)(5, import_TlvNumber.TlvUInt8),
    endHour: (0, import_TlvObject.TlvOptionalField)(6, import_TlvNumber.TlvUInt8),
    endMinute: (0, import_TlvObject.TlvOptionalField)(7, import_TlvNumber.TlvUInt8)
  });
  DoorLock2.TlvClearWeekDayScheduleRequest = (0, import_TlvObject.TlvObject)({
    weekDayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    userIndex: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16)
  });
  DoorLock2.TlvSetYearDayScheduleRequest = (0, import_TlvObject.TlvObject)({
    yearDayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    userIndex: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16),
    localStartTime: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvEpochS),
    localEndTime: (0, import_TlvObject.TlvField)(3, import_TlvNumber.TlvEpochS)
  });
  DoorLock2.TlvGetYearDayScheduleRequest = (0, import_TlvObject.TlvObject)({
    yearDayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    userIndex: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16)
  });
  DoorLock2.TlvGetYearDayScheduleResponse = (0, import_TlvObject.TlvObject)({
    yearDayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    userIndex: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16),
    status: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvEnum)()),
    localStartTime: (0, import_TlvObject.TlvOptionalField)(3, import_TlvNumber.TlvEpochS),
    localEndTime: (0, import_TlvObject.TlvOptionalField)(4, import_TlvNumber.TlvEpochS)
  });
  DoorLock2.TlvClearYearDayScheduleRequest = (0, import_TlvObject.TlvObject)({
    yearDayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    userIndex: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16)
  });
  DoorLock2.TlvSetHolidayScheduleRequest = (0, import_TlvObject.TlvObject)({
    holidayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    localStartTime: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvEpochS),
    localEndTime: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvEpochS),
    operatingMode: (0, import_TlvObject.TlvField)(3, (0, import_TlvNumber.TlvEnum)())
  });
  DoorLock2.TlvGetHolidayScheduleRequest = (0, import_TlvObject.TlvObject)({ holidayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8) });
  DoorLock2.TlvGetHolidayScheduleResponse = (0, import_TlvObject.TlvObject)({
    holidayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    status: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvEnum)()),
    localStartTime: (0, import_TlvObject.TlvOptionalField)(2, import_TlvNumber.TlvEpochS),
    localEndTime: (0, import_TlvObject.TlvOptionalField)(3, import_TlvNumber.TlvEpochS),
    operatingMode: (0, import_TlvObject.TlvOptionalField)(4, (0, import_TlvNumber.TlvEnum)())
  });
  DoorLock2.TlvClearHolidayScheduleRequest = (0, import_TlvObject.TlvObject)({ holidayIndex: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8) });
  DoorLock2.KeypadOperationEventMask = {
    unknownOrManufacturerSpecificKeypadOperationEvent: (0, import_BitmapSchema.BitFlag)(0),
    lockSourceKeypad: (0, import_BitmapSchema.BitFlag)(1),
    unlockSourceKeypad: (0, import_BitmapSchema.BitFlag)(2),
    lockSourceKeypadErrorInvalidPin: (0, import_BitmapSchema.BitFlag)(3),
    lockSourceKeypadErrorInvalidSchedule: (0, import_BitmapSchema.BitFlag)(4),
    unlockSourceKeypadErrorInvalidCode: (0, import_BitmapSchema.BitFlag)(5),
    unlockSourceKeypadErrorInvalidSchedule: (0, import_BitmapSchema.BitFlag)(6),
    nonAccessUserOperationEventSourceKeypad: (0, import_BitmapSchema.BitFlag)(15)
  };
  DoorLock2.KeypadProgrammingEventMask = {
    unknown: (0, import_BitmapSchema.BitFlag)(0),
    pinCodeChanged: (0, import_BitmapSchema.BitFlag)(1),
    pinAdded: (0, import_BitmapSchema.BitFlag)(2),
    pinCleared: (0, import_BitmapSchema.BitFlag)(3),
    pinChanged: (0, import_BitmapSchema.BitFlag)(4)
  };
  DoorLock2.RemoteOperationEventMask = {
    unknownOrManufacturerSpecificRemoteOperationEvent: (0, import_BitmapSchema.BitFlag)(0),
    lockSourceRemote: (0, import_BitmapSchema.BitFlag)(1),
    unlockSourceRemote: (0, import_BitmapSchema.BitFlag)(2),
    lockSourceRemoteErrorInvalidCode: (0, import_BitmapSchema.BitFlag)(3),
    lockSourceRemoteErrorInvalidSchedule: (0, import_BitmapSchema.BitFlag)(4),
    unlockSourceRemoteErrorInvalidCode: (0, import_BitmapSchema.BitFlag)(5),
    unlockSourceRemoteErrorInvalidSchedule: (0, import_BitmapSchema.BitFlag)(6)
  };
  DoorLock2.ManualOperationEventMask = {
    unknownOrManufacturerSpecificManualOperationEvent: (0, import_BitmapSchema.BitFlag)(0),
    thumbturnLock: (0, import_BitmapSchema.BitFlag)(1),
    thumbturnUnlock: (0, import_BitmapSchema.BitFlag)(2),
    oneTouchLock: (0, import_BitmapSchema.BitFlag)(7),
    keyLock: (0, import_BitmapSchema.BitFlag)(8),
    keyUnlock: (0, import_BitmapSchema.BitFlag)(9),
    autoLock: (0, import_BitmapSchema.BitFlag)(10),
    scheduleLock: (0, import_BitmapSchema.BitFlag)(11),
    scheduleUnlock: (0, import_BitmapSchema.BitFlag)(12),
    manualLock: (0, import_BitmapSchema.BitFlag)(13),
    manualUnlock: (0, import_BitmapSchema.BitFlag)(14)
  };
  DoorLock2.RemoteProgrammingEventMask = {
    unknown: (0, import_BitmapSchema.BitFlag)(0),
    pinAdded: (0, import_BitmapSchema.BitFlag)(2),
    pinCleared: (0, import_BitmapSchema.BitFlag)(3),
    pinChanged: (0, import_BitmapSchema.BitFlag)(4),
    rfidCodeAdded: (0, import_BitmapSchema.BitFlag)(5),
    rfidCodeCleared: (0, import_BitmapSchema.BitFlag)(6)
  };
  DoorLock2.RfidOperationEventMask = {
    unknownOrManufacturerSpecificKeypadOperationEvent: (0, import_BitmapSchema.BitFlag)(0),
    lockSourceRfid: (0, import_BitmapSchema.BitFlag)(1),
    unlockSourceRfid: (0, import_BitmapSchema.BitFlag)(2),
    lockSourceRfidErrorInvalidRfidId: (0, import_BitmapSchema.BitFlag)(3),
    lockSourceRfidErrorInvalidSchedule: (0, import_BitmapSchema.BitFlag)(4),
    unlockSourceRfidErrorInvalidRfidId: (0, import_BitmapSchema.BitFlag)(5),
    unlockSourceRfidErrorInvalidSchedule: (0, import_BitmapSchema.BitFlag)(6)
  };
  DoorLock2.RfidProgrammingEventMask = { unknown: (0, import_BitmapSchema.BitFlag)(0), idAdded: (0, import_BitmapSchema.BitFlag)(5), idCleared: (0, import_BitmapSchema.BitFlag)(6) };
  let Feature;
  ((Feature2) => {
    Feature2["PinCredential"] = "PinCredential";
    Feature2["RfidCredential"] = "RfidCredential";
    Feature2["FingerCredentials"] = "FingerCredentials";
    Feature2["Logging"] = "Logging";
    Feature2["WeekDayAccessSchedules"] = "WeekDayAccessSchedules";
    Feature2["DoorPositionSensor"] = "DoorPositionSensor";
    Feature2["FaceCredentials"] = "FaceCredentials";
    Feature2["CredentialOverTheAirAccess"] = "CredentialOverTheAirAccess";
    Feature2["User"] = "User";
    Feature2["Notification"] = "Notification";
    Feature2["YearDayAccessSchedules"] = "YearDayAccessSchedules";
    Feature2["HolidaySchedules"] = "HolidaySchedules";
  })(Feature = DoorLock2.Feature || (DoorLock2.Feature = {}));
  DoorLock2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 257,
    name: "DoorLock",
    revision: 6,
    features: {
      /**
       * PinCredential
       *
       * Lock supports PIN credentials (via keypad, or over- the-air)
       */
      pinCredential: (0, import_BitmapSchema.BitFlag)(0),
      /**
       * RfidCredential
       *
       * Lock supports RFID credentials
       */
      rfidCredential: (0, import_BitmapSchema.BitFlag)(1),
      /**
       * FingerCredentials
       *
       * Lock supports finger related credentials (fingerprint, finger vein)
       */
      fingerCredentials: (0, import_BitmapSchema.BitFlag)(2),
      /**
       * Logging
       *
       * Lock supports local/on-lock logging when Events are not supported
       */
      logging: (0, import_BitmapSchema.BitFlag)(3),
      /**
       * WeekDayAccessSchedules
       *
       * Lock supports week day user access schedules
       */
      weekDayAccessSchedules: (0, import_BitmapSchema.BitFlag)(4),
      /**
       * DoorPositionSensor
       *
       * Lock supports a door position sensor that indicates door’s state
       */
      doorPositionSensor: (0, import_BitmapSchema.BitFlag)(5),
      /**
       * FaceCredentials
       *
       * Lock supports face related credentials (face, iris, retina)
       */
      faceCredentials: (0, import_BitmapSchema.BitFlag)(6),
      /**
       * CredentialOverTheAirAccess
       *
       * PIN codes over- the-air supported for lock/unlock operations
       */
      credentialOverTheAirAccess: (0, import_BitmapSchema.BitFlag)(7),
      /**
       * User
       *
       * Lock supports the user commands and database
       */
      user: (0, import_BitmapSchema.BitFlag)(8),
      /**
       * Notification
       *
       * Operation and Programming Notifications
       */
      notification: (0, import_BitmapSchema.BitFlag)(9),
      /**
       * YearDayAccessSchedules
       *
       * Lock supports year day user access schedules
       */
      yearDayAccessSchedules: (0, import_BitmapSchema.BitFlag)(10),
      /**
       * HolidaySchedules
       *
       * Lock supports holiday schedules
       */
      holidaySchedules: (0, import_BitmapSchema.BitFlag)(11)
    },
    attributes: {
      /**
       * The LockState Attribute may be NULL if the lock hardware does not currently know the status of the
       * locking mechanism. For example, a lock may not know the LockState status after a power cycle until the
       * first lock actuation is completed.
       *
       * The Not Fully Locked value is used by a lock to indicate that the state of the lock is somewhere between
       * Locked and Unlocked so it is only partially secured. For example, a deadbolt could be partially extended
       * and not in a dead latched state.
       *
       * If the Scenes server cluster is implemented on the same endpoint, the following extension field shall be
       * added to the Scene Table:
       *
       *   • LockState
       *
       * When the LockState attribute is part of a Scene table, the attribute is treated as a writable command;
       * that is, setting the LockState to lock will command the lock to lock, and setting the LockState to
       * unlock will command the lock to unlock. Setting the LockState attribute to “not fully locked” is not
       * supported. The Transition Time field in the Scene table will be treated as a delay before setting the
       * LockState attribute; that is, it is possible to activate a scene with the lock actuation some seconds
       * later.
       *
       * Locks that do not have an actuation mechanism SHOULD not support the Scene table extension.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.1
       */
      lockState: (0, import_Cluster.Attribute)(0, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)()), { scene: true }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.2
       */
      lockType: (0, import_Cluster.Attribute)(1, (0, import_TlvNumber.TlvEnum)()),
      /**
       * The ActuatorEnabled attribute indicates if the lock is currently able to (Enabled) or not able to
       * (Disabled) process remote Lock, Unlock, or Unlock with Timeout commands.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.3
       */
      actuatorEnabled: (0, import_Cluster.Attribute)(2, import_TlvBoolean.TlvBoolean),
      /**
       * Modifies the language for the on-screen or audible user interface using a 2-byte language code from
       * ISO-639-1.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.22
       */
      language: (0, import_Cluster.OptionalWritableAttribute)(
        33,
        import_TlvString.TlvString.bound({ maxLength: 3 }),
        { writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The settings for the LED support three different modes
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.25
       */
      ledSettings: (0, import_Cluster.OptionalWritableAttribute)(
        34,
        import_TlvNumber.TlvUInt8.bound({ max: 2 }),
        { default: 0, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The number of seconds to wait after unlocking a lock before it automatically locks again. 0=disabled. If
       * set, unlock operations from any source will be timed. For one time unlock with timeout use the specific
       * command.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.26
       */
      autoRelockTime: (0, import_Cluster.OptionalWritableAttribute)(35, import_TlvNumber.TlvUInt32, { writeAcl: import_Cluster.AccessLevel.Manage }),
      /**
       * The sound volume on a door lock has four possible settings: silent, low, high and medium volumes
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.27
       */
      soundVolume: (0, import_Cluster.OptionalWritableAttribute)(
        36,
        import_TlvNumber.TlvUInt8.bound({ max: 3 }),
        { default: 0, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The current operating mode of the lock (see OperatingModeEnum).
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.23
       */
      operatingMode: (0, import_Cluster.WritableAttribute)(
        37,
        (0, import_TlvNumber.TlvEnum)(),
        { default: 0 /* Normal */, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This bitmap contains all operating bits of the Operating Mode Attribute supported by the lock. All
       * operating modes NOT supported by a lock shall be set to one. The value of the OperatingMode enumeration
       * defines the related bit to be set, as
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.24
       */
      supportedOperatingModes: (0, import_Cluster.FixedAttribute)(
        38,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, DoorLock2.SupportedOperatingModes),
        { default: (0, import_BitmapSchema.BitsFromPartial)(DoorLock2.SupportedOperatingModes, { vacation: true, privacy: true, passage: true }) }
      ),
      /**
       * This attribute represents the default configurations as they are physically set on the device (example:
       * hardware dip switch setting, etc…) and represents the default setting for some of the attributes within
       * this cluster (for example: LED, Auto Lock, Sound Volume, and Operating Mode attributes).
       *
       * This is a read-only attribute and is intended to allow clients to determine what changes may need to be
       * made without having to query all the included attributes. It may be beneficial for the clients to know
       * what the device’s original settings were in the event that the device needs to be restored to factory
       * default settings.
       *
       * If the Client device would like to query and modify the door lock server’s operating settings, it SHOULD
       * send read and write attribute requests to the specific attributes.
       *
       * For example, the Sound Volume attribute default value is Silent Mode. However, it is possible that the
       * current Sound Volume is High Volume. Therefore, if the client wants to query/modify the current Sound
       * Volume setting on the server, the client SHOULD read/write to the Sound Volume attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.28
       */
      defaultConfigurationRegister: (0, import_Cluster.OptionalAttribute)(39, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, DoorLock2.DefaultConfigurationRegister)),
      /**
       * Enable/disable local programming on the door lock of certain features (see LocalProgrammingFeatures
       * attribute). If this value is set to TRUE then local programming is enabled on the door lock for all
       * features. If it is set to FALSE then local programming is disabled on the door lock for those features
       * whose bit is set to 0 in the LocalProgrammingFeatures attribute. Local programming shall be enabled by
       * default.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.29
       */
      enableLocalProgramming: (0, import_Cluster.OptionalWritableAttribute)(
        40,
        import_TlvBoolean.TlvBoolean,
        { default: true, writeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * Enable/disable the ability to lock the door lock with a single touch on the door lock.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.30
       */
      enableOneTouchLocking: (0, import_Cluster.OptionalWritableAttribute)(
        41,
        import_TlvBoolean.TlvBoolean,
        { default: true, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * Enable/disable an inside LED that allows the user to see at a glance if the door is locked.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.31
       */
      enableInsideStatusLed: (0, import_Cluster.OptionalWritableAttribute)(
        42,
        import_TlvBoolean.TlvBoolean,
        { default: true, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * Enable/disable a button inside the door that is used to put the lock into privacy mode. When the lock is
       * in privacy mode it cannot be manipulated from the outside.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.32
       */
      enablePrivacyModeButton: (0, import_Cluster.OptionalWritableAttribute)(
        43,
        import_TlvBoolean.TlvBoolean,
        { default: true, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The local programming features that will be disabled when EnableLocalProgramming attribute is set to
       * False. If a door lock doesn’t support disabling one aspect of local programming it shall return
       * CONSTRAINT_ERROR during a write operation of this attribute. If the EnableLocalProgramming attribute is
       * set to True then all local programming features shall be enabled regardless of the bits set to 0 in this
       * attribute.
       *
       * The features that can be disabled from local programming are defined in the following bitmap.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.33
       */
      localProgrammingFeatures: (0, import_Cluster.OptionalWritableAttribute)(
        44,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, DoorLock2.LocalProgrammingFeatures),
        { writeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * This attribute is only supported if the Alarms cluster is on the same endpoint. The alarm mask is used
       * to turn on/off alarms for particular functions. Alarms for an alarm group are enabled if the associated
       * alarm mask bit is set. Each bit represents a group of alarms. Entire alarm groups can be turned on or
       * off by setting or clearing the associated bit in the alarm mask.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.39
       */
      alarmMask: (0, import_Cluster.OptionalWritableAttribute)(
        64,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, DoorLock2.AlarmMask),
        {
          default: (0, import_BitmapSchema.BitsFromPartial)(DoorLock2.AlarmMask, { lockingMechanismJammed: true, lockResetToFactoryDefaults: true, reserved: true, rfModulePowerCycled: true, tamperAlarmWrongCodeEntryLimit: true, tamperAlarmFrontEscutcheonRemovedFromMain: true, forcedDoorOpenUnderDoorLockedCondition: true }),
          writeAcl: import_Cluster.AccessLevel.Administer
        }
      )
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      lockDoor: (0, import_Cluster.Command)(0, DoorLock2.TlvLockDoorRequest, 0, import_Cluster.TlvNoResponse, { timed: true }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      unlockDoor: (0, import_Cluster.Command)(1, DoorLock2.TlvUnlockDoorRequest, 1, import_Cluster.TlvNoResponse, { timed: true }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      unlockWithTimeout: (0, import_Cluster.OptionalCommand)(3, DoorLock2.TlvUnlockWithTimeoutRequest, 3, import_Cluster.TlvNoResponse, { timed: true })
    },
    events: {
      /**
       * The door lock cluster provides several alarms which can be sent when there is a critical state on the
       * door lock. The alarms available for the door lock cluster are listed in the AlarmCodeEnum section below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.1
       */
      doorLockAlarm: (0, import_Cluster.Event)(0, import_Cluster.EventPriority.Critical, DoorLock2.TlvDoorLockAlarmEvent),
      /**
       * The door lock server sends out a LockOperation event when the event is triggered by the various lock
       * operation sources.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3
       */
      lockOperation: (0, import_Cluster.Event)(2, import_Cluster.EventPriority.Critical, DoorLock2.TlvLockOperationEvent),
      /**
       * The door lock server sends out a LockOperationError event when a lock operation fails for various
       * reasons.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4
       */
      lockOperationError: (0, import_Cluster.Event)(3, import_Cluster.EventPriority.Critical, DoorLock2.TlvLockOperationErrorEvent)
    }
  });
  DoorLock2.DoorPositionSensorComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The current door state as defined in DoorStateEnum.
       *
       * This attribute shall be null only if an internal error prevents the retrieval of the current door state.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.4
       */
      doorState: (0, import_Cluster.Attribute)(3, (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)())),
      /**
       * This attribute holds the number of door open events that have occurred since it was last zeroed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.5
       */
      doorOpenEvents: (0, import_Cluster.OptionalWritableAttribute)(4, import_TlvNumber.TlvUInt32, { writeAcl: import_Cluster.AccessLevel.Manage }),
      /**
       * This attribute holds the number of door closed events that have occurred since it was last zeroed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.6
       */
      doorClosedEvents: (0, import_Cluster.OptionalWritableAttribute)(5, import_TlvNumber.TlvUInt32, { writeAcl: import_Cluster.AccessLevel.Manage }),
      /**
       * This attribute holds the number of minutes the door has been open since the last time it transitioned
       * from closed to open.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.7
       */
      openPeriod: (0, import_Cluster.OptionalWritableAttribute)(6, import_TlvNumber.TlvUInt16, { writeAcl: import_Cluster.AccessLevel.Manage })
    },
    events: {
      /**
       * The door lock server sends out a DoorStateChange event when the door lock door state changes.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.2
       */
      doorStateChange: (0, import_Cluster.Event)(1, import_Cluster.EventPriority.Critical, DoorLock2.TlvDoorStateChangeEvent)
    }
  });
  DoorLock2.LoggingComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The number of available log records.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.8
       */
      numberOfLogRecordsSupported: (0, import_Cluster.FixedAttribute)(16, import_TlvNumber.TlvUInt16, { default: 0 }),
      /**
       * Enable/disable event logging. When event logging is enabled, all event messages are stored on the lock
       * for retrieval. Logging events can be but not limited to Tamper Alarm, Lock, Unlock, AutoRelock, User
       * Code Added, User Code Cleared, Schedule Added, and Schedule Cleared. For a full detail of all the
       * possible alarms and events, please refer to the full list in the Alarm and Event Masks Attribute Set.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.21
       */
      enableLogging: (0, import_Cluster.WritableAttribute)(32, import_TlvBoolean.TlvBoolean, { default: true, writeAcl: import_Cluster.AccessLevel.Administer })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getLogRecord: (0, import_Cluster.Command)(4, import_TlvNoArguments.TlvNoArguments, 4, import_TlvNoArguments.TlvNoArguments, { invokeAcl: import_Cluster.AccessLevel.Manage })
    }
  });
  DoorLock2.UserComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * Number of total users supported by the lock.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.9
       */
      numberOfTotalUsersSupported: (0, import_Cluster.FixedAttribute)(17, import_TlvNumber.TlvUInt16, { default: 0 }),
      /**
       * This bitmap contains a bit for every value of CredentialRuleEnum supported on this device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.19
       */
      credentialRulesSupport: (0, import_Cluster.FixedAttribute)(
        27,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, DoorLock2.CredentialRulesSupport),
        { default: (0, import_BitmapSchema.BitsFromPartial)(DoorLock2.CredentialRulesSupport, { single: true }) }
      ),
      /**
       * The number of credentials that could be assigned for each user.
       *
       * Depending on the value of NumberOfRFIDUsersSupported and NumberOfPINUsersSupported it may not be
       * possible to assign that number of credentials for a user.
       *
       * For example, if the device supports only PIN and RFID credential types,
       * NumberOfCredentialsSupportedPerUser is set to 10, NumberOfPINUsersSupported is set to 5 and
       * NumberOfRFIDUsersSupported is set to 3, it will not be possible to actually assign 10 credentials for a
       * user because maximum number of credentials in the database is 8.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.20
       */
      numberOfCredentialsSupportedPerUser: (0, import_Cluster.FixedAttribute)(28, import_TlvNumber.TlvUInt8, { default: 0 }),
      /**
       * Number of minutes a PIN, RFID, Fingerprint, or other credential associated with a user of type
       * ExpiringUser shall remain valid after its first use before expiring. When the credential expires the
       * UserStatus for the corresponding user record shall be set to OccupiedDisabled.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.38
       */
      expiringUserTimeout: (0, import_Cluster.OptionalWritableAttribute)(
        53,
        import_TlvNumber.TlvUInt16.bound({ min: 1, max: 2880 }),
        { writeAcl: import_Cluster.AccessLevel.Administer }
      )
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setUser: (0, import_Cluster.Command)(
        26,
        DoorLock2.TlvSetUserRequest,
        26,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getUser: (0, import_Cluster.Command)(27, DoorLock2.TlvGetUserRequest, 28, DoorLock2.TlvGetUserResponse, { invokeAcl: import_Cluster.AccessLevel.Administer }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearUser: (0, import_Cluster.Command)(
        29,
        DoorLock2.TlvClearUserRequest,
        29,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setCredential: (0, import_Cluster.Command)(
        34,
        DoorLock2.TlvSetCredentialRequest,
        35,
        DoorLock2.TlvSetCredentialResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getCredentialStatus: (0, import_Cluster.Command)(
        36,
        DoorLock2.TlvGetCredentialStatusRequest,
        37,
        DoorLock2.TlvGetCredentialStatusResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearCredential: (0, import_Cluster.Command)(
        38,
        DoorLock2.TlvClearCredentialRequest,
        38,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer, timed: true }
      )
    },
    events: {
      /**
       * The door lock server sends out a LockUserChange event when a lock user, schedule, or credential change
       * has occurred.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5
       */
      lockUserChange: (0, import_Cluster.Event)(4, import_Cluster.EventPriority.Info, DoorLock2.TlvLockUserChangeEvent)
    }
  });
  DoorLock2.PinCredentialComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The number of PIN users supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.10
       */
      numberOfPinUsersSupported: (0, import_Cluster.FixedAttribute)(18, import_TlvNumber.TlvUInt16, { default: 0 }),
      /**
       * An 8 bit value indicates the maximum length in bytes of a PIN Code on this device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.15
       */
      maxPinCodeLength: (0, import_Cluster.FixedAttribute)(23, import_TlvNumber.TlvUInt8),
      /**
       * An 8 bit value indicates the minimum length in bytes of a PIN Code on this device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.16
       */
      minPinCodeLength: (0, import_Cluster.FixedAttribute)(24, import_TlvNumber.TlvUInt8),
      /**
       * Boolean set to True if it is ok for the door lock server to send PINs over the air. This attribute
       * determines the behavior of the server’s TX operation. If it is false, then it is not ok for the device
       * to send PIN in any messages over the air.
       *
       * The PIN field within any door lock cluster message shall keep the first octet unchanged and masks the
       * actual code by replacing with 0xFF. For example (PIN "1234" ): If the attribute value is True, 0x04 0x31
       * 0x32 0x33 0x34 shall be used in the PIN field in any door lock cluster message payload. If the attribute
       * value is False, 0x04 0xFF 0xFF 0xFF 0xFF shall be used.
       *
       * If the USR feature is supported by the device then this attribute shall NOT be supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.36
       */
      sendPinOverTheAir: (0, import_Cluster.OptionalWritableAttribute)(
        50,
        import_TlvBoolean.TlvBoolean,
        { default: true, writeAcl: import_Cluster.AccessLevel.Administer }
      )
    }
  });
  DoorLock2.RfidCredentialComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The number of RFID users supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.11
       */
      numberOfRfidUsersSupported: (0, import_Cluster.FixedAttribute)(19, import_TlvNumber.TlvUInt16, { default: 0 }),
      /**
       * An 8 bit value indicates the maximum length in bytes of a RFID Code on this device. The value depends on
       * the RFID code range specified by the manufacturer, if media anti-collision identifiers (UID) are used as
       * RFID code, a value of 20 (equals 10 Byte ISO 14443A UID) is recommended.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.17
       */
      maxRfidCodeLength: (0, import_Cluster.FixedAttribute)(25, import_TlvNumber.TlvUInt8),
      /**
       * An 8 bit value indicates the minimum length in bytes of a RFID Code on this device. The value depends on
       * the RFID code range specified by the manufacturer, if media anti-collision identifiers (UID) are used as
       * RFID code, a value of 8 (equals 4 Byte ISO 14443A UID) is recommended.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.18
       */
      minRfidCodeLength: (0, import_Cluster.FixedAttribute)(26, import_TlvNumber.TlvUInt8)
    }
  });
  DoorLock2.WeekDayAccessSchedulesComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The number of configurable week day schedule supported per user.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.12
       */
      numberOfWeekDaySchedulesSupportedPerUser: (0, import_Cluster.FixedAttribute)(20, import_TlvNumber.TlvUInt8, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setWeekDaySchedule: (0, import_Cluster.Command)(
        11,
        DoorLock2.TlvSetWeekDayScheduleRequest,
        11,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getWeekDaySchedule: (0, import_Cluster.Command)(
        12,
        DoorLock2.TlvGetWeekDayScheduleRequest,
        12,
        DoorLock2.TlvGetWeekDayScheduleResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearWeekDaySchedule: (0, import_Cluster.Command)(
        13,
        DoorLock2.TlvClearWeekDayScheduleRequest,
        13,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      )
    }
  });
  DoorLock2.YearDayAccessSchedulesComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The number of configurable year day schedule supported per user.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.13
       */
      numberOfYearDaySchedulesSupportedPerUser: (0, import_Cluster.FixedAttribute)(21, import_TlvNumber.TlvUInt8, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setYearDaySchedule: (0, import_Cluster.Command)(
        14,
        DoorLock2.TlvSetYearDayScheduleRequest,
        14,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getYearDaySchedule: (0, import_Cluster.Command)(
        15,
        DoorLock2.TlvGetYearDayScheduleRequest,
        15,
        DoorLock2.TlvGetYearDayScheduleResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearYearDaySchedule: (0, import_Cluster.Command)(
        16,
        DoorLock2.TlvClearYearDayScheduleRequest,
        16,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      )
    }
  });
  DoorLock2.HolidaySchedulesComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The number of holiday schedules supported for the entire door lock device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.14
       */
      numberOfHolidaySchedulesSupported: (0, import_Cluster.FixedAttribute)(22, import_TlvNumber.TlvUInt8, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setHolidaySchedule: (0, import_Cluster.Command)(
        17,
        DoorLock2.TlvSetHolidayScheduleRequest,
        17,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getHolidaySchedule: (0, import_Cluster.Command)(
        18,
        DoorLock2.TlvGetHolidayScheduleRequest,
        18,
        DoorLock2.TlvGetHolidayScheduleResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearHolidaySchedule: (0, import_Cluster.Command)(
        19,
        DoorLock2.TlvClearHolidayScheduleRequest,
        19,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      )
    }
  });
  DoorLock2.PinCredentialOrRfidCredentialComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The number of incorrect Pin codes or RFID presentment attempts a user is allowed to enter before the
       * lock will enter a lockout state. The value of this attribute is compared to all failing forms of
       * credential presentation, including Pin codes used in an Unlock Command when RequirePINforRemoteOperation
       * is set to true. Valid range is 1-255 incorrect attempts. The lockout state will be for the duration of
       * UserCodeTemporaryDisableTime. If the attribute accepts writes and an attempt to write the value 0 is
       * made, the device shall respond with CONSTRAINT_ERROR.
       *
       * The lock may reset the counter used to track incorrect credential presentations as required by internal
       * logic, environmental events, or other reasons. The lock shall reset the counter if a valid credential is
       * presented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.34
       */
      wrongCodeEntryLimit: (0, import_Cluster.WritableAttribute)(
        48,
        import_TlvNumber.TlvUInt8.bound({ min: 1 }),
        { writeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * The number of seconds that the lock shuts down following wrong code entry. Valid range is 1-255 seconds.
       * Device can shut down to lock user out for specified amount of time. (Makes it difficult to try and guess
       * a PIN for the device.) If the attribute accepts writes and an attempt to write the attribute to 0 is
       * made, the device shall respond with CONSTRAINT_ERROR.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.35
       */
      userCodeTemporaryDisableTime: (0, import_Cluster.WritableAttribute)(
        49,
        import_TlvNumber.TlvUInt8.bound({ min: 1 }),
        { writeAcl: import_Cluster.AccessLevel.Administer }
      )
    }
  });
  DoorLock2.CredentialOverTheAirAccessAndPinCredentialComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * Boolean set to True if the door lock server requires that an optional PINs be included in the payload of
       * remote lock operation events like Lock, Unlock, Unlock with Timeout and Toggle in order to function.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.37
       */
      requirePinForRemoteOperation: (0, import_Cluster.WritableAttribute)(
        51,
        import_TlvBoolean.TlvBoolean,
        { default: true, writeAcl: import_Cluster.AccessLevel.Administer }
      )
    }
  });
  DoorLock2.NotificationAndPinCredentialComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * Event mask used to turn on and off the transmission of keypad operation events. This mask DOES NOT apply
       * to the storing of events in the event log. This mask only applies to the Operation Event Notification
       * Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.40
       */
      keypadOperationEventMask: (0, import_Cluster.OptionalWritableAttribute)(
        65,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, DoorLock2.KeypadOperationEventMask),
        {
          default: (0, import_BitmapSchema.BitsFromPartial)(DoorLock2.KeypadOperationEventMask, { unknownOrManufacturerSpecificKeypadOperationEvent: true, lockSourceKeypad: true, unlockSourceKeypad: true, lockSourceKeypadErrorInvalidPin: true, lockSourceKeypadErrorInvalidSchedule: true, unlockSourceKeypadErrorInvalidCode: true, unlockSourceKeypadErrorInvalidSchedule: true, nonAccessUserOperationEventSourceKeypad: true }),
          writeAcl: import_Cluster.AccessLevel.Administer
        }
      ),
      /**
       * Event mask used to turn on and off keypad programming events. This mask DOES NOT apply to the storing of
       * events in the event log. This mask only applies to the Programming Event Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.44
       */
      keypadProgrammingEventMask: (0, import_Cluster.OptionalWritableAttribute)(
        69,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, DoorLock2.KeypadProgrammingEventMask),
        {
          default: (0, import_BitmapSchema.BitsFromPartial)(DoorLock2.KeypadProgrammingEventMask, { unknown: true, pinCodeChanged: true, pinAdded: true, pinCleared: true, pinChanged: true }),
          writeAcl: import_Cluster.AccessLevel.Administer
        }
      )
    }
  });
  DoorLock2.NotificationComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * Event mask used to turn on and off the transmission of remote operation events. This mask DOES NOT apply
       * to the storing of events in the event log. This mask only applies to the Operation Event
       *
       * Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.41
       */
      remoteOperationEventMask: (0, import_Cluster.OptionalWritableAttribute)(
        66,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, DoorLock2.RemoteOperationEventMask),
        {
          default: (0, import_BitmapSchema.BitsFromPartial)(DoorLock2.RemoteOperationEventMask, { unknownOrManufacturerSpecificRemoteOperationEvent: true, lockSourceRemote: true, unlockSourceRemote: true, lockSourceRemoteErrorInvalidCode: true, lockSourceRemoteErrorInvalidSchedule: true, unlockSourceRemoteErrorInvalidCode: true, unlockSourceRemoteErrorInvalidSchedule: true }),
          writeAcl: import_Cluster.AccessLevel.Administer
        }
      ),
      /**
       * Event mask used to turn on and off manual operation events. This mask DOES NOT apply to the storing of
       * events in the event log. This mask only applies to the Operation Event Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.42
       */
      manualOperationEventMask: (0, import_Cluster.OptionalWritableAttribute)(
        67,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, DoorLock2.ManualOperationEventMask),
        {
          default: (0, import_BitmapSchema.BitsFromPartial)(DoorLock2.ManualOperationEventMask, { unknownOrManufacturerSpecificManualOperationEvent: true, thumbturnLock: true, thumbturnUnlock: true, oneTouchLock: true, keyLock: true, keyUnlock: true, autoLock: true, scheduleLock: true, scheduleUnlock: true, manualLock: true, manualUnlock: true }),
          writeAcl: import_Cluster.AccessLevel.Administer
        }
      ),
      /**
       * Event mask used to turn on and off remote programming events. This mask DOES NOT apply to the storing of
       * events in the event log. This mask only applies to the Programming Event Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.45
       */
      remoteProgrammingEventMask: (0, import_Cluster.OptionalWritableAttribute)(
        70,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, DoorLock2.RemoteProgrammingEventMask),
        {
          default: (0, import_BitmapSchema.BitsFromPartial)(DoorLock2.RemoteProgrammingEventMask, { unknown: true, pinAdded: true, pinCleared: true, pinChanged: true, rfidCodeAdded: true, rfidCodeCleared: true }),
          writeAcl: import_Cluster.AccessLevel.Administer
        }
      )
    }
  });
  DoorLock2.NotificationAndRfidCredentialComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * Event mask used to turn on and off RFID operation events. This mask DOES NOT apply to the storing of
       * events in the event log. This mask only applies to the Operation Event Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.43
       */
      rfidOperationEventMask: (0, import_Cluster.OptionalWritableAttribute)(
        68,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, DoorLock2.RfidOperationEventMask),
        {
          default: (0, import_BitmapSchema.BitsFromPartial)(DoorLock2.RfidOperationEventMask, { unknownOrManufacturerSpecificKeypadOperationEvent: true, lockSourceRfid: true, unlockSourceRfid: true, lockSourceRfidErrorInvalidRfidId: true, lockSourceRfidErrorInvalidSchedule: true, unlockSourceRfidErrorInvalidRfidId: true, unlockSourceRfidErrorInvalidSchedule: true }),
          writeAcl: import_Cluster.AccessLevel.Administer
        }
      ),
      /**
       * Event mask used to turn on and off RFID programming events. This mask DOES NOT apply to the storing of
       * events in the event log. This mask only applies to the Programming Event Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.46
       */
      rfidProgrammingEventMask: (0, import_Cluster.OptionalWritableAttribute)(
        71,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, DoorLock2.RfidProgrammingEventMask),
        {
          default: (0, import_BitmapSchema.BitsFromPartial)(DoorLock2.RfidProgrammingEventMask, { unknown: true, idAdded: true, idCleared: true }),
          writeAcl: import_Cluster.AccessLevel.Administer
        }
      )
    }
  });
  DoorLock2.PinCredentialNotUserComponent = import_ClusterFactory.ClusterFactory.Component({
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setPinCode: (0, import_Cluster.Command)(
        5,
        import_TlvNoArguments.TlvNoArguments,
        5,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getPinCode: (0, import_Cluster.Command)(6, import_TlvNoArguments.TlvNoArguments, 6, import_TlvNoArguments.TlvNoArguments, { invokeAcl: import_Cluster.AccessLevel.Administer }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearPinCode: (0, import_Cluster.Command)(
        7,
        import_TlvNoArguments.TlvNoArguments,
        7,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearAllPinCodes: (0, import_Cluster.Command)(
        8,
        import_TlvNoArguments.TlvNoArguments,
        8,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer, timed: true }
      )
    }
  });
  DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent = import_ClusterFactory.ClusterFactory.Component({
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setUserStatus: (0, import_Cluster.OptionalCommand)(
        9,
        import_TlvNoArguments.TlvNoArguments,
        9,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getUserStatus: (0, import_Cluster.OptionalCommand)(
        10,
        import_TlvNoArguments.TlvNoArguments,
        10,
        import_TlvNoArguments.TlvNoArguments,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setUserType: (0, import_Cluster.OptionalCommand)(
        20,
        import_TlvNoArguments.TlvNoArguments,
        20,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getUserType: (0, import_Cluster.OptionalCommand)(
        21,
        import_TlvNoArguments.TlvNoArguments,
        21,
        import_TlvNoArguments.TlvNoArguments,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      )
    }
  });
  DoorLock2.NotUserComponent = import_ClusterFactory.ClusterFactory.Component({});
  DoorLock2.RfidCredentialNotUserComponent = import_ClusterFactory.ClusterFactory.Component({
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setRfidCode: (0, import_Cluster.Command)(
        22,
        import_TlvNoArguments.TlvNoArguments,
        22,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getRfidCode: (0, import_Cluster.Command)(23, import_TlvNoArguments.TlvNoArguments, 23, import_TlvNoArguments.TlvNoArguments, { invokeAcl: import_Cluster.AccessLevel.Administer }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearRfidCode: (0, import_Cluster.Command)(
        24,
        import_TlvNoArguments.TlvNoArguments,
        24,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearAllRfidCodes: (0, import_Cluster.Command)(
        25,
        import_TlvNoArguments.TlvNoArguments,
        25,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer, timed: true }
      )
    }
  });
  DoorLock2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    DoorLock2.Base,
    /**
     * Use this factory method to create a DoorLock cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a DoorLock cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...DoorLock2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(DoorLock2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.DoorPositionSensorComponent, { doorPositionSensor: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.LoggingComponent, { logging: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.UserComponent, { user: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.PinCredentialComponent, { pinCredential: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.RfidCredentialComponent, { rfidCredential: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.WeekDayAccessSchedulesComponent, { weekDayAccessSchedules: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.YearDayAccessSchedulesComponent, { yearDayAccessSchedules: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.HolidaySchedulesComponent, { holidaySchedules: true });
      import_ClusterFactory.ClusterFactory.extend(
        cluster,
        DoorLock2.PinCredentialOrRfidCredentialComponent,
        { pinCredential: true },
        { rfidCredential: true }
      );
      import_ClusterFactory.ClusterFactory.extend(
        cluster,
        DoorLock2.CredentialOverTheAirAccessAndPinCredentialComponent,
        { credentialOverTheAirAccess: true, pinCredential: true }
      );
      import_ClusterFactory.ClusterFactory.extend(
        cluster,
        DoorLock2.NotificationAndPinCredentialComponent,
        { notification: true, pinCredential: true }
      );
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.NotificationComponent, { notification: true });
      import_ClusterFactory.ClusterFactory.extend(
        cluster,
        DoorLock2.NotificationAndRfidCredentialComponent,
        { notification: true, rfidCredential: true }
      );
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.PinCredentialNotUserComponent, { pinCredential: true, user: false });
      import_ClusterFactory.ClusterFactory.extend(
        cluster,
        DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent,
        { pinCredential: true, rfidCredential: true, fingerCredentials: true, user: false }
      );
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.NotUserComponent, { user: false });
      import_ClusterFactory.ClusterFactory.extend(cluster, DoorLock2.RfidCredentialNotUserComponent, { rfidCredential: true, user: false });
      import_ClusterFactory.ClusterFactory.prevent(
        cluster,
        {
          user: true,
          pinCredential: false,
          rfidCredential: false,
          fingerCredentials: false,
          faceCredentials: false
        }
      );
      return cluster;
    }
  );
  const DPS = { doorPositionSensor: true };
  const LOG = { logging: true };
  const USR = { user: true };
  const PIN = { pinCredential: true };
  const RID = { rfidCredential: true };
  const WDSCH = { weekDayAccessSchedules: true };
  const YDSCH = { yearDayAccessSchedules: true };
  const HDSCH = { holidaySchedules: true };
  const COTA_PIN = { credentialOverTheAirAccess: true, pinCredential: true };
  const NOT_PIN = { notification: true, pinCredential: true };
  const NOT = { notification: true };
  const NOT_RID = { notification: true, rfidCredential: true };
  const PIN_NOT_USR = { pinCredential: true, user: false };
  const PIN_RID_FGP_NOT_USR = { pinCredential: true, rfidCredential: true, fingerCredentials: true, user: false };
  const RID_NOT_USR = { rfidCredential: true, user: false };
  DoorLock2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: DoorLock2.Cluster.id,
    name: DoorLock2.Cluster.name,
    revision: DoorLock2.Cluster.revision,
    features: DoorLock2.Cluster.features,
    attributes: {
      ...DoorLock2.Cluster.attributes,
      doorState: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.DoorPositionSensorComponent.attributes.doorState,
        { mandatoryIf: [DPS] }
      ),
      doorOpenEvents: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.DoorPositionSensorComponent.attributes.doorOpenEvents,
        { optionalIf: [DPS] }
      ),
      doorClosedEvents: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.DoorPositionSensorComponent.attributes.doorClosedEvents,
        { optionalIf: [DPS] }
      ),
      openPeriod: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.DoorPositionSensorComponent.attributes.openPeriod,
        { optionalIf: [DPS] }
      ),
      numberOfLogRecordsSupported: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.LoggingComponent.attributes.numberOfLogRecordsSupported,
        { mandatoryIf: [LOG] }
      ),
      numberOfTotalUsersSupported: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.UserComponent.attributes.numberOfTotalUsersSupported,
        { mandatoryIf: [USR] }
      ),
      numberOfPinUsersSupported: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialComponent.attributes.numberOfPinUsersSupported,
        { mandatoryIf: [PIN] }
      ),
      numberOfRfidUsersSupported: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialComponent.attributes.numberOfRfidUsersSupported,
        { mandatoryIf: [RID] }
      ),
      numberOfWeekDaySchedulesSupportedPerUser: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.WeekDayAccessSchedulesComponent.attributes.numberOfWeekDaySchedulesSupportedPerUser,
        { mandatoryIf: [WDSCH] }
      ),
      numberOfYearDaySchedulesSupportedPerUser: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.YearDayAccessSchedulesComponent.attributes.numberOfYearDaySchedulesSupportedPerUser,
        { mandatoryIf: [YDSCH] }
      ),
      numberOfHolidaySchedulesSupported: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.HolidaySchedulesComponent.attributes.numberOfHolidaySchedulesSupported,
        { mandatoryIf: [HDSCH] }
      ),
      maxPinCodeLength: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialComponent.attributes.maxPinCodeLength,
        { mandatoryIf: [PIN] }
      ),
      minPinCodeLength: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialComponent.attributes.minPinCodeLength,
        { mandatoryIf: [PIN] }
      ),
      maxRfidCodeLength: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialComponent.attributes.maxRfidCodeLength,
        { mandatoryIf: [RID] }
      ),
      minRfidCodeLength: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialComponent.attributes.minRfidCodeLength,
        { mandatoryIf: [RID] }
      ),
      credentialRulesSupport: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.UserComponent.attributes.credentialRulesSupport,
        { mandatoryIf: [USR] }
      ),
      numberOfCredentialsSupportedPerUser: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.UserComponent.attributes.numberOfCredentialsSupportedPerUser,
        { mandatoryIf: [USR] }
      ),
      enableLogging: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.LoggingComponent.attributes.enableLogging,
        { mandatoryIf: [LOG] }
      ),
      wrongCodeEntryLimit: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialOrRfidCredentialComponent.attributes.wrongCodeEntryLimit,
        { mandatoryIf: [PIN, RID] }
      ),
      userCodeTemporaryDisableTime: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialOrRfidCredentialComponent.attributes.userCodeTemporaryDisableTime,
        { mandatoryIf: [PIN, RID] }
      ),
      sendPinOverTheAir: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialComponent.attributes.sendPinOverTheAir,
        { optionalIf: [PIN] }
      ),
      requirePinForRemoteOperation: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.CredentialOverTheAirAccessAndPinCredentialComponent.attributes.requirePinForRemoteOperation,
        { mandatoryIf: [COTA_PIN] }
      ),
      expiringUserTimeout: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.UserComponent.attributes.expiringUserTimeout,
        { optionalIf: [USR] }
      ),
      keypadOperationEventMask: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.NotificationAndPinCredentialComponent.attributes.keypadOperationEventMask,
        { optionalIf: [NOT_PIN] }
      ),
      remoteOperationEventMask: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.NotificationComponent.attributes.remoteOperationEventMask,
        { optionalIf: [NOT] }
      ),
      manualOperationEventMask: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.NotificationComponent.attributes.manualOperationEventMask,
        { optionalIf: [NOT] }
      ),
      rfidOperationEventMask: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.NotificationAndRfidCredentialComponent.attributes.rfidOperationEventMask,
        { optionalIf: [NOT_RID] }
      ),
      keypadProgrammingEventMask: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.NotificationAndPinCredentialComponent.attributes.keypadProgrammingEventMask,
        { optionalIf: [NOT_PIN] }
      ),
      remoteProgrammingEventMask: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.NotificationComponent.attributes.remoteProgrammingEventMask,
        { optionalIf: [NOT] }
      ),
      rfidProgrammingEventMask: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.NotificationAndRfidCredentialComponent.attributes.rfidProgrammingEventMask,
        { optionalIf: [NOT_RID] }
      )
    },
    commands: {
      ...DoorLock2.Cluster.commands,
      getLogRecord: import_ClusterFactory.ClusterFactory.AsConditional(DoorLock2.LoggingComponent.commands.getLogRecord, { mandatoryIf: [LOG] }),
      setPinCode: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialNotUserComponent.commands.setPinCode,
        { mandatoryIf: [PIN_NOT_USR] }
      ),
      getPinCode: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialNotUserComponent.commands.getPinCode,
        { mandatoryIf: [PIN_NOT_USR] }
      ),
      clearPinCode: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialNotUserComponent.commands.clearPinCode,
        { mandatoryIf: [PIN_NOT_USR] }
      ),
      clearAllPinCodes: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialNotUserComponent.commands.clearAllPinCodes,
        { mandatoryIf: [PIN_NOT_USR] }
      ),
      setUserStatus: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent.commands.setUserStatus,
        { optionalIf: [PIN_RID_FGP_NOT_USR] }
      ),
      getUserStatus: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent.commands.getUserStatus,
        { optionalIf: [PIN_RID_FGP_NOT_USR] }
      ),
      setWeekDaySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.WeekDayAccessSchedulesComponent.commands.setWeekDaySchedule,
        { mandatoryIf: [WDSCH] }
      ),
      getWeekDaySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.WeekDayAccessSchedulesComponent.commands.getWeekDaySchedule,
        { mandatoryIf: [WDSCH] }
      ),
      clearWeekDaySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.WeekDayAccessSchedulesComponent.commands.clearWeekDaySchedule,
        { mandatoryIf: [WDSCH] }
      ),
      setYearDaySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.YearDayAccessSchedulesComponent.commands.setYearDaySchedule,
        { mandatoryIf: [YDSCH] }
      ),
      getYearDaySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.YearDayAccessSchedulesComponent.commands.getYearDaySchedule,
        { mandatoryIf: [YDSCH] }
      ),
      clearYearDaySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.YearDayAccessSchedulesComponent.commands.clearYearDaySchedule,
        { mandatoryIf: [YDSCH] }
      ),
      setHolidaySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.HolidaySchedulesComponent.commands.setHolidaySchedule,
        { mandatoryIf: [HDSCH] }
      ),
      getHolidaySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.HolidaySchedulesComponent.commands.getHolidaySchedule,
        { mandatoryIf: [HDSCH] }
      ),
      clearHolidaySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.HolidaySchedulesComponent.commands.clearHolidaySchedule,
        { mandatoryIf: [HDSCH] }
      ),
      setUserType: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent.commands.setUserType,
        { optionalIf: [PIN_RID_FGP_NOT_USR] }
      ),
      getUserType: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent.commands.getUserType,
        { optionalIf: [PIN_RID_FGP_NOT_USR] }
      ),
      setRfidCode: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialNotUserComponent.commands.setRfidCode,
        { mandatoryIf: [RID_NOT_USR] }
      ),
      getRfidCode: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialNotUserComponent.commands.getRfidCode,
        { mandatoryIf: [RID_NOT_USR] }
      ),
      clearRfidCode: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialNotUserComponent.commands.clearRfidCode,
        { mandatoryIf: [RID_NOT_USR] }
      ),
      clearAllRfidCodes: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialNotUserComponent.commands.clearAllRfidCodes,
        { mandatoryIf: [RID_NOT_USR] }
      ),
      setUser: import_ClusterFactory.ClusterFactory.AsConditional(DoorLock2.UserComponent.commands.setUser, { mandatoryIf: [USR] }),
      getUser: import_ClusterFactory.ClusterFactory.AsConditional(DoorLock2.UserComponent.commands.getUser, { mandatoryIf: [USR] }),
      clearUser: import_ClusterFactory.ClusterFactory.AsConditional(DoorLock2.UserComponent.commands.clearUser, { mandatoryIf: [USR] }),
      setCredential: import_ClusterFactory.ClusterFactory.AsConditional(DoorLock2.UserComponent.commands.setCredential, { mandatoryIf: [USR] }),
      getCredentialStatus: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.UserComponent.commands.getCredentialStatus,
        { mandatoryIf: [USR] }
      ),
      clearCredential: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.UserComponent.commands.clearCredential,
        { mandatoryIf: [USR] }
      )
    },
    events: {
      ...DoorLock2.Cluster.events,
      doorStateChange: import_ClusterFactory.ClusterFactory.AsConditional(
        DoorLock2.DoorPositionSensorComponent.events.doorStateChange,
        { mandatoryIf: [DPS] }
      ),
      lockUserChange: import_ClusterFactory.ClusterFactory.AsConditional(DoorLock2.UserComponent.events.lockUserChange, { mandatoryIf: [USR] })
    }
  });
})(DoorLock || (DoorLock = {}));
const DoorLockCluster = DoorLock.Cluster;
