// @project-chip/matter.js/dist/cjs/cluster/definitions/ElectricalMeasurementCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ElectricalMeasurementCluster_exports = {};
__export(ElectricalMeasurementCluster_exports, {
  ElectricalMeasurement: () => ElectricalMeasurement,
  ElectricalMeasurementCluster: () => ElectricalMeasurementCluster
});
module.exports = __toCommonJS(ElectricalMeasurementCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ElectricalMeasurement;
((ElectricalMeasurement2) => {
  ElectricalMeasurement2.TlvGetMeasurementProfileCommandRequest = (0, import_TlvObject.TlvObject)({
    attributeId: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16),
    startTime: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt32),
    numberOfIntervals: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8)
  });
  ElectricalMeasurement2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 2820,
    name: "ElectricalMeasurement",
    revision: 1,
    attributes: {
      measurementType: (0, import_Cluster.OptionalAttribute)(0, import_TlvNumber.TlvUInt32),
      dcVoltage: (0, import_Cluster.OptionalAttribute)(256, import_TlvNumber.TlvInt16, { default: 32768 }),
      dcVoltageMin: (0, import_Cluster.OptionalAttribute)(257, import_TlvNumber.TlvInt16, { default: 32768 }),
      dcVoltageMax: (0, import_Cluster.OptionalAttribute)(258, import_TlvNumber.TlvInt16, { default: 32768 }),
      dcCurrent: (0, import_Cluster.OptionalAttribute)(259, import_TlvNumber.TlvInt16, { default: 32768 }),
      dcCurrentMin: (0, import_Cluster.OptionalAttribute)(260, import_TlvNumber.TlvInt16, { default: 32768 }),
      dcCurrentMax: (0, import_Cluster.OptionalAttribute)(261, import_TlvNumber.TlvInt16, { default: 32768 }),
      dcPower: (0, import_Cluster.OptionalAttribute)(262, import_TlvNumber.TlvInt16, { default: 32768 }),
      dcPowerMin: (0, import_Cluster.OptionalAttribute)(263, import_TlvNumber.TlvInt16, { default: 32768 }),
      dcPowerMax: (0, import_Cluster.OptionalAttribute)(264, import_TlvNumber.TlvInt16, { default: 32768 }),
      dcVoltageMultiplier: (0, import_Cluster.OptionalAttribute)(512, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      dcVoltageDivisor: (0, import_Cluster.OptionalAttribute)(513, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      dcCurrentMultiplier: (0, import_Cluster.OptionalAttribute)(514, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      dcCurrentDivisor: (0, import_Cluster.OptionalAttribute)(515, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      dcPowerMultiplier: (0, import_Cluster.OptionalAttribute)(516, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      dcPowerDivisor: (0, import_Cluster.OptionalAttribute)(517, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      acFrequency: (0, import_Cluster.OptionalAttribute)(768, import_TlvNumber.TlvUInt16, { default: 65535 }),
      acFrequencyMin: (0, import_Cluster.OptionalAttribute)(769, import_TlvNumber.TlvUInt16, { default: 65535 }),
      acFrequencyMax: (0, import_Cluster.OptionalAttribute)(770, import_TlvNumber.TlvUInt16, { default: 65535 }),
      neutralCurrent: (0, import_Cluster.OptionalAttribute)(771, import_TlvNumber.TlvUInt16, { default: 0 }),
      totalActivePower: (0, import_Cluster.OptionalAttribute)(772, import_TlvNumber.TlvInt32, { default: 0 }),
      totalReactivePower: (0, import_Cluster.OptionalAttribute)(773, import_TlvNumber.TlvInt32, { default: 0 }),
      totalApparentPower: (0, import_Cluster.OptionalAttribute)(774, import_TlvNumber.TlvUInt32.bound({ max: 16777215 }), { default: 1 }),
      measured1StHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(775, import_TlvNumber.TlvInt16, { default: 32768 }),
      measured3RdHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(776, import_TlvNumber.TlvInt16, { default: 32768 }),
      measured5ThHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(777, import_TlvNumber.TlvInt16, { default: 32768 }),
      measured7ThHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(778, import_TlvNumber.TlvInt16, { default: 32768 }),
      measured9ThHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(779, import_TlvNumber.TlvInt16, { default: 32768 }),
      measured11ThHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(780, import_TlvNumber.TlvInt16, { default: 32768 }),
      measuredPhase1StHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(781, import_TlvNumber.TlvInt16, { default: 32768 }),
      measuredPhase3RdHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(782, import_TlvNumber.TlvInt16, { default: 32768 }),
      measuredPhase5ThHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(783, import_TlvNumber.TlvInt16, { default: 32768 }),
      measuredPhase7ThHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(784, import_TlvNumber.TlvInt16, { default: 32768 }),
      measuredPhase9ThHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(785, import_TlvNumber.TlvInt16, { default: 32768 }),
      measuredPhase11ThHarmonicCurrent: (0, import_Cluster.OptionalAttribute)(786, import_TlvNumber.TlvInt16, { default: 32768 }),
      acFrequencyMultiplier: (0, import_Cluster.OptionalAttribute)(1024, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      acFrequencyDivisor: (0, import_Cluster.OptionalAttribute)(1025, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      powerMultiplier: (0, import_Cluster.OptionalAttribute)(1026, import_TlvNumber.TlvUInt32.bound({ max: 16777215 }), { default: 1 }),
      powerDivisor: (0, import_Cluster.OptionalAttribute)(1027, import_TlvNumber.TlvUInt32.bound({ max: 16777215 }), { default: 1 }),
      harmonicCurrentMultiplier: (0, import_Cluster.OptionalAttribute)(1028, import_TlvNumber.TlvInt8.bound({ min: -127 }), { default: 0 }),
      phaseHarmonicCurrentMultiplier: (0, import_Cluster.OptionalAttribute)(1029, import_TlvNumber.TlvInt8.bound({ min: -127 }), { default: 0 }),
      instantaneousVoltage: (0, import_Cluster.OptionalAttribute)(1280, import_TlvNumber.TlvInt16, { default: 65535 }),
      instantaneousLineCurrent: (0, import_Cluster.OptionalAttribute)(1281, import_TlvNumber.TlvUInt16, { default: 65535 }),
      instantaneousActiveCurrent: (0, import_Cluster.OptionalAttribute)(1282, import_TlvNumber.TlvInt16, { default: 65535 }),
      instantaneousReactiveCurrent: (0, import_Cluster.OptionalAttribute)(1283, import_TlvNumber.TlvInt16, { default: 65535 }),
      instantaneousPower: (0, import_Cluster.OptionalAttribute)(1284, import_TlvNumber.TlvInt16, { default: 65535 }),
      rmsVoltage: (0, import_Cluster.OptionalAttribute)(1285, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsVoltageMin: (0, import_Cluster.OptionalAttribute)(1286, import_TlvNumber.TlvUInt16, { default: 32768 }),
      rmsVoltageMax: (0, import_Cluster.OptionalAttribute)(1287, import_TlvNumber.TlvUInt16, { default: 32768 }),
      rmsCurrent: (0, import_Cluster.OptionalAttribute)(1288, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsCurrentMin: (0, import_Cluster.OptionalAttribute)(1289, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsCurrentMax: (0, import_Cluster.OptionalAttribute)(1290, import_TlvNumber.TlvUInt16, { default: 65535 }),
      activePower: (0, import_Cluster.OptionalAttribute)(1291, import_TlvNumber.TlvInt16, { default: 65535 }),
      activePowerMin: (0, import_Cluster.OptionalAttribute)(1292, import_TlvNumber.TlvInt16, { default: 65535 }),
      activePowerMax: (0, import_Cluster.OptionalAttribute)(1293, import_TlvNumber.TlvInt16, { default: 65535 }),
      reactivePower: (0, import_Cluster.OptionalAttribute)(1294, import_TlvNumber.TlvInt16, { default: 65535 }),
      apparentPower: (0, import_Cluster.OptionalAttribute)(1295, import_TlvNumber.TlvUInt16, { default: 65535 }),
      acPowerFactor: (0, import_Cluster.OptionalAttribute)(1296, import_TlvNumber.TlvInt8.bound({ min: -100, max: 100 }), { default: 0 }),
      averageRmsVoltageMeasurementPeriod: (0, import_Cluster.OptionalWritableAttribute)(1297, import_TlvNumber.TlvUInt16, { default: 0 }),
      averageRmsUnderVoltageCounter: (0, import_Cluster.OptionalWritableAttribute)(1299, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsExtremeOverVoltagePeriod: (0, import_Cluster.OptionalWritableAttribute)(1300, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsExtremeUnderVoltagePeriod: (0, import_Cluster.OptionalWritableAttribute)(1301, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsVoltageSagPeriod: (0, import_Cluster.OptionalWritableAttribute)(1302, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsVoltageSwellPeriod: (0, import_Cluster.OptionalWritableAttribute)(1303, import_TlvNumber.TlvUInt16, { default: 0 }),
      acVoltageMultiplier: (0, import_Cluster.OptionalAttribute)(1536, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      acVoltageDivisor: (0, import_Cluster.OptionalAttribute)(1537, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      acCurrentMultiplier: (0, import_Cluster.OptionalAttribute)(1538, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      acCurrentDivisor: (0, import_Cluster.OptionalAttribute)(1539, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      acPowerMultiplier: (0, import_Cluster.OptionalAttribute)(1540, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      acPowerDivisor: (0, import_Cluster.OptionalAttribute)(1541, import_TlvNumber.TlvUInt16.bound({ min: 1 }), { default: 1 }),
      dcOverloadAlarmsMask: (0, import_Cluster.OptionalWritableAttribute)(1792, import_TlvNumber.TlvUInt8),
      dcVoltageOverload: (0, import_Cluster.OptionalAttribute)(1793, import_TlvNumber.TlvInt16, { default: 65535 }),
      dcCurrentOverload: (0, import_Cluster.OptionalAttribute)(1794, import_TlvNumber.TlvInt16, { default: 65535 }),
      acOverloadAlarmsMask: (0, import_Cluster.OptionalWritableAttribute)(2048, import_TlvNumber.TlvUInt16),
      acVoltageOverload: (0, import_Cluster.OptionalAttribute)(2049, import_TlvNumber.TlvInt16, { default: 65535 }),
      acCurrentOverload: (0, import_Cluster.OptionalAttribute)(2050, import_TlvNumber.TlvInt16, { default: 65535 }),
      acPowerOverload: (0, import_Cluster.OptionalAttribute)(2051, import_TlvNumber.TlvInt16, { default: 65535 }),
      acReactivePowerOverload: (0, import_Cluster.OptionalAttribute)(2052, import_TlvNumber.TlvInt16, { default: 65535 }),
      averageRmsOverVoltage: (0, import_Cluster.OptionalAttribute)(2053, import_TlvNumber.TlvInt16, { default: 65535 }),
      averageRmsUnderVoltage: (0, import_Cluster.OptionalAttribute)(2054, import_TlvNumber.TlvInt16, { default: 65535 }),
      rmsExtremeOverVoltage: (0, import_Cluster.OptionalAttribute)(2055, import_TlvNumber.TlvInt16, { default: 65535 }),
      rmsExtremeUnderVoltage: (0, import_Cluster.OptionalAttribute)(2056, import_TlvNumber.TlvInt16, { default: 65535 }),
      rmsVoltageSag: (0, import_Cluster.OptionalAttribute)(2057, import_TlvNumber.TlvInt16, { default: 65535 }),
      rmsVoltageSwell: (0, import_Cluster.OptionalAttribute)(2058, import_TlvNumber.TlvInt16, { default: 65535 }),
      lineCurrentPhaseB: (0, import_Cluster.OptionalAttribute)(2305, import_TlvNumber.TlvUInt16, { default: 65535 }),
      activeCurrentPhaseB: (0, import_Cluster.OptionalAttribute)(2306, import_TlvNumber.TlvInt16, { default: 65535 }),
      reactiveCurrentPhaseB: (0, import_Cluster.OptionalAttribute)(2307, import_TlvNumber.TlvInt16, { default: 65535 }),
      rmsVoltagePhaseB: (0, import_Cluster.OptionalAttribute)(2309, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsVoltageMinPhaseB: (0, import_Cluster.OptionalAttribute)(2310, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsVoltageMaxPhaseB: (0, import_Cluster.OptionalAttribute)(2311, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsCurrentPhaseB: (0, import_Cluster.OptionalAttribute)(2312, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsCurrentMinPhaseB: (0, import_Cluster.OptionalAttribute)(2313, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsCurrentMaxPhaseB: (0, import_Cluster.OptionalAttribute)(2314, import_TlvNumber.TlvUInt16, { default: 65535 }),
      activePowerPhaseB: (0, import_Cluster.OptionalAttribute)(2315, import_TlvNumber.TlvInt16, { default: 65535 }),
      activePowerMinPhaseB: (0, import_Cluster.OptionalAttribute)(2316, import_TlvNumber.TlvInt16, { default: 65535 }),
      activePowerMaxPhaseB: (0, import_Cluster.OptionalAttribute)(2317, import_TlvNumber.TlvInt16, { default: 65535 }),
      reactivePowerPhaseB: (0, import_Cluster.OptionalAttribute)(2318, import_TlvNumber.TlvInt16, { default: 65535 }),
      apparentPowerPhaseB: (0, import_Cluster.OptionalAttribute)(2319, import_TlvNumber.TlvUInt16, { default: 65535 }),
      powerFactorPhaseB: (0, import_Cluster.OptionalAttribute)(2320, import_TlvNumber.TlvInt8.bound({ min: -100, max: 100 }), { default: 0 }),
      averageRmsVoltageMeasurementPeriodPhaseB: (0, import_Cluster.OptionalAttribute)(2321, import_TlvNumber.TlvUInt16, { default: 0 }),
      averageRmsOverVoltageCounterPhaseB: (0, import_Cluster.OptionalAttribute)(2322, import_TlvNumber.TlvUInt16, { default: 0 }),
      averageRmsUnderVoltageCounterPhaseB: (0, import_Cluster.OptionalAttribute)(2323, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsExtremeOverVoltagePeriodPhaseB: (0, import_Cluster.OptionalAttribute)(2324, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsExtremeUnderVoltagePeriodPhaseB: (0, import_Cluster.OptionalAttribute)(2325, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsVoltageSagPeriodPhaseB: (0, import_Cluster.OptionalAttribute)(2326, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsVoltageSwellPeriodPhaseB: (0, import_Cluster.OptionalAttribute)(2327, import_TlvNumber.TlvUInt16, { default: 0 }),
      lineCurrentPhaseC: (0, import_Cluster.OptionalAttribute)(2561, import_TlvNumber.TlvUInt16, { default: 65535 }),
      activeCurrentPhaseC: (0, import_Cluster.OptionalAttribute)(2562, import_TlvNumber.TlvInt16, { default: 65535 }),
      reactiveCurrentPhaseC: (0, import_Cluster.OptionalAttribute)(2563, import_TlvNumber.TlvInt16, { default: 65535 }),
      rmsVoltagePhaseC: (0, import_Cluster.OptionalAttribute)(2565, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsVoltageMinPhaseC: (0, import_Cluster.OptionalAttribute)(2566, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsVoltageMaxPhaseC: (0, import_Cluster.OptionalAttribute)(2567, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsCurrentPhaseC: (0, import_Cluster.OptionalAttribute)(2568, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsCurrentMinPhaseC: (0, import_Cluster.OptionalAttribute)(2569, import_TlvNumber.TlvUInt16, { default: 65535 }),
      rmsCurrentMaxPhaseC: (0, import_Cluster.OptionalAttribute)(2570, import_TlvNumber.TlvUInt16, { default: 65535 }),
      activePowerPhaseC: (0, import_Cluster.OptionalAttribute)(2571, import_TlvNumber.TlvInt16, { default: 65535 }),
      activePowerMinPhaseC: (0, import_Cluster.OptionalAttribute)(2572, import_TlvNumber.TlvInt16, { default: 65535 }),
      activePowerMaxPhaseC: (0, import_Cluster.OptionalAttribute)(2573, import_TlvNumber.TlvInt16, { default: 65535 }),
      reactivePowerPhaseC: (0, import_Cluster.OptionalAttribute)(2574, import_TlvNumber.TlvInt16, { default: 65535 }),
      apparentPowerPhaseC: (0, import_Cluster.OptionalAttribute)(2575, import_TlvNumber.TlvUInt16, { default: 65535 }),
      powerFactorPhaseC: (0, import_Cluster.OptionalAttribute)(2576, import_TlvNumber.TlvInt8.bound({ min: -100, max: 100 }), { default: 0 }),
      averageRmsVoltageMeasurementPeriodPhaseC: (0, import_Cluster.OptionalAttribute)(2577, import_TlvNumber.TlvUInt16, { default: 0 }),
      averageRmsOverVoltageCounterPhaseC: (0, import_Cluster.OptionalAttribute)(2578, import_TlvNumber.TlvUInt16, { default: 0 }),
      averageRmsUnderVoltageCounterPhaseC: (0, import_Cluster.OptionalAttribute)(2579, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsExtremeOverVoltagePeriodPhaseC: (0, import_Cluster.OptionalAttribute)(2580, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsExtremeUnderVoltagePeriodPhaseC: (0, import_Cluster.OptionalAttribute)(2581, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsVoltageSagPeriodPhaseC: (0, import_Cluster.OptionalAttribute)(2582, import_TlvNumber.TlvUInt16, { default: 0 }),
      rmsVoltageSwellPeriodPhaseC: (0, import_Cluster.OptionalAttribute)(2583, import_TlvNumber.TlvUInt16, { default: 0 })
    },
    commands: {
      getProfileInfoCommand: (0, import_Cluster.OptionalCommand)(0, import_TlvNoArguments.TlvNoArguments, 0, import_Cluster.TlvNoResponse),
      getMeasurementProfileCommand: (0, import_Cluster.OptionalCommand)(
        1,
        ElectricalMeasurement2.TlvGetMeasurementProfileCommandRequest,
        1,
        import_Cluster.TlvNoResponse
      )
    }
  });
})(ElectricalMeasurement || (ElectricalMeasurement = {}));
const ElectricalMeasurementCluster = ElectricalMeasurement.Cluster;
