// @project-chip/matter.js/dist/cjs/cluster/definitions/FanControlCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FanControlCluster_exports = {};
__export(FanControlCluster_exports, {
  FanControl: () => FanControl,
  FanControlCluster: () => FanControlCluster
});
module.exports = __toCommonJS(FanControlCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var FanControl;
((FanControl2) => {
  let FanMode;
  ((FanMode2) => {
    FanMode2[FanMode2["Off"] = 0] = "Off";
    FanMode2[FanMode2["Low"] = 1] = "Low";
    FanMode2[FanMode2["Medium"] = 2] = "Medium";
    FanMode2[FanMode2["High"] = 3] = "High";
    FanMode2[FanMode2["On"] = 4] = "On";
    FanMode2[FanMode2["Auto"] = 5] = "Auto";
    FanMode2[FanMode2["Smart"] = 6] = "Smart";
  })(FanMode = FanControl2.FanMode || (FanControl2.FanMode = {}));
  let FanModeSequence;
  ((FanModeSequence2) => {
    FanModeSequence2[FanModeSequence2["OffLowMedHigh"] = 0] = "OffLowMedHigh";
    FanModeSequence2[FanModeSequence2["OffLowHigh"] = 1] = "OffLowHigh";
    FanModeSequence2[FanModeSequence2["OffLowMedHighAuto"] = 2] = "OffLowMedHighAuto";
    FanModeSequence2[FanModeSequence2["OffLowHighAuto"] = 3] = "OffLowHighAuto";
    FanModeSequence2[FanModeSequence2["OffOnAuto"] = 4] = "OffOnAuto";
    FanModeSequence2[FanModeSequence2["OffOn"] = 5] = "OffOn";
  })(FanModeSequence = FanControl2.FanModeSequence || (FanControl2.FanModeSequence = {}));
  FanControl2.RockSupport = { rockLeftRight: (0, import_BitmapSchema.BitFlag)(0), rockUpDown: (0, import_BitmapSchema.BitFlag)(1), rockRound: (0, import_BitmapSchema.BitFlag)(2) };
  FanControl2.RockSetting = { rockLeftRight: (0, import_BitmapSchema.BitFlag)(0), rockUpDown: (0, import_BitmapSchema.BitFlag)(1), rockRound: (0, import_BitmapSchema.BitFlag)(2) };
  FanControl2.WindSupport = { sleepWind: (0, import_BitmapSchema.BitFlag)(0), naturalWind: (0, import_BitmapSchema.BitFlag)(1) };
  FanControl2.WindSetting = { sleepWind: (0, import_BitmapSchema.BitFlag)(0), naturalWind: (0, import_BitmapSchema.BitFlag)(1) };
  let Feature;
  ((Feature2) => {
    Feature2["MultiSpeed"] = "MultiSpeed";
    Feature2["Auto"] = "Auto";
    Feature2["Rocking"] = "Rocking";
    Feature2["Wind"] = "Wind";
  })(Feature = FanControl2.Feature || (FanControl2.Feature = {}));
  FanControl2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 514,
    name: "FanControl",
    revision: 2,
    features: {
      /**
       * MultiSpeed
       *
       * 1-100 speeds
       */
      multiSpeed: (0, import_BitmapSchema.BitFlag)(0),
      /**
       * Auto
       *
       * Automatic mode supported for fan speed
       */
      auto: (0, import_BitmapSchema.BitFlag)(1),
      /**
       * Rocking
       *
       * Rocking movement supported
       */
      rocking: (0, import_BitmapSchema.BitFlag)(2),
      /**
       * Wind
       *
       * Wind emulation supported
       */
      wind: (0, import_BitmapSchema.BitFlag)(3)
    },
    attributes: {
      /**
       * This attribute shall indicate the current speed mode of the fan. This attribute may be written by the
       * client to indicate a new speed mode of the fan. This attribute shall be set to one of the values in the
       * table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.1
       */
      fanMode: (0, import_Cluster.WritableAttribute)(0, (0, import_TlvNumber.TlvEnum)(), { persistent: true, default: 0 /* Off */ }),
      /**
       * This indicates the fan speed ranges that shall be supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.2
       */
      fanModeSequence: (0, import_Cluster.WritableAttribute)(
        1,
        (0, import_TlvNumber.TlvEnum)(),
        { persistent: true, default: 2 /* OffLowMedHighAuto */ }
      ),
      /**
       * This attribute shall indicate the speed setting for the fan. This attribute may be written by the client
       * to indicate a new fan speed. If the client writes null to this attribute, the attribute value shall NOT
       * change. If this is set to 0, the server shall set the FanMode attribute value to Off.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.3
       */
      percentSetting: (0, import_Cluster.WritableAttribute)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ max: 100 })), { default: 0 }),
      /**
       * This attribute shall indicate the actual currently operating fan speed, or zero to indicate that the fan
       * is off. See Section 4.4.6.3.1 for more details.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.4
       */
      percentCurrent: (0, import_Cluster.Attribute)(3, import_TlvNumber.TlvUInt8.bound({ max: 100 }), { default: 0 })
    }
  });
  FanControl2.MultiSpeedComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute shall indicate that the fan has one speed (value of 1) or the maximum speed, if the fan
       * is capable of multiple speeds.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.5
       */
      speedMax: (0, import_Cluster.FixedAttribute)(4, import_TlvNumber.TlvUInt8.bound({ min: 1, max: 100 }), { default: 1 }),
      /**
       * This attribute shall indicate the speed setting for the fan. This attribute may be written by the client
       * to indicate a new fan speed. If the client writes null to this attribute, the attribute value shall NOT
       * change. If this is set to 0, the server shall set the FanMode attribute value to Off. Please see the
       * Section 4.4.6.6.1 for details on other values.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.6
       */
      speedSetting: (0, import_Cluster.WritableAttribute)(5, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ max: 100 })), { default: 0 }),
      /**
       * This attribute shall indicate the actual currently operating fan speed, or zero to indicate that the fan
       * is off.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.7
       */
      speedCurrent: (0, import_Cluster.Attribute)(6, import_TlvNumber.TlvUInt8.bound({ max: 100 }), { default: 0 })
    }
  });
  FanControl2.RockingComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute is a bitmap that indicates what rocking motions the server supports. The bitmap is shown
       * in the table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.8
       */
      rockSupport: (0, import_Cluster.FixedAttribute)(7, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, FanControl2.RockSupport)),
      /**
       * This attribute is a bitmap that indicates the current active fan rocking motion settings. Each bit shall
       * only be set to 1, if the corresponding bit in the RockSupport attribute is set to 1, otherwise a status
       * code of CONSTRAINT_ERROR shall be returned.
       *
       * If a combination of supported bits is set by the client, and the server does not support the
       * combination, the lowest supported single bit in the combination shall be set and active, and all other
       * bits shall indicate zero.
       *
       * For example: If RockUpDown and RockRound are both set, but this combination is not possible, then only
       * RockUpDown becomes active.
       *
       * The bitmap is shown in the table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.9
       */
      rockSetting: (0, import_Cluster.WritableAttribute)(8, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, FanControl2.RockSetting))
    }
  });
  FanControl2.WindComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute is a bitmap that indicates what wind modes the server supports. At least one wind mode
       * bit shall be set. The bitmap is shown in the table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.10
       */
      windSupport: (0, import_Cluster.FixedAttribute)(9, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, FanControl2.WindSupport)),
      /**
       * This attribute is a bitmap that indicates the current active fan wind feature settings. Each bit shall
       * only be set to 1, if the corresponding bit in the WindSupport attribute is set to 1, otherwise a status
       * code of CONSTRAINT_ERROR shall be returned.
       *
       * If a combination of supported bits is set by the client, and the server does not support the
       * combination, the lowest supported single bit in the combination shall be set and active, and all other
       * bits shall indicate zero.
       *
       * For example: If Sleep Wind and Natural Wind are set, but this combination is not possible, then only
       * Sleep Wind becomes active.
       *
       * The bitmap is shown in the table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.11
       */
      windSetting: (0, import_Cluster.WritableAttribute)(10, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, FanControl2.WindSetting))
    }
  });
  FanControl2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    FanControl2.Base,
    /**
     * Use this factory method to create a FanControl cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a FanControl cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...FanControl2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(FanControl2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, FanControl2.MultiSpeedComponent, { multiSpeed: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, FanControl2.RockingComponent, { rocking: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, FanControl2.WindComponent, { wind: true });
      return cluster;
    }
  );
  const SPD = { multiSpeed: true };
  const RCK = { rocking: true };
  const WND = { wind: true };
  FanControl2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: FanControl2.Cluster.id,
    name: FanControl2.Cluster.name,
    revision: FanControl2.Cluster.revision,
    features: FanControl2.Cluster.features,
    attributes: {
      ...FanControl2.Cluster.attributes,
      speedMax: import_ClusterFactory.ClusterFactory.AsConditional(FanControl2.MultiSpeedComponent.attributes.speedMax, { mandatoryIf: [SPD] }),
      speedSetting: import_ClusterFactory.ClusterFactory.AsConditional(
        FanControl2.MultiSpeedComponent.attributes.speedSetting,
        { mandatoryIf: [SPD] }
      ),
      speedCurrent: import_ClusterFactory.ClusterFactory.AsConditional(
        FanControl2.MultiSpeedComponent.attributes.speedCurrent,
        { mandatoryIf: [SPD] }
      ),
      rockSupport: import_ClusterFactory.ClusterFactory.AsConditional(FanControl2.RockingComponent.attributes.rockSupport, { mandatoryIf: [RCK] }),
      rockSetting: import_ClusterFactory.ClusterFactory.AsConditional(FanControl2.RockingComponent.attributes.rockSetting, { mandatoryIf: [RCK] }),
      windSupport: import_ClusterFactory.ClusterFactory.AsConditional(FanControl2.WindComponent.attributes.windSupport, { mandatoryIf: [WND] }),
      windSetting: import_ClusterFactory.ClusterFactory.AsConditional(FanControl2.WindComponent.attributes.windSetting, { mandatoryIf: [WND] })
    }
  });
})(FanControl || (FanControl = {}));
const FanControlCluster = FanControl.Cluster;
