// @project-chip/matter.js/dist/cjs/cluster/definitions/FaultInjectionCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FaultInjectionCluster_exports = {};
__export(FaultInjectionCluster_exports, {
  FaultInjection: () => FaultInjection,
  FaultInjectionCluster: () => FaultInjectionCluster
});
module.exports = __toCommonJS(FaultInjectionCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var FaultInjection;
((FaultInjection2) => {
  let FaultType;
  ((FaultType2) => {
    FaultType2[FaultType2["Unspecified"] = 0] = "Unspecified";
    FaultType2[FaultType2["SystemFault"] = 1] = "SystemFault";
    FaultType2[FaultType2["InetFault"] = 2] = "InetFault";
    FaultType2[FaultType2["ChipFault"] = 3] = "ChipFault";
    FaultType2[FaultType2["CertFault"] = 4] = "CertFault";
  })(FaultType = FaultInjection2.FaultType || (FaultInjection2.FaultType = {}));
  FaultInjection2.TlvFailAtFaultRequest = (0, import_TlvObject.TlvObject)({
    type: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    id: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt32),
    numCallsToSkip: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt32),
    numCallsToFail: (0, import_TlvObject.TlvField)(3, import_TlvNumber.TlvUInt32),
    takeMutex: (0, import_TlvObject.TlvField)(4, import_TlvBoolean.TlvBoolean)
  });
  FaultInjection2.TlvFailRandomlyAtFaultRequest = (0, import_TlvObject.TlvObject)({
    type: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    id: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt32),
    percentage: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8)
  });
  FaultInjection2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 4294048774,
    name: "FaultInjection",
    revision: 1,
    commands: {
      failAtFault: (0, import_Cluster.Command)(0, FaultInjection2.TlvFailAtFaultRequest, 0, import_Cluster.TlvNoResponse, { invokeAcl: import_Cluster.AccessLevel.Manage }),
      failRandomlyAtFault: (0, import_Cluster.Command)(
        1,
        FaultInjection2.TlvFailRandomlyAtFaultRequest,
        1,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      )
    }
  });
})(FaultInjection || (FaultInjection = {}));
const FaultInjectionCluster = FaultInjection.Cluster;
