// @project-chip/matter.js/dist/cjs/cluster/definitions/FixedLabelCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FixedLabelCluster_exports = {};
__export(FixedLabelCluster_exports, {
  FixedLabel: () => FixedLabel,
  FixedLabelCluster: () => FixedLabelCluster
});
module.exports = __toCommonJS(FixedLabelCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_LabelCluster = require("../../cluster/definitions/LabelCluster.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var FixedLabel;
((FixedLabel2) => {
  FixedLabel2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 64,
    name: "FixedLabel",
    revision: 1,
    attributes: {
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.8.4
       */
      labelList: (0, import_Cluster.Attribute)(0, (0, import_TlvArray.TlvArray)(import_LabelCluster.Label.TlvLabelStruct), { persistent: true, default: [] })
    }
  });
})(FixedLabel || (FixedLabel = {}));
const FixedLabelCluster = FixedLabel.Cluster;
