// @project-chip/matter.js/dist/cjs/cluster/definitions/GroupsCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GroupsCluster_exports = {};
__export(GroupsCluster_exports, {
  Groups: () => Groups,
  GroupsCluster: () => GroupsCluster
});
module.exports = __toCommonJS(GroupsCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_GroupId = require("../../datatype/GroupId.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Groups;
((Groups2) => {
  Groups2.NameSupport = {
    /**
     * The ability to store a name for a group.
     */
    nameSupport: (0, import_BitmapSchema.BitFlag)(7)
  };
  Groups2.TlvAddGroupRequest = (0, import_TlvObject.TlvObject)({
    groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId),
    groupName: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvString.bound({ maxLength: 16 }))
  });
  Groups2.TlvAddGroupResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    groupId: (0, import_TlvObject.TlvField)(1, import_GroupId.TlvGroupId)
  });
  Groups2.TlvViewGroupRequest = (0, import_TlvObject.TlvObject)({ groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId) });
  Groups2.TlvViewGroupResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    groupId: (0, import_TlvObject.TlvField)(1, import_GroupId.TlvGroupId),
    groupName: (0, import_TlvObject.TlvField)(2, import_TlvString.TlvString.bound({ maxLength: 16 }))
  });
  Groups2.TlvGetGroupMembershipRequest = (0, import_TlvObject.TlvObject)({ groupList: (0, import_TlvObject.TlvField)(0, (0, import_TlvArray.TlvArray)(import_GroupId.TlvGroupId)) });
  Groups2.TlvGetGroupMembershipResponse = (0, import_TlvObject.TlvObject)({
    capacity: (0, import_TlvObject.TlvField)(0, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8)),
    groupList: (0, import_TlvObject.TlvField)(1, (0, import_TlvArray.TlvArray)(import_GroupId.TlvGroupId))
  });
  Groups2.TlvRemoveGroupRequest = (0, import_TlvObject.TlvObject)({ groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId) });
  Groups2.TlvRemoveGroupResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    groupId: (0, import_TlvObject.TlvField)(1, import_GroupId.TlvGroupId)
  });
  Groups2.TlvAddGroupIfIdentifyingRequest = (0, import_TlvObject.TlvObject)({
    groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId),
    groupName: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvString.bound({ maxLength: 16 }))
  });
  let Feature;
  ((Feature2) => {
    Feature2["GroupNames"] = "GroupNames";
  })(Feature = Groups2.Feature || (Groups2.Feature = {}));
  Groups2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 4,
    name: "Groups",
    revision: 4,
    features: {
      /**
       * GroupNames
       *
       * The ability to store a name for a group.
       */
      groupNames: (0, import_BitmapSchema.BitFlag)(0)
    },
    attributes: {
      /**
       * This attribute provides legacy, read-only access to whether the Group Names feature is supported. The
       * most significant bit, bit 7, shall be equal to bit 0 of the FeatureMap attribute. All other bits shall
       * be 0.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.6.1
       */
      nameSupport: (0, import_Cluster.FixedAttribute)(0, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Groups2.NameSupport))
    },
    commands: {
      /**
       * The AddGroup command allows a client to add group membership in a particular group for the server
       * endpoint.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.1
       */
      addGroup: (0, import_Cluster.Command)(0, Groups2.TlvAddGroupRequest, 0, Groups2.TlvAddGroupResponse, { invokeAcl: import_Cluster.AccessLevel.Manage }),
      /**
       * The ViewGroup command allows a client to request that the server responds with a ViewGroupResponse
       * command containing the name string for a particular group.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.2
       */
      viewGroup: (0, import_Cluster.Command)(1, Groups2.TlvViewGroupRequest, 1, Groups2.TlvViewGroupResponse),
      /**
       * The GetGroupMembership command allows a client to inquire about the group membership of the server
       * endpoint, in a number of ways.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.3
       */
      getGroupMembership: (0, import_Cluster.Command)(2, Groups2.TlvGetGroupMembershipRequest, 2, Groups2.TlvGetGroupMembershipResponse),
      /**
       * The RemoveGroup command allows a client to request that the server removes the membership for the server
       * endpoint, if any, in a particular group.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.4
       */
      removeGroup: (0, import_Cluster.Command)(
        3,
        Groups2.TlvRemoveGroupRequest,
        3,
        Groups2.TlvRemoveGroupResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The RemoveAllGroups command allows a client to direct the server to remove all group associations for
       * the server endpoint.
       *
       * The RemoveAllGroups command has no data fields.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.5
       */
      removeAllGroups: (0, import_Cluster.Command)(4, import_TlvNoArguments.TlvNoArguments, 4, import_Cluster.TlvNoResponse, { invokeAcl: import_Cluster.AccessLevel.Manage }),
      /**
       * The AddGroupIfIdentifying command allows a client to add group membership in a particular group for the
       * server endpoint, on condition that the endpoint is identifying itself. Identifying functionality is
       * controlled using the Identify cluster, (see Identify).
       *
       * This command might be used to assist configuring group membership in the absence of a commissioning tool.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.6
       */
      addGroupIfIdentifying: (0, import_Cluster.Command)(
        5,
        Groups2.TlvAddGroupIfIdentifyingRequest,
        5,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      )
    }
  });
  Groups2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    { ...Groups2.Base, supportedFeatures: { groupNames: true } },
    /**
     * Use this factory method to create a Groups cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a Groups cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...Groups2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(Groups2.Base.features, ...features)
      });
      return cluster;
    }
  );
})(Groups || (Groups = {}));
const GroupsCluster = Groups.Cluster;
