// @project-chip/matter.js/dist/cjs/cluster/definitions/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var definitions_exports = {};
__export(definitions_exports, {
  AccessControl: () => import_AccessControlCluster.AccessControl,
  AccessControlCluster: () => import_AccessControlCluster.AccessControlCluster,
  AccountLogin: () => import_AccountLoginCluster.AccountLogin,
  AccountLoginCluster: () => import_AccountLoginCluster.AccountLoginCluster,
  Actions: () => import_ActionsCluster.Actions,
  ActionsCluster: () => import_ActionsCluster.ActionsCluster,
  AdministratorCommissioning: () => import_AdministratorCommissioningCluster.AdministratorCommissioning,
  AdministratorCommissioningCluster: () => import_AdministratorCommissioningCluster.AdministratorCommissioningCluster,
  ApplicationBasic: () => import_ApplicationBasicCluster.ApplicationBasic,
  ApplicationBasicCluster: () => import_ApplicationBasicCluster.ApplicationBasicCluster,
  ApplicationLauncher: () => import_ApplicationLauncherCluster.ApplicationLauncher,
  ApplicationLauncherCluster: () => import_ApplicationLauncherCluster.ApplicationLauncherCluster,
  AudioOutput: () => import_AudioOutputCluster.AudioOutput,
  AudioOutputCluster: () => import_AudioOutputCluster.AudioOutputCluster,
  BallastConfiguration: () => import_BallastConfigurationCluster.BallastConfiguration,
  BallastConfigurationCluster: () => import_BallastConfigurationCluster.BallastConfigurationCluster,
  BarrierControl: () => import_BarrierControlCluster.BarrierControl,
  BarrierControlCluster: () => import_BarrierControlCluster.BarrierControlCluster,
  BasicInformation: () => import_BasicInformationCluster.BasicInformation,
  BasicInformationCluster: () => import_BasicInformationCluster.BasicInformationCluster,
  BinaryInputBasic: () => import_BinaryInputBasicCluster.BinaryInputBasic,
  BinaryInputBasicCluster: () => import_BinaryInputBasicCluster.BinaryInputBasicCluster,
  Binding: () => import_BindingCluster.Binding,
  BindingCluster: () => import_BindingCluster.BindingCluster,
  BooleanState: () => import_BooleanStateCluster.BooleanState,
  BooleanStateCluster: () => import_BooleanStateCluster.BooleanStateCluster,
  BridgedDeviceBasicInformation: () => import_BridgedDeviceBasicInformationCluster.BridgedDeviceBasicInformation,
  BridgedDeviceBasicInformationCluster: () => import_BridgedDeviceBasicInformationCluster.BridgedDeviceBasicInformationCluster,
  Channel: () => import_ChannelCluster.Channel,
  ChannelCluster: () => import_ChannelCluster.ChannelCluster,
  ClientMonitoring: () => import_ClientMonitoringCluster.ClientMonitoring,
  ClientMonitoringCluster: () => import_ClientMonitoringCluster.ClientMonitoringCluster,
  ColorControl: () => import_ColorControlCluster.ColorControl,
  ColorControlCluster: () => import_ColorControlCluster.ColorControlCluster,
  ContentLauncher: () => import_ContentLauncherCluster.ContentLauncher,
  ContentLauncherCluster: () => import_ContentLauncherCluster.ContentLauncherCluster,
  Descriptor: () => import_DescriptorCluster.Descriptor,
  DescriptorCluster: () => import_DescriptorCluster.DescriptorCluster,
  DiagnosticLogs: () => import_DiagnosticLogsCluster.DiagnosticLogs,
  DiagnosticLogsCluster: () => import_DiagnosticLogsCluster.DiagnosticLogsCluster,
  DoorLock: () => import_DoorLockCluster.DoorLock,
  DoorLockCluster: () => import_DoorLockCluster.DoorLockCluster,
  ElectricalMeasurement: () => import_ElectricalMeasurementCluster.ElectricalMeasurement,
  ElectricalMeasurementCluster: () => import_ElectricalMeasurementCluster.ElectricalMeasurementCluster,
  EthernetNetworkDiagnostics: () => import_EthernetNetworkDiagnosticsCluster.EthernetNetworkDiagnostics,
  EthernetNetworkDiagnosticsCluster: () => import_EthernetNetworkDiagnosticsCluster.EthernetNetworkDiagnosticsCluster,
  FanControl: () => import_FanControlCluster.FanControl,
  FanControlCluster: () => import_FanControlCluster.FanControlCluster,
  FaultInjection: () => import_FaultInjectionCluster.FaultInjection,
  FaultInjectionCluster: () => import_FaultInjectionCluster.FaultInjectionCluster,
  FixedLabel: () => import_FixedLabelCluster.FixedLabel,
  FixedLabelCluster: () => import_FixedLabelCluster.FixedLabelCluster,
  FlowMeasurement: () => import_FlowMeasurementCluster.FlowMeasurement,
  FlowMeasurementCluster: () => import_FlowMeasurementCluster.FlowMeasurementCluster,
  GeneralCommissioning: () => import_GeneralCommissioningCluster.GeneralCommissioning,
  GeneralCommissioningCluster: () => import_GeneralCommissioningCluster.GeneralCommissioningCluster,
  GeneralDiagnostics: () => import_GeneralDiagnosticsCluster.GeneralDiagnostics,
  GeneralDiagnosticsCluster: () => import_GeneralDiagnosticsCluster.GeneralDiagnosticsCluster,
  GroupKeyManagement: () => import_GroupKeyManagementCluster.GroupKeyManagement,
  GroupKeyManagementCluster: () => import_GroupKeyManagementCluster.GroupKeyManagementCluster,
  Groups: () => import_GroupsCluster.Groups,
  GroupsCluster: () => import_GroupsCluster.GroupsCluster,
  Identify: () => import_IdentifyCluster.Identify,
  IdentifyCluster: () => import_IdentifyCluster.IdentifyCluster,
  IlluminanceMeasurement: () => import_IlluminanceMeasurementCluster.IlluminanceMeasurement,
  IlluminanceMeasurementCluster: () => import_IlluminanceMeasurementCluster.IlluminanceMeasurementCluster,
  KeypadInput: () => import_KeypadInputCluster.KeypadInput,
  KeypadInputCluster: () => import_KeypadInputCluster.KeypadInputCluster,
  LeafWetnessMeasurement: () => import_LeafWetnessMeasurementCluster.LeafWetnessMeasurement,
  LeafWetnessMeasurementCluster: () => import_LeafWetnessMeasurementCluster.LeafWetnessMeasurementCluster,
  LevelControl: () => import_LevelControlCluster.LevelControl,
  LevelControlCluster: () => import_LevelControlCluster.LevelControlCluster,
  LocalizationConfiguration: () => import_LocalizationConfigurationCluster.LocalizationConfiguration,
  LocalizationConfigurationCluster: () => import_LocalizationConfigurationCluster.LocalizationConfigurationCluster,
  LowPower: () => import_LowPowerCluster.LowPower,
  LowPowerCluster: () => import_LowPowerCluster.LowPowerCluster,
  MediaInput: () => import_MediaInputCluster.MediaInput,
  MediaInputCluster: () => import_MediaInputCluster.MediaInputCluster,
  MediaPlayback: () => import_MediaPlaybackCluster.MediaPlayback,
  MediaPlaybackCluster: () => import_MediaPlaybackCluster.MediaPlaybackCluster,
  ModeSelect: () => import_ModeSelectCluster.ModeSelect,
  ModeSelectCluster: () => import_ModeSelectCluster.ModeSelectCluster,
  NetworkCommissioning: () => import_NetworkCommissioningCluster.NetworkCommissioning,
  NetworkCommissioningCluster: () => import_NetworkCommissioningCluster.NetworkCommissioningCluster,
  OccupancySensing: () => import_OccupancySensingCluster.OccupancySensing,
  OccupancySensingCluster: () => import_OccupancySensingCluster.OccupancySensingCluster,
  OnOff: () => import_OnOffCluster.OnOff,
  OnOffCluster: () => import_OnOffCluster.OnOffCluster,
  OnOffSwitchConfiguration: () => import_OnOffSwitchConfigurationCluster.OnOffSwitchConfiguration,
  OnOffSwitchConfigurationCluster: () => import_OnOffSwitchConfigurationCluster.OnOffSwitchConfigurationCluster,
  OperationalCredentials: () => import_OperationalCredentialsCluster.OperationalCredentials,
  OperationalCredentialsCluster: () => import_OperationalCredentialsCluster.OperationalCredentialsCluster,
  OtaSoftwareUpdateProvider: () => import_OtaSoftwareUpdateProviderCluster.OtaSoftwareUpdateProvider,
  OtaSoftwareUpdateProviderCluster: () => import_OtaSoftwareUpdateProviderCluster.OtaSoftwareUpdateProviderCluster,
  OtaSoftwareUpdateRequestor: () => import_OtaSoftwareUpdateRequestorCluster.OtaSoftwareUpdateRequestor,
  OtaSoftwareUpdateRequestorCluster: () => import_OtaSoftwareUpdateRequestorCluster.OtaSoftwareUpdateRequestorCluster,
  PowerSource: () => import_PowerSourceCluster.PowerSource,
  PowerSourceCluster: () => import_PowerSourceCluster.PowerSourceCluster,
  PowerSourceConfiguration: () => import_PowerSourceConfigurationCluster.PowerSourceConfiguration,
  PowerSourceConfigurationCluster: () => import_PowerSourceConfigurationCluster.PowerSourceConfigurationCluster,
  PressureMeasurement: () => import_PressureMeasurementCluster.PressureMeasurement,
  PressureMeasurementCluster: () => import_PressureMeasurementCluster.PressureMeasurementCluster,
  ProxyConfiguration: () => import_ProxyConfigurationCluster.ProxyConfiguration,
  ProxyConfigurationCluster: () => import_ProxyConfigurationCluster.ProxyConfigurationCluster,
  ProxyDiscovery: () => import_ProxyDiscoveryCluster.ProxyDiscovery,
  ProxyDiscoveryCluster: () => import_ProxyDiscoveryCluster.ProxyDiscoveryCluster,
  PulseWidthModulation: () => import_PulseWidthModulationCluster.PulseWidthModulation,
  PulseWidthModulationCluster: () => import_PulseWidthModulationCluster.PulseWidthModulationCluster,
  PumpConfigurationAndControl: () => import_PumpConfigurationAndControlCluster.PumpConfigurationAndControl,
  PumpConfigurationAndControlCluster: () => import_PumpConfigurationAndControlCluster.PumpConfigurationAndControlCluster,
  RelativeHumidityMeasurement: () => import_RelativeHumidityMeasurementCluster.RelativeHumidityMeasurement,
  RelativeHumidityMeasurementCluster: () => import_RelativeHumidityMeasurementCluster.RelativeHumidityMeasurementCluster,
  Scenes: () => import_ScenesCluster.Scenes,
  ScenesCluster: () => import_ScenesCluster.ScenesCluster,
  SoftwareDiagnostics: () => import_SoftwareDiagnosticsCluster.SoftwareDiagnostics,
  SoftwareDiagnosticsCluster: () => import_SoftwareDiagnosticsCluster.SoftwareDiagnosticsCluster,
  SoilMoistureMeasurement: () => import_SoilMoistureMeasurementCluster.SoilMoistureMeasurement,
  SoilMoistureMeasurementCluster: () => import_SoilMoistureMeasurementCluster.SoilMoistureMeasurementCluster,
  Switch: () => import_SwitchCluster.Switch,
  SwitchCluster: () => import_SwitchCluster.SwitchCluster,
  TargetNavigator: () => import_TargetNavigatorCluster.TargetNavigator,
  TargetNavigatorCluster: () => import_TargetNavigatorCluster.TargetNavigatorCluster,
  TemperatureMeasurement: () => import_TemperatureMeasurementCluster.TemperatureMeasurement,
  TemperatureMeasurementCluster: () => import_TemperatureMeasurementCluster.TemperatureMeasurementCluster,
  Thermostat: () => import_ThermostatCluster.Thermostat,
  ThermostatCluster: () => import_ThermostatCluster.ThermostatCluster,
  ThermostatUserInterfaceConfiguration: () => import_ThermostatUserInterfaceConfigurationCluster.ThermostatUserInterfaceConfiguration,
  ThermostatUserInterfaceConfigurationCluster: () => import_ThermostatUserInterfaceConfigurationCluster.ThermostatUserInterfaceConfigurationCluster,
  ThreadNetworkDiagnostics: () => import_ThreadNetworkDiagnosticsCluster.ThreadNetworkDiagnostics,
  ThreadNetworkDiagnosticsCluster: () => import_ThreadNetworkDiagnosticsCluster.ThreadNetworkDiagnosticsCluster,
  TimeFormatLocalization: () => import_TimeFormatLocalizationCluster.TimeFormatLocalization,
  TimeFormatLocalizationCluster: () => import_TimeFormatLocalizationCluster.TimeFormatLocalizationCluster,
  TimeSync: () => import_TimeSyncCluster.TimeSync,
  TimeSyncCluster: () => import_TimeSyncCluster.TimeSyncCluster,
  UnitLocalization: () => import_UnitLocalizationCluster.UnitLocalization,
  UnitLocalizationCluster: () => import_UnitLocalizationCluster.UnitLocalizationCluster,
  UserLabel: () => import_UserLabelCluster.UserLabel,
  UserLabelCluster: () => import_UserLabelCluster.UserLabelCluster,
  ValidProxies: () => import_ValidProxiesCluster.ValidProxies,
  ValidProxiesCluster: () => import_ValidProxiesCluster.ValidProxiesCluster,
  WakeOnLan: () => import_WakeOnLanCluster.WakeOnLan,
  WakeOnLanCluster: () => import_WakeOnLanCluster.WakeOnLanCluster,
  WiFiNetworkDiagnostics: () => import_WiFiNetworkDiagnosticsCluster.WiFiNetworkDiagnostics,
  WiFiNetworkDiagnosticsCluster: () => import_WiFiNetworkDiagnosticsCluster.WiFiNetworkDiagnosticsCluster,
  WindowCovering: () => import_WindowCoveringCluster.WindowCovering,
  WindowCoveringCluster: () => import_WindowCoveringCluster.WindowCoveringCluster
});
module.exports = __toCommonJS(definitions_exports);
var import_IdentifyCluster = require("./IdentifyCluster.js");
var import_GroupsCluster = require("./GroupsCluster.js");
var import_ScenesCluster = require("./ScenesCluster.js");
var import_OnOffCluster = require("./OnOffCluster.js");
var import_LevelControlCluster = require("./LevelControlCluster.js");
var import_PulseWidthModulationCluster = require("./PulseWidthModulationCluster.js");
var import_BooleanStateCluster = require("./BooleanStateCluster.js");
var import_ModeSelectCluster = require("./ModeSelectCluster.js");
var import_LowPowerCluster = require("./LowPowerCluster.js");
var import_WakeOnLanCluster = require("./WakeOnLanCluster.js");
var import_SwitchCluster = require("./SwitchCluster.js");
var import_IlluminanceMeasurementCluster = require("./IlluminanceMeasurementCluster.js");
var import_TemperatureMeasurementCluster = require("./TemperatureMeasurementCluster.js");
var import_PressureMeasurementCluster = require("./PressureMeasurementCluster.js");
var import_FlowMeasurementCluster = require("./FlowMeasurementCluster.js");
var import_RelativeHumidityMeasurementCluster = require("./RelativeHumidityMeasurementCluster.js");
var import_LeafWetnessMeasurementCluster = require("./LeafWetnessMeasurementCluster.js");
var import_SoilMoistureMeasurementCluster = require("./SoilMoistureMeasurementCluster.js");
var import_OccupancySensingCluster = require("./OccupancySensingCluster.js");
var import_ColorControlCluster = require("./ColorControlCluster.js");
var import_BallastConfigurationCluster = require("./BallastConfigurationCluster.js");
var import_PumpConfigurationAndControlCluster = require("./PumpConfigurationAndControlCluster.js");
var import_ThermostatCluster = require("./ThermostatCluster.js");
var import_FanControlCluster = require("./FanControlCluster.js");
var import_ThermostatUserInterfaceConfigurationCluster = require("./ThermostatUserInterfaceConfigurationCluster.js");
var import_DoorLockCluster = require("./DoorLockCluster.js");
var import_WindowCoveringCluster = require("./WindowCoveringCluster.js");
var import_AccountLoginCluster = require("./AccountLoginCluster.js");
var import_ApplicationBasicCluster = require("./ApplicationBasicCluster.js");
var import_ApplicationLauncherCluster = require("./ApplicationLauncherCluster.js");
var import_AudioOutputCluster = require("./AudioOutputCluster.js");
var import_ChannelCluster = require("./ChannelCluster.js");
var import_ContentLauncherCluster = require("./ContentLauncherCluster.js");
var import_KeypadInputCluster = require("./KeypadInputCluster.js");
var import_MediaInputCluster = require("./MediaInputCluster.js");
var import_MediaPlaybackCluster = require("./MediaPlaybackCluster.js");
var import_TargetNavigatorCluster = require("./TargetNavigatorCluster.js");
var import_DescriptorCluster = require("./DescriptorCluster.js");
var import_BindingCluster = require("./BindingCluster.js");
var import_FixedLabelCluster = require("./FixedLabelCluster.js");
var import_UserLabelCluster = require("./UserLabelCluster.js");
var import_AccessControlCluster = require("./AccessControlCluster.js");
var import_BridgedDeviceBasicInformationCluster = require("./BridgedDeviceBasicInformationCluster.js");
var import_ActionsCluster = require("./ActionsCluster.js");
var import_ProxyDiscoveryCluster = require("./ProxyDiscoveryCluster.js");
var import_ProxyConfigurationCluster = require("./ProxyConfigurationCluster.js");
var import_ValidProxiesCluster = require("./ValidProxiesCluster.js");
var import_BasicInformationCluster = require("./BasicInformationCluster.js");
var import_GroupKeyManagementCluster = require("./GroupKeyManagementCluster.js");
var import_LocalizationConfigurationCluster = require("./LocalizationConfigurationCluster.js");
var import_TimeFormatLocalizationCluster = require("./TimeFormatLocalizationCluster.js");
var import_UnitLocalizationCluster = require("./UnitLocalizationCluster.js");
var import_PowerSourceConfigurationCluster = require("./PowerSourceConfigurationCluster.js");
var import_PowerSourceCluster = require("./PowerSourceCluster.js");
var import_NetworkCommissioningCluster = require("./NetworkCommissioningCluster.js");
var import_GeneralCommissioningCluster = require("./GeneralCommissioningCluster.js");
var import_DiagnosticLogsCluster = require("./DiagnosticLogsCluster.js");
var import_GeneralDiagnosticsCluster = require("./GeneralDiagnosticsCluster.js");
var import_SoftwareDiagnosticsCluster = require("./SoftwareDiagnosticsCluster.js");
var import_ThreadNetworkDiagnosticsCluster = require("./ThreadNetworkDiagnosticsCluster.js");
var import_WiFiNetworkDiagnosticsCluster = require("./WiFiNetworkDiagnosticsCluster.js");
var import_EthernetNetworkDiagnosticsCluster = require("./EthernetNetworkDiagnosticsCluster.js");
var import_TimeSyncCluster = require("./TimeSyncCluster.js");
var import_OperationalCredentialsCluster = require("./OperationalCredentialsCluster.js");
var import_AdministratorCommissioningCluster = require("./AdministratorCommissioningCluster.js");
var import_OtaSoftwareUpdateProviderCluster = require("./OtaSoftwareUpdateProviderCluster.js");
var import_OtaSoftwareUpdateRequestorCluster = require("./OtaSoftwareUpdateRequestorCluster.js");
var import_ClientMonitoringCluster = require("./ClientMonitoringCluster.js");
var import_FaultInjectionCluster = require("./FaultInjectionCluster.js");
var import_BarrierControlCluster = require("./BarrierControlCluster.js");
var import_ElectricalMeasurementCluster = require("./ElectricalMeasurementCluster.js");
var import_BinaryInputBasicCluster = require("./BinaryInputBasicCluster.js");
var import_OnOffSwitchConfigurationCluster = require("./OnOffSwitchConfigurationCluster.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
