// @project-chip/matter.js/dist/cjs/cluster/definitions/LowPowerCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LowPowerCluster_exports = {};
__export(LowPowerCluster_exports, {
  LowPower: () => LowPower,
  LowPowerCluster: () => LowPowerCluster
});
module.exports = __toCommonJS(LowPowerCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var LowPower;
((LowPower2) => {
  LowPower2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 1288,
    name: "LowPower",
    revision: 1,
    commands: {
      /**
       * This command shall put the device into low power mode.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.9.3.1
       */
      sleep: (0, import_Cluster.Command)(0, import_TlvNoArguments.TlvNoArguments, 0, import_Cluster.TlvNoResponse)
    }
  });
})(LowPower || (LowPower = {}));
const LowPowerCluster = LowPower.Cluster;
