// @project-chip/matter.js/dist/cjs/cluster/definitions/ModeSelectCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ModeSelectCluster_exports = {};
__export(ModeSelectCluster_exports, {
  ModeSelect: () => ModeSelect,
  ModeSelectCluster: () => ModeSelectCluster
});
module.exports = __toCommonJS(ModeSelectCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_VendorId = require("../../datatype/VendorId.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ModeSelect;
((ModeSelect2) => {
  ModeSelect2.TlvSemanticTagStruct = (0, import_TlvObject.TlvObject)({
    /**
     * If this field is null, the Value field shall be defined in a standard namespace as indicated by the
     * StandardNamespace attribute. If this field is not null, it shall indicate a manufacturer code (Vendor ID),
     * and the Value field shall indicate a semantic tag defined by the manufacturer. Each manufacturer code
     * supports a single namespace of values. The same manufacturer code and semantic tag value in separate cluster
     * instances are part of the same namespace and have the same meaning. For example: a manufacturer tag meaning
     * "pinch", has the same meaning in a cluster whose purpose is to choose the amount of sugar, or amount of salt.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.8.2.2
     */
    mfgCode: (0, import_TlvObject.TlvField)(0, (0, import_TlvNullable.TlvNullable)(import_VendorId.TlvVendorId)),
    /**
     * This field shall indicate the semantic tag within a semantic tag namespace which is either manufacturer
     * specific or standard. For semantic tags in a standard namespace, see Standard Namespace.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.8.2.1
     */
    value: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16)
  });
  ModeSelect2.TlvModeOptionStruct = (0, import_TlvObject.TlvObject)({
    /**
     * This field is readable text that describes the mode option that can be used by a client to indicate to the
     * user what this option means. This field is meant to be readable and understandable by the user.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.8.1.1
     */
    label: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvString.bound({ maxLength: 64 })),
    /**
     * The Mode field is used to identify the mode option. The value shall be unique for every item in the
     * SupportedModes attribute.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.8.1.2
     */
    mode: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8),
    /**
     * This field is a list of semantic tags that map to the mode option. This may be used by clients to determine
     * the meaning of the mode option as defined in a standard or manufacturer specific namespace. Semantic tags
     * can help clients look for options that meet certain criteria. A semantic tag shall be either a standard tag
     * or manufacturer specific tag as defined in each SemanticTagStruct list entry.
     *
     * A mode option may have more than one semantic tag. A mode option may be mapped to a mixture of standard and
     * manufacturer specific semantic tags.
     *
     * All standard semantic tags are from a single namespace indicated by the StandardNamespace attribute.
     *
     * For example: A mode labeled "100%" can have both the HIGH (MS) and MAX (standard) semantic tag. Clients
     * seeking the option for either HIGH or MAX will find the same option in this case.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.8.1.3
     */
    semanticTags: (0, import_TlvObject.TlvField)(2, (0, import_TlvArray.TlvArray)(ModeSelect2.TlvSemanticTagStruct, { maxLength: 64 }))
  });
  ModeSelect2.TlvChangeToModeRequest = (0, import_TlvObject.TlvObject)({ newMode: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8) });
  let Feature;
  ((Feature2) => {
    Feature2["OnOff"] = "OnOff";
  })(Feature = ModeSelect2.Feature || (ModeSelect2.Feature = {}));
  ModeSelect2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 80,
    name: "ModeSelect",
    revision: 1,
    features: {
      /**
       * OnOff
       *
       * Dependency with the On/Off cluster
       */
      onOff: (0, import_BitmapSchema.BitFlag)(0)
    },
    attributes: {
      /**
       * This attribute describes the purpose of the server, in readable text.
       *
       * For example, a coffee machine may have a Mode Select cluster for the amount of milk to add, and another
       * Mode Select cluster for the amount of sugar to add. In this case, the first instance can have the
       * description Milk and the second instance can have the description Sugar. This allows the user to tell
       * the purpose of each of the instances.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.5.1
       */
      description: (0, import_Cluster.FixedAttribute)(0, import_TlvString.TlvString.bound({ maxLength: 64 })),
      /**
       * This attribute, when not null, shall indicate a single standard namespace for any standard semantic tag
       * value supported in this or any other cluster instance with the same value of this attribute. A null
       * value indicates no standard namespace, and therefore, no standard semantic tags are provided in this
       * cluster instance. Each standard namespace and corresponding values and value meanings shall be defined
       * in another document.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.5.2
       */
      standardNamespace: (0, import_Cluster.FixedAttribute)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16), { default: null }),
      /**
       * This attribute is the list of supported modes that may be selected for the CurrentMode attribute. Each
       * item in this list represents a unique mode as indicated by the Mode field of the ModeOptionStruct. Each
       * entry in this list shall have a unique value for the Mode field.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.5.3
       */
      supportedModes: (0, import_Cluster.FixedAttribute)(2, (0, import_TlvArray.TlvArray)(ModeSelect2.TlvModeOptionStruct, { maxLength: 255 }), { default: [] }),
      /**
       * This attribute represents the current mode of the server.
       *
       * The value of this field must match the Mode field of one of the entries in the SupportedModes
       *
       * attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.5.4
       */
      currentMode: (0, import_Cluster.Attribute)(3, import_TlvNumber.TlvUInt8, { scene: true, persistent: true }),
      /**
       * The StartUpMode attribute value indicates the desired startup mode for the server when it is supplied
       * with power.
       *
       * If this attribute is not null, the CurrentMode attribute shall be set to the StartUpMode value, when the
       * server is powered up, except in the case when the OnMode attribute overrides the StartUpMode attribute.
       *
       * This behavior does not apply to reboots associated with OTA. After an OTA restart, the CurrentMode
       * attribute shall return to its value prior to the restart.
       *
       * The value of this field shall match the Mode field of one of the entries in the SupportedModes
       *
       * attribute.
       *
       * If this attribute is not implemented, or is set to the null value, it shall have no effect.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.5.5
       */
      startUpMode: (0, import_Cluster.OptionalWritableAttribute)(4, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8), { persistent: true })
    },
    commands: {
      /**
       * On receipt of this command, if the NewMode field indicates a valid mode transition within the supported
       * list, the server shall set the CurrentMode attribute to the NewMode value, otherwise, the
       *
       * server shall respond with an INVALID_COMMAND status response.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.6.1
       */
      changeToMode: (0, import_Cluster.Command)(0, ModeSelect2.TlvChangeToModeRequest, 0, import_Cluster.TlvNoResponse)
    }
  });
  ModeSelect2.OnOffComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute shall indicate the value of CurrentMode that depends on the state of the On/Off cluster
       * on the same endpoint. If this attribute is not present or is set to null, it shall NOT have an effect,
       * otherwise the CurrentMode attribute shall depend on the OnOff attribute of the On/Off cluster
       *
       * The value of this field shall match the Mode field of one of the entries in the SupportedModes attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.8.5.6
       */
      onMode: (0, import_Cluster.WritableAttribute)(5, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8), { persistent: true, default: null })
    }
  });
  ModeSelect2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    ModeSelect2.Base,
    /**
     * Use this factory method to create a ModeSelect cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a ModeSelect cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...ModeSelect2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(ModeSelect2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, ModeSelect2.OnOffComponent, { onOff: true });
      return cluster;
    }
  );
  const DEPONOFF = { onOff: true };
  ModeSelect2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: ModeSelect2.Cluster.id,
    name: ModeSelect2.Cluster.name,
    revision: ModeSelect2.Cluster.revision,
    features: ModeSelect2.Cluster.features,
    attributes: {
      ...ModeSelect2.Cluster.attributes,
      onMode: import_ClusterFactory.ClusterFactory.AsConditional(ModeSelect2.OnOffComponent.attributes.onMode, { mandatoryIf: [DEPONOFF] })
    },
    commands: ModeSelect2.Cluster.commands
  });
})(ModeSelect || (ModeSelect = {}));
const ModeSelectCluster = ModeSelect.Cluster;
