// @project-chip/matter.js/dist/cjs/cluster/definitions/OccupancySensingCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OccupancySensingCluster_exports = {};
__export(OccupancySensingCluster_exports, {
  OccupancySensing: () => OccupancySensing,
  OccupancySensingCluster: () => OccupancySensingCluster
});
module.exports = __toCommonJS(OccupancySensingCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var OccupancySensing;
((OccupancySensing2) => {
  OccupancySensing2.OccupancyBitmap = {
    /**
     * Indicates the sensed occupancy state; 1 = occupied, 0 = unoccupied.
     */
    occupied: (0, import_BitmapSchema.BitFlag)(0)
  };
  let OccupancySensorType;
  ((OccupancySensorType2) => {
    OccupancySensorType2[OccupancySensorType2["Pir"] = 0] = "Pir";
    OccupancySensorType2[OccupancySensorType2["Ultrasonic"] = 1] = "Ultrasonic";
    OccupancySensorType2[OccupancySensorType2["PirAndUltrasonic"] = 2] = "PirAndUltrasonic";
    OccupancySensorType2[OccupancySensorType2["PhysicalContact"] = 3] = "PhysicalContact";
  })(OccupancySensorType = OccupancySensing2.OccupancySensorType || (OccupancySensing2.OccupancySensorType = {}));
  OccupancySensing2.OccupancySensorTypeBitmap = {
    /**
     * Indicates a passive infrared sensor.
     */
    pir: (0, import_BitmapSchema.BitFlag)(0),
    /**
     * Indicates a ultrasonic sensor.
     */
    ultrasonic: (0, import_BitmapSchema.BitFlag)(1),
    /**
     * Indicates a physical contact sensor.
     */
    physicalContact: (0, import_BitmapSchema.BitFlag)(2)
  };
  OccupancySensing2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 1030,
    name: "OccupancySensing",
    revision: 3,
    attributes: {
      /**
       * The Occupancy attribute indicates the status of occupancy.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.6.1
       */
      occupancy: (0, import_Cluster.Attribute)(0, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, OccupancySensing2.OccupancyBitmap)),
      /**
       * The OccupancySensorType attribute specifies the type of the occupancy sensor.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.6.2
       */
      occupancySensorType: (0, import_Cluster.Attribute)(1, (0, import_TlvNumber.TlvEnum)()),
      /**
       * The OccupancySensorTypeBitmap attribute specifies the types of the occupancy sensor. A ‘1’ in each bit
       * position indicates the capability is implemented.
       *
       * The value of the OccupancySensorTypeBitmap attribute and the OccupancySensorType attribute shall be
       * aligned as defined below.
       *
       * Table 32. Mapping between OccupancySensorTypeBitmap and OccupancySensorType Attributes
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.6.3
       */
      occupancySensorTypeBitmap: (0, import_Cluster.Attribute)(2, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, OccupancySensing2.OccupancySensorTypeBitmap)),
      /**
       * The PIROccupiedToUnoccupiedDelay attribute specifies the time delay, in seconds, before the PIR sensor
       * changes to its unoccupied state after the last detection of movement in the sensed area.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.7.1
       */
      pirOccupiedToUnoccupiedDelay: (0, import_Cluster.OptionalWritableAttribute)(
        16,
        import_TlvNumber.TlvUInt16,
        { default: 0, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The PIRUnoccupiedToOccupiedDelay attribute specifies the time delay, in seconds, before the PIR sensor
       * changes to its occupied state after the detection of movement in the sensed area. This attribute is
       * mandatory if the PIRUnoccupiedToOccupiedThreshold attribute is implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.7.2
       */
      pirUnoccupiedToOccupiedDelay: (0, import_Cluster.OptionalWritableAttribute)(
        17,
        import_TlvNumber.TlvUInt16,
        { default: 0, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The PIRUnoccupiedToOccupiedThreshold attribute specifies the number of movement detection events that
       * must occur in the period PIRUnoccupiedToOccupiedDelay, before the PIR sensor changes to its occupied
       * state. This attribute is mandatory if the PIRUnoccupiedToOccupiedDelay attribute is implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.7.3
       */
      pirUnoccupiedToOccupiedThreshold: (0, import_Cluster.OptionalWritableAttribute)(
        18,
        import_TlvNumber.TlvUInt8.bound({ min: 1, max: 254 }),
        { default: 1, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The UltrasonicOccupiedToUnoccupiedDelay attribute and specifies the time delay, in seconds, before the
       * Ultrasonic sensor changes to its unoccupied state after the last detection of movement in the sensed
       * area.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.8.1
       */
      ultrasonicOccupiedToUnoccupiedDelay: (0, import_Cluster.OptionalWritableAttribute)(
        32,
        import_TlvNumber.TlvUInt16,
        { default: 0, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The UltrasonicUnoccupiedToOccupiedDelay attribute and specifies the time delay, in seconds, before the
       * Ultrasonic sensor changes to its occupied state after the detection of movement in the sensed area. This
       * attribute is mandatory if the UltrasonicUnoccupiedToOccupiedThreshold attribute is implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.8.2
       */
      ultrasonicUnoccupiedToOccupiedDelay: (0, import_Cluster.OptionalWritableAttribute)(
        33,
        import_TlvNumber.TlvUInt16,
        { default: 0, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The UltrasonicUnoccupiedToOccupiedThreshold attribute specifies the number of movement detection events
       * that must occur in the period UltrasonicUnoccupiedToOccupiedDelay, before the Ultrasonic sensor changes
       * to its occupied state. This attribute is mandatory if the UltrasonicUnoccupiedToOccupiedDelay attribute
       * is implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.8.3
       */
      ultrasonicUnoccupiedToOccupiedThreshold: (0, import_Cluster.OptionalWritableAttribute)(
        34,
        import_TlvNumber.TlvUInt8.bound({ min: 1, max: 254 }),
        { default: 1, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The PhysicalContactOccupiedToUnoccupiedDelay attribute specifies the time delay, in seconds, before the
       * physical contact occupancy sensor changes to its unoccupied state after detecting the unoccupied event.
       * The null value indicates that the sensor does not report occupied to unoccupied transition.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.9.1
       */
      physicalContactOccupiedToUnoccupiedDelay: (0, import_Cluster.OptionalWritableAttribute)(
        48,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16),
        { default: 0, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The PhysicalContactUnoccupiedToOccupiedDelay attribute specifies the time delay, in seconds, before the
       * physical contact sensor changes to its occupied state after the detection of the occupied event.
       *
       * The null value indicates that the sensor does not report unoccupied to occupied transition.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.9.2
       */
      physicalContactUnoccupiedToOccupiedDelay: (0, import_Cluster.OptionalWritableAttribute)(
        49,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16),
        { default: 0, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The PhysicalContactUnoccupiedToOccupiedThreshold attribute specifies the number of movement detection
       * events that must occur in the period PhysicalContactUnoccupiedToOccupiedDelay, before the PIR sensor
       * changes to its occupied state. This attribute is mandatory if the
       * PhysicalContactUnoccupiedToOccupiedDelay attribute is implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.9.3
       */
      physicalContactUnoccupiedToOccupiedThreshold: (0, import_Cluster.OptionalWritableAttribute)(
        50,
        import_TlvNumber.TlvUInt8.bound({ min: 1, max: 254 }),
        { default: 1, writeAcl: import_Cluster.AccessLevel.Manage }
      )
    }
  });
})(OccupancySensing || (OccupancySensing = {}));
const OccupancySensingCluster = OccupancySensing.Cluster;
