// @project-chip/matter.js/dist/cjs/cluster/definitions/OnOffSwitchConfigurationCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OnOffSwitchConfigurationCluster_exports = {};
__export(OnOffSwitchConfigurationCluster_exports, {
  OnOffSwitchConfiguration: () => OnOffSwitchConfiguration,
  OnOffSwitchConfigurationCluster: () => OnOffSwitchConfigurationCluster
});
module.exports = __toCommonJS(OnOffSwitchConfigurationCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var OnOffSwitchConfiguration;
((OnOffSwitchConfiguration2) => {
  let SwitchType;
  ((SwitchType2) => {
    SwitchType2[SwitchType2["Toggle"] = 0] = "Toggle";
    SwitchType2[SwitchType2["Momentary"] = 1] = "Momentary";
    SwitchType2[SwitchType2["MultiFunction"] = 2] = "MultiFunction";
  })(SwitchType = OnOffSwitchConfiguration2.SwitchType || (OnOffSwitchConfiguration2.SwitchType = {}));
  let SwitchActions;
  ((SwitchActions2) => {
    SwitchActions2[SwitchActions2["On"] = 0] = "On";
    SwitchActions2[SwitchActions2["Off"] = 1] = "Off";
    SwitchActions2[SwitchActions2["Toggle"] = 2] = "Toggle";
  })(SwitchActions = OnOffSwitchConfiguration2.SwitchActions || (OnOffSwitchConfiguration2.SwitchActions = {}));
  OnOffSwitchConfiguration2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 7,
    name: "OnOffSwitchConfiguration",
    revision: 1,
    attributes: {
      switchType: (0, import_Cluster.Attribute)(0, (0, import_TlvNumber.TlvEnum)()),
      switchActions: (0, import_Cluster.WritableAttribute)(16, (0, import_TlvNumber.TlvEnum)(), { default: 0 /* On */ })
    }
  });
})(OnOffSwitchConfiguration || (OnOffSwitchConfiguration = {}));
const OnOffSwitchConfigurationCluster = OnOffSwitchConfiguration.Cluster;
