// @project-chip/matter.js/dist/cjs/cluster/definitions/OnOffCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OnOffCluster_exports = {};
__export(OnOffCluster_exports, {
  OnOff: () => OnOff,
  OnOffCluster: () => OnOffCluster
});
module.exports = __toCommonJS(OnOffCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var OnOff;
((OnOff2) => {
  let StartUpOnOff;
  ((StartUpOnOff2) => {
    StartUpOnOff2[StartUpOnOff2["Off"] = 0] = "Off";
    StartUpOnOff2[StartUpOnOff2["On"] = 1] = "On";
    StartUpOnOff2[StartUpOnOff2["Toggle"] = 2] = "Toggle";
  })(StartUpOnOff = OnOff2.StartUpOnOff || (OnOff2.StartUpOnOff = {}));
  let OnOffEffectIdentifier;
  ((OnOffEffectIdentifier2) => {
    OnOffEffectIdentifier2[OnOffEffectIdentifier2["DelayedAllOff"] = 0] = "DelayedAllOff";
    OnOffEffectIdentifier2[OnOffEffectIdentifier2["DyingLight"] = 1] = "DyingLight";
  })(OnOffEffectIdentifier = OnOff2.OnOffEffectIdentifier || (OnOff2.OnOffEffectIdentifier = {}));
  OnOff2.TlvOffWithEffectRequest = (0, import_TlvObject.TlvObject)({
    /**
     * The EffectIdentifier field specifies the fading effect to use when turning the device off. This field shall
     * contain one of the non-reserved values listed in Values of the EffectIdentifier Field of the OffWithEffect
     * Command.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.4.1
     */
    effectIdentifier: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The EffectVariant field is used to indicate which variant of the effect, indicated in the EffectIdentifier
     * field, SHOULD be triggered. If the server does not support the given variant, it shall use the default
     * variant. This field is dependent on the value of the EffectIdentifier field and shall contain one of the
     * non-reserved values listed in Values of the EffectVariant Field of the OffWithEffect Command.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.4.2
     */
    effectVariant: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8)
  });
  OnOff2.OnOffControl = { acceptOnlyWhenOn: (0, import_BitmapSchema.BitFlag)(0), reserved: (0, import_BitmapSchema.BitField)(1, 7) };
  OnOff2.TlvOnWithTimedOffRequest = (0, import_TlvObject.TlvObject)({
    /**
     * The OnOffControl field contains information on how the server is to be operated. This field shall be
     * formatted as illustrated in Format of the OnOffControl Field of the OnWithTimedOff Command.
     *
     * The AcceptOnlyWhenOn sub-field is 1 bit in length and specifies whether the OnWithTimedOff command is to be
     * processed unconditionally or only when the OnOff attribute is equal to TRUE. If this sub-field is set to 1,
     * the OnWithTimedOff command shall only be accepted if the OnOff attribute is equal to TRUE. If this sub-field
     * is set to 0, the OnWithTimedOff command shall be processed unconditionally.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.6.1
     */
    onOffControl: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, OnOff2.OnOffControl)),
    /**
     * The OnTime field is used to adjust the value of the OnTime attribute.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.6.2
     */
    onTime: (0, import_TlvObject.TlvField)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16)),
    /**
     * The OffWaitTime field is used to adjust the value of the OffWaitTime attribute.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.6.3
     */
    offWaitTime: (0, import_TlvObject.TlvField)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16))
  });
  let Feature;
  ((Feature2) => {
    Feature2["LevelControlForLighting"] = "LevelControlForLighting";
  })(Feature = OnOff2.Feature || (OnOff2.Feature = {}));
  OnOff2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 6,
    name: "OnOff",
    revision: 4,
    features: {
      /**
       * LevelControlForLighting
       *
       * Behavior that supports lighting applications.
       */
      levelControlForLighting: (0, import_BitmapSchema.BitFlag)(0)
    },
    attributes: {
      /**
       * The OnOff attribute indicates whether the device type implemented on the endpoint is turned off or
       * turned on, in these cases the value of the OnOff attribute equals FALSE, or TRUE respectively.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.6.1
       */
      onOff: (0, import_Cluster.Attribute)(0, import_TlvBoolean.TlvBoolean, { scene: true, persistent: true, default: true })
    },
    commands: {
      /**
       * This command does not have any data fields.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.1
       */
      off: (0, import_Cluster.Command)(0, import_TlvNoArguments.TlvNoArguments, 0, import_Cluster.TlvNoResponse),
      /**
       * This command does not have any data fields.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.2
       */
      on: (0, import_Cluster.Command)(1, import_TlvNoArguments.TlvNoArguments, 1, import_Cluster.TlvNoResponse),
      /**
       * This command does not have any data fields.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.3
       */
      toggle: (0, import_Cluster.Command)(2, import_TlvNoArguments.TlvNoArguments, 2, import_Cluster.TlvNoResponse)
    }
  });
  OnOff2.LevelControlForLightingComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * In order to support the use case where the user gets back the last setting of a set of devices (e.g.
       * level settings for lights), a global scene is introduced which is stored when the devices are turned off
       * and recalled when the devices are turned on. The global scene is defined as the scene that is stored
       * with group identifier 0 and scene identifier 0.
       *
       * The GlobalSceneControl attribute is defined in order to prevent a second Off command storing the
       * all-devices-off situation as a global scene, and to prevent a second On command destroying the current
       * settings by going back to the global scene.
       *
       * The GlobalSceneControl attribute shall be set to TRUE after the reception of a command which causes the
       * OnOff attribute to be set to TRUE, such as a standard On command, a MoveToLevel(WithOnOff) command, a
       * RecallScene command or a OnWithRecallGlobalScene command (see OnWithRecallGlobalScene Command).
       *
       * The GlobalSceneControl attribute is set to FALSE after reception of a OffWithEffect command.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.6.2
       */
      globalSceneControl: (0, import_Cluster.Attribute)(16384, import_TlvBoolean.TlvBoolean, { default: true }),
      /**
       * The OnTime attribute specifies the length of time (in 1/10ths second) that the ‘On’ state shall be
       * maintained before automatically transitioning to the ‘Off’ state when using the OnWithTimedOff command.
       * This attribute can be written at any time, but writing a value only has effect when in the ‘Timed On’
       * state. See OnWithTimedOff Command for more details.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.6.3
       */
      onTime: (0, import_Cluster.WritableAttribute)(16385, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16), { default: 0 }),
      /**
       * The OffWaitTime attribute specifies the length of time (in 1/10ths second) that the ‘Off’ state shall be
       * guarded to prevent another OnWithTimedOff command turning the server back to its ‘On’ state (e.g., when
       * leaving a room, the lights are turned off but an occupancy sensor detects the leaving person and
       * attempts to turn the lights back on). This attribute can be written at any time, but writing a value
       * only has an effect when in the ‘Timed On’ state followed by a transition to the ‘Delayed Off' state, or
       * in the ‘Delayed Off’ state. See OnWithTimedOff Command for more details.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.6.4
       */
      offWaitTime: (0, import_Cluster.WritableAttribute)(16386, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16), { default: 0 }),
      /**
       * The StartUpOnOff attribute shall define the desired startup behavior of a device when it is supplied
       * with power and this state shall be reflected in the OnOff attribute. If the value is null, the OnOff
       * attribute is set to its previous value. Otherwise, the behavior is defined in the table defining
       * StartUpOnOffEnum.
       *
       * This behavior does not apply to reboots associated with OTA. After an OTA restart, the OnOff attribute
       * shall return to its value prior to the restart.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.6.5
       */
      startUpOnOff: (0, import_Cluster.WritableAttribute)(
        16387,
        (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)()),
        { persistent: true, writeAcl: import_Cluster.AccessLevel.Manage }
      )
    },
    commands: {
      /**
       * The OffWithEffect command allows devices to be turned off using enhanced ways of fading.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.4
       */
      offWithEffect: (0, import_Cluster.Command)(64, OnOff2.TlvOffWithEffectRequest, 64, import_Cluster.TlvNoResponse),
      /**
       * The OnWithRecallGlobalScene command allows the recall of the settings when the device was turned off.
       *
       * The OnWithRecallGlobalScene command shall have no parameters.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.5
       */
      onWithRecallGlobalScene: (0, import_Cluster.Command)(65, import_TlvNoArguments.TlvNoArguments, 65, import_Cluster.TlvNoResponse),
      /**
       * The OnWithTimedOff command allows devices to be turned on for a specific duration with a guarded off
       * duration so that SHOULD the device be subsequently turned off, further OnWithTimedOff commands, received
       * during this time, are prevented from turning the devices back on. Further
       *
       * OnWithTimedOff commands received while the server is turned on, will update the period that the device
       * is turned on.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.5.7.6
       */
      onWithTimedOff: (0, import_Cluster.Command)(66, OnOff2.TlvOnWithTimedOffRequest, 66, import_Cluster.TlvNoResponse)
    }
  });
  OnOff2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    OnOff2.Base,
    /**
     * Use this factory method to create an OnOff cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns an OnOff cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...OnOff2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(OnOff2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, OnOff2.LevelControlForLightingComponent, { levelControlForLighting: true });
      return cluster;
    }
  );
  const LT = { levelControlForLighting: true };
  OnOff2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: OnOff2.Cluster.id,
    name: OnOff2.Cluster.name,
    revision: OnOff2.Cluster.revision,
    features: OnOff2.Cluster.features,
    attributes: {
      ...OnOff2.Cluster.attributes,
      globalSceneControl: import_ClusterFactory.ClusterFactory.AsConditional(
        OnOff2.LevelControlForLightingComponent.attributes.globalSceneControl,
        { mandatoryIf: [LT] }
      ),
      onTime: import_ClusterFactory.ClusterFactory.AsConditional(
        OnOff2.LevelControlForLightingComponent.attributes.onTime,
        { mandatoryIf: [LT] }
      ),
      offWaitTime: import_ClusterFactory.ClusterFactory.AsConditional(
        OnOff2.LevelControlForLightingComponent.attributes.offWaitTime,
        { mandatoryIf: [LT] }
      ),
      startUpOnOff: import_ClusterFactory.ClusterFactory.AsConditional(
        OnOff2.LevelControlForLightingComponent.attributes.startUpOnOff,
        { mandatoryIf: [LT] }
      )
    },
    commands: {
      ...OnOff2.Cluster.commands,
      offWithEffect: import_ClusterFactory.ClusterFactory.AsConditional(
        OnOff2.LevelControlForLightingComponent.commands.offWithEffect,
        { mandatoryIf: [LT] }
      ),
      onWithRecallGlobalScene: import_ClusterFactory.ClusterFactory.AsConditional(
        OnOff2.LevelControlForLightingComponent.commands.onWithRecallGlobalScene,
        { mandatoryIf: [LT] }
      ),
      onWithTimedOff: import_ClusterFactory.ClusterFactory.AsConditional(
        OnOff2.LevelControlForLightingComponent.commands.onWithTimedOff,
        { mandatoryIf: [LT] }
      )
    }
  });
})(OnOff || (OnOff = {}));
const OnOffCluster = OnOff.Cluster;
