// @project-chip/matter.js/dist/cjs/cluster/definitions/OtaSoftwareUpdateRequestorCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OtaSoftwareUpdateRequestorCluster_exports = {};
__export(OtaSoftwareUpdateRequestorCluster_exports, {
  OtaSoftwareUpdateRequestor: () => OtaSoftwareUpdateRequestor,
  OtaSoftwareUpdateRequestorCluster: () => OtaSoftwareUpdateRequestorCluster
});
module.exports = __toCommonJS(OtaSoftwareUpdateRequestorCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_NodeId = require("../../datatype/NodeId.js");
var import_EndpointNumber = require("../../datatype/EndpointNumber.js");
var import_FabricIndex = require("../../datatype/FabricIndex.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_VendorId = require("../../datatype/VendorId.js");
var import_TlvString = require("../../tlv/TlvString.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var OtaSoftwareUpdateRequestor;
((OtaSoftwareUpdateRequestor2) => {
  OtaSoftwareUpdateRequestor2.TlvProviderLocationStruct = (0, import_TlvObject.TlvObject)({
    providerNodeId: (0, import_TlvObject.TlvField)(1, import_NodeId.TlvNodeId),
    endpoint: (0, import_TlvObject.TlvField)(2, import_EndpointNumber.TlvEndpointNumber),
    fabricIndex: (0, import_TlvObject.TlvField)(254, import_FabricIndex.TlvFabricIndex)
  });
  let UpdateState;
  ((UpdateState2) => {
    UpdateState2[UpdateState2["Unknown"] = 0] = "Unknown";
    UpdateState2[UpdateState2["Idle"] = 1] = "Idle";
    UpdateState2[UpdateState2["Querying"] = 2] = "Querying";
    UpdateState2[UpdateState2["DelayedOnQuery"] = 3] = "DelayedOnQuery";
    UpdateState2[UpdateState2["Downloading"] = 4] = "Downloading";
    UpdateState2[UpdateState2["Applying"] = 5] = "Applying";
    UpdateState2[UpdateState2["DelayedOnApply"] = 6] = "DelayedOnApply";
    UpdateState2[UpdateState2["RollingBack"] = 7] = "RollingBack";
    UpdateState2[UpdateState2["DelayedOnUserConsent"] = 8] = "DelayedOnUserConsent";
  })(UpdateState = OtaSoftwareUpdateRequestor2.UpdateState || (OtaSoftwareUpdateRequestor2.UpdateState = {}));
  let AnnouncementReason;
  ((AnnouncementReason2) => {
    AnnouncementReason2[AnnouncementReason2["SimpleAnnouncement"] = 0] = "SimpleAnnouncement";
    AnnouncementReason2[AnnouncementReason2["UpdateAvailable"] = 1] = "UpdateAvailable";
    AnnouncementReason2[AnnouncementReason2["UrgentUpdateAvailable"] = 2] = "UrgentUpdateAvailable";
  })(AnnouncementReason = OtaSoftwareUpdateRequestor2.AnnouncementReason || (OtaSoftwareUpdateRequestor2.AnnouncementReason = {}));
  OtaSoftwareUpdateRequestor2.TlvAnnounceOtaProviderRequest = (0, import_TlvObject.TlvObject)({
    providerNodeId: (0, import_TlvObject.TlvField)(0, import_NodeId.TlvNodeId),
    vendorId: (0, import_TlvObject.TlvField)(1, import_VendorId.TlvVendorId),
    announcementReason: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvEnum)()),
    metadataForNode: (0, import_TlvObject.TlvOptionalField)(3, import_TlvString.TlvByteString.bound({ maxLength: 512 })),
    endpoint: (0, import_TlvObject.TlvField)(4, import_EndpointNumber.TlvEndpointNumber),
    fabricIndex: (0, import_TlvObject.TlvField)(254, import_FabricIndex.TlvFabricIndex)
  });
  let ChangeReason;
  ((ChangeReason2) => {
    ChangeReason2[ChangeReason2["Unknown"] = 0] = "Unknown";
    ChangeReason2[ChangeReason2["Success"] = 1] = "Success";
    ChangeReason2[ChangeReason2["Failure"] = 2] = "Failure";
    ChangeReason2[ChangeReason2["TimeOut"] = 3] = "TimeOut";
    ChangeReason2[ChangeReason2["DelayByProvider"] = 4] = "DelayByProvider";
  })(ChangeReason = OtaSoftwareUpdateRequestor2.ChangeReason || (OtaSoftwareUpdateRequestor2.ChangeReason = {}));
  OtaSoftwareUpdateRequestor2.TlvStateTransitionEvent = (0, import_TlvObject.TlvObject)({
    previousState: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    newState: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvEnum)()),
    reason: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvEnum)()),
    targetSoftwareVersion: (0, import_TlvObject.TlvField)(3, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt32))
  });
  OtaSoftwareUpdateRequestor2.TlvVersionAppliedEvent = (0, import_TlvObject.TlvObject)({
    softwareVersion: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt32),
    productId: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16)
  });
  OtaSoftwareUpdateRequestor2.TlvDownloadErrorEvent = (0, import_TlvObject.TlvObject)({
    softwareVersion: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt32),
    bytesDownloaded: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt64),
    progressPercent: (0, import_TlvObject.TlvField)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ max: 100 }))),
    platformCode: (0, import_TlvObject.TlvField)(3, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt64))
  });
  OtaSoftwareUpdateRequestor2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 42,
    name: "OtaSoftwareUpdateRequestor",
    revision: 1,
    attributes: {
      /**
       * This field is a list of ProviderLocationStruct whose entries shall be set by Administrators, either
       * during Commissioning or at a later time, to set the Provider Location for the default OTA Provider Node
       * to use for software updates on a given Fabric.
       *
       * There shall NOT be more than one entry per Fabric. On a list update that would introduce more than one
       * entry per fabric, the write shall fail with CONSTRAINT_ERROR status code.
       *
       * Provider Locations obtained using the AnnounceOTAProvider command shall NOT overwrite values set in the
       * DefaultOTAProviders attribute.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.5.1
       */
      defaultOtaProviders: (0, import_Cluster.WritableFabricScopedAttribute)(
        0,
        (0, import_TlvArray.TlvArray)(OtaSoftwareUpdateRequestor2.TlvProviderLocationStruct),
        { default: [], writeAcl: import_Cluster.AccessLevel.Administer }
      ),
      /**
       * This field shall be set to True if the OTA Requestor is currently able to be updated. Otherwise, it
       * shall be set to False in case of any condition preventing update being possible, such as insufficient
       * capacity of an internal battery. This field is merely informational for diagnostics purposes and shall
       * NOT affect the responses provided by an OTA Provider to an OTA Requestor.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.5.2
       */
      updatePossible: (0, import_Cluster.Attribute)(1, import_TlvBoolean.TlvBoolean, { default: true }),
      /**
       * This field shall reflect the current state of the OTA Requestor with regards to obtaining software
       * updates. See Section 11.19.7.4.2, “UpdateStateEnum” for possible values.
       *
       * This field SHOULD be updated in a timely manner whenever OTA Requestor internal state updates.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.5.3
       */
      updateState: (0, import_Cluster.Attribute)(2, (0, import_TlvNumber.TlvEnum)(), { default: 0 /* Unknown */ }),
      /**
       * This field shall reflect the percentage value of progress, relative to the current UpdateState, if
       * applicable to the state.
       *
       * The value of this field shall be null if a progress indication does not apply to the current state.
       *
       * A value of 0 shall indicate that the beginning has occurred. A value of 100 shall indicate completion.
       *
       * This field may be updated infrequently. Some care SHOULD be taken by Nodes to avoid over- reporting
       * progress when this attribute is part of a subscription.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.5.4
       */
      updateStateProgress: (0, import_Cluster.Attribute)(3, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ max: 100 })), { default: null })
    },
    commands: {
      /**
       * This command may be invoked by Administrators to announce the presence of a particular OTA Provider.
       *
       * This command shall be scoped to the accessing fabric.
       *
       * If the accessing fabric index is 0, this command shall fail with an UNSUPPORTED_ACCESS status code.
       *
       * This field shall contain the Node ID of a Node implementing the OTA Provider cluster server, on the
       * accessing fabric.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.6.1
       */
      announceOtaProvider: (0, import_Cluster.OptionalCommand)(
        0,
        OtaSoftwareUpdateRequestor2.TlvAnnounceOtaProviderRequest,
        0,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Administer }
      )
    },
    events: {
      /**
       * This event shall be generated when a change of the UpdateState attribute occurs due to an OTA Requestor
       * moving through the states necessary to query for updates.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.7.1
       */
      stateTransition: (0, import_Cluster.Event)(0, import_Cluster.EventPriority.Info, OtaSoftwareUpdateRequestor2.TlvStateTransitionEvent),
      /**
       * This event shall be generated whenever a new version starts executing after being applied due to a
       * software update. This event SHOULD be generated even if a software update was done using means outside
       * of this cluster.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.7.6
       */
      versionApplied: (0, import_Cluster.Event)(1, import_Cluster.EventPriority.Critical, OtaSoftwareUpdateRequestor2.TlvVersionAppliedEvent),
      /**
       * This event shall be generated whenever an error occurs during OTA Requestor download operation.
       *
       * This field shall be set to the value of the SoftwareVersion being downloaded, matching the
       * SoftwareVersion field of the QueryImageResponse that caused the failing download to take place.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.7.9
       */
      downloadError: (0, import_Cluster.Event)(2, import_Cluster.EventPriority.Info, OtaSoftwareUpdateRequestor2.TlvDownloadErrorEvent)
    }
  });
})(OtaSoftwareUpdateRequestor || (OtaSoftwareUpdateRequestor = {}));
const OtaSoftwareUpdateRequestorCluster = OtaSoftwareUpdateRequestor.Cluster;
