// @project-chip/matter.js/dist/cjs/cluster/definitions/OtaSoftwareUpdateProviderCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OtaSoftwareUpdateProviderCluster_exports = {};
__export(OtaSoftwareUpdateProviderCluster_exports, {
  OtaSoftwareUpdateProvider: () => OtaSoftwareUpdateProvider,
  OtaSoftwareUpdateProviderCluster: () => OtaSoftwareUpdateProviderCluster
});
module.exports = __toCommonJS(OtaSoftwareUpdateProviderCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_VendorId = require("../../datatype/VendorId.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var OtaSoftwareUpdateProvider;
((OtaSoftwareUpdateProvider2) => {
  let DownloadProtocol;
  ((DownloadProtocol2) => {
    DownloadProtocol2[DownloadProtocol2["BdxSynchronous"] = 0] = "BdxSynchronous";
    DownloadProtocol2[DownloadProtocol2["BdxAsynchronous"] = 1] = "BdxAsynchronous";
    DownloadProtocol2[DownloadProtocol2["Https"] = 2] = "Https";
    DownloadProtocol2[DownloadProtocol2["VendorSpecific"] = 3] = "VendorSpecific";
  })(DownloadProtocol = OtaSoftwareUpdateProvider2.DownloadProtocol || (OtaSoftwareUpdateProvider2.DownloadProtocol = {}));
  OtaSoftwareUpdateProvider2.TlvQueryImageRequest = (0, import_TlvObject.TlvObject)({
    vendorId: (0, import_TlvObject.TlvField)(0, import_VendorId.TlvVendorId),
    productId: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16),
    softwareVersion: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt32),
    protocolsSupported: (0, import_TlvObject.TlvField)(3, (0, import_TlvArray.TlvArray)((0, import_TlvNumber.TlvEnum)(), { maxLength: 8 })),
    hardwareVersion: (0, import_TlvObject.TlvOptionalField)(4, import_TlvNumber.TlvUInt16),
    location: (0, import_TlvObject.TlvOptionalField)(5, import_TlvString.TlvString.bound({ length: 2 })),
    requestorCanConsent: (0, import_TlvObject.TlvOptionalField)(6, import_TlvBoolean.TlvBoolean),
    metadataForProvider: (0, import_TlvObject.TlvOptionalField)(7, import_TlvString.TlvByteString.bound({ maxLength: 512 }))
  });
  let Status;
  ((Status2) => {
    Status2[Status2["UpdateAvailable"] = 0] = "UpdateAvailable";
    Status2[Status2["Busy"] = 1] = "Busy";
    Status2[Status2["NotAvailable"] = 2] = "NotAvailable";
    Status2[Status2["DownloadProtocolNotSupported"] = 3] = "DownloadProtocolNotSupported";
  })(Status = OtaSoftwareUpdateProvider2.Status || (OtaSoftwareUpdateProvider2.Status = {}));
  OtaSoftwareUpdateProvider2.TlvQueryImageResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    delayedActionTime: (0, import_TlvObject.TlvOptionalField)(1, import_TlvNumber.TlvUInt32),
    imageUri: (0, import_TlvObject.TlvOptionalField)(2, import_TlvString.TlvString.bound({ maxLength: 256 })),
    softwareVersion: (0, import_TlvObject.TlvOptionalField)(3, import_TlvNumber.TlvUInt32),
    softwareVersionString: (0, import_TlvObject.TlvOptionalField)(4, import_TlvString.TlvString.bound({ minLength: 1, maxLength: 64 })),
    updateToken: (0, import_TlvObject.TlvOptionalField)(5, import_TlvString.TlvByteString.bound({ minLength: 8, maxLength: 32 })),
    userConsentNeeded: (0, import_TlvObject.TlvOptionalField)(6, import_TlvBoolean.TlvBoolean),
    metadataForRequestor: (0, import_TlvObject.TlvOptionalField)(7, import_TlvString.TlvByteString.bound({ maxLength: 512 }))
  });
  OtaSoftwareUpdateProvider2.TlvApplyUpdateRequestRequest = (0, import_TlvObject.TlvObject)({
    updateToken: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvByteString.bound({ minLength: 8, maxLength: 32 })),
    newVersion: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt32)
  });
  let ApplyUpdateAction;
  ((ApplyUpdateAction2) => {
    ApplyUpdateAction2[ApplyUpdateAction2["Proceed"] = 0] = "Proceed";
    ApplyUpdateAction2[ApplyUpdateAction2["AwaitNextAction"] = 1] = "AwaitNextAction";
    ApplyUpdateAction2[ApplyUpdateAction2["Discontinue"] = 2] = "Discontinue";
  })(ApplyUpdateAction = OtaSoftwareUpdateProvider2.ApplyUpdateAction || (OtaSoftwareUpdateProvider2.ApplyUpdateAction = {}));
  OtaSoftwareUpdateProvider2.TlvApplyUpdateResponse = (0, import_TlvObject.TlvObject)({
    action: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    delayedActionTime: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt32)
  });
  OtaSoftwareUpdateProvider2.TlvNotifyUpdateAppliedRequest = (0, import_TlvObject.TlvObject)({
    updateToken: (0, import_TlvObject.TlvField)(0, import_TlvString.TlvByteString.bound({ minLength: 8, maxLength: 32 })),
    softwareVersion: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt32)
  });
  OtaSoftwareUpdateProvider2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 41,
    name: "OtaSoftwareUpdateProvider",
    revision: 1,
    commands: {
      /**
       * Upon receipt, this command shall trigger an attempt to find an updated Software Image by the OTA
       * Provider to match the OTA Requestor’s constraints provided in the payload fields.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.6.5.1
       */
      queryImage: (0, import_Cluster.Command)(0, OtaSoftwareUpdateProvider2.TlvQueryImageRequest, 1, OtaSoftwareUpdateProvider2.TlvQueryImageResponse),
      /**
       * This field shall contain the UpdateToken as specified in Section 11.19.3.6.1, “UpdateToken usage”. This
       * field may be used by the OTA Provider to track minimal lifecycle state to allow finer-grained scheduling
       * of the application of Software Images by OTA Requestors.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.6.5.18
       */
      applyUpdateRequest: (0, import_Cluster.Command)(2, OtaSoftwareUpdateProvider2.TlvApplyUpdateRequestRequest, 3, OtaSoftwareUpdateProvider2.TlvApplyUpdateResponse),
      /**
       * This field shall contain the UpdateToken as specified in Section 11.19.3.6.1, “UpdateToken usage”.
       *
       * The SoftwareVersion included in the request payload shall provide the same value as the SoftwareVersion
       * attribute in the invoking OTA Requestor’s Basic Information Cluster, and SHOULD be consistent with the
       * value representing a new version running on the Node invoking the command.
       *
       * When Generated
       *
       * The NotifyUpdateApplied command SHOULD be invoked in the following two circumstances:
       *
       *   1. An OTA Requestor has just successfully applied a Software Image it had obtained from a previous
       *      QueryImage response.
       *
       *   2. An OTA Requestor has just successfully applied a Software Image it had obtained through means
       *      different than those of this Cluster.
       *
       * An OTA Provider may use the state of invocation of this command to help track the progress of update for
       * OTA Requestors it knows require a new OTA Software Image. However, due to the possibility that an OTA
       * Requestor may never come back (e.g. device removed from Fabric altogether, or a critical malfunction),
       * an OTA Provider shall NOT expect every OTA Requestor to invoke this command for correct operation of the
       * OTA Provider.
       *
       * This command shall be considered optional and shall not result in reduced availability of the OTA
       * Provider functionality if OTA Requestors never invoke this command.
       *
       * Effect on Receipt
       *
       * An OTA Provider receiving an invocation of this command may log it internally.
       *
       * On receiving this command, an OTA Provider may use the information to update its bookkeeping of cached
       * Software Images, or use it for other similar administrative purposes.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.6.5.22
       */
      notifyUpdateApplied: (0, import_Cluster.Command)(4, OtaSoftwareUpdateProvider2.TlvNotifyUpdateAppliedRequest, 4, import_Cluster.TlvNoResponse)
    }
  });
})(OtaSoftwareUpdateProvider || (OtaSoftwareUpdateProvider = {}));
const OtaSoftwareUpdateProviderCluster = OtaSoftwareUpdateProvider.Cluster;
