// @project-chip/matter.js/dist/cjs/cluster/definitions/PressureMeasurementCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PressureMeasurementCluster_exports = {};
__export(PressureMeasurementCluster_exports, {
  PressureMeasurement: () => PressureMeasurement,
  PressureMeasurementCluster: () => PressureMeasurementCluster
});
module.exports = __toCommonJS(PressureMeasurementCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var PressureMeasurement;
((PressureMeasurement2) => {
  let Feature;
  ((Feature2) => {
    Feature2["Extended"] = "Extended";
  })(Feature = PressureMeasurement2.Feature || (PressureMeasurement2.Feature = {}));
  PressureMeasurement2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 1027,
    name: "PressureMeasurement",
    revision: 3,
    features: {
      /**
       * Extended
       *
       * The cluster is capable of extended range and resolution
       */
      extended: (0, import_BitmapSchema.BitFlag)(0)
    },
    attributes: {
      /**
       * This attribute represents the pressure in kPa as follows:
       *
       * MeasuredValue = 10 x Pressure [kPa]
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.1
       */
      measuredValue: (0, import_Cluster.Attribute)(0, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16.bound({ min: -32767 }))),
      /**
       * This attribute indicates the minimum value of MeasuredValue that can be measured. See Measured Value for
       * more details.
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.2
       */
      minMeasuredValue: (0, import_Cluster.Attribute)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16.bound({ min: -32767 }))),
      /**
       * This attribute indicates the maximum value of MeasuredValue that can be measured. See Measured Value for
       * more details.
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.3
       */
      maxMeasuredValue: (0, import_Cluster.Attribute)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16)),
      /**
       * This attribute indicates the magnitude of the possible error that is associated with ScaledValue.
       *
       * See Measured Value.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.4
       */
      tolerance: (0, import_Cluster.OptionalAttribute)(3, import_TlvNumber.TlvUInt16.bound({ max: 2048 }), { default: 0 })
    }
  });
  PressureMeasurement2.ExtendedComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * ScaledValue represents the pressure in Pascals as follows:
       *
       * ScaledValue = 10Scale x Pressure [Pa]
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.5
       */
      scaledValue: (0, import_Cluster.Attribute)(16, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16.bound({ min: -32767 })), { default: 0 }),
      /**
       * The MinScaledValue attribute indicates the minimum value of ScaledValue that can be measured. The null
       * value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.6
       */
      minScaledValue: (0, import_Cluster.Attribute)(17, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16.bound({ min: -32767 })), { default: 0 }),
      /**
       * This attribute indicates the maximum value of ScaledValue that can be measured. MaxScaledValue shall be
       * greater than MinScaledValue.
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.7
       */
      maxScaledValue: (0, import_Cluster.Attribute)(18, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16), { default: 0 }),
      /**
       * This attribute indicates the magnitude of the possible error that is associated with ScaledValue. The
       * true value is located in the range
       *
       * (ScaledValue – ScaledTolerance) to (ScaledValue + ScaledTolerance).
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.8
       */
      scaledTolerance: (0, import_Cluster.OptionalAttribute)(19, import_TlvNumber.TlvUInt16.bound({ max: 2048 }), { default: 0 }),
      /**
       * This attribute indicates the base 10 exponent used to obtain ScaledValue (see ScaledValue Attribute).
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.9
       */
      scale: (0, import_Cluster.Attribute)(20, import_TlvNumber.TlvInt8.bound({ min: -127 }), { default: 0 })
    }
  });
  PressureMeasurement2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    PressureMeasurement2.Base,
    /**
     * Use this factory method to create a PressureMeasurement cluster with support for optional features. Include
     * each {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a PressureMeasurement cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...PressureMeasurement2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(PressureMeasurement2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, PressureMeasurement2.ExtendedComponent, { extended: true });
      return cluster;
    }
  );
  const EXT = { extended: true };
  PressureMeasurement2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: PressureMeasurement2.Cluster.id,
    name: PressureMeasurement2.Cluster.name,
    revision: PressureMeasurement2.Cluster.revision,
    features: PressureMeasurement2.Cluster.features,
    attributes: {
      ...PressureMeasurement2.Cluster.attributes,
      scaledValue: import_ClusterFactory.ClusterFactory.AsConditional(PressureMeasurement2.ExtendedComponent.attributes.scaledValue, { mandatoryIf: [EXT] }),
      minScaledValue: import_ClusterFactory.ClusterFactory.AsConditional(
        PressureMeasurement2.ExtendedComponent.attributes.minScaledValue,
        { mandatoryIf: [EXT] }
      ),
      maxScaledValue: import_ClusterFactory.ClusterFactory.AsConditional(
        PressureMeasurement2.ExtendedComponent.attributes.maxScaledValue,
        { mandatoryIf: [EXT] }
      ),
      scaledTolerance: import_ClusterFactory.ClusterFactory.AsConditional(
        PressureMeasurement2.ExtendedComponent.attributes.scaledTolerance,
        { optionalIf: [EXT] }
      ),
      scale: import_ClusterFactory.ClusterFactory.AsConditional(PressureMeasurement2.ExtendedComponent.attributes.scale, { mandatoryIf: [EXT] })
    }
  });
})(PressureMeasurement || (PressureMeasurement = {}));
const PressureMeasurementCluster = PressureMeasurement.Cluster;
