// @project-chip/matter.js/dist/cjs/cluster/definitions/ProxyDiscoveryCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ProxyDiscoveryCluster_exports = {};
__export(ProxyDiscoveryCluster_exports, {
  ProxyDiscovery: () => ProxyDiscovery,
  ProxyDiscoveryCluster: () => ProxyDiscoveryCluster
});
module.exports = __toCommonJS(ProxyDiscoveryCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ProxyDiscovery;
((ProxyDiscovery2) => {
  ProxyDiscovery2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 67,
    name: "ProxyDiscovery",
    revision: 1,
    commands: {
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.15.13.5
       */
      proxyDiscoverRequest: (0, import_Cluster.Command)(0, import_TlvNoArguments.TlvNoArguments, 0, import_Cluster.TlvNoResponse)
    }
  });
})(ProxyDiscovery || (ProxyDiscovery = {}));
const ProxyDiscoveryCluster = ProxyDiscovery.Cluster;
