// @project-chip/matter.js/dist/cjs/cluster/definitions/ProxyConfigurationCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ProxyConfigurationCluster_exports = {};
__export(ProxyConfigurationCluster_exports, {
  ProxyConfiguration: () => ProxyConfiguration,
  ProxyConfigurationCluster: () => ProxyConfigurationCluster
});
module.exports = __toCommonJS(ProxyConfigurationCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_NodeId = require("../../datatype/NodeId.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ProxyConfiguration;
((ProxyConfiguration2) => {
  ProxyConfiguration2.TlvConfigurationStruct = (0, import_TlvObject.TlvObject)({
    proxyAllNodes: (0, import_TlvObject.TlvField)(1, import_TlvBoolean.TlvBoolean),
    sourceList: (0, import_TlvObject.TlvField)(2, (0, import_TlvArray.TlvArray)(import_NodeId.TlvNodeId))
  });
  ProxyConfiguration2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 66,
    name: "ProxyConfiguration",
    revision: 1,
    attributes: {
      /**
       * List of proxy configurations. There shall NOT be multiple entries in this list for the same fabric.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.15.14.5.1
       */
      configurationList: (0, import_Cluster.WritableAttribute)(
        0,
        (0, import_TlvArray.TlvArray)(ProxyConfiguration2.TlvConfigurationStruct),
        { persistent: true, default: [] }
      )
    }
  });
})(ProxyConfiguration || (ProxyConfiguration = {}));
const ProxyConfigurationCluster = ProxyConfiguration.Cluster;
