// @project-chip/matter.js/dist/cjs/cluster/definitions/PulseWidthModulationCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PulseWidthModulationCluster_exports = {};
__export(PulseWidthModulationCluster_exports, {
  PulseWidthModulation: () => PulseWidthModulation,
  PulseWidthModulationCluster: () => PulseWidthModulationCluster
});
module.exports = __toCommonJS(PulseWidthModulationCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var PulseWidthModulation;
((PulseWidthModulation2) => {
  PulseWidthModulation2.Options = { executeIfOff: (0, import_BitmapSchema.BitFlag)(0), coupleColorTempToLevel: (0, import_BitmapSchema.BitFlag)(1) };
  PulseWidthModulation2.TlvMoveToLevelRequest = (0, import_TlvObject.TlvObject)({
    level: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8.bound({ max: 254 })),
    transitionTime: (0, import_TlvObject.TlvField)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16)),
    optionsMask: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, PulseWidthModulation2.Options)),
    optionsOverride: (0, import_TlvObject.TlvField)(3, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, PulseWidthModulation2.Options))
  });
  let MoveMode;
  ((MoveMode2) => {
    MoveMode2[MoveMode2["Up"] = 0] = "Up";
    MoveMode2[MoveMode2["Down"] = 1] = "Down";
  })(MoveMode = PulseWidthModulation2.MoveMode || (PulseWidthModulation2.MoveMode = {}));
  PulseWidthModulation2.TlvMoveRequest = (0, import_TlvObject.TlvObject)({
    /**
     * The MoveMode field shall be one of the non-reserved values in Values of the MoveMode Field.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.2.1
     */
    moveMode: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The Rate field specifies the rate of movement in units per second. The actual rate of movement SHOULD be as
     * close to this rate as the device is able. If the Rate field is equal to null, then the value in
     * DefaultMoveRate attribute shall be used. However, if the Rate field is equal to null and the DefaultMoveRate
     * attribute is not supported, or if the Rate field is equal to null and the value of the DefaultMoveRate
     * attribute is equal to null, then the device SHOULD move as fast as it is able. If the device is not able to
     * move at a variable rate, this field may be disregarded.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.2.2
     */
    rate: (0, import_TlvObject.TlvField)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8)),
    optionsMask: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, PulseWidthModulation2.Options)),
    optionsOverride: (0, import_TlvObject.TlvField)(3, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, PulseWidthModulation2.Options))
  });
  PulseWidthModulation2.TlvStepRequest = (0, import_TlvObject.TlvObject)({
    stepMode: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    stepSize: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8),
    transitionTime: (0, import_TlvObject.TlvField)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16)),
    optionsMask: (0, import_TlvObject.TlvField)(3, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, PulseWidthModulation2.Options)),
    optionsOverride: (0, import_TlvObject.TlvField)(4, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, PulseWidthModulation2.Options))
  });
  PulseWidthModulation2.TlvStopRequest = (0, import_TlvObject.TlvObject)({
    optionsMask: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, PulseWidthModulation2.Options)),
    optionsOverride: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, PulseWidthModulation2.Options))
  });
  PulseWidthModulation2.TlvMoveToClosestFrequencyRequest = (0, import_TlvObject.TlvObject)({ frequency: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16) });
  let Feature;
  ((Feature2) => {
    Feature2["OnOff"] = "OnOff";
    Feature2["Lighting"] = "Lighting";
    Feature2["Frequency"] = "Frequency";
  })(Feature = PulseWidthModulation2.Feature || (PulseWidthModulation2.Feature = {}));
  PulseWidthModulation2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 28,
    name: "PulseWidthModulation",
    revision: 5,
    features: {
      /**
       * OnOff
       *
       * Dependency with the On/Off cluster
       */
      onOff: (0, import_BitmapSchema.BitFlag)(0),
      /**
       * Lighting
       *
       * Behavior that supports lighting applications
       */
      lighting: (0, import_BitmapSchema.BitFlag)(1),
      /**
       * Frequency
       *
       * Supports frequency attributes and behavior. The Pulse Width Modulation cluster was created for frequency
       * control.
       */
      frequency: (0, import_BitmapSchema.BitFlag)(2)
    },
    attributes: {
      /**
       * The CurrentLevel attribute represents the current level of this device. The meaning of 'level' is device
       * dependent.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.1
       */
      currentLevel: (0, import_Cluster.Attribute)(
        0,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ max: 254 })),
        { scene: true, persistent: true, default: null }
      ),
      /**
       * The MinLevel attribute indicates the minimum value of CurrentLevel that is capable of being assigned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.3
       */
      minLevel: (0, import_Cluster.OptionalAttribute)(2, import_TlvNumber.TlvUInt8),
      /**
       * The MaxLevel attribute indicates the maximum value of CurrentLevel that is capable of being assigned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.4
       */
      maxLevel: (0, import_Cluster.OptionalAttribute)(3, import_TlvNumber.TlvUInt8.bound({ max: 254 }), { default: 254 }),
      /**
       * The Options attribute is meant to be changed only during commissioning. The Options attribute is a
       * bitmap that determines the default behavior of some cluster commands. Each command that is dependent on
       * the Options attribute shall first construct a temporary Options bitmap that is in effect during the
       * command processing. The temporary Options bitmap has the same format and meaning as the Options
       * attribute, but includes any bits that may be overridden by command fields.
       *
       * Below is the format and description of the Options attribute and temporary Options bitmap and the effect
       * on dependent commands.
       *
       * Table 19. Options Attribute
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.8
       */
      options: (0, import_Cluster.WritableAttribute)(15, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, PulseWidthModulation2.Options)),
      /**
       * The OnOffTransitionTime attribute represents the time taken to move to or from the target level when On
       * or Off commands are received by an On/Off cluster on the same endpoint. It is specified in 1/10ths of a
       * second.
       *
       * The actual time taken SHOULD be as close to OnOffTransitionTime as the device is able. Please note that
       * if the device is not able to move at a variable rate, the OnOffTransitionTime attribute SHOULD NOT be
       * implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.9
       */
      onOffTransitionTime: (0, import_Cluster.OptionalWritableAttribute)(16, import_TlvNumber.TlvUInt16, { default: 0 }),
      /**
       * The OnLevel attribute determines the value that the CurrentLevel attribute is set to when the OnOff
       * attribute of an On/Off cluster on the same endpoint is set to TRUE, as a result of processing an On/Off
       * cluster command. If the OnLevel attribute is not implemented, or is set to the null value, it has no
       * effect. For more details see Effect of On/Off Commands on the CurrentLevel Attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.10
       */
      onLevel: (0, import_Cluster.WritableAttribute)(17, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ max: 254 })), { default: null }),
      /**
       * The OnTransitionTime attribute represents the time taken to move the current level from the minimum
       * level to the maximum level when an On command is received by an On/Off cluster on the same endpoint. It
       * is specified in 10ths of a second. If this attribute is not implemented, or contains a null value, the
       * OnOffTransitionTime will be used instead.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.11
       */
      onTransitionTime: (0, import_Cluster.OptionalWritableAttribute)(18, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16), { default: null }),
      /**
       * The OffTransitionTime attribute represents the time taken to move the current level from the maximum
       * level to the minimum level when an Off command is received by an On/Off cluster on the same endpoint. It
       * is specified in 10ths of a second. If this attribute is not implemented, or contains a null value, the
       * OnOffTransitionTime will be used instead.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.12
       */
      offTransitionTime: (0, import_Cluster.OptionalWritableAttribute)(19, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16), { default: null }),
      /**
       * The DefaultMoveRate attribute determines the movement rate, in units per second, when a Move command is
       * received with a null value Rate parameter.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.13
       */
      defaultMoveRate: (0, import_Cluster.OptionalWritableAttribute)(20, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8))
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.1
       */
      moveToLevel: (0, import_Cluster.Command)(0, PulseWidthModulation2.TlvMoveToLevelRequest, 0, import_Cluster.TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.2
       */
      move: (0, import_Cluster.Command)(1, PulseWidthModulation2.TlvMoveRequest, 1, import_Cluster.TlvNoResponse),
      /**
       * The StepMode field shall be one of the non-reserved values in Values of the StepMode Field.
       *
       * The TransitionTime field specifies the time that shall be taken to perform the step, in tenths of a
       * second. A step is a change in the CurrentLevel of StepSize units. The actual time taken SHOULD be as
       * close to this as the device is able. If the TransitionTime field is equal to null, the device SHOULD
       * move as fast as it is able.
       *
       * If the device is not able to move at a variable rate, the TransitionTime field may be disregarded.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.3
       */
      step: (0, import_Cluster.Command)(2, PulseWidthModulation2.TlvStepRequest, 2, import_Cluster.TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.4
       */
      stop: (0, import_Cluster.Command)(3, PulseWidthModulation2.TlvStopRequest, 3, import_Cluster.TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6
       */
      moveToLevelWithOnOff: (0, import_Cluster.Command)(4, import_TlvNoArguments.TlvNoArguments, 4, import_Cluster.TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6
       */
      moveWithOnOff: (0, import_Cluster.Command)(5, import_TlvNoArguments.TlvNoArguments, 5, import_Cluster.TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6
       */
      stepWithOnOff: (0, import_Cluster.Command)(6, import_TlvNoArguments.TlvNoArguments, 6, import_Cluster.TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6
       */
      stopWithOnOff: (0, import_Cluster.Command)(7, import_TlvNoArguments.TlvNoArguments, 7, import_Cluster.TlvNoResponse)
    }
  });
  PulseWidthModulation2.LightingComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The RemainingTime attribute represents the time remaining until the current command is complete - it is
       * specified in 1/10ths of a second.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.2
       */
      remainingTime: (0, import_Cluster.Attribute)(1, import_TlvNumber.TlvUInt16, { default: 0 }),
      /**
       * The StartUpCurrentLevel attribute shall define the desired startup level for a device when it is
       * supplied with power and this level shall be reflected in the CurrentLevel attribute. The values of the
       * StartUpCurrentLevel attribute are listed below:
       *
       * Table 20. Values of the StartUpCurrentLevel attribute
       *
       * This behavior does not apply to reboots associated with OTA. After an OTA restart, the CurrentLevel
       * attribute shall return to its value prior to the restart.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.14
       */
      startUpCurrentLevel: (0, import_Cluster.WritableAttribute)(
        16384,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8),
        { persistent: true, writeAcl: import_Cluster.AccessLevel.Manage }
      )
    }
  });
  PulseWidthModulation2.FrequencyComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The CurrentFrequency attribute represents the frequency at which the device is at CurrentLevel. A
       * CurrentFrequency of 0 is unknown.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.5
       */
      currentFrequency: (0, import_Cluster.Attribute)(4, import_TlvNumber.TlvUInt16, { scene: true, default: 0 }),
      /**
       * The MinFrequency attribute indicates the minimum value of CurrentFrequency that is capable of being
       * assigned. MinFrequency shall be less than or equal to MaxFrequency. A value of 0 indicates undefined.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.6
       */
      minFrequency: (0, import_Cluster.Attribute)(5, import_TlvNumber.TlvUInt16, { default: 0 }),
      /**
       * The MaxFrequency attribute indicates the maximum value of CurrentFrequency that is capable of being
       * assigned. MaxFrequency shall be greater than or equal to MinFrequency. A value of 0 indicates undefined.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.7
       */
      maxFrequency: (0, import_Cluster.Attribute)(6, import_TlvNumber.TlvUInt16, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.5
       */
      moveToClosestFrequency: (0, import_Cluster.Command)(8, PulseWidthModulation2.TlvMoveToClosestFrequencyRequest, 8, import_Cluster.TlvNoResponse)
    }
  });
  PulseWidthModulation2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    { ...PulseWidthModulation2.Base, supportedFeatures: { onOff: true } },
    /**
     * Use this factory method to create a PulseWidthModulation cluster with support for optional features. Include
     * each {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a PulseWidthModulation cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...PulseWidthModulation2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(PulseWidthModulation2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, PulseWidthModulation2.LightingComponent, { lighting: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, PulseWidthModulation2.FrequencyComponent, { frequency: true });
      return cluster;
    }
  );
  const LT = { lighting: true };
  const FQ = { frequency: true };
  PulseWidthModulation2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: PulseWidthModulation2.Cluster.id,
    name: PulseWidthModulation2.Cluster.name,
    revision: PulseWidthModulation2.Cluster.revision,
    features: PulseWidthModulation2.Cluster.features,
    attributes: {
      ...PulseWidthModulation2.Cluster.attributes,
      remainingTime: import_ClusterFactory.ClusterFactory.AsConditional(
        PulseWidthModulation2.LightingComponent.attributes.remainingTime,
        { mandatoryIf: [LT] }
      ),
      currentFrequency: import_ClusterFactory.ClusterFactory.AsConditional(
        PulseWidthModulation2.FrequencyComponent.attributes.currentFrequency,
        { mandatoryIf: [FQ] }
      ),
      minFrequency: import_ClusterFactory.ClusterFactory.AsConditional(
        PulseWidthModulation2.FrequencyComponent.attributes.minFrequency,
        { mandatoryIf: [FQ] }
      ),
      maxFrequency: import_ClusterFactory.ClusterFactory.AsConditional(
        PulseWidthModulation2.FrequencyComponent.attributes.maxFrequency,
        { mandatoryIf: [FQ] }
      ),
      startUpCurrentLevel: import_ClusterFactory.ClusterFactory.AsConditional(
        PulseWidthModulation2.LightingComponent.attributes.startUpCurrentLevel,
        { mandatoryIf: [LT] }
      )
    },
    commands: {
      ...PulseWidthModulation2.Cluster.commands,
      moveToClosestFrequency: import_ClusterFactory.ClusterFactory.AsConditional(
        PulseWidthModulation2.FrequencyComponent.commands.moveToClosestFrequency,
        { mandatoryIf: [FQ] }
      )
    }
  });
})(PulseWidthModulation || (PulseWidthModulation = {}));
const PulseWidthModulationCluster = PulseWidthModulation.Cluster;
