// @project-chip/matter.js/dist/cjs/cluster/definitions/ScenesCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ScenesCluster_exports = {};
__export(ScenesCluster_exports, {
  Scenes: () => Scenes,
  ScenesCluster: () => ScenesCluster
});
module.exports = __toCommonJS(ScenesCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_GroupId = require("../../datatype/GroupId.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_NodeId = require("../../datatype/NodeId.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_ClusterId = require("../../datatype/ClusterId.js");
var import_AttributeId = require("../../datatype/AttributeId.js");
var import_TlvAny = require("../../tlv/TlvAny.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Scenes;
((Scenes2) => {
  Scenes2.NameSupport = {
    /**
     * The ability to store a name for a scene.
     */
    nameSupport: (0, import_BitmapSchema.BitFlag)(7)
  };
  Scenes2.TlvAttributeValuePair = (0, import_TlvObject.TlvObject)({
    /**
     * This field shall be present or not present, for all instances in the Scenes cluster. If this field is not
     * present, then the data type of AttributeValue shall be determined by the order and data type defined in the
     * cluster specification. Otherwise the data type of AttributeValue shall be the data type of the attribute
     * indicated by AttributeID.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.6.1.1
     */
    attributeId: (0, import_TlvObject.TlvOptionalField)(0, import_AttributeId.TlvAttributeId),
    /**
     * This is the attribute value as part of an extension field set. See AttributeID to determine the data type
     * for this field.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.6.1.2
     */
    attributeValue: (0, import_TlvObject.TlvField)(1, import_TlvAny.TlvAny)
  });
  Scenes2.TlvExtensionFieldSet = (0, import_TlvObject.TlvObject)({
    clusterId: (0, import_TlvObject.TlvField)(0, import_ClusterId.TlvClusterId),
    attributeValueList: (0, import_TlvObject.TlvField)(1, (0, import_TlvArray.TlvArray)(Scenes2.TlvAttributeValuePair))
  });
  Scenes2.TlvAddSceneRequest = (0, import_TlvObject.TlvObject)({
    groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId),
    sceneId: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8),
    transitionTime: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt16),
    sceneName: (0, import_TlvObject.TlvField)(3, import_TlvString.TlvString.bound({ maxLength: 16 })),
    extensionFieldSets: (0, import_TlvObject.TlvField)(4, (0, import_TlvArray.TlvArray)(Scenes2.TlvExtensionFieldSet))
  });
  Scenes2.TlvAddSceneResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    groupId: (0, import_TlvObject.TlvField)(1, import_GroupId.TlvGroupId),
    sceneId: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8)
  });
  Scenes2.TlvViewSceneRequest = (0, import_TlvObject.TlvObject)({ groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId), sceneId: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8) });
  Scenes2.TlvViewSceneResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    groupId: (0, import_TlvObject.TlvField)(1, import_GroupId.TlvGroupId),
    sceneId: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8),
    transitionTime: (0, import_TlvObject.TlvOptionalField)(3, import_TlvNumber.TlvUInt16),
    sceneName: (0, import_TlvObject.TlvOptionalField)(4, import_TlvString.TlvString.bound({ maxLength: 16 })),
    extensionFieldSets: (0, import_TlvObject.TlvOptionalField)(5, (0, import_TlvArray.TlvArray)(Scenes2.TlvExtensionFieldSet))
  });
  Scenes2.TlvRemoveSceneRequest = (0, import_TlvObject.TlvObject)({
    groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId),
    sceneId: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8)
  });
  Scenes2.TlvRemoveSceneResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    groupId: (0, import_TlvObject.TlvField)(1, import_GroupId.TlvGroupId),
    sceneId: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8)
  });
  Scenes2.TlvRemoveAllScenesRequest = (0, import_TlvObject.TlvObject)({ groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId) });
  Scenes2.TlvRemoveAllScenesResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    groupId: (0, import_TlvObject.TlvField)(1, import_GroupId.TlvGroupId)
  });
  Scenes2.TlvStoreSceneRequest = (0, import_TlvObject.TlvObject)({ groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId), sceneId: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8) });
  Scenes2.TlvStoreSceneResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    groupId: (0, import_TlvObject.TlvField)(1, import_GroupId.TlvGroupId),
    sceneId: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8)
  });
  Scenes2.TlvRecallSceneRequest = (0, import_TlvObject.TlvObject)({
    groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId),
    sceneId: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8),
    transitionTime: (0, import_TlvObject.TlvOptionalField)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16))
  });
  Scenes2.TlvGetSceneMembershipRequest = (0, import_TlvObject.TlvObject)({ groupId: (0, import_TlvObject.TlvField)(0, import_GroupId.TlvGroupId) });
  Scenes2.TlvGetSceneMembershipResponse = (0, import_TlvObject.TlvObject)({
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    capacity: (0, import_TlvObject.TlvField)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8)),
    groupId: (0, import_TlvObject.TlvField)(2, import_GroupId.TlvGroupId),
    sceneList: (0, import_TlvObject.TlvOptionalField)(3, (0, import_TlvArray.TlvArray)(import_TlvNumber.TlvUInt8))
  });
  Scenes2.Mode = { copyAllScenes: (0, import_BitmapSchema.BitFlag)(0), reserved: (0, import_BitmapSchema.BitField)(1, 7) };
  Scenes2.TlvCopySceneRequest = (0, import_TlvObject.TlvObject)({
    /**
     * The Mode field contains information of how the scene copy is to proceed. This field shall be formatted as
     * illustrated in Format of the Mode Field of the CopyScene Command.
     *
     * The CopyAllScenes subfield is 1-bit in length and indicates whether all scenes are to be copied. If this
     * value is set to 1, all scenes are to be copied and the SceneIdentifierFrom and SceneIdentifierTo fields
     * shall be ignored. Otherwise this field is set to 0.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11.1
     */
    mode: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Scenes2.Mode)),
    /**
     * The GroupIdentifierFrom field specifies the identifier of the group from which the scene is to be copied.
     * Together with the SceneIdentifierFrom field, this field uniquely identifies the scene to copy from the Scene
     * Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11.2
     */
    groupIdentifierFrom: (0, import_TlvObject.TlvField)(1, import_GroupId.TlvGroupId),
    /**
     * The SceneIdentifierFrom field specifies the identifier of the scene from which the scene is to be copied.
     * Together with the GroupIdentifierFrom field, this field uniquely identifies the scene to copy from the Scene
     * Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11.3
     */
    sceneIdentifierFrom: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8),
    /**
     * The GroupIdentifierTo field specifies the identifier of the group to which the scene is to be copied.
     * Together with the SceneIdentifierTo field, this field uniquely identifies the scene to copy to the Scene
     * Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11.4
     */
    groupIdentifierTo: (0, import_TlvObject.TlvField)(3, import_GroupId.TlvGroupId),
    /**
     * The SceneIdentifierTo field specifies the identifier of the scene to which the scene is to be copied.
     * Together with the GroupIdentifierTo field, this field uniquely identifies the scene to copy to the Scene
     * Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11.5
     */
    sceneIdentifierTo: (0, import_TlvObject.TlvField)(4, import_TlvNumber.TlvUInt8)
  });
  Scenes2.TlvCopySceneResponse = (0, import_TlvObject.TlvObject)({
    /**
     * The Status field contains the status of the copy scene attempt. This field shall be set to one of the
     * non-reserved values listed in Values of the Status Field of the CopySceneResponse Command.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.20.1
     */
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * The GroupIdentifierFrom field specifies the identifier of the group from which the scene was copied, as
     * specified in the CopyScene command. Together with the SceneIdentifierFrom field, this field uniquely
     * identifies the scene that was copied from the Scene Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.20.2
     */
    groupIdentifierFrom: (0, import_TlvObject.TlvField)(1, import_GroupId.TlvGroupId),
    /**
     * The SceneIdentifierFrom field is specifies the identifier of the scene from which the scene was copied, as
     * specified in the CopyScene command. Together with the GroupIdentifierFrom field, this field uniquely
     * identifies the scene that was copied from the Scene Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.20.3
     */
    sceneIdentifierFrom: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8)
  });
  let Feature;
  ((Feature2) => {
    Feature2["SceneNames"] = "SceneNames";
  })(Feature = Scenes2.Feature || (Scenes2.Feature = {}));
  Scenes2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 5,
    name: "Scenes",
    revision: 4,
    features: {
      /**
       * SceneNames
       *
       * The ability to store a name for a scene.
       */
      sceneNames: (0, import_BitmapSchema.BitFlag)(0)
    },
    attributes: {
      /**
       * The SceneCount attribute specifies the number of scenes currently in the server’s Scene Table.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.1
       */
      sceneCount: (0, import_Cluster.Attribute)(0, import_TlvNumber.TlvUInt8, { default: 0 }),
      /**
       * The CurrentScene attribute holds the scene identifier of the scene last invoked.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.2
       */
      currentScene: (0, import_Cluster.Attribute)(1, import_TlvNumber.TlvUInt8, { default: 0 }),
      /**
       * The CurrentGroup attribute holds the group identifier of the scene last invoked, or 0 if the scene last
       * invoked is not associated with a group.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.3
       */
      currentGroup: (0, import_Cluster.Attribute)(2, import_GroupId.TlvGroupId, { default: (0, import_GroupId.GroupId)(0) }),
      /**
       * The SceneValid attribute indicates whether the state of the server corresponds to that associated with
       * the CurrentScene and CurrentGroup attributes. TRUE indicates that these attributes are valid, FALSE
       * indicates that they are not valid.
       *
       * Before a scene has been stored or recalled, this attribute is set to FALSE. After a successful
       * StoreScene or RecallScene command it is set to TRUE. If, after a scene is stored or recalled, the state
       * of the server is modified, this attribute is set to FALSE.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.4
       */
      sceneValid: (0, import_Cluster.Attribute)(3, import_TlvBoolean.TlvBoolean, { default: true }),
      /**
       * This attribute provides legacy, read-only access to whether the Scene Names feature is supported. The
       * most significant bit, bit 7, shall be equal to bit 0 of the FeatureMap attribute. All other bits shall
       * be 0.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.5
       */
      nameSupport: (0, import_Cluster.Attribute)(4, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Scenes2.NameSupport)),
      /**
       * The LastConfiguredBy attribute holds the Node ID (the IEEE address in case of Zigbee) of the node that
       * last configured the Scene Table.
       *
       * The null value indicates that the server has not been configured, or that the identifier of the node
       * that last configured the Scenes cluster is not known.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.6
       */
      lastConfiguredBy: (0, import_Cluster.OptionalAttribute)(5, (0, import_TlvNullable.TlvNullable)(import_NodeId.TlvNodeId), { default: null })
    },
    commands: {
      /**
       * It is not mandatory for an extension field set to be included in the command for every cluster on that
       * endpoint that has a defined extension field set. Extension field sets may be omitted, including the case
       * of no extension field sets at all.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.2
       */
      addScene: (0, import_Cluster.Command)(0, Scenes2.TlvAddSceneRequest, 0, Scenes2.TlvAddSceneResponse, { invokeAcl: import_Cluster.AccessLevel.Manage }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.3
       */
      viewScene: (0, import_Cluster.Command)(1, Scenes2.TlvViewSceneRequest, 1, Scenes2.TlvViewSceneResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.4
       */
      removeScene: (0, import_Cluster.Command)(
        2,
        Scenes2.TlvRemoveSceneRequest,
        2,
        Scenes2.TlvRemoveSceneResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.5
       */
      removeAllScenes: (0, import_Cluster.Command)(
        3,
        Scenes2.TlvRemoveAllScenesRequest,
        3,
        Scenes2.TlvRemoveAllScenesResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.6
       */
      storeScene: (0, import_Cluster.Command)(
        4,
        Scenes2.TlvStoreSceneRequest,
        4,
        Scenes2.TlvStoreSceneResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.7
       */
      recallScene: (0, import_Cluster.Command)(5, Scenes2.TlvRecallSceneRequest, 5, import_Cluster.TlvNoResponse),
      /**
       * The GetSceneMembership command can be used to find an unused scene identifier within a certain group
       * when no commissioning tool is in the network, or for a commissioning tool to get the used scene
       * identifiers within a certain group, for the endpoint that implements this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.8
       */
      getSceneMembership: (0, import_Cluster.Command)(6, Scenes2.TlvGetSceneMembershipRequest, 6, Scenes2.TlvGetSceneMembershipResponse),
      /**
       * The EnhancedAddScene command allows a scene to be added using a finer scene transition time than the
       * AddScene command.
       *
       * This command shall have the same data fields as the AddScene command, with the following difference:
       *
       * The TransitionTime data field shall be measured in tenths of a second rather than in seconds.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.9
       */
      enhancedAddScene: (0, import_Cluster.OptionalCommand)(
        64,
        Scenes2.TlvAddSceneRequest,
        64,
        Scenes2.TlvAddSceneResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The EnhancedViewScene command allows a scene to be retrieved using a finer scene transition time than
       * the ViewScene command.
       *
       * This command shall have the same data fields as the ViewScene command.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.10
       */
      enhancedViewScene: (0, import_Cluster.OptionalCommand)(65, Scenes2.TlvViewSceneRequest, 65, Scenes2.TlvViewSceneResponse),
      /**
       * The CopyScene command allows a client to efficiently copy scenes from one group/scene identifier pair to
       * another group/scene identifier pair.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11
       */
      copyScene: (0, import_Cluster.OptionalCommand)(
        66,
        Scenes2.TlvCopySceneRequest,
        66,
        Scenes2.TlvCopySceneResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      )
    }
  });
  Scenes2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    { ...Scenes2.Base, supportedFeatures: { sceneNames: true } },
    /**
     * Use this factory method to create a Scenes cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a Scenes cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...Scenes2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(Scenes2.Base.features, ...features)
      });
      return cluster;
    }
  );
})(Scenes || (Scenes = {}));
const ScenesCluster = Scenes.Cluster;
