// @project-chip/matter.js/dist/cjs/cluster/definitions/SwitchCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SwitchCluster_exports = {};
__export(SwitchCluster_exports, {
  Switch: () => Switch,
  SwitchCluster: () => SwitchCluster
});
module.exports = __toCommonJS(SwitchCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Switch;
((Switch2) => {
  Switch2.TlvMultiPressOngoingEvent = (0, import_TlvObject.TlvObject)({
    newPosition: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    currentNumberOfPressesCounted: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8.bound({ min: 2 }))
  });
  Switch2.TlvMultiPressCompleteEvent = (0, import_TlvObject.TlvObject)({
    previousPosition: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    totalNumberOfPressesCounted: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt8.bound({ min: 1 }))
  });
  Switch2.TlvSwitchLatchedEvent = (0, import_TlvObject.TlvObject)({ newPosition: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8) });
  Switch2.TlvInitialPressEvent = (0, import_TlvObject.TlvObject)({ newPosition: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8) });
  Switch2.TlvLongPressEvent = (0, import_TlvObject.TlvObject)({ newPosition: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8) });
  Switch2.TlvLongReleaseEvent = (0, import_TlvObject.TlvObject)({ previousPosition: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8) });
  Switch2.TlvShortReleaseEvent = (0, import_TlvObject.TlvObject)({ previousPosition: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8) });
  let Feature;
  ((Feature2) => {
    Feature2["LatchingSwitch"] = "LatchingSwitch";
    Feature2["MomentarySwitch"] = "MomentarySwitch";
    Feature2["MomentarySwitchRelease"] = "MomentarySwitchRelease";
    Feature2["MomentarySwitchLongPress"] = "MomentarySwitchLongPress";
    Feature2["MomentarySwitchMultiPress"] = "MomentarySwitchMultiPress";
  })(Feature = Switch2.Feature || (Switch2.Feature = {}));
  Switch2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 59,
    name: "Switch",
    revision: 1,
    features: {
      /**
       * LatchingSwitch
       */
      latchingSwitch: (0, import_BitmapSchema.BitFlag)(0),
      /**
       * MomentarySwitch
       */
      momentarySwitch: (0, import_BitmapSchema.BitFlag)(1),
      /**
       * MomentarySwitchRelease
       */
      momentarySwitchRelease: (0, import_BitmapSchema.BitFlag)(2),
      /**
       * MomentarySwitchLongPress
       */
      momentarySwitchLongPress: (0, import_BitmapSchema.BitFlag)(3),
      /**
       * MomentarySwitchMultiPress
       */
      momentarySwitchMultiPress: (0, import_BitmapSchema.BitFlag)(4)
    },
    attributes: {
      /**
       * This attribute shall indicate the maximum number of positions the switch has. Any kind of switch has a
       * minimum of 2 positions. Also see Section 1.11.10, “NumberOfPositions > 2” for the case
       * NumberOfPositions>2.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.11.5.1
       */
      numberOfPositions: (0, import_Cluster.FixedAttribute)(0, import_TlvNumber.TlvUInt8.bound({ min: 2 }), { default: 2 }),
      /**
       * This attribute shall indicate the position of the switch. The valid range is zero to
       * NumberOfPositions-1. CurrentPosition value 0 shall be assigned to the default position of the switch:
       * for example the "open" state of a rocker switch, or the "idle" state of a push button switch.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.11.5.2
       */
      currentPosition: (0, import_Cluster.Attribute)(1, import_TlvNumber.TlvUInt8, { persistent: true, default: 0 })
    }
  });
  Switch2.MomentarySwitchMultiPressComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute shall indicate how many consecutive presses can be detected and reported by a momentary
       * switch which supports multi-press (e.g. it will report the value 3 if it can detect single press, double
       * press and triple press, but not quad press and beyond).
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.11.5.3
       */
      multiPressMax: (0, import_Cluster.FixedAttribute)(2, import_TlvNumber.TlvUInt8.bound({ min: 2 }), { default: 2 })
    },
    events: {
      /**
       * This event shall be generated to indicate how many times the momentary switch has been pressed in a
       * multi-press sequence, during that sequence. See Section 1.11.9, “Sequence of events for MultiPress”
       * below.
       *
       * The NewPosition field shall indicate the new value of the CurrentPosition attribute, i.e. while pressed.
       *
       * The CurrentNumberOfPressesCounted field shall contain:
       *
       *   • a value of 2 when the second press of a multi-press sequence has been detected,
       *
       *   • a value of 3 when the third press of a multi-press sequence has been detected,
       *
       *   • a value of N when the Nth press of a multi-press sequence has been detected.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.11.7.6
       */
      multiPressOngoing: (0, import_Cluster.Event)(5, import_Cluster.EventPriority.Info, Switch2.TlvMultiPressOngoingEvent),
      /**
       * This event shall be generated to indicate how many times the momentary switch has been pressed in a
       * multi-press sequence, after it has been detected that the sequence has ended. See Section 1.11.9,
       * “Sequence of events for MultiPress” below.
       *
       * The PreviousPosition field shall indicate the previous value of the CurrentPosition attribute, i.e. just
       * prior to release.
       *
       * The TotalNumberOfPressesCounted field shall contain:
       *
       *   • a value of 1 when there was one press in a multi-press sequence (and the sequence has ended),
       *
       *     i.e. there was no double press (or more),
       *
       *   • a value of 2 when there were exactly two presses in a multi-press sequence (and the sequence has
       *     ended),
       *
       *   • a value of 3 when there were exactly three presses in a multi-press sequence (and the sequence has
       *     ended),
       *
       *   • a value of N when there were exactly N presses in a multi-press sequence (and the sequence has
       *     ended).
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.11.7.7
       */
      multiPressComplete: (0, import_Cluster.Event)(6, import_Cluster.EventPriority.Info, Switch2.TlvMultiPressCompleteEvent)
    }
  });
  Switch2.LatchingSwitchComponent = import_ClusterFactory.ClusterFactory.Component({
    events: {
      /**
       * This event shall be generated, when the latching switch is moved to a new position. It may have been
       * delayed by debouncing within the switch.
       *
       * The NewPosition field shall indicate the new value of the CurrentPosition attribute, i.e. after the move.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.11.7.1
       */
      switchLatched: (0, import_Cluster.Event)(0, import_Cluster.EventPriority.Info, Switch2.TlvSwitchLatchedEvent)
    }
  });
  Switch2.MomentarySwitchComponent = import_ClusterFactory.ClusterFactory.Component({
    events: {
      /**
       * This event shall be generated, when the momentary switch starts to be pressed (after debouncing).
       *
       * The NewPosition field shall indicate the new value of the CurrentPosition attribute, i.e. while pressed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.11.7.2
       */
      initialPress: (0, import_Cluster.Event)(1, import_Cluster.EventPriority.Info, Switch2.TlvInitialPressEvent)
    }
  });
  Switch2.MomentarySwitchLongPressComponent = import_ClusterFactory.ClusterFactory.Component({
    events: {
      /**
       * This event shall be generated, when the momentary switch has been pressed for a "long" time (this time
       * interval is manufacturer determined (e.g. since it depends on the switch physics)).
       *
       * The NewPosition field shall indicate the new value of the CurrentPosition attribute, i.e. while pressed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.11.7.3
       */
      longPress: (0, import_Cluster.Event)(2, import_Cluster.EventPriority.Info, Switch2.TlvLongPressEvent),
      /**
       * This event shall be generated, when the momentary switch has been released (after debouncing) and after
       * having been pressed for a long time, i.e. this event shall be generated when the switch is released if a
       * LongPress event has been generated since since the previous InitialPress event. Also see Section 1.11.8,
       * “Sequence of generated events”.
       *
       * The PreviousPosition field shall indicate the previous value of the CurrentPosition attribute, i.e. just
       * prior to release.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.11.7.5
       */
      longRelease: (0, import_Cluster.Event)(4, import_Cluster.EventPriority.Info, Switch2.TlvLongReleaseEvent)
    }
  });
  Switch2.MomentarySwitchReleaseComponent = import_ClusterFactory.ClusterFactory.Component({
    events: {
      /**
       * This event shall be generated, when the momentary switch has been released (after debouncing).
       *
       *   • If the server supports the Momentary Switch LongPress (MSL) feature, this event shall be generated
       *     when the switch is released if no LongPress event had been generated since the previous InitialPress
       *     event.
       *
       *   • If the server does not support the Momentary Switch LongPress (MSL) feature, this event shall be
       *     generated when the switch is released - even when the switch was pressed for a long time.
       *
       *   • Also see Section 1.11.8, “Sequence of generated events”.
       *
       * The PreviousPosition field shall indicate the previous value of the CurrentPosition attribute, i.e. just
       * prior to release.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.11.7.4
       */
      shortRelease: (0, import_Cluster.Event)(3, import_Cluster.EventPriority.Info, Switch2.TlvShortReleaseEvent)
    }
  });
  Switch2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    Switch2.Base,
    /**
     * Use this factory method to create a Switch cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a Switch cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...Switch2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(Switch2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, Switch2.MomentarySwitchMultiPressComponent, { momentarySwitchMultiPress: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Switch2.LatchingSwitchComponent, { latchingSwitch: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Switch2.MomentarySwitchComponent, { momentarySwitch: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Switch2.MomentarySwitchLongPressComponent, { momentarySwitchLongPress: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Switch2.MomentarySwitchReleaseComponent, { momentarySwitchRelease: true });
      import_ClusterFactory.ClusterFactory.prevent(
        cluster,
        { momentarySwitchRelease: true, momentarySwitch: false },
        { momentarySwitchLongPress: true, momentarySwitch: false },
        { momentarySwitchLongPress: true, momentarySwitchRelease: false },
        { momentarySwitchMultiPress: true, momentarySwitch: false },
        { momentarySwitchMultiPress: true, momentarySwitchRelease: false },
        { latchingSwitch: true, momentarySwitch: true },
        { latchingSwitch: false, momentarySwitch: false }
      );
      return cluster;
    }
  );
  const MSM = { momentarySwitchMultiPress: true };
  const LS = { latchingSwitch: true };
  const MS = { momentarySwitch: true };
  const MSL = { momentarySwitchLongPress: true };
  const MSR = { momentarySwitchRelease: true };
  Switch2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: Switch2.Cluster.id,
    name: Switch2.Cluster.name,
    revision: Switch2.Cluster.revision,
    features: Switch2.Cluster.features,
    attributes: {
      ...Switch2.Cluster.attributes,
      multiPressMax: import_ClusterFactory.ClusterFactory.AsConditional(
        Switch2.MomentarySwitchMultiPressComponent.attributes.multiPressMax,
        { mandatoryIf: [MSM] }
      )
    },
    events: {
      switchLatched: import_ClusterFactory.ClusterFactory.AsConditional(
        Switch2.LatchingSwitchComponent.events.switchLatched,
        { mandatoryIf: [LS] }
      ),
      initialPress: import_ClusterFactory.ClusterFactory.AsConditional(
        Switch2.MomentarySwitchComponent.events.initialPress,
        { mandatoryIf: [MS] }
      ),
      longPress: import_ClusterFactory.ClusterFactory.AsConditional(
        Switch2.MomentarySwitchLongPressComponent.events.longPress,
        { mandatoryIf: [MSL] }
      ),
      shortRelease: import_ClusterFactory.ClusterFactory.AsConditional(
        Switch2.MomentarySwitchReleaseComponent.events.shortRelease,
        { mandatoryIf: [MSR] }
      ),
      longRelease: import_ClusterFactory.ClusterFactory.AsConditional(
        Switch2.MomentarySwitchLongPressComponent.events.longRelease,
        { mandatoryIf: [MSL] }
      ),
      multiPressOngoing: import_ClusterFactory.ClusterFactory.AsConditional(
        Switch2.MomentarySwitchMultiPressComponent.events.multiPressOngoing,
        { mandatoryIf: [MSM] }
      ),
      multiPressComplete: import_ClusterFactory.ClusterFactory.AsConditional(
        Switch2.MomentarySwitchMultiPressComponent.events.multiPressComplete,
        { mandatoryIf: [MSM] }
      )
    }
  });
})(Switch || (Switch = {}));
const SwitchCluster = Switch.Cluster;
