// @project-chip/matter.js/dist/cjs/cluster/definitions/TargetNavigatorCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TargetNavigatorCluster_exports = {};
__export(TargetNavigatorCluster_exports, {
  TargetNavigator: () => TargetNavigator,
  TargetNavigatorCluster: () => TargetNavigatorCluster
});
module.exports = __toCommonJS(TargetNavigatorCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var TargetNavigator;
((TargetNavigator2) => {
  TargetNavigator2.TlvTargetInfoStruct = (0, import_TlvObject.TlvObject)({
    /**
     * An unique id within the TargetList.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.5.1.1
     */
    identifier: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    /**
     * A name string for the TargetInfoStruct.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.5.1.2
     */
    name: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvString.bound({ maxLength: 32 }))
  });
  TargetNavigator2.TlvNavigateTargetRequest = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the Identifier for the target for UX navigation. The Target shall be an Identifier value
     * contained within one of the TargetInfoStruct objects in the TargetList attribute list.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.4.1.1
     */
    target: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.4.1.2
     */
    data: (0, import_TlvObject.TlvOptionalField)(1, import_TlvString.TlvString)
  });
  let Status;
  ((Status2) => {
    Status2[Status2["Success"] = 0] = "Success";
    Status2[Status2["TargetNotFound"] = 1] = "TargetNotFound";
    Status2[Status2["NotAllowed"] = 2] = "NotAllowed";
  })(Status = TargetNavigator2.Status || (TargetNavigator2.Status = {}));
  TargetNavigator2.TlvNavigateTargetResponse = (0, import_TlvObject.TlvObject)({
    /**
     * This shall indicate the of the command.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.4.2.1
     */
    status: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.4.2.2
     */
    data: (0, import_TlvObject.TlvOptionalField)(1, import_TlvString.TlvString)
  });
  TargetNavigator2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 1285,
    name: "TargetNavigator",
    revision: 1,
    attributes: {
      /**
       * The TargetList attribute shall represent a list of targets that can be navigated to within the
       * experience presented to the user by the Endpoint (Video Player or Content App). The list shall not
       * contain any entries with the same Identifier in the TargetInfoStruct object.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.3.1
       */
      targetList: (0, import_Cluster.Attribute)(0, (0, import_TlvArray.TlvArray)(TargetNavigator2.TlvTargetInfoStruct, { maxLength: 254 }), { default: [] }),
      /**
       * The CurrentTarget attribute shall represent the Identifier for the target which is currently in
       * foreground on the corresponding Endpoint (Video Player or Content App), or null to indicate that no
       * target is in the foreground.
       *
       * When not null, the CurrentTarget shall be an Identifier value contained within one of the
       * TargetInfoStruct objects in the TargetList attribute list.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.3.2
       */
      currentTarget: (0, import_Cluster.OptionalAttribute)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8), { default: null })
    },
    commands: {
      /**
       * Upon receipt, this shall navigation the UX to the target identified.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.4.1
       */
      navigateTarget: (0, import_Cluster.Command)(0, TargetNavigator2.TlvNavigateTargetRequest, 1, TargetNavigator2.TlvNavigateTargetResponse)
    }
  });
})(TargetNavigator || (TargetNavigator = {}));
const TargetNavigatorCluster = TargetNavigator.Cluster;
