// @project-chip/matter.js/dist/cjs/cluster/definitions/TemperatureMeasurementCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TemperatureMeasurementCluster_exports = {};
__export(TemperatureMeasurementCluster_exports, {
  TemperatureMeasurement: () => TemperatureMeasurement,
  TemperatureMeasurementCluster: () => TemperatureMeasurementCluster
});
module.exports = __toCommonJS(TemperatureMeasurementCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var TemperatureMeasurement;
((TemperatureMeasurement2) => {
  TemperatureMeasurement2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 1026,
    name: "TemperatureMeasurement",
    revision: 4,
    attributes: {
      /**
       * Represents the temperature in degrees Celsius as follows:
       *
       * MeasuredValue = 100 x temperature [°C]
       *
       * Where -273.15°C ≤ temperature ≤ 327.67°C, with a resolution of 0.01°C. The null value indicates that the
       * temperature is unknown.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.3.4.1
       */
      measuredValue: (0, import_Cluster.Attribute)(0, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16.bound({ min: -27315 }))),
      /**
       * The MinMeasuredValue attribute indicates the minimum value of MeasuredValue that is capable of being
       * measured. See Measured Value for more details.
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.3.4.2
       */
      minMeasuredValue: (0, import_Cluster.Attribute)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16.bound({ min: -27315 })), { default: -27315 }),
      /**
       * The MaxMeasuredValue attribute indicates the maximum value of MeasuredValue that is capable of being
       * measured. See Measured Value for more details.
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.3.4.3
       */
      maxMeasuredValue: (0, import_Cluster.Attribute)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16), { default: 32767 }),
      /**
       * See Measured Value.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.3.4.4
       */
      tolerance: (0, import_Cluster.OptionalAttribute)(3, import_TlvNumber.TlvUInt16.bound({ max: 2048 }), { default: 0 })
    }
  });
})(TemperatureMeasurement || (TemperatureMeasurement = {}));
const TemperatureMeasurementCluster = TemperatureMeasurement.Cluster;
