// @project-chip/matter.js/dist/cjs/cluster/definitions/ThermostatCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ThermostatCluster_exports = {};
__export(ThermostatCluster_exports, {
  Thermostat: () => Thermostat,
  ThermostatCluster: () => ThermostatCluster
});
module.exports = __toCommonJS(ThermostatCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Thermostat;
((Thermostat2) => {
  Thermostat2.RemoteSensing = {
    /**
     * When set, LocalTemperature Value is derived from a remote node
     */
    localTemperature: (0, import_BitmapSchema.BitFlag)(0),
    /**
     * When set, OutdoorTemperature is derived from a remote node
     */
    outdoorTemperature: (0, import_BitmapSchema.BitFlag)(1),
    /**
     * When set, Occupancy is derived from a remote node
     */
    occupancy: (0, import_BitmapSchema.BitFlag)(2)
  };
  let ControlSequenceOfOperation;
  ((ControlSequenceOfOperation2) => {
    ControlSequenceOfOperation2[ControlSequenceOfOperation2["CoolingOnly"] = 0] = "CoolingOnly";
    ControlSequenceOfOperation2[ControlSequenceOfOperation2["CoolingWithReheat"] = 1] = "CoolingWithReheat";
    ControlSequenceOfOperation2[ControlSequenceOfOperation2["HeatingOnly"] = 2] = "HeatingOnly";
    ControlSequenceOfOperation2[ControlSequenceOfOperation2["HeatingWithReheat"] = 3] = "HeatingWithReheat";
    ControlSequenceOfOperation2[ControlSequenceOfOperation2["CoolingAndHeating"] = 4] = "CoolingAndHeating";
    ControlSequenceOfOperation2[ControlSequenceOfOperation2["CoolingAndHeatingWithReheat"] = 5] = "CoolingAndHeatingWithReheat";
  })(ControlSequenceOfOperation = Thermostat2.ControlSequenceOfOperation || (Thermostat2.ControlSequenceOfOperation = {}));
  let SystemMode;
  ((SystemMode2) => {
    SystemMode2[SystemMode2["Off"] = 0] = "Off";
    SystemMode2[SystemMode2["Auto"] = 1] = "Auto";
    SystemMode2[SystemMode2["Cool"] = 3] = "Cool";
    SystemMode2[SystemMode2["Heat"] = 4] = "Heat";
    SystemMode2[SystemMode2["EmergencyHeat"] = 5] = "EmergencyHeat";
    SystemMode2[SystemMode2["Precooling"] = 6] = "Precooling";
    SystemMode2[SystemMode2["FanOnly"] = 7] = "FanOnly";
    SystemMode2[SystemMode2["Dry"] = 8] = "Dry";
    SystemMode2[SystemMode2["Sleep"] = 9] = "Sleep";
  })(SystemMode = Thermostat2.SystemMode || (Thermostat2.SystemMode = {}));
  Thermostat2.AlarmMask = {
    initializationFailure: (0, import_BitmapSchema.BitFlag)(0),
    hardwareFailure: (0, import_BitmapSchema.BitFlag)(1),
    selfCalibrationFailure: (0, import_BitmapSchema.BitFlag)(2)
  };
  let TemperatureSetpointHold;
  ((TemperatureSetpointHold2) => {
    TemperatureSetpointHold2[TemperatureSetpointHold2["SetpointHoldOff"] = 0] = "SetpointHoldOff";
    TemperatureSetpointHold2[TemperatureSetpointHold2["SetpointHoldOn"] = 1] = "SetpointHoldOn";
  })(TemperatureSetpointHold = Thermostat2.TemperatureSetpointHold || (Thermostat2.TemperatureSetpointHold = {}));
  Thermostat2.ThermostatProgrammingOperationMode = {
    /**
     * Schedule programming mode. This enables any programmed weekly schedule configurations.
     */
    scheduleActive: (0, import_BitmapSchema.BitFlag)(0),
    /**
     * Auto/recovery mode
     */
    autoRecovery: (0, import_BitmapSchema.BitFlag)(1),
    /**
     * Economy/EnergyStar mode
     */
    economy: (0, import_BitmapSchema.BitFlag)(2)
  };
  Thermostat2.ThermostatRunningState = {
    /**
     * Heat State On
     */
    heat: (0, import_BitmapSchema.BitFlag)(0),
    /**
     * Cool State On
     */
    cool: (0, import_BitmapSchema.BitFlag)(1),
    /**
     * Fan State On
     */
    fan: (0, import_BitmapSchema.BitFlag)(2),
    /**
     * Heat 2nd Stage State On
     */
    heatStage2: (0, import_BitmapSchema.BitFlag)(3),
    /**
     * Cool 2nd Stage State On
     */
    coolStage2: (0, import_BitmapSchema.BitFlag)(4),
    /**
     * Fan 2nd Stage State On
     */
    fanStage2: (0, import_BitmapSchema.BitFlag)(5),
    /**
     * Fan 3rd Stage Stage On
     */
    fanStage3: (0, import_BitmapSchema.BitFlag)(6)
  };
  let SetpointChangeSource;
  ((SetpointChangeSource2) => {
    SetpointChangeSource2[SetpointChangeSource2["Manual"] = 0] = "Manual";
    SetpointChangeSource2[SetpointChangeSource2["Schedule"] = 1] = "Schedule";
    SetpointChangeSource2[SetpointChangeSource2["External"] = 2] = "External";
  })(SetpointChangeSource = Thermostat2.SetpointChangeSource || (Thermostat2.SetpointChangeSource = {}));
  let AcType;
  ((AcType2) => {
    AcType2[AcType2["Unknown"] = 0] = "Unknown";
    AcType2[AcType2["CoolingFixed"] = 1] = "CoolingFixed";
    AcType2[AcType2["HeatPumpFixed"] = 2] = "HeatPumpFixed";
    AcType2[AcType2["CoolingInverter"] = 3] = "CoolingInverter";
    AcType2[AcType2["HeatPumpInverter"] = 4] = "HeatPumpInverter";
  })(AcType = Thermostat2.AcType || (Thermostat2.AcType = {}));
  let AcRefrigerantType;
  ((AcRefrigerantType2) => {
    AcRefrigerantType2[AcRefrigerantType2["Unknown"] = 0] = "Unknown";
    AcRefrigerantType2[AcRefrigerantType2["R22"] = 1] = "R22";
    AcRefrigerantType2[AcRefrigerantType2["R410A"] = 2] = "R410A";
    AcRefrigerantType2[AcRefrigerantType2["R407C"] = 3] = "R407C";
  })(AcRefrigerantType = Thermostat2.AcRefrigerantType || (Thermostat2.AcRefrigerantType = {}));
  let AcCompressorType;
  ((AcCompressorType2) => {
    AcCompressorType2[AcCompressorType2["Unknown"] = 0] = "Unknown";
    AcCompressorType2[AcCompressorType2["T1"] = 1] = "T1";
    AcCompressorType2[AcCompressorType2["T2"] = 2] = "T2";
    AcCompressorType2[AcCompressorType2["T3"] = 3] = "T3";
  })(AcCompressorType = Thermostat2.AcCompressorType || (Thermostat2.AcCompressorType = {}));
  Thermostat2.AcErrorCode = {
    /**
     * Compressor Failure or Refrigerant Leakage
     */
    compressorFail: (0, import_BitmapSchema.BitFlag)(0),
    /**
     * Room Temperature Sensor Failure
     */
    roomSensorFail: (0, import_BitmapSchema.BitFlag)(1),
    /**
     * Outdoor Temperature Sensor Failure
     */
    outdoorSensorFail: (0, import_BitmapSchema.BitFlag)(2),
    /**
     * Indoor Coil Temperature Sensor Failure
     */
    coilSensorFail: (0, import_BitmapSchema.BitFlag)(3),
    /**
     * Fan Failure
     */
    fanFail: (0, import_BitmapSchema.BitFlag)(4)
  };
  let AcLouverPosition;
  ((AcLouverPosition2) => {
    AcLouverPosition2[AcLouverPosition2["Closed"] = 1] = "Closed";
    AcLouverPosition2[AcLouverPosition2["Open"] = 2] = "Open";
    AcLouverPosition2[AcLouverPosition2["Quarter"] = 3] = "Quarter";
    AcLouverPosition2[AcLouverPosition2["Half"] = 4] = "Half";
    AcLouverPosition2[AcLouverPosition2["ThreeQuarters"] = 5] = "ThreeQuarters";
  })(AcLouverPosition = Thermostat2.AcLouverPosition || (Thermostat2.AcLouverPosition = {}));
  let AcCapacityFormat;
  ((AcCapacityFormat2) => {
    AcCapacityFormat2[AcCapacityFormat2["BtUh"] = 0] = "BtUh";
  })(AcCapacityFormat = Thermostat2.AcCapacityFormat || (Thermostat2.AcCapacityFormat = {}));
  let SetpointAdjustMode;
  ((SetpointAdjustMode2) => {
    SetpointAdjustMode2[SetpointAdjustMode2["Heat"] = 0] = "Heat";
    SetpointAdjustMode2[SetpointAdjustMode2["Cool"] = 1] = "Cool";
    SetpointAdjustMode2[SetpointAdjustMode2["Both"] = 2] = "Both";
  })(SetpointAdjustMode = Thermostat2.SetpointAdjustMode || (Thermostat2.SetpointAdjustMode = {}));
  Thermostat2.TlvSetpointRaiseLowerRequest = (0, import_TlvObject.TlvObject)({
    mode: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvEnum)()),
    amount: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvInt8)
  });
  Thermostat2.Occupancy = { occupied: (0, import_BitmapSchema.BitFlag)(0) };
  let ThermostatRunningMode;
  ((ThermostatRunningMode2) => {
    ThermostatRunningMode2[ThermostatRunningMode2["Off"] = 0] = "Off";
    ThermostatRunningMode2[ThermostatRunningMode2["Cool"] = 3] = "Cool";
    ThermostatRunningMode2[ThermostatRunningMode2["Heat"] = 4] = "Heat";
  })(ThermostatRunningMode = Thermostat2.ThermostatRunningMode || (Thermostat2.ThermostatRunningMode = {}));
  let StartOfWeek;
  ((StartOfWeek2) => {
    StartOfWeek2[StartOfWeek2["Sunday"] = 0] = "Sunday";
    StartOfWeek2[StartOfWeek2["Monday"] = 1] = "Monday";
    StartOfWeek2[StartOfWeek2["Tuesday"] = 2] = "Tuesday";
    StartOfWeek2[StartOfWeek2["Wednesday"] = 3] = "Wednesday";
    StartOfWeek2[StartOfWeek2["Thursday"] = 4] = "Thursday";
    StartOfWeek2[StartOfWeek2["Friday"] = 5] = "Friday";
    StartOfWeek2[StartOfWeek2["Saturday"] = 6] = "Saturday";
  })(StartOfWeek = Thermostat2.StartOfWeek || (Thermostat2.StartOfWeek = {}));
  Thermostat2.DayOfWeek = {
    sunday: (0, import_BitmapSchema.BitFlag)(0),
    monday: (0, import_BitmapSchema.BitFlag)(1),
    tuesday: (0, import_BitmapSchema.BitFlag)(2),
    wednesday: (0, import_BitmapSchema.BitFlag)(3),
    thursday: (0, import_BitmapSchema.BitFlag)(4),
    friday: (0, import_BitmapSchema.BitFlag)(5),
    saturday: (0, import_BitmapSchema.BitFlag)(6),
    away: (0, import_BitmapSchema.BitFlag)(7)
  };
  Thermostat2.ModeForSequence = { heatSetpointPresent: (0, import_BitmapSchema.BitFlag)(0), coolSetpointPresent: (0, import_BitmapSchema.BitFlag)(1) };
  Thermostat2.TlvThermostatScheduleTransition = (0, import_TlvObject.TlvObject)({
    /**
     * This field represents the start time of the schedule transition during the associated day. The time will be
     * represented by a 16 bits unsigned integer to designate the minutes since midnight. For example, 6am will be
     * represented by 360 minutes since midnight and 11:30pm will be represented by 1410 minutes since midnight.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.9.5.1
     */
    transitionTime: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16.bound({ max: 1439 })),
    heatSetpoint: (0, import_TlvObject.TlvField)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16)),
    coolSetpoint: (0, import_TlvObject.TlvField)(2, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16))
  });
  Thermostat2.TlvSetWeeklyScheduleRequest = (0, import_TlvObject.TlvObject)({
    numberOfTransitionsForSequence: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    dayOfWeekForSequence: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Thermostat2.DayOfWeek)),
    modeForSequence: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Thermostat2.ModeForSequence)),
    transitions: (0, import_TlvObject.TlvField)(3, (0, import_TlvArray.TlvArray)(Thermostat2.TlvThermostatScheduleTransition))
  });
  Thermostat2.TlvGetWeeklyScheduleRequest = (0, import_TlvObject.TlvObject)({
    daysToReturn: (0, import_TlvObject.TlvField)(0, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Thermostat2.DayOfWeek)),
    modeToReturn: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Thermostat2.ModeForSequence))
  });
  Thermostat2.TlvGetWeeklyScheduleResponse = (0, import_TlvObject.TlvObject)({
    numberOfTransitionsForSequence: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt8),
    dayOfWeekForSequence: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Thermostat2.DayOfWeek)),
    modeForSequence: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Thermostat2.ModeForSequence)),
    transitions: (0, import_TlvObject.TlvField)(3, (0, import_TlvArray.TlvArray)(Thermostat2.TlvThermostatScheduleTransition))
  });
  let Feature;
  ((Feature2) => {
    Feature2["Heating"] = "Heating";
    Feature2["Cooling"] = "Cooling";
    Feature2["Occupancy"] = "Occupancy";
    Feature2["ScheduleConfiguration"] = "ScheduleConfiguration";
    Feature2["Setback"] = "Setback";
    Feature2["AutoMode"] = "AutoMode";
    Feature2["LocalTemperatureNotExposed"] = "LocalTemperatureNotExposed";
  })(Feature = Thermostat2.Feature || (Thermostat2.Feature = {}));
  Thermostat2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 513,
    name: "Thermostat",
    revision: 6,
    features: {
      /**
       * Heating
       *
       * Thermostat is capable of managing a heating device
       */
      heating: (0, import_BitmapSchema.BitFlag)(0),
      /**
       * Cooling
       *
       * Thermostat is capable of managing a cooling device
       */
      cooling: (0, import_BitmapSchema.BitFlag)(1),
      /**
       * Occupancy
       *
       * Supports Occupied and Unoccupied setpoints
       */
      occupancy: (0, import_BitmapSchema.BitFlag)(2),
      /**
       * ScheduleConfiguration
       *
       * Supports remote configuration of a weekly schedule of setpoint transitions
       */
      scheduleConfiguration: (0, import_BitmapSchema.BitFlag)(3),
      /**
       * Setback
       *
       * Supports configurable setback (or span)
       */
      setback: (0, import_BitmapSchema.BitFlag)(4),
      /**
       * AutoMode
       *
       * Supports a System Mode of Auto
       */
      autoMode: (0, import_BitmapSchema.BitFlag)(5),
      /**
       * LocalTemperatureNotExposed
       *
       * Thermostat does not expose the LocalTemperature Value in the LocalTemperature attribute
       */
      localTemperatureNotExposed: (0, import_BitmapSchema.BitFlag)(6)
    },
    attributes: {
      /**
       * This attribute indicates the current LocalTemperature value, when available. The value may be local, or
       * remote, depending on the value of the RemoteSensing attribute when supported.
       *
       *   • If the LTNE feature is not supported:
       *
       *     ◦ If the temperature measurement is invalid or currently unavailable, the attribute shall report
       *       null.
       *
       *     ◦ If the temperature measurement is valid, the attribute shall report that value.
       *
       *   • Otherwise, if the LTNE feature is supported, there is no feedback externally available for the
       *     LocalTemperature value. In that case, the LocalTemperature attribute shall always report null.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.2
       */
      localTemperature: (0, import_Cluster.Attribute)(0, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16), { default: null }),
      /**
       * This attribute represents the outdoor temperature, as measured locally or remotely (over the network).
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.3
       */
      outdoorTemperature: (0, import_Cluster.OptionalAttribute)(1, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16), { default: null }),
      /**
       * This attribute indicates when the local temperature, outdoor temperature and occupancy are being sensed
       * by remote networked sensors, rather than internal sensors.
       *
       * A bit set to 1 indicates remote sensing of the relevant value.
       *
       * Table 72. RemoteSensing Attribute Bit Values
       *
       * If the LTNE feature is present in the server, the LocalTemperature RemoteSensing bit value shall always
       * report a value of 0.
       *
       * If the LocalTemperature RemoteSensing bit is written with a value of 1 when the LTNE feature is present,
       * the write shall fail and the server shall report a CONSTRAINT_ERROR.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.22
       */
      remoteSensing: (0, import_Cluster.OptionalWritableAttribute)(
        26,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Thermostat2.RemoteSensing),
        { persistent: true, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute specifies the overall operating environment of the thermostat, and thus the possible
       * system modes that the thermostat can operate in. It shall be set to one of the following values.
       *
       * Table 73. ControlSequenceOfOperation Values
       *
       * NOTE
       *
       * CoolingAndHeating
       *
       * A thermostat indicating it supports CoolingAndHeating (or CoolingAndHeatingWithReheat) SHOULD be able to
       * request heating or cooling on demand and will usually support the Auto SystemMode.
       *
       * Systems which support cooling or heating, requiring external intervention to change modes or where the
       * whole building must be in the same mode, SHOULD report CoolingOnly or HeatingOnly based on the current
       * capability.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.23
       */
      controlSequenceOfOperation: (0, import_Cluster.WritableAttribute)(
        27,
        (0, import_TlvNumber.TlvEnum)(),
        {
          persistent: true,
          default: 4 /* CoolingAndHeating */,
          writeAcl: import_Cluster.AccessLevel.Manage
        }
      ),
      /**
       * This attribute specifies the current operating mode of the thermostat, It shall be set to one of the
       * following values, as limited by the ControlSequenceOfOperation Attribute.
       *
       * NOTE It is not mandatory to support all values.
       *
       * Table 75. Interpretation of Heat, Cool and Auto SystemMode Values
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.24
       */
      systemMode: (0, import_Cluster.WritableAttribute)(
        28,
        (0, import_TlvNumber.TlvEnum)(),
        { scene: true, persistent: true, default: 1 /* Auto */, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute specifies whether each of the alarms listed below is enabled. When the bit number
       * corresponding to the alarm code is set to 1, the alarm is enabled, else it is disabled. Bits not
       * corresponding to a code in the table are reserved.
       *
       * When the Alarms cluster is implemented on a device, and one of the alarm conditions included in this
       * table occurs, an alarm notification is generated, with the alarm code field set as listed in the table.
       *
       * Table 76. Alarm Codes
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.25
       */
      alarmMask: (0, import_Cluster.OptionalAttribute)(29, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Thermostat2.AlarmMask)),
      /**
       * This attribute specifies the temperature hold status on the thermostat. If hold status is on, the
       * thermostat SHOULD maintain the temperature set point for the current mode until a system mode change. If
       * hold status is off, the thermostat SHOULD follow the setpoint transitions specified by its internal
       * scheduling program. If the thermostat supports setpoint hold for a specific duration, it SHOULD also
       * implement the TemperatureSetpointHoldDuration attribute.
       *
       * Table 78. TemperatureSetpointHold Values
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.30
       */
      temperatureSetpointHold: (0, import_Cluster.OptionalWritableAttribute)(
        35,
        (0, import_TlvNumber.TlvEnum)(),
        { persistent: true, default: 0 /* SetpointHoldOff */, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute sets the period in minutes for which a setpoint hold is active. Thermostats that support
       * hold for a specified duration SHOULD implement this attribute. The null value indicates the field is
       * unused. All other values are reserved.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.31
       */
      temperatureSetpointHoldDuration: (0, import_Cluster.OptionalWritableAttribute)(
        36,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16.bound({ max: 1440 })),
        { persistent: true, default: null, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute determines the operational state of the thermostat’s programming. The thermostat shall
       * modify its programming operation when this attribute is modified by a client and update this attribute
       * when its programming operation is modified locally by a user. The thermostat may support more than one
       * active ThermostatProgrammingOperationMode. For example, the thermostat may operate simultaneously in
       * Schedule Programming Mode and Recovery Mode.
       *
       * Thermostats which contain a schedule may use this attribute to control how that schedule is used, even
       * if they do not support the Schedule Configuration feature.
       *
       * Table 79. ThermostatProgrammingOperationMode Attribute Values
       *
       * When ScheduleActive is not set, the setpoint is altered only by manual up/down changes at the thermostat
       * or remotely, not by internal schedule programming.
       *
       * NOTE Modifying the ScheduleActive bit does not clear or delete previous weekly schedule programming
       * configurations.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.32
       */
      thermostatProgrammingOperationMode: (0, import_Cluster.OptionalWritableAttribute)(
        37,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Thermostat2.ThermostatProgrammingOperationMode),
        { writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute represents the current relay state of the heat, cool, and fan relays.
       *
       * Table 80. HVAC Relay State Values
       *
       * Unimplemented outputs shall be treated as if they were Off.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.33
       */
      thermostatRunningState: (0, import_Cluster.OptionalAttribute)(41, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, Thermostat2.ThermostatRunningState)),
      /**
       * This attribute specifies the source of the current active OccupiedCoolingSetpoint or
       * OccupiedHeatingSetpoint (i.e., who or what determined the current setpoint).
       *
       * This attribute enables service providers to determine whether changes to setpoints were initiated due to
       * occupant comfort, scheduled programming or some other source (e.g., electric utility or other service
       * provider). Because automation services may initiate frequent setpoint changes, this attribute clearly
       * differentiates the source of setpoint changes made at the thermostat.
       *
       * Table 81. SetpointChangeSource Values
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.34
       */
      setpointChangeSource: (0, import_Cluster.OptionalAttribute)(
        48,
        (0, import_TlvNumber.TlvEnum)(),
        { default: 0 /* Manual */ }
      ),
      /**
       * This attribute specifies the delta between the current active OccupiedCoolingSetpoint or
       * OccupiedHeatingSetpoint and the previous active setpoint. This attribute is meant to accompany the
       * SetpointChangeSource attribute; devices implementing SetpointChangeAmount SHOULD also implement
       * SetpointChangeSource.
       *
       * The null value indicates that the previous setpoint was unknown.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.35
       */
      setpointChangeAmount: (0, import_Cluster.OptionalAttribute)(49, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16), { default: null }),
      /**
       * This attribute specifies the time in UTC at which the SetpointChangeSourceAmount attribute change was
       * recorded.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.36
       */
      setpointChangeSourceTimestamp: (0, import_Cluster.OptionalAttribute)(50, import_TlvNumber.TlvUInt32, { default: 0 }),
      /**
       * This attribute specifies the delta between the LocalTemperature Value and the OccupiedHeatingSetpoint or
       * UnoccupiedHeatingSetpoint attributes at which the Thermostat server will operate in emergency heat mode.
       *
       * If the difference between LocalTemperature Value and Un/OccupiedHeatingSetpoint is greater than or equal
       * to the EmergencyHeatDelta and the Thermostat server’s SystemMode attribute is in a heating-related mode,
       * then the Thermostat server shall immediately switch to the SystemMode attribute value that provides the
       * highest stage of heating (e.g., emergency heat) and continue operating in that running state until the
       * OccupiedHeatingSetpoint value is reached. For example:
       *
       *   • LocalTemperature = 10.0°C
       *
       *   • OccupiedHeatingSetpoint = 16.0°C
       *
       *   • EmergencyHeatDelta = 2.0°C
       *
       * ⇒ OccupiedHeatingSetpoint - LocalTemperature ≥? EmergencyHeatDelta
       *
       * ⇒ 16°C - 10°C ≥? 2°C
       *
       * ⇒ TRUE >>> Thermostat server changes its SystemMode to operate in 2nd stage or emergency heat mode
       *
       * The purpose of this attribute is to provide Thermostat clients the ability to configure rapid heating
       * when a setpoint is of a specified amount greater than the measured temperature. This allows the heated
       * space to be quickly heated to the desired level set by the user.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.43
       */
      emergencyHeatDelta: (0, import_Cluster.OptionalWritableAttribute)(
        58,
        import_TlvNumber.TlvUInt8,
        { persistent: true, default: 250, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute indicates the type of Mini Split ACType of Mini Split AC is defined depending on how
       * Cooling and Heating condition is achieved by Mini Split AC.
       *
       * Table 82. ACType Values
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.44
       */
      acType: (0, import_Cluster.OptionalWritableAttribute)(
        64,
        (0, import_TlvNumber.TlvEnum)(),
        { persistent: true, default: 0 /* Unknown */, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute indicates capacity of Mini Split AC in terms of the format defined by the
       * ACCapacityFormat attribute
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.45
       */
      acCapacity: (0, import_Cluster.OptionalWritableAttribute)(
        65,
        import_TlvNumber.TlvUInt16,
        { persistent: true, default: 0, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute indicates type of refrigerant used within the Mini Split AC.
       *
       * Table 83. ACRefrigerantType Values
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.46
       */
      acRefrigerantType: (0, import_Cluster.OptionalWritableAttribute)(
        66,
        (0, import_TlvNumber.TlvEnum)(),
        { persistent: true, default: 0 /* Unknown */, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute indicates type of Compressor used within the Mini Split AC.
       *
       * Table 84. ACCompressorType Values
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.47
       */
      acCompressorType: (0, import_Cluster.OptionalWritableAttribute)(
        67,
        (0, import_TlvNumber.TlvEnum)(),
        { persistent: true, default: 0 /* Unknown */, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute indicates the type of errors encountered within the Mini Split AC. Error values are
       * reported with four bytes values. Each bit within the four bytes indicates the unique error.
       *
       * Table 85. ACErrorCode Values
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.48
       */
      acErrorCode: (0, import_Cluster.OptionalWritableAttribute)(
        68,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt32, Thermostat2.AcErrorCode),
        { writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute indicates the position of Louver on the AC.
       *
       * Table 86. ACLouverPosition Values
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.49
       */
      acLouverPosition: (0, import_Cluster.OptionalWritableAttribute)(
        69,
        (0, import_TlvNumber.TlvEnum)(),
        { persistent: true, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute represents the temperature of the AC coil, as measured locally or remotely (over the
       * network).
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.50
       */
      acCoilTemperature: (0, import_Cluster.OptionalAttribute)(70, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvInt16.bound({ min: -27315 })), { default: null }),
      /**
       * This attribute specifies the format for the ACCapacity attribute.
       *
       * Table 87. ACCapacity Values
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.51
       */
      acCapacityFormat: (0, import_Cluster.OptionalWritableAttribute)(
        71,
        (0, import_TlvNumber.TlvEnum)(),
        { persistent: true, default: 0 /* BtUh */, writeAcl: import_Cluster.AccessLevel.Manage }
      )
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.8
       */
      setpointRaiseLower: (0, import_Cluster.Command)(0, Thermostat2.TlvSetpointRaiseLowerRequest, 0, import_Cluster.TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.8
       */
      getRelayStatusLog: (0, import_Cluster.OptionalCommand)(4, import_TlvNoArguments.TlvNoArguments, 1, import_TlvNoArguments.TlvNoArguments)
    }
  });
  Thermostat2.OccupancyComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute specifies whether the heated/cooled space is occupied or not, as measured locally or
       * remotely (over the network). If bit 0 = 1, the space is occupied, else it is unoccupied. All other bits
       * are reserved.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.4
       */
      occupancy: (0, import_Cluster.Attribute)(
        2,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, Thermostat2.Occupancy),
        { default: (0, import_BitmapSchema.BitsFromPartial)(Thermostat2.Occupancy, { occupied: true }) }
      )
    }
  });
  Thermostat2.HeatingComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute specifies the absolute minimum level that the heating setpoint may be set to. This is a
       * limitation imposed by the manufacturer.
       *
       * Refer to Setpoint Limits for constraints
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.5
       */
      absMinHeatSetpointLimit: (0, import_Cluster.OptionalFixedAttribute)(3, import_TlvNumber.TlvInt16, { default: 700 }),
      /**
       * This attribute specifies the absolute maximum level that the heating setpoint may be set to. This is a
       * limitation imposed by the manufacturer.
       *
       * Refer to Setpoint Limits for constraints
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.6
       */
      absMaxHeatSetpointLimit: (0, import_Cluster.OptionalFixedAttribute)(4, import_TlvNumber.TlvInt16, { default: 3e3 }),
      /**
       * This attribute specifies the level of heating demanded by the PI loop in percent. This value is 0 when
       * the thermostat is in “off” or “cooling” mode.
       *
       * This attribute is reported regularly and may be used to control a heating device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.10
       */
      piHeatingDemand: (0, import_Cluster.OptionalAttribute)(8, import_TlvNumber.TlvUInt8.bound({ max: 100 })),
      /**
       * This attribute specifies the heating mode setpoint when the room is occupied.
       *
       * Refer to Setpoint Limits for constraints. If an attempt is made to set this attribute such that these
       * constraints are violated, a response with the status code CONSTRAINT_ERROR shall be returned.
       *
       * If the occupancy status of the room is unknown, this attribute shall be used as the heating mode
       * setpoint.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.14
       */
      occupiedHeatingSetpoint: (0, import_Cluster.WritableAttribute)(
        18,
        import_TlvNumber.TlvInt16.bound({ min: -27315 }),
        { scene: true, persistent: true, default: 2e3 }
      ),
      /**
       * This attribute specifies the minimum level that the heating setpoint may be set to.
       *
       * This attribute, and the following three attributes, allow the user to define setpoint limits more
       * constrictive than the manufacturer imposed ones. Limiting users (e.g., in a commercial building) to such
       * setpoint limits can help conserve power.
       *
       * Refer to Setpoint Limits for constraints. If an attempt is made to set this attribute to a value which
       * conflicts with setpoint values then those setpoints shall be adjusted by the minimum amount to permit
       * this attribute to be set to the desired value. If an attempt is made to set this attribute to a value
       * which is not consistent with the constraints and cannot be resolved by modifying setpoints then a
       * response with the status code CONSTRAINT_ERROR shall be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.17
       */
      minHeatSetpointLimit: (0, import_Cluster.OptionalWritableAttribute)(
        21,
        import_TlvNumber.TlvInt16.bound({ min: -27315 }),
        { persistent: true, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute specifies the maximum level that the heating setpoint may be set to.
       *
       * Refer to Setpoint Limits for constraints. If an attempt is made to set this attribute to a value which
       * conflicts with setpoint values then those setpoints shall be adjusted by the minimum amount to permit
       * this attribute to be set to the desired value. If an attempt is made to set this attribute to a value
       * which is not consistent with the constraints and cannot be resolved by modifying setpoints then a
       * response with the status code CONSTRAINT_ERROR shall be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.18
       */
      maxHeatSetpointLimit: (0, import_Cluster.OptionalWritableAttribute)(
        22,
        import_TlvNumber.TlvInt16.bound({ min: -27315 }),
        { persistent: true, writeAcl: import_Cluster.AccessLevel.Manage }
      )
    }
  });
  Thermostat2.CoolingComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute specifies the absolute minimum level that the cooling setpoint may be set to. This is a
       * limitation imposed by the manufacturer.
       *
       * Refer to Setpoint Limits for constraints
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.7
       */
      absMinCoolSetpointLimit: (0, import_Cluster.OptionalFixedAttribute)(5, import_TlvNumber.TlvInt16, { default: 1600 }),
      /**
       * This attribute specifies the absolute maximum level that the cooling setpoint may be set to. This is a
       * limitation imposed by the manufacturer.
       *
       * Refer to Setpoint Limits for constraints
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.8
       */
      absMaxCoolSetpointLimit: (0, import_Cluster.OptionalFixedAttribute)(6, import_TlvNumber.TlvInt16, { default: 3200 }),
      /**
       * This attribute specifies the level of cooling demanded by the PI (proportional integral) control loop in
       * use by the thermostat (if any), in percent. This value is 0 when the thermostat is in “off” or “heating”
       * mode.
       *
       * This attribute is reported regularly and may be used to control a cooling device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.9
       */
      piCoolingDemand: (0, import_Cluster.OptionalAttribute)(7, import_TlvNumber.TlvUInt8.bound({ max: 100 })),
      /**
       * This attribute specifies the cooling mode setpoint when the room is occupied.
       *
       * Refer to Setpoint Limits for constraints. If an attempt is made to set this attribute such that these
       * constraints are violated, a response with the status code CONSTRAINT_ERROR shall be returned.
       *
       * If the occupancy status of the room is unknown, this attribute shall be used as the cooling mode
       * setpoint.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.13
       */
      occupiedCoolingSetpoint: (0, import_Cluster.WritableAttribute)(
        17,
        import_TlvNumber.TlvInt16.bound({ min: -27315 }),
        { scene: true, persistent: true, default: 2600 }
      ),
      /**
       * This attribute specifies the minimum level that the cooling setpoint may be set to.
       *
       * Refer to Setpoint Limits for constraints. If an attempt is made to set this attribute to a value which
       * conflicts with setpoint values then those setpoints shall be adjusted by the minimum amount to permit
       * this attribute to be set to the desired value. If an attempt is made to set this attribute to a value
       * which is not consistent with the constraints and cannot be resolved by modifying setpoints then a
       * response with the status code CONSTRAINT_ERROR shall be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.19
       */
      minCoolSetpointLimit: (0, import_Cluster.OptionalWritableAttribute)(
        23,
        import_TlvNumber.TlvInt16.bound({ min: -27315 }),
        { persistent: true, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute specifies the maximum level that the cooling setpoint may be set to.
       *
       * Refer to Setpoint Limits for constraints. If an attempt is made to set this attribute to a value which
       * conflicts with setpoint values then those setpoints shall be adjusted by the minimum amount to permit
       * this attribute to be set to the desired value. If an attempt is made to set this attribute to a value
       * which is not consistent with the constraints and cannot be resolved by modifying setpoints
       *
       * then a response with the status code CONSTRAINT_ERROR shall be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.20
       */
      maxCoolSetpointLimit: (0, import_Cluster.OptionalWritableAttribute)(
        24,
        import_TlvNumber.TlvInt16.bound({ min: -27315 }),
        { persistent: true, writeAcl: import_Cluster.AccessLevel.Manage }
      )
    }
  });
  Thermostat2.NotLocalTemperatureNotExposedComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute specifies the offset the Thermostat server shall make to the measured temperature
       * (locally or remotely) to adjust the LocalTemperature Value prior to using, displaying or reporting it.
       *
       * The purpose of this attribute is to adjust the calibration of the Thermostat server per the user’s
       * preferences (e.g., to match if there are multiple servers displaying different values for the same HVAC
       * area) or compensate for variability amongst temperature sensors.
       *
       * If a Thermostat client attempts to write LocalTemperatureCalibration attribute to an unsupported value
       * (e.g., out of the range supported by the Thermostat server), the Thermostat server shall respond with a
       * status of SUCCESS and set the value of LocalTemperatureCalibration to the upper or lower limit reached.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.12
       */
      localTemperatureCalibration: (0, import_Cluster.OptionalWritableAttribute)(
        16,
        import_TlvNumber.TlvInt8.bound({ min: -25, max: 25 }),
        { persistent: true, default: 0, writeAcl: import_Cluster.AccessLevel.Manage }
      )
    }
  });
  Thermostat2.CoolingAndOccupancyComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute specifies the cooling mode setpoint when the room is unoccupied.
       *
       * Refer to Setpoint Limits for constraints. If an attempt is made to set this attribute such that these
       * constraints are violated, a response with the status code CONSTRAINT_ERROR shall be returned.
       *
       * If the occupancy status of the room is unknown, this attribute shall not be used.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.15
       */
      unoccupiedCoolingSetpoint: (0, import_Cluster.WritableAttribute)(
        19,
        import_TlvNumber.TlvInt16.bound({ min: -27315 }),
        { persistent: true, default: 2600 }
      )
    }
  });
  Thermostat2.HeatingAndOccupancyComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute specifies the heating mode setpoint when the room is unoccupied.
       *
       * Refer to Setpoint Limits for constraints. If an attempt is made to set this attribute such that these
       * constraints are violated, a response with the status code CONSTRAINT_ERROR shall be returned.
       *
       * If the occupancy status of the room is unknown, this attribute shall not be used.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.16
       */
      unoccupiedHeatingSetpoint: (0, import_Cluster.WritableAttribute)(
        20,
        import_TlvNumber.TlvInt16.bound({ min: -27315 }),
        { persistent: true, default: 2e3 }
      )
    }
  });
  Thermostat2.AutoModeComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * On devices which support the AUTO feature, this attribute specifies the minimum difference between the
       * Heat Setpoint and the Cool Setpoint.
       *
       * Refer to Setpoint Limits for constraints. If an attempt is made to set this attribute to a value which
       * conflicts with setpoint values then those setpoints shall be adjusted by the minimum amount to permit
       * this attribute to be set to the desired value. If an attempt is made to set this attribute to a value
       * which is not consistent with the constraints and cannot be resolved by modifying setpoints then a
       * response with the status code CONSTRAINT_ERROR shall be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.21
       */
      minSetpointDeadBand: (0, import_Cluster.WritableAttribute)(
        25,
        import_TlvNumber.TlvInt8.bound({ min: 0, max: 25 }),
        { persistent: true, default: 20, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7
       */
      thermostatRunningMode: (0, import_Cluster.OptionalAttribute)(
        30,
        (0, import_TlvNumber.TlvEnum)(),
        { default: 0 /* Off */ }
      )
    }
  });
  Thermostat2.ScheduleConfigurationComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute represents the day of the week that this thermostat considers to be the start of week for
       * weekly set point scheduling.
       *
       * Table 77. StartofWeek Values
       *
       * This attribute may be able to be used as the base to determine if the device supports weekly scheduling
       * by reading the attribute. Successful response means that the weekly scheduling is supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.27
       */
      startOfWeek: (0, import_Cluster.FixedAttribute)(32, (0, import_TlvNumber.TlvEnum)()),
      /**
       * This attribute determines how many weekly schedule transitions the thermostat is capable of handling.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.28
       */
      numberOfWeeklyTransitions: (0, import_Cluster.FixedAttribute)(33, import_TlvNumber.TlvUInt8, { default: 0 }),
      /**
       * This attribute determines how many daily schedule transitions the thermostat is capable of handling.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.29
       */
      numberOfDailyTransitions: (0, import_Cluster.FixedAttribute)(34, import_TlvNumber.TlvUInt8, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.8
       */
      setWeeklySchedule: (0, import_Cluster.Command)(
        1,
        Thermostat2.TlvSetWeeklyScheduleRequest,
        1,
        import_Cluster.TlvNoResponse,
        { invokeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.8
       */
      getWeeklySchedule: (0, import_Cluster.Command)(2, Thermostat2.TlvGetWeeklyScheduleRequest, 0, Thermostat2.TlvGetWeeklyScheduleResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.8
       */
      clearWeeklySchedule: (0, import_Cluster.Command)(3, import_TlvNoArguments.TlvNoArguments, 3, import_Cluster.TlvNoResponse, { invokeAcl: import_Cluster.AccessLevel.Manage })
    }
  });
  Thermostat2.SetbackComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute specifies the amount that the Thermostat server will allow the LocalTemperature Value to
       * float above the OccupiedCooling setpoint (i.e., OccupiedCooling + OccupiedSetback) or below the
       * OccupiedHeating setpoint (i.e., OccupiedHeating – OccupiedSetback) before initiating a state change to
       * bring the temperature back to the user’s desired setpoint. This attribute is sometimes also referred to
       * as the “span.”
       *
       * The purpose of this attribute is to allow remote configuration of the span between the desired set
       *
       * point and the measured temperature to help prevent over-cycling and reduce energy bills, though this may
       * result in lower comfort on the part of some users.
       *
       * The null value indicates the attribute is unused.
       *
       * If the Thermostat client attempts to write OccupiedSetback to a value greater than OccupiedSetbackMax,
       * the Thermostat server shall set its OccupiedSetback value to OccupiedSetbackMax and shall send a Write
       * Attribute Response command with a Status Code field enumeration of SUCCESS response.
       *
       * If the Thermostat client attempts to write OccupiedSetback to a value less than OccupiedSetbackMin, the
       * Thermostat server shall set its OccupiedSetback value to OccupiedSetbackMin and shall send a Write
       * Attribute Response command with a Status Code field enumeration of SUCCESS response.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.37
       */
      occupiedSetback: (0, import_Cluster.WritableAttribute)(
        52,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ max: 254 })),
        { persistent: true, default: null, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute specifies the minimum value that the Thermostat server will allow the OccupiedSetback
       * attribute to be configured by a user.
       *
       * The null value indicates the attribute is unused.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.38
       */
      occupiedSetbackMin: (0, import_Cluster.FixedAttribute)(
        53,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ min: 0, max: 254 })),
        { default: null }
      ),
      /**
       * This attribute specifies the maximum value that the Thermostat server will allow the OccupiedSetback
       * attribute to be configured by a user.
       *
       * The null value indicates the attribute is unused.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.39
       */
      occupiedSetbackMax: (0, import_Cluster.FixedAttribute)(54, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ max: 254 })), { default: null })
    }
  });
  Thermostat2.SetbackAndOccupancyComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * This attribute specifies the amount that the Thermostat server will allow the LocalTemperature Value to
       * float above the UnoccupiedCooling setpoint (i.e., UnoccupiedCooling + UnoccupiedSetback) or below the
       * UnoccupiedHeating setpoint (i.e., UnoccupiedHeating - UnoccupiedSetback) before initiating a state
       * change to bring the temperature back to the user’s desired setpoint. This attribute is sometimes also
       * referred to as the “span.”
       *
       * The purpose of this attribute is to allow remote configuration of the span between the desired setpoint
       * and the measured temperature to help prevent over-cycling and reduce energy bills, though this may
       * result in lower comfort on the part of some users.
       *
       * The null value indicates the attribute is unused.
       *
       * If the Thermostat client attempts to write UnoccupiedSetback to a value greater than
       * UnoccupiedSetbackMax, the Thermostat server shall set its UnoccupiedSetback value to
       * UnoccupiedSetbackMax and shall send a Write Attribute Response command with a Status Code field
       * enumeration of SUCCESS response.
       *
       * If the Thermostat client attempts to write UnoccupiedSetback to a value less than UnoccupiedSetbackMin,
       * the Thermostat server shall set its UnoccupiedSetback value to UnoccupiedSetbackMin and shall send a
       * Write Attribute Response command with a Status Code field enumeration of SUCCESS response.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.40
       */
      unoccupiedSetback: (0, import_Cluster.WritableAttribute)(
        55,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ max: 254 })),
        { persistent: true, default: null, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * This attribute specifies the minimum value that the Thermostat server will allow the UnoccupiedSetback
       * attribute to be configured by a user.
       *
       * The null value indicates the attribute is unused.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.41
       */
      unoccupiedSetbackMin: (0, import_Cluster.FixedAttribute)(
        56,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ min: 0, max: 254 })),
        { default: null }
      ),
      /**
       * This attribute specifies the maximum value that the Thermostat server will allow the UnoccupiedSetback
       * attribute to be configured by a user.
       *
       * The null value indicates the attribute is unused.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.3.7.42
       */
      unoccupiedSetbackMax: (0, import_Cluster.FixedAttribute)(57, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt8.bound({ max: 254 })), { default: null })
    }
  });
  Thermostat2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    Thermostat2.Base,
    /**
     * Use this factory method to create a Thermostat cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a Thermostat cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...Thermostat2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(Thermostat2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, Thermostat2.OccupancyComponent, { occupancy: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Thermostat2.HeatingComponent, { heating: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Thermostat2.CoolingComponent, { cooling: true });
      import_ClusterFactory.ClusterFactory.extend(
        cluster,
        Thermostat2.NotLocalTemperatureNotExposedComponent,
        { localTemperatureNotExposed: false }
      );
      import_ClusterFactory.ClusterFactory.extend(cluster, Thermostat2.CoolingAndOccupancyComponent, { cooling: true, occupancy: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Thermostat2.HeatingAndOccupancyComponent, { heating: true, occupancy: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Thermostat2.AutoModeComponent, { autoMode: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Thermostat2.ScheduleConfigurationComponent, { scheduleConfiguration: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Thermostat2.SetbackComponent, { setback: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, Thermostat2.SetbackAndOccupancyComponent, { setback: true, occupancy: true });
      import_ClusterFactory.ClusterFactory.prevent(
        cluster,
        { autoMode: true, heating: false },
        { autoMode: true, cooling: false },
        { heating: false, cooling: false }
      );
      return cluster;
    }
  );
  const OCC = { occupancy: true };
  const HEAT = { heating: true };
  const COOL = { cooling: true };
  const COOL_OCC = { cooling: true, occupancy: true };
  const HEAT_OCC = { heating: true, occupancy: true };
  const AUTO = { autoMode: true };
  const SCH = { scheduleConfiguration: true };
  const SB = { setback: true };
  const SB_OCC = { setback: true, occupancy: true };
  Thermostat2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: Thermostat2.Cluster.id,
    name: Thermostat2.Cluster.name,
    revision: Thermostat2.Cluster.revision,
    features: Thermostat2.Cluster.features,
    attributes: {
      ...Thermostat2.Cluster.attributes,
      occupancy: import_ClusterFactory.ClusterFactory.AsConditional(Thermostat2.OccupancyComponent.attributes.occupancy, { mandatoryIf: [OCC] }),
      absMinHeatSetpointLimit: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.HeatingComponent.attributes.absMinHeatSetpointLimit,
        { optionalIf: [HEAT] }
      ),
      absMaxHeatSetpointLimit: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.HeatingComponent.attributes.absMaxHeatSetpointLimit,
        { optionalIf: [HEAT] }
      ),
      absMinCoolSetpointLimit: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.CoolingComponent.attributes.absMinCoolSetpointLimit,
        { optionalIf: [COOL] }
      ),
      absMaxCoolSetpointLimit: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.CoolingComponent.attributes.absMaxCoolSetpointLimit,
        { optionalIf: [COOL] }
      ),
      piCoolingDemand: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.CoolingComponent.attributes.piCoolingDemand,
        { optionalIf: [COOL] }
      ),
      piHeatingDemand: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.HeatingComponent.attributes.piHeatingDemand,
        { optionalIf: [HEAT] }
      ),
      localTemperatureCalibration: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.NotLocalTemperatureNotExposedComponent.attributes.localTemperatureCalibration,
        { optionalIf: [] }
      ),
      occupiedCoolingSetpoint: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.CoolingComponent.attributes.occupiedCoolingSetpoint,
        { mandatoryIf: [COOL] }
      ),
      occupiedHeatingSetpoint: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.HeatingComponent.attributes.occupiedHeatingSetpoint,
        { mandatoryIf: [HEAT] }
      ),
      unoccupiedCoolingSetpoint: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.CoolingAndOccupancyComponent.attributes.unoccupiedCoolingSetpoint,
        { mandatoryIf: [COOL_OCC] }
      ),
      unoccupiedHeatingSetpoint: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.HeatingAndOccupancyComponent.attributes.unoccupiedHeatingSetpoint,
        { mandatoryIf: [HEAT_OCC] }
      ),
      minHeatSetpointLimit: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.HeatingComponent.attributes.minHeatSetpointLimit,
        { optionalIf: [HEAT] }
      ),
      maxHeatSetpointLimit: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.HeatingComponent.attributes.maxHeatSetpointLimit,
        { optionalIf: [HEAT] }
      ),
      minCoolSetpointLimit: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.CoolingComponent.attributes.minCoolSetpointLimit,
        { optionalIf: [COOL] }
      ),
      maxCoolSetpointLimit: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.CoolingComponent.attributes.maxCoolSetpointLimit,
        { optionalIf: [COOL] }
      ),
      minSetpointDeadBand: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.AutoModeComponent.attributes.minSetpointDeadBand,
        { mandatoryIf: [AUTO] }
      ),
      thermostatRunningMode: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.AutoModeComponent.attributes.thermostatRunningMode,
        { optionalIf: [AUTO] }
      ),
      startOfWeek: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.ScheduleConfigurationComponent.attributes.startOfWeek,
        { mandatoryIf: [SCH] }
      ),
      numberOfWeeklyTransitions: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.ScheduleConfigurationComponent.attributes.numberOfWeeklyTransitions,
        { mandatoryIf: [SCH] }
      ),
      numberOfDailyTransitions: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.ScheduleConfigurationComponent.attributes.numberOfDailyTransitions,
        { mandatoryIf: [SCH] }
      ),
      occupiedSetback: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.SetbackComponent.attributes.occupiedSetback,
        { mandatoryIf: [SB] }
      ),
      occupiedSetbackMin: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.SetbackComponent.attributes.occupiedSetbackMin,
        { mandatoryIf: [SB] }
      ),
      occupiedSetbackMax: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.SetbackComponent.attributes.occupiedSetbackMax,
        { mandatoryIf: [SB] }
      ),
      unoccupiedSetback: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.SetbackAndOccupancyComponent.attributes.unoccupiedSetback,
        { mandatoryIf: [SB_OCC] }
      ),
      unoccupiedSetbackMin: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.SetbackAndOccupancyComponent.attributes.unoccupiedSetbackMin,
        { mandatoryIf: [SB_OCC] }
      ),
      unoccupiedSetbackMax: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.SetbackAndOccupancyComponent.attributes.unoccupiedSetbackMax,
        { mandatoryIf: [SB_OCC] }
      )
    },
    commands: {
      ...Thermostat2.Cluster.commands,
      setWeeklySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.ScheduleConfigurationComponent.commands.setWeeklySchedule,
        { mandatoryIf: [SCH] }
      ),
      getWeeklySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.ScheduleConfigurationComponent.commands.getWeeklySchedule,
        { mandatoryIf: [SCH] }
      ),
      clearWeeklySchedule: import_ClusterFactory.ClusterFactory.AsConditional(
        Thermostat2.ScheduleConfigurationComponent.commands.clearWeeklySchedule,
        { mandatoryIf: [SCH] }
      )
    }
  });
})(Thermostat || (Thermostat = {}));
const ThermostatCluster = Thermostat.Cluster;
