// @project-chip/matter.js/dist/cjs/cluster/definitions/TimeSyncCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeSyncCluster_exports = {};
__export(TimeSyncCluster_exports, {
  TimeSync: () => TimeSync,
  TimeSyncCluster: () => TimeSyncCluster
});
module.exports = __toCommonJS(TimeSyncCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_NodeId = require("../../datatype/NodeId.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var TimeSync;
((TimeSync2) => {
  let Granularity;
  ((Granularity2) => {
    Granularity2[Granularity2["NoTimeGranularity"] = 0] = "NoTimeGranularity";
    Granularity2[Granularity2["MinutesGranularity"] = 1] = "MinutesGranularity";
    Granularity2[Granularity2["SecondsGranularity"] = 2] = "SecondsGranularity";
    Granularity2[Granularity2["MillisecondsGranularity"] = 3] = "MillisecondsGranularity";
    Granularity2[Granularity2["MicrosecondsGranularity"] = 4] = "MicrosecondsGranularity";
  })(Granularity = TimeSync2.Granularity || (TimeSync2.Granularity = {}));
  let TimeSource;
  ((TimeSource2) => {
    TimeSource2[TimeSource2["None"] = 0] = "None";
    TimeSource2[TimeSource2["Unknown"] = 1] = "Unknown";
    TimeSource2[TimeSource2["Admin"] = 2] = "Admin";
    TimeSource2[TimeSource2["NodeTimeCluster"] = 3] = "NodeTimeCluster";
    TimeSource2[TimeSource2["NonFabricSntp"] = 4] = "NonFabricSntp";
    TimeSource2[TimeSource2["NonFabricNtp"] = 5] = "NonFabricNtp";
    TimeSource2[TimeSource2["FabricSntp"] = 6] = "FabricSntp";
    TimeSource2[TimeSource2["FabricNtp"] = 7] = "FabricNtp";
    TimeSource2[TimeSource2["MixedNtp"] = 8] = "MixedNtp";
    TimeSource2[TimeSource2["NonFabricSntpNts"] = 9] = "NonFabricSntpNts";
    TimeSource2[TimeSource2["NonFabricNtpNts"] = 10] = "NonFabricNtpNts";
    TimeSource2[TimeSource2["FabricSntpNts"] = 11] = "FabricSntpNts";
    TimeSource2[TimeSource2["FabricNtpNts"] = 12] = "FabricNtpNts";
    TimeSource2[TimeSource2["MixedNtpNts"] = 13] = "MixedNtpNts";
    TimeSource2[TimeSource2["CloudSource"] = 14] = "CloudSource";
    TimeSource2[TimeSource2["Ptp"] = 15] = "Ptp";
    TimeSource2[TimeSource2["Gnss"] = 16] = "Gnss";
  })(TimeSource = TimeSync2.TimeSource || (TimeSync2.TimeSource = {}));
  TimeSync2.TlvSetUtcTimeRequest = (0, import_TlvObject.TlvObject)({
    /**
     * This shall give the Client’s UTC Time.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.9.1.1
     */
    utcTime: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvEpochUs),
    /**
     * This shall give the Client’s Granularity, as described in Section 11.16.8.2, “Granularity Attribute”.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.9.1.2
     */
    granularity: (0, import_TlvObject.TlvField)(1, (0, import_TlvNumber.TlvEnum)()),
    /**
     * This shall give the Client’s TimeSource, as described in Section 11.16.8.3, “TimeSource Attribute”.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.9.1.3
     */
    timeSource: (0, import_TlvObject.TlvOptionalField)(2, (0, import_TlvNumber.TlvEnum)())
  });
  let StatusCode;
  ((StatusCode2) => {
    StatusCode2[StatusCode2["TimeNotAccepted"] = 2] = "TimeNotAccepted";
  })(StatusCode = TimeSync2.StatusCode || (TimeSync2.StatusCode = {}));
  TimeSync2.TlvTimeZoneStruct = (0, import_TlvObject.TlvObject)({
    /**
     * The time zone offset from UTC in seconds.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.6.3.1
     */
    offset: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvInt32.bound({ min: -43200, max: 50400 })),
    /**
     * The UTC time when the offset shall be applied.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.6.3.2
     */
    validAt: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvEpochUs),
    /**
     * The time zone name SHOULD provide a human-readable time zone name and it SHOULD use the country/city format
     * specified by the IANA time zone database [https://www.iana.org/time-zones].
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.6.3.3
     */
    name: (0, import_TlvObject.TlvOptionalField)(2, import_TlvString.TlvString.bound({ minLength: 0, maxLength: 64 }))
  });
  TimeSync2.TlvDSTOffsetStruct = (0, import_TlvObject.TlvObject)({
    offset: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvInt32),
    /**
     * The UTC time when the offset shall be applied.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.6.4.1
     */
    validStarting: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvEpochUs),
    /**
     * The UTC time when the offset shall stop being applied. This value shall be larger than the ValidStarting
     * time.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.6.4.2
     */
    validUntil: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvEpochUs)
  });
  TimeSync2.TlvTimeZoneStatusEvent = (0, import_TlvObject.TlvObject)({
    offset: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvInt32.bound({ min: -43200, max: 50400 })),
    name: (0, import_TlvObject.TlvOptionalField)(1, import_TlvString.TlvString.bound({ minLength: 0, maxLength: 64 }))
  });
  let Feature;
  ((Feature2) => {
    Feature2["TimeZone"] = "TimeZone";
    Feature2["NtpClient"] = "NtpClient";
    Feature2["NtpServer"] = "NtpServer";
  })(Feature = TimeSync2.Feature || (TimeSync2.Feature = {}));
  TimeSync2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 56,
    name: "TimeSync",
    revision: 1,
    features: {
      /**
       * TimeZone
       *
       * Server supports time zone.
       */
      timeZone: (0, import_BitmapSchema.BitFlag)(0),
      /**
       * NtpClient
       *
       * Server supports an NTP or SNTP client.
       */
      ntpClient: (0, import_BitmapSchema.BitFlag)(1),
      /**
       * NtpServer
       *
       * Server supports an NTP server role.
       */
      ntpServer: (0, import_BitmapSchema.BitFlag)(2)
    },
    attributes: {
      /**
       * If the server has achieved time synchronization, this shall indicate the current time as a UTC epoch-us
       * (Epoch Time in Microseconds).
       *
       * If the server has not achieved time synchronization, this shall be null. This attribute may be set when
       * a Section 11.16.9.1, “SetUtcTime Command” is received.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.1
       */
      utcTime: (0, import_Cluster.Attribute)(0, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvEpochUs), { omitChanges: true, default: null }),
      /**
       * The granularity of the error that the server is willing to guarantee on the time synchronization. It is
       * of type GranularityEnum.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.2
       */
      granularity: (0, import_Cluster.Attribute)(1, (0, import_TlvNumber.TlvEnum)(), { default: 0 /* NoTimeGranularity */ }),
      /**
       * The server’s time source. This attribute indicates what method the server is using to sync, whether the
       * source uses NTS or not and whether the source is internal or external to the Fabric. This attribute may
       * be used by a client to determine its level of trust in the UTCTime. It is of type TimeSourceEnum.
       *
       * If a server is unsure if the selected NTP server is within the Fabric, it SHOULD indicate the server is
       * NonFabric.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.3
       */
      timeSource: (0, import_Cluster.OptionalAttribute)(2, (0, import_TlvNumber.TlvEnum)(), { default: 0 /* None */ }),
      /**
       * The Node ID of a trusted Time Cluster. The TrustedTimeNodeId Node is used as a check on external time
       * sync sources and may be used as the primary time source if other time sources are unavailable. See
       * Section 11.16.13, “Time source prioritization”. This attribute is writeable only by an Administrator. It
       * SHOULD be set by the Commissioner during commissioning. If no appropriate TrustedTimeNodeId is
       * available, the commissioner may set this value to null.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.5
       */
      trustedTimeNodeId: (0, import_Cluster.WritableAttribute)(
        3,
        (0, import_TlvNullable.TlvNullable)(import_NodeId.TlvNodeId),
        { default: null, writeAcl: import_Cluster.AccessLevel.Administer }
      )
    },
    commands: {
      /**
       * This command may be issued by Administrator to set the time. If the Commissioner does not have a valid
       * time source, it may send a Granularity of NoTimeGranularity.
       *
       * Upon receipt of this command, the server may update its UTCTime attribute to match the time specified in
       * the command, if the stated Granularity and TimeSource are acceptable. The server shall update its
       * UTCTime attribute if its current Granularity is NoTimeGranularity.
       *
       * If the time is updated, the server shall also update its Granularity attribute as appropriate
       *
       * server does not plan to maintain time). It shall also update its TimeSource attribute to Admin. It shall
       * also update its last known good UTC time.
       *
       * If the server updates its UTCTime attribute, it shall accept the command with a status code of SUCCESS.
       * If it opts to not update its time, it shall fail the command with a cluster specific Status Code of
       * TimeNotAccepted.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.9.1
       */
      setUtcTime: (0, import_Cluster.Command)(0, TimeSync2.TlvSetUtcTimeRequest, 0, import_Cluster.TlvNoResponse, { invokeAcl: import_Cluster.AccessLevel.Administer })
    }
  });
  TimeSync2.NtpClientComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The default NTP server the server’s Node may use if other time sources are unavailable. This attribute
       * may contain a domain name or a static IPv6 address in text format as specified in RFC 5952
       * [https://tools.ietf.org/html/rfc5952]. See Section 11.16.13, “Time source prioritization”. This
       * attribute is writeable only by an Administrator. It SHOULD be set by the Commissioner during
       * commissioning. If no default NTP is available, the Commissioner may set this value to null.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.4
       */
      defaultNtp: (0, import_Cluster.WritableAttribute)(
        4,
        (0, import_TlvNullable.TlvNullable)(import_TlvString.TlvString.bound({ maxLength: 128 })),
        { default: null, writeAcl: import_Cluster.AccessLevel.Administer }
      )
    }
  });
  TimeSync2.TimeZoneComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * A list of time zone offsets from UTC and when they shall take effect. This attribute uses a list of time
       * offset configurations to allow Nodes to handle scheduled regulatory time zone changes. This attribute
       * shall NOT be used to indicate daylight savings time changes (see Section 11.16.8.7, “DSTOffset
       * Attribute” for daylight savings time).
       *
       * The first entry shall have a ValidAt entry of 0. If there is a second entry, it shall have a non-zero
       * ValidAt time.
       *
       * If a server supports a TimeZoneDatabase, the server may update its own DSTOffset list (Section
       * 11.16.8.7, “DSTOffset Attribute”) to add new DST change times as required, based on the Name fields of
       * the TimeZoneStruct. Administrators may add additional entries to the DSTOffset of other Nodes with the
       * same time zone, if required.
       *
       * If a server does not support a TimeZoneDatabase, the Name field of the TimeZoneStruct is only applicable
       * for client-side localization. In particular:
       *
       *   • If the server does not support a TimeZoneDatabase, the Name field shall NOT be used to calculate the
       *     local time.
       *
       *   • If the server does not support a TimeZoneDatabase, the Name field shall NOT be used to calculate DST
       *     start or end dates.
       *
       * Upon writing this attribute, the server shall recompute its LocalTime, taking into account the Offset of
       * the currently used TimeZoneStruct.
       *
       * When time passes, the server SHOULD remove any entries which are no longer active and change the ValidAt
       * time for the currently used TimeZoneStruct list item to zero.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.6
       */
      timeZone: (0, import_Cluster.WritableAttribute)(
        5,
        (0, import_TlvArray.TlvArray)(TimeSync2.TlvTimeZoneStruct, { minLength: 1, maxLength: 2 }),
        { writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * A list of offsets to apply for daylight savings time, and their validity period. List entries shall be
       * sorted by ValidStarting time.
       *
       * A list entry shall NOT have a ValidStarting time that is smaller than the ValidUntil time of the
       * previous entry.
       *
       * Upon writing this attribute, the server shall recompute its LocalTime.
       *
       * This list may hold up to 20 entries. If a server does not have sufficient storage for 20 entries, it may
       * truncate the list by removing entries with the largest ValidStarting times. The server shall reserve
       * sufficient storage for at least one entry.
       *
       * Over time, the server SHOULD remove any entries which are no longer active from the list.
       *
       * Over time, if the server supports a TimeZoneDatabase, it may update its own list to add additional
       * entries.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.7
       */
      dstOffset: (0, import_Cluster.WritableAttribute)(
        6,
        (0, import_TlvArray.TlvArray)(TimeSync2.TlvDSTOffsetStruct, { maxLength: 20 }),
        { default: [], writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The computed current local time of the server as a epoch-us (Epoch Time in Microseconds). The local time
       * offset of the value is the sum of the currently used TimeZoneEntry’s offset and the currently used DST
       * offset, if any.
       *
       * If the server has not achieved time synchronization, this shall be null.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.8
       */
      localTime: (0, import_Cluster.Attribute)(7, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvEpochUs), { omitChanges: true, default: 0 }),
      /**
       * Indicates whether the server has access to a time zone database. Nodes with a time zone database may
       * update their own DSTOffset attribute to add new entries and may push DSTOffset updates to other Nodes in
       * the same time zone as required.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.9
       */
      timeZoneDatabase: (0, import_Cluster.FixedAttribute)(8, import_TlvBoolean.TlvBoolean, { default: true })
    },
    events: {
      /**
       * This event shall be generated when the server stops applying the current DSTOffset and there are no
       * entries in the list with a larger ValidStarting time, indicating the need to possibly get new DST data.
       *
       * There is no data for this event.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.10.1
       */
      dstTableEmpty: (0, import_Cluster.Event)(0, import_Cluster.EventPriority.Info, import_TlvNoArguments.TlvNoArguments),
      /**
       * This event shall be generated when the server starts or stops applying a DST offset.
       *
       * This event contains a boolean predicate that indicates whether the server is applying the DST offset.
       * When the value is "true", the current DST offset is being applied (i.e, daylight savings time is
       * applied, as opposed to standard time).
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.10.2
       */
      dstStatus: (0, import_Cluster.Event)(1, import_Cluster.EventPriority.Info, import_TlvNoArguments.TlvNoArguments),
      /**
       * This event shall be generated when the server changes its time zone offset or name. It shall NOT be sent
       * for DST changes that are not accompanied by a time zone change.
       *
       * This event returns a structure as follows:
       *
       * Current time zone offset from UTC in seconds.
       *
       * Current time zone name. This name SHOULD use the country/city format specified by the IANA time zone
       * database [https://www.iana.org/time-zones].
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.10.3
       */
      timeZoneStatus: (0, import_Cluster.Event)(2, import_Cluster.EventPriority.Info, TimeSync2.TlvTimeZoneStatusEvent)
    }
  });
  TimeSync2.NtpServerComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * If the server is running an NTP server, this value shall be the port number for the service. If the
       * server is not currently running an NTP server, this value shall be null.
       *
       * This attribute shall be present if this server is capable of providing an NTP server instance. See
       * Section 11.16.15, “Acting as an NTP Server” for more information.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.10
       */
      ntpServerPort: (0, import_Cluster.Attribute)(9, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16), { default: null })
    }
  });
  TimeSync2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    TimeSync2.Base,
    /**
     * Use this factory method to create a TimeSync cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a TimeSync cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...TimeSync2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(TimeSync2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, TimeSync2.NtpClientComponent, { ntpClient: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, TimeSync2.TimeZoneComponent, { timeZone: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, TimeSync2.NtpServerComponent, { ntpServer: true });
      return cluster;
    }
  );
  const NTPC = { ntpClient: true };
  const TZ = { timeZone: true };
  const NTPS = { ntpServer: true };
  TimeSync2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: TimeSync2.Cluster.id,
    name: TimeSync2.Cluster.name,
    revision: TimeSync2.Cluster.revision,
    features: TimeSync2.Cluster.features,
    attributes: {
      ...TimeSync2.Cluster.attributes,
      defaultNtp: import_ClusterFactory.ClusterFactory.AsConditional(TimeSync2.NtpClientComponent.attributes.defaultNtp, { mandatoryIf: [NTPC] }),
      timeZone: import_ClusterFactory.ClusterFactory.AsConditional(TimeSync2.TimeZoneComponent.attributes.timeZone, { mandatoryIf: [TZ] }),
      dstOffset: import_ClusterFactory.ClusterFactory.AsConditional(TimeSync2.TimeZoneComponent.attributes.dstOffset, { mandatoryIf: [TZ] }),
      localTime: import_ClusterFactory.ClusterFactory.AsConditional(TimeSync2.TimeZoneComponent.attributes.localTime, { mandatoryIf: [TZ] }),
      timeZoneDatabase: import_ClusterFactory.ClusterFactory.AsConditional(
        TimeSync2.TimeZoneComponent.attributes.timeZoneDatabase,
        { mandatoryIf: [TZ] }
      ),
      ntpServerPort: import_ClusterFactory.ClusterFactory.AsConditional(
        TimeSync2.NtpServerComponent.attributes.ntpServerPort,
        { mandatoryIf: [NTPS] }
      )
    },
    commands: TimeSync2.Cluster.commands,
    events: {
      dstTableEmpty: import_ClusterFactory.ClusterFactory.AsConditional(TimeSync2.TimeZoneComponent.events.dstTableEmpty, { mandatoryIf: [TZ] }),
      dstStatus: import_ClusterFactory.ClusterFactory.AsConditional(TimeSync2.TimeZoneComponent.events.dstStatus, { mandatoryIf: [TZ] }),
      timeZoneStatus: import_ClusterFactory.ClusterFactory.AsConditional(
        TimeSync2.TimeZoneComponent.events.timeZoneStatus,
        { mandatoryIf: [TZ] }
      )
    }
  });
})(TimeSync || (TimeSync = {}));
const TimeSyncCluster = TimeSync.Cluster;
