// @project-chip/matter.js/dist/cjs/cluster/definitions/UnitLocalizationCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var UnitLocalizationCluster_exports = {};
__export(UnitLocalizationCluster_exports, {
  UnitLocalization: () => UnitLocalization,
  UnitLocalizationCluster: () => UnitLocalizationCluster
});
module.exports = __toCommonJS(UnitLocalizationCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var UnitLocalization;
((UnitLocalization2) => {
  let TempUnit;
  ((TempUnit2) => {
    TempUnit2[TempUnit2["Fahrenheit"] = 0] = "Fahrenheit";
    TempUnit2[TempUnit2["Celsius"] = 1] = "Celsius";
    TempUnit2[TempUnit2["Kelvin"] = 2] = "Kelvin";
  })(TempUnit = UnitLocalization2.TempUnit || (UnitLocalization2.TempUnit = {}));
  let Feature;
  ((Feature2) => {
    Feature2["TemperatureUnit"] = "TemperatureUnit";
  })(Feature = UnitLocalization2.Feature || (UnitLocalization2.Feature = {}));
  UnitLocalization2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 45,
    name: "UnitLocalization",
    revision: 1,
    features: {
      /**
       * TemperatureUnit
       *
       * The Node can be configured to use different units of temperature when conveying values to a user.
       */
      temperatureUnit: (0, import_BitmapSchema.BitFlag)(0)
    }
  });
  UnitLocalization2.TemperatureUnitComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The TemperatureUnit attribute shall indicate the unit for the Node to use only when conveying
       * temperature in communication to the user. If provided, this value shall take priority over any unit
       * implied through the ActiveLocale Attribute.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.5.6.1
       */
      temperatureUnit: (0, import_Cluster.WritableAttribute)(
        0,
        (0, import_TlvNullable.TlvNullable)((0, import_TlvNumber.TlvEnum)()),
        { persistent: true, default: null, writeAcl: import_Cluster.AccessLevel.Manage }
      )
    }
  });
  UnitLocalization2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    UnitLocalization2.Base,
    /**
     * Use this factory method to create an UnitLocalization cluster with support for optional features. Include
     * each {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns an UnitLocalization cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...UnitLocalization2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(UnitLocalization2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(cluster, UnitLocalization2.TemperatureUnitComponent, { temperatureUnit: true });
      return cluster;
    }
  );
  const TEMP = { temperatureUnit: true };
  UnitLocalization2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: UnitLocalization2.Cluster.id,
    name: UnitLocalization2.Cluster.name,
    revision: UnitLocalization2.Cluster.revision,
    features: UnitLocalization2.Cluster.features,
    attributes: {
      temperatureUnit: import_ClusterFactory.ClusterFactory.AsConditional(
        UnitLocalization2.TemperatureUnitComponent.attributes.temperatureUnit,
        { mandatoryIf: [TEMP] }
      )
    }
  });
})(UnitLocalization || (UnitLocalization = {}));
const UnitLocalizationCluster = UnitLocalization.Cluster;
