// @project-chip/matter.js/dist/cjs/cluster/definitions/ValidProxiesCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ValidProxiesCluster_exports = {};
__export(ValidProxiesCluster_exports, {
  ValidProxies: () => ValidProxies,
  ValidProxiesCluster: () => ValidProxiesCluster
});
module.exports = __toCommonJS(ValidProxiesCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_NodeId = require("../../datatype/NodeId.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ValidProxies;
((ValidProxies2) => {
  ValidProxies2.TlvValidProxyStruct = (0, import_TlvObject.TlvObject)({ nodeId: (0, import_TlvObject.TlvField)(1, import_NodeId.TlvNodeId) });
  ValidProxies2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 68,
    name: "ValidProxies",
    revision: 1,
    attributes: {
      /**
       * List of valid proxies that can proxy this Node. Each entry in this list is fabric-scoped.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.15.15.5.1
       */
      validProxyList: (0, import_Cluster.WritableFixedAttribute)(
        0,
        (0, import_TlvArray.TlvArray)(ValidProxies2.TlvValidProxyStruct),
        { persistent: true, default: [] }
      )
    },
    commands: {
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.15.15.6
       */
      getValidProxiesRequest: (0, import_Cluster.Command)(0, import_TlvNoArguments.TlvNoArguments, 1, import_TlvNoArguments.TlvNoArguments)
    }
  });
})(ValidProxies || (ValidProxies = {}));
const ValidProxiesCluster = ValidProxies.Cluster;
