// @project-chip/matter.js/dist/cjs/cluster/definitions/WakeOnLanCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var WakeOnLanCluster_exports = {};
__export(WakeOnLanCluster_exports, {
  WakeOnLan: () => WakeOnLan,
  WakeOnLanCluster: () => WakeOnLanCluster
});
module.exports = __toCommonJS(WakeOnLanCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvString = require("../../tlv/TlvString.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var WakeOnLan;
((WakeOnLan2) => {
  WakeOnLan2.Cluster = import_ClusterFactory.ClusterFactory.Definition({
    id: 1283,
    name: "WakeOnLan",
    revision: 1,
    attributes: {
      /**
       * This shall indicate the current MAC address of the device. Only 48-bit MAC Addresses shall be used for
       * this attribute as required by the Wake on LAN protocol.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.10.2.1
       */
      macAddress: (0, import_Cluster.OptionalFixedAttribute)(0, import_TlvString.TlvByteString.bound({ minLength: 6, maxLength: 32 })),
      /**
       * This shall indicate the current link-local address of the device. Only 128-bit IPv6 link-local addresses
       * shall be used for this attribute.
       *
       * NOTE
       *
       * Some companies may consider MAC Address to be protected data subject to PII handling considerations and
       * will therefore choose not to include it or read it. The MAC Address can often be determined using ARP in
       * IPv4 or NDP in IPv6.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.10.2.2
       */
      linkLocalAddress: (0, import_Cluster.OptionalFixedAttribute)(1, import_TlvString.TlvByteString.bound({ length: 16 }))
    }
  });
})(WakeOnLan || (WakeOnLan = {}));
const WakeOnLanCluster = WakeOnLan.Cluster;
