// @project-chip/matter.js/dist/cjs/cluster/definitions/WindowCoveringCluster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var WindowCoveringCluster_exports = {};
__export(WindowCoveringCluster_exports, {
  WindowCovering: () => WindowCovering,
  WindowCoveringCluster: () => WindowCoveringCluster
});
module.exports = __toCommonJS(WindowCoveringCluster_exports);
var import_ClusterFactory = require("../../cluster/ClusterFactory.js");
var import_BitmapSchema = require("../../schema/BitmapSchema.js");
var import_Cluster = require("../../cluster/Cluster.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvNoArguments = require("../../tlv/TlvNoArguments.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var WindowCovering;
((WindowCovering2) => {
  let WindowCoveringType;
  ((WindowCoveringType2) => {
    WindowCoveringType2[WindowCoveringType2["Rollershade"] = 0] = "Rollershade";
    WindowCoveringType2[WindowCoveringType2["Rollershade2Motor"] = 1] = "Rollershade2Motor";
    WindowCoveringType2[WindowCoveringType2["RollershadeExterior"] = 2] = "RollershadeExterior";
    WindowCoveringType2[WindowCoveringType2["RollershadeExterior2Motor"] = 3] = "RollershadeExterior2Motor";
    WindowCoveringType2[WindowCoveringType2["Drapery"] = 4] = "Drapery";
    WindowCoveringType2[WindowCoveringType2["Awning"] = 5] = "Awning";
    WindowCoveringType2[WindowCoveringType2["Shutter"] = 6] = "Shutter";
    WindowCoveringType2[WindowCoveringType2["TiltBlindTiltOnly"] = 7] = "TiltBlindTiltOnly";
    WindowCoveringType2[WindowCoveringType2["TiltBlindLift"] = 8] = "TiltBlindLift";
    WindowCoveringType2[WindowCoveringType2["ProjectorScreen"] = 9] = "ProjectorScreen";
    WindowCoveringType2[WindowCoveringType2["Unknown"] = 255] = "Unknown";
  })(WindowCoveringType = WindowCovering2.WindowCoveringType || (WindowCovering2.WindowCoveringType = {}));
  WindowCovering2.ConfigStatus = {
    /**
     * Operational: This status bit defines if the Window Covering is operational.The SafetyStatus & Mode
     * attributes might affect this bit
     */
    operational: (0, import_BitmapSchema.BitFlag)(0),
    /**
     * deprecated
     */
    onlineReserved: (0, import_BitmapSchema.BitFlag)(1),
    /**
     * Reversal: This status bit identifies if the directions of the lift/slide movements have been reversed in
     * order for commands (e.g: Open, Close, GoTos) to match the physical installation conditionsThis bit can be
     * adjusted by setting the appropriate reversal bit value in the Mode attribute
     */
    liftMovementReversed: (0, import_BitmapSchema.BitFlag)(2),
    /**
     * Control - Lift: This status bit identifies if the window covering supports the Position Aware Lift Control
     */
    liftPositionAware: (0, import_BitmapSchema.BitFlag)(3),
    /**
     * Control - Tilt: This status bit identifies if the window covering supports the Position Aware Tilt Control
     */
    tiltPositionAware: (0, import_BitmapSchema.BitFlag)(4),
    /**
     * Encoder - Lift: This status bit identifies if a Position Aware Controlled Window Covering is employing an
     * encoder for positioning the height of the window covering.
     */
    liftEncoderControlled: (0, import_BitmapSchema.BitFlag)(5),
    /**
     * Encoder - Tilt: This status bit identifies if a Position Aware Controlled Window Covering is employing an
     * encoder for tilting the window covering.
     */
    tiltEncoderControlled: (0, import_BitmapSchema.BitFlag)(6)
  };
  WindowCovering2.OperationalStatus = {
    /**
     * Movement status of the cover
     */
    global: (0, import_BitmapSchema.BitFieldEnum)(0, 2),
    /**
     * Movement status of the cover's lift function
     */
    lift: (0, import_BitmapSchema.BitFieldEnum)(2, 2),
    /**
     * Movement status of the cover's tilt function
     */
    tilt: (0, import_BitmapSchema.BitFieldEnum)(4, 2)
  };
  let MovementStatus;
  ((MovementStatus2) => {
    MovementStatus2[MovementStatus2["Stopped"] = 0] = "Stopped";
    MovementStatus2[MovementStatus2["Opening"] = 1] = "Opening";
    MovementStatus2[MovementStatus2["Closing"] = 2] = "Closing";
  })(MovementStatus = WindowCovering2.MovementStatus || (WindowCovering2.MovementStatus = {}));
  let EndProductType;
  ((EndProductType2) => {
    EndProductType2[EndProductType2["RollerShade"] = 0] = "RollerShade";
    EndProductType2[EndProductType2["RomanShade"] = 1] = "RomanShade";
    EndProductType2[EndProductType2["BalloonShade"] = 2] = "BalloonShade";
    EndProductType2[EndProductType2["WovenWood"] = 3] = "WovenWood";
    EndProductType2[EndProductType2["PleatedShade"] = 4] = "PleatedShade";
    EndProductType2[EndProductType2["CellularShade"] = 5] = "CellularShade";
    EndProductType2[EndProductType2["LayeredShade"] = 6] = "LayeredShade";
    EndProductType2[EndProductType2["LayeredShade2D"] = 7] = "LayeredShade2D";
    EndProductType2[EndProductType2["SheerShade"] = 8] = "SheerShade";
    EndProductType2[EndProductType2["TiltOnlyInteriorBlind"] = 9] = "TiltOnlyInteriorBlind";
    EndProductType2[EndProductType2["InteriorBlind"] = 10] = "InteriorBlind";
    EndProductType2[EndProductType2["VerticalBlindStripCurtain"] = 11] = "VerticalBlindStripCurtain";
    EndProductType2[EndProductType2["InteriorVenetianBlind"] = 12] = "InteriorVenetianBlind";
    EndProductType2[EndProductType2["ExteriorVenetianBlind"] = 13] = "ExteriorVenetianBlind";
    EndProductType2[EndProductType2["LateralLeftCurtain"] = 14] = "LateralLeftCurtain";
    EndProductType2[EndProductType2["LateralRightCurtain"] = 15] = "LateralRightCurtain";
    EndProductType2[EndProductType2["CentralCurtain"] = 16] = "CentralCurtain";
    EndProductType2[EndProductType2["RollerShutter"] = 17] = "RollerShutter";
    EndProductType2[EndProductType2["ExteriorVerticalScreen"] = 18] = "ExteriorVerticalScreen";
    EndProductType2[EndProductType2["AwningTerrace"] = 19] = "AwningTerrace";
    EndProductType2[EndProductType2["AwningVerticalScreen"] = 20] = "AwningVerticalScreen";
    EndProductType2[EndProductType2["TiltOnlyPergola"] = 21] = "TiltOnlyPergola";
    EndProductType2[EndProductType2["SwingingShutter"] = 22] = "SwingingShutter";
    EndProductType2[EndProductType2["SlidingShutter"] = 23] = "SlidingShutter";
    EndProductType2[EndProductType2["Unknown"] = 255] = "Unknown";
  })(EndProductType = WindowCovering2.EndProductType || (WindowCovering2.EndProductType = {}));
  WindowCovering2.Mode = {
    /**
     * Disables (0) or Enables (1) Lift reversal
     */
    motorDirectionReversed: (0, import_BitmapSchema.BitFlag)(0),
    /**
     * Disabled (0) or Enabled (1) placing the Window Covering into calibration Mode where limits are either setup
     * using tools or learned by the Window Covering by doing self-calibration.If in calibration mode, all commands
     * (e.g: UpOrOpen, DownOrClose, GoTos) that can result in movement, could be accepted and result in a
     * self-calibration being initiated before the command is executed. In case the Window Covering does not have
     * the ability or is not able to perform a self-calibration, the command SHOULD be ignored and a FAILURE status
     * SHOULD be returned.In a write interaction, setting this bit to 0, while the device is in calibration mode,
     * is not allowed and SHALL generate a FAILURE error status. In order to leave calibration mode, the device
     * must perform its calibration routine, either as a self- calibration or assisted by external tool(s),
     * depending on the device/manufacturer implementation.A manufacturer might choose to set the operational bit
     * to its not operational value, if applicable during calibration mode
     */
    calibrationMode: (0, import_BitmapSchema.BitFlag)(1),
    /**
     * Disables (0) or Enables (1) placing the Window Covering into Maintenance Mode where it cannot be moved over
     * the network or by a switch connected to a Local Switch Input.While in maintenance mode, all commands (e.g:
     * UpOrOpen, DownOrClose, GoTos) that can result in movement, must be ignored and respond with a BUSY status.
     * Additionally, the operational bit of the ConfigStatus attribute should be set to its not operational value.
     */
    maintenanceMode: (0, import_BitmapSchema.BitFlag)(2),
    /**
     * Disables (0) or Enables (1) the display of any feedback LEDs resident especially on the packaging of an
     * endpoint where they may cause distraction to the occupant.
     */
    ledFeedback: (0, import_BitmapSchema.BitFlag)(3)
  };
  WindowCovering2.SafetyStatus = {
    /**
     * Movement commands are ignored (locked out). e.g. not granted authorization, outside some time/date range.
     */
    remoteLockout: (0, import_BitmapSchema.BitFlag)(0),
    /**
     * Tampering detected on sensors or any other safety equipment. Ex: a device has been forcedly moved without
     * its actuator(s).
     */
    tamperDetection: (0, import_BitmapSchema.BitFlag)(1),
    /**
     * Communication failure to sensors or other safety equipment.
     */
    failedCommunication: (0, import_BitmapSchema.BitFlag)(2),
    /**
     * Device has failed to reach the desired position. e.g. with Position Aware device, time expired before
     * TargetPosition is reached.
     */
    positionFailure: (0, import_BitmapSchema.BitFlag)(3),
    /**
     * Motor(s) and/or electric circuit thermal protection activated.
     */
    thermalProtection: (0, import_BitmapSchema.BitFlag)(4),
    /**
     * An obstacle is preventing actuator movement.
     */
    obstacleDetected: (0, import_BitmapSchema.BitFlag)(5),
    /**
     * Device has power related issue or limitation e.g. device is running w/ the help of a backup battery or power
     * might not be fully available at the moment.
     */
    power: (0, import_BitmapSchema.BitFlag)(6),
    /**
     * Local safety sensor (not a direct obstacle) is preventing movements (e.g. Safety EU Standard EN60335).
     */
    stopInput: (0, import_BitmapSchema.BitFlag)(7),
    /**
     * Mechanical problem related to the motor(s) detected.
     */
    motorJammed: (0, import_BitmapSchema.BitFlag)(8),
    /**
     * PCB, fuse and other electrics problems.
     */
    hardwareFailure: (0, import_BitmapSchema.BitFlag)(9),
    /**
     * Actuator is manually operated and is preventing actuator movement (e.g. actuator is disengaged/decoupled).
     */
    manualOperation: (0, import_BitmapSchema.BitFlag)(10),
    /**
     * Protection is activated.
     */
    protection: (0, import_BitmapSchema.BitFlag)(11)
  };
  WindowCovering2.TlvGoToLiftPercentageRequest = (0, import_TlvObject.TlvObject)({ liftPercent100thsValue: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvPercent100ths) });
  WindowCovering2.TlvGoToTiltPercentageRequest = (0, import_TlvObject.TlvObject)({ tiltPercent100thsValue: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvPercent100ths) });
  WindowCovering2.TlvGoToLiftValueRequest = (0, import_TlvObject.TlvObject)({ liftValue: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16) });
  WindowCovering2.TlvGoToTiltValueRequest = (0, import_TlvObject.TlvObject)({ tiltValue: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16) });
  let Feature;
  ((Feature2) => {
    Feature2["Lift"] = "Lift";
    Feature2["Tilt"] = "Tilt";
    Feature2["PositionAwareLift"] = "PositionAwareLift";
    Feature2["AbsolutePosition"] = "AbsolutePosition";
    Feature2["PositionAwareTilt"] = "PositionAwareTilt";
  })(Feature = WindowCovering2.Feature || (WindowCovering2.Feature = {}));
  WindowCovering2.Base = import_ClusterFactory.ClusterFactory.Definition({
    id: 258,
    name: "WindowCovering",
    revision: 5,
    features: {
      /**
       * Lift
       *
       * Lift Control and behavior for lifting/sliding window coverings
       */
      lift: (0, import_BitmapSchema.BitFlag)(0),
      /**
       * Tilt
       *
       * Tilt Control and behavior for tilting window coverings
       */
      tilt: (0, import_BitmapSchema.BitFlag)(1),
      /**
       * PositionAwareLift
       *
       * Position Aware lift control is supported.
       */
      positionAwareLift: (0, import_BitmapSchema.BitFlag)(2),
      /**
       * AbsolutePosition
       *
       * Absolute positioning is supported.
       */
      absolutePosition: (0, import_BitmapSchema.BitFlag)(3),
      /**
       * PositionAwareTilt
       *
       * Position Aware tilt control is supported.
       */
      positionAwareTilt: (0, import_BitmapSchema.BitFlag)(4)
    },
    attributes: {
      /**
       * The Type attribute identifies the type of window covering being controlled by this endpoint and shall be
       * set to one of the non-reserved values in the table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.1
       */
      type: (0, import_Cluster.FixedAttribute)(0, (0, import_TlvNumber.TlvEnum)(), { default: 0 /* Rollershade */ }),
      /**
       * The ConfigStatus attribute makes configuration and status information available. To change settings,
       * devices shall write to the Mode attribute of the Window Covering Settings Attribute Set. The behavior
       * causing the setting or clearing of each bit is vendor specific. See table below for details on each bit.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.8
       */
      configStatus: (0, import_Cluster.Attribute)(
        7,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, WindowCovering2.ConfigStatus),
        {
          persistent: true,
          default: (0, import_BitmapSchema.BitsFromPartial)(WindowCovering2.ConfigStatus, { operational: true, onlineReserved: true })
        }
      ),
      /**
       * The OperationalStatus attribute keeps track of currently ongoing operations and applies to all type of
       * devices. See below for details about the meaning of individual bits.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.15
       */
      operationalStatus: (0, import_Cluster.Attribute)(10, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, WindowCovering2.OperationalStatus)),
      /**
       * The EndProductType attribute identifies the product type in complement of the main category indicated by
       * the Type attribute. The window covering shall set this value to one of the values in the table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.16
       */
      endProductType: (0, import_Cluster.FixedAttribute)(13, (0, import_TlvNumber.TlvEnum)(), { default: 0 /* RollerShade */ }),
      /**
       * The Mode attribute allows configuration of the Window Covering, such as: reversing the motor direction,
       * placing the Window Covering into calibration mode, placing the motor into maintenance mode, disabling
       * the network, and disabling status LEDs. See below for details.
       *
       * In the case a device does not support or implement a specific mode, e.g. the device has a specific
       * installation method and reversal is not relevant or the device does not include a maintenance mode, any
       * write interaction to the Mode attribute, with an unsupported mode bit or any out of bounds bits set,
       * must be ignored and a response containing the status of CONSTRAINT_ERROR will be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.21
       */
      mode: (0, import_Cluster.WritableAttribute)(
        23,
        (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt8, WindowCovering2.Mode),
        { persistent: true, writeAcl: import_Cluster.AccessLevel.Manage }
      ),
      /**
       * The SafetyStatus attribute reflects the state of the safety sensors and the common issues preventing
       * movements. By default for nominal operation all flags are cleared (0). A device might support none, one
       * or several bit flags from this attribute (all optional). See below for details about the meaning of
       * individual bits.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.22
       */
      safetyStatus: (0, import_Cluster.OptionalAttribute)(26, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt16, WindowCovering2.SafetyStatus))
    },
    commands: {
      /**
       * Upon receipt of this command, the Window Covering will adjust its position so the physical lift/slide
       * and tilt is at the maximum open/up position. This will happen as fast as possible. The server attributes
       * shall be updated as follows:
       *
       * if Position Aware feature is supported:
       *
       *   • TargetPositionLiftPercent100ths attribute shall be set to 0.00%.
       *
       *   • TargetPositionTiltPercent100ths attribute shall be set to 0.00%.
       *
       * The server positioning attributes will follow the movements, once the movement has successfully
       * finished, the server attributes shall be updated as follows:
       *
       * if Position Aware feature is supported:
       *
       *   • CurrentPositionLiftPercent100ths attribute shall be 0.00%.
       *
       *   • CurrentPositionLiftPercentage attribute shall be 0%.
       *
       *   • CurrentPositionTiltPercent100ths attribute shall be 0.00%.
       *
       *   • CurrentPositionTiltPercentage attribute shall be 0%. if Absolute Position feature is supported:
       *
       *   • CurrentPositionLift attribute shall be equal to the InstalledOpenLimitLift attribute.
       *
       *   • CurrentPositionTilt attribute shall be equal to the InstalledOpenLimitTilt attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.6.1
       */
      upOrOpen: (0, import_Cluster.Command)(0, import_TlvNoArguments.TlvNoArguments, 0, import_Cluster.TlvNoResponse),
      /**
       * Upon receipt of this command, the Window Covering will adjust its position so the physical lift/slide
       * and tilt is at the maximum closed/down position. This will happen as fast as possible. The server
       * attributes supported shall be updated as follows:
       *
       * if Position Aware feature is supported:
       *
       *   • TargetPositionLiftPercent100ths attribute shall be set to 100.00%.
       *
       *   • TargetPositionTiltPercent100ths attribute shall be set to 100.00%.
       *
       * The server positioning attributes will follow the movements, once the movement has successfully
       * finished, the server attributes shall be updated as follows:
       *
       * if Position Aware feature is supported:
       *
       *   • CurrentPositionLiftPercent100ths attribute shall be 100.00%.
       *
       *   • CurrentPositionLiftPercentage attribute shall be 100%.
       *
       *   • CurrentPositionTiltPercent100ths attribute shall be 100.00%.
       *
       *   • CurrentPositionTiltPercentage attribute shall be 100%. if Absolute Position feature is supported:
       *
       *   • CurrentPositionLift attribute shall be equal to the InstalledClosedLimitLift attribute.
       *
       *   • CurrentPositionTilt attribute shall be equal to the InstalledClosedLimitTilt attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.6.2
       */
      downOrClose: (0, import_Cluster.Command)(1, import_TlvNoArguments.TlvNoArguments, 1, import_Cluster.TlvNoResponse),
      /**
       * Upon receipt of this command, the Window Covering will stop any adjusting to the physical tilt and
       * lift/slide that is currently occurring. The server attributes supported shall be updated as follows:
       *
       *   • TargetPositionLiftPercent100ths attribute will be set to CurrentPositionLiftPercent100ths attribute
       *     value.
       *
       *   • TargetPositionTiltPercent100ths attribute will be set to CurrentPositionTiltPercent100ths attribute
       *     value.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.6.3
       */
      stopMotion: (0, import_Cluster.Command)(2, import_TlvNoArguments.TlvNoArguments, 2, import_Cluster.TlvNoResponse)
    }
  });
  WindowCovering2.LiftAndPositionAwareLiftAndAbsolutePositionComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The PhysicalClosedLimitLift attribute identifies the maximum possible encoder position possible (in
       * centimeters) to position the height of the window covering Lift.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.2
       */
      physicalClosedLimitLift: (0, import_Cluster.OptionalFixedAttribute)(1, import_TlvNumber.TlvUInt16, { default: 0 }),
      /**
       * The CurrentPositionLift attribute identifies the actual Lift position (in centimeters) of the window
       * covering from the fully-open position.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.4
       */
      currentPositionLift: (0, import_Cluster.OptionalAttribute)(
        3,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16.bound({ max: 65534 })),
        { persistent: true, default: null }
      ),
      /**
       * The InstalledOpenLimitLift attribute identifies the Open Limit for Lifting the Window Covering whether
       * position (in centimeters) is encoded or timed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.17
       */
      installedOpenLimitLift: (0, import_Cluster.Attribute)(16, import_TlvNumber.TlvUInt16.bound({ max: 65534 }), { persistent: true, default: 0 }),
      /**
       * The InstalledClosedLimitLift attribute identifies the Closed Limit for Lifting the Window Covering
       * whether position (in centimeters) is encoded or timed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.18
       */
      installedClosedLimitLift: (0, import_Cluster.Attribute)(
        17,
        import_TlvNumber.TlvUInt16.bound({ max: 65534 }),
        { persistent: true, default: 65534 }
      )
    }
  });
  WindowCovering2.TiltAndPositionAwareTiltAndAbsolutePositionComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The PhysicalClosedLimitTilt attribute identifies the maximum possible encoder position possible (tenth
       * of a degrees) to position the angle of the window covering Tilt.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.3
       */
      physicalClosedLimitTilt: (0, import_Cluster.OptionalFixedAttribute)(2, import_TlvNumber.TlvUInt16, { default: 0 }),
      /**
       * The CurrentPositionTilt attribute identifies the actual Tilt position (in tenth of an degree) of the
       * window covering from the fully-open position.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.5
       */
      currentPositionTilt: (0, import_Cluster.OptionalAttribute)(
        4,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16.bound({ max: 65534 })),
        { persistent: true, default: null }
      ),
      /**
       * The InstalledOpenLimitTilt attribute identifies the Open Limit for Tilting the Window Covering whether
       * position (in tenth of a degree) is encoded or timed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.19
       */
      installedOpenLimitTilt: (0, import_Cluster.Attribute)(18, import_TlvNumber.TlvUInt16.bound({ max: 65534 }), { persistent: true, default: 0 }),
      /**
       * The InstalledClosedLimitTilt attribute identifies the Closed Limit for Tilting the Window Covering
       * whether position (in tenth of a degree) is encoded or timed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.20
       */
      installedClosedLimitTilt: (0, import_Cluster.Attribute)(
        19,
        import_TlvNumber.TlvUInt16.bound({ max: 65534 }),
        { persistent: true, default: 65534 }
      )
    }
  });
  WindowCovering2.LiftComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The NumberOfActuationsLift attribute identifies the total number of lift/slide actuations applied to the
       * Window Covering since the device was installed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.6
       */
      numberOfActuationsLift: (0, import_Cluster.OptionalAttribute)(5, import_TlvNumber.TlvUInt16, { persistent: true, default: 0 })
    },
    commands: {
      /**
       * Upon receipt of this command, the server will adjust the window covering to the lift/slide percentage
       * specified in the payload of this command.
       *
       * If the command includes LiftPercent100thsValue, then TargetPositionLiftPercent100ths attribute shall be
       * set to LiftPercent100thsValue. Otherwise the TargetPositionLiftPercent100ths attribute shall be set to
       * LiftPercentageValue * 100.
       *
       * If a client includes LiftPercent100thsValue in the command, the LiftPercentageValue shall be set to to
       * LiftPercent100thsValue / 100, so a legacy server which only supports LiftPercentageValue (not
       * LiftPercent100thsValue) has a value to set the target position.
       *
       * If the server does not support the Position Aware feature, then a zero percentage shall be treated as a
       * UpOrOpen command and a non-zero percentage shall be treated as an DownOrClose command. If the device is
       * only a tilt control device, then the command SHOULD be ignored and a UNSUPPORTED_COMMAND status SHOULD
       * be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.6.5
       */
      goToLiftPercentage: (0, import_Cluster.OptionalCommand)(5, WindowCovering2.TlvGoToLiftPercentageRequest, 5, import_Cluster.TlvNoResponse)
    }
  });
  WindowCovering2.TiltComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The NumberOfActuationsTilt attribute identifies the total number of tilt actuations applied to the
       * Window Covering since the device was installed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.7
       */
      numberOfActuationsTilt: (0, import_Cluster.OptionalAttribute)(6, import_TlvNumber.TlvUInt16, { persistent: true, default: 0 })
    },
    commands: {
      /**
       * Upon receipt of this command, the server will adjust the window covering to the tilt percentage
       * specified in the payload of this command.
       *
       * If the command includes TiltPercent100thsValue, then TargetPositionTiltPercent100ths attribute
       *
       * shall be set to TiltPercent100thsValue. Otherwise the TargetPositionTiltPercent100ths attribute shall be
       * set to TiltPercentageValue * 100.
       *
       * If a client includes TiltPercent100thsValue in the command, the TiltPercentageValue shall be set to to
       * TiltPercent100thsValue / 100, so a legacy server which only supports TiltPercentageValue (not
       * TiltPercent100thsValue) has a value to set the target position.
       *
       * If the server does not support the Position Aware feature, then a zero percentage shall be treated as a
       * UpOrOpen command and a non-zero percentage shall be treated as an DownOrClose command. If the device is
       * only a tilt control device, then the command SHOULD be ignored and a UNSUPPORTED_COMMAND status SHOULD
       * be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.6.7
       */
      goToTiltPercentage: (0, import_Cluster.OptionalCommand)(8, WindowCovering2.TlvGoToTiltPercentageRequest, 8, import_Cluster.TlvNoResponse)
    }
  });
  WindowCovering2.LiftAndPositionAwareLiftComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The CurrentPositionLiftPercentage attribute identifies the actual position as a percentage from 0% to
       * 100% with 1% default step. This attribute is equal to CurrentPositionLiftPercent100ths attribute divided
       * by 100.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.11
       */
      currentPositionLiftPercentage: (0, import_Cluster.OptionalAttribute)(
        8,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvPercent),
        { scene: true, persistent: true, default: null }
      ),
      /**
       * The TargetPositionLiftPercent100ths attribute identifies the position where the Window Covering Lift
       * will go or is moving to as a percentage.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.13
       */
      targetPositionLiftPercent100ths: (0, import_Cluster.Attribute)(
        11,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvPercent100ths),
        { scene: true, default: null }
      ),
      /**
       * The CurrentPositionLiftPercent100ths attribute identifies the actual position as a percentage with a
       * minimal step of 0.01%. E.g Max 10000 equals 100.00%.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.9
       */
      currentPositionLiftPercent100ths: (0, import_Cluster.Attribute)(
        14,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvPercent100ths),
        { persistent: true, default: null }
      )
    },
    commands: {
      /**
       * Upon receipt of this command, the server will adjust the window covering to the lift/slide percentage
       * specified in the payload of this command.
       *
       * If the command includes LiftPercent100thsValue, then TargetPositionLiftPercent100ths attribute shall be
       * set to LiftPercent100thsValue. Otherwise the TargetPositionLiftPercent100ths attribute shall be set to
       * LiftPercentageValue * 100.
       *
       * If a client includes LiftPercent100thsValue in the command, the LiftPercentageValue shall be set to to
       * LiftPercent100thsValue / 100, so a legacy server which only supports LiftPercentageValue (not
       * LiftPercent100thsValue) has a value to set the target position.
       *
       * If the server does not support the Position Aware feature, then a zero percentage shall be treated as a
       * UpOrOpen command and a non-zero percentage shall be treated as an DownOrClose command. If the device is
       * only a tilt control device, then the command SHOULD be ignored and a UNSUPPORTED_COMMAND status SHOULD
       * be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.6.5
       */
      goToLiftPercentage: (0, import_Cluster.Command)(5, WindowCovering2.TlvGoToLiftPercentageRequest, 5, import_Cluster.TlvNoResponse)
    }
  });
  WindowCovering2.TiltAndPositionAwareTiltComponent = import_ClusterFactory.ClusterFactory.Component({
    attributes: {
      /**
       * The CurrentPositionTiltPercentage attribute identifies the actual position as a percentage from 0% to
       * 100% with 1% default step. This attribute is equal to CurrentPositionTiltPercent100ths attribute divided
       * by 100.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.12
       */
      currentPositionTiltPercentage: (0, import_Cluster.OptionalAttribute)(
        9,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvPercent),
        { scene: true, persistent: true, default: null }
      ),
      /**
       * The TargetPositionTiltPercent100ths attribute identifies the position where the Window Covering Tilt
       * will go or is moving to as a percentage.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.14
       */
      targetPositionTiltPercent100ths: (0, import_Cluster.Attribute)(
        12,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvPercent100ths),
        { scene: true, default: null }
      ),
      /**
       * The CurrentPositionTiltPercent100ths attribute identifies the actual position as a percentage with a
       * minimal step of 0.01%. E.g Max 10000 equals 100.00%.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.5.10
       */
      currentPositionTiltPercent100ths: (0, import_Cluster.Attribute)(
        15,
        (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvPercent100ths),
        { persistent: true, default: null }
      )
    },
    commands: {
      /**
       * Upon receipt of this command, the server will adjust the window covering to the tilt percentage
       * specified in the payload of this command.
       *
       * If the command includes TiltPercent100thsValue, then TargetPositionTiltPercent100ths attribute
       *
       * shall be set to TiltPercent100thsValue. Otherwise the TargetPositionTiltPercent100ths attribute shall be
       * set to TiltPercentageValue * 100.
       *
       * If a client includes TiltPercent100thsValue in the command, the TiltPercentageValue shall be set to to
       * TiltPercent100thsValue / 100, so a legacy server which only supports TiltPercentageValue (not
       * TiltPercent100thsValue) has a value to set the target position.
       *
       * If the server does not support the Position Aware feature, then a zero percentage shall be treated as a
       * UpOrOpen command and a non-zero percentage shall be treated as an DownOrClose command. If the device is
       * only a tilt control device, then the command SHOULD be ignored and a UNSUPPORTED_COMMAND status SHOULD
       * be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.6.7
       */
      goToTiltPercentage: (0, import_Cluster.Command)(8, WindowCovering2.TlvGoToTiltPercentageRequest, 8, import_Cluster.TlvNoResponse)
    }
  });
  WindowCovering2.LiftAndAbsolutePositionComponent = import_ClusterFactory.ClusterFactory.Component({
    commands: {
      /**
       * Upon receipt of this command, the Window Covering will adjust the window so the physical lift/slide is
       * at the value specified in the payload of this command as long as that value is not larger than
       * InstalledOpenLimitLift attribute and not smaller than InstalledClosedLimitLift attribute. Once the
       * command is received the TargetPositionLiftPercent100ths attribute will update its value accordingly. If
       * the value is out of bounds a response containing the status of CONSTRAINT_ERROR will be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.6.4
       */
      goToLiftValue: (0, import_Cluster.OptionalCommand)(4, WindowCovering2.TlvGoToLiftValueRequest, 4, import_Cluster.TlvNoResponse)
    }
  });
  WindowCovering2.TiltAndAbsolutePositionComponent = import_ClusterFactory.ClusterFactory.Component({
    commands: {
      /**
       * Upon receipt of this command, the Window Covering will adjust the window so the physical tilt is at the
       * tilt value specified in the payload of this command as long as that value is not larger than
       * InstalledOpenLimitTilt attribute and not smaller than InstalledClosedLimitTilt attribute. Once the
       * command is received the TargetPositionTiltPercent100ths attribute will update its value accordingly. If
       * the tilt value is out of bounds a response containing the status of CONSTRAINT_ERROR will be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.3.6.6
       */
      goToTiltValue: (0, import_Cluster.OptionalCommand)(7, WindowCovering2.TlvGoToTiltValueRequest, 7, import_Cluster.TlvNoResponse)
    }
  });
  WindowCovering2.Cluster = import_ClusterFactory.ClusterFactory.Extensible(
    WindowCovering2.Base,
    /**
     * Use this factory method to create a WindowCovering cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a WindowCovering cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      import_ClusterFactory.ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = import_ClusterFactory.ClusterFactory.Definition({
        ...WindowCovering2.Base,
        supportedFeatures: (0, import_BitmapSchema.BitFlags)(WindowCovering2.Base.features, ...features)
      });
      import_ClusterFactory.ClusterFactory.extend(
        cluster,
        WindowCovering2.LiftAndPositionAwareLiftAndAbsolutePositionComponent,
        { lift: true, positionAwareLift: true, absolutePosition: true }
      );
      import_ClusterFactory.ClusterFactory.extend(
        cluster,
        WindowCovering2.TiltAndPositionAwareTiltAndAbsolutePositionComponent,
        { tilt: true, positionAwareTilt: true, absolutePosition: true }
      );
      import_ClusterFactory.ClusterFactory.extend(cluster, WindowCovering2.LiftComponent, { lift: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, WindowCovering2.TiltComponent, { tilt: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, WindowCovering2.LiftAndPositionAwareLiftComponent, { lift: true, positionAwareLift: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, WindowCovering2.TiltAndPositionAwareTiltComponent, { tilt: true, positionAwareTilt: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, WindowCovering2.LiftAndAbsolutePositionComponent, { lift: true, absolutePosition: true });
      import_ClusterFactory.ClusterFactory.extend(cluster, WindowCovering2.TiltAndAbsolutePositionComponent, { tilt: true, absolutePosition: true });
      import_ClusterFactory.ClusterFactory.prevent(
        cluster,
        { positionAwareLift: true, lift: false },
        { positionAwareTilt: true, tilt: false },
        { lift: false, tilt: false }
      );
      return cluster;
    }
  );
  const LF_PA_LF_ABS = { lift: true, positionAwareLift: true, absolutePosition: true };
  const TL_PA_TL_ABS = { tilt: true, positionAwareTilt: true, absolutePosition: true };
  const LF = { lift: true };
  const TL = { tilt: true };
  const LF_PA_LF = { lift: true, positionAwareLift: true };
  const TL_PA_TL = { tilt: true, positionAwareTilt: true };
  const LF_ABS = { lift: true, absolutePosition: true };
  const TL_ABS = { tilt: true, absolutePosition: true };
  WindowCovering2.Complete = import_ClusterFactory.ClusterFactory.Definition({
    id: WindowCovering2.Cluster.id,
    name: WindowCovering2.Cluster.name,
    revision: WindowCovering2.Cluster.revision,
    features: WindowCovering2.Cluster.features,
    attributes: {
      ...WindowCovering2.Cluster.attributes,
      physicalClosedLimitLift: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.LiftAndPositionAwareLiftAndAbsolutePositionComponent.attributes.physicalClosedLimitLift,
        { optionalIf: [LF_PA_LF_ABS] }
      ),
      physicalClosedLimitTilt: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.TiltAndPositionAwareTiltAndAbsolutePositionComponent.attributes.physicalClosedLimitTilt,
        { optionalIf: [TL_PA_TL_ABS] }
      ),
      currentPositionLift: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.LiftAndPositionAwareLiftAndAbsolutePositionComponent.attributes.currentPositionLift,
        { optionalIf: [LF_PA_LF_ABS] }
      ),
      currentPositionTilt: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.TiltAndPositionAwareTiltAndAbsolutePositionComponent.attributes.currentPositionTilt,
        { optionalIf: [TL_PA_TL_ABS] }
      ),
      numberOfActuationsLift: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.LiftComponent.attributes.numberOfActuationsLift,
        { optionalIf: [LF] }
      ),
      numberOfActuationsTilt: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.TiltComponent.attributes.numberOfActuationsTilt,
        { optionalIf: [TL] }
      ),
      currentPositionLiftPercentage: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.LiftAndPositionAwareLiftComponent.attributes.currentPositionLiftPercentage,
        { optionalIf: [LF_PA_LF] }
      ),
      currentPositionTiltPercentage: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.TiltAndPositionAwareTiltComponent.attributes.currentPositionTiltPercentage,
        { optionalIf: [TL_PA_TL] }
      ),
      targetPositionLiftPercent100ths: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.LiftAndPositionAwareLiftComponent.attributes.targetPositionLiftPercent100ths,
        { mandatoryIf: [LF_PA_LF] }
      ),
      targetPositionTiltPercent100ths: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.TiltAndPositionAwareTiltComponent.attributes.targetPositionTiltPercent100ths,
        { mandatoryIf: [TL_PA_TL] }
      ),
      currentPositionLiftPercent100ths: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.LiftAndPositionAwareLiftComponent.attributes.currentPositionLiftPercent100ths,
        { mandatoryIf: [LF_PA_LF] }
      ),
      currentPositionTiltPercent100ths: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.TiltAndPositionAwareTiltComponent.attributes.currentPositionTiltPercent100ths,
        { mandatoryIf: [TL_PA_TL] }
      ),
      installedOpenLimitLift: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.LiftAndPositionAwareLiftAndAbsolutePositionComponent.attributes.installedOpenLimitLift,
        { mandatoryIf: [LF_PA_LF_ABS] }
      ),
      installedClosedLimitLift: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.LiftAndPositionAwareLiftAndAbsolutePositionComponent.attributes.installedClosedLimitLift,
        { mandatoryIf: [LF_PA_LF_ABS] }
      ),
      installedOpenLimitTilt: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.TiltAndPositionAwareTiltAndAbsolutePositionComponent.attributes.installedOpenLimitTilt,
        { mandatoryIf: [TL_PA_TL_ABS] }
      ),
      installedClosedLimitTilt: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.TiltAndPositionAwareTiltAndAbsolutePositionComponent.attributes.installedClosedLimitTilt,
        { mandatoryIf: [TL_PA_TL_ABS] }
      )
    },
    commands: {
      ...WindowCovering2.Cluster.commands,
      goToLiftValue: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.LiftAndAbsolutePositionComponent.commands.goToLiftValue,
        { optionalIf: [LF_ABS] }
      ),
      goToLiftPercentage: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.LiftComponent.commands.goToLiftPercentage,
        { optionalIf: [LF], mandatoryIf: [LF_PA_LF] }
      ),
      goToTiltValue: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.TiltAndAbsolutePositionComponent.commands.goToTiltValue,
        { optionalIf: [TL_ABS] }
      ),
      goToTiltPercentage: import_ClusterFactory.ClusterFactory.AsConditional(
        WindowCovering2.TiltComponent.commands.goToTiltPercentage,
        { optionalIf: [TL], mandatoryIf: [TL_PA_TL] }
      )
    }
  });
})(WindowCovering || (WindowCovering = {}));
const WindowCoveringCluster = WindowCovering.Cluster;
