// @project-chip/matter.js/dist/esm/schema/PairingCodeSchema.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { UnexpectedDataError } from "../common/MatterError.js";
import { VendorId } from "../datatype/VendorId.js";
import { Verhoeff } from "../math/Verhoeff.js";
import { Base38 } from "./Base38Schema.js";
import {
  BitField,
  BitFieldEnum,
  BitFlag,
  BitmapSchema,
  ByteArrayBitmapSchema
} from "./BitmapSchema.js";
import { Schema } from "./Schema.js";
var CommissionningFlowType = /* @__PURE__ */ ((CommissionningFlowType2) => {
  CommissionningFlowType2[CommissionningFlowType2["Standard"] = 0] = "Standard";
  CommissionningFlowType2[CommissionningFlowType2["UserIntent"] = 1] = "UserIntent";
  CommissionningFlowType2[CommissionningFlowType2["Custom"] = 2] = "Custom";
  return CommissionningFlowType2;
})(CommissionningFlowType || {});
const DiscoveryCapabilitiesBitmap = {
  /** Device supports hosting a Soft-AP when not commissioned. */
  softAccessPoint: BitFlag(0),
  /** Device supports BLE for discovery when not commissioned. */
  ble: BitFlag(1),
  /** Device is already on the IP network. */
  onIpNetwork: BitFlag(2)
};
const DiscoveryCapabilitiesSchema = BitmapSchema(DiscoveryCapabilitiesBitmap);
const QrCodeDataSchema = ByteArrayBitmapSchema({
  version: BitField(0, 3),
  vendorId: BitField(3, 16),
  productId: BitField(19, 16),
  flowType: BitFieldEnum(35, 2),
  discoveryCapabilities: BitField(37, 8),
  discriminator: BitField(45, 12),
  passcode: BitField(57, 27)
});
const PREFIX = "MT:";
class QrPairingCodeSchema extends Schema {
  encodeInternal(payloadData) {
    return PREFIX + Base38.encode(QrCodeDataSchema.encode(payloadData));
  }
  decodeInternal(encoded) {
    if (!encoded.startsWith(PREFIX))
      throw new UnexpectedDataError("The pairing code should start with MT:");
    return QrCodeDataSchema.decode(Base38.decode(encoded.slice(PREFIX.length)));
  }
}
const QrPairingCodeCodec = new QrPairingCodeSchema();
class ManualPairingCodeSchema extends Schema {
  encodeInternal({ discriminator, passcode, vendorId, productId }) {
    if (discriminator === void 0)
      throw new UnexpectedDataError("discriminator is required");
    if (discriminator > 4095)
      throw new UnexpectedDataError("discriminator value must be less than 4096");
    let result = "";
    const hasVendorProductIds = vendorId !== void 0 && productId !== void 0;
    result += discriminator >> 10 | (hasVendorProductIds ? 1 << 2 : 0);
    result += ((discriminator & 768) << 6 | passcode & 16383).toString().padStart(5, "0");
    result += (passcode >> 14).toString().padStart(4, "0");
    if (hasVendorProductIds) {
      result += vendorId.toString().padStart(5, "0");
      result += productId.toString().padStart(5, "0");
    }
    result += new Verhoeff().computeChecksum(result);
    return result;
  }
  decodeInternal(encoded) {
    if (encoded.length !== 11 && encoded.length != 21) {
      throw new UnexpectedDataError("Invalid pairing code");
    }
    if (new Verhoeff().computeChecksum(encoded.slice(0, -1)) !== parseInt(encoded.slice(-1))) {
      throw new UnexpectedDataError("Invalid checksum");
    }
    const hasVendorProductIds = !!(parseInt(encoded[0]) & 1 << 2);
    const shortDiscriminator = (parseInt(encoded[0]) & 3) << 2 | parseInt(encoded.slice(1, 6)) >> 14 & 3;
    const passcode = parseInt(encoded.slice(1, 6)) & 16383 | parseInt(encoded.slice(6, 10)) << 14;
    let vendorId;
    let productId;
    if (hasVendorProductIds) {
      vendorId = VendorId(parseInt(encoded.slice(10, 15)));
      productId = parseInt(encoded.slice(15, 20));
    }
    return { shortDiscriminator, passcode, vendorId, productId };
  }
}
const ManualPairingCodeCodec = new ManualPairingCodeSchema();
export {
  CommissionningFlowType,
  DiscoveryCapabilitiesBitmap,
  DiscoveryCapabilitiesSchema,
  ManualPairingCodeCodec,
  QrPairingCodeCodec
};
