// @project-chip/matter.js/dist/esm/fabric/FabricManager.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { InternalError, MatterError, MatterFlowError } from "../common/MatterError.js";
import { FabricIndex } from "../datatype/FabricIndex.js";
import { Fabric } from "./Fabric.js";
class FabricNotFoundError extends MatterError {
}
class FabricTableFullError extends MatterError {
}
class FabricManager {
  constructor(storage, fabricRemoveCallback) {
    this.fabricRemoveCallback = fabricRemoveCallback;
    this.nextFabricIndex = 1;
    this.fabrics = /* @__PURE__ */ new Map();
    this.fabricStorage = storage.createContext("FabricManager");
    const fabrics = this.fabricStorage.get("fabrics", []);
    fabrics.forEach((fabric) => this.addFabric(Fabric.createFromStorageObject(fabric)));
    this.nextFabricIndex = this.fabricStorage.get("nextFabricIndex", this.nextFabricIndex);
  }
  getNextFabricIndex() {
    for (let i = 0; i < 254; i++) {
      const fabricIndex = this.nextFabricIndex++;
      if (this.nextFabricIndex > 254)
        this.nextFabricIndex = 1;
      if (!this.fabrics.has(FabricIndex(fabricIndex))) {
        return FabricIndex(fabricIndex);
      }
    }
    throw new FabricTableFullError("No free fabric index available.");
  }
  persistFabrics() {
    this.fabricStorage.set(
      "fabrics",
      Array.from(this.fabrics.values()).map((fabric) => fabric.toStorageObject())
    );
    this.fabricStorage.set("nextFabricIndex", this.nextFabricIndex);
  }
  addFabric(fabric) {
    const { fabricIndex } = fabric;
    if (this.fabrics.has(fabricIndex)) {
      throw new MatterFlowError(`Fabric with index ${fabricIndex} already exists.`);
    }
    this.fabrics.set(fabricIndex, fabric);
    fabric.addRemoveCallback(() => this.removeFabric(fabricIndex));
    fabric.setPersistCallback(() => this.persistFabrics());
  }
  removeFabric(fabricIndex) {
    const fabric = this.fabrics.get(fabricIndex);
    if (fabric === void 0)
      throw new FabricNotFoundError(
        `Fabric with index ${fabricIndex} cannot be removed because it does not exist.`
      );
    this.fabrics.delete(fabricIndex);
    this.persistFabrics();
    this.fabricRemoveCallback?.(fabricIndex, fabric.rootNodeId);
  }
  getFabrics() {
    return Array.from(this.fabrics.values());
  }
  findFabricFromDestinationId(destinationId, initiatorRandom) {
    for (const fabric of this.fabrics.values()) {
      const candidateDestinationId = fabric.getDestinationId(fabric.nodeId, initiatorRandom);
      if (!candidateDestinationId.equals(destinationId))
        continue;
      return fabric;
    }
    throw new InternalError("Fabric cannot be found from destinationId");
  }
  updateFabric(fabric) {
    const { fabricIndex } = fabric;
    if (!this.fabrics.has(fabricIndex)) {
      throw new FabricNotFoundError(
        `Fabric with index ${fabricIndex} cannot be updated because it does not exist.`
      );
    }
    this.fabrics.set(fabricIndex, fabric);
  }
  async revokeFabric(fabricIndex) {
    const fabric = this.fabrics.get(fabricIndex);
    if (fabric === void 0) {
      throw new MatterFlowError(`Fabric with index ${fabricIndex} does not exist to revoke.`);
    }
    await fabric.remove();
  }
}
export {
  FabricManager,
  FabricNotFoundError,
  FabricTableFullError
};
