// @project-chip/matter.js/dist/esm/cluster/client/AttributeClient.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { tryCatch } from "../../common/TryCatchHandler.js";
import { FabricIndex } from "../../datatype/FabricIndex.js";
import { Globals } from "../../model/index.js";
import { NoAssociatedFabricError } from "../../session/SecureSession.js";
import { AttributeError } from "../Cluster.js";
function createAttributeClient(attribute, name, endpointId, clusterId, interactionClient, present = false) {
  if (attribute.unknown) {
    return new UnknownSupportedAttributeClient(attribute, name, endpointId, clusterId, interactionClient);
  }
  if (present) {
    return new SupportedAttributeClient(attribute, name, endpointId, clusterId, interactionClient);
  }
  return new AttributeClient(attribute, name, endpointId, clusterId, interactionClient);
}
class AttributeClient {
  constructor(attribute, name, endpointId, clusterId, interactionClient) {
    this.attribute = attribute;
    this.name = name;
    this.endpointId = endpointId;
    this.clusterId = clusterId;
    this.interactionClient = interactionClient;
    this.listeners = new Array();
    const { schema, writable, fabricScoped, id } = attribute;
    this.schema = schema;
    this.isWritable = writable;
    this.isFabricScoped = fabricScoped;
    this.id = id;
  }
  /**
   * Set the value of the attribute. When dataVersion parameter is provided the value is only set when the
   * cluster dataVersion of the server matches. If it does not match it is rejected with an Error.
   */
  async set(value, dataVersion) {
    if (!this.isWritable)
      throw new AttributeError(`Attribute ${this.name} is not writable`);
    value = this.schema.injectField(
      value,
      Globals.FabricIndex.id,
      FabricIndex.OMIT_FABRIC,
      (existingFieldIndex) => existingFieldIndex === void 0
    );
    this.schema.validate(value);
    if (this.isFabricScoped) {
      value = this.schema.removeField(
        value,
        Globals.FabricIndex.id,
        (existingFieldIndex) => existingFieldIndex === FabricIndex.OMIT_FABRIC
      );
      value = tryCatch(
        () => {
          const sessionFabric = this.interactionClient.session.getAssociatedFabric();
          return this.schema.removeField(
            value,
            Globals.FabricIndex.id,
            (existingFieldIndex) => existingFieldIndex.index === sessionFabric.fabricIndex
          );
        },
        NoAssociatedFabricError,
        value
      );
    }
    return await this.interactionClient.setAttribute({
      attributeData: {
        endpointId: this.endpointId,
        clusterId: this.clusterId,
        attribute: this.attribute,
        value,
        dataVersion
      }
    });
  }
  /** Get the value of the attribute. Fabric scoped reads are always done with the remote. */
  async get(alwaysRequestFromRemote, isFabricFiltered = true) {
    if (alwaysRequestFromRemote === void 0) {
      alwaysRequestFromRemote = this.isFabricScoped;
    } else if (!alwaysRequestFromRemote && this.isFabricScoped) {
      alwaysRequestFromRemote = true;
    }
    return await this.interactionClient.getAttribute({
      endpointId: this.endpointId,
      clusterId: this.clusterId,
      attribute: this.attribute,
      isFabricFiltered,
      alwaysRequestFromRemote
    });
  }
  /**
   * Get the value with version of the attribute. Fabric scoped reads are always done with the remote.
   * */
  async getWithVersion(alwaysRequestFromRemote, isFabricFiltered = true) {
    if (alwaysRequestFromRemote === void 0) {
      alwaysRequestFromRemote = this.isFabricScoped;
    } else if (!alwaysRequestFromRemote && this.isFabricScoped) {
      alwaysRequestFromRemote = true;
    }
    return await this.interactionClient.getAttributeWithVersion({
      endpointId: this.endpointId,
      clusterId: this.clusterId,
      attribute: this.attribute,
      isFabricFiltered,
      alwaysRequestFromRemote
    });
  }
  /** Subscribe to the attribute. */
  async subscribe(minIntervalFloorSeconds, maxIntervalCeilingSeconds, knownDataVersion, isFabricFiltered = true) {
    return await this.interactionClient.subscribeAttribute({
      endpointId: this.endpointId,
      clusterId: this.clusterId,
      attribute: this.attribute,
      minIntervalFloorSeconds,
      maxIntervalCeilingSeconds,
      isFabricFiltered,
      knownDataVersion,
      listener: this.update.bind(this)
    });
  }
  /**
   * Update the value of the attribute. Just internally used!
   * @private
   */
  update(value) {
    this.listeners.forEach((listener) => listener(value));
  }
  /** Add a listener to the attribute. */
  addListener(listener) {
    this.listeners.push(listener);
  }
  /** Remove a listener from the attribute. */
  removeListener(listener) {
    const entryIndex = this.listeners.indexOf(listener);
    if (entryIndex !== -1) {
      this.listeners.splice(entryIndex, 1);
    }
  }
}
class SupportedAttributeClient extends AttributeClient {
}
class UnknownSupportedAttributeClient extends SupportedAttributeClient {
}
export {
  AttributeClient,
  SupportedAttributeClient,
  UnknownSupportedAttributeClient,
  createAttributeClient
};
