// @project-chip/matter.js/dist/esm/cluster/ClusterHelper.js
/**
 * @license
 * Copyright 2022 The node-matter Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger } from "../log/Logger.js";
import { toHexString } from "../util/Number.js";
import { Cluster } from "./Cluster.js";
import * as Clusters from "./definitions/index.js";
const logger = Logger.get("ClusterHelper");
const AllClustersMap = {
  [Clusters.AccessControlCluster.id]: Clusters.AccessControlCluster,
  [Clusters.AccountLoginCluster.id]: Clusters.AccountLoginCluster,
  [Clusters.ActionsCluster.id]: Clusters.ActionsCluster,
  [Clusters.AdministratorCommissioningCluster.id]: Clusters.AdministratorCommissioning.Complete,
  [Clusters.ApplicationBasicCluster.id]: Clusters.ApplicationBasicCluster,
  [Clusters.ApplicationLauncherCluster.id]: Clusters.ApplicationLauncher.Complete,
  [Clusters.AudioOutputCluster.id]: Clusters.AudioOutput.Complete,
  [Clusters.BallastConfigurationCluster.id]: Clusters.BallastConfigurationCluster,
  [Clusters.BarrierControlCluster.id]: Clusters.BarrierControlCluster,
  [Clusters.BasicInformationCluster.id]: Clusters.BasicInformationCluster,
  [Clusters.BinaryInputBasicCluster.id]: Clusters.BinaryInputBasicCluster,
  [Clusters.BindingCluster.id]: Clusters.BindingCluster,
  [Clusters.BooleanStateCluster.id]: Clusters.BooleanStateCluster,
  [Clusters.BridgedDeviceBasicInformationCluster.id]: Clusters.BridgedDeviceBasicInformationCluster,
  [Clusters.ChannelCluster.id]: Clusters.Channel.Complete,
  [Clusters.ClientMonitoringCluster.id]: Clusters.ClientMonitoringCluster,
  [Clusters.ColorControlCluster.id]: Clusters.ColorControl.Complete,
  [Clusters.ContentLauncherCluster.id]: Clusters.ContentLauncher.Complete,
  [Clusters.DescriptorCluster.id]: Clusters.DescriptorCluster,
  [Clusters.DiagnosticLogsCluster.id]: Clusters.DiagnosticLogsCluster,
  [Clusters.DoorLockCluster.id]: Clusters.DoorLock.Complete,
  [Clusters.ElectricalMeasurementCluster.id]: Clusters.ElectricalMeasurementCluster,
  [Clusters.EthernetNetworkDiagnosticsCluster.id]: Clusters.EthernetNetworkDiagnostics.Complete,
  [Clusters.FanControlCluster.id]: Clusters.FanControl.Complete,
  [Clusters.FaultInjectionCluster.id]: Clusters.FaultInjectionCluster,
  [Clusters.FixedLabelCluster.id]: Clusters.FixedLabelCluster,
  [Clusters.FlowMeasurementCluster.id]: Clusters.FlowMeasurementCluster,
  [Clusters.GeneralCommissioningCluster.id]: Clusters.GeneralCommissioningCluster,
  [Clusters.GeneralDiagnosticsCluster.id]: Clusters.GeneralDiagnosticsCluster,
  [Clusters.GroupKeyManagementCluster.id]: Clusters.GroupKeyManagementCluster,
  [Clusters.GroupsCluster.id]: Clusters.GroupsCluster,
  [Clusters.IdentifyCluster.id]: Clusters.Identify.Complete,
  [Clusters.IlluminanceMeasurementCluster.id]: Clusters.IlluminanceMeasurementCluster,
  [Clusters.KeypadInputCluster.id]: Clusters.KeypadInputCluster,
  [Clusters.LeafWetnessMeasurementCluster.id]: Clusters.LeafWetnessMeasurementCluster,
  [Clusters.LevelControlCluster.id]: Clusters.LevelControl.Complete,
  [Clusters.LocalizationConfigurationCluster.id]: Clusters.LocalizationConfigurationCluster,
  [Clusters.LowPowerCluster.id]: Clusters.LowPowerCluster,
  [Clusters.MediaInputCluster.id]: Clusters.MediaInput.Complete,
  [Clusters.MediaPlaybackCluster.id]: Clusters.MediaPlayback.Complete,
  [Clusters.ModeSelectCluster.id]: Clusters.ModeSelect.Complete,
  [Clusters.NetworkCommissioningCluster.id]: Clusters.NetworkCommissioning.Complete,
  [Clusters.OccupancySensingCluster.id]: Clusters.OccupancySensingCluster,
  [Clusters.OnOffCluster.id]: Clusters.OnOff.Complete,
  [Clusters.OnOffSwitchConfigurationCluster.id]: Clusters.OnOffSwitchConfigurationCluster,
  [Clusters.OperationalCredentialsCluster.id]: Clusters.OperationalCredentialsCluster,
  [Clusters.OtaSoftwareUpdateProviderCluster.id]: Clusters.OtaSoftwareUpdateProviderCluster,
  [Clusters.OtaSoftwareUpdateRequestorCluster.id]: Clusters.OtaSoftwareUpdateRequestorCluster,
  [Clusters.PowerSourceCluster.id]: Clusters.PowerSource.Complete,
  [Clusters.PowerSourceConfigurationCluster.id]: Clusters.PowerSourceConfigurationCluster,
  [Clusters.PressureMeasurementCluster.id]: Clusters.PressureMeasurement.Complete,
  [Clusters.ProxyConfigurationCluster.id]: Clusters.ProxyConfigurationCluster,
  [Clusters.ProxyDiscoveryCluster.id]: Clusters.ProxyDiscoveryCluster,
  [Clusters.PulseWidthModulationCluster.id]: Clusters.PulseWidthModulation.Complete,
  [Clusters.PumpConfigurationAndControlCluster.id]: Clusters.PumpConfigurationAndControl.Complete,
  [Clusters.RelativeHumidityMeasurementCluster.id]: Clusters.RelativeHumidityMeasurementCluster,
  [Clusters.ScenesCluster.id]: Clusters.ScenesCluster,
  [Clusters.SoftwareDiagnosticsCluster.id]: Clusters.SoftwareDiagnostics.Complete,
  [Clusters.SoilMoistureMeasurementCluster.id]: Clusters.SoilMoistureMeasurementCluster,
  [Clusters.SwitchCluster.id]: Clusters.Switch.Complete,
  [Clusters.TargetNavigatorCluster.id]: Clusters.TargetNavigatorCluster,
  [Clusters.TemperatureMeasurementCluster.id]: Clusters.TemperatureMeasurementCluster,
  [Clusters.ThermostatCluster.id]: Clusters.Thermostat.Complete,
  [Clusters.ThermostatUserInterfaceConfigurationCluster.id]: Clusters.ThermostatUserInterfaceConfigurationCluster,
  [Clusters.ThreadNetworkDiagnosticsCluster.id]: Clusters.ThreadNetworkDiagnostics.Complete,
  [Clusters.TimeFormatLocalizationCluster.id]: Clusters.TimeFormatLocalization.Complete,
  [Clusters.TimeSyncCluster.id]: Clusters.TimeSync.Complete,
  [Clusters.UnitLocalizationCluster.id]: Clusters.UnitLocalization.Complete,
  [Clusters.UserLabelCluster.id]: Clusters.UserLabelCluster,
  [Clusters.ValidProxiesCluster.id]: Clusters.ValidProxiesCluster,
  [Clusters.WakeOnLanCluster.id]: Clusters.WakeOnLanCluster,
  [Clusters.WiFiNetworkDiagnosticsCluster.id]: Clusters.WiFiNetworkDiagnostics.Complete,
  [Clusters.WindowCoveringCluster.id]: Clusters.WindowCovering.Complete
};
const clusterAttributeCache = /* @__PURE__ */ new Map();
const clusterEventCache = /* @__PURE__ */ new Map();
const clusterCommandCache = /* @__PURE__ */ new Map();
const UnknownCluster = (clusterId) => Cluster({
  id: clusterId,
  name: `Unknown cluster ${toHex(clusterId)}`,
  revision: 0,
  unknown: true
});
function getClusterNameById(clusterId) {
  return AllClustersMap[clusterId]?.name ?? `Unknown cluster ${toHexString(clusterId)}`;
}
function getClusterById(clusterId) {
  if (AllClustersMap[clusterId] === void 0) {
    logger.info(`Unknown cluster ${toHex(clusterId)} requested: UnknownCluster instance added.`);
    AllClustersMap[clusterId] = UnknownCluster(clusterId);
  }
  return AllClustersMap[clusterId];
}
function getClusterAttributeById(clusterDef, attributeId) {
  if (!clusterAttributeCache.has(clusterDef.id)) {
    const attributeMap2 = /* @__PURE__ */ new Map();
    const { attributes } = clusterDef;
    for (const attributeName in attributes) {
      const attribute = attributes[attributeName];
      attributeMap2.set(attribute.id, { attribute, name: attributeName });
    }
    clusterAttributeCache.set(clusterDef.id, attributeMap2);
    return attributeMap2.get(attributeId);
  }
  const attributeMap = clusterAttributeCache.get(clusterDef.id);
  if (attributeMap === void 0) {
    return void 0;
  }
  return attributeMap.get(attributeId);
}
function getClusterEventById(clusterDef, eventId) {
  if (!clusterEventCache.has(clusterDef.id)) {
    const eventMap2 = /* @__PURE__ */ new Map();
    const { events } = clusterDef;
    for (const eventName in events) {
      const event = events[eventName];
      eventMap2.set(event.id, { event, name: eventName });
    }
    clusterEventCache.set(clusterDef.id, eventMap2);
    return eventMap2.get(eventId);
  }
  const eventMap = clusterEventCache.get(clusterDef.id);
  if (eventMap === void 0) {
    return void 0;
  }
  return eventMap.get(eventId);
}
function getClusterCommandById(clusterDef, commandId) {
  if (!clusterCommandCache.has(clusterDef.id)) {
    const commandMap2 = /* @__PURE__ */ new Map();
    const { commands } = clusterDef;
    for (const commandName in commands) {
      const command = commands[commandName];
      commandMap2.set(command.requestId, { command, name: commandName });
    }
    clusterCommandCache.set(clusterDef.id, commandMap2);
    return commandMap2.get(commandId);
  }
  const commandMap = clusterCommandCache.get(clusterDef.id);
  if (commandMap === void 0) {
    return void 0;
  }
  return commandMap.get(commandId);
}
function toHex(value) {
  return value === void 0 ? "*" : `0x${value.toString(16)}`;
}
function resolveEndpointClusterName(nodeId, endpointId, clusterId) {
  let elementName = nodeId === void 0 ? "" : `${toHex(nodeId)}/`;
  if (endpointId === void 0) {
    elementName += "*";
  } else {
    elementName += `${toHex(endpointId)}`;
  }
  if (clusterId === void 0) {
    return `${elementName}/*`;
  }
  const cluster = getClusterById(clusterId);
  if (cluster.unknown) {
    return `${elementName}/unknown(${toHex(clusterId)})`;
  }
  return `${elementName}/${cluster.name}(${toHex(clusterId)})`;
}
function resolveAttributeName({
  nodeId,
  endpointId,
  clusterId,
  attributeId
}) {
  const endpointClusterName = resolveEndpointClusterName(nodeId, endpointId, clusterId);
  if (endpointId === void 0 || clusterId === void 0 || attributeId === void 0) {
    return `${endpointClusterName}/${toHex(attributeId)}`;
  }
  const cluster = getClusterById(clusterId);
  const attribute = getClusterAttributeById(cluster, attributeId);
  if (attribute === void 0) {
    return `${endpointClusterName}/unknown(${toHex(attributeId)})`;
  }
  return `${endpointClusterName}/${attribute.name}(${toHex(attributeId)})`;
}
function resolveEventName({ nodeId, endpointId, clusterId, eventId }) {
  const endpointClusterName = resolveEndpointClusterName(nodeId, endpointId, clusterId);
  if (endpointId === void 0 || clusterId === void 0 || eventId === void 0) {
    return `${endpointClusterName}/${toHex(eventId)}`;
  }
  const event = getClusterEventById(getClusterById(clusterId), eventId);
  if (event === void 0) {
    return `${endpointClusterName}/unknown(${toHex(eventId)})`;
  }
  return `${endpointClusterName}/${event.name}(${toHex(eventId)})`;
}
function resolveCommandName({ endpointId, clusterId, commandId }) {
  const endpointClusterName = resolveEndpointClusterName(void 0, endpointId, clusterId);
  if (endpointId === void 0 || clusterId === void 0 || commandId === void 0) {
    return `${endpointClusterName}/${toHex(commandId)}`;
  }
  const command = getClusterCommandById(getClusterById(clusterId), commandId);
  if (command === void 0) {
    return `${endpointClusterName}/unknown(${toHex(commandId)})`;
  }
  return `${endpointClusterName}/${command.name}(${toHex(commandId)})`;
}
export {
  AllClustersMap,
  UnknownCluster,
  getClusterAttributeById,
  getClusterById,
  getClusterCommandById,
  getClusterEventById,
  getClusterNameById,
  resolveAttributeName,
  resolveCommandName,
  resolveEventName
};
