// @project-chip/matter.js/dist/esm/cluster/ClusterFactory.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterError } from "../common/MatterError.js";
import { ClusterId } from "../datatype/ClusterId.js";
import { serialize } from "../util/String.js";
import {
  GlobalAttributes
} from "./Cluster.js";
class IllegalClusterError extends MatterError {
}
var ClusterFactory;
((ClusterFactory2) => {
  function Component(component) {
    return {
      attributes: {},
      commands: {},
      events: {},
      ...component
    };
  }
  ClusterFactory2.Component = Component;
  function Definition(definition) {
    return {
      id: ClusterId(definition.id),
      name: definition.name,
      revision: definition.revision,
      features: definition.features ?? {},
      supportedFeatures: definition.supportedFeatures ?? {},
      attributes: {
        ...GlobalAttributes(definition.features ?? {}),
        ...definition.attributes
      },
      commands: definition.commands ?? {},
      events: definition.events ?? {},
      unknown: definition.unknown
    };
  }
  ClusterFactory2.Definition = Definition;
  function Extensible(definition, extender) {
    return {
      ...Definition(definition),
      with: extender
    };
  }
  ClusterFactory2.Extensible = Extensible;
  function ExtensibleOnly(extender) {
    return {
      with: extender
    };
  }
  ClusterFactory2.ExtensibleOnly = ExtensibleOnly;
  function extend(definition, elements, ...applicableFeatures) {
    let applicable = false;
    pool:
      for (const features of applicableFeatures) {
        for (const k in features) {
          if (!!definition.supportedFeatures[k] !== !!features[k]) {
            continue pool;
          }
        }
        applicable = true;
        break;
      }
    if (!applicable) {
      return;
    }
    if (elements.attributes) {
      if (definition.attributes) {
        definition.attributes = { ...definition.attributes, ...elements.attributes };
      } else {
        definition.attributes = elements.attributes;
      }
    }
    if (elements.commands) {
      if (definition.commands) {
        definition.commands = { ...definition.commands, ...elements.commands };
      } else {
        definition.commands = elements.commands;
      }
    }
    if (elements.events) {
      if (definition.events) {
        definition.events = { ...definition.events, ...elements.events };
      } else {
        definition.events = elements.events;
      }
    }
  }
  ClusterFactory2.extend = extend;
  function validateFeatureSelection(features, validFeatures) {
    for (const f of features) {
      if (!validFeatures[f]) {
        throw new IllegalClusterError(`"${f}" is not a valid feature identifier`);
      }
    }
  }
  ClusterFactory2.validateFeatureSelection = validateFeatureSelection;
  function prevent(definition, ...illegalFeatureCombinations) {
    pool:
      for (const bitmap of illegalFeatureCombinations) {
        for (const k in bitmap) {
          if (!!definition.supportedFeatures[k] !== !!bitmap[k]) {
            continue pool;
          }
        }
        throw new IllegalClusterError(
          `Feature combination ${serialize(bitmap)} is disallowed by the Matter specification`
        );
      }
  }
  ClusterFactory2.prevent = prevent;
  function AsConditional(element, { optionalIf, mandatoryIf }) {
    return {
      ...element,
      optional: true,
      isConditional: true,
      optionalIf: optionalIf ?? [],
      mandatoryIf: mandatoryIf ?? []
    };
  }
  ClusterFactory2.AsConditional = AsConditional;
})(ClusterFactory || (ClusterFactory = {}));
export {
  ClusterFactory,
  IllegalClusterError
};
