// @project-chip/matter.js/dist/esm/crypto/Crypto.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import BN from "bn.js";
import { MatterError, NoProviderError } from "../common/MatterError.js";
import { Endian } from "../util/ByteArray.js";
import { DataReader } from "../util/DataReader.js";
const CRYPTO_RANDOM_LENGTH = 32;
const CRYPTO_ENCRYPT_ALGORITHM = "aes-128-ccm";
const CRYPTO_HASH_ALGORITHM = "sha256";
const CRYPTO_EC_CURVE = "prime256v1";
const CRYPTO_EC_KEY_BYTES = 32;
const CRYPTO_AUTH_TAG_LENGTH = 16;
const CRYPTO_SYMMETRIC_KEY_LENGTH = 16;
class CryptoError extends MatterError {
}
class Crypto {
  static {
    this.get = () => {
      throw new NoProviderError("No provider configured");
    };
  }
  static {
    this.encrypt = (key, data, nonce, aad) => Crypto.get().encrypt(key, data, nonce, aad);
  }
  static {
    this.decrypt = (key, data, nonce, aad) => Crypto.get().decrypt(key, data, nonce, aad);
  }
  static {
    this.getRandomData = (length) => Crypto.get().getRandomData(length);
  }
  static {
    this.getRandom = () => Crypto.get().getRandomData(CRYPTO_RANDOM_LENGTH);
  }
  static {
    this.getRandomUInt16 = () => new DataReader(Crypto.get().getRandomData(2), Endian.Little).readUInt16();
  }
  static {
    this.getRandomUInt32 = () => new DataReader(Crypto.get().getRandomData(4), Endian.Little).readUInt32();
  }
  static {
    this.getRandomBigUInt64 = () => new DataReader(Crypto.get().getRandomData(8), Endian.Little).readUInt64();
  }
  static {
    this.getRandomBN = (size, maxValue) => {
      while (true) {
        const random = new BN(Crypto.getRandomData(size));
        if (random < maxValue)
          return random;
      }
    };
  }
  static {
    this.ecdhGeneratePublicKey = () => Crypto.get().ecdhGeneratePublicKey();
  }
  static {
    this.ecdhGeneratePublicKeyAndSecret = (peerPublicKey) => Crypto.get().ecdhGeneratePublicKeyAndSecret(peerPublicKey);
  }
  static {
    this.ecdhGenerateSecret = (peerPublicKey, ecdh) => Crypto.get().ecdhGenerateSecret(peerPublicKey, ecdh);
  }
  static {
    this.hash = (data) => Crypto.get().hash(data);
  }
  static {
    this.pbkdf2 = (secret, salt, iteration, keyLength) => Crypto.get().pbkdf2(secret, salt, iteration, keyLength);
  }
  static {
    this.hkdf = (secret, salt, info, length) => Crypto.get().hkdf(secret, salt, info, length);
  }
  static {
    this.hmac = (key, data) => Crypto.get().hmac(key, data);
  }
  static {
    this.sign = (privateKey, data, dsaEncoding) => Crypto.get().sign(privateKey, data, dsaEncoding);
  }
  static {
    this.verify = (publicKey, data, signature, dsaEncoding) => Crypto.get().verify(publicKey, data, signature, dsaEncoding);
  }
  static {
    this.createKeyPair = () => Crypto.get().createKeyPair();
  }
}
import elliptic from "elliptic";
const { ec } = elliptic;
if (typeof MatterHooks !== "undefined") {
  MatterHooks.cryptoSetup?.(Crypto);
}
export {
  CRYPTO_AUTH_TAG_LENGTH,
  CRYPTO_EC_CURVE,
  CRYPTO_EC_KEY_BYTES,
  CRYPTO_ENCRYPT_ALGORITHM,
  CRYPTO_HASH_ALGORITHM,
  CRYPTO_RANDOM_LENGTH,
  CRYPTO_SYMMETRIC_KEY_LENGTH,
  Crypto,
  CryptoError,
  ec
};
