// @project-chip/matter.js/dist/esm/crypto/Spake2p.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import BN from "bn.js";
import { InternalError } from "../common/MatterError.js";
import { ByteArray, Endian } from "../util/ByteArray.js";
import { DataWriter } from "../util/DataWriter.js";
import { Crypto, ec } from "./Crypto.js";
const P256_CURVE = new ec("p256").curve;
const M = P256_CURVE.decodePoint("02886e2f97ace46e55ba9dd7242579f2993b64e16ef3dcab95afd497333d8fa12f", "hex");
const N = P256_CURVE.decodePoint("03d8bbd6c639c62937b04d997f38c3770719c629d7014d49a24b4f98baa1292b49", "hex");
class Spake2p {
  constructor(context, random, w0) {
    this.context = context;
    this.random = random;
    this.w0 = w0;
  }
  static async computeW0W1({ iterations, salt }, pin) {
    const pinWriter = new DataWriter(Endian.Little);
    pinWriter.writeUInt32(pin);
    const ws = await Crypto.pbkdf2(pinWriter.toByteArray(), salt, iterations, 80);
    const w0 = new BN(ws.slice(0, 40)).mod(P256_CURVE.n);
    const w1 = new BN(ws.slice(40, 80)).mod(P256_CURVE.n);
    return { w0, w1 };
  }
  static async computeW0L(pbkdfParameters, pin) {
    const { w0, w1 } = await this.computeW0W1(pbkdfParameters, pin);
    const L = P256_CURVE.g.mul(w1).encode();
    return { w0, L };
  }
  static create(context, w0) {
    const random = Crypto.getRandomBN(32, P256_CURVE.p);
    return new Spake2p(context, random, w0);
  }
  computeX() {
    const X = P256_CURVE.g.mul(this.random).add(M.mul(this.w0));
    return ByteArray.from(X.encode());
  }
  computeY() {
    const Y = P256_CURVE.g.mul(this.random).add(N.mul(this.w0));
    return ByteArray.from(Y.encode());
  }
  async computeSecretAndVerifiersFromY(w1, X, Y) {
    const YPoint = P256_CURVE.decodePoint(Y);
    if (!YPoint.validate())
      throw new InternalError("Y is not on the curve");
    const yNwo = YPoint.add(N.mul(this.w0).neg());
    const Z = yNwo.mul(this.random);
    const V = yNwo.mul(w1);
    return this.computeSecretAndVerifiers(X, Y, ByteArray.from(Z.encode()), ByteArray.from(V.encode()));
  }
  async computeSecretAndVerifiersFromX(L, X, Y) {
    const XPoint = P256_CURVE.decodePoint(X);
    const LPoint = P256_CURVE.decodePoint(L);
    if (!XPoint.validate())
      throw new InternalError("X is not on the curve");
    const Z = XPoint.add(M.mul(this.w0).neg()).mul(this.random);
    const V = LPoint.mul(this.random);
    return this.computeSecretAndVerifiers(X, Y, ByteArray.from(Z.encode()), ByteArray.from(V.encode()));
  }
  async computeSecretAndVerifiers(X, Y, Z, V) {
    const TT_HASH = this.computeTranscriptHash(X, Y, Z, V);
    const Ka = TT_HASH.slice(0, 16);
    const Ke = TT_HASH.slice(16, 32);
    const KcAB = await Crypto.hkdf(Ka, new ByteArray(0), ByteArray.fromString("ConfirmationKeys"), 32);
    const KcA = KcAB.slice(0, 16);
    const KcB = KcAB.slice(16, 32);
    const hAY = Crypto.hmac(KcA, Y);
    const hBX = Crypto.hmac(KcB, X);
    return { Ke, hAY, hBX };
  }
  computeTranscriptHash(X, Y, Z, V) {
    const TTwriter = new DataWriter(Endian.Little);
    this.addToContext(TTwriter, this.context);
    this.addToContext(TTwriter, ByteArray.fromString(""));
    this.addToContext(TTwriter, ByteArray.fromString(""));
    this.addToContext(TTwriter, ByteArray.from(M.encode()));
    this.addToContext(TTwriter, ByteArray.from(N.encode()));
    this.addToContext(TTwriter, X);
    this.addToContext(TTwriter, Y);
    this.addToContext(TTwriter, Z);
    this.addToContext(TTwriter, V);
    this.addToContext(TTwriter, ByteArray.from(this.w0.toArray()));
    return Crypto.hash(TTwriter.toByteArray());
  }
  addToContext(TTwriter, data) {
    TTwriter.writeUInt64(data.length);
    TTwriter.writeByteArray(data);
  }
}
export {
  Spake2p
};
