// @project-chip/matter.js/dist/esm/model/logic/record-validation/ValidatorImplementation.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ValidatorBuilder } from "./ValidatorBuilder.js";
class ValidatorImplementation {
  constructor(fields, definedFeatures, enabledFeatures) {
    this.fields = fields;
    this.enabledFeatures = enabledFeatures;
    const builder = new ValidatorBuilder(fields, definedFeatures, enabledFeatures);
    this.validate = builder.compile();
    this.logFailure = () => builder.logFailure();
  }
  // Compiled validator invokes to test group values
  testGroup(...args) {
    for (const a of args) {
      if (a === true) {
        return true;
      }
      if (a === null) {
        return false;
      }
    }
    return false;
  }
  // Compiled validator invokes to record active choice.  A counter is kept
  // using a key of the form ".a" in values.  After conformance is tested on
  // each field, choices must be reexamined to ensure choices are correct
  testChoice(name, result, hasValue, num, orMore) {
    if (!this.choices) {
      this.choices = {};
    }
    let choice = this.choices[name];
    if (!choice) {
      choice = this.choices[name] = {
        count: 0,
        num,
        orMore
      };
    }
    if (hasValue) {
      choice.count++;
    }
    return result;
  }
  // Compiled validator invokes to check choices
  checkChoices() {
    for (const name in this.choices) {
      const choice = this.choices[name];
      if (choice.count < choice.num) {
        this.error("CHOICE_THRESHOLD_NOT_MET", name, `Choice minimum of ${choice.num} not met`);
      } else if (choice.count > choice.num && !choice.orMore) {
        this.error("CHOICE_THRESHOLD_EXCEEDED", name, `Choice maximum of ${choice.num} exceeded`);
      }
    }
  }
  // Compiled validator calls to add a validation error
  error(code, source, message) {
    const result = this.result;
    if (!result) {
      return;
    }
    if (!result.errors) {
      result.valid = false;
      result.errors = [];
    }
    result.errors.push({ code, source, message });
  }
}
export {
  ValidatorImplementation
};
