// @project-chip/matter.js/dist/cjs/storage/StorageContext.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var StorageContext_exports = {};
__export(StorageContext_exports, {
  StorageContext: () => StorageContext
});
module.exports = __toCommonJS(StorageContext_exports);
var import_Storage = require("./Storage.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class StorageContext {
  constructor(storage, contexts) {
    this.storage = storage;
    this.contexts = contexts;
  }
  get(key, defaultValue) {
    const value = this.storage.get(this.contexts, key);
    if (value !== void 0)
      return value;
    if (defaultValue === void 0) {
      throw new import_Storage.StorageError(
        `No value found for key ${key} in context ${this.contexts} and no default value specified!`
      );
    }
    return defaultValue;
  }
  has(key) {
    return this.storage.get(this.contexts, key) !== void 0;
  }
  set(key, value) {
    this.storage.set(this.contexts, key, value);
  }
  delete(key) {
    this.storage.delete(this.contexts, key);
  }
  createContext(context) {
    if (context.length === 0)
      throw new import_Storage.StorageError("Context must not be an empty string");
    if (context.includes("."))
      throw new import_Storage.StorageError("Context must not contain dots!");
    return new StorageContext(this.storage, [...this.contexts, context]);
  }
  keys() {
    return this.storage.keys(this.contexts);
  }
  /** Clears all keys in this context */
  clear() {
    for (const key of this.keys()) {
      this.delete(key);
    }
  }
  /** Clears all keys in this context and all created sub-contexts. */
  clearAll() {
    this.storage.clearAll(this.contexts);
  }
}
