// @project-chip/matter.js/dist/esm/net/UdpInterface.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Network, NetworkError } from "./Network.js";
class UdpInterface {
  constructor(server) {
    this.server = server;
  }
  static async create(type, port, host, netInterface) {
    return new UdpInterface(
      await Network.get().createUdpChannel({ listeningPort: port, type, netInterface, listeningAddress: host })
    );
  }
  async openChannel(address) {
    if (address.type !== "udp") {
      throw new NetworkError(`Unsupported address type ${address.type}`);
    }
    const { ip, port } = address;
    return Promise.resolve(new UdpConnection(this.server, ip, port));
  }
  onData(listener) {
    return this.server.onData(
      (_netInterface, peerHost, peerPort, data) => listener(new UdpConnection(this.server, peerHost, peerPort), data)
    );
  }
  async close() {
    this.server.close();
  }
}
class UdpConnection {
  constructor(server, peerAddress, peerPort) {
    this.server = server;
    this.peerAddress = peerAddress;
    this.peerPort = peerPort;
  }
  send(data) {
    return this.server.send(this.peerAddress, this.peerPort, data);
  }
  get name() {
    return `udp://${this.peerAddress}:${this.peerPort}`;
  }
  async close() {
  }
}
export {
  UdpInterface
};
