// @project-chip/matter.js/dist/cjs/model/models/ClusterModel.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClusterModel_exports = {};
__export(ClusterModel_exports, {
  ClusterModel: () => ClusterModel
});
module.exports = __toCommonJS(ClusterModel_exports);
var import_elements = require("../elements/index.js");
var import_AttributeModel = require("./AttributeModel.js");
var import_CommandModel = require("./CommandModel.js");
var import_DatatypeModel = require("./DatatypeModel.js");
var import_EventModel = require("./EventModel.js");
var import_Model = require("./Model.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class ClusterModel extends import_Model.Model {
  constructor(definition) {
    super(definition);
    this.tag = import_elements.ClusterElement.Tag;
    this.isTypeScope = true;
  }
  get attributes() {
    return this.all(import_AttributeModel.AttributeModel);
  }
  get commands() {
    return this.all(import_CommandModel.CommandModel);
  }
  get events() {
    return this.all(import_EventModel.EventModel);
  }
  get datatypes() {
    return this.all(import_DatatypeModel.DatatypeModel);
  }
  get revision() {
    let revision = 1;
    const revisionAttr = this.get(import_AttributeModel.AttributeModel, import_elements.Globals.ClusterRevision.id);
    if (typeof revisionAttr?.default === "number") {
      revision = revisionAttr.default;
    }
    return revision;
  }
  get features() {
    return this.featureMap?.children ?? [];
  }
  get featureMap() {
    return this.get(import_AttributeModel.AttributeModel, import_elements.Globals.FeatureMap.id);
  }
  get children() {
    return super.children;
  }
  set children(children) {
    super.children = children;
  }
  static {
    import_Model.Model.constructors[import_elements.ClusterElement.Tag] = this;
  }
}
