// @project-chip/matter.js/dist/cjs/model/models/MatterModel.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MatterModel_exports = {};
__export(MatterModel_exports, {
  MatterModel: () => MatterModel
});
module.exports = __toCommonJS(MatterModel_exports);
var import_elements = require("../elements/index.js");
var import_standard = require("../standard/index.js");
var import_AttributeModel = require("./AttributeModel.js");
var import_ClusterModel = require("./ClusterModel.js");
var import_DatatypeModel = require("./DatatypeModel.js");
var import_DeviceTypeModel = require("./DeviceTypeModel.js");
var import_FabricModel = require("./FabricModel.js");
var import_Model = require("./Model.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class MatterModel extends import_Model.Model {
  constructor(definition = import_standard.Matter, globals = Object.values(import_elements.Globals)) {
    const children = [...globals, ...definition.children || []];
    super({ ...definition, children });
    this.tag = import_elements.MatterElement.Tag;
    this.isTypeScope = true;
  }
  /**
   * Clusters.
   */
  get clusters() {
    return this.all(import_ClusterModel.ClusterModel);
  }
  /**
   * Device types.
   */
  get deviceTypes() {
    return this.all(import_DeviceTypeModel.DeviceTypeModel);
  }
  /**
   * Global datatypes.
   */
  get datatypes() {
    return this.all(import_DatatypeModel.DatatypeModel);
  }
  /**
   * Global attributes.
   */
  get attributes() {
    return this.all(import_AttributeModel.AttributeModel);
  }
  /**
   * Fabrics.
   */
  get fabrics() {
    return this.all(import_FabricModel.FabricModel);
  }
  get children() {
    return super.children;
  }
  set children(children) {
    super.children = children;
  }
  static {
    import_Model.Model.constructors[import_elements.MatterElement.Tag] = this;
  }
}
