// @project-chip/matter.js/dist/esm/storage/StorageBackendMemory.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { StorageError } from "./Storage.js";
class StorageBackendMemory {
  constructor(store = {}) {
    this.store = store;
  }
  createContextKey(contexts) {
    const key = contexts.join(".");
    if (!key.length || key.includes("..") || key.startsWith(".") || key.endsWith("."))
      throw new StorageError("Context must not be an empty string.");
    return key;
  }
  async initialize() {
  }
  async close() {
  }
  clear() {
    this.store = {};
  }
  get(contexts, key) {
    if (!contexts.length || !key.length)
      throw new StorageError("Context and key must not be empty.");
    return this.store[this.createContextKey(contexts)]?.[key];
  }
  set(contexts, key, value) {
    if (!contexts.length || !key.length)
      throw new StorageError("Context and key must not be empty.");
    const contextKey = this.createContextKey(contexts);
    if (this.store[contextKey] === void 0) {
      this.store[contextKey] = {};
    }
    this.store[contextKey][key] = value;
  }
  delete(contexts, key) {
    if (!contexts.length || !key.length)
      throw new StorageError("Context and key must not be empty.");
    delete this.store[this.createContextKey(contexts)]?.[key];
  }
  keys(contexts) {
    if (!contexts.length)
      throw new StorageError("Context must not be empty!");
    return Object.keys(this.store[this.createContextKey(contexts)] ?? {});
  }
  clearAll(contexts) {
    const contextKey = this.createContextKey(contexts);
    if (contextKey.length) {
      delete this.store[contextKey];
    }
    const startContextKey = contextKey.length ? `${contextKey}.` : "";
    Object.keys(this.store).forEach((key) => {
      if (key.startsWith(startContextKey)) {
        delete this.store[key];
      }
    });
  }
}
export {
  StorageBackendMemory
};
