// @project-chip/matter.js/dist/esm/storage/StorageContext.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { StorageError } from "./Storage.js";
class StorageContext {
  constructor(storage, contexts) {
    this.storage = storage;
    this.contexts = contexts;
  }
  get(key, defaultValue) {
    const value = this.storage.get(this.contexts, key);
    if (value !== void 0)
      return value;
    if (defaultValue === void 0) {
      throw new StorageError(
        `No value found for key ${key} in context ${this.contexts} and no default value specified!`
      );
    }
    return defaultValue;
  }
  has(key) {
    return this.storage.get(this.contexts, key) !== void 0;
  }
  set(key, value) {
    this.storage.set(this.contexts, key, value);
  }
  delete(key) {
    this.storage.delete(this.contexts, key);
  }
  createContext(context) {
    if (context.length === 0)
      throw new StorageError("Context must not be an empty string");
    if (context.includes("."))
      throw new StorageError("Context must not contain dots!");
    return new StorageContext(this.storage, [...this.contexts, context]);
  }
  keys() {
    return this.storage.keys(this.contexts);
  }
  /** Clears all keys in this context */
  clear() {
    for (const key of this.keys()) {
      this.delete(key);
    }
  }
  /** Clears all keys in this context and all created sub-contexts. */
  clearAll() {
    this.storage.clearAll(this.contexts);
  }
}
export {
  StorageContext
};
