// @project-chip/matter.js/dist/esm/log/Logger.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { NotImplementedError } from "../common/MatterError.js";
import { Time } from "../time/Time.js";
import { ByteArray } from "../util/ByteArray.js";
var Level = /* @__PURE__ */ ((Level2) => {
  Level2[Level2["DEBUG"] = 0] = "DEBUG";
  Level2[Level2["INFO"] = 1] = "INFO";
  Level2[Level2["WARN"] = 2] = "WARN";
  Level2[Level2["ERROR"] = 3] = "ERROR";
  Level2[Level2["FATAL"] = 4] = "FATAL";
  return Level2;
})(Level || {});
var Format = /* @__PURE__ */ ((Format2) => {
  Format2["PLAIN"] = "plain";
  Format2["ANSI"] = "ansi";
  Format2["HTML"] = "html";
  return Format2;
})(Format || {});
function formatValue(value, keyFormatter, valueFormatter) {
  if (value instanceof ByteArray) {
    return valueFormatter(value.toHex());
  }
  if (value instanceof Error) {
    let stack;
    if (value.stack === void 0) {
      stack = "(details unavailable)";
    } else {
      stack = value.stack;
      stack = stack.replace(/^.*?\n\nError: /gs, "Error: ");
      if (stack.startsWith("Error: ")) {
        return stack.slice(7);
      }
    }
    return valueFormatter(stack);
  }
  if (value instanceof DiagnosticDictionary) {
    return value.serialize(keyFormatter, valueFormatter);
  }
  if (value === void 0) {
    return "undefined";
  }
  if (value === null) {
    return "null";
  }
  return valueFormatter(value.toString());
}
function formatValues(values, keyFormatter = defaultKeyFormatter, valueFormatter = defaultValueFormatter) {
  return values.map((value) => formatValue(value, keyFormatter, valueFormatter)).join(" ");
}
function formatTime(time) {
  return `${time.getFullYear()}-${(time.getMonth() + 1).toString().padStart(2, "0")}-${time.getDate().toString().padStart(2, "0")} ${time.getHours().toString().padStart(2, "0")}:${time.getMinutes().toString().padStart(2, "0")}:${time.getSeconds().toString().padStart(2, "0")}.${time.getMilliseconds().toString().padStart(3, "0")}`;
}
function nestingPrefix() {
  if (Logger.nestingLevel) {
    return "\u23B8".padEnd(Logger.nestingLevel * 2);
  }
  return "";
}
function plainLogFormatter(now, level, facility, values) {
  const formattedValues = formatValues(values);
  return `${formatTime(now)} ${Level[level]} ${facility} ${nestingPrefix()}${formattedValues}`;
}
const ANSI_CODES = {
  PREFIX: "\x1B[90m\x1B[2m",
  // dark gray, dim
  FACILITY: "\x1B[90m\x1B[1m",
  // dark grey, bold
  LEVEL_DEBUG: "\x1B[90m",
  // dark gray
  LEVEL_INFO: "\x1B[32m",
  // green
  LEVEL_WARN: "\x1B[33m",
  // yellow
  LEVEL_ERROR: "\x1B[31m",
  // red
  LEVEL_FATAL: "\x1B[31m\x1B[1m",
  // red, bold
  KEY: "\x1B[34m",
  // blue
  NONE: "\x1B[0m"
  // reset
};
function ansiLogFormatter(now, level, facility, values) {
  const levelCode = ANSI_CODES[`LEVEL_${Level[level]}`];
  const formattedPrefix = `${ANSI_CODES.PREFIX}${formatTime(now)} ${Level[level].padEnd(5)}${ANSI_CODES.NONE}`;
  let formattedFacility = facility.length > 20 ? `${facility.slice(0, 10)}~${facility.slice(facility.length - 9)}` : facility.padEnd(20);
  formattedFacility = `${ANSI_CODES.FACILITY}${formattedFacility}${ANSI_CODES.NONE}`;
  let formattedValues = formatValues(
    values,
    (key) => `${ANSI_CODES.KEY}${key}:${levelCode} `,
    (value) => value.replace(/([✓✔])/g, `${ANSI_CODES.LEVEL_INFO}$1${levelCode}`).replace(/([✗✘])/g, `${ANSI_CODES.LEVEL_ERROR}$1${levelCode}`)
  );
  formattedValues = `${nestingPrefix()}${formattedValues}`;
  const prefixedMatch = formattedValues.match(/^(⎸? *\d+ )(.*)$/);
  if (prefixedMatch) {
    formattedValues = `${ANSI_CODES.PREFIX}${prefixedMatch[1]}${ANSI_CODES.NONE}${levelCode}${prefixedMatch[2]}${ANSI_CODES.NONE}`;
  } else {
    formattedValues = `${levelCode}${formattedValues}${ANSI_CODES.NONE}`;
  }
  return `${formattedPrefix} ${formattedFacility} ${formattedValues}`;
}
function htmlSpan(type, value) {
  return `<span class="matter-log-${type}">${value}</span>`;
}
function htmlEscape(value) {
  return value.toString().replace(/\n/g, "<br/>").replace(/</g, "&amp").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}
function htmlLogFormatter(now, level, facility, values) {
  const formattedValues = formatValues(
    values,
    (key) => htmlSpan("key", htmlEscape(`${key}:`)) + " ",
    (value) => htmlSpan("value", htmlEscape(value))
  );
  const np = nestingPrefix().replace(/ /g, "&nbsp;");
  return htmlSpan(
    "matter-log-line",
    `${htmlSpan("time", formatTime(now))} ${htmlSpan("level", Level[level])} ${htmlSpan(
      "facility",
      facility
    )} ${np}${formattedValues}`
  );
}
function consoleLogger(level, formattedLog) {
  const console2 = consoleLogger.console;
  switch (level) {
    case 0 /* DEBUG */:
      console2.debug(formattedLog);
      break;
    case 1 /* INFO */:
      console2.info(formattedLog);
      break;
    case 2 /* WARN */:
      console2.warn(formattedLog);
      break;
    case 3 /* ERROR */:
      console2.error(formattedLog);
      break;
    case 4 /* FATAL */:
      console2.error(formattedLog);
      break;
  }
}
const globalConsole = console;
((consoleLogger2) => {
  consoleLogger2.console = globalConsole;
})(consoleLogger || (consoleLogger = {}));
const defaultKeyFormatter = (key) => `${key}: `;
const defaultValueFormatter = (value) => value;
class DiagnosticDictionary {
  /**
   * Create a new dictionary with optional entry values.
   *
   * @param entries the entries as [ "KEY", value ] tuples
   */
  constructor(entries = {}) {
    this.entries = entries;
  }
  /**
   * Format the dictionary for human consumption.
   *
   * @param keyFormatter formats keys
   * @param valueFormatter formats values
   * @returns the formatted value
   */
  serialize(keyFormatter = defaultKeyFormatter, valueFormatter = defaultValueFormatter) {
    return Object.getOwnPropertyNames(this.entries).map(
      (k) => this.entries[k] === void 0 ? void 0 : `${keyFormatter(k)}${formatValue(this.entries[k], keyFormatter, valueFormatter)}`
    ).filter((v) => v !== void 0).join(" ");
  }
  toString() {
    return this.serialize();
  }
}
function determineLoggerForFormat(format) {
  switch (format) {
    case "plain" /* PLAIN */:
      return plainLogFormatter;
    case "ansi" /* ANSI */:
      return ansiLogFormatter;
    case "html" /* HTML */:
      return htmlLogFormatter;
    default:
      throw new NotImplementedError(`Unsupported log format "${format}"`);
  }
}
class Logger {
  constructor(name) {
    this.name = name;
    this.debug = (...values) => this.log(0 /* DEBUG */, values);
    this.info = (...values) => this.log(1 /* INFO */, values);
    this.warn = (...values) => this.log(2 /* WARN */, values);
    this.error = (...values) => this.log(3 /* ERROR */, values);
    this.fatal = (...values) => this.log(4 /* FATAL */, values);
  }
  static {
    this.logger = new Array({
      logIdentifier: "default",
      logFormatter: plainLogFormatter,
      log: consoleLogger,
      defaultLogLevel: 0 /* DEBUG */,
      logLevels: {}
    });
  }
  static {
    this.nestingLevel = 0;
  }
  /** Add additional logger to the list of loggers including the default configuration. */
  static addLogger(identifier, logger, options) {
    if (Logger.logger.some((logger2) => logger2.logIdentifier === identifier)) {
      throw new NotImplementedError(`Logger "${identifier}" already exists`);
    }
    Logger.logger.push({
      logIdentifier: identifier,
      logFormatter: determineLoggerForFormat(options?.logFormat ?? "plain" /* PLAIN */),
      log: logger,
      defaultLogLevel: options?.defaultLogLevel ?? 0 /* DEBUG */,
      logLevels: options?.logLevels ?? {}
    });
  }
  static removeLogger(identifier) {
    const index = Logger.logger.findIndex((logger) => logger.logIdentifier === identifier);
    if (index === -1) {
      throw new NotImplementedError(`Logger "${identifier}" does not exist`);
    }
    Logger.logger.splice(index, 1);
  }
  /**
   * Get the logger with the matching identifier.
   * @param identifier The identifier of the logger
   */
  static getLoggerforIdentifier(identifier) {
    const logger = Logger.logger.find((logger2) => logger2.logIdentifier === identifier);
    if (logger === void 0) {
      throw new NotImplementedError(`Unknown logger "${identifier}"`);
    }
    return logger;
  }
  /**
   * Set logFormatter using configuration-style format name for the default logger.
   *
   * @param format the name of the formatter (see Format enum)
   */
  static set format(format) {
    Logger.setFormatForLogger("default", format);
  }
  /**
   * Set facility loglevels for the default logger.
   * @param levels The levels to set
   */
  static set logLevels(levels) {
    Logger.setLogLevelsForLogger("default", levels);
  }
  /**
   * Get facility loglevels for the default logger.
   */
  static get logLevels() {
    return Logger.getLoggerforIdentifier("default").logLevels;
  }
  /**
   * Set default loglevel for the default logger.
   *
   * @param level The level to set
   */
  static set defaultLogLevel(level) {
    Logger.setDefaultLoglevelForLogger("default", level);
  }
  /**
   * Get default loglevel for the default logger.
   */
  static get defaultLogLevel() {
    return Logger.getLoggerforIdentifier("default").defaultLogLevel;
  }
  /**
   * Set the log function for the default logger.
   *
   * @param log The log function to set
   */
  static set log(log) {
    Logger.setLogger("default", log);
  }
  /**
   * Get the log function for the default logger.
   */
  static get log() {
    return Logger.getLoggerforIdentifier("default").log;
  }
  /**
   * Set the log formatter for the default logger.
   *
   * @param logFormatter
   */
  static set logFormatter(logFormatter) {
    Logger.setLogFormatterForLogger("default", logFormatter);
  }
  /**
   * Get the log formatter for the default logger.
   */
  static get logFormatter() {
    return Logger.getLoggerforIdentifier("default").logFormatter;
  }
  /**
   * Set logFormatter using configuration-style format name for the logger with the matching identifier.
   *
   * @param identifier The identifier of the logger
   * @param format the name of the formatter (see Format enum)
   */
  static setFormatForLogger(identifier, format) {
    const logger = Logger.logger.find((logger2) => logger2.logIdentifier === identifier);
    if (logger) {
      logger.logFormatter = determineLoggerForFormat(format);
    } else {
      throw new NotImplementedError(`Unknown logger "${identifier}"`);
    }
  }
  /**
   * Set default loglevel for the logger with the matching identifier.
   *
   * @param identifier The identifier of the logger
   * @param level The level to set
   */
  static setDefaultLoglevelForLogger(identifier, level) {
    const logger = Logger.logger.find((logger2) => logger2.logIdentifier === identifier);
    if (logger) {
      logger.defaultLogLevel = level;
    } else {
      throw new NotImplementedError(`Unknown logger "${identifier}"`);
    }
  }
  /**
   * Set facility loglevels for the logger with the matching identifier.
   *
   * @param identifier The identifier of the logger
   * @param levels The levels to set
   */
  static setLogLevelsForLogger(identifier, levels) {
    const logger = Logger.logger.find((logger2) => logger2.logIdentifier === identifier);
    if (logger) {
      logger.logLevels = levels;
    } else {
      throw new NotImplementedError(`Unknown logger "${identifier}"`);
    }
  }
  /**
   * Set the log function for the logger with the matching identifier.
   *
   * @param identifier The identifier of the logger
   * @param log The log function to set
   */
  static setLogger(identifier, log) {
    const logger = Logger.logger.find((logger2) => logger2.logIdentifier === identifier);
    if (logger) {
      logger.log = log;
    } else {
      throw new NotImplementedError(`Unknown logger "${identifier}"`);
    }
  }
  /**
   * Set the log formatter for the logger with the matching identifier.
   *
   * @param identifier The identifier of the logger
   * @param logFormatter The log formatter to set
   */
  static setLogFormatterForLogger(identifier, logFormatter) {
    const logger = Logger.logger.find((logger2) => logger2.logIdentifier === identifier);
    if (logger) {
      logger.logFormatter = logFormatter;
    } else {
      throw new NotImplementedError(`Unknown logger "${identifier}"`);
    }
  }
  /**
   * Create a new facility.
   *
   * @param name the name of the facility
   * @returns a new facility
   */
  static get(name) {
    return new Logger(name);
  }
  /**
   * Stringify a value (BigInt aware) as JSON.
   *
   * @param data the value to stringify
   * @returns the stringified value
   */
  static toJSON(data) {
    return JSON.stringify(data, (_, value) => {
      if (typeof value === "bigint") {
        return value.toString();
      }
      if (value instanceof ByteArray) {
        return value.toHex();
      }
      return value;
    });
  }
  /**
   * Mask a string with a given character. If unmaskedLength is provided then these number of characters will be
   * shown unmasked.
   *
   * @param str String to mask
   * @param maskChar character to mask with
   * @param unmaskedLength number of characters to show unmasked in the beginning
   */
  static maskString(str, maskChar = "*", unmaskedLength) {
    return str.substring(0, unmaskedLength ?? 0) + str.substring(unmaskedLength ?? 0).replace(/./g, maskChar);
  }
  /**
   * Shortcut for new DiagnosticDictionary().
   *
   * @param entries initial dictionary entries
   */
  static dict(entries) {
    return new DiagnosticDictionary(entries);
  }
  /**
   * Perform operations in a nested logging context.  Messages will be
   * indented while the context executes.
   */
  static nest(context) {
    this.nestingLevel++;
    try {
      return context();
    } finally {
      this.nestingLevel--;
    }
  }
  /**
   * Async version of nest().
   */
  static async nestAsync(context) {
    this.nestingLevel++;
    try {
      return await context();
    } finally {
      this.nestingLevel--;
    }
  }
  log(level, values) {
    Logger.logger.forEach((logger) => {
      if (level < (logger.logLevels[this.name] ?? logger.defaultLogLevel))
        return;
      logger.log(level, logger.logFormatter(Time.now(), level, this.name, values));
    });
  }
}
if (typeof MatterHooks !== "undefined") {
  MatterHooks.loggerSetup?.(Logger);
}
export {
  DiagnosticDictionary,
  Format,
  Level,
  Logger,
  consoleLogger
};
