// @project-chip/matter.js/dist/esm/protocol/interaction/EventHandler.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { EventPriority } from "../../cluster/Cluster.js";
import { resolveEventName } from "../../cluster/ClusterHelper.js";
import { Logger } from "../../log/Logger.js";
const logger = Logger.get("EventHandler");
const MAX_EVENTS = 1e4;
class EventHandler {
  constructor(storage) {
    this.eventNumber = 0;
    this.storedEventCount = 0;
    this.events = {
      [EventPriority.Critical]: new Array(),
      [EventPriority.Info]: new Array(),
      [EventPriority.Debug]: new Array()
    };
    this.eventStorage = storage.createContext("EventHandler");
    this.eventNumber = this.eventStorage.get("lastEventNumber", this.eventNumber);
    logger.debug(`Set/Restore last event number: ${this.eventNumber}`);
  }
  getEvents(eventPath, filters) {
    const eventFilter = filters !== void 0 && filters.length > 0 ? (event) => filters.some((filter) => filter.eventMin !== void 0 && event.eventNumber >= filter.eventMin) : () => true;
    const events = new Array();
    const { endpointId, clusterId, eventId } = eventPath;
    for (const priority of [EventPriority.Critical, EventPriority.Info, EventPriority.Debug]) {
      const eventsToCheck = this.events[priority];
      for (const event of eventsToCheck) {
        if (eventFilter(event) && endpointId === event.endpointId && clusterId === event.clusterId && eventId === event.eventId) {
          events.push(event);
        }
      }
    }
    logger.debug(
      `Got ${events.length} events for ${resolveEventName(eventPath)} with filters: ${Logger.toJSON(filters)}`
    );
    return events;
  }
  pushEvent(event) {
    const eventData = {
      eventNumber: ++this.eventNumber,
      ...event
    };
    logger.debug(`Received event: ${JSON.stringify(eventData)}`);
    this.events[event.priority].push(eventData);
    this.storedEventCount++;
    this.eventStorage.set("lastEventNumber", this.eventNumber);
    this.cleanUpEvents();
    return eventData;
  }
  cleanUpEvents() {
    if (this.storedEventCount < MAX_EVENTS)
      return;
    const eventsToDelete = this.storedEventCount - MAX_EVENTS;
    for (const priority of [EventPriority.Debug, EventPriority.Info, EventPriority.Critical]) {
      const events = this.events[priority];
      if (events.length > 0) {
        const removedEvents = events.splice(0, events.length - eventsToDelete);
        logger.debug(`Removed ${removedEvents.length} events from priority ${priority}`);
        return;
      }
    }
  }
}
export {
  EventHandler
};
