// @project-chip/matter.js/dist/esm/protocol/interaction/InteractionMessenger.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SessionType } from "../../codec/MessageCodec.js";
import {
  ImplementationError,
  MatterFlowError,
  NotImplementedError,
  UnexpectedDataError
} from "../../common/MatterError.js";
import { tryCatchAsync } from "../../common/TryCatchHandler.js";
import { Logger } from "../../log/Logger.js";
import {
  RetransmissionLimitReachedError,
  UnexpectedMessageError
} from "../../protocol/MessageExchange.js";
import { TlvAny } from "../../tlv/TlvAny.js";
import {
  canAttributePayloadBeChunked,
  chunkAttributePayload,
  encodeAttributePayload,
  encodeEventPayload
} from "./AttributeDataEncoder.js";
import {
  TlvDataReport,
  TlvDataReportForSend,
  TlvInvokeRequest,
  TlvInvokeResponse,
  TlvReadRequest,
  TlvStatusResponse,
  TlvSubscribeRequest,
  TlvSubscribeResponse,
  TlvTimedRequest,
  TlvWriteRequest,
  TlvWriteResponse
} from "./InteractionProtocol.js";
import { INTERACTION_MODEL_REVISION } from "./InteractionServer.js";
import { StatusCode, StatusResponseError } from "./StatusCode.js";
var MessageType = /* @__PURE__ */ ((MessageType2) => {
  MessageType2[MessageType2["StatusResponse"] = 1] = "StatusResponse";
  MessageType2[MessageType2["ReadRequest"] = 2] = "ReadRequest";
  MessageType2[MessageType2["SubscribeRequest"] = 3] = "SubscribeRequest";
  MessageType2[MessageType2["SubscribeResponse"] = 4] = "SubscribeResponse";
  MessageType2[MessageType2["ReportData"] = 5] = "ReportData";
  MessageType2[MessageType2["WriteRequest"] = 6] = "WriteRequest";
  MessageType2[MessageType2["WriteResponse"] = 7] = "WriteResponse";
  MessageType2[MessageType2["InvokeCommandRequest"] = 8] = "InvokeCommandRequest";
  MessageType2[MessageType2["InvokeCommandResponse"] = 9] = "InvokeCommandResponse";
  MessageType2[MessageType2["TimedRequest"] = 10] = "TimedRequest";
  return MessageType2;
})(MessageType || {});
const MAX_SPDU_LENGTH = 1024;
const logger = Logger.get("InteractionMessenger");
class InteractionMessenger {
  constructor(exchange) {
    this.exchange = exchange;
  }
  async send(messageType, payload, options) {
    return this.exchange.send(messageType, payload, options);
  }
  sendStatus(status) {
    return this.send(
      1 /* StatusResponse */,
      TlvStatusResponse.encode({ status, interactionModelRevision: INTERACTION_MODEL_REVISION })
    );
  }
  async waitForSuccess() {
    await this.nextMessage(1 /* StatusResponse */);
  }
  async nextMessage(expectedMessageType) {
    const message = await this.exchange.nextMessage();
    const messageType = message.payloadHeader.messageType;
    this.throwIfErrorStatusMessage(message);
    if (expectedMessageType !== void 0 && messageType !== expectedMessageType) {
      throw new UnexpectedDataError(
        `Received unexpected message type: ${messageType}, expected: ${expectedMessageType}`
      );
    }
    return message;
  }
  async close() {
    await this.exchange.close();
  }
  throwIfErrorStatusMessage(message) {
    const {
      payloadHeader: { messageType },
      payload
    } = message;
    if (messageType !== 1 /* StatusResponse */)
      return;
    const { status } = TlvStatusResponse.decode(payload);
    if (status !== StatusCode.Success)
      throw new StatusResponseError(`Received error status: ${status}`, status);
  }
  getExchangeChannelName() {
    return this.exchange.channel.name;
  }
}
class InteractionServerMessenger extends InteractionMessenger {
  async handleRequest(handleReadRequest, handleWriteRequest, handleSubscribeRequest, handleInvokeRequest, handleTimedRequest) {
    let continueExchange = true;
    try {
      while (continueExchange) {
        const message = await this.exchange.nextMessage();
        const isGroupSession = message.packetHeader.sessionType === SessionType.Group;
        continueExchange = false;
        switch (message.payloadHeader.messageType) {
          case 2 /* ReadRequest */: {
            const readRequest = TlvReadRequest.decode(message.payload);
            await this.sendDataReport(handleReadRequest(readRequest));
            break;
          }
          case 6 /* WriteRequest */: {
            const writeRequest = TlvWriteRequest.decode(message.payload);
            const { suppressResponse } = writeRequest;
            const writeResponse = handleWriteRequest(writeRequest, message);
            if (!suppressResponse && !isGroupSession) {
              await this.send(7 /* WriteResponse */, TlvWriteResponse.encode(writeResponse));
            }
            break;
          }
          case 3 /* SubscribeRequest */: {
            const subscribeRequest = TlvSubscribeRequest.decode(message.payload);
            await handleSubscribeRequest(subscribeRequest, this);
            break;
          }
          case 8 /* InvokeCommandRequest */: {
            const invokeRequest = TlvInvokeRequest.decode(message.payload);
            const { suppressResponse } = invokeRequest;
            const invokeResponse = await handleInvokeRequest(invokeRequest, message);
            if (!suppressResponse && !isGroupSession) {
              await this.send(
                9 /* InvokeCommandResponse */,
                TlvInvokeResponse.encode(invokeResponse)
              );
            }
            break;
          }
          case 10 /* TimedRequest */: {
            const timedRequest = TlvTimedRequest.decode(message.payload);
            handleTimedRequest(timedRequest);
            await this.sendStatus(StatusCode.Success);
            continueExchange = true;
            break;
          }
          default:
            throw new NotImplementedError(`Unsupported message type ${message.payloadHeader.messageType}`);
        }
      }
    } catch (error) {
      if (error instanceof StatusResponseError) {
        logger.info(`Sending status response ${error.code} for interaction error: ${error.message}`);
        await this.sendStatus(error.code);
      } else {
        logger.error(error);
        await this.sendStatus(StatusCode.Failure);
      }
    } finally {
      await this.exchange.close();
    }
  }
  /**
   * Handle DataReportPayload with the content of a DataReport to send, split them into multiple DataReport
   * messages and send them out based on the size.
   */
  async sendDataReport(dataReportPayload) {
    const {
      subscriptionId,
      attributeReportsPayload,
      eventReportsPayload,
      suppressResponse,
      interactionModelRevision
    } = dataReportPayload;
    const dataReport = {
      subscriptionId,
      suppressResponse,
      interactionModelRevision,
      attributeReports: void 0,
      eventReports: void 0
    };
    if (attributeReportsPayload !== void 0 || eventReportsPayload !== void 0) {
      const attributeReportsToSend = [...attributeReportsPayload ?? []];
      const eventReportsToSend = [...eventReportsPayload ?? []];
      dataReport.moreChunkedMessages = true;
      const emptyDataReportBytes = TlvDataReportForSend.encode(dataReport);
      let firstAttributeAddedToReportMessage = false;
      let firstEventAddedToReportMessage = false;
      const sendAndResetReport = async () => {
        await this.sendDataReportMessage(dataReport);
        dataReport.attributeReports = void 0;
        dataReport.eventReports = void 0;
        messageSize = emptyDataReportBytes.length;
        firstAttributeAddedToReportMessage = false;
        firstEventAddedToReportMessage = false;
      };
      let messageSize = emptyDataReportBytes.length;
      while (true) {
        if (attributeReportsToSend.length > 0) {
          const attributeReport = attributeReportsToSend.shift();
          if (attributeReport !== void 0) {
            if (!firstAttributeAddedToReportMessage) {
              firstAttributeAddedToReportMessage = true;
              messageSize += 3;
            }
            const encodedAttribute = encodeAttributePayload(attributeReport);
            const attributeReportBytes = TlvAny.getEncodedByteLength(encodedAttribute);
            if (messageSize + attributeReportBytes > MAX_SPDU_LENGTH) {
              if (canAttributePayloadBeChunked(attributeReport)) {
                attributeReportsToSend.unshift(...chunkAttributePayload(attributeReport));
                continue;
              }
              await sendAndResetReport();
            }
            messageSize += attributeReportBytes;
            if (dataReport.attributeReports === void 0)
              dataReport.attributeReports = [];
            dataReport.attributeReports.push(encodedAttribute);
          }
        } else if (eventReportsToSend.length > 0) {
          const eventReport = eventReportsToSend.shift();
          if (eventReport === void 0) {
            dataReport.moreChunkedMessages = void 0;
            break;
          }
          if (!firstEventAddedToReportMessage) {
            firstEventAddedToReportMessage = true;
            messageSize += 3;
          }
          const encodedEvent = encodeEventPayload(eventReport);
          const eventReportBytes = TlvAny.getEncodedByteLength(encodedEvent);
          if (messageSize + eventReportBytes > MAX_SPDU_LENGTH) {
            await sendAndResetReport();
          }
          messageSize += eventReportBytes;
          if (dataReport.eventReports === void 0)
            dataReport.eventReports = [];
          dataReport.eventReports.push(encodedEvent);
        } else {
          dataReport.moreChunkedMessages = void 0;
          break;
        }
      }
    }
    await this.sendDataReportMessage(dataReport);
  }
  async sendDataReportMessage(dataReport) {
    const encodedMessage = TlvDataReportForSend.encode(dataReport);
    if (encodedMessage.length > MAX_SPDU_LENGTH) {
      throw new MatterFlowError(
        `DataReport is too long to fit in a single chunk, This should not happen! Data: ${Logger.toJSON(
          dataReport
        )}`
      );
    }
    logger.debug(
      `Sending DataReport chunk with ${dataReport.attributeReports?.length ?? 0} attributes and ${dataReport.eventReports?.length ?? 0} events: ${encodedMessage.length} bytes`
    );
    if (dataReport.suppressResponse) {
      await tryCatchAsync(
        async () => await this.exchange.send(5 /* ReportData */, encodedMessage, {
          expectAckOnly: true
        }),
        UnexpectedMessageError,
        async (error) => {
          const { receivedMessage } = error;
          this.throwIfErrorStatusMessage(receivedMessage);
        }
      );
    } else {
      await this.exchange.send(5 /* ReportData */, encodedMessage);
      await this.waitForSuccess();
    }
  }
}
class IncomingInteractionClientMessenger extends InteractionMessenger {
  async readDataReport() {
    let subscriptionId;
    const attributeValues = [];
    const eventValues = [];
    while (true) {
      const dataReportMessage = await this.exchange.waitFor(5 /* ReportData */);
      const report = TlvDataReport.decode(dataReportMessage.payload);
      if (subscriptionId === void 0 && report.subscriptionId !== void 0) {
        subscriptionId = report.subscriptionId;
      } else if ((subscriptionId !== void 0 || report.subscriptionId !== void 0) && report.subscriptionId !== subscriptionId) {
        throw new UnexpectedDataError(`Invalid subscription ID ${report.subscriptionId} received`);
      }
      if (Array.isArray(report.attributeReports) && report.attributeReports.length > 0) {
        attributeValues.push(...report.attributeReports);
      }
      if (Array.isArray(report.eventReports) && report.eventReports.length > 0) {
        eventValues.push(...report.eventReports);
      }
      if (!report.moreChunkedMessages) {
        report.attributeReports = attributeValues;
        report.eventReports = eventValues;
        return report;
      }
      await this.sendStatus(StatusCode.Success);
    }
  }
}
class InteractionClientMessenger extends IncomingInteractionClientMessenger {
  constructor(exchangeProvider) {
    super(exchangeProvider.initiateExchange());
    this.exchangeProvider = exchangeProvider;
  }
  /** Implements a send method with an automatic reconnection mechanism */
  async send(messageType, payload, options) {
    if (this.exchange.channel.closed) {
      throw new ImplementationError("The exchange channel is closed. Please connect the device first.");
    }
    try {
      return await this.exchange.send(messageType, payload, options);
    } catch (error) {
      if (error instanceof RetransmissionLimitReachedError) {
        await this.exchange.close();
        if (await this.exchangeProvider.reconnectChannel()) {
          this.exchange = this.exchangeProvider.initiateExchange();
          return await this.exchange.send(messageType, payload);
        }
      }
      throw error;
    }
  }
  sendReadRequest(readRequest) {
    return this.request(
      2 /* ReadRequest */,
      TlvReadRequest,
      5 /* ReportData */,
      TlvDataReport,
      readRequest
    );
  }
  async sendSubscribeRequest(subscribeRequest) {
    await this.send(3 /* SubscribeRequest */, TlvSubscribeRequest.encode(subscribeRequest));
    const report = await this.readDataReport();
    const { subscriptionId } = report;
    if (subscriptionId === void 0) {
      throw new UnexpectedDataError(`Subscription ID not provided in report`);
    }
    await this.sendStatus(StatusCode.Success);
    const subscribeResponseMessage = await this.nextMessage(4 /* SubscribeResponse */);
    const subscribeResponse = TlvSubscribeResponse.decode(subscribeResponseMessage.payload);
    if (subscribeResponse.subscriptionId !== subscriptionId) {
      throw new MatterFlowError(
        `Received subscription ID ${subscribeResponse.subscriptionId} instead of ${subscriptionId}`
      );
    }
    return {
      subscribeResponse,
      report
    };
  }
  async sendInvokeCommand(invokeRequest, minimumResponseTimeoutMs) {
    if (invokeRequest.suppressResponse) {
      await this.requestWithSuppressedResponse(
        8 /* InvokeCommandRequest */,
        TlvInvokeRequest,
        invokeRequest,
        minimumResponseTimeoutMs
      );
    } else {
      return await this.request(
        8 /* InvokeCommandRequest */,
        TlvInvokeRequest,
        9 /* InvokeCommandResponse */,
        TlvInvokeResponse,
        invokeRequest,
        minimumResponseTimeoutMs
      );
    }
  }
  async sendWriteCommand(writeRequest) {
    if (writeRequest.suppressResponse) {
      await this.requestWithSuppressedResponse(6 /* WriteRequest */, TlvWriteRequest, writeRequest);
    } else {
      return await this.request(
        6 /* WriteRequest */,
        TlvWriteRequest,
        7 /* WriteResponse */,
        TlvWriteResponse,
        writeRequest
      );
    }
  }
  sendTimedRequest(timeoutSeconds) {
    return this.request(10 /* TimedRequest */, TlvTimedRequest, 1 /* StatusResponse */, TlvStatusResponse, {
      timeout: timeoutSeconds,
      interactionModelRevision: INTERACTION_MODEL_REVISION
    });
  }
  async requestWithSuppressedResponse(requestMessageType, requestSchema, request, minimumResponseTimeoutMs) {
    await this.send(requestMessageType, requestSchema.encode(request), {
      expectAckOnly: true,
      minimumResponseTimeoutMs
    });
  }
  async request(requestMessageType, requestSchema, responseMessageType, responseSchema, request, minimumResponseTimeoutMs) {
    await this.send(requestMessageType, requestSchema.encode(request), {
      expectAckOnly: false,
      minimumResponseTimeoutMs
    });
    const responseMessage = await this.nextMessage(responseMessageType);
    return responseSchema.decode(responseMessage.payload);
  }
}
export {
  IncomingInteractionClientMessenger,
  InteractionClientMessenger,
  InteractionServerMessenger,
  MessageType
};
