// @project-chip/matter.js/dist/esm/protocol/interaction/StatusCode.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterError } from "../../common/MatterError.js";
var StatusCode = /* @__PURE__ */ ((StatusCode2) => {
  StatusCode2[StatusCode2["Success"] = 0] = "Success";
  StatusCode2[StatusCode2["Failure"] = 1] = "Failure";
  StatusCode2[StatusCode2["InvalidSubscription"] = 125] = "InvalidSubscription";
  StatusCode2[StatusCode2["UnsupportedAccess"] = 126] = "UnsupportedAccess";
  StatusCode2[StatusCode2["UnsupportedEndpoint"] = 127] = "UnsupportedEndpoint";
  StatusCode2[StatusCode2["InvalidAction"] = 128] = "InvalidAction";
  StatusCode2[StatusCode2["UnsupportedCommand"] = 129] = "UnsupportedCommand";
  StatusCode2[StatusCode2["InvalidCommand"] = 133] = "InvalidCommand";
  StatusCode2[StatusCode2["UnsupportedAttribute"] = 134] = "UnsupportedAttribute";
  StatusCode2[StatusCode2["ConstraintError"] = 135] = "ConstraintError";
  StatusCode2[StatusCode2["UnsupportedWrite"] = 136] = "UnsupportedWrite";
  StatusCode2[StatusCode2["ResourceExhausted"] = 137] = "ResourceExhausted";
  StatusCode2[StatusCode2["NotFound"] = 139] = "NotFound";
  StatusCode2[StatusCode2["UnreportableAttribute"] = 140] = "UnreportableAttribute";
  StatusCode2[StatusCode2["InvalidDataType"] = 141] = "InvalidDataType";
  StatusCode2[StatusCode2["UnsupportedRead"] = 143] = "UnsupportedRead";
  StatusCode2[StatusCode2["DataVersionMismatch"] = 146] = "DataVersionMismatch";
  StatusCode2[StatusCode2["Timeout"] = 148] = "Timeout";
  StatusCode2[StatusCode2["UnsupportedMode"] = 155] = "UnsupportedMode";
  StatusCode2[StatusCode2["Busy"] = 156] = "Busy";
  StatusCode2[StatusCode2["UnsupportedCluster"] = 195] = "UnsupportedCluster";
  StatusCode2[StatusCode2["NoUpstreamSubscription"] = 197] = "NoUpstreamSubscription";
  StatusCode2[StatusCode2["NeedsTimedInteraction"] = 198] = "NeedsTimedInteraction";
  StatusCode2[StatusCode2["UnsupportedEvent"] = 199] = "UnsupportedEvent";
  StatusCode2[StatusCode2["PathsExhausted"] = 200] = "PathsExhausted";
  StatusCode2[StatusCode2["TimedRequestMismatch"] = 201] = "TimedRequestMismatch";
  StatusCode2[StatusCode2["FailsafeRequired"] = 202] = "FailsafeRequired";
  return StatusCode2;
})(StatusCode || {});
class StatusResponseError extends MatterError {
  constructor(message, code, clusterCode) {
    super();
    this.code = code;
    this.clusterCode = clusterCode;
    this.message = `(${code}) ${message}`;
  }
}
export {
  StatusCode,
  StatusResponseError
};
