// @project-chip/matter.js/dist/esm/protocol/interaction/SubscriptionHandler.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { FabricScopedAttributeServer } from "../../cluster/server/AttributeServer.js";
import { InternalError } from "../../common/MatterError.js";
import { tryCatch, tryCatchAsync } from "../../common/TryCatchHandler.js";
import { Logger } from "../../log/Logger.js";
import { Time } from "../../time/Time.js";
import { RetransmissionLimitReachedError } from "../MessageExchange.js";
import { InteractionServerMessenger } from "./InteractionMessenger.js";
import {
  INTERACTION_MODEL_REVISION,
  INTERACTION_PROTOCOL_ID,
  attributePathToId,
  clusterPathToId,
  eventPathToId
} from "./InteractionServer.js";
import { StatusCode, StatusResponseError } from "./StatusCode.js";
const logger = Logger.get("SubscriptionHandler");
const SUBSCRIPTION_MAX_INTERVAL_PUBLISHER_LIMIT_S = 3 * 60;
const SUBSCRIPTION_MIN_INTERVAL_S = 2;
const SUBSCRIPTION_DEFAULT_RANDOMIZATION_WINDOW_S = 10;
class SubscriptionHandler {
  constructor(subscriptionId, session, endpointStructure, attributeRequests, dataVersionFilters, eventRequests, eventFilters, eventHandler, isFabricFiltered, minIntervalFloor, maxIntervalCeiling, subscriptionMaxIntervalSeconds, subscriptionMinIntervalSeconds, subscriptionRandomizationWindowSeconds, cancelCallback) {
    this.subscriptionId = subscriptionId;
    this.session = session;
    this.endpointStructure = endpointStructure;
    this.attributeRequests = attributeRequests;
    this.dataVersionFilters = dataVersionFilters;
    this.eventRequests = eventRequests;
    this.eventFilters = eventFilters;
    this.eventHandler = eventHandler;
    this.isFabricFiltered = isFabricFiltered;
    this.cancelCallback = cancelCallback;
    this.lastUpdateTimeMs = 0;
    this.outstandingAttributeUpdates = /* @__PURE__ */ new Map();
    this.outstandingEventUpdates = /* @__PURE__ */ new Set();
    this.attributeListeners = /* @__PURE__ */ new Map();
    this.eventListeners = /* @__PURE__ */ new Map();
    this.sendUpdatesActivated = false;
    this.sendingUpdateInProgress = false;
    this.sendNextUpdateImmediately = false;
    this.sendUpdateErrorCounter = 0;
    this.server = this.session.getContext();
    this.fabric = this.session.getAssociatedFabric();
    this.peerNodeId = this.session.getPeerNodeId();
    this.minIntervalFloorMs = minIntervalFloor * 1e3;
    this.maxIntervalCeilingMs = maxIntervalCeiling * 1e3;
    const { maxInterval, sendInterval } = this.determineSendingIntervals(
      (subscriptionMinIntervalSeconds ?? SUBSCRIPTION_MIN_INTERVAL_S) * 1e3,
      (subscriptionMaxIntervalSeconds ?? SUBSCRIPTION_MAX_INTERVAL_PUBLISHER_LIMIT_S) * 1e3,
      (subscriptionRandomizationWindowSeconds ?? SUBSCRIPTION_DEFAULT_RANDOMIZATION_WINDOW_S) * 1e3
    );
    this.maxInterval = maxInterval;
    this.sendInterval = sendInterval;
    this.updateTimer = Time.getTimer(this.sendInterval, () => this.prepareDataUpdate());
    this.sendDelayTimer = Time.getTimer(50, () => this.sendUpdate());
  }
  determineSendingIntervals(subscriptionMinIntervalMs, subscriptionMaxIntervalMs, subscriptionRandomizationWindowMs) {
    const maxInterval = Math.max(
      Math.max(subscriptionMinIntervalMs, SUBSCRIPTION_MIN_INTERVAL_S * 1e3),
      Math.max(this.minIntervalFloorMs, Math.min(subscriptionMaxIntervalMs, this.maxIntervalCeilingMs))
    ) + Math.floor(subscriptionRandomizationWindowMs * Math.random());
    let sendInterval = Math.floor(maxInterval / 2);
    if (sendInterval < 6e4) {
      sendInterval = Math.max(this.minIntervalFloorMs, Math.floor(maxInterval * 0.8));
    }
    if (sendInterval < subscriptionMinIntervalMs) {
      logger.warn(
        `Determined subscription send interval of ${sendInterval}ms is too low. Using maxInterval (${maxInterval}ms) instead.`
      );
      sendInterval = Math.max(subscriptionMinIntervalMs, SUBSCRIPTION_MIN_INTERVAL_S * 1e3);
    }
    return { maxInterval, sendInterval };
  }
  registerNewAttributes() {
    const newAttributes = new Array();
    const attributeErrors = new Array();
    const formerAttributes = new Set(this.attributeListeners.keys());
    if (this.attributeRequests !== void 0) {
      this.attributeRequests.forEach((path) => {
        const attributes = this.endpointStructure.getAttributes([path]);
        if (attributes.length === 0) {
          const { endpointId, clusterId, attributeId } = path;
          if (endpointId === void 0 || clusterId === void 0 || attributeId === void 0) {
            logger.debug(
              `Subscription attribute ${this.endpointStructure.resolveAttributeName(
                path
              )}: ignore non-existing attribute`
            );
          } else {
            attributeErrors.push(
              tryCatch(
                () => {
                  if (this.endpointStructure.validateConcreteAttributePath(
                    endpointId,
                    clusterId,
                    attributeId
                  )) {
                    throw new StatusResponseError(
                      `Attribute ${attributeId} is not writable.`,
                      StatusCode.UnsupportedWrite
                    );
                  }
                  throw new InternalError(
                    "validateConcreteAttributePath check should throw StatusResponseError but did not."
                  );
                },
                StatusResponseError,
                (error) => {
                  logger.debug(
                    `Subscription attribute ${this.endpointStructure.resolveAttributeName(
                      path
                    )}: unsupported path: Status=${error.code}`
                  );
                  return { path, status: { status: error.code } };
                }
              )
            );
          }
          return;
        }
        attributes.forEach(({ path: path2, attribute }) => {
          formerAttributes.delete(attributePathToId(path2));
          const existingAttributeListener = this.attributeListeners.get(attributePathToId(path2));
          if (existingAttributeListener !== void 0) {
            const { attribute: existingAttribute, listener: existingListener } = existingAttributeListener;
            if (existingAttribute !== attribute) {
              if (existingListener !== void 0) {
                existingAttribute.removeValueChangeListener(existingListener);
              }
              this.attributeListeners.delete(attributePathToId(path2));
            } else {
              return;
            }
          }
          if (attribute.isSubscribable) {
            const listener = (value, version) => this.attributeChangeListener(path2, attribute.schema, version, value);
            attribute.addValueChangeListener(listener);
            this.attributeListeners.set(attributePathToId(path2), { attribute, listener });
          } else {
            this.attributeListeners.set(attributePathToId(path2), { attribute });
          }
          newAttributes.push({ path: path2, attribute });
        });
      });
    }
    this.unregisterAttributeListeners(Array.from(formerAttributes.values()));
    return { newAttributes, attributeErrors };
  }
  unregisterAttributeListeners(list) {
    for (const pathId of list) {
      const existingAttributeListener = this.attributeListeners.get(pathId);
      if (existingAttributeListener !== void 0) {
        const { attribute, listener } = existingAttributeListener;
        if (listener !== void 0) {
          attribute.removeValueChangeListener(listener);
        }
        this.attributeListeners.delete(pathId);
      }
    }
  }
  registerNewEvents() {
    const newEvents = new Array();
    const eventErrors = new Array();
    const formerEvents = new Set(this.eventListeners.keys());
    if (this.eventRequests !== void 0) {
      this.eventRequests.forEach((path) => {
        const events = this.endpointStructure.getEvents([path]);
        if (events.length === 0) {
          const { endpointId, clusterId, eventId } = path;
          if (endpointId === void 0 || clusterId === void 0 || eventId === void 0) {
            logger.debug(
              `Subscription event ${this.endpointStructure.resolveEventName(
                path
              )}: ignore non-existing event`
            );
          } else {
            eventErrors.push(
              tryCatch(
                () => {
                  this.endpointStructure.validateConcreteEventPath(endpointId, clusterId, eventId);
                  throw new InternalError(
                    "validateConcreteEventPath should throw StatusResponseError but did not."
                  );
                },
                StatusResponseError,
                (error) => {
                  logger.debug(
                    `Subscription event ${this.endpointStructure.resolveEventName(
                      path
                    )}: unsupported path: Status=${error.code}`
                  );
                  return { path, status: { status: error.code } };
                }
              )
            );
          }
          return;
        }
        events.forEach(({ path: path2, event }) => {
          formerEvents.delete(eventPathToId(path2));
          const existingEventListener = this.eventListeners.get(eventPathToId(path2));
          if (existingEventListener !== void 0) {
            const { event: existingEvent, listener: existingListener } = existingEventListener;
            if (existingEvent !== event) {
              if (existingListener !== void 0) {
                existingEvent.removeListener(existingListener);
              }
              this.eventListeners.delete(eventPathToId(path2));
            } else {
              return;
            }
          }
          const listener = (newEvent) => this.eventChangeListener(path2, event.schema, newEvent);
          event.addListener(listener);
          newEvents.push({ path: path2, event });
          this.eventListeners.set(eventPathToId(path2), { event, listener });
        });
      });
    }
    this.unregisterEventListeners(Array.from(formerEvents.values()));
    return { newEvents, eventErrors };
  }
  unregisterEventListeners(list) {
    for (const pathId of list) {
      const existingEventListener = this.eventListeners.get(pathId);
      if (existingEventListener !== void 0) {
        const { event, listener } = existingEventListener;
        if (listener !== void 0) {
          event.removeListener(listener);
        }
        this.eventListeners.delete(pathId);
      }
    }
  }
  /**
   * Update the session after an endpoint structure change. The method will initialize all missing new attributes and
   * events and will remove listeners no longer needed.
   * Newly added attributes are then treated ad "changed values" and will be sent as subscription data update to the
   * controller. The data of newly added events are not sent automatically.
   */
  updateSubscription() {
    const { newAttributes } = this.registerNewAttributes();
    for (const attributeWitPath of newAttributes) {
      const { path, attribute } = attributeWitPath;
      const { version, value } = attribute.getWithVersion(this.session, true);
      this.outstandingAttributeUpdates.set(attributePathToId(path), {
        path,
        schema: attribute.schema,
        version,
        value
      });
    }
    this.prepareDataUpdate();
  }
  getMaxInterval() {
    return Math.ceil(this.maxInterval / 1e3);
  }
  getSendInterval() {
    return Math.ceil(this.sendInterval / 1e3);
  }
  activateSendingUpdates() {
    if (this.eventFilters !== void 0)
      this.eventFilters.length = 0;
    if (this.dataVersionFilters !== void 0)
      this.dataVersionFilters.length = 0;
    this.sendUpdatesActivated = true;
    if (this.outstandingAttributeUpdates.size > 0 || this.outstandingEventUpdates.size > 0) {
      void this.sendUpdate();
    }
    this.updateTimer = Time.getTimer(this.sendInterval, () => this.prepareDataUpdate()).start();
  }
  /**
   * Check if data should be sent straight away or delayed because the minimum interval is not reached. Delay real
   * sending by 50ms in any case to mke sure to catch all updates.
   */
  prepareDataUpdate() {
    if (this.sendDelayTimer.isRunning) {
      return;
    }
    if (!this.sendUpdatesActivated) {
      return;
    }
    this.updateTimer.stop();
    const now = Time.nowMs();
    const timeSinceLastUpdateMs = now - this.lastUpdateTimeMs;
    if (timeSinceLastUpdateMs < this.minIntervalFloorMs) {
      this.updateTimer = Time.getTimer(
        this.minIntervalFloorMs - timeSinceLastUpdateMs,
        () => this.prepareDataUpdate()
      ).start();
      return;
    }
    this.sendDelayTimer.start();
    this.updateTimer = Time.getTimer(this.sendInterval, () => this.prepareDataUpdate()).start();
  }
  /**
   * Determine all attributes that have changed since the last update and send them tout to the subscriber.
   */
  async sendUpdate() {
    if (this.sendingUpdateInProgress) {
      logger.debug("Sending update already in progress, delaying update ...");
      this.sendNextUpdateImmediately = true;
      return;
    }
    const attributeUpdatesToSend = Array.from(this.outstandingAttributeUpdates.values());
    this.outstandingAttributeUpdates.clear();
    const eventUpdatesToSend = Array.from(this.outstandingEventUpdates.values());
    this.outstandingEventUpdates.clear();
    this.lastUpdateTimeMs = Time.nowMs();
    this.sendingUpdateInProgress = true;
    try {
      await this.sendUpdateMessage(attributeUpdatesToSend, eventUpdatesToSend);
      this.sendUpdateErrorCounter = 0;
    } catch (error) {
      this.sendUpdateErrorCounter++;
      logger.error(
        `Error sending subscription update message (error count=${this.sendUpdateErrorCounter}):`,
        error
      );
      if (this.sendUpdateErrorCounter > 2) {
        logger.error(
          `Sending update failed 3 times in a row, canceling subscription ${this.subscriptionId} and let controller subscribe again.`
        );
        this.sendNextUpdateImmediately = false;
        if (error instanceof RetransmissionLimitReachedError) {
          await this.session.destroy(false);
        } else {
          throw error;
        }
      }
    }
    this.sendingUpdateInProgress = false;
    if (this.sendNextUpdateImmediately) {
      logger.debug("Sending delayed update immediately after last one was sent.");
      this.sendNextUpdateImmediately = false;
      await this.sendUpdate();
    }
  }
  async sendInitialReport(messenger) {
    this.updateTimer.stop();
    const { newAttributes, attributeErrors } = this.registerNewAttributes();
    const dataVersionFilterMap = new Map(
      this.dataVersionFilters?.map(({ path, dataVersion }) => [clusterPathToId(path), dataVersion]) ?? []
    );
    const attributes = newAttributes.flatMap(({ path, attribute }) => {
      const { value, version } = attribute.getWithVersion(this.session, this.isFabricFiltered);
      if (value === void 0)
        return [];
      const { nodeId, endpointId, clusterId } = path;
      const versionFilterValue = endpointId !== void 0 && clusterId !== void 0 ? dataVersionFilterMap.get(clusterPathToId({ nodeId, endpointId, clusterId })) : void 0;
      if (versionFilterValue !== void 0 && versionFilterValue === version)
        return [];
      return [{ path, value, version, schema: attribute.schema }];
    });
    const attributeReportsPayload = attributes.map(
      ({ path, schema, value, version }) => ({
        attributeData: {
          path,
          dataVersion: version,
          payload: value,
          schema
        }
      })
    );
    attributeErrors.forEach((attributeStatus) => attributeReportsPayload.push({ attributeStatus }));
    const { newEvents, eventErrors } = this.registerNewEvents();
    const eventReportsPayload = newEvents.flatMap(({ path, event: { schema } }) => {
      const matchingEvents = this.eventHandler.getEvents(path, this.eventFilters);
      return matchingEvents.map(({ eventNumber, priority, epochTimestamp, data }) => ({
        eventData: {
          path,
          eventNumber,
          priority,
          epochTimestamp,
          payload: data,
          schema
        }
      }));
    }).sort((a, b) => {
      const eventNumberA = a.eventData?.eventNumber ?? 0;
      const eventNumberB = b.eventData?.eventNumber ?? 0;
      if (eventNumberA > eventNumberB) {
        return 1;
      } else if (eventNumberA < eventNumberB) {
        return -1;
      } else {
        return 0;
      }
    });
    if (newAttributes.length === 0 && newEvents.length === 0) {
      throw new StatusResponseError(
        "Subscription failed because no attributes or events are matching the query",
        StatusCode.InvalidAction
      );
    }
    eventErrors.forEach((eventStatus) => eventReportsPayload.push({ eventStatus }));
    logger.debug(
      `Initialize Subscription with ${attributes.length} attributes and ${eventReportsPayload.length} events.`
    );
    this.lastUpdateTimeMs = Time.nowMs();
    await messenger.sendDataReport({
      suppressResponse: false,
      subscriptionId: this.subscriptionId,
      interactionModelRevision: INTERACTION_MODEL_REVISION,
      attributeReportsPayload,
      // TODO Return compressed response once https://github.com/project-chip/connectedhomeip/issues/29359 is solved
      eventReportsPayload
    });
  }
  attributeChangeListener(path, schema, version, value) {
    const attributeListenerData = this.attributeListeners.get(attributePathToId(path));
    if (attributeListenerData === void 0)
      return;
    const { attribute } = attributeListenerData;
    if (attribute instanceof FabricScopedAttributeServer) {
      value = attribute.get(this.session, this.isFabricFiltered);
    }
    this.outstandingAttributeUpdates.set(attributePathToId(path), { path, schema, version, value });
    this.prepareDataUpdate();
  }
  eventChangeListener(path, schema, newEvent) {
    this.outstandingEventUpdates.add({ path, schema, event: newEvent });
    if (path.isUrgent) {
      this.prepareDataUpdate();
    }
  }
  async flush() {
    this.sendDelayTimer.stop();
    logger.debug(
      `Flushing subscription ${this.subscriptionId} with ${this.outstandingAttributeUpdates.size} attributes and ${this.outstandingEventUpdates.size} events`
    );
    if (this.outstandingAttributeUpdates.size > 0 || this.outstandingEventUpdates.size > 0) {
      void this.sendUpdate();
    }
  }
  async cancel(flush = false, cancelledByPeer = false) {
    this.sendUpdatesActivated = false;
    this.updateTimer.stop();
    this.sendDelayTimer.stop();
    this.unregisterAttributeListeners(Array.from(this.attributeListeners.keys()));
    this.unregisterEventListeners(Array.from(this.eventListeners.keys()));
    if (flush) {
      await this.flush();
    }
    this.session.removeSubscription(this.subscriptionId);
    this.cancelCallback();
    if (cancelledByPeer) {
      await this.session.getContext().startAnnouncement();
    }
  }
  async sendUpdateMessage(attributes, events) {
    logger.debug(
      `Sending subscription update message for ID ${this.subscriptionId} with ${attributes.length} attributes and ${events.length} events`
    );
    const exchange = this.server.initiateExchange(this.fabric, this.peerNodeId, INTERACTION_PROTOCOL_ID);
    if (exchange === void 0)
      return;
    logger.debug(
      `Sending subscription changes for ID ${this.subscriptionId}: ${attributes.map(
        ({ path, value, version }) => `${this.endpointStructure.resolveAttributeName(path)}=${Logger.toJSON(value)} (${version})`
      ).join(", ")}`
    );
    const messenger = new InteractionServerMessenger(exchange);
    try {
      await tryCatchAsync(
        async () => {
          if (attributes.length === 0 && events.length === 0) {
            await messenger.sendDataReport({
              suppressResponse: true,
              // suppressResponse ok for empty DataReports
              subscriptionId: this.subscriptionId,
              interactionModelRevision: INTERACTION_MODEL_REVISION
            });
          } else {
            await messenger.sendDataReport({
              suppressResponse: false,
              subscriptionId: this.subscriptionId,
              interactionModelRevision: INTERACTION_MODEL_REVISION,
              // TODO Return compressed response once https://github.com/project-chip/connectedhomeip/issues/29359 is solved
              attributeReportsPayload: attributes.map(({ path, schema, value, version }) => ({
                attributeData: {
                  path,
                  dataVersion: version,
                  schema,
                  payload: value
                }
              })),
              eventReportsPayload: events.map(
                ({ path, schema, event: { eventNumber, priority, epochTimestamp, data } }) => ({
                  eventData: {
                    path,
                    eventNumber,
                    priority,
                    epochTimestamp,
                    schema,
                    payload: data
                  }
                })
              )
            });
          }
        },
        StatusResponseError,
        async (error) => {
          if (error.code === StatusCode.InvalidSubscription || error.code === StatusCode.Failure) {
            logger.info(`Subscription ${this.subscriptionId} cancelled by peer.`);
            await this.cancel(false, true);
          } else {
            await this.cancel(false);
            throw error;
          }
        }
      );
    } finally {
      await messenger.close();
    }
  }
}
export {
  SubscriptionHandler
};
