// @project-chip/matter.js/dist/esm/util/Cache.js
/**
 * Cache computed values or resources for a specified duration to improve performances.
 *
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Time } from "../time/Time.js";
class Cache {
  constructor(generator, expirationMs, expireCallback) {
    this.generator = generator;
    this.expirationMs = expirationMs;
    this.expireCallback = expireCallback;
    this.knownKeys = /* @__PURE__ */ new Set();
    this.values = /* @__PURE__ */ new Map();
    this.timestamps = /* @__PURE__ */ new Map();
    this.periodicTimer = Time.getPeriodicTimer(expirationMs, () => this.expire()).start();
  }
  get(...params) {
    const key = params.join(",");
    let value = this.values.get(key);
    if (value === void 0) {
      value = this.generator(...params);
      this.values.set(key, value);
      this.knownKeys.add(key);
    }
    this.timestamps.set(key, Time.nowMs());
    return value;
  }
  keys() {
    return Array.from(this.knownKeys.values());
  }
  async deleteEntry(key) {
    const value = this.values.get(key);
    if (this.expireCallback !== void 0 && value !== void 0) {
      await this.expireCallback(key, value);
    }
    this.values.delete(key);
    this.timestamps.delete(key);
  }
  async clear() {
    for (const key of this.values.keys()) {
      await this.deleteEntry(key);
    }
    this.values.clear();
    this.timestamps.clear();
  }
  async close() {
    await this.clear();
    this.knownKeys.clear();
    this.periodicTimer.stop();
  }
  async expire() {
    const now = Time.nowMs();
    for (const [key, timestamp] of this.timestamps.entries()) {
      if (now - timestamp < this.expirationMs)
        continue;
      await this.deleteEntry(key);
    }
  }
}
export {
  Cache
};
