// @project-chip/matter.js/dist/cjs/cluster/server/ClusterServer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClusterServer_exports = {};
__export(ClusterServer_exports, {
  ClusterServer: () => ClusterServer
});
module.exports = __toCommonJS(ClusterServer_exports);
var import_MatterError = require("../../common/MatterError.js");
var import_Crypto = require("../../crypto/Crypto.js");
var import_AttributeId = require("../../datatype/AttributeId.js");
var import_Logger = require("../../log/Logger.js");
var import_String = require("../../util/String.js");
var import_Cluster = require("../Cluster.js");
var import_AttributeServer = require("./AttributeServer.js");
var import_CommandServer = require("./CommandServer.js");
var import_EventServer = require("./EventServer.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("ClusterServer");
function isConditionMatching(featureSets, supportedFeatures) {
  for (const features of featureSets) {
    if (Object.keys(features).every((feature) => !!features[feature] === !!supportedFeatures[feature])) {
      return true;
    }
  }
  return false;
}
function ClusterServer(clusterDef, attributesInitialValues, handlers, supportedEvents = {}) {
  const {
    id: clusterId,
    name,
    commands: commandDef,
    attributes: attributeDef,
    events: eventDef,
    supportedFeatures
  } = clusterDef;
  let clusterStorage = null;
  const attributeStorageListeners = /* @__PURE__ */ new Map();
  const sceneAttributeList = new Array();
  const attributes = {};
  const commands = {};
  const events = {};
  let assignedEndpoint = void 0;
  const result = {
    id: clusterId,
    name,
    _type: "ClusterServer",
    clusterDataVersion: import_Crypto.Crypto.getRandomUInt32(),
    attributes,
    _commands: commands,
    _events: events,
    _assignToEndpoint: (endpoint) => {
      for (const name2 in attributes) {
        attributes[name2].assignToEndpoint(endpoint);
      }
      for (const name2 in events) {
        events[name2].assignToEndpoint(endpoint);
      }
      assignedEndpoint = endpoint;
    },
    _destroy: () => {
      if (typeof handlers.destroyClusterServer === "function") {
        handlers.destroyClusterServer();
      }
    },
    _setStorage: (storageContext) => {
      clusterStorage = storageContext;
      result.clusterDataVersion = storageContext.get("_clusterDataVersion", result.clusterDataVersion);
      logger.debug(
        `${storageContext.has("_clusterDataVersion") ? "Restore" : "Set"} cluster data version ${result.clusterDataVersion} in cluster ${name} (${clusterId})`
      );
      storageContext.set("_clusterDataVersion", result.clusterDataVersion);
      for (const attributeName in attributes) {
        const attribute = attributes[attributeName];
        if (!attributeStorageListeners.has(attribute.id))
          continue;
        if (!storageContext.has(attribute.name))
          continue;
        try {
          const value = storageContext.get(attribute.name);
          logger.debug(
            `Restoring attribute ${attributeName} (${attribute.id}) in cluster ${name} (${clusterId})`
          );
          attribute.init(value);
        } catch (error) {
          logger.warn(
            `Failed to restore attribute ${attributeName} (${attribute.id}) in cluster ${name} (${clusterId})`,
            error
          );
          storageContext.delete(attribute.name);
        }
      }
      if (assignedEndpoint === void 0) {
        throw new import_MatterError.InternalError(
          "The Endpoint always needs to be existing before storage is initialized for an Endpoint."
        );
      }
      if (typeof handlers.initializeClusterServer === "function") {
        handlers.initializeClusterServer({
          attributes,
          events,
          endpoint: assignedEndpoint
        });
      }
    },
    _registerEventHandler: (eventHandler) => {
      for (const eventName in events) {
        events[eventName].bindToEventHandler(eventHandler);
      }
    },
    _getSceneExtensionFieldSets: () => {
      const values = new Array();
      for (const name2 of sceneAttributeList) {
        const attributeServer = attributes[name2];
        values.push({
          attributeId: attributeServer.id,
          attributeValue: attributeServer.schema.encodeTlv(attributeServer.get())
        });
      }
      return values;
    },
    _setSceneExtensionFieldSets: (values, _transitionTime) => {
      for (const { attributeId, attributeValue } of values) {
        if (attributeId === void 0) {
          logger.warn(`Empty attributeId in scene extension field not supported for "set" yet`);
          continue;
        }
        const attributeName = sceneAttributeList.find((name2) => attributes[name2].id === attributeId);
        if (attributeName) {
          const attributeServer = attributes[attributeName];
          attributeServer.setLocal(attributeServer.schema.decodeTlv(attributeValue));
        }
      }
    },
    _verifySceneExtensionFieldSets(values) {
      for (const { attributeId, attributeValue } of values) {
        if (attributeId === void 0) {
          logger.warn(`Empty attributeId in scene extension field not supported for "verify" yet`);
          continue;
        }
        const attributeName = sceneAttributeList.find((name2) => attributes[name2].id === attributeId);
        if (attributeName) {
          const attributeServer = attributes[attributeName];
          if (attributeServer.getLocal() !== attributeServer.schema.decodeTlv(attributeValue))
            return false;
        }
      }
      return true;
    },
    isAttributeSupported: (attributeId) => {
      return attributes.attributeList.getLocal().includes(attributeId);
    },
    isAttributeSupportedByName: (attributeName) => {
      return attributesInitialValues[attributeName] !== void 0;
    },
    isEventSupported: (eventId) => {
      return attributes.eventList.getLocal().includes(eventId);
    },
    isEventSupportedByName: (eventName) => {
      return supportedEvents[eventName] === true;
    },
    isCommandSupported: (commandId) => {
      return attributes.acceptedCommandList.getLocal().includes(commandId);
    },
    isCommandSupportedByName: (commandName) => {
      return commands[commandName] !== void 0;
    }
  };
  const attributeStorageListener = (attributeName, value) => {
    if (!clusterStorage || value === void 0)
      return;
    logger.debug(`Storing attribute ${attributeName} in cluster ${name} (${clusterId})`);
    clusterStorage.set(attributeName, value);
  };
  attributesInitialValues = {
    ...attributesInitialValues,
    clusterRevision: clusterDef.revision,
    featureMap: supportedFeatures,
    attributeList: new Array(),
    acceptedCommandList: new Array(),
    generatedCommandList: new Array(),
    eventList: new Array()
  };
  const attributeList = new Array();
  for (const attributeName in attributeDef) {
    const capitalizedAttributeName = (0, import_String.capitalize)(attributeName);
    if (attributeDef[attributeName].isConditional) {
      const { mandatoryIf, optionalIf } = attributeDef[attributeName];
      let conditionHasMatched = false;
      if (mandatoryIf !== void 0 && mandatoryIf.length > 0) {
        const conditionMatched = isConditionMatching(mandatoryIf, supportedFeatures);
        if (conditionMatched && attributesInitialValues[attributeName] === void 0) {
          logger.warn(
            `InitialAttributeValue for "${clusterDef.name}/${attributeName}" is REQUIRED by supportedFeatures: ${JSON.stringify(
              supportedFeatures
            )} but is not set!`
          );
        }
        conditionHasMatched = conditionHasMatched || conditionMatched;
      }
      if (!conditionHasMatched && optionalIf !== void 0 && optionalIf.length > 0) {
        const conditionMatched = isConditionMatching(optionalIf, supportedFeatures);
        if (conditionMatched && attributesInitialValues[attributeName] === void 0) {
          logger.debug(
            `InitialAttributeValue for "${clusterDef.name}/${attributeName}" is optional by supportedFeatures: ${JSON.stringify(
              supportedFeatures
            )} and is not set!`
          );
        }
        conditionHasMatched = conditionHasMatched || conditionMatched;
      }
      if (!conditionHasMatched && attributesInitialValues[attributeName] !== void 0) {
        logger.warn(
          `InitialAttributeValue for "${clusterDef.name}/${attributeName}" is provided but it's neither optional or mandatory for supportedFeatures: ${JSON.stringify(
            supportedFeatures
          )} but is set!`
        );
      }
    }
    const { id, persistent, fabricScoped, scene, fixed } = attributeDef[attributeName];
    if (attributesInitialValues[attributeName] !== void 0) {
      const getter = handlers[`${attributeName}AttributeGetter`];
      const setter = handlers[`${attributeName}AttributeSetter`];
      const validator = handlers[`${attributeName}AttributeValidator`];
      attributes[attributeName] = (0, import_AttributeServer.createAttributeServer)(
        clusterDef,
        attributeDef[attributeName],
        attributeName,
        attributesInitialValues[attributeName],
        () => result.clusterDataVersion,
        () => {
          if (result.clusterDataVersion === 4294967295) {
            result.clusterDataVersion = -1;
          }
          clusterStorage?.set("_clusterDataVersion", ++result.clusterDataVersion);
          return result.clusterDataVersion;
        },
        getter ? (session, endpoint, isFabricFiltered) => getter({
          attributes,
          endpoint,
          session,
          isFabricFiltered
        }) : void 0,
        setter ? (value, session, endpoint) => setter(value, {
          attributes,
          endpoint,
          session
        }) : void 0,
        validator ? (value, session, endpoint) => validator(value, {
          attributes,
          endpoint,
          session
        }) : void 0
      );
      if (fixed) {
        result[`get${capitalizedAttributeName}Attribute`] = () => attributes[attributeName].getLocal();
      } else if (fabricScoped) {
        result[`get${capitalizedAttributeName}Attribute`] = (fabric) => attributes[attributeName].getLocalForFabric(fabric);
        result[`set${capitalizedAttributeName}Attribute`] = (value, fabric) => attributes[attributeName].setLocalForFabric(value, fabric);
        result[`subscribe${capitalizedAttributeName}Attribute`] = (listener) => attributes[attributeName].addValueSetListener(listener);
      } else {
        if (scene) {
          sceneAttributeList.push(attributeName);
        }
        result[`get${capitalizedAttributeName}Attribute`] = () => attributes[attributeName].getLocal();
        result[`set${capitalizedAttributeName}Attribute`] = (value) => attributes[attributeName].setLocal(value);
        result[`subscribe${capitalizedAttributeName}Attribute`] = (listener) => attributes[attributeName].addValueSetListener(listener);
      }
      if (persistent || getter || setter) {
        const listener = (value) => attributeStorageListener(attributeName, fabricScoped || getter || setter ? void 0 : value);
        attributeStorageListeners.set(id, listener);
        attributes[attributeName].addValueChangeListener(listener);
      }
      attributeList.push((0, import_AttributeId.AttributeId)(id));
    } else {
      result[`get${capitalizedAttributeName}Attribute`] = () => void 0;
      if (!fixed) {
        result[`set${capitalizedAttributeName}Attribute`] = () => {
          throw new import_MatterError.ImplementationError(
            `Attribute ${attributeName} is optional and not initialized. To use it please initialize it first.`
          );
        };
        result[`subscribe${capitalizedAttributeName}Attribute`] = () => {
          throw new import_MatterError.ImplementationError(
            `Attribute ${attributeName} is optional and not initialized. To use it please initialize it first.`
          );
        };
      }
    }
  }
  attributes.attributeList.setLocal(attributeList.sort((a, b) => a - b));
  const acceptedCommandList = new Array();
  const generatedCommandList = new Array();
  for (const name2 in commandDef) {
    const handler = handlers[name2];
    if (commandDef[name2].isConditional) {
      const { mandatoryIf, optionalIf } = commandDef[name2];
      let conditionHasMatched = false;
      if (mandatoryIf !== void 0 && mandatoryIf.length > 0) {
        const conditionMatched = isConditionMatching(mandatoryIf, supportedFeatures);
        if (conditionMatched && handler === void 0) {
          logger.warn(
            `Command "${clusterDef.name}/${name2}" is REQUIRED by supportedFeatures: ${JSON.stringify(
              supportedFeatures
            )} but is not set!`
          );
        }
        conditionHasMatched = conditionHasMatched || conditionMatched;
      }
      if (!conditionHasMatched && optionalIf !== void 0 && optionalIf.length > 0) {
        const conditionMatched = isConditionMatching(optionalIf, supportedFeatures);
        if (conditionMatched && handler === void 0) {
          logger.debug(
            `Command "${clusterDef.name}/${name2}" is optional by supportedFeatures: ${JSON.stringify(
              supportedFeatures
            )} and is not set!`
          );
        }
        conditionHasMatched = conditionHasMatched || conditionMatched;
      }
      if (!conditionHasMatched && handler !== void 0) {
        logger.warn(
          `Command "${clusterDef.name}/${name2}" is provided but it's neither optional or mandatory for supportedFeatures: ${JSON.stringify(
            supportedFeatures
          )} but is set!`
        );
      }
    }
    if (handler === void 0)
      continue;
    const { requestId, requestSchema, responseId, responseSchema, timed } = commandDef[name2];
    commands[name2] = new import_CommandServer.CommandServer(
      requestId,
      responseId,
      name2,
      requestSchema,
      responseSchema,
      timed,
      (request, session, message, endpoint) => handler({
        request,
        attributes,
        events,
        session,
        message,
        endpoint
      })
    );
    if (!acceptedCommandList.includes(requestId)) {
      acceptedCommandList.push(requestId);
    }
    if (responseSchema !== import_Cluster.TlvNoResponse) {
      if (!generatedCommandList.includes(responseId)) {
        generatedCommandList.push(responseId);
      }
    }
  }
  attributes.acceptedCommandList.setLocal(acceptedCommandList.sort((a, b) => a - b));
  attributes.generatedCommandList.setLocal(generatedCommandList.sort((a, b) => a - b));
  const eventList = new Array();
  for (const eventName in eventDef) {
    const { id, schema, priority, optional } = eventDef[eventName];
    if (!optional && supportedEvents[eventName] !== true) {
      throw new import_MatterError.ImplementationError(`Event ${eventName} needs to be supported by cluster ${name} (${clusterId})`);
    }
    if (eventDef[eventName].isConditional) {
      const { mandatoryIf, optionalIf } = eventDef[eventName];
      let conditionHasMatched = false;
      if (mandatoryIf !== void 0) {
        const conditionMatched = isConditionMatching(mandatoryIf, supportedFeatures);
        if (conditionMatched && supportedEvents[eventName] === void 0) {
          logger.warn(
            `Event "${clusterDef.name}/${eventName}" is REQUIRED by supportedFeatures: ${JSON.stringify(
              supportedFeatures
            )} but is not set!`
          );
        }
        conditionHasMatched = conditionHasMatched || conditionMatched;
      }
      if (!conditionHasMatched && optionalIf !== void 0 && optionalIf.length > 0) {
        const conditionMatched = isConditionMatching(optionalIf, supportedFeatures);
        if (conditionMatched && supportedEvents[eventName] === void 0) {
          logger.debug(
            `Event "${clusterDef.name}/${eventName}" is optional by supportedFeatures: ${JSON.stringify(
              supportedFeatures
            )} and is not set!`
          );
        }
        conditionHasMatched = conditionHasMatched || conditionMatched;
      }
      if (!conditionHasMatched && supportedEvents[eventName] !== void 0) {
        logger.warn(
          `Event "${clusterDef.name}/${eventName}" is provided but it's neither optional or mandatory for supportedFeatures: ${JSON.stringify(
            supportedFeatures
          )} but is set!`
        );
      }
    }
    if (supportedEvents[eventName] === true) {
      events[eventName] = new import_EventServer.EventServer(id, clusterId, eventName, schema, priority);
      const capitalizedEventName = (0, import_String.capitalize)(eventName);
      result[`trigger${capitalizedEventName}Event`] = (event) => events[eventName].triggerEvent(event);
      eventList.push(id);
    }
  }
  attributes.eventList.setLocal(eventList.sort((a, b) => a - b));
  return result;
}
