// @project-chip/matter.js/dist/cjs/cluster/server/EventServer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EventServer_exports = {};
__export(EventServer_exports, {
  EventServer: () => EventServer
});
module.exports = __toCommonJS(EventServer_exports);
var import_MatterError = require("../../common/MatterError.js");
var import_Time = require("../../time/Time.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class EventServer {
  constructor(id, clusterId, name, schema, priority) {
    this.id = id;
    this.clusterId = clusterId;
    this.name = name;
    this.schema = schema;
    this.priority = priority;
    this.eventList = new Array();
    this.listeners = new Array();
  }
  assignToEndpoint(endpoint) {
    this.endpoint = endpoint;
  }
  // TODO Try to get rid of that late binding and simply things again
  //      potentially with refactoring out MatterDevice and MatterController
  bindToEventHandler(eventHandler) {
    this.eventHandler = eventHandler;
    for (const event of this.eventList) {
      const finalEvent = this.eventHandler.pushEvent(event);
      this.listeners.forEach((listener) => listener(finalEvent));
    }
    this.eventList = [];
  }
  triggerEvent(data) {
    if (this.endpoint === void 0 || this.endpoint.id === void 0) {
      throw new import_MatterError.InternalError("Endpoint not assigned");
    }
    const event = {
      eventId: this.id,
      clusterId: this.clusterId,
      endpointId: this.endpoint.id,
      epochTimestamp: import_Time.Time.nowMs(),
      priority: this.priority,
      data
    };
    if (this.eventHandler === void 0) {
      this.eventList.push(event);
    } else {
      const finalEvent = this.eventHandler.pushEvent(event);
      this.listeners.forEach((listener) => listener(finalEvent));
    }
  }
  addListener(listener) {
    this.listeners.push(listener);
  }
  removeListener(listener) {
    const entryIndex = this.listeners.indexOf(listener);
    if (entryIndex !== -1) {
      this.listeners.splice(entryIndex, 1);
    }
  }
}
