// @project-chip/matter.js/dist/cjs/cluster/server/GeneralCommissioningServer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GeneralCommissioningServer_exports = {};
__export(GeneralCommissioningServer_exports, {
  GeneralCommissioningClusterHandler: () => GeneralCommissioningClusterHandler
});
module.exports = __toCommonJS(GeneralCommissioningServer_exports);
var import_MatterError = require("../../common/MatterError.js");
var import_TryCatchHandler = require("../../common/TryCatchHandler.js");
var import_Logger = require("../../log/Logger.js");
var import_StatusCode = require("../../protocol/interaction/StatusCode.js");
var import_SecureSession = require("../../session/SecureSession.js");
var import_AdministratorCommissioningCluster = require("../definitions/AdministratorCommissioningCluster.js");
var import_BasicInformationCluster = require("../definitions/BasicInformationCluster.js");
var import_GeneralCommissioningCluster = require("../definitions/GeneralCommissioningCluster.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const SuccessResponse = { errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.Ok, debugText: "" };
const logger = import_Logger.Logger.get("GeneralCommissioningClusterHandler");
const GeneralCommissioningClusterHandler = (options) => ({
  initializeClusterServer: ({ attributes: { breadcrumb } }) => {
    breadcrumb.setLocal(BigInt(0));
  },
  armFailSafe: async ({
    request: { breadcrumb: breadcrumbStep, expiryLengthSeconds },
    attributes: { breadcrumb, basicCommissioningInfo },
    session,
    endpoint
  }) => {
    (0, import_SecureSession.assertSecureSession)(session, "armFailSafe can only be called on a secure session");
    const device = session.getContext();
    try {
      if (!device.isFailsafeArmed() && endpoint.getClusterServer(import_AdministratorCommissioningCluster.AdministratorCommissioning.Cluster)?.getWindowStatusAttribute() !== import_AdministratorCommissioningCluster.AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen && !session.isPase()) {
        throw new import_MatterError.MatterFlowError("Failed to arm failsafe using CASE while commissioning window is opened.");
      }
      await device.armFailSafe(
        expiryLengthSeconds,
        basicCommissioningInfo.getLocal().maxCumulativeFailsafeSeconds,
        session.getFabric(),
        endpoint
      );
      if (device.isFailsafeArmed()) {
        breadcrumb.setLocal(breadcrumbStep);
      }
    } catch (error) {
      if (error instanceof import_MatterError.MatterFlowError) {
        logger.debug(`Error while arming failSafe timer`, error);
        return {
          errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.BusyWithOtherAdmin,
          debugText: error.message
        };
      } else {
        throw error;
      }
    }
    return SuccessResponse;
  },
  setRegulatoryConfig: async ({
    request: { breadcrumb: breadcrumbStep, newRegulatoryConfig, countryCode },
    attributes: { breadcrumb, regulatoryConfig, locationCapability },
    endpoint
  }) => {
    const locationCapabilityValue = locationCapability.getLocal();
    const basicInformationCluster = endpoint.getClusterServer(import_BasicInformationCluster.BasicInformationCluster);
    if (basicInformationCluster === void 0) {
      throw new import_MatterError.ImplementationError("BasicInformationCluster needs to be present on the root endpoint");
    }
    const currentLocationCountryCode = basicInformationCluster.getLocationAttribute();
    if (currentLocationCountryCode !== countryCode) {
      if (options?.allowCountryCodeChange === false && countryCode !== "XX") {
        return {
          errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.ValueOutsideRange,
          debugText: `Country code change not allowed: ${countryCode}`
        };
      }
      if (options?.countryCodeWhitelist !== void 0 && !options?.countryCodeWhitelist.includes(countryCode)) {
        return {
          errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.ValueOutsideRange,
          debugText: `Country code change not allowed: ${countryCode}`
        };
      }
      if (countryCode !== "XX") {
        basicInformationCluster.setLocationAttribute(countryCode);
      }
    }
    let validValues;
    switch (locationCapabilityValue) {
      case import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Outdoor:
        validValues = [import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Outdoor];
        break;
      case import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Indoor:
        validValues = [import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Indoor];
        break;
      case import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.IndoorOutdoor:
        validValues = [
          import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Indoor,
          import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Outdoor,
          import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.IndoorOutdoor
        ];
        break;
      default:
        return {
          errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.ValueOutsideRange,
          debugText: `Invalid regulatory location: ${newRegulatoryConfig === import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Indoor ? "Indoor" : "Outdoor"}`
        };
    }
    if (!validValues.includes(newRegulatoryConfig)) {
      return {
        errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.ValueOutsideRange,
        debugText: `Invalid regulatory location: ${newRegulatoryConfig === import_GeneralCommissioningCluster.GeneralCommissioning.RegulatoryLocationType.Indoor ? "Indoor" : "Outdoor"}`
      };
    }
    regulatoryConfig.setLocal(newRegulatoryConfig);
    breadcrumb.setLocal(breadcrumbStep);
    return SuccessResponse;
  },
  commissioningComplete: async ({ session, attributes: { breadcrumb } }) => {
    const fabric = (0, import_TryCatchHandler.tryCatch)(
      () => session.getAssociatedFabric(),
      import_SecureSession.NoAssociatedFabricError,
      () => {
        throw new import_StatusCode.StatusResponseError("No associated fabric existing", import_StatusCode.StatusCode.UnsupportedAccess);
      }
    );
    if (session.isPase()) {
      return {
        errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.InvalidAuthentication,
        debugText: "Command not executed over CASE session."
      };
    }
    const device = session.getContext();
    if (!device.isFailsafeArmed()) {
      return { errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.NoFailSafe, debugText: "FailSafe not armed." };
    }
    (0, import_SecureSession.assertSecureSession)(session, "commissioningComplete can only be called on a secure session");
    const failSafeContext = device.getFailSafeContext();
    if (fabric.fabricIndex !== failSafeContext.associatedFabric?.fabricIndex) {
      return {
        errorCode: import_GeneralCommissioningCluster.GeneralCommissioning.CommissioningError.InvalidAuthentication,
        debugText: `Associated fabric ${fabric.fabricIndex} does not match the one from the failsafe context ${failSafeContext.associatedFabric?.fabricIndex}.`
      };
    }
    await device.completeCommission();
    breadcrumb.setLocal(BigInt(0));
    logger.info(`Commissioning completed on fabric #${fabric.fabricId} as node #${fabric.nodeId}.`);
    return SuccessResponse;
  }
});
