// @project-chip/matter.js/dist/cjs/cluster/server/GroupsServer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GroupsServer_exports = {};
__export(GroupsServer_exports, {
  GroupsClusterHandler: () => GroupsClusterHandler,
  GroupsManager: () => GroupsManager,
  createDefaultGroupsClusterServer: () => createDefaultGroupsClusterServer
});
module.exports = __toCommonJS(GroupsServer_exports);
var import_StatusCode = require("../../protocol/interaction/StatusCode.js");
var import_SecureSession = require("../../session/SecureSession.js");
var import_GroupsCluster = require("../definitions/GroupsCluster.js");
var import_IdentifyCluster = require("../definitions/IdentifyCluster.js");
var import_ClusterServer = require("./ClusterServer.js");
var import_ScenesServer = require("./ScenesServer.js");
/**
 * @license
 * Copyright 2022 The node-matter Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class GroupsManager {
  static setGroup(fabric, endpointId, groupId, groupName) {
    let endpointGroups = fabric.getScopedClusterDataValue(
      import_GroupsCluster.GroupsCluster,
      endpointId.toString()
    );
    if (endpointGroups === void 0) {
      endpointGroups = /* @__PURE__ */ new Map();
    }
    endpointGroups.set(groupId, groupName || "");
    fabric.setScopedClusterDataValue(import_GroupsCluster.GroupsCluster, endpointId.toString(), endpointGroups);
  }
  static getGroupName(fabric, endpointId, groupId) {
    const endpointGroups = fabric.getScopedClusterDataValue(
      import_GroupsCluster.GroupsCluster,
      endpointId.toString()
    );
    return endpointGroups?.get(groupId);
  }
  static hasGroup(fabric, endpointId, groupId) {
    const endpointGroups = fabric.getScopedClusterDataValue(
      import_GroupsCluster.GroupsCluster,
      endpointId.toString()
    );
    return endpointGroups?.has(groupId) ?? false;
  }
  static getGroups(fabric, endpointId) {
    return fabric.getScopedClusterDataValue(import_GroupsCluster.GroupsCluster, endpointId.toString()) ?? /* @__PURE__ */ new Map();
  }
  static removeGroup(fabric, endpointId, groupId) {
    const endpointGroups = fabric.getScopedClusterDataValue(
      import_GroupsCluster.GroupsCluster,
      endpointId.toString()
    );
    if (endpointGroups !== void 0) {
      if (endpointGroups.delete(groupId)) {
        fabric.persist();
        return true;
      }
    }
    return false;
  }
  static removeAllGroups(fabric, endpointId) {
    fabric.deleteScopedClusterDataValue(import_GroupsCluster.GroupsCluster, endpointId.toString());
  }
}
const GroupsClusterHandler = () => {
  const addGroupLogic = (groupId, groupName, fabric, endpointId) => {
    if (groupId < 1) {
      return { status: import_StatusCode.StatusCode.ConstraintError, groupId };
    }
    if (groupName.length > 16) {
      return { status: import_StatusCode.StatusCode.ConstraintError, groupId };
    }
    GroupsManager.setGroup(fabric, endpointId, groupId, groupName);
    return { status: import_StatusCode.StatusCode.Success, groupId };
  };
  return {
    addGroup: async ({ request: { groupId, groupName }, session, endpoint }) => {
      (0, import_SecureSession.assertSecureSession)(session);
      return addGroupLogic(groupId, groupName, session.getAssociatedFabric(), endpoint.getId());
    },
    viewGroup: async ({ request: { groupId }, session, endpoint }) => {
      if (groupId < 1) {
        return { status: import_StatusCode.StatusCode.ConstraintError, groupId, groupName: "" };
      }
      (0, import_SecureSession.assertSecureSession)(session);
      const groupName = GroupsManager.getGroupName(session.getAssociatedFabric(), endpoint.getId(), groupId);
      if (groupName !== void 0) {
        return { status: import_StatusCode.StatusCode.Success, groupId, groupName };
      }
      return { status: import_StatusCode.StatusCode.NotFound, groupId, groupName: "" };
    },
    getGroupMembership: async ({ request: { groupList }, session, endpoint }) => {
      (0, import_SecureSession.assertSecureSession)(session);
      const endpointGroups = GroupsManager.getGroups(session.getAssociatedFabric(), endpoint.getId());
      const fabricGroupsList = Array.from(endpointGroups.keys());
      const capacity = fabricGroupsList.length < 255 ? 254 - fabricGroupsList.length : 0;
      if (groupList.length === 0) {
        return { capacity, groupList: fabricGroupsList };
      }
      const filteredGroupsList = groupList.filter((groupId) => endpointGroups.get(groupId));
      if (filteredGroupsList.length === 0) {
        return { capacity, groupList: [] };
      }
      return { capacity, groupList: filteredGroupsList };
    },
    removeGroup: async ({ request: { groupId }, session, endpoint }) => {
      if (groupId < 1) {
        return { status: import_StatusCode.StatusCode.ConstraintError, groupId };
      }
      (0, import_SecureSession.assertSecureSession)(session);
      const fabric = session.getAssociatedFabric();
      if (GroupsManager.removeGroup(session.getAssociatedFabric(), endpoint.getId(), groupId)) {
        import_ScenesServer.ScenesManager.removeAllScenesForGroup(fabric, endpoint.getId(), groupId);
        return { status: import_StatusCode.StatusCode.Success, groupId };
      }
      return { status: import_StatusCode.StatusCode.NotFound, groupId };
    },
    removeAllGroups: async ({ session, endpoint }) => {
      (0, import_SecureSession.assertSecureSession)(session);
      const fabric = session.getAssociatedFabric();
      GroupsManager.removeAllGroups(fabric, endpoint.getId());
      import_ScenesServer.ScenesManager.removeAllNonGlobalScenesForEndpoint(fabric, endpoint.getId());
      return;
    },
    addGroupIfIdentifying: async ({ request: { groupId, groupName }, session, endpoint }) => {
      const identifyCluster = endpoint.getClusterServer(import_IdentifyCluster.IdentifyCluster);
      if (identifyCluster) {
        if (identifyCluster.attributes.identifyTime.getLocal() > 0) {
          (0, import_SecureSession.assertSecureSession)(session);
          addGroupLogic(groupId, groupName, session.getAssociatedFabric(), endpoint.getId());
        }
      }
    }
  };
};
const createDefaultGroupsClusterServer = () => (0, import_ClusterServer.ClusterServer)(
  import_GroupsCluster.GroupsCluster,
  {
    nameSupport: {
      nameSupport: true
    }
  },
  GroupsClusterHandler()
);
