// @project-chip/matter.js/dist/cjs/cluster/server/OperationalCredentialsServer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OperationalCredentialsServer_exports = {};
__export(OperationalCredentialsServer_exports, {
  OperationalCredentialsClusterHandler: () => OperationalCredentialsClusterHandler,
  TlvAttestation: () => TlvAttestation,
  TlvCertSigningRequest: () => TlvCertSigningRequest
});
module.exports = __toCommonJS(OperationalCredentialsServer_exports);
var import_FailSafeManager = require("../../common/FailSafeManager.js");
var import_MatterError = require("../../common/MatterError.js");
var import_TryCatchHandler = require("../../common/TryCatchHandler.js");
var import_Crypto = require("../../crypto/Crypto.js");
var import_Key = require("../../crypto/Key.js");
var import_FabricIndex = require("../../datatype/FabricIndex.js");
var import_FabricManager = require("../../fabric/FabricManager.js");
var import_Logger = require("../../log/Logger.js");
var import_StatusCode = require("../../protocol/interaction/StatusCode.js");
var import_SecureSession = require("../../session/SecureSession.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvString = require("../../tlv/TlvString.js");
var import_BasicInformationCluster = require("../definitions/BasicInformationCluster.js");
var import_OperationalCredentialsCluster = require("../definitions/OperationalCredentialsCluster.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("OperationalCredentialsServer");
const TlvAttestation = (0, import_TlvObject.TlvObject)({
  declaration: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvByteString),
  attestationNonce: (0, import_TlvObject.TlvField)(2, import_TlvString.TlvByteString.bound({ length: 32 })),
  timestamp: (0, import_TlvObject.TlvField)(3, import_TlvNumber.TlvUInt32),
  // TODO: check actual max length in specs
  firmwareInfo: (0, import_TlvObject.TlvOptionalField)(4, import_TlvString.TlvByteString)
});
const TlvCertSigningRequest = (0, import_TlvObject.TlvObject)({
  certSigningRequest: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvByteString),
  csrNonce: (0, import_TlvObject.TlvField)(2, import_TlvString.TlvByteString.bound({ length: 32 })),
  vendorReserved1: (0, import_TlvObject.TlvOptionalField)(3, import_TlvString.TlvByteString),
  vendorReserved2: (0, import_TlvObject.TlvOptionalField)(4, import_TlvString.TlvByteString),
  vendorReserved3: (0, import_TlvObject.TlvOptionalField)(5, import_TlvString.TlvByteString)
});
function signWithDeviceKey(conf, session, data) {
  return import_Crypto.Crypto.sign((0, import_Key.PrivateKey)(conf.devicePrivateKey), [data, session.getAttestationChallengeKey()]);
}
const OperationalCredentialsClusterHandler = (conf) => ({
  attestationRequest: async ({ request: { attestationNonce }, session }) => {
    (0, import_SecureSession.assertSecureSession)(session);
    const elements = TlvAttestation.encode({
      declaration: conf.certificationDeclaration,
      attestationNonce,
      timestamp: 0
    });
    return { attestationElements: elements, attestationSignature: signWithDeviceKey(conf, session, elements) };
  },
  csrRequest: async ({ request: { csrNonce, isForUpdateNoc }, session }) => {
    (0, import_SecureSession.assertSecureSession)(session);
    if (isForUpdateNoc && session.isPase()) {
      throw new import_StatusCode.StatusResponseError(
        "csrRequest for UpdateNoc received on a PASE session.",
        import_StatusCode.StatusCode.InvalidCommand
      );
    }
    const device = session.getContext();
    device.assertFailSafeArmed("csrRequest received while failsafe is not armed.");
    const failSafeContext = device.getFailSafeContext();
    if (failSafeContext.fabricIndex !== void 0) {
      throw new import_StatusCode.StatusResponseError(
        `csrRequest received after ${failSafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"} already invoked.`,
        import_StatusCode.StatusCode.ConstraintError
      );
    }
    const certSigningRequest = failSafeContext.createCertificateSigningRequest(
      isForUpdateNoc ?? false,
      session.getId()
    );
    const nocsrElements = TlvCertSigningRequest.encode({ certSigningRequest, csrNonce });
    return { nocsrElements, attestationSignature: signWithDeviceKey(conf, session, nocsrElements) };
  },
  certificateChainRequest: async ({ request: { certificateType } }) => {
    switch (certificateType) {
      case import_OperationalCredentialsCluster.OperationalCredentials.CertificateChainType.DacCertificate:
        return { certificate: conf.deviceCertificate };
      case import_OperationalCredentialsCluster.OperationalCredentials.CertificateChainType.PaiCertificate:
        return { certificate: conf.deviceIntermediateCertificate };
      default:
        throw new import_MatterError.MatterFlowError(`Unsupported certificate type: ${certificateType}`);
    }
  },
  addNoc: async ({
    request: { nocValue, icacValue, ipkValue, caseAdminSubject, adminVendorId },
    attributes: { nocs, commissionedFabrics, fabrics, trustedRootCertificates, supportedFabrics },
    session
  }) => {
    if (!session.isSecure())
      throw new import_MatterError.MatterFlowError("addOperationalCert should be called on a secure session.");
    const device = session.getContext();
    device.assertFailSafeArmed("addNoc received while failsafe is not armed.");
    const failSafeContext = device.getFailSafeContext();
    if (failSafeContext.fabricIndex !== void 0) {
      throw new import_StatusCode.StatusResponseError(
        `addNoc received after ${failSafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"} already invoked.`,
        import_StatusCode.StatusCode.ConstraintError
      );
    }
    if (!failSafeContext.fabricBuilder.hasRootCert()) {
      return {
        statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.InvalidNoc,
        debugText: "Root certificate not found."
      };
    }
    if (failSafeContext.csrSessionId !== session.getId()) {
      return {
        statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.MissingCsr,
        debugText: "CSR not found in failsafe context."
      };
    }
    if (failSafeContext.forUpdateNoc) {
      throw new import_StatusCode.StatusResponseError(
        `addNoc received after csr request was invoked for UpdateNOC.`,
        import_StatusCode.StatusCode.ConstraintError
      );
    }
    if (device.getFabrics().length === supportedFabrics.getLocal()) {
      return {
        statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.TableFull,
        debugText: `No more fabrics can be added because limit ${supportedFabrics.getLocal()} reached.`
      };
    }
    let fabric;
    try {
      fabric = await failSafeContext.buildFabric({
        nocValue,
        icacValue,
        adminVendorId,
        ipkValue,
        caseAdminSubject
      });
    } catch (error) {
      if (error instanceof import_FailSafeManager.MatterFabricConflictError) {
        return {
          statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.FabricConflict,
          debugText: error.message
        };
      } else if (error instanceof import_FabricManager.FabricTableFullError) {
        return {
          statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.TableFull,
          debugText: error.message
        };
      } else {
        throw error;
      }
    }
    await device.addFabric(fabric);
    (0, import_SecureSession.assertSecureSession)(session);
    if (session.isPase()) {
      logger.debug(`Add Fabric with index ${fabric.fabricIndex} to PASE session ${session.name}.`);
      session.addAssociatedFabric(fabric);
    }
    nocs.updated(session);
    commissionedFabrics.updated(session);
    fabrics.updated(session);
    trustedRootCertificates.updated(session);
    logger.info(`addNoc success, adminVendorId ${adminVendorId}, caseAdminSubject ${caseAdminSubject}`);
    return { statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.Ok, fabricIndex: fabric.fabricIndex };
  },
  fabricsAttributeGetter: ({ session, isFabricFiltered }) => {
    if (session === void 0 || !session.isSecure())
      return [];
    const fabrics = isFabricFiltered ? [session.getAssociatedFabric()] : session.getContext().getFabrics();
    return fabrics.map((fabric) => ({
      fabricId: fabric.fabricId,
      label: fabric.label,
      nodeId: fabric.nodeId,
      rootPublicKey: fabric.rootPublicKey,
      vendorId: fabric.rootVendorId,
      fabricIndex: fabric.fabricIndex
    }));
  },
  nocsAttributeGetter: ({ session, isFabricFiltered }) => {
    if (session === void 0 || !session.isSecure())
      return [];
    const fabrics = isFabricFiltered ? [session.getAssociatedFabric()] : session.getContext().getFabrics();
    return fabrics.map((fabric) => ({
      noc: fabric.operationalCert,
      icac: fabric.intermediateCACert ?? null,
      fabricIndex: fabric.fabricIndex
    }));
  },
  commissionedFabricsAttributeGetter: ({ session }) => {
    if (session === void 0 || !session.isSecure())
      return 0;
    return session.getContext().getFabrics().length;
  },
  trustedRootCertificatesAttributeGetter: ({ session, isFabricFiltered }) => {
    if (session === void 0 || !session.isSecure())
      return [];
    const fabrics = isFabricFiltered ? [session.getAssociatedFabric()] : session.getContext().getFabrics();
    return fabrics.map((fabric) => fabric.rootCert);
  },
  currentFabricIndexAttributeGetter: ({ session }) => {
    if (session === void 0 || !session.isSecure())
      return import_FabricIndex.FabricIndex.NO_FABRIC;
    (0, import_SecureSession.assertSecureSession)(session);
    return session.getFabric()?.fabricIndex ?? import_FabricIndex.FabricIndex.NO_FABRIC;
  },
  updateNoc: async ({ request: { nocValue, icacValue }, attributes: { nocs, fabrics }, session }) => {
    (0, import_SecureSession.assertSecureSession)(session);
    (0, import_TryCatchHandler.tryCatch)(
      () => session.getAssociatedFabric(),
      import_SecureSession.NoAssociatedFabricError,
      () => {
        throw new import_StatusCode.StatusResponseError("No associated fabric existing", import_StatusCode.StatusCode.UnsupportedAccess);
      }
    );
    const device = session.getContext();
    device.assertFailSafeArmed("updateNoc received while failsafe is not armed.");
    const failSafeContext = device.getFailSafeContext();
    if (failSafeContext.fabricIndex !== void 0) {
      throw new import_StatusCode.StatusResponseError(
        `updateNoc received after ${failSafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"} already invoked.`,
        import_StatusCode.StatusCode.ConstraintError
      );
    }
    if (failSafeContext.forUpdateNoc) {
      throw new import_StatusCode.StatusResponseError(
        `addNoc received after csr request was invoked for UpdateNOC.`,
        import_StatusCode.StatusCode.ConstraintError
      );
    }
    if (failSafeContext.fabricBuilder.hasRootCert()) {
      throw new import_StatusCode.StatusResponseError(
        "Trusted root certificate added in this session which is now allowed for UpdateNOC.",
        import_StatusCode.StatusCode.ConstraintError
      );
    }
    if (!failSafeContext.forUpdateNoc) {
      throw new import_StatusCode.StatusResponseError("csrRequest not invoked for UpdateNOC.", import_StatusCode.StatusCode.ConstraintError);
    }
    if (session.getAssociatedFabric().fabricIndex !== failSafeContext.associatedFabric?.fabricIndex) {
      throw new import_StatusCode.StatusResponseError(
        "Fabric of this session and the failsafe context do not match.",
        import_StatusCode.StatusCode.ConstraintError
      );
    }
    const updateFabric = await failSafeContext.buildUpdatedFabric(nocValue, icacValue);
    device.updateFabric(updateFabric);
    nocs.updated(session);
    fabrics.updated(session);
    return {
      statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.Ok,
      fabricIndex: updateFabric.fabricIndex
    };
  },
  updateFabricLabel: async ({ request: { label }, attributes: { fabrics }, session }) => {
    (0, import_SecureSession.assertSecureSession)(session, "updateOperationalCert should be called on a secure session.");
    const fabric = (0, import_TryCatchHandler.tryCatch)(
      () => session.getAssociatedFabric(),
      import_SecureSession.NoAssociatedFabricError,
      () => {
        throw new import_StatusCode.StatusResponseError("No associated fabric existing", import_StatusCode.StatusCode.UnsupportedAccess);
      }
    );
    const currentFabricIndex = fabric.fabricIndex;
    const device = session.getContext();
    const conflictingLabelFabric = device.getFabrics().find((f) => f.label === label && f.fabricIndex !== currentFabricIndex);
    if (conflictingLabelFabric !== void 0) {
      return {
        statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.LabelConflict,
        debugText: `Label ${label} already used by fabric ${conflictingLabelFabric.fabricIndex}`
      };
    }
    fabric.setLabel(label);
    fabrics.updated(session);
    return { statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.Ok, fabricIndex: fabric.fabricIndex };
  },
  removeFabric: async ({
    request: { fabricIndex },
    attributes: { nocs, commissionedFabrics, fabrics, trustedRootCertificates },
    session,
    endpoint
  }) => {
    const device = session.getContext();
    const fabric = device.getFabricByIndex(fabricIndex);
    if (fabric === void 0) {
      return {
        statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.InvalidFabricIndex,
        debugText: `Fabric ${fabricIndex} not found`
      };
    }
    const basicInformationCluster = endpoint.getClusterServer(import_BasicInformationCluster.BasicInformation.Cluster);
    basicInformationCluster?.triggerLeaveEvent?.({ fabricIndex });
    (0, import_SecureSession.assertSecureSession)(session);
    await fabric.remove(session.getId());
    nocs.updated(session);
    commissionedFabrics.updated(session);
    fabrics.updated(session);
    trustedRootCertificates.updated(session);
    return {
      statusCode: import_OperationalCredentialsCluster.OperationalCredentials.NodeOperationalCertStatus.Ok,
      fabricIndex
    };
  },
  addTrustedRootCertificate: async ({ request: { rootCaCertificate }, session }) => {
    const device = session.getContext();
    device.assertFailSafeArmed("addTrustedRootCertificate received while failsafe is not armed.");
    const failSafeContext = device.getFailSafeContext();
    if (failSafeContext.fabricBuilder.hasRootCert()) {
      throw new import_StatusCode.StatusResponseError(
        "Trusted root certificate already added in this FailSafe context.",
        import_StatusCode.StatusCode.ConstraintError
      );
    }
    if (failSafeContext.fabricIndex !== void 0) {
      throw new import_StatusCode.StatusResponseError(
        `Can not add trusted root certificates after ${failSafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"}.`,
        import_StatusCode.StatusCode.ConstraintError
      );
    }
    session.getContext().getFailSafeContext().setRootCert(rootCaCertificate);
  }
});
