// @project-chip/matter.js/dist/esm/mdns/MdnsBroadcaster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { AAAARecord, ARecord, PtrRecord, SrvRecord, TxtRecord } from "../codec/DnsCodec.js";
import {
  PairingHintBitmapSchema
} from "../common/InstanceBroadcaster.js";
import { ImplementationError } from "../common/MatterError.js";
import { Crypto } from "../crypto/Crypto.js";
import { NodeId } from "../datatype/NodeId.js";
import { Logger } from "../log/Logger.js";
import { Network } from "../net/Network.js";
import {
  SESSION_ACTIVE_INTERVAL_MS,
  SESSION_ACTIVE_THRESHOLD_MS,
  SESSION_IDLE_INTERVAL_MS
} from "../session/Session.js";
import { isIPv4, isIPv6 } from "../util/Ip.js";
import {
  MATTER_COMMISSIONER_SERVICE_QNAME,
  MATTER_COMMISSION_SERVICE_QNAME,
  MATTER_SERVICE_QNAME,
  SERVICE_DISCOVERY_QNAME,
  getCommissioningModeQname,
  getDeviceInstanceQname,
  getDeviceMatterQname,
  getDeviceTypeQname,
  getFabricQname,
  getLongDiscriminatorQname,
  getShortDiscriminatorQname,
  getVendorQname
} from "./MdnsConsts.js";
import { AnnouncementType, MdnsServer } from "./MdnsServer.js";
const logger = Logger.get("MdnsBroadcaster");
const TCP_SUPPORTED = 0;
const ICD_SUPPORTED = 0;
const DEFAULT_PAIRING_HINT = {
  powerCycle: true,
  deviceManual: true
};
class MdnsBroadcaster {
  constructor(mdnsServer, enableIpv4) {
    this.mdnsServer = mdnsServer;
    this.enableIpv4 = enableIpv4;
    this.network = Network.get();
    this.activeCommissioningAnnouncements = /* @__PURE__ */ new Set();
    this.activeOperationalAnnouncements = /* @__PURE__ */ new Map();
  }
  static async create(options) {
    const { enableIpv4, multicastInterface } = options ?? {};
    return new MdnsBroadcaster(
      await MdnsServer.create({ enableIpv4, netInterface: multicastInterface }),
      enableIpv4
    );
  }
  validatePairingInstructions(pairingHint, pairingInstructions) {
    const needsInstructions = [
      "customInstruction",
      "pressRestButtonForNumberOfSeconds",
      "pressResetButtonUntilLightBlinks",
      "pressResetButtonForNumberOfSecondsWithApplicationOfPower",
      "pressResetButtonUntilLightBlinksWithApplicationOfPower",
      "pressResetButtonNumberOfTimes",
      "pressSetupButtonForNumberOfSeconds",
      "pressSetupButtonUntilLightBlinks",
      "pressSetupButtonForNumberOfSecondsWithApplicationOfPower",
      "pressSetupButtonUntilLightBlinksWithApplicationOfPower",
      "pressSetupButtonNumberOfTimes"
    ].find((hint) => pairingHint[hint] === true);
    if (needsInstructions && pairingInstructions.length === 0) {
      throw new ImplementationError(
        `Pairing instructions required for Pairing Hint of type "${needsInstructions}"`
      );
    }
  }
  getIpRecords(hostname, ips) {
    const records = new Array();
    ips.forEach((ip) => {
      if (isIPv6(ip)) {
        records.push(AAAARecord(hostname, ip));
      } else if (isIPv4(ip)) {
        if (this.enableIpv4) {
          records.push(ARecord(hostname, ip));
        }
      } else {
        logger.warn(`Unknown IP address type: ${ip}`);
      }
    });
    return records;
  }
  /** Set the Broadcaster data to announce a device ready for commissioning in a special mode */
  async setCommissionMode(announcedNetPort, mode, {
    deviceName,
    deviceType,
    vendorId,
    productId,
    discriminator,
    sessionIdleInterval = SESSION_ACTIVE_INTERVAL_MS,
    sessionActiveInterval = SESSION_IDLE_INTERVAL_MS,
    sessionActiveThreshold = SESSION_ACTIVE_THRESHOLD_MS,
    pairingHint = DEFAULT_PAIRING_HINT,
    pairingInstructions = ""
  }) {
    await this.expireCommissioningAnnouncement(announcedNetPort);
    logger.debug(
      `Announce commissioning mode ${mode} ${deviceName} ${deviceType} ${vendorId} ${productId} ${discriminator} ${announcedNetPort}`
    );
    this.activeCommissioningAnnouncements.add(announcedNetPort);
    const shortDiscriminator = discriminator >> 8 & 15;
    const instanceId = Crypto.getRandomData(8).toHex().toUpperCase();
    const vendorQname = getVendorQname(vendorId);
    const deviceTypeQname = getDeviceTypeQname(deviceType);
    const shortDiscriminatorQname = getShortDiscriminatorQname(shortDiscriminator);
    const longDiscriminatorQname = getLongDiscriminatorQname(discriminator);
    const commissionModeQname = getCommissioningModeQname();
    const deviceQname = getDeviceInstanceQname(instanceId);
    this.validatePairingInstructions(pairingHint, pairingInstructions);
    await this.mdnsServer.setRecordsGenerator(announcedNetPort, AnnouncementType.Commissionable, (netInterface) => {
      const ipMac = this.network.getIpMac(netInterface);
      if (ipMac === void 0)
        return [];
      const { mac, ips } = ipMac;
      const hostname = mac.replace(/:/g, "").toUpperCase() + "0000.local";
      logger.debug(
        "Announcement Generator: Commission mode ",
        Logger.dict({
          mode,
          qname: deviceQname,
          port: announcedNetPort,
          interface: netInterface
        })
      );
      const records = [
        PtrRecord(SERVICE_DISCOVERY_QNAME, MATTER_COMMISSION_SERVICE_QNAME),
        PtrRecord(SERVICE_DISCOVERY_QNAME, vendorQname),
        PtrRecord(SERVICE_DISCOVERY_QNAME, deviceTypeQname),
        PtrRecord(SERVICE_DISCOVERY_QNAME, shortDiscriminatorQname),
        PtrRecord(SERVICE_DISCOVERY_QNAME, longDiscriminatorQname),
        PtrRecord(SERVICE_DISCOVERY_QNAME, commissionModeQname),
        PtrRecord(MATTER_COMMISSION_SERVICE_QNAME, deviceQname),
        PtrRecord(vendorQname, deviceQname),
        PtrRecord(deviceTypeQname, deviceQname),
        PtrRecord(shortDiscriminatorQname, deviceQname),
        PtrRecord(longDiscriminatorQname, deviceQname),
        PtrRecord(commissionModeQname, deviceQname),
        SrvRecord(deviceQname, { priority: 0, weight: 0, port: announcedNetPort, target: hostname }),
        TxtRecord(deviceQname, [
          `VP=${vendorId}+${productId}`,
          `DT=${deviceType}`,
          `DN=${deviceName}`,
          `SII=${sessionIdleInterval}`,
          `SAI=${sessionActiveInterval}`,
          `SAT=${sessionActiveThreshold}`,
          `T=${TCP_SUPPORTED}`,
          `D=${discriminator}`,
          `CM=${mode}`,
          `PH=${PairingHintBitmapSchema.encode(pairingHint)}`,
          `PI=${pairingInstructions}`,
          `ICD=${ICD_SUPPORTED}`
        ])
      ];
      records.push(...this.getIpRecords(hostname, ips));
      return records;
    });
  }
  /** Set the Broadcaster Data to announce a device for operative discovery (aka "already paired") */
  async setFabrics(announcedNetPort, fabrics, {
    sessionIdleInterval = SESSION_ACTIVE_INTERVAL_MS,
    sessionActiveInterval = SESSION_IDLE_INTERVAL_MS,
    sessionActiveThreshold = SESSION_ACTIVE_THRESHOLD_MS
  } = {}) {
    const currentOperationalFabrics = this.activeOperationalAnnouncements.get(announcedNetPort);
    if (currentOperationalFabrics !== void 0) {
      const fabricIndexesSet = new Set(fabrics.map((f) => f.fabricIndex));
      if (!currentOperationalFabrics.every((fabricIndex) => fabricIndexesSet.has(fabricIndex))) {
        await this.expireFabricAnnouncement(announcedNetPort);
      }
    }
    this.activeOperationalAnnouncements.set(
      announcedNetPort,
      fabrics.map((f) => f.fabricIndex)
    );
    await this.mdnsServer.setRecordsGenerator(announcedNetPort, AnnouncementType.Operative, (netInterface) => {
      const ipMac = this.network.getIpMac(netInterface);
      if (ipMac === void 0)
        return [];
      const { mac, ips } = ipMac;
      const hostname = mac.replace(/:/g, "").toUpperCase() + "0000.local";
      const records = [PtrRecord(SERVICE_DISCOVERY_QNAME, MATTER_SERVICE_QNAME)];
      fabrics.forEach((fabric) => {
        const { operationalId, nodeId } = fabric;
        const operationalIdString = operationalId.toHex().toUpperCase();
        const fabricQname = getFabricQname(operationalIdString);
        const deviceMatterQname = getDeviceMatterQname(operationalIdString, NodeId.toHexString(nodeId));
        logger.debug(
          "Announcement Generator: Fabric",
          Logger.dict({
            id: `${operationalId.toHex()}/${nodeId}`,
            qname: deviceMatterQname,
            port: announcedNetPort,
            interface: netInterface
          })
        );
        const fabricRecords = [
          PtrRecord(SERVICE_DISCOVERY_QNAME, fabricQname),
          PtrRecord(MATTER_SERVICE_QNAME, deviceMatterQname),
          PtrRecord(fabricQname, deviceMatterQname),
          SrvRecord(deviceMatterQname, { priority: 0, weight: 0, port: announcedNetPort, target: hostname }),
          TxtRecord(deviceMatterQname, [
            `SII=${sessionIdleInterval}`,
            `SAI=${sessionActiveInterval}`,
            `SAT=${sessionActiveThreshold}`,
            `T=${TCP_SUPPORTED}`,
            `ICD=${ICD_SUPPORTED}`
          ])
        ];
        records.push(...fabricRecords);
      });
      records.push(...this.getIpRecords(hostname, ips));
      return records;
    });
  }
  /** Set the Broadcaster data to announce a Commissioner (aka Commissioner discovery) */
  async setCommissionerInfo(announcedNetPort, {
    deviceName,
    deviceType,
    vendorId,
    productId,
    sessionIdleInterval = SESSION_ACTIVE_INTERVAL_MS,
    sessionActiveInterval = SESSION_IDLE_INTERVAL_MS,
    sessionActiveThreshold = SESSION_ACTIVE_THRESHOLD_MS
  }) {
    logger.debug(
      "Announcement: Commissioner",
      Logger.dict({
        port: announcedNetPort,
        deviceType
      })
    );
    const instanceId = Crypto.getRandomData(8).toHex().toUpperCase();
    const deviceTypeQname = `_T${deviceType}._sub.${MATTER_COMMISSIONER_SERVICE_QNAME}`;
    const vendorQname = `_V${vendorId}._sub.${MATTER_COMMISSIONER_SERVICE_QNAME}`;
    const deviceQname = `${instanceId}.${MATTER_COMMISSIONER_SERVICE_QNAME}`;
    this.activeCommissioningAnnouncements.add(announcedNetPort);
    await this.mdnsServer.setRecordsGenerator(announcedNetPort, AnnouncementType.Commissionable, (netInterface) => {
      const ipMac = this.network.getIpMac(netInterface);
      if (ipMac === void 0)
        return [];
      const { mac, ips } = ipMac;
      const hostname = mac.replace(/:/g, "").toUpperCase() + "0000.local";
      const records = [
        PtrRecord(SERVICE_DISCOVERY_QNAME, MATTER_COMMISSIONER_SERVICE_QNAME),
        PtrRecord(MATTER_COMMISSIONER_SERVICE_QNAME, vendorQname),
        PtrRecord(vendorQname, deviceQname),
        SrvRecord(deviceQname, { priority: 0, weight: 0, port: announcedNetPort, target: hostname }),
        TxtRecord(deviceQname, [
          `VP=${vendorId}+${productId}`,
          `DT=${deviceType}`,
          `DN=${deviceName}`,
          `SII=${sessionIdleInterval}`,
          `SAI=${sessionActiveInterval}`,
          `SAT=${sessionActiveThreshold}`,
          `T=${TCP_SUPPORTED}`,
          `ICD=${ICD_SUPPORTED}`
        ])
      ];
      if (deviceType !== void 0) {
        records.push(PtrRecord(SERVICE_DISCOVERY_QNAME, deviceTypeQname));
        records.push(PtrRecord(deviceTypeQname, deviceQname));
      }
      records.push(...this.getIpRecords(hostname, ips));
      return records;
    });
  }
  async announce(announcementPort) {
    this.mdnsServer.announce(announcementPort).catch((error) => logger.error(error));
  }
  async expireFabricAnnouncement(announcementPort) {
    if (this.activeOperationalAnnouncements.has(announcementPort)) {
      await this.mdnsServer.expireAnnouncements(announcementPort, AnnouncementType.Operative);
      this.activeOperationalAnnouncements.delete(announcementPort);
    }
  }
  async expireCommissioningAnnouncement(announcementPort) {
    if (this.activeCommissioningAnnouncements.has(announcementPort)) {
      await this.mdnsServer.expireAnnouncements(announcementPort, AnnouncementType.Commissionable);
      this.activeCommissioningAnnouncements.delete(announcementPort);
    }
  }
  async expireAllAnnouncements(announcementPort) {
    if (!this.activeCommissioningAnnouncements.has(announcementPort) && !this.activeOperationalAnnouncements.has(announcementPort))
      return;
    await this.mdnsServer.expireAnnouncements(announcementPort);
    this.activeCommissioningAnnouncements.delete(announcementPort);
    this.activeOperationalAnnouncements.delete(announcementPort);
  }
  async close() {
    await this.mdnsServer.expireAnnouncements();
    this.activeCommissioningAnnouncements.clear();
    this.activeOperationalAnnouncements.clear();
    await this.mdnsServer.close();
  }
}
export {
  MdnsBroadcaster
};
