// @project-chip/matter.js/dist/esm/mdns/MdnsInstanceBroadcaster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class MdnsInstanceBroadcaster {
  constructor(instancePort, mdnsBroadcaster) {
    this.instancePort = instancePort;
    this.mdnsBroadcaster = mdnsBroadcaster;
  }
  async setCommissionMode(mode, deviceData) {
    await this.mdnsBroadcaster.setCommissionMode(this.instancePort, mode, deviceData);
  }
  /** Set the Broadcaster Data to announce a device for operative discovery (aka "already paired") */
  async setFabrics(fabrics, expireCommissioningAnnouncements = false) {
    if (expireCommissioningAnnouncements) {
      await this.mdnsBroadcaster.expireCommissioningAnnouncement(this.instancePort);
    }
    await this.mdnsBroadcaster.setFabrics(this.instancePort, fabrics);
  }
  /** Set the Broadcaster data to announce a Commissioner (aka Commissioner discovery) */
  async setCommissionerInfo(commissionerData) {
    await this.mdnsBroadcaster.setCommissionerInfo(this.instancePort, commissionerData);
  }
  async announce() {
    await this.mdnsBroadcaster.announce(this.instancePort);
  }
  async expireFabricAnnouncement() {
    await this.mdnsBroadcaster.expireFabricAnnouncement(this.instancePort);
  }
  async expireCommissioningAnnouncement() {
    await this.mdnsBroadcaster.expireCommissioningAnnouncement(this.instancePort);
  }
  async expireAllAnnouncements() {
    await this.mdnsBroadcaster.expireAllAnnouncements(this.instancePort);
  }
}
export {
  MdnsInstanceBroadcaster
};
