// @project-chip/matter.js/dist/cjs/cluster/ClusterFactory.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClusterFactory_exports = {};
__export(ClusterFactory_exports, {
  ClusterFactory: () => ClusterFactory,
  IllegalClusterError: () => IllegalClusterError
});
module.exports = __toCommonJS(ClusterFactory_exports);
var import_MatterError = require("../common/MatterError.js");
var import_ClusterId = require("../datatype/ClusterId.js");
var import_String = require("../util/String.js");
var import_Cluster = require("./Cluster.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class IllegalClusterError extends import_MatterError.MatterError {
}
var ClusterFactory;
((ClusterFactory2) => {
  function Component(component) {
    return {
      attributes: {},
      commands: {},
      events: {},
      ...component
    };
  }
  ClusterFactory2.Component = Component;
  function Definition(definition) {
    return {
      id: (0, import_ClusterId.ClusterId)(definition.id),
      name: definition.name,
      revision: definition.revision,
      features: definition.features ?? {},
      supportedFeatures: definition.supportedFeatures ?? {},
      attributes: {
        ...(0, import_Cluster.GlobalAttributes)(definition.features ?? {}),
        ...definition.attributes
      },
      commands: definition.commands ?? {},
      events: definition.events ?? {},
      unknown: definition.unknown
    };
  }
  ClusterFactory2.Definition = Definition;
  function Extensible(definition, extender) {
    return {
      ...Definition(definition),
      with: extender
    };
  }
  ClusterFactory2.Extensible = Extensible;
  function ExtensibleOnly(extender) {
    return {
      with: extender
    };
  }
  ClusterFactory2.ExtensibleOnly = ExtensibleOnly;
  function extend(definition, elements, ...applicableFeatures) {
    let applicable = false;
    pool:
      for (const features of applicableFeatures) {
        for (const k in features) {
          if (!!definition.supportedFeatures[k] !== !!features[k]) {
            continue pool;
          }
        }
        applicable = true;
        break;
      }
    if (!applicable) {
      return;
    }
    if (elements.attributes) {
      if (definition.attributes) {
        definition.attributes = { ...definition.attributes, ...elements.attributes };
      } else {
        definition.attributes = elements.attributes;
      }
    }
    if (elements.commands) {
      if (definition.commands) {
        definition.commands = { ...definition.commands, ...elements.commands };
      } else {
        definition.commands = elements.commands;
      }
    }
    if (elements.events) {
      if (definition.events) {
        definition.events = { ...definition.events, ...elements.events };
      } else {
        definition.events = elements.events;
      }
    }
  }
  ClusterFactory2.extend = extend;
  function validateFeatureSelection(features, validFeatures) {
    for (const f of features) {
      if (!validFeatures[f]) {
        throw new IllegalClusterError(`"${f}" is not a valid feature identifier`);
      }
    }
  }
  ClusterFactory2.validateFeatureSelection = validateFeatureSelection;
  function prevent(definition, ...illegalFeatureCombinations) {
    pool:
      for (const bitmap of illegalFeatureCombinations) {
        for (const k in bitmap) {
          if (!!definition.supportedFeatures[k] !== !!bitmap[k]) {
            continue pool;
          }
        }
        throw new IllegalClusterError(
          `Feature combination ${(0, import_String.serialize)(bitmap)} is disallowed by the Matter specification`
        );
      }
  }
  ClusterFactory2.prevent = prevent;
  function AsConditional(element, { optionalIf, mandatoryIf }) {
    return {
      ...element,
      optional: true,
      isConditional: true,
      optionalIf: optionalIf ?? [],
      mandatoryIf: mandatoryIf ?? []
    };
  }
  ClusterFactory2.AsConditional = AsConditional;
})(ClusterFactory || (ClusterFactory = {}));
