// @project-chip/matter.js/dist/cjs/util/Cache.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Cache_exports = {};
__export(Cache_exports, {
  Cache: () => Cache
});
module.exports = __toCommonJS(Cache_exports);
var import_Time = require("../time/Time.js");
/**
 * Cache computed values or resources for a specified duration to improve performances.
 *
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class Cache {
  constructor(generator, expirationMs, expireCallback) {
    this.generator = generator;
    this.expirationMs = expirationMs;
    this.expireCallback = expireCallback;
    this.knownKeys = /* @__PURE__ */ new Set();
    this.values = /* @__PURE__ */ new Map();
    this.timestamps = /* @__PURE__ */ new Map();
    this.periodicTimer = import_Time.Time.getPeriodicTimer(expirationMs, () => this.expire()).start();
  }
  get(...params) {
    const key = params.join(",");
    let value = this.values.get(key);
    if (value === void 0) {
      value = this.generator(...params);
      this.values.set(key, value);
      this.knownKeys.add(key);
    }
    this.timestamps.set(key, import_Time.Time.nowMs());
    return value;
  }
  keys() {
    return Array.from(this.knownKeys.values());
  }
  async deleteEntry(key) {
    const value = this.values.get(key);
    if (this.expireCallback !== void 0 && value !== void 0) {
      await this.expireCallback(key, value);
    }
    this.values.delete(key);
    this.timestamps.delete(key);
  }
  async clear() {
    for (const key of this.values.keys()) {
      await this.deleteEntry(key);
    }
    this.values.clear();
    this.timestamps.clear();
  }
  async close() {
    await this.clear();
    this.knownKeys.clear();
    this.periodicTimer.stop();
  }
  async expire() {
    const now = import_Time.Time.nowMs();
    for (const [key, timestamp] of this.timestamps.entries()) {
      if (now - timestamp < this.expirationMs)
        continue;
      await this.deleteEntry(key);
    }
  }
}
