// @project-chip/matter.js/dist/esm/cluster/definitions/ApplicationLauncherCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { OptionalAttribute, Command, Attribute } from "../../cluster/Cluster.js";
import { TlvObject, TlvField, TlvOptionalField } from "../../tlv/TlvObject.js";
import { TlvUInt16, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvString, TlvByteString } from "../../tlv/TlvString.js";
import { TlvEndpointNumber } from "../../datatype/EndpointNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvArray } from "../../tlv/TlvArray.js";
var ApplicationLauncher;
((ApplicationLauncher2) => {
  ApplicationLauncher2.TlvApplicationStruct = TlvObject({
    /**
     * This shall indicate the CSA-issued vendor ID for the catalog. The DIAL registry shall use value 0x0000.
     *
     * Content App Platform providers will have their own catalog vendor ID (set to their own Vendor ID) and will
     * assign an ApplicationID to each Content App.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.5.2.1
     */
    catalogVendorId: TlvField(0, TlvUInt16),
    /**
     * This shall indicate the application identifier, expressed as a string, such as "PruneVideo" or "Company X".
     * This field shall be unique within a catalog.
     *
     * For the DIAL registry catalog, this value shall be the DIAL prefix (see [DIAL Registry]).
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.5.2.2
     */
    applicationId: TlvField(1, TlvString)
  });
  ApplicationLauncher2.TlvApplicationEPStruct = TlvObject({
    application: TlvField(0, ApplicationLauncher2.TlvApplicationStruct),
    endpoint: TlvOptionalField(1, TlvEndpointNumber)
  });
  ApplicationLauncher2.TlvLaunchAppRequest = TlvObject({
    /**
     * This field shall specify the Application to launch.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.4.1.1
     */
    application: TlvOptionalField(0, ApplicationLauncher2.TlvApplicationStruct),
    /**
     * This field shall specify optional app-specific data to be sent to the app.
     *
     * Note: This format and meaning of this value is proprietary and outside the specification. It provides a
     * transition path for device makers that use other protocols (like DIAL) which allow for proprietary data.
     * Apps that are not yet Matter aware can be launched via Matter, while retaining the existing ability to
     * launch with proprietary data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.4.1.2
     */
    data: TlvOptionalField(1, TlvByteString)
  });
  let Status;
  ((Status2) => {
    Status2[Status2["Success"] = 0] = "Success";
    Status2[Status2["AppNotAvailable"] = 1] = "AppNotAvailable";
    Status2[Status2["SystemBusy"] = 2] = "SystemBusy";
  })(Status = ApplicationLauncher2.Status || (ApplicationLauncher2.Status = {}));
  ApplicationLauncher2.TlvLauncherResponse = TlvObject({
    /**
     * This shall indicate the status of the command which resulted in this response.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.4.4.1
     */
    status: TlvField(0, TlvEnum()),
    /**
     * This shall specify Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.4.4.2
     */
    data: TlvOptionalField(1, TlvByteString)
  });
  ApplicationLauncher2.TlvStopAppRequest = TlvObject({
    /**
     * This field shall specify the Application to stop.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.4.2.1
     */
    application: TlvOptionalField(0, ApplicationLauncher2.TlvApplicationStruct)
  });
  ApplicationLauncher2.TlvHideAppRequest = TlvObject({
    /**
     * This field shall specify the Application to hide.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.4.3.1
     */
    application: TlvOptionalField(0, ApplicationLauncher2.TlvApplicationStruct)
  });
  let Feature;
  ((Feature2) => {
    Feature2["ApplicationPlatform"] = "ApplicationPlatform";
  })(Feature = ApplicationLauncher2.Feature || (ApplicationLauncher2.Feature = {}));
  ApplicationLauncher2.Base = ClusterFactory.Definition({
    id: 1292,
    name: "ApplicationLauncher",
    revision: 1,
    features: {
      /**
       * ApplicationPlatform
       *
       * Support for attributes and commands required for endpoint to support launching any application within
       * the supported application catalogs
       */
      applicationPlatform: BitFlag(0)
    },
    attributes: {
      /**
       * This attribute shall specify the current in-focus application, identified using an Application ID,
       * catalog vendor ID and the corresponding endpoint number when the application is represented by a Content
       * App endpoint. A null shall be used to indicate there is no current in-focus application.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.3.2
       */
      currentApp: OptionalAttribute(1, TlvNullable(ApplicationLauncher2.TlvApplicationEPStruct), { default: null })
    },
    commands: {
      /**
       * Upon receipt of this command, the server shall launch the application with optional data. The
       * application shall be either
       *
       *   • the specified application, if the Application Platform feature is supported;
       *
       *   • otherwise the application corresponding to the endpoint.
       *
       * The endpoint shall launch and bring to foreground the requisite application if the application is not
       * already launched and in foreground. The Status attribute shall be updated to ACTIVE_VISIBLE_FOCUS on the
       * Application Basic cluster of the Endpoint corresponding to the launched application. The Status
       * attribute shall be updated on any other application whose Status may have changed as a result of this
       * command. The CurrentApp attribute, if supported, shall be updated to reflect the new application in the
       * foreground.
       *
       * This command returns a Launcher Response.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.4.1
       */
      launchApp: Command(0, ApplicationLauncher2.TlvLaunchAppRequest, 3, ApplicationLauncher2.TlvLauncherResponse),
      /**
       * Upon receipt of this command, the server shall stop the application if it is running. The application
       * shall be either
       *
       *   • the specified application, if the Application Platform feature is supported;
       *
       *   • otherwise the application corresponding to the endpoint.
       *
       * The Status attribute shall be updated to STOPPED on the Application Basic cluster of the Endpoint
       * corresponding to the stopped application. The Status attribute shall be updated on any other application
       * whose Status may have changed as a result of this command.
       *
       * This command returns a Launcher Response.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.4.2
       */
      stopApp: Command(1, ApplicationLauncher2.TlvStopAppRequest, 3, ApplicationLauncher2.TlvLauncherResponse),
      /**
       * Upon receipt of this command, the server shall hide the application. The application shall be either
       *
       *   • the specified application, if the Application Platform feature is supported;
       *
       *   • otherwise the application corresponding to the endpoint.
       *
       * The endpoint may decide to stop the application based on manufacturer specific behavior or resource
       * constraints if any. The Status attribute shall be updated to ACTIVE_HIDDEN or STOPPED, depending on the
       * action taken, on the Application Basic cluster of the Endpoint corresponding to the application on which
       * the action was taken. The Status attribute shall be updated on any other application whose Status may
       * have changed as a result of this command.
       *
       * This command returns a Launcher Response.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.4.3
       */
      hideApp: Command(2, ApplicationLauncher2.TlvHideAppRequest, 3, ApplicationLauncher2.TlvLauncherResponse)
    }
  });
  ApplicationLauncher2.ApplicationPlatformComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This attribute shall specify the list of supported application catalogs, where each entry in the list is
       * the CSA-issued vendor ID for the catalog. The DIAL registry (see [DIAL Registry]) shall use value 0x0000.
       *
       * It is expected that Content App Platform providers will have their own catalog vendor ID (set to their
       * own Vendor ID) and will assign an ApplicationID to each Content App.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.4.3.1
       */
      catalogList: Attribute(0, TlvArray(TlvUInt16), { persistent: true })
    }
  });
  ApplicationLauncher2.Cluster = ClusterFactory.Extensible(
    ApplicationLauncher2.Base,
    /**
     * Use this factory method to create an ApplicationLauncher cluster with support for optional features. Include
     * each {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns an ApplicationLauncher cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...ApplicationLauncher2.Base,
        supportedFeatures: BitFlags(ApplicationLauncher2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, ApplicationLauncher2.ApplicationPlatformComponent, { applicationPlatform: true });
      return cluster;
    }
  );
  const AP = { applicationPlatform: true };
  ApplicationLauncher2.Complete = ClusterFactory.Definition({
    id: ApplicationLauncher2.Cluster.id,
    name: ApplicationLauncher2.Cluster.name,
    revision: ApplicationLauncher2.Cluster.revision,
    features: ApplicationLauncher2.Cluster.features,
    attributes: {
      ...ApplicationLauncher2.Cluster.attributes,
      catalogList: ClusterFactory.AsConditional(
        ApplicationLauncher2.ApplicationPlatformComponent.attributes.catalogList,
        { mandatoryIf: [AP] }
      )
    },
    commands: ApplicationLauncher2.Cluster.commands
  });
})(ApplicationLauncher || (ApplicationLauncher = {}));
const ApplicationLauncherCluster = ApplicationLauncher.Cluster;
export {
  ApplicationLauncher,
  ApplicationLauncherCluster
};
