// @project-chip/matter.js/dist/esm/cluster/definitions/AudioOutputCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { Attribute, Command, TlvNoResponse, AccessLevel } from "../../cluster/Cluster.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvUInt8, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvString } from "../../tlv/TlvString.js";
var AudioOutput;
((AudioOutput2) => {
  let OutputType;
  ((OutputType2) => {
    OutputType2[OutputType2["Hdmi"] = 0] = "Hdmi";
    OutputType2[OutputType2["Bt"] = 1] = "Bt";
    OutputType2[OutputType2["Optical"] = 2] = "Optical";
    OutputType2[OutputType2["Headphone"] = 3] = "Headphone";
    OutputType2[OutputType2["Internal"] = 4] = "Internal";
    OutputType2[OutputType2["Other"] = 5] = "Other";
  })(OutputType = AudioOutput2.OutputType || (AudioOutput2.OutputType = {}));
  AudioOutput2.TlvOutputInfoStruct = TlvObject({
    /**
     * This shall indicate the unique index into the list of outputs.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.5.1.1
     */
    index: TlvField(0, TlvUInt8),
    /**
     * This shall indicate the type of output
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.5.1.2
     */
    outputType: TlvField(1, TlvEnum()),
    /**
     * The device defined and user editable output name, such as “Soundbar”, “Speakers”. This field may be blank,
     * but SHOULD be provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.5.1.3
     */
    name: TlvField(2, TlvString.bound({ maxLength: 32 }))
  });
  AudioOutput2.TlvSelectOutputRequest = TlvObject({ index: TlvField(0, TlvUInt8) });
  AudioOutput2.TlvRenameOutputRequest = TlvObject({ index: TlvField(0, TlvUInt8), name: TlvField(1, TlvString) });
  let Feature;
  ((Feature2) => {
    Feature2["NameUpdates"] = "NameUpdates";
  })(Feature = AudioOutput2.Feature || (AudioOutput2.Feature = {}));
  AudioOutput2.Base = ClusterFactory.Definition({
    id: 1291,
    name: "AudioOutput",
    revision: 1,
    features: {
      /**
       * NameUpdates
       *
       * Supports updates to output names
       */
      nameUpdates: BitFlag(0)
    },
    attributes: {
      /**
       * This list provides the outputs supported by the device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.3.1
       */
      outputList: Attribute(0, TlvArray(AudioOutput2.TlvOutputInfoStruct)),
      /**
       * This field contains the value of the index field of the currently selected OutputInfoStruct.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.3.2
       */
      currentOutput: Attribute(1, TlvUInt8, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.4
       */
      selectOutput: Command(0, AudioOutput2.TlvSelectOutputRequest, 0, TlvNoResponse)
    }
  });
  AudioOutput2.NameUpdatesComponent = ClusterFactory.Component({
    commands: {
      /**
       * Upon receipt, this shall rename the output at a specific index in the Output List.
       *
       * Updates to the output name shall appear in the device’s settings menus. Name updates may automatically
       * be sent to the actual device to which the output connects.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.5.4.2
       */
      renameOutput: Command(1, AudioOutput2.TlvRenameOutputRequest, 1, TlvNoResponse, { invokeAcl: AccessLevel.Manage })
    }
  });
  AudioOutput2.Cluster = ClusterFactory.Extensible(
    AudioOutput2.Base,
    /**
     * Use this factory method to create an AudioOutput cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns an AudioOutput cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...AudioOutput2.Base,
        supportedFeatures: BitFlags(AudioOutput2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, AudioOutput2.NameUpdatesComponent, { nameUpdates: true });
      return cluster;
    }
  );
  const NU = { nameUpdates: true };
  AudioOutput2.Complete = ClusterFactory.Definition({
    id: AudioOutput2.Cluster.id,
    name: AudioOutput2.Cluster.name,
    revision: AudioOutput2.Cluster.revision,
    features: AudioOutput2.Cluster.features,
    attributes: AudioOutput2.Cluster.attributes,
    commands: {
      ...AudioOutput2.Cluster.commands,
      renameOutput: ClusterFactory.AsConditional(
        AudioOutput2.NameUpdatesComponent.commands.renameOutput,
        { mandatoryIf: [NU] }
      )
    }
  });
})(AudioOutput || (AudioOutput = {}));
const AudioOutputCluster = AudioOutput.Cluster;
export {
  AudioOutput,
  AudioOutputCluster
};
