// @project-chip/matter.js/dist/esm/cluster/definitions/BarrierControlCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { Attribute, OptionalWritableAttribute, Command, TlvNoResponse } from "../../cluster/Cluster.js";
import { TlvUInt8, TlvUInt16 } from "../../tlv/TlvNumber.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var BarrierControl;
((BarrierControl2) => {
  BarrierControl2.TlvBarrierControlGoToPercentRequest = TlvObject({ percentOpen: TlvField(0, TlvUInt8) });
  BarrierControl2.Cluster = ClusterFactory.Definition({
    id: 259,
    name: "BarrierControl",
    revision: 1,
    attributes: {
      barrierMovingState: Attribute(1, TlvUInt8),
      barrierSafetyStatus: Attribute(2, TlvUInt16),
      barrierCapabilities: Attribute(3, TlvUInt8),
      barrierOpenEvents: OptionalWritableAttribute(4, TlvUInt16.bound({ max: 65534 }), { default: 0 }),
      barrierCloseEvents: OptionalWritableAttribute(5, TlvUInt16.bound({ max: 65534 }), { default: 0 }),
      barrierCommandOpenEvents: OptionalWritableAttribute(6, TlvUInt16.bound({ max: 65534 }), { default: 0 }),
      barrierCommandCloseEvents: OptionalWritableAttribute(7, TlvUInt16.bound({ max: 65534 }), { default: 0 }),
      barrierOpenPeriod: OptionalWritableAttribute(8, TlvUInt16.bound({ max: 65534 })),
      barrierClosePeriod: OptionalWritableAttribute(9, TlvUInt16.bound({ max: 65534 })),
      barrierPosition: Attribute(10, TlvUInt8)
    },
    commands: {
      barrierControlGoToPercent: Command(0, BarrierControl2.TlvBarrierControlGoToPercentRequest, 0, TlvNoResponse),
      barrierControlStop: Command(1, TlvNoArguments, 1, TlvNoResponse)
    }
  });
})(BarrierControl || (BarrierControl = {}));
const BarrierControlCluster = BarrierControl.Cluster;
export {
  BarrierControl,
  BarrierControlCluster
};
