// @project-chip/matter.js/dist/esm/cluster/definitions/BinaryInputBasicCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { OptionalWritableAttribute, WritableAttribute, OptionalAttribute, Attribute } from "../../cluster/Cluster.js";
import { TlvString } from "../../tlv/TlvString.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
import { TlvUInt8, TlvUInt32 } from "../../tlv/TlvNumber.js";
var BinaryInputBasic;
((BinaryInputBasic2) => {
  BinaryInputBasic2.Cluster = ClusterFactory.Definition({
    id: 15,
    name: "BinaryInputBasic",
    revision: 1,
    attributes: {
      activeText: OptionalWritableAttribute(4, TlvString.bound({ maxLength: 16 }), { default: "" }),
      description: OptionalWritableAttribute(28, TlvString.bound({ maxLength: 16 }), { default: "" }),
      inactiveText: OptionalWritableAttribute(46, TlvString.bound({ maxLength: 16 }), { default: "" }),
      outOfService: WritableAttribute(81, TlvBoolean, { default: true }),
      polarity: OptionalAttribute(84, TlvUInt8),
      presentValue: WritableAttribute(85, TlvBoolean),
      reliability: OptionalWritableAttribute(103, TlvUInt8),
      statusFlags: Attribute(111, TlvUInt8),
      applicationType: OptionalAttribute(256, TlvUInt32)
    }
  });
})(BinaryInputBasic || (BinaryInputBasic = {}));
const BinaryInputBasicCluster = BinaryInputBasic.Cluster;
export {
  BinaryInputBasic,
  BinaryInputBasicCluster
};
