// @project-chip/matter.js/dist/esm/cluster/definitions/BooleanStateCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { Attribute, OptionalEvent, EventPriority } from "../../cluster/Cluster.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
var BooleanState;
((BooleanState2) => {
  BooleanState2.TlvStateChangeEvent = TlvObject({ stateValue: TlvField(0, TlvBoolean) });
  BooleanState2.Cluster = ClusterFactory.Definition({
    id: 69,
    name: "BooleanState",
    revision: 1,
    attributes: {
      /**
       * This represents a Boolean state.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.7.4.1
       */
      stateValue: Attribute(0, TlvBoolean)
    },
    events: {
      /**
       * This event shall be generated when the StateValue attribute changes.
       *
       * The StateValue field shall indicate the new value of the StateValue attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.7.5.1
       */
      stateChange: OptionalEvent(0, EventPriority.Info, BooleanState2.TlvStateChangeEvent)
    }
  });
})(BooleanState || (BooleanState = {}));
const BooleanStateCluster = BooleanState.Cluster;
export {
  BooleanState,
  BooleanStateCluster
};
