// @project-chip/matter.js/dist/esm/cluster/definitions/BridgedDeviceBasicInformationCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import {
  OptionalFixedAttribute,
  OptionalWritableAttribute,
  AccessLevel,
  Attribute,
  OptionalEvent,
  EventPriority,
  Event
} from "../../cluster/Cluster.js";
import { TlvString } from "../../tlv/TlvString.js";
import { TlvVendorId } from "../../datatype/VendorId.js";
import { TlvUInt16, TlvUInt32, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var BridgedDeviceBasicInformation;
((BridgedDeviceBasicInformation2) => {
  let ProductFinish;
  ((ProductFinish2) => {
    ProductFinish2[ProductFinish2["Other"] = 0] = "Other";
    ProductFinish2[ProductFinish2["Matte"] = 1] = "Matte";
    ProductFinish2[ProductFinish2["Satin"] = 2] = "Satin";
    ProductFinish2[ProductFinish2["Polished"] = 3] = "Polished";
    ProductFinish2[ProductFinish2["Rugged"] = 4] = "Rugged";
    ProductFinish2[ProductFinish2["Fabric"] = 5] = "Fabric";
  })(ProductFinish = BridgedDeviceBasicInformation2.ProductFinish || (BridgedDeviceBasicInformation2.ProductFinish = {}));
  let Color;
  ((Color2) => {
    Color2[Color2["Black"] = 0] = "Black";
    Color2[Color2["Navy"] = 1] = "Navy";
    Color2[Color2["Green"] = 2] = "Green";
    Color2[Color2["Teal"] = 3] = "Teal";
    Color2[Color2["Maroon"] = 4] = "Maroon";
    Color2[Color2["Purple"] = 5] = "Purple";
    Color2[Color2["Olive"] = 6] = "Olive";
    Color2[Color2["Gray"] = 7] = "Gray";
    Color2[Color2["Blue"] = 8] = "Blue";
    Color2[Color2["Lime"] = 9] = "Lime";
    Color2[Color2["Aqua"] = 10] = "Aqua";
    Color2[Color2["Red"] = 11] = "Red";
    Color2[Color2["Fuchsia"] = 12] = "Fuchsia";
    Color2[Color2["Yellow"] = 13] = "Yellow";
    Color2[Color2["White"] = 14] = "White";
    Color2[Color2["Nickel"] = 15] = "Nickel";
    Color2[Color2["Chrome"] = 16] = "Chrome";
    Color2[Color2["Brass"] = 17] = "Brass";
    Color2[Color2["Copper"] = 18] = "Copper";
    Color2[Color2["Silver"] = 19] = "Silver";
    Color2[Color2["Gold"] = 20] = "Gold";
  })(Color = BridgedDeviceBasicInformation2.Color || (BridgedDeviceBasicInformation2.Color = {}));
  BridgedDeviceBasicInformation2.TlvProductAppearanceStruct = TlvObject({
    finish: TlvField(0, TlvEnum()),
    primaryColor: TlvField(1, TlvNullable(TlvEnum()))
  });
  BridgedDeviceBasicInformation2.TlvStartUpEvent = TlvObject({ softwareVersion: TlvField(0, TlvUInt32) });
  BridgedDeviceBasicInformation2.TlvReachableChangedEvent = TlvObject({ reachableNewValue: TlvField(0, TlvBoolean) });
  BridgedDeviceBasicInformation2.Cluster = ClusterFactory.Definition({
    id: 57,
    name: "BridgedDeviceBasicInformation",
    revision: 2,
    attributes: {
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      vendorName: OptionalFixedAttribute(1, TlvString.bound({ maxLength: 32 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      vendorId: OptionalFixedAttribute(2, TlvVendorId),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      productName: OptionalFixedAttribute(3, TlvString.bound({ maxLength: 32 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      nodeLabel: OptionalWritableAttribute(
        5,
        TlvString.bound({ maxLength: 32 }),
        { persistent: true, default: "", writeAcl: AccessLevel.Manage }
      ),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      hardwareVersion: OptionalFixedAttribute(7, TlvUInt16, { default: 0 }),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      hardwareVersionString: OptionalFixedAttribute(8, TlvString.bound({ minLength: 1, maxLength: 64 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      softwareVersion: OptionalFixedAttribute(9, TlvUInt32, { default: 0 }),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      softwareVersionString: OptionalFixedAttribute(10, TlvString.bound({ minLength: 1, maxLength: 64 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      manufacturingDate: OptionalFixedAttribute(11, TlvString.bound({ minLength: 8, maxLength: 16 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      partNumber: OptionalFixedAttribute(12, TlvString.bound({ maxLength: 32 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      productUrl: OptionalFixedAttribute(13, TlvString.bound({ maxLength: 256 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      productLabel: OptionalFixedAttribute(14, TlvString.bound({ maxLength: 64 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      serialNumber: OptionalFixedAttribute(15, TlvString.bound({ maxLength: 32 })),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      reachable: Attribute(17, TlvBoolean, { default: true }),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.4
       */
      uniqueId: OptionalFixedAttribute(18, TlvString.bound({ maxLength: 32 })),
      productAppearance: OptionalFixedAttribute(20, BridgedDeviceBasicInformation2.TlvProductAppearanceStruct)
    },
    events: {
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.5
       */
      startUp: OptionalEvent(0, EventPriority.Critical, BridgedDeviceBasicInformation2.TlvStartUpEvent),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.5
       */
      shutDown: OptionalEvent(1, EventPriority.Critical, TlvNoArguments),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.5
       */
      leave: OptionalEvent(2, EventPriority.Info, TlvNoArguments),
      /**
       * This event shall be generated when there is a change in the Reachable attribute. Its purpose is to
       * provide an indication towards interested parties that the reachability of a bridged device (over the
       * non-Matter network) has changed, so they may take appropriate action.
       *
       * After (re)start of a bridge this event may be generated.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.13.5.1
       */
      reachableChanged: Event(3, EventPriority.Info, BridgedDeviceBasicInformation2.TlvReachableChangedEvent)
    }
  });
})(BridgedDeviceBasicInformation || (BridgedDeviceBasicInformation = {}));
const BridgedDeviceBasicInformationCluster = BridgedDeviceBasicInformation.Cluster;
export {
  BridgedDeviceBasicInformation,
  BridgedDeviceBasicInformationCluster
};
