// @project-chip/matter.js/dist/esm/cluster/definitions/ChannelCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { OptionalAttribute, Command, TlvNoResponse, Attribute } from "../../cluster/Cluster.js";
import { TlvObject, TlvField, TlvOptionalField } from "../../tlv/TlvObject.js";
import { TlvUInt16, TlvInt16, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvString, TlvByteString } from "../../tlv/TlvString.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvArray } from "../../tlv/TlvArray.js";
var Channel;
((Channel2) => {
  Channel2.TlvChannelInfoStruct = TlvObject({
    /**
     * This shall indicate the channel major number value (for example, using ATSC format). When the channel number
     * is expressed as a string, such as "13.1" or "256", the major number would be 13 or 256, respectively.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.1.1
     */
    majorNumber: TlvField(0, TlvUInt16),
    /**
     * This shall indicate the channel minor number value (for example, using ATSC format). When the channel number
     * is expressed as a string, such as "13.1" or "256", the minor number would be 1 or 0, respectively.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.1.2
     */
    minorNumber: TlvField(1, TlvUInt16),
    /**
     * This shall indicate the marketing name for the channel, such as “The CW" or "Comedy Central". This field is
     * optional, but SHOULD be provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.1.3
     */
    name: TlvOptionalField(2, TlvString),
    /**
     * This shall indicate the call sign of the channel, such as "PBS". This field is optional, but SHOULD be
     * provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.1.4
     */
    callSign: TlvOptionalField(3, TlvString),
    /**
     * This shall indicate the local affiliate call sign, such as "KCTS". This field is optional, but SHOULD be
     * provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.1.5
     */
    affiliateCallSign: TlvOptionalField(4, TlvString)
  });
  Channel2.TlvChangeChannelByNumberRequest = TlvObject({
    /**
     * This shall indicate the channel major number value (ATSC format) to which the channel should change.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.3.1
     */
    majorNumber: TlvField(0, TlvUInt16),
    /**
     * This shall indicate the channel minor number value (ATSC format) to which the channel should change.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.3.2
     */
    minorNumber: TlvField(1, TlvUInt16)
  });
  Channel2.TlvSkipChannelRequest = TlvObject({
    /**
     * This shall indicate the number of steps to increase (Count is positive) or decrease (Count is negative) the
     * current channel.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.4.1
     */
    count: TlvField(0, TlvInt16)
  });
  let LineupInfoType;
  ((LineupInfoType2) => {
    LineupInfoType2[LineupInfoType2["Mso"] = 0] = "Mso";
  })(LineupInfoType = Channel2.LineupInfoType || (Channel2.LineupInfoType = {}));
  Channel2.TlvLineupInfoStruct = TlvObject({
    /**
     * This shall indicate the name of the operator, for example “Comcast”.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.2.1
     */
    operatorName: TlvField(0, TlvString),
    lineupName: TlvOptionalField(1, TlvString),
    postalCode: TlvOptionalField(2, TlvString),
    /**
     * This shall indicate the type of lineup. This field is optional, but SHOULD be provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.5.2.4
     */
    lineupInfoType: TlvField(3, TlvEnum())
  });
  Channel2.TlvChangeChannelRequest = TlvObject({
    /**
     * This shall contain a user-input string to match in order to identify the target channel.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.1.1
     */
    match: TlvField(0, TlvString)
  });
  let Status;
  ((Status2) => {
    Status2[Status2["Success"] = 0] = "Success";
    Status2[Status2["MultipleMatches"] = 1] = "MultipleMatches";
    Status2[Status2["NoMatches"] = 2] = "NoMatches";
  })(Status = Channel2.Status || (Channel2.Status = {}));
  Channel2.TlvChangeChannelResponse = TlvObject({
    /**
     * This shall indicate the status of the command which resulted in this response.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.2.1
     */
    status: TlvField(0, TlvEnum()),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.2.2
     */
    data: TlvOptionalField(1, TlvByteString)
  });
  let Feature;
  ((Feature2) => {
    Feature2["ChannelList"] = "ChannelList";
    Feature2["LineupInfo"] = "LineupInfo";
  })(Feature = Channel2.Feature || (Channel2.Feature = {}));
  Channel2.Base = ClusterFactory.Definition({
    id: 1284,
    name: "Channel",
    revision: 1,
    features: {
      /**
       * ChannelList
       *
       * Provides list of available channels.
       */
      channelList: BitFlag(0),
      /**
       * LineupInfo
       *
       * Provides lineup info, which is a reference to an external source of lineup information.
       */
      lineupInfo: BitFlag(1)
    },
    attributes: {
      /**
       * This optional field contains the current channel. When supported but a channel is not currently tuned to
       * (if a content application is in foreground), the value of the field shall be null.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.3.3
       */
      currentChannel: OptionalAttribute(2, TlvNullable(Channel2.TlvChannelInfoStruct), { default: null })
    },
    commands: {
      /**
       * Change the channel to the channel with the given Number in the ChannelList attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.3
       */
      changeChannelByNumber: Command(2, Channel2.TlvChangeChannelByNumberRequest, 2, TlvNoResponse),
      /**
       * This command provides channel up and channel down functionality, but allows channel index jumps of size
       * Count.
       *
       * When the value of the increase or decrease is larger than the number of channels remaining in the given
       * direction, then the behavior shall be to return to the beginning (or end) of the channel list and
       * continue. For example, if the current channel is at index 0 and count value of -1 is given, then the
       * current channel should change to the last channel.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.4
       */
      skipChannel: Command(3, Channel2.TlvSkipChannelRequest, 3, TlvNoResponse)
    }
  });
  Channel2.ChannelListComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This optional list provides the channels supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.3.1
       */
      channelList: Attribute(0, TlvArray(Channel2.TlvChannelInfoStruct), { default: [] })
    }
  });
  Channel2.LineupInfoComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This optional field identifies the channel lineup using external data sources.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.3.2
       */
      lineup: Attribute(1, TlvNullable(Channel2.TlvLineupInfoStruct), { default: null })
    }
  });
  Channel2.ChannelListOrLineupInfoComponent = ClusterFactory.Component({
    commands: {
      /**
       * Change the channel to the channel case-insensitive exact matching the value passed as an argument.
       *
       * The match priority order shall be: AffiliateCallSign ("KCTS"), CallSign ("PBS"), Name ("Comedy
       * Central"), Number ("13.1")
       *
       * Upon receipt, this shall generate a ChangeChannelResponse command.
       *
       * Upon success, the CurrentChannel attribute, if supported, shall be updated to reflect the change.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.6.4.1
       */
      changeChannel: Command(0, Channel2.TlvChangeChannelRequest, 1, Channel2.TlvChangeChannelResponse)
    }
  });
  Channel2.Cluster = ClusterFactory.Extensible(
    Channel2.Base,
    /**
     * Use this factory method to create a Channel cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a Channel cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...Channel2.Base,
        supportedFeatures: BitFlags(Channel2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, Channel2.ChannelListComponent, { channelList: true });
      ClusterFactory.extend(cluster, Channel2.LineupInfoComponent, { lineupInfo: true });
      ClusterFactory.extend(
        cluster,
        Channel2.ChannelListOrLineupInfoComponent,
        { channelList: true },
        { lineupInfo: true }
      );
      return cluster;
    }
  );
  const CL = { channelList: true };
  const LI = { lineupInfo: true };
  Channel2.Complete = ClusterFactory.Definition({
    id: Channel2.Cluster.id,
    name: Channel2.Cluster.name,
    revision: Channel2.Cluster.revision,
    features: Channel2.Cluster.features,
    attributes: {
      ...Channel2.Cluster.attributes,
      channelList: ClusterFactory.AsConditional(
        Channel2.ChannelListComponent.attributes.channelList,
        { mandatoryIf: [CL] }
      ),
      lineup: ClusterFactory.AsConditional(Channel2.LineupInfoComponent.attributes.lineup, { mandatoryIf: [LI] })
    },
    commands: {
      ...Channel2.Cluster.commands,
      changeChannel: ClusterFactory.AsConditional(
        Channel2.ChannelListOrLineupInfoComponent.commands.changeChannel,
        { mandatoryIf: [CL, LI] }
      )
    }
  });
})(Channel || (Channel = {}));
const ChannelCluster = Channel.Cluster;
export {
  Channel,
  ChannelCluster
};
