// @project-chip/matter.js/dist/esm/cluster/definitions/ClientMonitoringCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { Attribute, Command, TlvNoResponse, AccessLevel, OptionalCommand } from "../../cluster/Cluster.js";
import { TlvUInt32, TlvUInt16, TlvUInt64 } from "../../tlv/TlvNumber.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvNodeId } from "../../datatype/NodeId.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var ClientMonitoring;
((ClientMonitoring2) => {
  ClientMonitoring2.TlvMonitoringRegistration = TlvObject({
    clientNodeId: TlvField(1, TlvNodeId),
    iCid: TlvField(2, TlvUInt64)
  });
  ClientMonitoring2.TlvRegisterClientMonitoringRequest = TlvObject({
    clientNodeId: TlvField(0, TlvNodeId),
    iCid: TlvField(1, TlvUInt64)
  });
  ClientMonitoring2.TlvUnregisterClientMonitoringRequest = TlvObject({
    clientNodeId: TlvField(0, TlvNodeId),
    iCid: TlvField(1, TlvUInt64)
  });
  ClientMonitoring2.Cluster = ClusterFactory.Definition({
    id: 4166,
    name: "ClientMonitoring",
    revision: 1,
    attributes: {
      idleModeInterval: Attribute(0, TlvUInt32.bound({ min: 300, max: 864e5 }), { default: 18 }),
      activeModeInterval: Attribute(1, TlvUInt32.bound({ min: 300, max: 864e5 }), { default: 18 }),
      activeModeThreshold: Attribute(2, TlvUInt16.bound({ min: 300, max: 6e4 }), { default: 0 }),
      expectedClients: Attribute(3, TlvArray(ClientMonitoring2.TlvMonitoringRegistration, { maxLength: 4 }), { default: [] })
    },
    commands: {
      registerClientMonitoring: Command(
        0,
        ClientMonitoring2.TlvRegisterClientMonitoringRequest,
        0,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Manage }
      ),
      unregisterClientMonitoring: Command(
        1,
        ClientMonitoring2.TlvUnregisterClientMonitoringRequest,
        1,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Manage }
      ),
      stayAwakeRequest: OptionalCommand(
        2,
        TlvNoArguments,
        2,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Manage }
      )
    }
  });
})(ClientMonitoring || (ClientMonitoring = {}));
const ClientMonitoringCluster = ClientMonitoring.Cluster;
export {
  ClientMonitoring,
  ClientMonitoringCluster
};
