// @project-chip/matter.js/dist/esm/cluster/definitions/ContentLauncherCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { Attribute, Command } from "../../cluster/Cluster.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvString, TlvByteString } from "../../tlv/TlvString.js";
import { TlvUInt32, TlvBitmap, TlvDouble, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvObject, TlvField, TlvOptionalField } from "../../tlv/TlvObject.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
var ContentLauncher;
((ContentLauncher2) => {
  ContentLauncher2.SupportedStreamingProtocols = {
    /**
     * Device supports Dynamic Adaptive Streaming over HTTP (DASH)
     */
    dash: BitFlag(0),
    /**
     * Device supports HTTP Live Streaming (HLS)
     */
    hls: BitFlag(1)
  };
  let MetricType;
  ((MetricType2) => {
    MetricType2[MetricType2["Pixels"] = 0] = "Pixels";
    MetricType2[MetricType2["Percentage"] = 1] = "Percentage";
  })(MetricType = ContentLauncher2.MetricType || (ContentLauncher2.MetricType = {}));
  ContentLauncher2.TlvDimensionStruct = TlvObject({
    /**
     * This indicates the width using the metric defined in Metric
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.8.1
     */
    width: TlvField(0, TlvDouble),
    /**
     * This indicates the height using the metric defined in Metric
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.8.2
     */
    height: TlvField(1, TlvDouble),
    /**
     * This shall indicate metric used for defining Height/Width.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.8.3
     */
    metric: TlvField(2, TlvEnum())
  });
  ContentLauncher2.TlvStyleInformationStruct = TlvObject({
    /**
     * This shall indicate the URL of image used for Styling different Video Player sections like Logo, Watermark
     * etc.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.7.1
     */
    imageUrl: TlvOptionalField(0, TlvString.bound({ maxLength: 8192 })),
    /**
     * This shall indicate the color, in RGB or RGBA, used for styling different Video Player sections like Logo,
     * Watermark, etc. The value shall conform to the 6-digit or 8-digit format defined for CSS sRGB hexadecimal
     * color notation [https://www.w3.org/TR/css-color-4/#hex-notation]. Examples:
     *
     *   • #76DE19 for R=0x76, G=0xDE, B=0x19, A absent
     *
     *   • #76DE1980 for R=0x76, G=0xDE, B=0x19, A=0x80
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.7.2
     */
    color: TlvOptionalField(1, TlvString),
    /**
     * This shall indicate the size of the image used for Styling different Video Player sections like Logo,
     * Watermark etc.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.7.3
     */
    size: TlvOptionalField(2, ContentLauncher2.TlvDimensionStruct)
  });
  ContentLauncher2.TlvBrandingInformationStruct = TlvObject({
    /**
     * This shall indicate name of of the provider for the given content.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.1
     */
    providerName: TlvField(0, TlvString.bound({ maxLength: 256 })),
    /**
     * This shall indicate background of the Video Player while content launch request is being processed by it.
     * This background information may also be used by the Video Player when it is in idle state.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.2
     */
    background: TlvOptionalField(1, ContentLauncher2.TlvStyleInformationStruct),
    /**
     * This shall indicate the logo shown when the Video Player is launching. This is also used when the Video
     * Player is in the idle state and Splash field is not available.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.3
     */
    logo: TlvOptionalField(2, ContentLauncher2.TlvStyleInformationStruct),
    /**
     * This shall indicate the style of progress bar for media playback.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.4
     */
    progressBar: TlvOptionalField(3, ContentLauncher2.TlvStyleInformationStruct),
    /**
     * This shall indicate the screen shown when the Video Player is in an idle state. If this property is not
     * populated, the Video Player shall default to logo or the provider name.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.5
     */
    splash: TlvOptionalField(4, ContentLauncher2.TlvStyleInformationStruct),
    /**
     * This shall indicate watermark shown when the media is playing.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.6.6
     */
    waterMark: TlvOptionalField(5, ContentLauncher2.TlvStyleInformationStruct)
  });
  ContentLauncher2.TlvLaunchUrlRequest = TlvObject({
    /**
     * This shall indicate the URL of content to launch.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.2.1
     */
    contentUrl: TlvField(0, TlvString),
    /**
     * This field, if present, shall provide a string that may be used to describe the content being accessed at
     * the given URL.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.2.2
     */
    displayString: TlvOptionalField(1, TlvString),
    /**
     * This field, if present, shall indicate the branding information that may be displayed when playing back the
     * given content.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.2.3
     */
    brandingInformation: TlvOptionalField(2, ContentLauncher2.TlvBrandingInformationStruct)
  });
  let Status;
  ((Status2) => {
    Status2[Status2["Success"] = 0] = "Success";
    Status2[Status2["UrlNotAvailable"] = 1] = "UrlNotAvailable";
    Status2[Status2["AuthFailed"] = 2] = "AuthFailed";
  })(Status = ContentLauncher2.Status || (ContentLauncher2.Status = {}));
  ContentLauncher2.TlvLauncherResponse = TlvObject({
    /**
     * This shall indicate the status of the command which resulted in this response.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.3.1
     */
    status: TlvField(0, TlvEnum()),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.3.2
     */
    data: TlvOptionalField(1, TlvByteString)
  });
  let Parameter;
  ((Parameter2) => {
    Parameter2[Parameter2["Actor"] = 0] = "Actor";
    Parameter2[Parameter2["Channel"] = 1] = "Channel";
    Parameter2[Parameter2["Character"] = 2] = "Character";
    Parameter2[Parameter2["Director"] = 3] = "Director";
    Parameter2[Parameter2["Event"] = 4] = "Event";
    Parameter2[Parameter2["Franchise"] = 5] = "Franchise";
    Parameter2[Parameter2["Genre"] = 6] = "Genre";
    Parameter2[Parameter2["League"] = 7] = "League";
    Parameter2[Parameter2["Popularity"] = 8] = "Popularity";
    Parameter2[Parameter2["Provider"] = 9] = "Provider";
    Parameter2[Parameter2["Sport"] = 10] = "Sport";
    Parameter2[Parameter2["SportsTeam"] = 11] = "SportsTeam";
    Parameter2[Parameter2["Type"] = 12] = "Type";
    Parameter2[Parameter2["Video"] = 13] = "Video";
  })(Parameter = ContentLauncher2.Parameter || (ContentLauncher2.Parameter = {}));
  ContentLauncher2.TlvAdditionalInfoStruct = TlvObject({
    /**
     * This shall indicate the name of external id, ex. "musicbrainz".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.5.1
     */
    name: TlvField(0, TlvString.bound({ maxLength: 256 })),
    /**
     * This shall indicate the value for external id, ex. "ST0000000666661".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.5.2
     */
    value: TlvField(1, TlvString.bound({ maxLength: 8192 }))
  });
  ContentLauncher2.TlvParameterStruct = TlvObject({
    /**
     * This shall indicate the entity type.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.3.1
     */
    type: TlvField(0, TlvEnum()),
    /**
     * This shall indicate the entity value, which is a search string, ex. “Manchester by the Sea”.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.3.2
     */
    value: TlvField(1, TlvString.bound({ maxLength: 1024 })),
    /**
     * This shall indicate the list of additional external content identifiers.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.3.3
     */
    externalIdList: TlvOptionalField(2, TlvArray(ContentLauncher2.TlvAdditionalInfoStruct))
  });
  ContentLauncher2.TlvContentSearchStruct = TlvObject({
    /**
     * This shall indicate the list of parameters comprising the search. If multiple parameters are provided, the
     * search parameters shall be joined with 'AND' logic. e.g. action movies with Tom Cruise will be represented
     * as [{Actor: 'Tom Cruise'}, {Type: 'Movie'}, {Genre: 'Action'}]
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.5.2.1
     */
    parameterList: TlvField(0, TlvArray(ContentLauncher2.TlvParameterStruct))
  });
  ContentLauncher2.TlvLaunchContentRequest = TlvObject({
    /**
     * This shall indicate the content to launch.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.1.1
     */
    search: TlvField(0, ContentLauncher2.TlvContentSearchStruct),
    /**
     * This shall indicate whether to automatically start playing content, where: * TRUE means best match should
     * start playing automatically. * FALSE means matches should be displayed on screen for user selection.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.1.2
     */
    autoPlay: TlvField(1, TlvBoolean),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.1.3
     */
    data: TlvOptionalField(2, TlvByteString)
  });
  let Feature;
  ((Feature2) => {
    Feature2["ContentSearch"] = "ContentSearch";
    Feature2["UrlPlayback"] = "UrlPlayback";
  })(Feature = ContentLauncher2.Feature || (ContentLauncher2.Feature = {}));
  ContentLauncher2.Base = ClusterFactory.Definition({
    id: 1290,
    name: "ContentLauncher",
    revision: 1,
    features: {
      /**
       * ContentSearch
       *
       * Device supports content search (non-app specific)
       */
      contentSearch: BitFlag(0),
      /**
       * UrlPlayback
       *
       * Device supports basic URL-based file playback
       */
      urlPlayback: BitFlag(1)
    }
  });
  ContentLauncher2.UrlPlaybackComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This list provides list of content types supported by the Video Player or Content App in the form of
       * entries in the HTTP "Accept" request header.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.3.1
       */
      acceptHeader: Attribute(0, TlvArray(TlvString, { maxLength: 100 }), { persistent: true, default: [] }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.3.2.1
       */
      supportedStreamingProtocols: Attribute(
        1,
        TlvBitmap(TlvUInt32, ContentLauncher2.SupportedStreamingProtocols),
        { persistent: true }
      )
    },
    commands: {
      /**
       * Upon receipt, this shall launch content from the specified URL.
       *
       * The content types supported include those identified in the AcceptHeader and SupportedStreamingProtocols
       * attributes.
       *
       * A check shall be made to ensure the URL is secure (uses HTTPS). This command returns a Launch Response.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.2
       */
      launchUrl: Command(1, ContentLauncher2.TlvLaunchUrlRequest, 2, ContentLauncher2.TlvLauncherResponse)
    }
  });
  ContentLauncher2.ContentSearchComponent = ClusterFactory.Component({
    commands: {
      /**
       * Upon receipt, this shall launch the specified content with optional search criteria. This command
       * returns a Launch Response.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.7.4.1
       */
      launchContent: Command(0, ContentLauncher2.TlvLaunchContentRequest, 2, ContentLauncher2.TlvLauncherResponse)
    }
  });
  ContentLauncher2.ContentSearchOrUrlPlaybackComponent = ClusterFactory.Component({});
  ContentLauncher2.Cluster = ClusterFactory.Extensible(
    ContentLauncher2.Base,
    /**
     * Use this factory method to create a ContentLauncher cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a ContentLauncher cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...ContentLauncher2.Base,
        supportedFeatures: BitFlags(ContentLauncher2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, ContentLauncher2.UrlPlaybackComponent, { urlPlayback: true });
      ClusterFactory.extend(cluster, ContentLauncher2.ContentSearchComponent, { contentSearch: true });
      ClusterFactory.extend(
        cluster,
        ContentLauncher2.ContentSearchOrUrlPlaybackComponent,
        { contentSearch: true },
        { urlPlayback: true }
      );
      return cluster;
    }
  );
  const UP = { urlPlayback: true };
  const CS = { contentSearch: true };
  ContentLauncher2.Complete = ClusterFactory.Definition({
    id: ContentLauncher2.Cluster.id,
    name: ContentLauncher2.Cluster.name,
    revision: ContentLauncher2.Cluster.revision,
    features: ContentLauncher2.Cluster.features,
    attributes: {
      acceptHeader: ClusterFactory.AsConditional(
        ContentLauncher2.UrlPlaybackComponent.attributes.acceptHeader,
        { mandatoryIf: [UP] }
      ),
      supportedStreamingProtocols: ClusterFactory.AsConditional(
        ContentLauncher2.UrlPlaybackComponent.attributes.supportedStreamingProtocols,
        { mandatoryIf: [UP] }
      )
    },
    commands: {
      launchContent: ClusterFactory.AsConditional(
        ContentLauncher2.ContentSearchComponent.commands.launchContent,
        { mandatoryIf: [CS] }
      ),
      launchUrl: ClusterFactory.AsConditional(ContentLauncher2.UrlPlaybackComponent.commands.launchUrl, { mandatoryIf: [UP] })
    }
  });
})(ContentLauncher || (ContentLauncher = {}));
const ContentLauncherCluster = ContentLauncher.Cluster;
export {
  ContentLauncher,
  ContentLauncherCluster
};
