// @project-chip/matter.js/dist/esm/cluster/definitions/DoorLockCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitsFromPartial, BitFlags } from "../../schema/BitmapSchema.js";
import {
  Attribute,
  OptionalWritableAttribute,
  AccessLevel,
  WritableAttribute,
  FixedAttribute,
  OptionalAttribute,
  Command,
  TlvNoResponse,
  OptionalCommand,
  Event,
  EventPriority
} from "../../cluster/Cluster.js";
import { TlvEnum, TlvUInt8, TlvUInt32, TlvUInt16, TlvBitmap, TlvEpochS } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
import { TlvString, TlvByteString } from "../../tlv/TlvString.js";
import { TlvObject, TlvOptionalField, TlvField } from "../../tlv/TlvObject.js";
import { TlvFabricIndex } from "../../datatype/FabricIndex.js";
import { TlvNodeId } from "../../datatype/NodeId.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var DoorLock;
((DoorLock2) => {
  let LockState;
  ((LockState2) => {
    LockState2[LockState2["NotFullyLocked"] = 0] = "NotFullyLocked";
    LockState2[LockState2["Locked"] = 1] = "Locked";
    LockState2[LockState2["Unlocked"] = 2] = "Unlocked";
  })(LockState = DoorLock2.LockState || (DoorLock2.LockState = {}));
  let LockType;
  ((LockType2) => {
    LockType2[LockType2["Deadbolt"] = 0] = "Deadbolt";
    LockType2[LockType2["Magnetic"] = 1] = "Magnetic";
    LockType2[LockType2["Other"] = 2] = "Other";
    LockType2[LockType2["Mortise"] = 3] = "Mortise";
    LockType2[LockType2["Rim"] = 4] = "Rim";
    LockType2[LockType2["LatchBolt"] = 5] = "LatchBolt";
    LockType2[LockType2["CylindricalLock"] = 6] = "CylindricalLock";
    LockType2[LockType2["TubularLock"] = 7] = "TubularLock";
    LockType2[LockType2["InterconnectedLock"] = 8] = "InterconnectedLock";
    LockType2[LockType2["DeadLatch"] = 9] = "DeadLatch";
    LockType2[LockType2["DoorFurniture"] = 10] = "DoorFurniture";
  })(LockType = DoorLock2.LockType || (DoorLock2.LockType = {}));
  let OperatingMode;
  ((OperatingMode2) => {
    OperatingMode2[OperatingMode2["Normal"] = 0] = "Normal";
    OperatingMode2[OperatingMode2["Vacation"] = 1] = "Vacation";
    OperatingMode2[OperatingMode2["Privacy"] = 2] = "Privacy";
    OperatingMode2[OperatingMode2["NoRemoteLockUnlock"] = 3] = "NoRemoteLockUnlock";
    OperatingMode2[OperatingMode2["Passage"] = 4] = "Passage";
  })(OperatingMode = DoorLock2.OperatingMode || (DoorLock2.OperatingMode = {}));
  DoorLock2.SupportedOperatingModes = {
    normal: BitFlag(0),
    vacation: BitFlag(1),
    privacy: BitFlag(2),
    noRemoteLockUnlock: BitFlag(3),
    passage: BitFlag(4)
  };
  DoorLock2.DefaultConfigurationRegister = {
    enableLocalProgrammingEnabled: BitFlag(0),
    keypadInterfaceDefaultAccessEnabled: BitFlag(1),
    remoteInterfaceDefaultAccessIsEnabled: BitFlag(2),
    soundEnabled: BitFlag(5),
    autoRelockTimeSet: BitFlag(6),
    ledSettingsSet: BitFlag(7)
  };
  DoorLock2.LocalProgrammingFeatures = {
    addUsersCredentialsSchedulesLocally: BitFlag(0),
    modifyUsersCredentialsSchedulesLocally: BitFlag(1),
    clearUsersCredentialsSchedulesLocally: BitFlag(2),
    adjustLockSettingsLocally: BitFlag(3)
  };
  DoorLock2.AlarmMask = {
    lockingMechanismJammed: BitFlag(0),
    lockResetToFactoryDefaults: BitFlag(1),
    reserved: BitFlag(2),
    rfModulePowerCycled: BitFlag(3),
    tamperAlarmWrongCodeEntryLimit: BitFlag(4),
    tamperAlarmFrontEscutcheonRemovedFromMain: BitFlag(5),
    forcedDoorOpenUnderDoorLockedCondition: BitFlag(6)
  };
  DoorLock2.TlvLockDoorRequest = TlvObject({ pinCode: TlvOptionalField(0, TlvByteString) });
  DoorLock2.TlvUnlockDoorRequest = TlvObject({ pinCode: TlvOptionalField(0, TlvByteString) });
  DoorLock2.TlvUnlockWithTimeoutRequest = TlvObject({
    timeout: TlvField(0, TlvUInt16),
    pinCode: TlvOptionalField(1, TlvByteString)
  });
  let AlarmCode;
  ((AlarmCode2) => {
    AlarmCode2[AlarmCode2["LockJammed"] = 0] = "LockJammed";
    AlarmCode2[AlarmCode2["LockFactoryReset"] = 1] = "LockFactoryReset";
    AlarmCode2[AlarmCode2["LockRadioPowerCycled"] = 3] = "LockRadioPowerCycled";
    AlarmCode2[AlarmCode2["WrongCodeEntryLimit"] = 4] = "WrongCodeEntryLimit";
    AlarmCode2[AlarmCode2["FrontEsceutcheonRemoved"] = 5] = "FrontEsceutcheonRemoved";
    AlarmCode2[AlarmCode2["DoorForcedOpen"] = 6] = "DoorForcedOpen";
    AlarmCode2[AlarmCode2["DoorAjar"] = 7] = "DoorAjar";
    AlarmCode2[AlarmCode2["ForcedUser"] = 8] = "ForcedUser";
  })(AlarmCode = DoorLock2.AlarmCode || (DoorLock2.AlarmCode = {}));
  DoorLock2.TlvDoorLockAlarmEvent = TlvObject({
    /**
     * The alarm code of the event that has happened.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.1.1
     */
    alarmCode: TlvField(0, TlvEnum())
  });
  let LockOperationType;
  ((LockOperationType2) => {
    LockOperationType2[LockOperationType2["Lock"] = 0] = "Lock";
    LockOperationType2[LockOperationType2["Unlock"] = 1] = "Unlock";
    LockOperationType2[LockOperationType2["NonAccessUserEvent"] = 2] = "NonAccessUserEvent";
    LockOperationType2[LockOperationType2["ForcedUserEvent"] = 3] = "ForcedUserEvent";
  })(LockOperationType = DoorLock2.LockOperationType || (DoorLock2.LockOperationType = {}));
  let OperationSource;
  ((OperationSource2) => {
    OperationSource2[OperationSource2["Unspecified"] = 0] = "Unspecified";
    OperationSource2[OperationSource2["Manual"] = 1] = "Manual";
    OperationSource2[OperationSource2["ProprietaryRemote"] = 2] = "ProprietaryRemote";
    OperationSource2[OperationSource2["Keypad"] = 3] = "Keypad";
    OperationSource2[OperationSource2["Auto"] = 4] = "Auto";
    OperationSource2[OperationSource2["Button"] = 5] = "Button";
    OperationSource2[OperationSource2["Schedule"] = 6] = "Schedule";
    OperationSource2[OperationSource2["Remote"] = 7] = "Remote";
    OperationSource2[OperationSource2["Rfid"] = 8] = "Rfid";
    OperationSource2[OperationSource2["Biometric"] = 9] = "Biometric";
  })(OperationSource = DoorLock2.OperationSource || (DoorLock2.OperationSource = {}));
  let CredentialType;
  ((CredentialType2) => {
    CredentialType2[CredentialType2["ProgrammingPin"] = 0] = "ProgrammingPin";
    CredentialType2[CredentialType2["Pin"] = 1] = "Pin";
    CredentialType2[CredentialType2["Rfid"] = 2] = "Rfid";
    CredentialType2[CredentialType2["Fingerprint"] = 3] = "Fingerprint";
    CredentialType2[CredentialType2["FingerVein"] = 4] = "FingerVein";
    CredentialType2[CredentialType2["Face"] = 5] = "Face";
  })(CredentialType = DoorLock2.CredentialType || (DoorLock2.CredentialType = {}));
  DoorLock2.TlvCredentialStruct = TlvObject({
    /**
     * The credential type used to authorize the lock operation.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.6.3.1
     */
    credentialType: TlvField(0, TlvEnum()),
    /**
     * This is the index of the specific credential used to authorize the lock operation in the list of credentials
     * identified by CredentialType (e.g. schedule, PIN, RFID, etc.). This shall be set to 0 if CredentialType is
     * ProgrammingPIN or does not correspond to a list that can be indexed into.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.6.3.2
     */
    credentialIndex: TlvField(1, TlvUInt16)
  });
  DoorLock2.TlvLockOperationEvent = TlvObject({
    /**
     * The type of the lock operation that was performed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.1
     */
    lockOperationType: TlvField(0, TlvEnum()),
    /**
     * The source of the lock operation that was performed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.2
     */
    operationSource: TlvField(1, TlvEnum()),
    /**
     * The lock UserIndex who performed the lock operation. This shall be null if there is no user index that can
     * be determined for the given operation source. This shall NOT be null if a user index can be determined. In
     * particular, this shall NOT be null if the operation was associated with a valid credential.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.3
     */
    userIndex: TlvField(2, TlvNullable(TlvUInt16)),
    /**
     * The fabric index of the fabric that performed the lock operation. This shall be null if there is no fabric
     * that can be determined for the given operation source. This shall NOT be null if the operation source is
     * "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.4
     */
    fabricIndex: TlvField(3, TlvNullable(TlvFabricIndex)),
    /**
     * The Node ID of the node that performed the lock operation. This shall be null if there is no Node associated
     * with the given operation source. This shall NOT be null if the operation source is "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.5
     */
    sourceNode: TlvField(4, TlvNullable(TlvNodeId)),
    /**
     * The list of credentials used in performing the lock operation. This shall be null if no credentials were
     * involved.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3.6
     */
    credentials: TlvOptionalField(5, TlvNullable(TlvArray(DoorLock2.TlvCredentialStruct, { minLength: 1 })))
  });
  let OperationError;
  ((OperationError2) => {
    OperationError2[OperationError2["Unspecified"] = 0] = "Unspecified";
    OperationError2[OperationError2["InvalidCredential"] = 1] = "InvalidCredential";
    OperationError2[OperationError2["DisabledUserDenied"] = 2] = "DisabledUserDenied";
    OperationError2[OperationError2["Restricted"] = 3] = "Restricted";
    OperationError2[OperationError2["InsufficientBattery"] = 4] = "InsufficientBattery";
  })(OperationError = DoorLock2.OperationError || (DoorLock2.OperationError = {}));
  DoorLock2.TlvLockOperationErrorEvent = TlvObject({
    /**
     * The type of the lock operation that was performed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.1
     */
    lockOperationType: TlvField(0, TlvEnum()),
    /**
     * The source of the lock operation that was performed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.2
     */
    operationSource: TlvField(1, TlvEnum()),
    /**
     * The lock operation error triggered when the operation was performed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.3
     */
    operationError: TlvField(2, TlvEnum()),
    /**
     * The lock UserIndex who performed the lock operation. This shall be null if there is no user id that can be
     * determined for the given operation source.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.4
     */
    userIndex: TlvField(3, TlvNullable(TlvUInt16)),
    /**
     * The fabric index of the fabric that performed the lock operation. This shall be null if there is no fabric
     * that can be determined for the given operation source. This shall NOT be null if the operation source is
     * "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.5
     */
    fabricIndex: TlvField(4, TlvNullable(TlvFabricIndex)),
    /**
     * The Node ID of the node that performed the lock operation. This shall be null if there is no Node associated
     * with the given operation source. This shall NOT be null if the operation source is
     *
     * "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.6
     */
    sourceNode: TlvField(5, TlvNullable(TlvNodeId)),
    /**
     * The list of credentials used in performing the lock operation. This shall be null if no credentials were
     * involved.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4.7
     */
    credentials: TlvOptionalField(6, TlvNullable(TlvArray(DoorLock2.TlvCredentialStruct, { minLength: 1 })))
  });
  let DoorState;
  ((DoorState2) => {
    DoorState2[DoorState2["DoorOpen"] = 0] = "DoorOpen";
    DoorState2[DoorState2["DoorClosed"] = 1] = "DoorClosed";
    DoorState2[DoorState2["DoorJammed"] = 2] = "DoorJammed";
    DoorState2[DoorState2["DoorForcedOpen"] = 3] = "DoorForcedOpen";
    DoorState2[DoorState2["DoorUnspecifiedError"] = 4] = "DoorUnspecifiedError";
    DoorState2[DoorState2["DoorAjar"] = 5] = "DoorAjar";
  })(DoorState = DoorLock2.DoorState || (DoorLock2.DoorState = {}));
  DoorLock2.TlvDoorStateChangeEvent = TlvObject({
    /**
     * The new door state for this door event.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.2.1
     */
    doorState: TlvField(0, TlvEnum())
  });
  DoorLock2.CredentialRulesSupport = { single: BitFlag(0), dual: BitFlag(1), tri: BitFlag(2) };
  let DataOperationType;
  ((DataOperationType2) => {
    DataOperationType2[DataOperationType2["Add"] = 0] = "Add";
    DataOperationType2[DataOperationType2["Clear"] = 1] = "Clear";
    DataOperationType2[DataOperationType2["Modify"] = 2] = "Modify";
  })(DataOperationType = DoorLock2.DataOperationType || (DoorLock2.DataOperationType = {}));
  let UserStatus;
  ((UserStatus2) => {
    UserStatus2[UserStatus2["Available"] = 0] = "Available";
    UserStatus2[UserStatus2["OccupiedEnabled"] = 1] = "OccupiedEnabled";
    UserStatus2[UserStatus2["OccupiedDisabled"] = 3] = "OccupiedDisabled";
  })(UserStatus = DoorLock2.UserStatus || (DoorLock2.UserStatus = {}));
  let UserType;
  ((UserType2) => {
    UserType2[UserType2["UnrestrictedUser"] = 0] = "UnrestrictedUser";
    UserType2[UserType2["YearDayScheduleUser"] = 1] = "YearDayScheduleUser";
    UserType2[UserType2["WeekDayScheduleUser"] = 2] = "WeekDayScheduleUser";
    UserType2[UserType2["ProgrammingUser"] = 3] = "ProgrammingUser";
    UserType2[UserType2["NonAccessUser"] = 4] = "NonAccessUser";
    UserType2[UserType2["ForcedUser"] = 5] = "ForcedUser";
    UserType2[UserType2["DisposableUser"] = 6] = "DisposableUser";
    UserType2[UserType2["ExpiringUser"] = 7] = "ExpiringUser";
    UserType2[UserType2["ScheduleRestrictedUser"] = 8] = "ScheduleRestrictedUser";
    UserType2[UserType2["RemoteOnlyUser"] = 9] = "RemoteOnlyUser";
  })(UserType = DoorLock2.UserType || (DoorLock2.UserType = {}));
  let CredentialRule;
  ((CredentialRule2) => {
    CredentialRule2[CredentialRule2["Single"] = 0] = "Single";
    CredentialRule2[CredentialRule2["Dual"] = 1] = "Dual";
    CredentialRule2[CredentialRule2["Tri"] = 2] = "Tri";
  })(CredentialRule = DoorLock2.CredentialRule || (DoorLock2.CredentialRule = {}));
  DoorLock2.TlvSetUserRequest = TlvObject({
    operationType: TlvField(0, TlvEnum()),
    userIndex: TlvField(1, TlvUInt16),
    userName: TlvField(2, TlvNullable(TlvString)),
    userUniqueId: TlvField(3, TlvNullable(TlvUInt32)),
    userStatus: TlvField(4, TlvNullable(TlvEnum())),
    userType: TlvField(5, TlvNullable(TlvEnum())),
    credentialRule: TlvField(6, TlvNullable(TlvEnum()))
  });
  DoorLock2.TlvGetUserRequest = TlvObject({ userIndex: TlvField(0, TlvUInt16) });
  DoorLock2.TlvGetUserResponse = TlvObject({
    userIndex: TlvField(0, TlvUInt16),
    userName: TlvField(1, TlvNullable(TlvString)),
    userUniqueId: TlvField(2, TlvNullable(TlvUInt32)),
    userStatus: TlvField(3, TlvNullable(TlvEnum())),
    userType: TlvField(4, TlvNullable(TlvEnum())),
    credentialRule: TlvField(5, TlvNullable(TlvEnum())),
    credentials: TlvField(6, TlvNullable(TlvArray(DoorLock2.TlvCredentialStruct))),
    creatorFabricIndex: TlvField(7, TlvNullable(TlvFabricIndex)),
    lastModifiedFabricIndex: TlvField(8, TlvNullable(TlvFabricIndex)),
    nextUserIndex: TlvField(9, TlvNullable(TlvUInt16))
  });
  DoorLock2.TlvClearUserRequest = TlvObject({ userIndex: TlvField(0, TlvUInt16) });
  DoorLock2.TlvSetCredentialRequest = TlvObject({
    operationType: TlvField(0, TlvEnum()),
    credential: TlvField(1, DoorLock2.TlvCredentialStruct),
    credentialData: TlvField(2, TlvByteString),
    userIndex: TlvField(3, TlvNullable(TlvUInt16)),
    userStatus: TlvField(4, TlvNullable(TlvEnum())),
    userType: TlvField(5, TlvNullable(TlvEnum()))
  });
  let DlStatus;
  ((DlStatus2) => {
    DlStatus2[DlStatus2["Success"] = 0] = "Success";
    DlStatus2[DlStatus2["Failure"] = 1] = "Failure";
    DlStatus2[DlStatus2["Duplicate"] = 2] = "Duplicate";
    DlStatus2[DlStatus2["Occupied"] = 3] = "Occupied";
    DlStatus2[DlStatus2["InvalidField"] = 133] = "InvalidField";
    DlStatus2[DlStatus2["ResourceExhausted"] = 137] = "ResourceExhausted";
    DlStatus2[DlStatus2["NotFound"] = 139] = "NotFound";
  })(DlStatus = DoorLock2.DlStatus || (DoorLock2.DlStatus = {}));
  DoorLock2.TlvSetCredentialResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    userIndex: TlvField(1, TlvNullable(TlvUInt16)),
    nextCredentialIndex: TlvField(2, TlvNullable(TlvUInt16))
  });
  DoorLock2.TlvGetCredentialStatusRequest = TlvObject({ credential: TlvField(0, DoorLock2.TlvCredentialStruct) });
  DoorLock2.TlvGetCredentialStatusResponse = TlvObject({
    credentialExists: TlvField(0, TlvBoolean),
    userIndex: TlvField(1, TlvNullable(TlvUInt16)),
    creatorFabricIndex: TlvField(2, TlvNullable(TlvFabricIndex)),
    lastModifiedFabricIndex: TlvField(3, TlvNullable(TlvFabricIndex)),
    nextCredentialIndex: TlvField(4, TlvNullable(TlvUInt16))
  });
  DoorLock2.TlvClearCredentialRequest = TlvObject({ credential: TlvField(0, TlvNullable(DoorLock2.TlvCredentialStruct)) });
  let LockDataType;
  ((LockDataType2) => {
    LockDataType2[LockDataType2["Unspecified"] = 0] = "Unspecified";
    LockDataType2[LockDataType2["ProgrammingCode"] = 1] = "ProgrammingCode";
    LockDataType2[LockDataType2["UserIndex"] = 2] = "UserIndex";
    LockDataType2[LockDataType2["WeekDaySchedule"] = 3] = "WeekDaySchedule";
    LockDataType2[LockDataType2["YearDaySchedule"] = 4] = "YearDaySchedule";
    LockDataType2[LockDataType2["HolidaySchedule"] = 5] = "HolidaySchedule";
    LockDataType2[LockDataType2["Pin"] = 6] = "Pin";
    LockDataType2[LockDataType2["Rfid"] = 7] = "Rfid";
    LockDataType2[LockDataType2["Fingerprint"] = 8] = "Fingerprint";
    LockDataType2[LockDataType2["FingerVein"] = 9] = "FingerVein";
    LockDataType2[LockDataType2["Face"] = 10] = "Face";
  })(LockDataType = DoorLock2.LockDataType || (DoorLock2.LockDataType = {}));
  DoorLock2.TlvLockUserChangeEvent = TlvObject({
    /**
     * The lock data type that was changed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.1
     */
    lockDataType: TlvField(0, TlvEnum()),
    /**
     * The data operation performed on the lock data type changed.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.2
     */
    dataOperationType: TlvField(1, TlvEnum()),
    /**
     * The source of the user data change.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.3
     */
    operationSource: TlvField(2, TlvEnum()),
    /**
     * The lock UserIndex associated with the change (if any). This shall be null if there is no specific user
     * associated with the data operation. This shall be 0xFFFE if all users are affected (e.g. Clear Users).
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.4
     */
    userIndex: TlvField(3, TlvNullable(TlvUInt16)),
    /**
     * The fabric index of the fabric that performed the change (if any). This shall be null if there is no fabric
     * that can be determined to have caused the change. This shall NOT be null if the operation source is "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.5
     */
    fabricIndex: TlvField(4, TlvNullable(TlvFabricIndex)),
    /**
     * The Node ID that that performed the change (if any). The Node ID of the node that performed the change. This
     * shall be null if there was no Node involved in the change. This shall NOT be null if the operation source is
     * "Remote".
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.6
     */
    sourceNode: TlvField(5, TlvNullable(TlvNodeId)),
    /**
     * This is the index of the specific item that was changed (e.g. schedule, PIN, RFID, etc.) in the list of
     * items identified by LockDataType. This shall be null if the LockDataType does not correspond to a list that
     * can be indexed into (e.g. ProgrammingUser). This shall be 0xFFFE if all indices are affected (e.g. Clear PIN
     * Code, Clear RFID Code, Clear Week Day Schedule, Clear Year Day Schedule, etc.).
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5.7
     */
    dataIndex: TlvField(6, TlvNullable(TlvUInt16))
  });
  DoorLock2.DaysMaskMap = {
    sunday: BitFlag(0),
    monday: BitFlag(1),
    tuesday: BitFlag(2),
    wednesday: BitFlag(3),
    thursday: BitFlag(4),
    friday: BitFlag(5),
    saturday: BitFlag(6)
  };
  DoorLock2.TlvSetWeekDayScheduleRequest = TlvObject({
    weekDayIndex: TlvField(0, TlvUInt8),
    userIndex: TlvField(1, TlvUInt16),
    daysMask: TlvField(2, TlvBitmap(TlvUInt8, DoorLock2.DaysMaskMap)),
    startHour: TlvField(3, TlvUInt8),
    startMinute: TlvField(4, TlvUInt8),
    endHour: TlvField(5, TlvUInt8),
    endMinute: TlvField(6, TlvUInt8)
  });
  DoorLock2.TlvGetWeekDayScheduleRequest = TlvObject({
    weekDayIndex: TlvField(0, TlvUInt8),
    userIndex: TlvField(1, TlvUInt16)
  });
  DoorLock2.TlvGetWeekDayScheduleResponse = TlvObject({
    weekDayIndex: TlvField(0, TlvUInt8),
    userIndex: TlvField(1, TlvUInt16),
    status: TlvField(2, TlvEnum()),
    daysMask: TlvOptionalField(3, TlvBitmap(TlvUInt8, DoorLock2.DaysMaskMap)),
    startHour: TlvOptionalField(4, TlvUInt8),
    startMinute: TlvOptionalField(5, TlvUInt8),
    endHour: TlvOptionalField(6, TlvUInt8),
    endMinute: TlvOptionalField(7, TlvUInt8)
  });
  DoorLock2.TlvClearWeekDayScheduleRequest = TlvObject({
    weekDayIndex: TlvField(0, TlvUInt8),
    userIndex: TlvField(1, TlvUInt16)
  });
  DoorLock2.TlvSetYearDayScheduleRequest = TlvObject({
    yearDayIndex: TlvField(0, TlvUInt8),
    userIndex: TlvField(1, TlvUInt16),
    localStartTime: TlvField(2, TlvEpochS),
    localEndTime: TlvField(3, TlvEpochS)
  });
  DoorLock2.TlvGetYearDayScheduleRequest = TlvObject({
    yearDayIndex: TlvField(0, TlvUInt8),
    userIndex: TlvField(1, TlvUInt16)
  });
  DoorLock2.TlvGetYearDayScheduleResponse = TlvObject({
    yearDayIndex: TlvField(0, TlvUInt8),
    userIndex: TlvField(1, TlvUInt16),
    status: TlvField(2, TlvEnum()),
    localStartTime: TlvOptionalField(3, TlvEpochS),
    localEndTime: TlvOptionalField(4, TlvEpochS)
  });
  DoorLock2.TlvClearYearDayScheduleRequest = TlvObject({
    yearDayIndex: TlvField(0, TlvUInt8),
    userIndex: TlvField(1, TlvUInt16)
  });
  DoorLock2.TlvSetHolidayScheduleRequest = TlvObject({
    holidayIndex: TlvField(0, TlvUInt8),
    localStartTime: TlvField(1, TlvEpochS),
    localEndTime: TlvField(2, TlvEpochS),
    operatingMode: TlvField(3, TlvEnum())
  });
  DoorLock2.TlvGetHolidayScheduleRequest = TlvObject({ holidayIndex: TlvField(0, TlvUInt8) });
  DoorLock2.TlvGetHolidayScheduleResponse = TlvObject({
    holidayIndex: TlvField(0, TlvUInt8),
    status: TlvField(1, TlvEnum()),
    localStartTime: TlvOptionalField(2, TlvEpochS),
    localEndTime: TlvOptionalField(3, TlvEpochS),
    operatingMode: TlvOptionalField(4, TlvEnum())
  });
  DoorLock2.TlvClearHolidayScheduleRequest = TlvObject({ holidayIndex: TlvField(0, TlvUInt8) });
  DoorLock2.KeypadOperationEventMask = {
    unknownOrManufacturerSpecificKeypadOperationEvent: BitFlag(0),
    lockSourceKeypad: BitFlag(1),
    unlockSourceKeypad: BitFlag(2),
    lockSourceKeypadErrorInvalidPin: BitFlag(3),
    lockSourceKeypadErrorInvalidSchedule: BitFlag(4),
    unlockSourceKeypadErrorInvalidCode: BitFlag(5),
    unlockSourceKeypadErrorInvalidSchedule: BitFlag(6),
    nonAccessUserOperationEventSourceKeypad: BitFlag(15)
  };
  DoorLock2.KeypadProgrammingEventMask = {
    unknown: BitFlag(0),
    pinCodeChanged: BitFlag(1),
    pinAdded: BitFlag(2),
    pinCleared: BitFlag(3),
    pinChanged: BitFlag(4)
  };
  DoorLock2.RemoteOperationEventMask = {
    unknownOrManufacturerSpecificRemoteOperationEvent: BitFlag(0),
    lockSourceRemote: BitFlag(1),
    unlockSourceRemote: BitFlag(2),
    lockSourceRemoteErrorInvalidCode: BitFlag(3),
    lockSourceRemoteErrorInvalidSchedule: BitFlag(4),
    unlockSourceRemoteErrorInvalidCode: BitFlag(5),
    unlockSourceRemoteErrorInvalidSchedule: BitFlag(6)
  };
  DoorLock2.ManualOperationEventMask = {
    unknownOrManufacturerSpecificManualOperationEvent: BitFlag(0),
    thumbturnLock: BitFlag(1),
    thumbturnUnlock: BitFlag(2),
    oneTouchLock: BitFlag(7),
    keyLock: BitFlag(8),
    keyUnlock: BitFlag(9),
    autoLock: BitFlag(10),
    scheduleLock: BitFlag(11),
    scheduleUnlock: BitFlag(12),
    manualLock: BitFlag(13),
    manualUnlock: BitFlag(14)
  };
  DoorLock2.RemoteProgrammingEventMask = {
    unknown: BitFlag(0),
    pinAdded: BitFlag(2),
    pinCleared: BitFlag(3),
    pinChanged: BitFlag(4),
    rfidCodeAdded: BitFlag(5),
    rfidCodeCleared: BitFlag(6)
  };
  DoorLock2.RfidOperationEventMask = {
    unknownOrManufacturerSpecificKeypadOperationEvent: BitFlag(0),
    lockSourceRfid: BitFlag(1),
    unlockSourceRfid: BitFlag(2),
    lockSourceRfidErrorInvalidRfidId: BitFlag(3),
    lockSourceRfidErrorInvalidSchedule: BitFlag(4),
    unlockSourceRfidErrorInvalidRfidId: BitFlag(5),
    unlockSourceRfidErrorInvalidSchedule: BitFlag(6)
  };
  DoorLock2.RfidProgrammingEventMask = { unknown: BitFlag(0), idAdded: BitFlag(5), idCleared: BitFlag(6) };
  let Feature;
  ((Feature2) => {
    Feature2["PinCredential"] = "PinCredential";
    Feature2["RfidCredential"] = "RfidCredential";
    Feature2["FingerCredentials"] = "FingerCredentials";
    Feature2["Logging"] = "Logging";
    Feature2["WeekDayAccessSchedules"] = "WeekDayAccessSchedules";
    Feature2["DoorPositionSensor"] = "DoorPositionSensor";
    Feature2["FaceCredentials"] = "FaceCredentials";
    Feature2["CredentialOverTheAirAccess"] = "CredentialOverTheAirAccess";
    Feature2["User"] = "User";
    Feature2["Notification"] = "Notification";
    Feature2["YearDayAccessSchedules"] = "YearDayAccessSchedules";
    Feature2["HolidaySchedules"] = "HolidaySchedules";
  })(Feature = DoorLock2.Feature || (DoorLock2.Feature = {}));
  DoorLock2.Base = ClusterFactory.Definition({
    id: 257,
    name: "DoorLock",
    revision: 6,
    features: {
      /**
       * PinCredential
       *
       * Lock supports PIN credentials (via keypad, or over- the-air)
       */
      pinCredential: BitFlag(0),
      /**
       * RfidCredential
       *
       * Lock supports RFID credentials
       */
      rfidCredential: BitFlag(1),
      /**
       * FingerCredentials
       *
       * Lock supports finger related credentials (fingerprint, finger vein)
       */
      fingerCredentials: BitFlag(2),
      /**
       * Logging
       *
       * Lock supports local/on-lock logging when Events are not supported
       */
      logging: BitFlag(3),
      /**
       * WeekDayAccessSchedules
       *
       * Lock supports week day user access schedules
       */
      weekDayAccessSchedules: BitFlag(4),
      /**
       * DoorPositionSensor
       *
       * Lock supports a door position sensor that indicates door’s state
       */
      doorPositionSensor: BitFlag(5),
      /**
       * FaceCredentials
       *
       * Lock supports face related credentials (face, iris, retina)
       */
      faceCredentials: BitFlag(6),
      /**
       * CredentialOverTheAirAccess
       *
       * PIN codes over- the-air supported for lock/unlock operations
       */
      credentialOverTheAirAccess: BitFlag(7),
      /**
       * User
       *
       * Lock supports the user commands and database
       */
      user: BitFlag(8),
      /**
       * Notification
       *
       * Operation and Programming Notifications
       */
      notification: BitFlag(9),
      /**
       * YearDayAccessSchedules
       *
       * Lock supports year day user access schedules
       */
      yearDayAccessSchedules: BitFlag(10),
      /**
       * HolidaySchedules
       *
       * Lock supports holiday schedules
       */
      holidaySchedules: BitFlag(11)
    },
    attributes: {
      /**
       * The LockState Attribute may be NULL if the lock hardware does not currently know the status of the
       * locking mechanism. For example, a lock may not know the LockState status after a power cycle until the
       * first lock actuation is completed.
       *
       * The Not Fully Locked value is used by a lock to indicate that the state of the lock is somewhere between
       * Locked and Unlocked so it is only partially secured. For example, a deadbolt could be partially extended
       * and not in a dead latched state.
       *
       * If the Scenes server cluster is implemented on the same endpoint, the following extension field shall be
       * added to the Scene Table:
       *
       *   • LockState
       *
       * When the LockState attribute is part of a Scene table, the attribute is treated as a writable command;
       * that is, setting the LockState to lock will command the lock to lock, and setting the LockState to
       * unlock will command the lock to unlock. Setting the LockState attribute to “not fully locked” is not
       * supported. The Transition Time field in the Scene table will be treated as a delay before setting the
       * LockState attribute; that is, it is possible to activate a scene with the lock actuation some seconds
       * later.
       *
       * Locks that do not have an actuation mechanism SHOULD not support the Scene table extension.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.1
       */
      lockState: Attribute(0, TlvNullable(TlvEnum()), { scene: true }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.2
       */
      lockType: Attribute(1, TlvEnum()),
      /**
       * The ActuatorEnabled attribute indicates if the lock is currently able to (Enabled) or not able to
       * (Disabled) process remote Lock, Unlock, or Unlock with Timeout commands.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.3
       */
      actuatorEnabled: Attribute(2, TlvBoolean),
      /**
       * Modifies the language for the on-screen or audible user interface using a 2-byte language code from
       * ISO-639-1.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.22
       */
      language: OptionalWritableAttribute(
        33,
        TlvString.bound({ maxLength: 3 }),
        { writeAcl: AccessLevel.Manage }
      ),
      /**
       * The settings for the LED support three different modes
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.25
       */
      ledSettings: OptionalWritableAttribute(
        34,
        TlvUInt8.bound({ max: 2 }),
        { default: 0, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The number of seconds to wait after unlocking a lock before it automatically locks again. 0=disabled. If
       * set, unlock operations from any source will be timed. For one time unlock with timeout use the specific
       * command.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.26
       */
      autoRelockTime: OptionalWritableAttribute(35, TlvUInt32, { writeAcl: AccessLevel.Manage }),
      /**
       * The sound volume on a door lock has four possible settings: silent, low, high and medium volumes
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.27
       */
      soundVolume: OptionalWritableAttribute(
        36,
        TlvUInt8.bound({ max: 3 }),
        { default: 0, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The current operating mode of the lock (see OperatingModeEnum).
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.23
       */
      operatingMode: WritableAttribute(
        37,
        TlvEnum(),
        { default: 0 /* Normal */, writeAcl: AccessLevel.Manage }
      ),
      /**
       * This bitmap contains all operating bits of the Operating Mode Attribute supported by the lock. All
       * operating modes NOT supported by a lock shall be set to one. The value of the OperatingMode enumeration
       * defines the related bit to be set, as
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.24
       */
      supportedOperatingModes: FixedAttribute(
        38,
        TlvBitmap(TlvUInt16, DoorLock2.SupportedOperatingModes),
        { default: BitsFromPartial(DoorLock2.SupportedOperatingModes, { vacation: true, privacy: true, passage: true }) }
      ),
      /**
       * This attribute represents the default configurations as they are physically set on the device (example:
       * hardware dip switch setting, etc…) and represents the default setting for some of the attributes within
       * this cluster (for example: LED, Auto Lock, Sound Volume, and Operating Mode attributes).
       *
       * This is a read-only attribute and is intended to allow clients to determine what changes may need to be
       * made without having to query all the included attributes. It may be beneficial for the clients to know
       * what the device’s original settings were in the event that the device needs to be restored to factory
       * default settings.
       *
       * If the Client device would like to query and modify the door lock server’s operating settings, it SHOULD
       * send read and write attribute requests to the specific attributes.
       *
       * For example, the Sound Volume attribute default value is Silent Mode. However, it is possible that the
       * current Sound Volume is High Volume. Therefore, if the client wants to query/modify the current Sound
       * Volume setting on the server, the client SHOULD read/write to the Sound Volume attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.28
       */
      defaultConfigurationRegister: OptionalAttribute(39, TlvBitmap(TlvUInt16, DoorLock2.DefaultConfigurationRegister)),
      /**
       * Enable/disable local programming on the door lock of certain features (see LocalProgrammingFeatures
       * attribute). If this value is set to TRUE then local programming is enabled on the door lock for all
       * features. If it is set to FALSE then local programming is disabled on the door lock for those features
       * whose bit is set to 0 in the LocalProgrammingFeatures attribute. Local programming shall be enabled by
       * default.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.29
       */
      enableLocalProgramming: OptionalWritableAttribute(
        40,
        TlvBoolean,
        { default: true, writeAcl: AccessLevel.Administer }
      ),
      /**
       * Enable/disable the ability to lock the door lock with a single touch on the door lock.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.30
       */
      enableOneTouchLocking: OptionalWritableAttribute(
        41,
        TlvBoolean,
        { default: true, writeAcl: AccessLevel.Manage }
      ),
      /**
       * Enable/disable an inside LED that allows the user to see at a glance if the door is locked.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.31
       */
      enableInsideStatusLed: OptionalWritableAttribute(
        42,
        TlvBoolean,
        { default: true, writeAcl: AccessLevel.Manage }
      ),
      /**
       * Enable/disable a button inside the door that is used to put the lock into privacy mode. When the lock is
       * in privacy mode it cannot be manipulated from the outside.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.32
       */
      enablePrivacyModeButton: OptionalWritableAttribute(
        43,
        TlvBoolean,
        { default: true, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The local programming features that will be disabled when EnableLocalProgramming attribute is set to
       * False. If a door lock doesn’t support disabling one aspect of local programming it shall return
       * CONSTRAINT_ERROR during a write operation of this attribute. If the EnableLocalProgramming attribute is
       * set to True then all local programming features shall be enabled regardless of the bits set to 0 in this
       * attribute.
       *
       * The features that can be disabled from local programming are defined in the following bitmap.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.33
       */
      localProgrammingFeatures: OptionalWritableAttribute(
        44,
        TlvBitmap(TlvUInt8, DoorLock2.LocalProgrammingFeatures),
        { writeAcl: AccessLevel.Administer }
      ),
      /**
       * This attribute is only supported if the Alarms cluster is on the same endpoint. The alarm mask is used
       * to turn on/off alarms for particular functions. Alarms for an alarm group are enabled if the associated
       * alarm mask bit is set. Each bit represents a group of alarms. Entire alarm groups can be turned on or
       * off by setting or clearing the associated bit in the alarm mask.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.39
       */
      alarmMask: OptionalWritableAttribute(
        64,
        TlvBitmap(TlvUInt16, DoorLock2.AlarmMask),
        {
          default: BitsFromPartial(DoorLock2.AlarmMask, { lockingMechanismJammed: true, lockResetToFactoryDefaults: true, reserved: true, rfModulePowerCycled: true, tamperAlarmWrongCodeEntryLimit: true, tamperAlarmFrontEscutcheonRemovedFromMain: true, forcedDoorOpenUnderDoorLockedCondition: true }),
          writeAcl: AccessLevel.Administer
        }
      )
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      lockDoor: Command(0, DoorLock2.TlvLockDoorRequest, 0, TlvNoResponse, { timed: true }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      unlockDoor: Command(1, DoorLock2.TlvUnlockDoorRequest, 1, TlvNoResponse, { timed: true }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      unlockWithTimeout: OptionalCommand(3, DoorLock2.TlvUnlockWithTimeoutRequest, 3, TlvNoResponse, { timed: true })
    },
    events: {
      /**
       * The door lock cluster provides several alarms which can be sent when there is a critical state on the
       * door lock. The alarms available for the door lock cluster are listed in the AlarmCodeEnum section below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.1
       */
      doorLockAlarm: Event(0, EventPriority.Critical, DoorLock2.TlvDoorLockAlarmEvent),
      /**
       * The door lock server sends out a LockOperation event when the event is triggered by the various lock
       * operation sources.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.3
       */
      lockOperation: Event(2, EventPriority.Critical, DoorLock2.TlvLockOperationEvent),
      /**
       * The door lock server sends out a LockOperationError event when a lock operation fails for various
       * reasons.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.4
       */
      lockOperationError: Event(3, EventPriority.Critical, DoorLock2.TlvLockOperationErrorEvent)
    }
  });
  DoorLock2.DoorPositionSensorComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The current door state as defined in DoorStateEnum.
       *
       * This attribute shall be null only if an internal error prevents the retrieval of the current door state.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.4
       */
      doorState: Attribute(3, TlvNullable(TlvEnum())),
      /**
       * This attribute holds the number of door open events that have occurred since it was last zeroed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.5
       */
      doorOpenEvents: OptionalWritableAttribute(4, TlvUInt32, { writeAcl: AccessLevel.Manage }),
      /**
       * This attribute holds the number of door closed events that have occurred since it was last zeroed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.6
       */
      doorClosedEvents: OptionalWritableAttribute(5, TlvUInt32, { writeAcl: AccessLevel.Manage }),
      /**
       * This attribute holds the number of minutes the door has been open since the last time it transitioned
       * from closed to open.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.7
       */
      openPeriod: OptionalWritableAttribute(6, TlvUInt16, { writeAcl: AccessLevel.Manage })
    },
    events: {
      /**
       * The door lock server sends out a DoorStateChange event when the door lock door state changes.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.2
       */
      doorStateChange: Event(1, EventPriority.Critical, DoorLock2.TlvDoorStateChangeEvent)
    }
  });
  DoorLock2.LoggingComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The number of available log records.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.8
       */
      numberOfLogRecordsSupported: FixedAttribute(16, TlvUInt16, { default: 0 }),
      /**
       * Enable/disable event logging. When event logging is enabled, all event messages are stored on the lock
       * for retrieval. Logging events can be but not limited to Tamper Alarm, Lock, Unlock, AutoRelock, User
       * Code Added, User Code Cleared, Schedule Added, and Schedule Cleared. For a full detail of all the
       * possible alarms and events, please refer to the full list in the Alarm and Event Masks Attribute Set.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.21
       */
      enableLogging: WritableAttribute(32, TlvBoolean, { default: true, writeAcl: AccessLevel.Administer })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getLogRecord: Command(4, TlvNoArguments, 4, TlvNoArguments, { invokeAcl: AccessLevel.Manage })
    }
  });
  DoorLock2.UserComponent = ClusterFactory.Component({
    attributes: {
      /**
       * Number of total users supported by the lock.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.9
       */
      numberOfTotalUsersSupported: FixedAttribute(17, TlvUInt16, { default: 0 }),
      /**
       * This bitmap contains a bit for every value of CredentialRuleEnum supported on this device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.19
       */
      credentialRulesSupport: FixedAttribute(
        27,
        TlvBitmap(TlvUInt8, DoorLock2.CredentialRulesSupport),
        { default: BitsFromPartial(DoorLock2.CredentialRulesSupport, { single: true }) }
      ),
      /**
       * The number of credentials that could be assigned for each user.
       *
       * Depending on the value of NumberOfRFIDUsersSupported and NumberOfPINUsersSupported it may not be
       * possible to assign that number of credentials for a user.
       *
       * For example, if the device supports only PIN and RFID credential types,
       * NumberOfCredentialsSupportedPerUser is set to 10, NumberOfPINUsersSupported is set to 5 and
       * NumberOfRFIDUsersSupported is set to 3, it will not be possible to actually assign 10 credentials for a
       * user because maximum number of credentials in the database is 8.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.20
       */
      numberOfCredentialsSupportedPerUser: FixedAttribute(28, TlvUInt8, { default: 0 }),
      /**
       * Number of minutes a PIN, RFID, Fingerprint, or other credential associated with a user of type
       * ExpiringUser shall remain valid after its first use before expiring. When the credential expires the
       * UserStatus for the corresponding user record shall be set to OccupiedDisabled.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.38
       */
      expiringUserTimeout: OptionalWritableAttribute(
        53,
        TlvUInt16.bound({ min: 1, max: 2880 }),
        { writeAcl: AccessLevel.Administer }
      )
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setUser: Command(
        26,
        DoorLock2.TlvSetUserRequest,
        26,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getUser: Command(27, DoorLock2.TlvGetUserRequest, 28, DoorLock2.TlvGetUserResponse, { invokeAcl: AccessLevel.Administer }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearUser: Command(
        29,
        DoorLock2.TlvClearUserRequest,
        29,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setCredential: Command(
        34,
        DoorLock2.TlvSetCredentialRequest,
        35,
        DoorLock2.TlvSetCredentialResponse,
        { invokeAcl: AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getCredentialStatus: Command(
        36,
        DoorLock2.TlvGetCredentialStatusRequest,
        37,
        DoorLock2.TlvGetCredentialStatusResponse,
        { invokeAcl: AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearCredential: Command(
        38,
        DoorLock2.TlvClearCredentialRequest,
        38,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer, timed: true }
      )
    },
    events: {
      /**
       * The door lock server sends out a LockUserChange event when a lock user, schedule, or credential change
       * has occurred.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.5.5
       */
      lockUserChange: Event(4, EventPriority.Info, DoorLock2.TlvLockUserChangeEvent)
    }
  });
  DoorLock2.PinCredentialComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The number of PIN users supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.10
       */
      numberOfPinUsersSupported: FixedAttribute(18, TlvUInt16, { default: 0 }),
      /**
       * An 8 bit value indicates the maximum length in bytes of a PIN Code on this device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.15
       */
      maxPinCodeLength: FixedAttribute(23, TlvUInt8),
      /**
       * An 8 bit value indicates the minimum length in bytes of a PIN Code on this device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.16
       */
      minPinCodeLength: FixedAttribute(24, TlvUInt8),
      /**
       * Boolean set to True if it is ok for the door lock server to send PINs over the air. This attribute
       * determines the behavior of the server’s TX operation. If it is false, then it is not ok for the device
       * to send PIN in any messages over the air.
       *
       * The PIN field within any door lock cluster message shall keep the first octet unchanged and masks the
       * actual code by replacing with 0xFF. For example (PIN "1234" ): If the attribute value is True, 0x04 0x31
       * 0x32 0x33 0x34 shall be used in the PIN field in any door lock cluster message payload. If the attribute
       * value is False, 0x04 0xFF 0xFF 0xFF 0xFF shall be used.
       *
       * If the USR feature is supported by the device then this attribute shall NOT be supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.36
       */
      sendPinOverTheAir: OptionalWritableAttribute(
        50,
        TlvBoolean,
        { default: true, writeAcl: AccessLevel.Administer }
      )
    }
  });
  DoorLock2.RfidCredentialComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The number of RFID users supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.11
       */
      numberOfRfidUsersSupported: FixedAttribute(19, TlvUInt16, { default: 0 }),
      /**
       * An 8 bit value indicates the maximum length in bytes of a RFID Code on this device. The value depends on
       * the RFID code range specified by the manufacturer, if media anti-collision identifiers (UID) are used as
       * RFID code, a value of 20 (equals 10 Byte ISO 14443A UID) is recommended.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.17
       */
      maxRfidCodeLength: FixedAttribute(25, TlvUInt8),
      /**
       * An 8 bit value indicates the minimum length in bytes of a RFID Code on this device. The value depends on
       * the RFID code range specified by the manufacturer, if media anti-collision identifiers (UID) are used as
       * RFID code, a value of 8 (equals 4 Byte ISO 14443A UID) is recommended.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.18
       */
      minRfidCodeLength: FixedAttribute(26, TlvUInt8)
    }
  });
  DoorLock2.WeekDayAccessSchedulesComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The number of configurable week day schedule supported per user.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.12
       */
      numberOfWeekDaySchedulesSupportedPerUser: FixedAttribute(20, TlvUInt8, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setWeekDaySchedule: Command(
        11,
        DoorLock2.TlvSetWeekDayScheduleRequest,
        11,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getWeekDaySchedule: Command(
        12,
        DoorLock2.TlvGetWeekDayScheduleRequest,
        12,
        DoorLock2.TlvGetWeekDayScheduleResponse,
        { invokeAcl: AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearWeekDaySchedule: Command(
        13,
        DoorLock2.TlvClearWeekDayScheduleRequest,
        13,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer }
      )
    }
  });
  DoorLock2.YearDayAccessSchedulesComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The number of configurable year day schedule supported per user.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.13
       */
      numberOfYearDaySchedulesSupportedPerUser: FixedAttribute(21, TlvUInt8, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setYearDaySchedule: Command(
        14,
        DoorLock2.TlvSetYearDayScheduleRequest,
        14,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getYearDaySchedule: Command(
        15,
        DoorLock2.TlvGetYearDayScheduleRequest,
        15,
        DoorLock2.TlvGetYearDayScheduleResponse,
        { invokeAcl: AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearYearDaySchedule: Command(
        16,
        DoorLock2.TlvClearYearDayScheduleRequest,
        16,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer }
      )
    }
  });
  DoorLock2.HolidaySchedulesComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The number of holiday schedules supported for the entire door lock device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.14
       */
      numberOfHolidaySchedulesSupported: FixedAttribute(22, TlvUInt8, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setHolidaySchedule: Command(
        17,
        DoorLock2.TlvSetHolidayScheduleRequest,
        17,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getHolidaySchedule: Command(
        18,
        DoorLock2.TlvGetHolidayScheduleRequest,
        18,
        DoorLock2.TlvGetHolidayScheduleResponse,
        { invokeAcl: AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearHolidaySchedule: Command(
        19,
        DoorLock2.TlvClearHolidayScheduleRequest,
        19,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer }
      )
    }
  });
  DoorLock2.PinCredentialOrRfidCredentialComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The number of incorrect Pin codes or RFID presentment attempts a user is allowed to enter before the
       * lock will enter a lockout state. The value of this attribute is compared to all failing forms of
       * credential presentation, including Pin codes used in an Unlock Command when RequirePINforRemoteOperation
       * is set to true. Valid range is 1-255 incorrect attempts. The lockout state will be for the duration of
       * UserCodeTemporaryDisableTime. If the attribute accepts writes and an attempt to write the value 0 is
       * made, the device shall respond with CONSTRAINT_ERROR.
       *
       * The lock may reset the counter used to track incorrect credential presentations as required by internal
       * logic, environmental events, or other reasons. The lock shall reset the counter if a valid credential is
       * presented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.34
       */
      wrongCodeEntryLimit: WritableAttribute(
        48,
        TlvUInt8.bound({ min: 1 }),
        { writeAcl: AccessLevel.Administer }
      ),
      /**
       * The number of seconds that the lock shuts down following wrong code entry. Valid range is 1-255 seconds.
       * Device can shut down to lock user out for specified amount of time. (Makes it difficult to try and guess
       * a PIN for the device.) If the attribute accepts writes and an attempt to write the attribute to 0 is
       * made, the device shall respond with CONSTRAINT_ERROR.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.35
       */
      userCodeTemporaryDisableTime: WritableAttribute(
        49,
        TlvUInt8.bound({ min: 1 }),
        { writeAcl: AccessLevel.Administer }
      )
    }
  });
  DoorLock2.CredentialOverTheAirAccessAndPinCredentialComponent = ClusterFactory.Component({
    attributes: {
      /**
       * Boolean set to True if the door lock server requires that an optional PINs be included in the payload of
       * remote lock operation events like Lock, Unlock, Unlock with Timeout and Toggle in order to function.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.37
       */
      requirePinForRemoteOperation: WritableAttribute(
        51,
        TlvBoolean,
        { default: true, writeAcl: AccessLevel.Administer }
      )
    }
  });
  DoorLock2.NotificationAndPinCredentialComponent = ClusterFactory.Component({
    attributes: {
      /**
       * Event mask used to turn on and off the transmission of keypad operation events. This mask DOES NOT apply
       * to the storing of events in the event log. This mask only applies to the Operation Event Notification
       * Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.40
       */
      keypadOperationEventMask: OptionalWritableAttribute(
        65,
        TlvBitmap(TlvUInt16, DoorLock2.KeypadOperationEventMask),
        {
          default: BitsFromPartial(DoorLock2.KeypadOperationEventMask, { unknownOrManufacturerSpecificKeypadOperationEvent: true, lockSourceKeypad: true, unlockSourceKeypad: true, lockSourceKeypadErrorInvalidPin: true, lockSourceKeypadErrorInvalidSchedule: true, unlockSourceKeypadErrorInvalidCode: true, unlockSourceKeypadErrorInvalidSchedule: true, nonAccessUserOperationEventSourceKeypad: true }),
          writeAcl: AccessLevel.Administer
        }
      ),
      /**
       * Event mask used to turn on and off keypad programming events. This mask DOES NOT apply to the storing of
       * events in the event log. This mask only applies to the Programming Event Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.44
       */
      keypadProgrammingEventMask: OptionalWritableAttribute(
        69,
        TlvBitmap(TlvUInt16, DoorLock2.KeypadProgrammingEventMask),
        {
          default: BitsFromPartial(DoorLock2.KeypadProgrammingEventMask, { unknown: true, pinCodeChanged: true, pinAdded: true, pinCleared: true, pinChanged: true }),
          writeAcl: AccessLevel.Administer
        }
      )
    }
  });
  DoorLock2.NotificationComponent = ClusterFactory.Component({
    attributes: {
      /**
       * Event mask used to turn on and off the transmission of remote operation events. This mask DOES NOT apply
       * to the storing of events in the event log. This mask only applies to the Operation Event
       *
       * Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.41
       */
      remoteOperationEventMask: OptionalWritableAttribute(
        66,
        TlvBitmap(TlvUInt16, DoorLock2.RemoteOperationEventMask),
        {
          default: BitsFromPartial(DoorLock2.RemoteOperationEventMask, { unknownOrManufacturerSpecificRemoteOperationEvent: true, lockSourceRemote: true, unlockSourceRemote: true, lockSourceRemoteErrorInvalidCode: true, lockSourceRemoteErrorInvalidSchedule: true, unlockSourceRemoteErrorInvalidCode: true, unlockSourceRemoteErrorInvalidSchedule: true }),
          writeAcl: AccessLevel.Administer
        }
      ),
      /**
       * Event mask used to turn on and off manual operation events. This mask DOES NOT apply to the storing of
       * events in the event log. This mask only applies to the Operation Event Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.42
       */
      manualOperationEventMask: OptionalWritableAttribute(
        67,
        TlvBitmap(TlvUInt16, DoorLock2.ManualOperationEventMask),
        {
          default: BitsFromPartial(DoorLock2.ManualOperationEventMask, { unknownOrManufacturerSpecificManualOperationEvent: true, thumbturnLock: true, thumbturnUnlock: true, oneTouchLock: true, keyLock: true, keyUnlock: true, autoLock: true, scheduleLock: true, scheduleUnlock: true, manualLock: true, manualUnlock: true }),
          writeAcl: AccessLevel.Administer
        }
      ),
      /**
       * Event mask used to turn on and off remote programming events. This mask DOES NOT apply to the storing of
       * events in the event log. This mask only applies to the Programming Event Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.45
       */
      remoteProgrammingEventMask: OptionalWritableAttribute(
        70,
        TlvBitmap(TlvUInt16, DoorLock2.RemoteProgrammingEventMask),
        {
          default: BitsFromPartial(DoorLock2.RemoteProgrammingEventMask, { unknown: true, pinAdded: true, pinCleared: true, pinChanged: true, rfidCodeAdded: true, rfidCodeCleared: true }),
          writeAcl: AccessLevel.Administer
        }
      )
    }
  });
  DoorLock2.NotificationAndRfidCredentialComponent = ClusterFactory.Component({
    attributes: {
      /**
       * Event mask used to turn on and off RFID operation events. This mask DOES NOT apply to the storing of
       * events in the event log. This mask only applies to the Operation Event Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.43
       */
      rfidOperationEventMask: OptionalWritableAttribute(
        68,
        TlvBitmap(TlvUInt16, DoorLock2.RfidOperationEventMask),
        {
          default: BitsFromPartial(DoorLock2.RfidOperationEventMask, { unknownOrManufacturerSpecificKeypadOperationEvent: true, lockSourceRfid: true, unlockSourceRfid: true, lockSourceRfidErrorInvalidRfidId: true, lockSourceRfidErrorInvalidSchedule: true, unlockSourceRfidErrorInvalidRfidId: true, unlockSourceRfidErrorInvalidSchedule: true }),
          writeAcl: AccessLevel.Administer
        }
      ),
      /**
       * Event mask used to turn on and off RFID programming events. This mask DOES NOT apply to the storing of
       * events in the event log. This mask only applies to the Programming Event Notification Command.
       *
       * This mask DOES NOT apply to the Events mechanism of this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.3.46
       */
      rfidProgrammingEventMask: OptionalWritableAttribute(
        71,
        TlvBitmap(TlvUInt16, DoorLock2.RfidProgrammingEventMask),
        {
          default: BitsFromPartial(DoorLock2.RfidProgrammingEventMask, { unknown: true, idAdded: true, idCleared: true }),
          writeAcl: AccessLevel.Administer
        }
      )
    }
  });
  DoorLock2.PinCredentialNotUserComponent = ClusterFactory.Component({
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setPinCode: Command(
        5,
        TlvNoArguments,
        5,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getPinCode: Command(6, TlvNoArguments, 6, TlvNoArguments, { invokeAcl: AccessLevel.Administer }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearPinCode: Command(
        7,
        TlvNoArguments,
        7,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearAllPinCodes: Command(
        8,
        TlvNoArguments,
        8,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer, timed: true }
      )
    }
  });
  DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent = ClusterFactory.Component({
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setUserStatus: OptionalCommand(
        9,
        TlvNoArguments,
        9,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getUserStatus: OptionalCommand(
        10,
        TlvNoArguments,
        10,
        TlvNoArguments,
        { invokeAcl: AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setUserType: OptionalCommand(
        20,
        TlvNoArguments,
        20,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getUserType: OptionalCommand(
        21,
        TlvNoArguments,
        21,
        TlvNoArguments,
        { invokeAcl: AccessLevel.Administer }
      )
    }
  });
  DoorLock2.NotUserComponent = ClusterFactory.Component({});
  DoorLock2.RfidCredentialNotUserComponent = ClusterFactory.Component({
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      setRfidCode: Command(
        22,
        TlvNoArguments,
        22,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      getRfidCode: Command(23, TlvNoArguments, 23, TlvNoArguments, { invokeAcl: AccessLevel.Administer }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearRfidCode: Command(
        24,
        TlvNoArguments,
        24,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer, timed: true }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 5.2.4
       */
      clearAllRfidCodes: Command(
        25,
        TlvNoArguments,
        25,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer, timed: true }
      )
    }
  });
  DoorLock2.Cluster = ClusterFactory.Extensible(
    DoorLock2.Base,
    /**
     * Use this factory method to create a DoorLock cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a DoorLock cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...DoorLock2.Base,
        supportedFeatures: BitFlags(DoorLock2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, DoorLock2.DoorPositionSensorComponent, { doorPositionSensor: true });
      ClusterFactory.extend(cluster, DoorLock2.LoggingComponent, { logging: true });
      ClusterFactory.extend(cluster, DoorLock2.UserComponent, { user: true });
      ClusterFactory.extend(cluster, DoorLock2.PinCredentialComponent, { pinCredential: true });
      ClusterFactory.extend(cluster, DoorLock2.RfidCredentialComponent, { rfidCredential: true });
      ClusterFactory.extend(cluster, DoorLock2.WeekDayAccessSchedulesComponent, { weekDayAccessSchedules: true });
      ClusterFactory.extend(cluster, DoorLock2.YearDayAccessSchedulesComponent, { yearDayAccessSchedules: true });
      ClusterFactory.extend(cluster, DoorLock2.HolidaySchedulesComponent, { holidaySchedules: true });
      ClusterFactory.extend(
        cluster,
        DoorLock2.PinCredentialOrRfidCredentialComponent,
        { pinCredential: true },
        { rfidCredential: true }
      );
      ClusterFactory.extend(
        cluster,
        DoorLock2.CredentialOverTheAirAccessAndPinCredentialComponent,
        { credentialOverTheAirAccess: true, pinCredential: true }
      );
      ClusterFactory.extend(
        cluster,
        DoorLock2.NotificationAndPinCredentialComponent,
        { notification: true, pinCredential: true }
      );
      ClusterFactory.extend(cluster, DoorLock2.NotificationComponent, { notification: true });
      ClusterFactory.extend(
        cluster,
        DoorLock2.NotificationAndRfidCredentialComponent,
        { notification: true, rfidCredential: true }
      );
      ClusterFactory.extend(cluster, DoorLock2.PinCredentialNotUserComponent, { pinCredential: true, user: false });
      ClusterFactory.extend(
        cluster,
        DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent,
        { pinCredential: true, rfidCredential: true, fingerCredentials: true, user: false }
      );
      ClusterFactory.extend(cluster, DoorLock2.NotUserComponent, { user: false });
      ClusterFactory.extend(cluster, DoorLock2.RfidCredentialNotUserComponent, { rfidCredential: true, user: false });
      ClusterFactory.prevent(
        cluster,
        {
          user: true,
          pinCredential: false,
          rfidCredential: false,
          fingerCredentials: false,
          faceCredentials: false
        }
      );
      return cluster;
    }
  );
  const DPS = { doorPositionSensor: true };
  const LOG = { logging: true };
  const USR = { user: true };
  const PIN = { pinCredential: true };
  const RID = { rfidCredential: true };
  const WDSCH = { weekDayAccessSchedules: true };
  const YDSCH = { yearDayAccessSchedules: true };
  const HDSCH = { holidaySchedules: true };
  const COTA_PIN = { credentialOverTheAirAccess: true, pinCredential: true };
  const NOT_PIN = { notification: true, pinCredential: true };
  const NOT = { notification: true };
  const NOT_RID = { notification: true, rfidCredential: true };
  const PIN_NOT_USR = { pinCredential: true, user: false };
  const PIN_RID_FGP_NOT_USR = { pinCredential: true, rfidCredential: true, fingerCredentials: true, user: false };
  const RID_NOT_USR = { rfidCredential: true, user: false };
  DoorLock2.Complete = ClusterFactory.Definition({
    id: DoorLock2.Cluster.id,
    name: DoorLock2.Cluster.name,
    revision: DoorLock2.Cluster.revision,
    features: DoorLock2.Cluster.features,
    attributes: {
      ...DoorLock2.Cluster.attributes,
      doorState: ClusterFactory.AsConditional(
        DoorLock2.DoorPositionSensorComponent.attributes.doorState,
        { mandatoryIf: [DPS] }
      ),
      doorOpenEvents: ClusterFactory.AsConditional(
        DoorLock2.DoorPositionSensorComponent.attributes.doorOpenEvents,
        { optionalIf: [DPS] }
      ),
      doorClosedEvents: ClusterFactory.AsConditional(
        DoorLock2.DoorPositionSensorComponent.attributes.doorClosedEvents,
        { optionalIf: [DPS] }
      ),
      openPeriod: ClusterFactory.AsConditional(
        DoorLock2.DoorPositionSensorComponent.attributes.openPeriod,
        { optionalIf: [DPS] }
      ),
      numberOfLogRecordsSupported: ClusterFactory.AsConditional(
        DoorLock2.LoggingComponent.attributes.numberOfLogRecordsSupported,
        { mandatoryIf: [LOG] }
      ),
      numberOfTotalUsersSupported: ClusterFactory.AsConditional(
        DoorLock2.UserComponent.attributes.numberOfTotalUsersSupported,
        { mandatoryIf: [USR] }
      ),
      numberOfPinUsersSupported: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialComponent.attributes.numberOfPinUsersSupported,
        { mandatoryIf: [PIN] }
      ),
      numberOfRfidUsersSupported: ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialComponent.attributes.numberOfRfidUsersSupported,
        { mandatoryIf: [RID] }
      ),
      numberOfWeekDaySchedulesSupportedPerUser: ClusterFactory.AsConditional(
        DoorLock2.WeekDayAccessSchedulesComponent.attributes.numberOfWeekDaySchedulesSupportedPerUser,
        { mandatoryIf: [WDSCH] }
      ),
      numberOfYearDaySchedulesSupportedPerUser: ClusterFactory.AsConditional(
        DoorLock2.YearDayAccessSchedulesComponent.attributes.numberOfYearDaySchedulesSupportedPerUser,
        { mandatoryIf: [YDSCH] }
      ),
      numberOfHolidaySchedulesSupported: ClusterFactory.AsConditional(
        DoorLock2.HolidaySchedulesComponent.attributes.numberOfHolidaySchedulesSupported,
        { mandatoryIf: [HDSCH] }
      ),
      maxPinCodeLength: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialComponent.attributes.maxPinCodeLength,
        { mandatoryIf: [PIN] }
      ),
      minPinCodeLength: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialComponent.attributes.minPinCodeLength,
        { mandatoryIf: [PIN] }
      ),
      maxRfidCodeLength: ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialComponent.attributes.maxRfidCodeLength,
        { mandatoryIf: [RID] }
      ),
      minRfidCodeLength: ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialComponent.attributes.minRfidCodeLength,
        { mandatoryIf: [RID] }
      ),
      credentialRulesSupport: ClusterFactory.AsConditional(
        DoorLock2.UserComponent.attributes.credentialRulesSupport,
        { mandatoryIf: [USR] }
      ),
      numberOfCredentialsSupportedPerUser: ClusterFactory.AsConditional(
        DoorLock2.UserComponent.attributes.numberOfCredentialsSupportedPerUser,
        { mandatoryIf: [USR] }
      ),
      enableLogging: ClusterFactory.AsConditional(
        DoorLock2.LoggingComponent.attributes.enableLogging,
        { mandatoryIf: [LOG] }
      ),
      wrongCodeEntryLimit: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialOrRfidCredentialComponent.attributes.wrongCodeEntryLimit,
        { mandatoryIf: [PIN, RID] }
      ),
      userCodeTemporaryDisableTime: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialOrRfidCredentialComponent.attributes.userCodeTemporaryDisableTime,
        { mandatoryIf: [PIN, RID] }
      ),
      sendPinOverTheAir: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialComponent.attributes.sendPinOverTheAir,
        { optionalIf: [PIN] }
      ),
      requirePinForRemoteOperation: ClusterFactory.AsConditional(
        DoorLock2.CredentialOverTheAirAccessAndPinCredentialComponent.attributes.requirePinForRemoteOperation,
        { mandatoryIf: [COTA_PIN] }
      ),
      expiringUserTimeout: ClusterFactory.AsConditional(
        DoorLock2.UserComponent.attributes.expiringUserTimeout,
        { optionalIf: [USR] }
      ),
      keypadOperationEventMask: ClusterFactory.AsConditional(
        DoorLock2.NotificationAndPinCredentialComponent.attributes.keypadOperationEventMask,
        { optionalIf: [NOT_PIN] }
      ),
      remoteOperationEventMask: ClusterFactory.AsConditional(
        DoorLock2.NotificationComponent.attributes.remoteOperationEventMask,
        { optionalIf: [NOT] }
      ),
      manualOperationEventMask: ClusterFactory.AsConditional(
        DoorLock2.NotificationComponent.attributes.manualOperationEventMask,
        { optionalIf: [NOT] }
      ),
      rfidOperationEventMask: ClusterFactory.AsConditional(
        DoorLock2.NotificationAndRfidCredentialComponent.attributes.rfidOperationEventMask,
        { optionalIf: [NOT_RID] }
      ),
      keypadProgrammingEventMask: ClusterFactory.AsConditional(
        DoorLock2.NotificationAndPinCredentialComponent.attributes.keypadProgrammingEventMask,
        { optionalIf: [NOT_PIN] }
      ),
      remoteProgrammingEventMask: ClusterFactory.AsConditional(
        DoorLock2.NotificationComponent.attributes.remoteProgrammingEventMask,
        { optionalIf: [NOT] }
      ),
      rfidProgrammingEventMask: ClusterFactory.AsConditional(
        DoorLock2.NotificationAndRfidCredentialComponent.attributes.rfidProgrammingEventMask,
        { optionalIf: [NOT_RID] }
      )
    },
    commands: {
      ...DoorLock2.Cluster.commands,
      getLogRecord: ClusterFactory.AsConditional(DoorLock2.LoggingComponent.commands.getLogRecord, { mandatoryIf: [LOG] }),
      setPinCode: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialNotUserComponent.commands.setPinCode,
        { mandatoryIf: [PIN_NOT_USR] }
      ),
      getPinCode: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialNotUserComponent.commands.getPinCode,
        { mandatoryIf: [PIN_NOT_USR] }
      ),
      clearPinCode: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialNotUserComponent.commands.clearPinCode,
        { mandatoryIf: [PIN_NOT_USR] }
      ),
      clearAllPinCodes: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialNotUserComponent.commands.clearAllPinCodes,
        { mandatoryIf: [PIN_NOT_USR] }
      ),
      setUserStatus: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent.commands.setUserStatus,
        { optionalIf: [PIN_RID_FGP_NOT_USR] }
      ),
      getUserStatus: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent.commands.getUserStatus,
        { optionalIf: [PIN_RID_FGP_NOT_USR] }
      ),
      setWeekDaySchedule: ClusterFactory.AsConditional(
        DoorLock2.WeekDayAccessSchedulesComponent.commands.setWeekDaySchedule,
        { mandatoryIf: [WDSCH] }
      ),
      getWeekDaySchedule: ClusterFactory.AsConditional(
        DoorLock2.WeekDayAccessSchedulesComponent.commands.getWeekDaySchedule,
        { mandatoryIf: [WDSCH] }
      ),
      clearWeekDaySchedule: ClusterFactory.AsConditional(
        DoorLock2.WeekDayAccessSchedulesComponent.commands.clearWeekDaySchedule,
        { mandatoryIf: [WDSCH] }
      ),
      setYearDaySchedule: ClusterFactory.AsConditional(
        DoorLock2.YearDayAccessSchedulesComponent.commands.setYearDaySchedule,
        { mandatoryIf: [YDSCH] }
      ),
      getYearDaySchedule: ClusterFactory.AsConditional(
        DoorLock2.YearDayAccessSchedulesComponent.commands.getYearDaySchedule,
        { mandatoryIf: [YDSCH] }
      ),
      clearYearDaySchedule: ClusterFactory.AsConditional(
        DoorLock2.YearDayAccessSchedulesComponent.commands.clearYearDaySchedule,
        { mandatoryIf: [YDSCH] }
      ),
      setHolidaySchedule: ClusterFactory.AsConditional(
        DoorLock2.HolidaySchedulesComponent.commands.setHolidaySchedule,
        { mandatoryIf: [HDSCH] }
      ),
      getHolidaySchedule: ClusterFactory.AsConditional(
        DoorLock2.HolidaySchedulesComponent.commands.getHolidaySchedule,
        { mandatoryIf: [HDSCH] }
      ),
      clearHolidaySchedule: ClusterFactory.AsConditional(
        DoorLock2.HolidaySchedulesComponent.commands.clearHolidaySchedule,
        { mandatoryIf: [HDSCH] }
      ),
      setUserType: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent.commands.setUserType,
        { optionalIf: [PIN_RID_FGP_NOT_USR] }
      ),
      getUserType: ClusterFactory.AsConditional(
        DoorLock2.PinCredentialAndRfidCredentialAndFingerCredentialsNotUserComponent.commands.getUserType,
        { optionalIf: [PIN_RID_FGP_NOT_USR] }
      ),
      setRfidCode: ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialNotUserComponent.commands.setRfidCode,
        { mandatoryIf: [RID_NOT_USR] }
      ),
      getRfidCode: ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialNotUserComponent.commands.getRfidCode,
        { mandatoryIf: [RID_NOT_USR] }
      ),
      clearRfidCode: ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialNotUserComponent.commands.clearRfidCode,
        { mandatoryIf: [RID_NOT_USR] }
      ),
      clearAllRfidCodes: ClusterFactory.AsConditional(
        DoorLock2.RfidCredentialNotUserComponent.commands.clearAllRfidCodes,
        { mandatoryIf: [RID_NOT_USR] }
      ),
      setUser: ClusterFactory.AsConditional(DoorLock2.UserComponent.commands.setUser, { mandatoryIf: [USR] }),
      getUser: ClusterFactory.AsConditional(DoorLock2.UserComponent.commands.getUser, { mandatoryIf: [USR] }),
      clearUser: ClusterFactory.AsConditional(DoorLock2.UserComponent.commands.clearUser, { mandatoryIf: [USR] }),
      setCredential: ClusterFactory.AsConditional(DoorLock2.UserComponent.commands.setCredential, { mandatoryIf: [USR] }),
      getCredentialStatus: ClusterFactory.AsConditional(
        DoorLock2.UserComponent.commands.getCredentialStatus,
        { mandatoryIf: [USR] }
      ),
      clearCredential: ClusterFactory.AsConditional(
        DoorLock2.UserComponent.commands.clearCredential,
        { mandatoryIf: [USR] }
      )
    },
    events: {
      ...DoorLock2.Cluster.events,
      doorStateChange: ClusterFactory.AsConditional(
        DoorLock2.DoorPositionSensorComponent.events.doorStateChange,
        { mandatoryIf: [DPS] }
      ),
      lockUserChange: ClusterFactory.AsConditional(DoorLock2.UserComponent.events.lockUserChange, { mandatoryIf: [USR] })
    }
  });
})(DoorLock || (DoorLock = {}));
const DoorLockCluster = DoorLock.Cluster;
export {
  DoorLock,
  DoorLockCluster
};
