// @project-chip/matter.js/dist/esm/cluster/definitions/EthernetNetworkDiagnosticsCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { OptionalAttribute, Attribute, Command, TlvNoResponse, AccessLevel } from "../../cluster/Cluster.js";
import { TlvEnum, TlvUInt64 } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var EthernetNetworkDiagnostics;
((EthernetNetworkDiagnostics2) => {
  let PHYRate;
  ((PHYRate2) => {
    PHYRate2[PHYRate2["Rate10M"] = 0] = "Rate10M";
    PHYRate2[PHYRate2["Rate100M"] = 1] = "Rate100M";
    PHYRate2[PHYRate2["Rate1G"] = 2] = "Rate1G";
    PHYRate2[PHYRate2["Rate25G"] = 3] = "Rate25G";
    PHYRate2[PHYRate2["Rate5G"] = 4] = "Rate5G";
    PHYRate2[PHYRate2["Rate10G"] = 5] = "Rate10G";
    PHYRate2[PHYRate2["Rate40G"] = 6] = "Rate40G";
    PHYRate2[PHYRate2["Rate100G"] = 7] = "Rate100G";
    PHYRate2[PHYRate2["Rate200G"] = 8] = "Rate200G";
    PHYRate2[PHYRate2["Rate400G"] = 9] = "Rate400G";
  })(PHYRate = EthernetNetworkDiagnostics2.PHYRate || (EthernetNetworkDiagnostics2.PHYRate = {}));
  let Feature;
  ((Feature2) => {
    Feature2["PacketCounts"] = "PacketCounts";
    Feature2["ErrorCounts"] = "ErrorCounts";
  })(Feature = EthernetNetworkDiagnostics2.Feature || (EthernetNetworkDiagnostics2.Feature = {}));
  EthernetNetworkDiagnostics2.Base = ClusterFactory.Definition({
    id: 55,
    name: "EthernetNetworkDiagnostics",
    revision: 1,
    features: {
      /**
       * PacketCounts
       *
       * Node makes available the counts for the number of received and transmitted packets on the ethernet
       * interface.
       */
      packetCounts: BitFlag(0),
      /**
       * ErrorCounts
       *
       * Node makes available the counts for the number of errors that have occurred during the reception and
       * transmission of packets on the ethernet interface.
       */
      errorCounts: BitFlag(1)
    },
    attributes: {
      /**
       * The PHYRate attribute shall indicate the current nominal, usable speed at the top of the physical layer
       * of the Node. A value of null shall indicate that the interface is not currently configured or
       * operational.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.15.6.1
       */
      phyRate: OptionalAttribute(0, TlvNullable(TlvEnum()), { default: null }),
      /**
       * The FullDuplex attribute shall indicate if the Node is currently utilizing the full-duplex operating
       * mode. A value of null shall indicate that the interface is not currently configured or operational.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.15.6.2
       */
      fullDuplex: OptionalAttribute(1, TlvNullable(TlvBoolean), { default: null }),
      /**
       * The CarrierDetect attribute shall indicate the value of the Carrier Detect control signal present on the
       * ethernet network interface. A value of null shall indicate that the interface is not currently
       * configured or operational.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.15.6.8
       */
      carrierDetect: OptionalAttribute(7, TlvNullable(TlvBoolean), { omitChanges: true, default: null }),
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 11.15.6
       */
      timeSinceReset: OptionalAttribute(8, TlvUInt64, { omitChanges: true, default: 0 })
    }
  });
  EthernetNetworkDiagnostics2.PacketCountsComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The PacketRxCount attribute shall indicate the number of packets that have been received on the ethernet
       * network interface. The PacketRxCount attribute shall be reset to 0 upon a reboot of the Node.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.15.6.3
       */
      packetRxCount: Attribute(2, TlvUInt64, { default: 0 }),
      /**
       * The PacketTxCount attribute shall indicate the number of packets that have been successfully transferred
       * on the ethernet network interface. The PacketTxCount attribute shall be reset to 0 upon a reboot of the
       * Node.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.15.6.4
       */
      packetTxCount: Attribute(3, TlvUInt64, { omitChanges: true, default: 0 })
    }
  });
  EthernetNetworkDiagnostics2.ErrorCountsComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The TxErrCount attribute shall indicate the number of failed packet transmissions that have occurred on
       * the ethernet network interface. The TxErrCount attribute shall be reset to 0 upon a reboot of the Node.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.15.6.5
       */
      txErrCount: Attribute(4, TlvUInt64, { omitChanges: true, default: 0 }),
      /**
       * The CollisionCount attribute shall indicate the number of collisions that have occurred while attempting
       * to transmit a packet on the ethernet network interface. The CollisionCount attribute shall be reset to 0
       * upon a reboot of the Node.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.15.6.6
       */
      collisionCount: Attribute(5, TlvUInt64, { omitChanges: true, default: 0 }),
      /**
       * The OverrunCount attribute shall indicate the number of packets dropped either at ingress or egress, due
       * to lack of buffer memory to retain all packets on the ethernet network interface. The OverrunCount
       * attribute shall be reset to 0 upon a reboot of the Node.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.15.6.7
       */
      overrunCount: Attribute(6, TlvUInt64, { omitChanges: true, default: 0 })
    }
  });
  EthernetNetworkDiagnostics2.PacketCountsOrErrorCountsComponent = ClusterFactory.Component({
    commands: {
      /**
       * Reception of this command shall reset the following attributes to 0:
       *
       *   • PacketRxCount
       *
       *   • PacketTxCount
       *
       *   • TxErrCount
       *
       *   • CollisionCount
       *
       *   • OverrunCount
       *
       * This command has no associated data.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.15.7.1
       */
      resetCounts: Command(0, TlvNoArguments, 0, TlvNoResponse, { invokeAcl: AccessLevel.Manage })
    }
  });
  EthernetNetworkDiagnostics2.Cluster = ClusterFactory.Extensible(
    EthernetNetworkDiagnostics2.Base,
    /**
     * Use this factory method to create an EthernetNetworkDiagnostics cluster with support for optional features.
     * Include each {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns an EthernetNetworkDiagnostics cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...EthernetNetworkDiagnostics2.Base,
        supportedFeatures: BitFlags(EthernetNetworkDiagnostics2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, EthernetNetworkDiagnostics2.PacketCountsComponent, { packetCounts: true });
      ClusterFactory.extend(cluster, EthernetNetworkDiagnostics2.ErrorCountsComponent, { errorCounts: true });
      ClusterFactory.extend(
        cluster,
        EthernetNetworkDiagnostics2.PacketCountsOrErrorCountsComponent,
        { packetCounts: true },
        { errorCounts: true }
      );
      return cluster;
    }
  );
  const PKTCNT = { packetCounts: true };
  const ERRCNT = { errorCounts: true };
  EthernetNetworkDiagnostics2.Complete = ClusterFactory.Definition({
    id: EthernetNetworkDiagnostics2.Cluster.id,
    name: EthernetNetworkDiagnostics2.Cluster.name,
    revision: EthernetNetworkDiagnostics2.Cluster.revision,
    features: EthernetNetworkDiagnostics2.Cluster.features,
    attributes: {
      ...EthernetNetworkDiagnostics2.Cluster.attributes,
      packetRxCount: ClusterFactory.AsConditional(
        EthernetNetworkDiagnostics2.PacketCountsComponent.attributes.packetRxCount,
        { mandatoryIf: [PKTCNT] }
      ),
      packetTxCount: ClusterFactory.AsConditional(
        EthernetNetworkDiagnostics2.PacketCountsComponent.attributes.packetTxCount,
        { mandatoryIf: [PKTCNT] }
      ),
      txErrCount: ClusterFactory.AsConditional(
        EthernetNetworkDiagnostics2.ErrorCountsComponent.attributes.txErrCount,
        { mandatoryIf: [ERRCNT] }
      ),
      collisionCount: ClusterFactory.AsConditional(
        EthernetNetworkDiagnostics2.ErrorCountsComponent.attributes.collisionCount,
        { mandatoryIf: [ERRCNT] }
      ),
      overrunCount: ClusterFactory.AsConditional(
        EthernetNetworkDiagnostics2.ErrorCountsComponent.attributes.overrunCount,
        { mandatoryIf: [ERRCNT] }
      )
    },
    commands: {
      resetCounts: ClusterFactory.AsConditional(
        EthernetNetworkDiagnostics2.PacketCountsOrErrorCountsComponent.commands.resetCounts,
        { mandatoryIf: [PKTCNT, ERRCNT] }
      )
    }
  });
})(EthernetNetworkDiagnostics || (EthernetNetworkDiagnostics = {}));
const EthernetNetworkDiagnosticsCluster = EthernetNetworkDiagnostics.Cluster;
export {
  EthernetNetworkDiagnostics,
  EthernetNetworkDiagnosticsCluster
};
