// @project-chip/matter.js/dist/esm/cluster/definitions/FanControlCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { WritableAttribute, Attribute, FixedAttribute } from "../../cluster/Cluster.js";
import { TlvEnum, TlvUInt8, TlvBitmap } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
var FanControl;
((FanControl2) => {
  let FanMode;
  ((FanMode2) => {
    FanMode2[FanMode2["Off"] = 0] = "Off";
    FanMode2[FanMode2["Low"] = 1] = "Low";
    FanMode2[FanMode2["Medium"] = 2] = "Medium";
    FanMode2[FanMode2["High"] = 3] = "High";
    FanMode2[FanMode2["On"] = 4] = "On";
    FanMode2[FanMode2["Auto"] = 5] = "Auto";
    FanMode2[FanMode2["Smart"] = 6] = "Smart";
  })(FanMode = FanControl2.FanMode || (FanControl2.FanMode = {}));
  let FanModeSequence;
  ((FanModeSequence2) => {
    FanModeSequence2[FanModeSequence2["OffLowMedHigh"] = 0] = "OffLowMedHigh";
    FanModeSequence2[FanModeSequence2["OffLowHigh"] = 1] = "OffLowHigh";
    FanModeSequence2[FanModeSequence2["OffLowMedHighAuto"] = 2] = "OffLowMedHighAuto";
    FanModeSequence2[FanModeSequence2["OffLowHighAuto"] = 3] = "OffLowHighAuto";
    FanModeSequence2[FanModeSequence2["OffOnAuto"] = 4] = "OffOnAuto";
    FanModeSequence2[FanModeSequence2["OffOn"] = 5] = "OffOn";
  })(FanModeSequence = FanControl2.FanModeSequence || (FanControl2.FanModeSequence = {}));
  FanControl2.RockSupport = { rockLeftRight: BitFlag(0), rockUpDown: BitFlag(1), rockRound: BitFlag(2) };
  FanControl2.RockSetting = { rockLeftRight: BitFlag(0), rockUpDown: BitFlag(1), rockRound: BitFlag(2) };
  FanControl2.WindSupport = { sleepWind: BitFlag(0), naturalWind: BitFlag(1) };
  FanControl2.WindSetting = { sleepWind: BitFlag(0), naturalWind: BitFlag(1) };
  let Feature;
  ((Feature2) => {
    Feature2["MultiSpeed"] = "MultiSpeed";
    Feature2["Auto"] = "Auto";
    Feature2["Rocking"] = "Rocking";
    Feature2["Wind"] = "Wind";
  })(Feature = FanControl2.Feature || (FanControl2.Feature = {}));
  FanControl2.Base = ClusterFactory.Definition({
    id: 514,
    name: "FanControl",
    revision: 2,
    features: {
      /**
       * MultiSpeed
       *
       * 1-100 speeds
       */
      multiSpeed: BitFlag(0),
      /**
       * Auto
       *
       * Automatic mode supported for fan speed
       */
      auto: BitFlag(1),
      /**
       * Rocking
       *
       * Rocking movement supported
       */
      rocking: BitFlag(2),
      /**
       * Wind
       *
       * Wind emulation supported
       */
      wind: BitFlag(3)
    },
    attributes: {
      /**
       * This attribute shall indicate the current speed mode of the fan. This attribute may be written by the
       * client to indicate a new speed mode of the fan. This attribute shall be set to one of the values in the
       * table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.1
       */
      fanMode: WritableAttribute(0, TlvEnum(), { persistent: true, default: 0 /* Off */ }),
      /**
       * This indicates the fan speed ranges that shall be supported.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.2
       */
      fanModeSequence: WritableAttribute(
        1,
        TlvEnum(),
        { persistent: true, default: 2 /* OffLowMedHighAuto */ }
      ),
      /**
       * This attribute shall indicate the speed setting for the fan. This attribute may be written by the client
       * to indicate a new fan speed. If the client writes null to this attribute, the attribute value shall NOT
       * change. If this is set to 0, the server shall set the FanMode attribute value to Off.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.3
       */
      percentSetting: WritableAttribute(2, TlvNullable(TlvUInt8.bound({ max: 100 })), { default: 0 }),
      /**
       * This attribute shall indicate the actual currently operating fan speed, or zero to indicate that the fan
       * is off. See Section 4.4.6.3.1 for more details.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.4
       */
      percentCurrent: Attribute(3, TlvUInt8.bound({ max: 100 }), { default: 0 })
    }
  });
  FanControl2.MultiSpeedComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This attribute shall indicate that the fan has one speed (value of 1) or the maximum speed, if the fan
       * is capable of multiple speeds.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.5
       */
      speedMax: FixedAttribute(4, TlvUInt8.bound({ min: 1, max: 100 }), { default: 1 }),
      /**
       * This attribute shall indicate the speed setting for the fan. This attribute may be written by the client
       * to indicate a new fan speed. If the client writes null to this attribute, the attribute value shall NOT
       * change. If this is set to 0, the server shall set the FanMode attribute value to Off. Please see the
       * Section 4.4.6.6.1 for details on other values.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.6
       */
      speedSetting: WritableAttribute(5, TlvNullable(TlvUInt8.bound({ max: 100 })), { default: 0 }),
      /**
       * This attribute shall indicate the actual currently operating fan speed, or zero to indicate that the fan
       * is off.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.7
       */
      speedCurrent: Attribute(6, TlvUInt8.bound({ max: 100 }), { default: 0 })
    }
  });
  FanControl2.RockingComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This attribute is a bitmap that indicates what rocking motions the server supports. The bitmap is shown
       * in the table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.8
       */
      rockSupport: FixedAttribute(7, TlvBitmap(TlvUInt8, FanControl2.RockSupport)),
      /**
       * This attribute is a bitmap that indicates the current active fan rocking motion settings. Each bit shall
       * only be set to 1, if the corresponding bit in the RockSupport attribute is set to 1, otherwise a status
       * code of CONSTRAINT_ERROR shall be returned.
       *
       * If a combination of supported bits is set by the client, and the server does not support the
       * combination, the lowest supported single bit in the combination shall be set and active, and all other
       * bits shall indicate zero.
       *
       * For example: If RockUpDown and RockRound are both set, but this combination is not possible, then only
       * RockUpDown becomes active.
       *
       * The bitmap is shown in the table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.9
       */
      rockSetting: WritableAttribute(8, TlvBitmap(TlvUInt8, FanControl2.RockSetting))
    }
  });
  FanControl2.WindComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This attribute is a bitmap that indicates what wind modes the server supports. At least one wind mode
       * bit shall be set. The bitmap is shown in the table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.10
       */
      windSupport: FixedAttribute(9, TlvBitmap(TlvUInt8, FanControl2.WindSupport)),
      /**
       * This attribute is a bitmap that indicates the current active fan wind feature settings. Each bit shall
       * only be set to 1, if the corresponding bit in the WindSupport attribute is set to 1, otherwise a status
       * code of CONSTRAINT_ERROR shall be returned.
       *
       * If a combination of supported bits is set by the client, and the server does not support the
       * combination, the lowest supported single bit in the combination shall be set and active, and all other
       * bits shall indicate zero.
       *
       * For example: If Sleep Wind and Natural Wind are set, but this combination is not possible, then only
       * Sleep Wind becomes active.
       *
       * The bitmap is shown in the table below.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 4.4.6.11
       */
      windSetting: WritableAttribute(10, TlvBitmap(TlvUInt8, FanControl2.WindSetting))
    }
  });
  FanControl2.Cluster = ClusterFactory.Extensible(
    FanControl2.Base,
    /**
     * Use this factory method to create a FanControl cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a FanControl cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...FanControl2.Base,
        supportedFeatures: BitFlags(FanControl2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, FanControl2.MultiSpeedComponent, { multiSpeed: true });
      ClusterFactory.extend(cluster, FanControl2.RockingComponent, { rocking: true });
      ClusterFactory.extend(cluster, FanControl2.WindComponent, { wind: true });
      return cluster;
    }
  );
  const SPD = { multiSpeed: true };
  const RCK = { rocking: true };
  const WND = { wind: true };
  FanControl2.Complete = ClusterFactory.Definition({
    id: FanControl2.Cluster.id,
    name: FanControl2.Cluster.name,
    revision: FanControl2.Cluster.revision,
    features: FanControl2.Cluster.features,
    attributes: {
      ...FanControl2.Cluster.attributes,
      speedMax: ClusterFactory.AsConditional(FanControl2.MultiSpeedComponent.attributes.speedMax, { mandatoryIf: [SPD] }),
      speedSetting: ClusterFactory.AsConditional(
        FanControl2.MultiSpeedComponent.attributes.speedSetting,
        { mandatoryIf: [SPD] }
      ),
      speedCurrent: ClusterFactory.AsConditional(
        FanControl2.MultiSpeedComponent.attributes.speedCurrent,
        { mandatoryIf: [SPD] }
      ),
      rockSupport: ClusterFactory.AsConditional(FanControl2.RockingComponent.attributes.rockSupport, { mandatoryIf: [RCK] }),
      rockSetting: ClusterFactory.AsConditional(FanControl2.RockingComponent.attributes.rockSetting, { mandatoryIf: [RCK] }),
      windSupport: ClusterFactory.AsConditional(FanControl2.WindComponent.attributes.windSupport, { mandatoryIf: [WND] }),
      windSetting: ClusterFactory.AsConditional(FanControl2.WindComponent.attributes.windSetting, { mandatoryIf: [WND] })
    }
  });
})(FanControl || (FanControl = {}));
const FanControlCluster = FanControl.Cluster;
export {
  FanControl,
  FanControlCluster
};
