// @project-chip/matter.js/dist/esm/cluster/definitions/FaultInjectionCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { Command, TlvNoResponse, AccessLevel } from "../../cluster/Cluster.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvEnum, TlvUInt32, TlvUInt8 } from "../../tlv/TlvNumber.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
var FaultInjection;
((FaultInjection2) => {
  let FaultType;
  ((FaultType2) => {
    FaultType2[FaultType2["Unspecified"] = 0] = "Unspecified";
    FaultType2[FaultType2["SystemFault"] = 1] = "SystemFault";
    FaultType2[FaultType2["InetFault"] = 2] = "InetFault";
    FaultType2[FaultType2["ChipFault"] = 3] = "ChipFault";
    FaultType2[FaultType2["CertFault"] = 4] = "CertFault";
  })(FaultType = FaultInjection2.FaultType || (FaultInjection2.FaultType = {}));
  FaultInjection2.TlvFailAtFaultRequest = TlvObject({
    type: TlvField(0, TlvEnum()),
    id: TlvField(1, TlvUInt32),
    numCallsToSkip: TlvField(2, TlvUInt32),
    numCallsToFail: TlvField(3, TlvUInt32),
    takeMutex: TlvField(4, TlvBoolean)
  });
  FaultInjection2.TlvFailRandomlyAtFaultRequest = TlvObject({
    type: TlvField(0, TlvEnum()),
    id: TlvField(1, TlvUInt32),
    percentage: TlvField(2, TlvUInt8)
  });
  FaultInjection2.Cluster = ClusterFactory.Definition({
    id: 4294048774,
    name: "FaultInjection",
    revision: 1,
    commands: {
      failAtFault: Command(0, FaultInjection2.TlvFailAtFaultRequest, 0, TlvNoResponse, { invokeAcl: AccessLevel.Manage }),
      failRandomlyAtFault: Command(
        1,
        FaultInjection2.TlvFailRandomlyAtFaultRequest,
        1,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Manage }
      )
    }
  });
})(FaultInjection || (FaultInjection = {}));
const FaultInjectionCluster = FaultInjection.Cluster;
export {
  FaultInjection,
  FaultInjectionCluster
};
