// @project-chip/matter.js/dist/esm/cluster/definitions/GroupsCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { FixedAttribute, Command, AccessLevel, TlvNoResponse } from "../../cluster/Cluster.js";
import { TlvUInt8, TlvBitmap, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvGroupId } from "../../datatype/GroupId.js";
import { TlvString } from "../../tlv/TlvString.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var Groups;
((Groups2) => {
  Groups2.NameSupport = {
    /**
     * The ability to store a name for a group.
     */
    nameSupport: BitFlag(7)
  };
  Groups2.TlvAddGroupRequest = TlvObject({
    groupId: TlvField(0, TlvGroupId),
    groupName: TlvField(1, TlvString.bound({ maxLength: 16 }))
  });
  Groups2.TlvAddGroupResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    groupId: TlvField(1, TlvGroupId)
  });
  Groups2.TlvViewGroupRequest = TlvObject({ groupId: TlvField(0, TlvGroupId) });
  Groups2.TlvViewGroupResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    groupId: TlvField(1, TlvGroupId),
    groupName: TlvField(2, TlvString.bound({ maxLength: 16 }))
  });
  Groups2.TlvGetGroupMembershipRequest = TlvObject({ groupList: TlvField(0, TlvArray(TlvGroupId)) });
  Groups2.TlvGetGroupMembershipResponse = TlvObject({
    capacity: TlvField(0, TlvNullable(TlvUInt8)),
    groupList: TlvField(1, TlvArray(TlvGroupId))
  });
  Groups2.TlvRemoveGroupRequest = TlvObject({ groupId: TlvField(0, TlvGroupId) });
  Groups2.TlvRemoveGroupResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    groupId: TlvField(1, TlvGroupId)
  });
  Groups2.TlvAddGroupIfIdentifyingRequest = TlvObject({
    groupId: TlvField(0, TlvGroupId),
    groupName: TlvField(1, TlvString.bound({ maxLength: 16 }))
  });
  let Feature;
  ((Feature2) => {
    Feature2["GroupNames"] = "GroupNames";
  })(Feature = Groups2.Feature || (Groups2.Feature = {}));
  Groups2.Base = ClusterFactory.Definition({
    id: 4,
    name: "Groups",
    revision: 4,
    features: {
      /**
       * GroupNames
       *
       * The ability to store a name for a group.
       */
      groupNames: BitFlag(0)
    },
    attributes: {
      /**
       * This attribute provides legacy, read-only access to whether the Group Names feature is supported. The
       * most significant bit, bit 7, shall be equal to bit 0 of the FeatureMap attribute. All other bits shall
       * be 0.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.6.1
       */
      nameSupport: FixedAttribute(0, TlvBitmap(TlvUInt8, Groups2.NameSupport))
    },
    commands: {
      /**
       * The AddGroup command allows a client to add group membership in a particular group for the server
       * endpoint.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.1
       */
      addGroup: Command(0, Groups2.TlvAddGroupRequest, 0, Groups2.TlvAddGroupResponse, { invokeAcl: AccessLevel.Manage }),
      /**
       * The ViewGroup command allows a client to request that the server responds with a ViewGroupResponse
       * command containing the name string for a particular group.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.2
       */
      viewGroup: Command(1, Groups2.TlvViewGroupRequest, 1, Groups2.TlvViewGroupResponse),
      /**
       * The GetGroupMembership command allows a client to inquire about the group membership of the server
       * endpoint, in a number of ways.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.3
       */
      getGroupMembership: Command(2, Groups2.TlvGetGroupMembershipRequest, 2, Groups2.TlvGetGroupMembershipResponse),
      /**
       * The RemoveGroup command allows a client to request that the server removes the membership for the server
       * endpoint, if any, in a particular group.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.4
       */
      removeGroup: Command(
        3,
        Groups2.TlvRemoveGroupRequest,
        3,
        Groups2.TlvRemoveGroupResponse,
        { invokeAcl: AccessLevel.Manage }
      ),
      /**
       * The RemoveAllGroups command allows a client to direct the server to remove all group associations for
       * the server endpoint.
       *
       * The RemoveAllGroups command has no data fields.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.5
       */
      removeAllGroups: Command(4, TlvNoArguments, 4, TlvNoResponse, { invokeAcl: AccessLevel.Manage }),
      /**
       * The AddGroupIfIdentifying command allows a client to add group membership in a particular group for the
       * server endpoint, on condition that the endpoint is identifying itself. Identifying functionality is
       * controlled using the Identify cluster, (see Identify).
       *
       * This command might be used to assist configuring group membership in the absence of a commissioning tool.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.3.7.6
       */
      addGroupIfIdentifying: Command(
        5,
        Groups2.TlvAddGroupIfIdentifyingRequest,
        5,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Manage }
      )
    }
  });
  Groups2.Cluster = ClusterFactory.Extensible(
    { ...Groups2.Base, supportedFeatures: { groupNames: true } },
    /**
     * Use this factory method to create a Groups cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a Groups cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...Groups2.Base,
        supportedFeatures: BitFlags(Groups2.Base.features, ...features)
      });
      return cluster;
    }
  );
})(Groups || (Groups = {}));
const GroupsCluster = Groups.Cluster;
export {
  Groups,
  GroupsCluster
};
