// @project-chip/matter.js/dist/esm/cluster/definitions/IdentifyCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import {
  WritableAttribute,
  Attribute,
  Command,
  TlvNoResponse,
  AccessLevel,
  OptionalCommand
} from "../../cluster/Cluster.js";
import { TlvUInt16, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var Identify;
((Identify2) => {
  let IdentifyType;
  ((IdentifyType2) => {
    IdentifyType2[IdentifyType2["None"] = 0] = "None";
    IdentifyType2[IdentifyType2["LightOutput"] = 1] = "LightOutput";
    IdentifyType2[IdentifyType2["VisibleIndicator"] = 2] = "VisibleIndicator";
    IdentifyType2[IdentifyType2["AudibleBeep"] = 3] = "AudibleBeep";
    IdentifyType2[IdentifyType2["Display"] = 4] = "Display";
    IdentifyType2[IdentifyType2["Actuator"] = 5] = "Actuator";
  })(IdentifyType = Identify2.IdentifyType || (Identify2.IdentifyType = {}));
  Identify2.TlvIdentifyRequest = TlvObject({ identifyTime: TlvField(0, TlvUInt16) });
  let EffectIdentifier;
  ((EffectIdentifier2) => {
    EffectIdentifier2[EffectIdentifier2["Blink"] = 0] = "Blink";
    EffectIdentifier2[EffectIdentifier2["Breathe"] = 1] = "Breathe";
    EffectIdentifier2[EffectIdentifier2["Okay"] = 2] = "Okay";
    EffectIdentifier2[EffectIdentifier2["ChannelChange"] = 11] = "ChannelChange";
    EffectIdentifier2[EffectIdentifier2["FinishEffect"] = 254] = "FinishEffect";
    EffectIdentifier2[EffectIdentifier2["StopEffect"] = 255] = "StopEffect";
  })(EffectIdentifier = Identify2.EffectIdentifier || (Identify2.EffectIdentifier = {}));
  let EffectVariant;
  ((EffectVariant2) => {
    EffectVariant2[EffectVariant2["Default"] = 0] = "Default";
  })(EffectVariant = Identify2.EffectVariant || (Identify2.EffectVariant = {}));
  Identify2.TlvTriggerEffectRequest = TlvObject({
    /**
     * This field specifies the identify effect to use. All values of the EffectIdentifier shall be supported.
     * Implementors may deviate from the example light effects in the table below, but they SHOULD indicate during
     * testing how they handle each effect.
     *
     * This field shall contain one of the non-reserved values listed below.
     *
     * Table 3. Values of the EffectIdentifier Field of the TriggerEffect Command
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.2.6.3.1
     */
    effectIdentifier: TlvField(0, TlvEnum()),
    /**
     * This field is used to indicate which variant of the effect, indicated in the EffectIdentifier field, SHOULD
     * be triggered. If a device does not support the given variant, it shall use the default variant. This field
     * shall contain one of the values listed below:
     *
     * Table 4. Values of the EffectVariant Field of the TriggerEffect Command
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.2.6.3.2
     */
    effectVariant: TlvField(1, TlvEnum())
  });
  Identify2.TlvIdentifyQueryResponse = TlvObject({
    /**
     * This field contains the current value of the IdentifyTime attribute, and specifies the length of time, in
     * seconds, that the device will continue to identify itself.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.2.6.4.1
     */
    timeout: TlvField(0, TlvUInt16)
  });
  let Feature;
  ((Feature2) => {
    Feature2["Query"] = "Query";
  })(Feature = Identify2.Feature || (Identify2.Feature = {}));
  Identify2.Base = ClusterFactory.Definition({
    id: 3,
    name: "Identify",
    revision: 4,
    features: {
      /**
       * Query
       *
       * Multicast query for identification state
       */
      query: BitFlag(0)
    },
    attributes: {
      /**
       * This attribute specifies the remaining length of time, in seconds, that the endpoint will continue to
       * identify itself.
       *
       * If this attribute is set to a value other than 0 then the device shall enter its identification state,
       * in order to indicate to an observer which of several nodes and/or endpoints it is. It is RECOMMENDED
       * that this state consists of flashing a light with a period of 0.5 seconds. The IdentifyTime attribute
       * shall be decremented every second while in this state.
       *
       * If this attribute reaches or is set to the value 0 then the device shall terminate its identification
       * state.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.2.5.1
       */
      identifyTime: WritableAttribute(0, TlvUInt16, { default: 0 }),
      /**
       * This attribute specifies how the identification state is presented to the user. This field shall contain
       * one of the values listed below:
       *
       * Table 2. Values of the IdentifyType attribute
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.2.5.2
       */
      identifyType: Attribute(1, TlvEnum(), { default: 0 /* None */ })
    },
    commands: {
      /**
       * This command starts or stops the receiving device identifying itself.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.2.6.1
       */
      identify: Command(0, Identify2.TlvIdentifyRequest, 0, TlvNoResponse, { invokeAcl: AccessLevel.Manage }),
      /**
       * This command allows the support of feedback to the user, such as a certain light effect. It is used to
       * allow an implementation to provide visual feedback to the user under certain circumstances such as a
       * color light turning green when it has successfully connected to a network. The use of this command and
       * the effects themselves are entirely up to the implementer to use whenever a visual feedback is useful
       * but it is not the same as and does not replace the identify mechanism used during commissioning.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.2.6.3
       */
      triggerEffect: OptionalCommand(
        64,
        Identify2.TlvTriggerEffectRequest,
        64,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Manage }
      )
    }
  });
  Identify2.QueryComponent = ClusterFactory.Component({
    commands: {
      /**
       * This command allows the sending device to request the target or targets to respond if they are currently
       * identifying themselves.
       *
       * This command has no data fields.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.2.6.2
       */
      identifyQuery: Command(
        1,
        TlvNoArguments,
        0,
        Identify2.TlvIdentifyQueryResponse,
        { invokeAcl: AccessLevel.Manage }
      )
    }
  });
  Identify2.Cluster = ClusterFactory.Extensible(
    Identify2.Base,
    /**
     * Use this factory method to create an Identify cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns an Identify cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...Identify2.Base,
        supportedFeatures: BitFlags(Identify2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, Identify2.QueryComponent, { query: true });
      return cluster;
    }
  );
  const QRY = { query: true };
  Identify2.Complete = ClusterFactory.Definition({
    id: Identify2.Cluster.id,
    name: Identify2.Cluster.name,
    revision: Identify2.Cluster.revision,
    features: Identify2.Cluster.features,
    attributes: Identify2.Cluster.attributes,
    commands: {
      ...Identify2.Cluster.commands,
      identifyQuery: ClusterFactory.AsConditional(Identify2.QueryComponent.commands.identifyQuery, { mandatoryIf: [QRY] })
    }
  });
})(Identify || (Identify = {}));
const IdentifyCluster = Identify.Cluster;
export {
  Identify,
  IdentifyCluster
};
