// @project-chip/matter.js/dist/esm/cluster/definitions/LocalizationConfigurationCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { WritableAttribute, AccessLevel, FixedAttribute } from "../../cluster/Cluster.js";
import { TlvString } from "../../tlv/TlvString.js";
import { TlvArray } from "../../tlv/TlvArray.js";
var LocalizationConfiguration;
((LocalizationConfiguration2) => {
  LocalizationConfiguration2.Cluster = ClusterFactory.Definition({
    id: 43,
    name: "LocalizationConfiguration",
    revision: 1,
    attributes: {
      /**
       * The ActiveLocale attribute shall represent the locale that the Node is currently configured to use when
       * conveying information. The ActiveLocale attribute shall be a Language Tag as defined by BCP47
       * [https://tools.ietf.org/rfc/bcp/bcp47.txt]. The ActiveLocale attribute shall have a default value
       * assigned by the Vendor and shall be a value contained within the SupportedLocales attribute list.
       *
       * An attempt to write a value to ActiveLocale that is not present in SupportedLocales shall result in a
       * CONSTRAINT_ERROR error.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.3.4.1
       */
      activeLocale: WritableAttribute(
        0,
        TlvString.bound({ maxLength: 35 }),
        { persistent: true, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The SupportedLocales attribute shall represent a list of locale strings that are valid values for the
       * ActiveLocale attribute. The list shall NOT contain any duplicate entries. The ordering of items within
       * the list SHOULD NOT express any meaning.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.3.4.2
       */
      supportedLocales: FixedAttribute(1, TlvArray(TlvString, { maxLength: 32 }), { default: [] })
    }
  });
})(LocalizationConfiguration || (LocalizationConfiguration = {}));
const LocalizationConfigurationCluster = LocalizationConfiguration.Cluster;
export {
  LocalizationConfiguration,
  LocalizationConfigurationCluster
};
