// @project-chip/matter.js/dist/esm/cluster/definitions/MediaPlaybackCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { Attribute, Command, OptionalCommand } from "../../cluster/Cluster.js";
import { TlvEnum, TlvUInt64, TlvEpochUs, TlvFloat } from "../../tlv/TlvNumber.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
import { TlvObject, TlvField, TlvOptionalField } from "../../tlv/TlvObject.js";
import { TlvByteString } from "../../tlv/TlvString.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
var MediaPlayback;
((MediaPlayback2) => {
  let PlaybackState;
  ((PlaybackState2) => {
    PlaybackState2[PlaybackState2["Playing"] = 0] = "Playing";
    PlaybackState2[PlaybackState2["Paused"] = 1] = "Paused";
    PlaybackState2[PlaybackState2["NotPlaying"] = 2] = "NotPlaying";
    PlaybackState2[PlaybackState2["Buffering"] = 3] = "Buffering";
  })(PlaybackState = MediaPlayback2.PlaybackState || (MediaPlayback2.PlaybackState = {}));
  let Status;
  ((Status2) => {
    Status2[Status2["Success"] = 0] = "Success";
    Status2[Status2["InvalidStateForCommand"] = 1] = "InvalidStateForCommand";
    Status2[Status2["NotAllowed"] = 2] = "NotAllowed";
    Status2[Status2["NotActive"] = 3] = "NotActive";
    Status2[Status2["SpeedOutOfRange"] = 4] = "SpeedOutOfRange";
    Status2[Status2["SeekOutOfRange"] = 5] = "SeekOutOfRange";
  })(Status = MediaPlayback2.Status || (MediaPlayback2.Status = {}));
  MediaPlayback2.TlvPlaybackResponse = TlvObject({
    /**
     * This shall indicate the status of the command which resulted in this response.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.12.1
     */
    status: TlvField(0, TlvEnum()),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.12.2
     */
    data: TlvOptionalField(1, TlvByteString)
  });
  MediaPlayback2.TlvSkipForwardRequest = TlvObject({
    /**
     * This shall indicate the duration of the time span to skip forward in the media, in milliseconds. In case the
     * resulting position falls in the middle of a frame, the server shall set the position to the beginning of
     * that frame and set the SampledPosition attribute on the cluster accordingly. If the resultant position falls
     * beyond the furthest valid position in the media the client may seek forward to, the position should be set
     * to that furthest valid position. If the SampledPosition attribute is supported it shall be updated on the
     * cluster accordingly.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.9.1
     */
    deltaPositionMilliseconds: TlvField(0, TlvUInt64)
  });
  MediaPlayback2.TlvSkipBackwardRequest = TlvObject({
    /**
     * This shall indicate the duration of the time span to skip backward in the media, in milliseconds. In case
     * the resulting position falls in the middle of a frame, the server shall set the position to the beginning of
     * that frame and set the SampledPosition attribute on the cluster accordingly. If the resultant position falls
     * before the earliest valid position to which a client may seek back to, the position should be set to that
     * earliest valid position. If the SampledPosition attribute is supported it shall be updated on the cluster
     * accordingly.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.10.1
     */
    deltaPositionMilliseconds: TlvField(0, TlvUInt64)
  });
  MediaPlayback2.TlvPlaybackPositionStruct = TlvObject({
    /**
     * This shall indicate the time when the position was last updated.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.5.3.1
     */
    updatedAt: TlvField(0, TlvEpochUs),
    /**
     * This shall indicate the associated discrete position within the media stream, in milliseconds from the
     * beginning of the stream, being associated with the time indicated by the UpdatedAt field. The Position shall
     * not be greater than the duration of the media if duration is specified. The Position shall not be greater
     * than the time difference between current time and start time of the media when start time is specified.
     *
     * A value of null shall indicate that playback position is not applicable for the current state of the media
     * playback (For example : Live media with no known duration and where seek is not supported).
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.5.3.2
     */
    position: TlvField(1, TlvNullable(TlvUInt64))
  });
  MediaPlayback2.TlvSeekRequest = TlvObject({
    /**
     * This shall indicate the position (in milliseconds) in the media to seek to. In case the position falls in
     * the middle of a frame, the server shall set the position to the beginning of that frame and set the
     * SampledPosition attribute on the cluster accordingly. If the position falls before the earliest valid
     * position or beyond the furthest valid position to which a client may seek back or forward to respectively,
     * the status of SEEK_OUT_OF_RANGE shall be returned and no change shall be made to the position of the
     * playback.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.11.1
     */
    position: TlvField(0, TlvUInt64)
  });
  let Feature;
  ((Feature2) => {
    Feature2["AdvancedSeek"] = "AdvancedSeek";
    Feature2["VariableSpeed"] = "VariableSpeed";
  })(Feature = MediaPlayback2.Feature || (MediaPlayback2.Feature = {}));
  MediaPlayback2.Base = ClusterFactory.Definition({
    id: 1286,
    name: "MediaPlayback",
    revision: 1,
    features: {
      /**
       * AdvancedSeek
       *
       * Enables clients to implement more advanced media seeking behavior in their user interface, such as for
       * example a "seek bar". Adds support for Attributes and Commands related to advanced seek support
       */
      advancedSeek: BitFlag(0),
      /**
       * VariableSpeed
       *
       * Support for commands to support variable speed playback on media that supports it.
       */
      variableSpeed: BitFlag(1)
    },
    attributes: {
      /**
       * This shall indicate the current playback state of media.
       *
       * During fast-forward, rewind, and other seek operations; this attribute shall be set to PLAYING.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.3.1
       */
      currentState: Attribute(0, TlvEnum(), { default: 0 /* Playing */ })
    },
    commands: {
      /**
       * Upon receipt, this shall play media. If content is currently in a FastForward or Rewind state. Play
       * shall return media to normal playback speed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.1
       */
      play: Command(0, TlvNoArguments, 10, MediaPlayback2.TlvPlaybackResponse),
      /**
       * Upon receipt, this shall pause playback of the media.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.2
       */
      pause: Command(1, TlvNoArguments, 10, MediaPlayback2.TlvPlaybackResponse),
      /**
       * Upon receipt, this shall stop playback of the media. User-visible outcome is context-specific. This may
       * navigate the user back to the location from where the media was originally launched.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.3
       */
      stop: Command(2, TlvNoArguments, 10, MediaPlayback2.TlvPlaybackResponse),
      /**
       * Upon receipt, this shall Start Over with the current media playback item.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.4
       */
      startOver: OptionalCommand(3, TlvNoArguments, 10, MediaPlayback2.TlvPlaybackResponse),
      /**
       * Upon receipt, this shall cause the handler to be invoked for "Previous". User experience is
       * context-specific. This will often Go back to the previous media playback item.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.5
       */
      previous: OptionalCommand(4, TlvNoArguments, 10, MediaPlayback2.TlvPlaybackResponse),
      /**
       * Upon receipt, this shall cause the handler to be invoked for "Next". User experience is context-
       * specific. This will often Go forward to the next media playback item.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.6
       */
      next: OptionalCommand(5, TlvNoArguments, 10, MediaPlayback2.TlvPlaybackResponse),
      /**
       * Upon receipt, this shall Skip forward in the media by the given number of milliseconds.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.9
       */
      skipForward: OptionalCommand(8, MediaPlayback2.TlvSkipForwardRequest, 10, MediaPlayback2.TlvPlaybackResponse),
      /**
       * Upon receipt, this shall Skip backward in the media by the given number of milliseconds.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.10
       */
      skipBackward: OptionalCommand(9, MediaPlayback2.TlvSkipBackwardRequest, 10, MediaPlayback2.TlvPlaybackResponse)
    }
  });
  MediaPlayback2.AdvancedSeekComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This shall indicate the start time of the media, in case the media has a fixed start time (for example,
       * live stream or television broadcast), or null when start time does not apply to the current
       *
       * media (for example, video-on-demand). This time is a UTC time. The client needs to handle conversion to
       * local time, as required, taking in account time zone and possible local DST offset.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.3.2
       */
      startTime: Attribute(1, TlvNullable(TlvEpochUs), { default: null }),
      /**
       * This shall indicate the duration, in milliseconds, of the current media being played back or null when
       * duration is not applicable (for example, in live streaming content with no known duration). This
       * attribute shall never be 0.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.3.3
       */
      duration: Attribute(2, TlvNullable(TlvUInt64), { default: null }),
      /**
       * This shall indicate the position of playback (Position field) at the time (UpdateAt field) specified in
       * the attribute. The client may use the SampledPosition attribute to compute the current position within
       * the media stream based on the PlaybackSpeed, PlaybackPositionStruct.UpdatedAt and
       * PlaybackPositionStruct.Position fields. To enable this, the SampledPosition attribute shall be updated
       * whenever a change in either the playback speed or the playback position is triggered outside the normal
       * playback of the media. The events which may cause this to happen include:
       *
       *   • Starting or resumption of playback
       *
       *   • Seeking
       *
       *   • Skipping forward or backward
       *
       *   • Fast-forwarding or rewinding
       *
       *   • Updating of playback speed as a result of explicit request, or as a result of buffering events
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.3.4
       */
      sampledPosition: Attribute(3, TlvNullable(MediaPlayback2.TlvPlaybackPositionStruct), { default: null }),
      /**
       * This shall indicate the speed at which the current media is being played. The new PlaybackSpeed
       *
       * shall be reflected in this attribute whenever any of the following occurs:
       *
       *   • Starting of playback
       *
       *   • Resuming of playback
       *
       *   • Fast-forwarding
       *
       *   • Rewinding
       *
       * The PlaybackSpeed shall reflect the ratio of time elapsed in the media to the actual time taken for the
       * playback assuming no changes to media playback (for example buffering events or requests to
       * pause/rewind/forward).
       *
       *   • A value for PlaybackSpeed of 1 shall indicate normal playback where, for example, playback for 1
       *     second causes the media to advance by 1 second within the duration of the media.
       *
       *   • A value for PlaybackSpeed which is greater than 0 shall indicate that as playback is happening the
       *     media is currently advancing in time within the duration of the media.
       *
       *   • A value for PlaybackSpeed which is less than 0 shall indicate that as playback is happening the
       *     media is currently going back in time within the duration of the media.
       *
       *   • A value for PlaybackSpeed of 0 shall indicate that the media is currently not playing back. When the
       *     CurrentState attribute has the value of PAUSED, NOT_PLAYING or BUFFERING, the Playback
       *
       * Speed shall be set to 0 to reflect that the media is not playing.
       *
       * Following examples illustrate the PlaybackSpeed attribute values in various conditions.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.3.5
       */
      playbackSpeed: Attribute(4, TlvFloat, { default: 0 }),
      /**
       * This shall indicate the furthest forward valid position to which a client may seek forward, in
       * milliseconds from the start of the media. When the media has an associated StartTime, a value of null
       * shall indicate that a seek forward is valid only until the current time within the media, using a
       * position computed from the difference between the current time offset and StartTime, in milliseconds
       * from start of the media, truncating fractional milliseconds towards 0. A value of Nas when StartTime is
       * not specified shall indicate that seeking forward is not allowed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.3.7
       */
      seekRangeEnd: Attribute(5, TlvNullable(TlvUInt64), { default: null }),
      /**
       * This shall indicate the earliest valid position to which a client may seek back, in milliseconds from
       * start of the media. A value of Nas shall indicate that seeking backwards is not allowed.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.3.6
       */
      seekRangeStart: Attribute(6, TlvNullable(TlvUInt64), { default: null })
    },
    commands: {
      /**
       * Upon receipt, this shall change the playback position in the media to the given position.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.11
       */
      seek: Command(11, MediaPlayback2.TlvSeekRequest, 10, MediaPlayback2.TlvPlaybackResponse)
    }
  });
  MediaPlayback2.VariableSpeedComponent = ClusterFactory.Component({
    commands: {
      /**
       * Upon receipt, this shall start playback of the media backward in case the media is currently playing in
       * the forward direction or is not playing. If the playback is already happening in the backwards direction
       * receipt of this command shall increase the speed of the media playback back
       *
       * wards.
       *
       * Different "rewind" speeds may be be reflected on the media playback device based upon the number of
       * sequential calls to this function and the capability of the device. This is to avoid needing to define
       * every speed (multiple fast, slow motion, etc). If the PlaybackSpeed attribute is supported it shall be
       * updated to reflect the new speed of playback. If the playback speed cannot be changed for the media
       * being played(for example, in live streaming content not supporting seek), the status of NOT_ALLOWED
       * shall be returned. If the playback speed has reached the maximum supported speed for media playing
       * backwards, the status of SPEED_OUT_OF_RANGE shall be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.7
       */
      rewind: Command(6, TlvNoArguments, 10, MediaPlayback2.TlvPlaybackResponse),
      /**
       * Upon receipt, this shall start playback of the media in the forward direction in case the media is
       * currently playing in the backward direction or is not playing. If the playback is already happening in
       * the forward direction receipt of this command shall increase the speed of the media playback.
       *
       * Different "fast-forward" speeds may be be reflected on the media playback device based upon the number
       * of sequential calls to this function and the capability of the device. This is to avoid needing to
       * define every speed (multiple fast, slow motion, etc). If the PlaybackSpeed attribute is supported it
       * shall be updated to reflect the new speed of playback. If the playback speed cannot be changed for the
       * media being played(for example, in live streaming content not supporting seek), the status of
       * NOT_ALLOWED shall be returned. If the playback speed has reached the maximum supported speed for media
       * playing forward, the status of SPEED_OUT_OF_RANGE shall be returned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.10.4.8
       */
      fastForward: Command(7, TlvNoArguments, 10, MediaPlayback2.TlvPlaybackResponse)
    }
  });
  MediaPlayback2.Cluster = ClusterFactory.Extensible(
    MediaPlayback2.Base,
    /**
     * Use this factory method to create a MediaPlayback cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a MediaPlayback cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...MediaPlayback2.Base,
        supportedFeatures: BitFlags(MediaPlayback2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, MediaPlayback2.AdvancedSeekComponent, { advancedSeek: true });
      ClusterFactory.extend(cluster, MediaPlayback2.VariableSpeedComponent, { variableSpeed: true });
      return cluster;
    }
  );
  const AS = { advancedSeek: true };
  const VS = { variableSpeed: true };
  MediaPlayback2.Complete = ClusterFactory.Definition({
    id: MediaPlayback2.Cluster.id,
    name: MediaPlayback2.Cluster.name,
    revision: MediaPlayback2.Cluster.revision,
    features: MediaPlayback2.Cluster.features,
    attributes: {
      ...MediaPlayback2.Cluster.attributes,
      startTime: ClusterFactory.AsConditional(MediaPlayback2.AdvancedSeekComponent.attributes.startTime, { mandatoryIf: [AS] }),
      duration: ClusterFactory.AsConditional(MediaPlayback2.AdvancedSeekComponent.attributes.duration, { mandatoryIf: [AS] }),
      sampledPosition: ClusterFactory.AsConditional(
        MediaPlayback2.AdvancedSeekComponent.attributes.sampledPosition,
        { mandatoryIf: [AS] }
      ),
      playbackSpeed: ClusterFactory.AsConditional(
        MediaPlayback2.AdvancedSeekComponent.attributes.playbackSpeed,
        { mandatoryIf: [AS] }
      ),
      seekRangeEnd: ClusterFactory.AsConditional(
        MediaPlayback2.AdvancedSeekComponent.attributes.seekRangeEnd,
        { mandatoryIf: [AS] }
      ),
      seekRangeStart: ClusterFactory.AsConditional(
        MediaPlayback2.AdvancedSeekComponent.attributes.seekRangeStart,
        { mandatoryIf: [AS] }
      )
    },
    commands: {
      ...MediaPlayback2.Cluster.commands,
      rewind: ClusterFactory.AsConditional(MediaPlayback2.VariableSpeedComponent.commands.rewind, { mandatoryIf: [VS] }),
      fastForward: ClusterFactory.AsConditional(
        MediaPlayback2.VariableSpeedComponent.commands.fastForward,
        { mandatoryIf: [VS] }
      ),
      seek: ClusterFactory.AsConditional(MediaPlayback2.AdvancedSeekComponent.commands.seek, { mandatoryIf: [AS] })
    }
  });
})(MediaPlayback || (MediaPlayback = {}));
const MediaPlaybackCluster = MediaPlayback.Cluster;
export {
  MediaPlayback,
  MediaPlaybackCluster
};
