// @project-chip/matter.js/dist/esm/cluster/definitions/MediaInputCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { Attribute, Command, TlvNoResponse, AccessLevel } from "../../cluster/Cluster.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvUInt8, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvString } from "../../tlv/TlvString.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var MediaInput;
((MediaInput2) => {
  let InputType;
  ((InputType2) => {
    InputType2[InputType2["Internal"] = 0] = "Internal";
    InputType2[InputType2["Aux"] = 1] = "Aux";
    InputType2[InputType2["Coax"] = 2] = "Coax";
    InputType2[InputType2["Composite"] = 3] = "Composite";
    InputType2[InputType2["Hdmi"] = 4] = "Hdmi";
    InputType2[InputType2["Input"] = 5] = "Input";
    InputType2[InputType2["Line"] = 6] = "Line";
    InputType2[InputType2["Optical"] = 7] = "Optical";
    InputType2[InputType2["Video"] = 8] = "Video";
    InputType2[InputType2["Scart"] = 9] = "Scart";
    InputType2[InputType2["Usb"] = 10] = "Usb";
    InputType2[InputType2["Other"] = 11] = "Other";
  })(InputType = MediaInput2.InputType || (MediaInput2.InputType = {}));
  MediaInput2.TlvInputInfoStruct = TlvObject({
    /**
     * This shall indicate the unique index into the list of Inputs.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.5.1.1
     */
    index: TlvField(0, TlvUInt8),
    /**
     * This shall indicate the type of input
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.5.1.2
     */
    inputType: TlvField(1, TlvEnum()),
    /**
     * This shall indicate the input name, such as “HDMI 1”. This field may be blank, but SHOULD be provided when
     * known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.5.1.3
     */
    name: TlvField(2, TlvString.bound({ maxLength: 32 })),
    /**
     * This shall indicate the user editable input description, such as “Living room Playstation”. This field may
     * be blank, but SHOULD be provided when known.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.5.1.4
     */
    description: TlvField(3, TlvString.bound({ maxLength: 32 }))
  });
  MediaInput2.TlvSelectInputRequest = TlvObject({
    /**
     * This shall indicate the index field of the InputInfoStruct from the InputList attribute in which to change
     * to.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.4.1.1
     */
    index: TlvField(0, TlvUInt8)
  });
  MediaInput2.TlvRenameInputRequest = TlvObject({ index: TlvField(0, TlvUInt8), name: TlvField(1, TlvString) });
  let Feature;
  ((Feature2) => {
    Feature2["NameUpdates"] = "NameUpdates";
  })(Feature = MediaInput2.Feature || (MediaInput2.Feature = {}));
  MediaInput2.Base = ClusterFactory.Definition({
    id: 1287,
    name: "MediaInput",
    revision: 1,
    features: {
      /**
       * NameUpdates
       *
       * Supports updates to the input names
       */
      nameUpdates: BitFlag(0)
    },
    attributes: {
      /**
       * This list provides the media inputs supported by the device.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.3.1
       */
      inputList: Attribute(0, TlvArray(MediaInput2.TlvInputInfoStruct, { maxLength: 254 }), { default: [] }),
      /**
       * This field contains the value of the index field of the currently selected InputInfoStruct.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.3.2
       */
      currentInput: Attribute(1, TlvUInt8, { default: 0 })
    },
    commands: {
      /**
       * Upon receipt, this shall change the media input on the device to the input at a specific index in the
       * Input List.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.4.1
       */
      selectInput: Command(0, MediaInput2.TlvSelectInputRequest, 0, TlvNoResponse),
      /**
       * Upon receipt, this shall display the active status of the input list on screen.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.4.2
       */
      showInputStatus: Command(1, TlvNoArguments, 1, TlvNoResponse),
      /**
       * Upon receipt, this shall hide the input list from the screen.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.4.3
       */
      hideInputStatus: Command(2, TlvNoArguments, 2, TlvNoResponse)
    }
  });
  MediaInput2.NameUpdatesComponent = ClusterFactory.Component({
    commands: {
      /**
       * Upon receipt, this shall rename the input at a specific index in the Input List. Updates to the input
       * name shall appear in the device’s settings menus.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.9.4.4
       */
      renameInput: Command(3, MediaInput2.TlvRenameInputRequest, 3, TlvNoResponse, { invokeAcl: AccessLevel.Manage })
    }
  });
  MediaInput2.Cluster = ClusterFactory.Extensible(
    MediaInput2.Base,
    /**
     * Use this factory method to create a MediaInput cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a MediaInput cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...MediaInput2.Base,
        supportedFeatures: BitFlags(MediaInput2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, MediaInput2.NameUpdatesComponent, { nameUpdates: true });
      return cluster;
    }
  );
  const NU = { nameUpdates: true };
  MediaInput2.Complete = ClusterFactory.Definition({
    id: MediaInput2.Cluster.id,
    name: MediaInput2.Cluster.name,
    revision: MediaInput2.Cluster.revision,
    features: MediaInput2.Cluster.features,
    attributes: MediaInput2.Cluster.attributes,
    commands: {
      ...MediaInput2.Cluster.commands,
      renameInput: ClusterFactory.AsConditional(MediaInput2.NameUpdatesComponent.commands.renameInput, { mandatoryIf: [NU] })
    }
  });
})(MediaInput || (MediaInput = {}));
const MediaInputCluster = MediaInput.Cluster;
export {
  MediaInput,
  MediaInputCluster
};
