// @project-chip/matter.js/dist/esm/cluster/definitions/OccupancySensingCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { Attribute, OptionalWritableAttribute, AccessLevel } from "../../cluster/Cluster.js";
import { BitFlag } from "../../schema/BitmapSchema.js";
import { TlvUInt8, TlvBitmap, TlvEnum, TlvUInt16 } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
var OccupancySensing;
((OccupancySensing2) => {
  OccupancySensing2.OccupancyBitmap = {
    /**
     * Indicates the sensed occupancy state; 1 = occupied, 0 = unoccupied.
     */
    occupied: BitFlag(0)
  };
  let OccupancySensorType;
  ((OccupancySensorType2) => {
    OccupancySensorType2[OccupancySensorType2["Pir"] = 0] = "Pir";
    OccupancySensorType2[OccupancySensorType2["Ultrasonic"] = 1] = "Ultrasonic";
    OccupancySensorType2[OccupancySensorType2["PirAndUltrasonic"] = 2] = "PirAndUltrasonic";
    OccupancySensorType2[OccupancySensorType2["PhysicalContact"] = 3] = "PhysicalContact";
  })(OccupancySensorType = OccupancySensing2.OccupancySensorType || (OccupancySensing2.OccupancySensorType = {}));
  OccupancySensing2.OccupancySensorTypeBitmap = {
    /**
     * Indicates a passive infrared sensor.
     */
    pir: BitFlag(0),
    /**
     * Indicates a ultrasonic sensor.
     */
    ultrasonic: BitFlag(1),
    /**
     * Indicates a physical contact sensor.
     */
    physicalContact: BitFlag(2)
  };
  OccupancySensing2.Cluster = ClusterFactory.Definition({
    id: 1030,
    name: "OccupancySensing",
    revision: 3,
    attributes: {
      /**
       * The Occupancy attribute indicates the status of occupancy.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.6.1
       */
      occupancy: Attribute(0, TlvBitmap(TlvUInt8, OccupancySensing2.OccupancyBitmap)),
      /**
       * The OccupancySensorType attribute specifies the type of the occupancy sensor.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.6.2
       */
      occupancySensorType: Attribute(1, TlvEnum()),
      /**
       * The OccupancySensorTypeBitmap attribute specifies the types of the occupancy sensor. A ‘1’ in each bit
       * position indicates the capability is implemented.
       *
       * The value of the OccupancySensorTypeBitmap attribute and the OccupancySensorType attribute shall be
       * aligned as defined below.
       *
       * Table 32. Mapping between OccupancySensorTypeBitmap and OccupancySensorType Attributes
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.6.3
       */
      occupancySensorTypeBitmap: Attribute(2, TlvBitmap(TlvUInt8, OccupancySensing2.OccupancySensorTypeBitmap)),
      /**
       * The PIROccupiedToUnoccupiedDelay attribute specifies the time delay, in seconds, before the PIR sensor
       * changes to its unoccupied state after the last detection of movement in the sensed area.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.7.1
       */
      pirOccupiedToUnoccupiedDelay: OptionalWritableAttribute(
        16,
        TlvUInt16,
        { default: 0, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The PIRUnoccupiedToOccupiedDelay attribute specifies the time delay, in seconds, before the PIR sensor
       * changes to its occupied state after the detection of movement in the sensed area. This attribute is
       * mandatory if the PIRUnoccupiedToOccupiedThreshold attribute is implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.7.2
       */
      pirUnoccupiedToOccupiedDelay: OptionalWritableAttribute(
        17,
        TlvUInt16,
        { default: 0, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The PIRUnoccupiedToOccupiedThreshold attribute specifies the number of movement detection events that
       * must occur in the period PIRUnoccupiedToOccupiedDelay, before the PIR sensor changes to its occupied
       * state. This attribute is mandatory if the PIRUnoccupiedToOccupiedDelay attribute is implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.7.3
       */
      pirUnoccupiedToOccupiedThreshold: OptionalWritableAttribute(
        18,
        TlvUInt8.bound({ min: 1, max: 254 }),
        { default: 1, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The UltrasonicOccupiedToUnoccupiedDelay attribute and specifies the time delay, in seconds, before the
       * Ultrasonic sensor changes to its unoccupied state after the last detection of movement in the sensed
       * area.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.8.1
       */
      ultrasonicOccupiedToUnoccupiedDelay: OptionalWritableAttribute(
        32,
        TlvUInt16,
        { default: 0, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The UltrasonicUnoccupiedToOccupiedDelay attribute and specifies the time delay, in seconds, before the
       * Ultrasonic sensor changes to its occupied state after the detection of movement in the sensed area. This
       * attribute is mandatory if the UltrasonicUnoccupiedToOccupiedThreshold attribute is implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.8.2
       */
      ultrasonicUnoccupiedToOccupiedDelay: OptionalWritableAttribute(
        33,
        TlvUInt16,
        { default: 0, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The UltrasonicUnoccupiedToOccupiedThreshold attribute specifies the number of movement detection events
       * that must occur in the period UltrasonicUnoccupiedToOccupiedDelay, before the Ultrasonic sensor changes
       * to its occupied state. This attribute is mandatory if the UltrasonicUnoccupiedToOccupiedDelay attribute
       * is implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.8.3
       */
      ultrasonicUnoccupiedToOccupiedThreshold: OptionalWritableAttribute(
        34,
        TlvUInt8.bound({ min: 1, max: 254 }),
        { default: 1, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The PhysicalContactOccupiedToUnoccupiedDelay attribute specifies the time delay, in seconds, before the
       * physical contact occupancy sensor changes to its unoccupied state after detecting the unoccupied event.
       * The null value indicates that the sensor does not report occupied to unoccupied transition.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.9.1
       */
      physicalContactOccupiedToUnoccupiedDelay: OptionalWritableAttribute(
        48,
        TlvNullable(TlvUInt16),
        { default: 0, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The PhysicalContactUnoccupiedToOccupiedDelay attribute specifies the time delay, in seconds, before the
       * physical contact sensor changes to its occupied state after the detection of the occupied event.
       *
       * The null value indicates that the sensor does not report unoccupied to occupied transition.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.9.2
       */
      physicalContactUnoccupiedToOccupiedDelay: OptionalWritableAttribute(
        49,
        TlvNullable(TlvUInt16),
        { default: 0, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The PhysicalContactUnoccupiedToOccupiedThreshold attribute specifies the number of movement detection
       * events that must occur in the period PhysicalContactUnoccupiedToOccupiedDelay, before the PIR sensor
       * changes to its occupied state. This attribute is mandatory if the
       * PhysicalContactUnoccupiedToOccupiedDelay attribute is implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.7.9.3
       */
      physicalContactUnoccupiedToOccupiedThreshold: OptionalWritableAttribute(
        50,
        TlvUInt8.bound({ min: 1, max: 254 }),
        { default: 1, writeAcl: AccessLevel.Manage }
      )
    }
  });
})(OccupancySensing || (OccupancySensing = {}));
const OccupancySensingCluster = OccupancySensing.Cluster;
export {
  OccupancySensing,
  OccupancySensingCluster
};
