// @project-chip/matter.js/dist/esm/cluster/definitions/OnOffSwitchConfigurationCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { Attribute, WritableAttribute } from "../../cluster/Cluster.js";
import { TlvEnum } from "../../tlv/TlvNumber.js";
var OnOffSwitchConfiguration;
((OnOffSwitchConfiguration2) => {
  let SwitchType;
  ((SwitchType2) => {
    SwitchType2[SwitchType2["Toggle"] = 0] = "Toggle";
    SwitchType2[SwitchType2["Momentary"] = 1] = "Momentary";
    SwitchType2[SwitchType2["MultiFunction"] = 2] = "MultiFunction";
  })(SwitchType = OnOffSwitchConfiguration2.SwitchType || (OnOffSwitchConfiguration2.SwitchType = {}));
  let SwitchActions;
  ((SwitchActions2) => {
    SwitchActions2[SwitchActions2["On"] = 0] = "On";
    SwitchActions2[SwitchActions2["Off"] = 1] = "Off";
    SwitchActions2[SwitchActions2["Toggle"] = 2] = "Toggle";
  })(SwitchActions = OnOffSwitchConfiguration2.SwitchActions || (OnOffSwitchConfiguration2.SwitchActions = {}));
  OnOffSwitchConfiguration2.Cluster = ClusterFactory.Definition({
    id: 7,
    name: "OnOffSwitchConfiguration",
    revision: 1,
    attributes: {
      switchType: Attribute(0, TlvEnum()),
      switchActions: WritableAttribute(16, TlvEnum(), { default: 0 /* On */ })
    }
  });
})(OnOffSwitchConfiguration || (OnOffSwitchConfiguration = {}));
const OnOffSwitchConfigurationCluster = OnOffSwitchConfiguration.Cluster;
export {
  OnOffSwitchConfiguration,
  OnOffSwitchConfigurationCluster
};
