// @project-chip/matter.js/dist/esm/cluster/definitions/OtaSoftwareUpdateRequestorCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import {
  WritableFabricScopedAttribute,
  AccessLevel,
  Attribute,
  OptionalCommand,
  TlvNoResponse,
  Event,
  EventPriority
} from "../../cluster/Cluster.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvObject, TlvField, TlvOptionalField } from "../../tlv/TlvObject.js";
import { TlvNodeId } from "../../datatype/NodeId.js";
import { TlvEndpointNumber } from "../../datatype/EndpointNumber.js";
import { TlvFabricIndex } from "../../datatype/FabricIndex.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
import { TlvEnum, TlvUInt8, TlvUInt32, TlvUInt16, TlvUInt64, TlvInt64 } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvVendorId } from "../../datatype/VendorId.js";
import { TlvByteString } from "../../tlv/TlvString.js";
var OtaSoftwareUpdateRequestor;
((OtaSoftwareUpdateRequestor2) => {
  OtaSoftwareUpdateRequestor2.TlvProviderLocationStruct = TlvObject({
    providerNodeId: TlvField(1, TlvNodeId),
    endpoint: TlvField(2, TlvEndpointNumber),
    fabricIndex: TlvField(254, TlvFabricIndex)
  });
  let UpdateState;
  ((UpdateState2) => {
    UpdateState2[UpdateState2["Unknown"] = 0] = "Unknown";
    UpdateState2[UpdateState2["Idle"] = 1] = "Idle";
    UpdateState2[UpdateState2["Querying"] = 2] = "Querying";
    UpdateState2[UpdateState2["DelayedOnQuery"] = 3] = "DelayedOnQuery";
    UpdateState2[UpdateState2["Downloading"] = 4] = "Downloading";
    UpdateState2[UpdateState2["Applying"] = 5] = "Applying";
    UpdateState2[UpdateState2["DelayedOnApply"] = 6] = "DelayedOnApply";
    UpdateState2[UpdateState2["RollingBack"] = 7] = "RollingBack";
    UpdateState2[UpdateState2["DelayedOnUserConsent"] = 8] = "DelayedOnUserConsent";
  })(UpdateState = OtaSoftwareUpdateRequestor2.UpdateState || (OtaSoftwareUpdateRequestor2.UpdateState = {}));
  let AnnouncementReason;
  ((AnnouncementReason2) => {
    AnnouncementReason2[AnnouncementReason2["SimpleAnnouncement"] = 0] = "SimpleAnnouncement";
    AnnouncementReason2[AnnouncementReason2["UpdateAvailable"] = 1] = "UpdateAvailable";
    AnnouncementReason2[AnnouncementReason2["UrgentUpdateAvailable"] = 2] = "UrgentUpdateAvailable";
  })(AnnouncementReason = OtaSoftwareUpdateRequestor2.AnnouncementReason || (OtaSoftwareUpdateRequestor2.AnnouncementReason = {}));
  OtaSoftwareUpdateRequestor2.TlvAnnounceOtaProviderRequest = TlvObject({
    providerNodeId: TlvField(0, TlvNodeId),
    vendorId: TlvField(1, TlvVendorId),
    announcementReason: TlvField(2, TlvEnum()),
    metadataForNode: TlvOptionalField(3, TlvByteString.bound({ maxLength: 512 })),
    endpoint: TlvField(4, TlvEndpointNumber),
    fabricIndex: TlvField(254, TlvFabricIndex)
  });
  let ChangeReason;
  ((ChangeReason2) => {
    ChangeReason2[ChangeReason2["Unknown"] = 0] = "Unknown";
    ChangeReason2[ChangeReason2["Success"] = 1] = "Success";
    ChangeReason2[ChangeReason2["Failure"] = 2] = "Failure";
    ChangeReason2[ChangeReason2["TimeOut"] = 3] = "TimeOut";
    ChangeReason2[ChangeReason2["DelayByProvider"] = 4] = "DelayByProvider";
  })(ChangeReason = OtaSoftwareUpdateRequestor2.ChangeReason || (OtaSoftwareUpdateRequestor2.ChangeReason = {}));
  OtaSoftwareUpdateRequestor2.TlvStateTransitionEvent = TlvObject({
    previousState: TlvField(0, TlvEnum()),
    newState: TlvField(1, TlvEnum()),
    reason: TlvField(2, TlvEnum()),
    targetSoftwareVersion: TlvField(3, TlvNullable(TlvUInt32))
  });
  OtaSoftwareUpdateRequestor2.TlvVersionAppliedEvent = TlvObject({
    softwareVersion: TlvField(0, TlvUInt32),
    productId: TlvField(1, TlvUInt16)
  });
  OtaSoftwareUpdateRequestor2.TlvDownloadErrorEvent = TlvObject({
    softwareVersion: TlvField(0, TlvUInt32),
    bytesDownloaded: TlvField(1, TlvUInt64),
    progressPercent: TlvField(2, TlvNullable(TlvUInt8.bound({ max: 100 }))),
    platformCode: TlvField(3, TlvNullable(TlvInt64))
  });
  OtaSoftwareUpdateRequestor2.Cluster = ClusterFactory.Definition({
    id: 42,
    name: "OtaSoftwareUpdateRequestor",
    revision: 1,
    attributes: {
      /**
       * This field is a list of ProviderLocationStruct whose entries shall be set by Administrators, either
       * during Commissioning or at a later time, to set the Provider Location for the default OTA Provider Node
       * to use for software updates on a given Fabric.
       *
       * There shall NOT be more than one entry per Fabric. On a list update that would introduce more than one
       * entry per fabric, the write shall fail with CONSTRAINT_ERROR status code.
       *
       * Provider Locations obtained using the AnnounceOTAProvider command shall NOT overwrite values set in the
       * DefaultOTAProviders attribute.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.5.1
       */
      defaultOtaProviders: WritableFabricScopedAttribute(
        0,
        TlvArray(OtaSoftwareUpdateRequestor2.TlvProviderLocationStruct),
        { default: [], writeAcl: AccessLevel.Administer }
      ),
      /**
       * This field shall be set to True if the OTA Requestor is currently able to be updated. Otherwise, it
       * shall be set to False in case of any condition preventing update being possible, such as insufficient
       * capacity of an internal battery. This field is merely informational for diagnostics purposes and shall
       * NOT affect the responses provided by an OTA Provider to an OTA Requestor.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.5.2
       */
      updatePossible: Attribute(1, TlvBoolean, { default: true }),
      /**
       * This field shall reflect the current state of the OTA Requestor with regards to obtaining software
       * updates. See Section 11.19.7.4.2, “UpdateStateEnum” for possible values.
       *
       * This field SHOULD be updated in a timely manner whenever OTA Requestor internal state updates.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.5.3
       */
      updateState: Attribute(2, TlvEnum(), { default: 0 /* Unknown */ }),
      /**
       * This field shall reflect the percentage value of progress, relative to the current UpdateState, if
       * applicable to the state.
       *
       * The value of this field shall be null if a progress indication does not apply to the current state.
       *
       * A value of 0 shall indicate that the beginning has occurred. A value of 100 shall indicate completion.
       *
       * This field may be updated infrequently. Some care SHOULD be taken by Nodes to avoid over- reporting
       * progress when this attribute is part of a subscription.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.5.4
       */
      updateStateProgress: Attribute(3, TlvNullable(TlvUInt8.bound({ max: 100 })), { default: null })
    },
    commands: {
      /**
       * This command may be invoked by Administrators to announce the presence of a particular OTA Provider.
       *
       * This command shall be scoped to the accessing fabric.
       *
       * If the accessing fabric index is 0, this command shall fail with an UNSUPPORTED_ACCESS status code.
       *
       * This field shall contain the Node ID of a Node implementing the OTA Provider cluster server, on the
       * accessing fabric.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.6.1
       */
      announceOtaProvider: OptionalCommand(
        0,
        OtaSoftwareUpdateRequestor2.TlvAnnounceOtaProviderRequest,
        0,
        TlvNoResponse,
        { invokeAcl: AccessLevel.Administer }
      )
    },
    events: {
      /**
       * This event shall be generated when a change of the UpdateState attribute occurs due to an OTA Requestor
       * moving through the states necessary to query for updates.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.7.1
       */
      stateTransition: Event(0, EventPriority.Info, OtaSoftwareUpdateRequestor2.TlvStateTransitionEvent),
      /**
       * This event shall be generated whenever a new version starts executing after being applied due to a
       * software update. This event SHOULD be generated even if a software update was done using means outside
       * of this cluster.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.7.6
       */
      versionApplied: Event(1, EventPriority.Critical, OtaSoftwareUpdateRequestor2.TlvVersionAppliedEvent),
      /**
       * This event shall be generated whenever an error occurs during OTA Requestor download operation.
       *
       * This field shall be set to the value of the SoftwareVersion being downloaded, matching the
       * SoftwareVersion field of the QueryImageResponse that caused the failing download to take place.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.7.7.9
       */
      downloadError: Event(2, EventPriority.Info, OtaSoftwareUpdateRequestor2.TlvDownloadErrorEvent)
    }
  });
})(OtaSoftwareUpdateRequestor || (OtaSoftwareUpdateRequestor = {}));
const OtaSoftwareUpdateRequestorCluster = OtaSoftwareUpdateRequestor.Cluster;
export {
  OtaSoftwareUpdateRequestor,
  OtaSoftwareUpdateRequestorCluster
};
