// @project-chip/matter.js/dist/esm/cluster/definitions/OtaSoftwareUpdateProviderCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { Command, TlvNoResponse } from "../../cluster/Cluster.js";
import { TlvObject, TlvField, TlvOptionalField } from "../../tlv/TlvObject.js";
import { TlvVendorId } from "../../datatype/VendorId.js";
import { TlvUInt16, TlvUInt32, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvString, TlvByteString } from "../../tlv/TlvString.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
var OtaSoftwareUpdateProvider;
((OtaSoftwareUpdateProvider2) => {
  let DownloadProtocol;
  ((DownloadProtocol2) => {
    DownloadProtocol2[DownloadProtocol2["BdxSynchronous"] = 0] = "BdxSynchronous";
    DownloadProtocol2[DownloadProtocol2["BdxAsynchronous"] = 1] = "BdxAsynchronous";
    DownloadProtocol2[DownloadProtocol2["Https"] = 2] = "Https";
    DownloadProtocol2[DownloadProtocol2["VendorSpecific"] = 3] = "VendorSpecific";
  })(DownloadProtocol = OtaSoftwareUpdateProvider2.DownloadProtocol || (OtaSoftwareUpdateProvider2.DownloadProtocol = {}));
  OtaSoftwareUpdateProvider2.TlvQueryImageRequest = TlvObject({
    vendorId: TlvField(0, TlvVendorId),
    productId: TlvField(1, TlvUInt16),
    softwareVersion: TlvField(2, TlvUInt32),
    protocolsSupported: TlvField(3, TlvArray(TlvEnum(), { maxLength: 8 })),
    hardwareVersion: TlvOptionalField(4, TlvUInt16),
    location: TlvOptionalField(5, TlvString.bound({ length: 2 })),
    requestorCanConsent: TlvOptionalField(6, TlvBoolean),
    metadataForProvider: TlvOptionalField(7, TlvByteString.bound({ maxLength: 512 }))
  });
  let Status;
  ((Status2) => {
    Status2[Status2["UpdateAvailable"] = 0] = "UpdateAvailable";
    Status2[Status2["Busy"] = 1] = "Busy";
    Status2[Status2["NotAvailable"] = 2] = "NotAvailable";
    Status2[Status2["DownloadProtocolNotSupported"] = 3] = "DownloadProtocolNotSupported";
  })(Status = OtaSoftwareUpdateProvider2.Status || (OtaSoftwareUpdateProvider2.Status = {}));
  OtaSoftwareUpdateProvider2.TlvQueryImageResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    delayedActionTime: TlvOptionalField(1, TlvUInt32),
    imageUri: TlvOptionalField(2, TlvString.bound({ maxLength: 256 })),
    softwareVersion: TlvOptionalField(3, TlvUInt32),
    softwareVersionString: TlvOptionalField(4, TlvString.bound({ minLength: 1, maxLength: 64 })),
    updateToken: TlvOptionalField(5, TlvByteString.bound({ minLength: 8, maxLength: 32 })),
    userConsentNeeded: TlvOptionalField(6, TlvBoolean),
    metadataForRequestor: TlvOptionalField(7, TlvByteString.bound({ maxLength: 512 }))
  });
  OtaSoftwareUpdateProvider2.TlvApplyUpdateRequestRequest = TlvObject({
    updateToken: TlvField(0, TlvByteString.bound({ minLength: 8, maxLength: 32 })),
    newVersion: TlvField(1, TlvUInt32)
  });
  let ApplyUpdateAction;
  ((ApplyUpdateAction2) => {
    ApplyUpdateAction2[ApplyUpdateAction2["Proceed"] = 0] = "Proceed";
    ApplyUpdateAction2[ApplyUpdateAction2["AwaitNextAction"] = 1] = "AwaitNextAction";
    ApplyUpdateAction2[ApplyUpdateAction2["Discontinue"] = 2] = "Discontinue";
  })(ApplyUpdateAction = OtaSoftwareUpdateProvider2.ApplyUpdateAction || (OtaSoftwareUpdateProvider2.ApplyUpdateAction = {}));
  OtaSoftwareUpdateProvider2.TlvApplyUpdateResponse = TlvObject({
    action: TlvField(0, TlvEnum()),
    delayedActionTime: TlvField(1, TlvUInt32)
  });
  OtaSoftwareUpdateProvider2.TlvNotifyUpdateAppliedRequest = TlvObject({
    updateToken: TlvField(0, TlvByteString.bound({ minLength: 8, maxLength: 32 })),
    softwareVersion: TlvField(1, TlvUInt32)
  });
  OtaSoftwareUpdateProvider2.Cluster = ClusterFactory.Definition({
    id: 41,
    name: "OtaSoftwareUpdateProvider",
    revision: 1,
    commands: {
      /**
       * Upon receipt, this command shall trigger an attempt to find an updated Software Image by the OTA
       * Provider to match the OTA Requestor’s constraints provided in the payload fields.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.6.5.1
       */
      queryImage: Command(0, OtaSoftwareUpdateProvider2.TlvQueryImageRequest, 1, OtaSoftwareUpdateProvider2.TlvQueryImageResponse),
      /**
       * This field shall contain the UpdateToken as specified in Section 11.19.3.6.1, “UpdateToken usage”. This
       * field may be used by the OTA Provider to track minimal lifecycle state to allow finer-grained scheduling
       * of the application of Software Images by OTA Requestors.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.6.5.18
       */
      applyUpdateRequest: Command(2, OtaSoftwareUpdateProvider2.TlvApplyUpdateRequestRequest, 3, OtaSoftwareUpdateProvider2.TlvApplyUpdateResponse),
      /**
       * This field shall contain the UpdateToken as specified in Section 11.19.3.6.1, “UpdateToken usage”.
       *
       * The SoftwareVersion included in the request payload shall provide the same value as the SoftwareVersion
       * attribute in the invoking OTA Requestor’s Basic Information Cluster, and SHOULD be consistent with the
       * value representing a new version running on the Node invoking the command.
       *
       * When Generated
       *
       * The NotifyUpdateApplied command SHOULD be invoked in the following two circumstances:
       *
       *   1. An OTA Requestor has just successfully applied a Software Image it had obtained from a previous
       *      QueryImage response.
       *
       *   2. An OTA Requestor has just successfully applied a Software Image it had obtained through means
       *      different than those of this Cluster.
       *
       * An OTA Provider may use the state of invocation of this command to help track the progress of update for
       * OTA Requestors it knows require a new OTA Software Image. However, due to the possibility that an OTA
       * Requestor may never come back (e.g. device removed from Fabric altogether, or a critical malfunction),
       * an OTA Provider shall NOT expect every OTA Requestor to invoke this command for correct operation of the
       * OTA Provider.
       *
       * This command shall be considered optional and shall not result in reduced availability of the OTA
       * Provider functionality if OTA Requestors never invoke this command.
       *
       * Effect on Receipt
       *
       * An OTA Provider receiving an invocation of this command may log it internally.
       *
       * On receiving this command, an OTA Provider may use the information to update its bookkeeping of cached
       * Software Images, or use it for other similar administrative purposes.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.19.6.5.22
       */
      notifyUpdateApplied: Command(4, OtaSoftwareUpdateProvider2.TlvNotifyUpdateAppliedRequest, 4, TlvNoResponse)
    }
  });
})(OtaSoftwareUpdateProvider || (OtaSoftwareUpdateProvider = {}));
const OtaSoftwareUpdateProviderCluster = OtaSoftwareUpdateProvider.Cluster;
export {
  OtaSoftwareUpdateProvider,
  OtaSoftwareUpdateProviderCluster
};
