// @project-chip/matter.js/dist/esm/cluster/definitions/PowerSourceConfigurationCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { Attribute } from "../../cluster/Cluster.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvEndpointNumber } from "../../datatype/EndpointNumber.js";
var PowerSourceConfiguration;
((PowerSourceConfiguration2) => {
  PowerSourceConfiguration2.Cluster = ClusterFactory.Definition({
    id: 46,
    name: "PowerSourceConfiguration",
    revision: 1,
    attributes: {
      /**
       * This list shall contain the set of all power sources capable of participating in the power system of
       * this Node. Each entry in the list shall be the endpoint number of an endpoint having a Power Source
       * cluster, which corresponds to a physical power source. The endpoint number shall be unique within the
       * list.
       *
       * The order of power sources on a Node is defined by the Order attribute of its associated Power Source
       * cluster provided on the endpoint. List entries shall be sorted in increasing order, that is, an entry
       * with a lower order shall have a lower index than any entry with a higher order. Multiple entries may
       * have the same order, there are no restrictions on their relative sorting.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.6.4.1
       */
      sources: Attribute(0, TlvArray(TlvEndpointNumber, { maxLength: 6 }), { persistent: true, default: [] })
    }
  });
})(PowerSourceConfiguration || (PowerSourceConfiguration = {}));
const PowerSourceConfigurationCluster = PowerSourceConfiguration.Cluster;
export {
  PowerSourceConfiguration,
  PowerSourceConfigurationCluster
};
