// @project-chip/matter.js/dist/esm/cluster/definitions/PowerSourceCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import {
  Attribute,
  FixedAttribute,
  OptionalAttribute,
  OptionalFixedAttribute,
  OptionalEvent,
  EventPriority
} from "../../cluster/Cluster.js";
import { TlvEnum, TlvUInt8, TlvUInt32, TlvUInt16 } from "../../tlv/TlvNumber.js";
import { TlvString } from "../../tlv/TlvString.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
var PowerSource;
((PowerSource2) => {
  let PowerSourceStatus;
  ((PowerSourceStatus2) => {
    PowerSourceStatus2[PowerSourceStatus2["Unspecified"] = 0] = "Unspecified";
    PowerSourceStatus2[PowerSourceStatus2["Active"] = 1] = "Active";
    PowerSourceStatus2[PowerSourceStatus2["Standby"] = 2] = "Standby";
    PowerSourceStatus2[PowerSourceStatus2["Unavailable"] = 3] = "Unavailable";
  })(PowerSourceStatus = PowerSource2.PowerSourceStatus || (PowerSource2.PowerSourceStatus = {}));
  let WiredCurrentType;
  ((WiredCurrentType2) => {
    WiredCurrentType2[WiredCurrentType2["Ac"] = 0] = "Ac";
    WiredCurrentType2[WiredCurrentType2["Dc"] = 1] = "Dc";
  })(WiredCurrentType = PowerSource2.WiredCurrentType || (PowerSource2.WiredCurrentType = {}));
  let WiredFault;
  ((WiredFault2) => {
    WiredFault2[WiredFault2["Unspecified"] = 0] = "Unspecified";
    WiredFault2[WiredFault2["OverVoltage"] = 1] = "OverVoltage";
    WiredFault2[WiredFault2["UnderVoltage"] = 2] = "UnderVoltage";
  })(WiredFault = PowerSource2.WiredFault || (PowerSource2.WiredFault = {}));
  PowerSource2.TlvWiredFaultChangeEvent = TlvObject({
    /**
     * This field shall represent the set of faults currently detected, as per Section 11.7.6.11,
     * “ActiveWiredFaults Attribute”.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.7.7.1.1
     */
    current: TlvField(0, TlvArray(TlvEnum(), { maxLength: 8 })),
    /**
     * This field shall represent the set of faults detected prior to this change event, as per Section 11.7.6.11,
     * “ActiveWiredFaults Attribute”.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.7.7.1.2
     */
    previous: TlvField(1, TlvArray(TlvEnum(), { maxLength: 8 }))
  });
  let BatChargeLevel;
  ((BatChargeLevel2) => {
    BatChargeLevel2[BatChargeLevel2["Ok"] = 0] = "Ok";
    BatChargeLevel2[BatChargeLevel2["Warning"] = 1] = "Warning";
    BatChargeLevel2[BatChargeLevel2["Critical"] = 2] = "Critical";
  })(BatChargeLevel = PowerSource2.BatChargeLevel || (PowerSource2.BatChargeLevel = {}));
  let BatReplaceability;
  ((BatReplaceability2) => {
    BatReplaceability2[BatReplaceability2["Unspecified"] = 0] = "Unspecified";
    BatReplaceability2[BatReplaceability2["NotReplaceable"] = 1] = "NotReplaceable";
    BatReplaceability2[BatReplaceability2["UserReplaceable"] = 2] = "UserReplaceable";
    BatReplaceability2[BatReplaceability2["FactoryReplaceable"] = 3] = "FactoryReplaceable";
  })(BatReplaceability = PowerSource2.BatReplaceability || (PowerSource2.BatReplaceability = {}));
  let BatFault;
  ((BatFault2) => {
    BatFault2[BatFault2["Unspecified"] = 0] = "Unspecified";
    BatFault2[BatFault2["OverTemp"] = 1] = "OverTemp";
    BatFault2[BatFault2["UnderTemp"] = 2] = "UnderTemp";
  })(BatFault = PowerSource2.BatFault || (PowerSource2.BatFault = {}));
  PowerSource2.TlvBatFaultChangeEvent = TlvObject({
    current: TlvField(0, TlvArray(TlvEnum(), { maxLength: 8 })),
    previous: TlvField(1, TlvArray(TlvEnum(), { maxLength: 8 }))
  });
  let BatCommonDesignation;
  ((BatCommonDesignation2) => {
    BatCommonDesignation2[BatCommonDesignation2["Unspecified"] = 0] = "Unspecified";
    BatCommonDesignation2[BatCommonDesignation2["Aaa"] = 1] = "Aaa";
    BatCommonDesignation2[BatCommonDesignation2["Aa"] = 2] = "Aa";
    BatCommonDesignation2[BatCommonDesignation2["C"] = 3] = "C";
    BatCommonDesignation2[BatCommonDesignation2["D"] = 4] = "D";
    BatCommonDesignation2[BatCommonDesignation2["4V5"] = 5] = "4V5";
    BatCommonDesignation2[BatCommonDesignation2["6V0"] = 6] = "6V0";
    BatCommonDesignation2[BatCommonDesignation2["9V0"] = 7] = "9V0";
    BatCommonDesignation2[BatCommonDesignation2["12Aa"] = 8] = "12Aa";
    BatCommonDesignation2[BatCommonDesignation2["Aaaa"] = 9] = "Aaaa";
    BatCommonDesignation2[BatCommonDesignation2["A"] = 10] = "A";
    BatCommonDesignation2[BatCommonDesignation2["B"] = 11] = "B";
    BatCommonDesignation2[BatCommonDesignation2["F"] = 12] = "F";
    BatCommonDesignation2[BatCommonDesignation2["N"] = 13] = "N";
    BatCommonDesignation2[BatCommonDesignation2["No6"] = 14] = "No6";
    BatCommonDesignation2[BatCommonDesignation2["SubC"] = 15] = "SubC";
    BatCommonDesignation2[BatCommonDesignation2["A23"] = 16] = "A23";
    BatCommonDesignation2[BatCommonDesignation2["A27"] = 17] = "A27";
    BatCommonDesignation2[BatCommonDesignation2["Ba5800"] = 18] = "Ba5800";
    BatCommonDesignation2[BatCommonDesignation2["Duplex"] = 19] = "Duplex";
    BatCommonDesignation2[BatCommonDesignation2["4Sr44"] = 20] = "4Sr44";
    BatCommonDesignation2[BatCommonDesignation2["E523"] = 21] = "E523";
    BatCommonDesignation2[BatCommonDesignation2["E531"] = 22] = "E531";
    BatCommonDesignation2[BatCommonDesignation2["15V0"] = 23] = "15V0";
    BatCommonDesignation2[BatCommonDesignation2["22V5"] = 24] = "22V5";
    BatCommonDesignation2[BatCommonDesignation2["30V0"] = 25] = "30V0";
    BatCommonDesignation2[BatCommonDesignation2["45V0"] = 26] = "45V0";
    BatCommonDesignation2[BatCommonDesignation2["67V5"] = 27] = "67V5";
    BatCommonDesignation2[BatCommonDesignation2["J"] = 28] = "J";
    BatCommonDesignation2[BatCommonDesignation2["Cr123A"] = 29] = "Cr123A";
    BatCommonDesignation2[BatCommonDesignation2["Cr2"] = 30] = "Cr2";
    BatCommonDesignation2[BatCommonDesignation2["2Cr5"] = 31] = "2Cr5";
    BatCommonDesignation2[BatCommonDesignation2["CrP2"] = 32] = "CrP2";
    BatCommonDesignation2[BatCommonDesignation2["CrV3"] = 33] = "CrV3";
    BatCommonDesignation2[BatCommonDesignation2["Sr41"] = 34] = "Sr41";
    BatCommonDesignation2[BatCommonDesignation2["Sr43"] = 35] = "Sr43";
    BatCommonDesignation2[BatCommonDesignation2["Sr44"] = 36] = "Sr44";
    BatCommonDesignation2[BatCommonDesignation2["Sr45"] = 37] = "Sr45";
    BatCommonDesignation2[BatCommonDesignation2["Sr48"] = 38] = "Sr48";
    BatCommonDesignation2[BatCommonDesignation2["Sr54"] = 39] = "Sr54";
    BatCommonDesignation2[BatCommonDesignation2["Sr55"] = 40] = "Sr55";
    BatCommonDesignation2[BatCommonDesignation2["Sr57"] = 41] = "Sr57";
    BatCommonDesignation2[BatCommonDesignation2["Sr58"] = 42] = "Sr58";
    BatCommonDesignation2[BatCommonDesignation2["Sr59"] = 43] = "Sr59";
    BatCommonDesignation2[BatCommonDesignation2["Sr60"] = 44] = "Sr60";
    BatCommonDesignation2[BatCommonDesignation2["Sr63"] = 45] = "Sr63";
    BatCommonDesignation2[BatCommonDesignation2["Sr64"] = 46] = "Sr64";
    BatCommonDesignation2[BatCommonDesignation2["Sr65"] = 47] = "Sr65";
    BatCommonDesignation2[BatCommonDesignation2["Sr66"] = 48] = "Sr66";
    BatCommonDesignation2[BatCommonDesignation2["Sr67"] = 49] = "Sr67";
    BatCommonDesignation2[BatCommonDesignation2["Sr68"] = 50] = "Sr68";
    BatCommonDesignation2[BatCommonDesignation2["Sr69"] = 51] = "Sr69";
    BatCommonDesignation2[BatCommonDesignation2["Sr516"] = 52] = "Sr516";
    BatCommonDesignation2[BatCommonDesignation2["Sr731"] = 53] = "Sr731";
    BatCommonDesignation2[BatCommonDesignation2["Sr712"] = 54] = "Sr712";
    BatCommonDesignation2[BatCommonDesignation2["Lr932"] = 55] = "Lr932";
    BatCommonDesignation2[BatCommonDesignation2["A5"] = 56] = "A5";
    BatCommonDesignation2[BatCommonDesignation2["A10"] = 57] = "A10";
    BatCommonDesignation2[BatCommonDesignation2["A13"] = 58] = "A13";
    BatCommonDesignation2[BatCommonDesignation2["A312"] = 59] = "A312";
    BatCommonDesignation2[BatCommonDesignation2["A675"] = 60] = "A675";
    BatCommonDesignation2[BatCommonDesignation2["Ac41E"] = 61] = "Ac41E";
    BatCommonDesignation2[BatCommonDesignation2["E10180"] = 62] = "E10180";
    BatCommonDesignation2[BatCommonDesignation2["E10280"] = 63] = "E10280";
    BatCommonDesignation2[BatCommonDesignation2["E10440"] = 64] = "E10440";
    BatCommonDesignation2[BatCommonDesignation2["E14250"] = 65] = "E14250";
    BatCommonDesignation2[BatCommonDesignation2["E14430"] = 66] = "E14430";
    BatCommonDesignation2[BatCommonDesignation2["E14500"] = 67] = "E14500";
    BatCommonDesignation2[BatCommonDesignation2["E14650"] = 68] = "E14650";
    BatCommonDesignation2[BatCommonDesignation2["E15270"] = 69] = "E15270";
    BatCommonDesignation2[BatCommonDesignation2["E16340"] = 70] = "E16340";
    BatCommonDesignation2[BatCommonDesignation2["Rcr123A"] = 71] = "Rcr123A";
    BatCommonDesignation2[BatCommonDesignation2["E17500"] = 72] = "E17500";
    BatCommonDesignation2[BatCommonDesignation2["E17670"] = 73] = "E17670";
    BatCommonDesignation2[BatCommonDesignation2["E18350"] = 74] = "E18350";
    BatCommonDesignation2[BatCommonDesignation2["E18500"] = 75] = "E18500";
    BatCommonDesignation2[BatCommonDesignation2["E18650"] = 76] = "E18650";
    BatCommonDesignation2[BatCommonDesignation2["E19670"] = 77] = "E19670";
    BatCommonDesignation2[BatCommonDesignation2["E25500"] = 78] = "E25500";
    BatCommonDesignation2[BatCommonDesignation2["E26650"] = 79] = "E26650";
    BatCommonDesignation2[BatCommonDesignation2["E32600"] = 80] = "E32600";
  })(BatCommonDesignation = PowerSource2.BatCommonDesignation || (PowerSource2.BatCommonDesignation = {}));
  let BatApprovedChemistry;
  ((BatApprovedChemistry2) => {
    BatApprovedChemistry2[BatApprovedChemistry2["Unspecified"] = 0] = "Unspecified";
    BatApprovedChemistry2[BatApprovedChemistry2["Alkaline"] = 1] = "Alkaline";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumCarbonFluoride"] = 2] = "LithiumCarbonFluoride";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumChromiumOxide"] = 3] = "LithiumChromiumOxide";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumCopperOxide"] = 4] = "LithiumCopperOxide";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumIronDisulfide"] = 5] = "LithiumIronDisulfide";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumManganeseDioxide"] = 6] = "LithiumManganeseDioxide";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumThionylChloride"] = 7] = "LithiumThionylChloride";
    BatApprovedChemistry2[BatApprovedChemistry2["Magnesium"] = 8] = "Magnesium";
    BatApprovedChemistry2[BatApprovedChemistry2["MercuryOxide"] = 9] = "MercuryOxide";
    BatApprovedChemistry2[BatApprovedChemistry2["NickelOxyhydride"] = 10] = "NickelOxyhydride";
    BatApprovedChemistry2[BatApprovedChemistry2["SilverOxide"] = 11] = "SilverOxide";
    BatApprovedChemistry2[BatApprovedChemistry2["ZincAir"] = 12] = "ZincAir";
    BatApprovedChemistry2[BatApprovedChemistry2["ZincCarbon"] = 13] = "ZincCarbon";
    BatApprovedChemistry2[BatApprovedChemistry2["ZincChloride"] = 14] = "ZincChloride";
    BatApprovedChemistry2[BatApprovedChemistry2["ZincManganeseDioxide"] = 15] = "ZincManganeseDioxide";
    BatApprovedChemistry2[BatApprovedChemistry2["LeadAcid"] = 16] = "LeadAcid";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumCobaltOxide"] = 17] = "LithiumCobaltOxide";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumIon"] = 18] = "LithiumIon";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumIonPolymer"] = 19] = "LithiumIonPolymer";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumIronPhosphate"] = 20] = "LithiumIronPhosphate";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumSulfur"] = 21] = "LithiumSulfur";
    BatApprovedChemistry2[BatApprovedChemistry2["LithiumTitanate"] = 22] = "LithiumTitanate";
    BatApprovedChemistry2[BatApprovedChemistry2["NickelCadmium"] = 23] = "NickelCadmium";
    BatApprovedChemistry2[BatApprovedChemistry2["NickelHydrogen"] = 24] = "NickelHydrogen";
    BatApprovedChemistry2[BatApprovedChemistry2["NickelIron"] = 25] = "NickelIron";
    BatApprovedChemistry2[BatApprovedChemistry2["NickelMetalHydride"] = 26] = "NickelMetalHydride";
    BatApprovedChemistry2[BatApprovedChemistry2["NickelZinc"] = 27] = "NickelZinc";
    BatApprovedChemistry2[BatApprovedChemistry2["SilverZinc"] = 28] = "SilverZinc";
    BatApprovedChemistry2[BatApprovedChemistry2["SodiumIon"] = 29] = "SodiumIon";
    BatApprovedChemistry2[BatApprovedChemistry2["SodiumSulfur"] = 30] = "SodiumSulfur";
    BatApprovedChemistry2[BatApprovedChemistry2["ZincBromide"] = 31] = "ZincBromide";
    BatApprovedChemistry2[BatApprovedChemistry2["ZincCerium"] = 32] = "ZincCerium";
  })(BatApprovedChemistry = PowerSource2.BatApprovedChemistry || (PowerSource2.BatApprovedChemistry = {}));
  let BatChargeState;
  ((BatChargeState2) => {
    BatChargeState2[BatChargeState2["Unknown"] = 0] = "Unknown";
    BatChargeState2[BatChargeState2["IsCharging"] = 1] = "IsCharging";
    BatChargeState2[BatChargeState2["IsAtFullCharge"] = 2] = "IsAtFullCharge";
    BatChargeState2[BatChargeState2["IsNotCharging"] = 3] = "IsNotCharging";
  })(BatChargeState = PowerSource2.BatChargeState || (PowerSource2.BatChargeState = {}));
  let BatChargeFault;
  ((BatChargeFault2) => {
    BatChargeFault2[BatChargeFault2["Unspecified"] = 0] = "Unspecified";
    BatChargeFault2[BatChargeFault2["AmbientTooHot"] = 1] = "AmbientTooHot";
    BatChargeFault2[BatChargeFault2["AmbientTooCold"] = 2] = "AmbientTooCold";
    BatChargeFault2[BatChargeFault2["BatteryTooHot"] = 3] = "BatteryTooHot";
    BatChargeFault2[BatChargeFault2["BatteryTooCold"] = 4] = "BatteryTooCold";
    BatChargeFault2[BatChargeFault2["BatteryAbsent"] = 5] = "BatteryAbsent";
    BatChargeFault2[BatChargeFault2["BatteryOverVoltage"] = 6] = "BatteryOverVoltage";
    BatChargeFault2[BatChargeFault2["BatteryUnderVoltage"] = 7] = "BatteryUnderVoltage";
    BatChargeFault2[BatChargeFault2["ChargerOverVoltage"] = 8] = "ChargerOverVoltage";
    BatChargeFault2[BatChargeFault2["ChargerUnderVoltage"] = 9] = "ChargerUnderVoltage";
    BatChargeFault2[BatChargeFault2["SafetyTimeout"] = 10] = "SafetyTimeout";
  })(BatChargeFault = PowerSource2.BatChargeFault || (PowerSource2.BatChargeFault = {}));
  PowerSource2.TlvBatChargeFaultChangeEvent = TlvObject({
    current: TlvField(0, TlvArray(TlvEnum(), { maxLength: 16 })),
    previous: TlvField(1, TlvArray(TlvEnum(), { maxLength: 16 }))
  });
  let Feature;
  ((Feature2) => {
    Feature2["Wired"] = "Wired";
    Feature2["Battery"] = "Battery";
    Feature2["Rechargeable"] = "Rechargeable";
    Feature2["Replaceable"] = "Replaceable";
  })(Feature = PowerSource2.Feature || (PowerSource2.Feature = {}));
  PowerSource2.Base = ClusterFactory.Definition({
    id: 47,
    name: "PowerSource",
    revision: 1,
    features: {
      /**
       * Wired
       *
       * A wired power source
       */
      wired: BitFlag(0),
      /**
       * Battery
       *
       * A battery power source
       */
      battery: BitFlag(1),
      /**
       * Rechargeable
       *
       * A rechargeable battery power source (requires Battery feature)
       */
      rechargeable: BitFlag(2),
      /**
       * Replaceable
       *
       * A replaceable battery power source (requires Battery feature)
       */
      replaceable: BitFlag(3)
    },
    attributes: {
      /**
       * This attribute shall indicate the participation of this power source in providing power to the Node as
       * specified in PowerSourceStatusEnum.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.1
       */
      status: Attribute(0, TlvEnum()),
      /**
       * This attribute shall indicate the relative preference with which the Node will select this source to
       * provide power. A source with a lower order shall be selected by the Node to provide power before any
       * other source with a higher order, if the lower order source is available (see Status).
       *
       * Note, Order is read-only and therefore NOT intended to allow clients control over power source selection.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.2
       */
      order: Attribute(1, TlvUInt8, { persistent: true }),
      /**
       * This attribute shall provide a user-facing description of this source, used to distinguish it from other
       * power sources, e.g. "DC Power", "Primary Battery" or "Battery back-up". This attribute shall NOT be used
       * to convey information such as battery form factor, or chemistry.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.3
       */
      description: FixedAttribute(2, TlvString.bound({ maxLength: 60 }))
    }
  });
  PowerSource2.WiredComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This attribute shall indicate the assessed RMS or DC voltage currently provided by the hard-wired
       * source, in mV (millivolts). A value of NULL shall indicate the Node is currently unable to assess the
       * value. If the wired source is not connected, but the Node is still able to assess a value, then the
       * assessed value may be reported.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.4
       */
      wiredAssessedInputVoltage: OptionalAttribute(3, TlvNullable(TlvUInt32), { omitChanges: true }),
      /**
       * This attribute shall indicate the assessed frequency of the voltage, currently provided by the
       * hard-wired source, in Hz. A value of NULL shall indicate the Node is currently unable to assess the
       * value. If the wired source is not connected, but the Node is still able to assess a value, then the
       * assessed value may be reported.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.5
       */
      wiredAssessedInputFrequency: OptionalAttribute(4, TlvNullable(TlvUInt16), { omitChanges: true }),
      /**
       * This attribute shall indicate the type of current the Node expects to be provided by the hard- wired
       * source as specified in WiredCurrentTypeEnum.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.6
       */
      wiredCurrentType: FixedAttribute(5, TlvEnum()),
      /**
       * This attribute shall indicate the assessed instantaneous current draw of the Node on the hard- wired
       * source, in mA (milliamps). A value of NULL shall indicate the Node is currently unable to assess the
       * value. If the wired source is not connected, but the Node is still able to assess a value, then the
       * assessed value may be reported.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.7
       */
      wiredAssessedCurrent: OptionalAttribute(6, TlvNullable(TlvUInt32), { omitChanges: true }),
      /**
       * This attribute shall indicate the nominal voltage, printed as part of the Node’s regulatory compliance
       * label in mV (millivolts), expected to be provided by the hard-wired source.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.8
       */
      wiredNominalVoltage: OptionalFixedAttribute(7, TlvUInt32),
      /**
       * This attribute shall indicate the maximum current, printed as part of the Node’s regulatory compliance
       * label in mA (milliamps), expected to be provided by the hard-wired source.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.9
       */
      wiredMaximumCurrent: OptionalFixedAttribute(8, TlvUInt32),
      /**
       * This attribute shall indicate if the Node detects that the hard-wired power source is properly connected.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.10
       */
      wiredPresent: OptionalAttribute(9, TlvBoolean),
      /**
       * This attribute shall indicate the set of wired faults currently detected by the Node on this power
       * source. This set is represented as a list of WiredFaultEnum. When the Node detects a fault has been
       * raised, the appropriate WiredFaultEnum value shall be added to this list, provided it is not already
       * present. This list shall NOT contain more than one instance of a specific WiredFaultEnum value. When the
       * Node detects all conditions contributing to a fault have been cleared, the corresponding WiredFaultEnum
       * value shall be removed from this list. An empty list shall indicate there are currently no active
       * faults. The order of this list SHOULD have no significance. Clients interested in monitoring changes in
       * active faults may subscribe to this attribute, or they may subscribe to WiredFaultChange.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.11
       */
      activeWiredFaults: OptionalAttribute(10, TlvArray(TlvEnum(), { length: 8 }))
    },
    events: {
      /**
       * The WiredFaultChange Event shall be generated when the set of wired faults currently detected by the
       * Node on this wired power source changes. This event shall correspond to a change in value of
       * ActiveWiredFaults.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.7.1
       */
      wiredFaultChange: OptionalEvent(0, EventPriority.Info, PowerSource2.TlvWiredFaultChangeEvent)
    }
  });
  PowerSource2.BatteryComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This attribute shall indicate the currently measured output voltage of the battery in mV (millivolts). A
       * value of NULL shall indicate the Node is currently unable to assess the value.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.12
       */
      batVoltage: OptionalAttribute(11, TlvNullable(TlvUInt32), { omitChanges: true }),
      /**
       * This attribute shall indicate the estimated percentage of battery charge remaining until the battery
       * will no longer be able to provide power to the Node. Values are expressed in half percent units, ranging
       * from 0 to 200. E.g. a value of 48 is equivalent to 24%. A value of NULL shall indicate the Node is
       * currently unable to assess the value.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.13
       */
      batPercentRemaining: OptionalAttribute(
        12,
        TlvNullable(TlvUInt8.bound({ max: 200 })),
        { omitChanges: true }
      ),
      /**
       * This attribute shall indicate the estimated time in seconds before the battery will no longer be able to
       * provide power to the Node. A value of NULL shall indicate the Node is currently unable to assess the
       * value.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.14
       */
      batTimeRemaining: OptionalAttribute(13, TlvNullable(TlvUInt32), { omitChanges: true }),
      /**
       * This attribute shall indicate a coarse ranking of the charge level of the battery, used to indicate when
       * intervention is required as specified in BatChargeLevelEnum.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.15
       */
      batChargeLevel: Attribute(14, TlvEnum()),
      /**
       * This attribute shall indicate if the battery needs to be replaced. Replacement may be simple routine
       * maintenance, such as with a single use, non-rechargeable cell. Replacement, however, may also indicate
       * end of life, or serious fault with a rechargeable or even non-replaceable cell.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.16
       */
      batReplacementNeeded: Attribute(15, TlvBoolean),
      /**
       * This attribute shall indicate the replaceability of the battery as specified in BatReplaceabilityEnum.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.17
       */
      batReplaceability: FixedAttribute(16, TlvEnum()),
      /**
       * This attribute shall indicate whether the Node detects that the batteries are properly installed.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.18
       */
      batPresent: OptionalAttribute(17, TlvBoolean),
      /**
       * This attribute shall indicate the set of battery faults currently detected by the Node on this power
       * source. This set is represented as a list of BatFaultEnum. When the Node detects a fault has been
       * raised, the appropriate BatFaultEnum value shall be added to this list, provided it is not already
       * present. This list shall NOT contain more than one instance of a specific BatFaultEnum value. When the
       * Node detects all conditions contributing to a fault have been cleared, the corresponding BatFaultEnum
       * value shall be removed from this list. An empty list shall indicate there are currently no active
       * faults. The order of this list SHOULD have no significance. Clients interested in monitoring changes in
       * active faults may subscribe to this attribute, or they may subscribe to Section 11.7.7.2,
       * “BatFaultChange Event”.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.19
       */
      activeBatFaults: OptionalAttribute(18, TlvArray(TlvEnum(), { length: 8 }))
    },
    events: {
      /**
       * The BatFaultChange Event shall be generated when the set of battery faults currently detected by the
       * Node on this battery power source changes. This event shall correspond to a change in value of
       * ActiveBatFaults.
       *
       * This field shall represent the set of faults currently detected, as per Section 11.7.6.19,
       * “ActiveBatFaults Attribute”.
       *
       * This field shall represent the set of faults detected prior to this change event, as per Section
       * 11.7.6.19, “ActiveBatFaults Attribute”.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.7.2
       */
      batFaultChange: OptionalEvent(1, EventPriority.Info, PowerSource2.TlvBatFaultChangeEvent)
    }
  });
  PowerSource2.ReplaceableComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This attribute shall provide a user-facing description of this battery, which SHOULD contain information
       * required to identify a replacement, such as form factor, chemistry or preferred manufacturer.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.20
       */
      batReplacementDescription: FixedAttribute(19, TlvString.bound({ maxLength: 60 })),
      /**
       * This attribute shall indicate the ID of the common or colloquial designation of the battery, as
       * specified in BatCommonDesignationEnum.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.21
       */
      batCommonDesignation: OptionalFixedAttribute(20, TlvEnum()),
      /**
       * This attribute shall indicate the string representing the ANSI designation for the battery as specified
       * in ANSI C18.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.22
       */
      batAnsiDesignation: OptionalFixedAttribute(21, TlvString.bound({ maxLength: 20 })),
      /**
       * This attribute shall indicate the string representing the IEC designation for the battery as specified
       * in IEC 60086.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.23
       */
      batIecDesignation: OptionalFixedAttribute(22, TlvString.bound({ maxLength: 20 })),
      /**
       * This attribute shall indicate the ID of the preferred chemistry of the battery source as specified in
       * BatApprovedChemistryEnum.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.24
       */
      batApprovedChemistry: OptionalFixedAttribute(23, TlvEnum()),
      /**
       * This attribute shall indicate the preferred minimum charge capacity rating in mAh of individual, user-
       * or factory-serviceable battery cells or packs in the battery source.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.25
       */
      batCapacity: OptionalFixedAttribute(24, TlvUInt32),
      /**
       * This attribute shall indicate the quantity of individual, user- or factory-serviceable battery cells or
       * packs in the battery source.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.26
       */
      batQuantity: FixedAttribute(25, TlvUInt8)
    }
  });
  PowerSource2.RechargeableComponent = ClusterFactory.Component({
    attributes: {
      /**
       * This attribute shall indicate the current state of the battery source with respect to charging as
       * specified in BatChargeStateEnum.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.27
       */
      batChargeState: Attribute(26, TlvEnum()),
      /**
       * This attribute shall indicate the estimated time in seconds before the battery source will be at full
       * charge. A value of NULL shall indicate the Node is currently unable to assess the value.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.28
       */
      batTimeToFullCharge: OptionalAttribute(27, TlvNullable(TlvUInt32), { omitChanges: true }),
      /**
       * This attribute shall indicate whether the Node can remain operational while the battery source is
       * charging.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.29
       */
      batFunctionalWhileCharging: Attribute(28, TlvBoolean),
      /**
       * This attribute shall indicate assessed current in mA (milliamps) presently supplied to charge the
       * battery source. A value of NULL shall indicate the Node is currently unable to assess the value.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.30
       */
      batChargingCurrent: OptionalAttribute(29, TlvNullable(TlvUInt32), { omitChanges: true }),
      /**
       * This attribute shall indicate the set of charge faults currently detected by the Node on this power
       * source. This set is represented as a list of BatChargeFaultEnum. When the Node detects a fault has been
       * raised, the appropriate BatChargeFaultEnum value shall be added to this list, provided it is not already
       * present. This list shall NOT contain more than one instance of a specific BatChargeFaultEnum value. When
       * the Node detects all conditions contributing to a fault have been cleared, the corresponding
       * BatChargeFaultEnum value shall be removed from this list. An empty list shall indicate there are
       * currently no active faults. The order of this list SHOULD have no significance. Clients interested in
       * monitoring changes in active faults may subscribe to this attribute, or they may subscribe to the
       * BatFaultChange event.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.6.31
       */
      activeBatChargeFaults: OptionalAttribute(30, TlvArray(TlvEnum(), { length: 16 }))
    },
    events: {
      /**
       * The BatChargeFaultChange Event shall be generated when the set of charge faults currently
       *
       * detected by the Node on this battery power source changes. This event shall correspond to a change in
       * value of ActiveBatChargeFaults.
       *
       * This field shall represent the set of faults currently detected, as per Section 11.7.6.31,
       * “ActiveBatChargeFaults Attribute”.
       *
       * This field shall represent the set of faults detected prior to this change event, as per Section
       * 11.7.6.31, “ActiveBatChargeFaults Attribute”.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.7.7.3
       */
      batChargeFaultChange: OptionalEvent(2, EventPriority.Info, PowerSource2.TlvBatChargeFaultChangeEvent)
    }
  });
  PowerSource2.Cluster = ClusterFactory.Extensible(
    PowerSource2.Base,
    /**
     * Use this factory method to create a PowerSource cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a PowerSource cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...PowerSource2.Base,
        supportedFeatures: BitFlags(PowerSource2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, PowerSource2.WiredComponent, { wired: true });
      ClusterFactory.extend(cluster, PowerSource2.BatteryComponent, { battery: true });
      ClusterFactory.extend(cluster, PowerSource2.ReplaceableComponent, { replaceable: true });
      ClusterFactory.extend(cluster, PowerSource2.RechargeableComponent, { rechargeable: true });
      return cluster;
    }
  );
  const WIRED = { wired: true };
  const BAT = { battery: true };
  const REPLC = { replaceable: true };
  const RECHG = { rechargeable: true };
  PowerSource2.Complete = ClusterFactory.Definition({
    id: PowerSource2.Cluster.id,
    name: PowerSource2.Cluster.name,
    revision: PowerSource2.Cluster.revision,
    features: PowerSource2.Cluster.features,
    attributes: {
      ...PowerSource2.Cluster.attributes,
      wiredAssessedInputVoltage: ClusterFactory.AsConditional(
        PowerSource2.WiredComponent.attributes.wiredAssessedInputVoltage,
        { optionalIf: [WIRED] }
      ),
      wiredAssessedInputFrequency: ClusterFactory.AsConditional(
        PowerSource2.WiredComponent.attributes.wiredAssessedInputFrequency,
        { optionalIf: [WIRED] }
      ),
      wiredCurrentType: ClusterFactory.AsConditional(
        PowerSource2.WiredComponent.attributes.wiredCurrentType,
        { mandatoryIf: [WIRED] }
      ),
      wiredAssessedCurrent: ClusterFactory.AsConditional(
        PowerSource2.WiredComponent.attributes.wiredAssessedCurrent,
        { optionalIf: [WIRED] }
      ),
      wiredNominalVoltage: ClusterFactory.AsConditional(
        PowerSource2.WiredComponent.attributes.wiredNominalVoltage,
        { optionalIf: [WIRED] }
      ),
      wiredMaximumCurrent: ClusterFactory.AsConditional(
        PowerSource2.WiredComponent.attributes.wiredMaximumCurrent,
        { optionalIf: [WIRED] }
      ),
      wiredPresent: ClusterFactory.AsConditional(PowerSource2.WiredComponent.attributes.wiredPresent, { optionalIf: [WIRED] }),
      activeWiredFaults: ClusterFactory.AsConditional(
        PowerSource2.WiredComponent.attributes.activeWiredFaults,
        { optionalIf: [WIRED] }
      ),
      batVoltage: ClusterFactory.AsConditional(PowerSource2.BatteryComponent.attributes.batVoltage, { optionalIf: [BAT] }),
      batPercentRemaining: ClusterFactory.AsConditional(
        PowerSource2.BatteryComponent.attributes.batPercentRemaining,
        { optionalIf: [BAT] }
      ),
      batTimeRemaining: ClusterFactory.AsConditional(
        PowerSource2.BatteryComponent.attributes.batTimeRemaining,
        { optionalIf: [BAT] }
      ),
      batChargeLevel: ClusterFactory.AsConditional(
        PowerSource2.BatteryComponent.attributes.batChargeLevel,
        { mandatoryIf: [BAT] }
      ),
      batReplacementNeeded: ClusterFactory.AsConditional(
        PowerSource2.BatteryComponent.attributes.batReplacementNeeded,
        { mandatoryIf: [BAT] }
      ),
      batReplaceability: ClusterFactory.AsConditional(
        PowerSource2.BatteryComponent.attributes.batReplaceability,
        { mandatoryIf: [BAT] }
      ),
      batPresent: ClusterFactory.AsConditional(PowerSource2.BatteryComponent.attributes.batPresent, { optionalIf: [BAT] }),
      activeBatFaults: ClusterFactory.AsConditional(
        PowerSource2.BatteryComponent.attributes.activeBatFaults,
        { optionalIf: [BAT] }
      ),
      batReplacementDescription: ClusterFactory.AsConditional(
        PowerSource2.ReplaceableComponent.attributes.batReplacementDescription,
        { mandatoryIf: [REPLC] }
      ),
      batCommonDesignation: ClusterFactory.AsConditional(
        PowerSource2.ReplaceableComponent.attributes.batCommonDesignation,
        { optionalIf: [REPLC] }
      ),
      batAnsiDesignation: ClusterFactory.AsConditional(
        PowerSource2.ReplaceableComponent.attributes.batAnsiDesignation,
        { optionalIf: [REPLC] }
      ),
      batIecDesignation: ClusterFactory.AsConditional(
        PowerSource2.ReplaceableComponent.attributes.batIecDesignation,
        { optionalIf: [REPLC] }
      ),
      batApprovedChemistry: ClusterFactory.AsConditional(
        PowerSource2.ReplaceableComponent.attributes.batApprovedChemistry,
        { optionalIf: [REPLC] }
      ),
      batCapacity: ClusterFactory.AsConditional(
        PowerSource2.ReplaceableComponent.attributes.batCapacity,
        { optionalIf: [REPLC] }
      ),
      batQuantity: ClusterFactory.AsConditional(
        PowerSource2.ReplaceableComponent.attributes.batQuantity,
        { mandatoryIf: [REPLC] }
      ),
      batChargeState: ClusterFactory.AsConditional(
        PowerSource2.RechargeableComponent.attributes.batChargeState,
        { mandatoryIf: [RECHG] }
      ),
      batTimeToFullCharge: ClusterFactory.AsConditional(
        PowerSource2.RechargeableComponent.attributes.batTimeToFullCharge,
        { optionalIf: [RECHG] }
      ),
      batFunctionalWhileCharging: ClusterFactory.AsConditional(
        PowerSource2.RechargeableComponent.attributes.batFunctionalWhileCharging,
        { mandatoryIf: [RECHG] }
      ),
      batChargingCurrent: ClusterFactory.AsConditional(
        PowerSource2.RechargeableComponent.attributes.batChargingCurrent,
        { optionalIf: [RECHG] }
      ),
      activeBatChargeFaults: ClusterFactory.AsConditional(
        PowerSource2.RechargeableComponent.attributes.activeBatChargeFaults,
        { optionalIf: [RECHG] }
      )
    },
    events: {
      wiredFaultChange: ClusterFactory.AsConditional(
        PowerSource2.WiredComponent.events.wiredFaultChange,
        { optionalIf: [WIRED] }
      ),
      batFaultChange: ClusterFactory.AsConditional(PowerSource2.BatteryComponent.events.batFaultChange, { optionalIf: [BAT] }),
      batChargeFaultChange: ClusterFactory.AsConditional(
        PowerSource2.RechargeableComponent.events.batChargeFaultChange,
        { optionalIf: [RECHG] }
      )
    }
  });
})(PowerSource || (PowerSource = {}));
const PowerSourceCluster = PowerSource.Cluster;
export {
  PowerSource,
  PowerSourceCluster
};
