// @project-chip/matter.js/dist/esm/cluster/definitions/PressureMeasurementCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { Attribute, OptionalAttribute } from "../../cluster/Cluster.js";
import { TlvInt16, TlvUInt16, TlvInt8 } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
var PressureMeasurement;
((PressureMeasurement2) => {
  let Feature;
  ((Feature2) => {
    Feature2["Extended"] = "Extended";
  })(Feature = PressureMeasurement2.Feature || (PressureMeasurement2.Feature = {}));
  PressureMeasurement2.Base = ClusterFactory.Definition({
    id: 1027,
    name: "PressureMeasurement",
    revision: 3,
    features: {
      /**
       * Extended
       *
       * The cluster is capable of extended range and resolution
       */
      extended: BitFlag(0)
    },
    attributes: {
      /**
       * This attribute represents the pressure in kPa as follows:
       *
       * MeasuredValue = 10 x Pressure [kPa]
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.1
       */
      measuredValue: Attribute(0, TlvNullable(TlvInt16.bound({ min: -32767 }))),
      /**
       * This attribute indicates the minimum value of MeasuredValue that can be measured. See Measured Value for
       * more details.
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.2
       */
      minMeasuredValue: Attribute(1, TlvNullable(TlvInt16.bound({ min: -32767 }))),
      /**
       * This attribute indicates the maximum value of MeasuredValue that can be measured. See Measured Value for
       * more details.
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.3
       */
      maxMeasuredValue: Attribute(2, TlvNullable(TlvInt16)),
      /**
       * This attribute indicates the magnitude of the possible error that is associated with ScaledValue.
       *
       * See Measured Value.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.4
       */
      tolerance: OptionalAttribute(3, TlvUInt16.bound({ max: 2048 }), { default: 0 })
    }
  });
  PressureMeasurement2.ExtendedComponent = ClusterFactory.Component({
    attributes: {
      /**
       * ScaledValue represents the pressure in Pascals as follows:
       *
       * ScaledValue = 10Scale x Pressure [Pa]
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.5
       */
      scaledValue: Attribute(16, TlvNullable(TlvInt16.bound({ min: -32767 })), { default: 0 }),
      /**
       * The MinScaledValue attribute indicates the minimum value of ScaledValue that can be measured. The null
       * value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.6
       */
      minScaledValue: Attribute(17, TlvNullable(TlvInt16.bound({ min: -32767 })), { default: 0 }),
      /**
       * This attribute indicates the maximum value of ScaledValue that can be measured. MaxScaledValue shall be
       * greater than MinScaledValue.
       *
       * The null value indicates that the value is not available.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.7
       */
      maxScaledValue: Attribute(18, TlvNullable(TlvInt16), { default: 0 }),
      /**
       * This attribute indicates the magnitude of the possible error that is associated with ScaledValue. The
       * true value is located in the range
       *
       * (ScaledValue – ScaledTolerance) to (ScaledValue + ScaledTolerance).
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.8
       */
      scaledTolerance: OptionalAttribute(19, TlvUInt16.bound({ max: 2048 }), { default: 0 }),
      /**
       * This attribute indicates the base 10 exponent used to obtain ScaledValue (see ScaledValue Attribute).
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 2.4.5.9
       */
      scale: Attribute(20, TlvInt8.bound({ min: -127 }), { default: 0 })
    }
  });
  PressureMeasurement2.Cluster = ClusterFactory.Extensible(
    PressureMeasurement2.Base,
    /**
     * Use this factory method to create a PressureMeasurement cluster with support for optional features. Include
     * each {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a PressureMeasurement cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...PressureMeasurement2.Base,
        supportedFeatures: BitFlags(PressureMeasurement2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, PressureMeasurement2.ExtendedComponent, { extended: true });
      return cluster;
    }
  );
  const EXT = { extended: true };
  PressureMeasurement2.Complete = ClusterFactory.Definition({
    id: PressureMeasurement2.Cluster.id,
    name: PressureMeasurement2.Cluster.name,
    revision: PressureMeasurement2.Cluster.revision,
    features: PressureMeasurement2.Cluster.features,
    attributes: {
      ...PressureMeasurement2.Cluster.attributes,
      scaledValue: ClusterFactory.AsConditional(PressureMeasurement2.ExtendedComponent.attributes.scaledValue, { mandatoryIf: [EXT] }),
      minScaledValue: ClusterFactory.AsConditional(
        PressureMeasurement2.ExtendedComponent.attributes.minScaledValue,
        { mandatoryIf: [EXT] }
      ),
      maxScaledValue: ClusterFactory.AsConditional(
        PressureMeasurement2.ExtendedComponent.attributes.maxScaledValue,
        { mandatoryIf: [EXT] }
      ),
      scaledTolerance: ClusterFactory.AsConditional(
        PressureMeasurement2.ExtendedComponent.attributes.scaledTolerance,
        { optionalIf: [EXT] }
      ),
      scale: ClusterFactory.AsConditional(PressureMeasurement2.ExtendedComponent.attributes.scale, { mandatoryIf: [EXT] })
    }
  });
})(PressureMeasurement || (PressureMeasurement = {}));
const PressureMeasurementCluster = PressureMeasurement.Cluster;
export {
  PressureMeasurement,
  PressureMeasurementCluster
};
