// @project-chip/matter.js/dist/esm/cluster/definitions/ProxyConfigurationCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { WritableAttribute } from "../../cluster/Cluster.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
import { TlvNodeId } from "../../datatype/NodeId.js";
var ProxyConfiguration;
((ProxyConfiguration2) => {
  ProxyConfiguration2.TlvConfigurationStruct = TlvObject({
    proxyAllNodes: TlvField(1, TlvBoolean),
    sourceList: TlvField(2, TlvArray(TlvNodeId))
  });
  ProxyConfiguration2.Cluster = ClusterFactory.Definition({
    id: 66,
    name: "ProxyConfiguration",
    revision: 1,
    attributes: {
      /**
       * List of proxy configurations. There shall NOT be multiple entries in this list for the same fabric.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.15.14.5.1
       */
      configurationList: WritableAttribute(
        0,
        TlvArray(ProxyConfiguration2.TlvConfigurationStruct),
        { persistent: true, default: [] }
      )
    }
  });
})(ProxyConfiguration || (ProxyConfiguration = {}));
const ProxyConfigurationCluster = ProxyConfiguration.Cluster;
export {
  ProxyConfiguration,
  ProxyConfigurationCluster
};
