// @project-chip/matter.js/dist/esm/cluster/definitions/PulseWidthModulationCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import {
  Attribute,
  OptionalAttribute,
  WritableAttribute,
  OptionalWritableAttribute,
  Command,
  TlvNoResponse,
  AccessLevel
} from "../../cluster/Cluster.js";
import { TlvUInt8, TlvBitmap, TlvUInt16, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var PulseWidthModulation;
((PulseWidthModulation2) => {
  PulseWidthModulation2.Options = { executeIfOff: BitFlag(0), coupleColorTempToLevel: BitFlag(1) };
  PulseWidthModulation2.TlvMoveToLevelRequest = TlvObject({
    level: TlvField(0, TlvUInt8.bound({ max: 254 })),
    transitionTime: TlvField(1, TlvNullable(TlvUInt16)),
    optionsMask: TlvField(2, TlvBitmap(TlvUInt8, PulseWidthModulation2.Options)),
    optionsOverride: TlvField(3, TlvBitmap(TlvUInt8, PulseWidthModulation2.Options))
  });
  let MoveMode;
  ((MoveMode2) => {
    MoveMode2[MoveMode2["Up"] = 0] = "Up";
    MoveMode2[MoveMode2["Down"] = 1] = "Down";
  })(MoveMode = PulseWidthModulation2.MoveMode || (PulseWidthModulation2.MoveMode = {}));
  PulseWidthModulation2.TlvMoveRequest = TlvObject({
    /**
     * The MoveMode field shall be one of the non-reserved values in Values of the MoveMode Field.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.2.1
     */
    moveMode: TlvField(0, TlvEnum()),
    /**
     * The Rate field specifies the rate of movement in units per second. The actual rate of movement SHOULD be as
     * close to this rate as the device is able. If the Rate field is equal to null, then the value in
     * DefaultMoveRate attribute shall be used. However, if the Rate field is equal to null and the DefaultMoveRate
     * attribute is not supported, or if the Rate field is equal to null and the value of the DefaultMoveRate
     * attribute is equal to null, then the device SHOULD move as fast as it is able. If the device is not able to
     * move at a variable rate, this field may be disregarded.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.2.2
     */
    rate: TlvField(1, TlvNullable(TlvUInt8)),
    optionsMask: TlvField(2, TlvBitmap(TlvUInt8, PulseWidthModulation2.Options)),
    optionsOverride: TlvField(3, TlvBitmap(TlvUInt8, PulseWidthModulation2.Options))
  });
  PulseWidthModulation2.TlvStepRequest = TlvObject({
    stepMode: TlvField(0, TlvUInt8),
    stepSize: TlvField(1, TlvUInt8),
    transitionTime: TlvField(2, TlvNullable(TlvUInt16)),
    optionsMask: TlvField(3, TlvBitmap(TlvUInt8, PulseWidthModulation2.Options)),
    optionsOverride: TlvField(4, TlvBitmap(TlvUInt8, PulseWidthModulation2.Options))
  });
  PulseWidthModulation2.TlvStopRequest = TlvObject({
    optionsMask: TlvField(0, TlvBitmap(TlvUInt8, PulseWidthModulation2.Options)),
    optionsOverride: TlvField(1, TlvBitmap(TlvUInt8, PulseWidthModulation2.Options))
  });
  PulseWidthModulation2.TlvMoveToClosestFrequencyRequest = TlvObject({ frequency: TlvField(0, TlvUInt16) });
  let Feature;
  ((Feature2) => {
    Feature2["OnOff"] = "OnOff";
    Feature2["Lighting"] = "Lighting";
    Feature2["Frequency"] = "Frequency";
  })(Feature = PulseWidthModulation2.Feature || (PulseWidthModulation2.Feature = {}));
  PulseWidthModulation2.Base = ClusterFactory.Definition({
    id: 28,
    name: "PulseWidthModulation",
    revision: 5,
    features: {
      /**
       * OnOff
       *
       * Dependency with the On/Off cluster
       */
      onOff: BitFlag(0),
      /**
       * Lighting
       *
       * Behavior that supports lighting applications
       */
      lighting: BitFlag(1),
      /**
       * Frequency
       *
       * Supports frequency attributes and behavior. The Pulse Width Modulation cluster was created for frequency
       * control.
       */
      frequency: BitFlag(2)
    },
    attributes: {
      /**
       * The CurrentLevel attribute represents the current level of this device. The meaning of 'level' is device
       * dependent.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.1
       */
      currentLevel: Attribute(
        0,
        TlvNullable(TlvUInt8.bound({ max: 254 })),
        { scene: true, persistent: true, default: null }
      ),
      /**
       * The MinLevel attribute indicates the minimum value of CurrentLevel that is capable of being assigned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.3
       */
      minLevel: OptionalAttribute(2, TlvUInt8),
      /**
       * The MaxLevel attribute indicates the maximum value of CurrentLevel that is capable of being assigned.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.4
       */
      maxLevel: OptionalAttribute(3, TlvUInt8.bound({ max: 254 }), { default: 254 }),
      /**
       * The Options attribute is meant to be changed only during commissioning. The Options attribute is a
       * bitmap that determines the default behavior of some cluster commands. Each command that is dependent on
       * the Options attribute shall first construct a temporary Options bitmap that is in effect during the
       * command processing. The temporary Options bitmap has the same format and meaning as the Options
       * attribute, but includes any bits that may be overridden by command fields.
       *
       * Below is the format and description of the Options attribute and temporary Options bitmap and the effect
       * on dependent commands.
       *
       * Table 19. Options Attribute
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.8
       */
      options: WritableAttribute(15, TlvBitmap(TlvUInt8, PulseWidthModulation2.Options)),
      /**
       * The OnOffTransitionTime attribute represents the time taken to move to or from the target level when On
       * or Off commands are received by an On/Off cluster on the same endpoint. It is specified in 1/10ths of a
       * second.
       *
       * The actual time taken SHOULD be as close to OnOffTransitionTime as the device is able. Please note that
       * if the device is not able to move at a variable rate, the OnOffTransitionTime attribute SHOULD NOT be
       * implemented.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.9
       */
      onOffTransitionTime: OptionalWritableAttribute(16, TlvUInt16, { default: 0 }),
      /**
       * The OnLevel attribute determines the value that the CurrentLevel attribute is set to when the OnOff
       * attribute of an On/Off cluster on the same endpoint is set to TRUE, as a result of processing an On/Off
       * cluster command. If the OnLevel attribute is not implemented, or is set to the null value, it has no
       * effect. For more details see Effect of On/Off Commands on the CurrentLevel Attribute.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.10
       */
      onLevel: WritableAttribute(17, TlvNullable(TlvUInt8.bound({ max: 254 })), { default: null }),
      /**
       * The OnTransitionTime attribute represents the time taken to move the current level from the minimum
       * level to the maximum level when an On command is received by an On/Off cluster on the same endpoint. It
       * is specified in 10ths of a second. If this attribute is not implemented, or contains a null value, the
       * OnOffTransitionTime will be used instead.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.11
       */
      onTransitionTime: OptionalWritableAttribute(18, TlvNullable(TlvUInt16), { default: null }),
      /**
       * The OffTransitionTime attribute represents the time taken to move the current level from the maximum
       * level to the minimum level when an Off command is received by an On/Off cluster on the same endpoint. It
       * is specified in 10ths of a second. If this attribute is not implemented, or contains a null value, the
       * OnOffTransitionTime will be used instead.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.12
       */
      offTransitionTime: OptionalWritableAttribute(19, TlvNullable(TlvUInt16), { default: null }),
      /**
       * The DefaultMoveRate attribute determines the movement rate, in units per second, when a Move command is
       * received with a null value Rate parameter.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.13
       */
      defaultMoveRate: OptionalWritableAttribute(20, TlvNullable(TlvUInt8))
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.1
       */
      moveToLevel: Command(0, PulseWidthModulation2.TlvMoveToLevelRequest, 0, TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.2
       */
      move: Command(1, PulseWidthModulation2.TlvMoveRequest, 1, TlvNoResponse),
      /**
       * The StepMode field shall be one of the non-reserved values in Values of the StepMode Field.
       *
       * The TransitionTime field specifies the time that shall be taken to perform the step, in tenths of a
       * second. A step is a change in the CurrentLevel of StepSize units. The actual time taken SHOULD be as
       * close to this as the device is able. If the TransitionTime field is equal to null, the device SHOULD
       * move as fast as it is able.
       *
       * If the device is not able to move at a variable rate, the TransitionTime field may be disregarded.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.3
       */
      step: Command(2, PulseWidthModulation2.TlvStepRequest, 2, TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.4
       */
      stop: Command(3, PulseWidthModulation2.TlvStopRequest, 3, TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6
       */
      moveToLevelWithOnOff: Command(4, TlvNoArguments, 4, TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6
       */
      moveWithOnOff: Command(5, TlvNoArguments, 5, TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6
       */
      stepWithOnOff: Command(6, TlvNoArguments, 6, TlvNoResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6
       */
      stopWithOnOff: Command(7, TlvNoArguments, 7, TlvNoResponse)
    }
  });
  PulseWidthModulation2.LightingComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The RemainingTime attribute represents the time remaining until the current command is complete - it is
       * specified in 1/10ths of a second.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.2
       */
      remainingTime: Attribute(1, TlvUInt16, { default: 0 }),
      /**
       * The StartUpCurrentLevel attribute shall define the desired startup level for a device when it is
       * supplied with power and this level shall be reflected in the CurrentLevel attribute. The values of the
       * StartUpCurrentLevel attribute are listed below:
       *
       * Table 20. Values of the StartUpCurrentLevel attribute
       *
       * This behavior does not apply to reboots associated with OTA. After an OTA restart, the CurrentLevel
       * attribute shall return to its value prior to the restart.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.14
       */
      startUpCurrentLevel: WritableAttribute(
        16384,
        TlvNullable(TlvUInt8),
        { persistent: true, writeAcl: AccessLevel.Manage }
      )
    }
  });
  PulseWidthModulation2.FrequencyComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The CurrentFrequency attribute represents the frequency at which the device is at CurrentLevel. A
       * CurrentFrequency of 0 is unknown.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.5
       */
      currentFrequency: Attribute(4, TlvUInt16, { scene: true, default: 0 }),
      /**
       * The MinFrequency attribute indicates the minimum value of CurrentFrequency that is capable of being
       * assigned. MinFrequency shall be less than or equal to MaxFrequency. A value of 0 indicates undefined.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.6
       */
      minFrequency: Attribute(5, TlvUInt16, { default: 0 }),
      /**
       * The MaxFrequency attribute indicates the maximum value of CurrentFrequency that is capable of being
       * assigned. MaxFrequency shall be greater than or equal to MinFrequency. A value of 0 indicates undefined.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.5.7
       */
      maxFrequency: Attribute(6, TlvUInt16, { default: 0 })
    },
    commands: {
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.6.6.5
       */
      moveToClosestFrequency: Command(8, PulseWidthModulation2.TlvMoveToClosestFrequencyRequest, 8, TlvNoResponse)
    }
  });
  PulseWidthModulation2.Cluster = ClusterFactory.Extensible(
    { ...PulseWidthModulation2.Base, supportedFeatures: { onOff: true } },
    /**
     * Use this factory method to create a PulseWidthModulation cluster with support for optional features. Include
     * each {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a PulseWidthModulation cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...PulseWidthModulation2.Base,
        supportedFeatures: BitFlags(PulseWidthModulation2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, PulseWidthModulation2.LightingComponent, { lighting: true });
      ClusterFactory.extend(cluster, PulseWidthModulation2.FrequencyComponent, { frequency: true });
      return cluster;
    }
  );
  const LT = { lighting: true };
  const FQ = { frequency: true };
  PulseWidthModulation2.Complete = ClusterFactory.Definition({
    id: PulseWidthModulation2.Cluster.id,
    name: PulseWidthModulation2.Cluster.name,
    revision: PulseWidthModulation2.Cluster.revision,
    features: PulseWidthModulation2.Cluster.features,
    attributes: {
      ...PulseWidthModulation2.Cluster.attributes,
      remainingTime: ClusterFactory.AsConditional(
        PulseWidthModulation2.LightingComponent.attributes.remainingTime,
        { mandatoryIf: [LT] }
      ),
      currentFrequency: ClusterFactory.AsConditional(
        PulseWidthModulation2.FrequencyComponent.attributes.currentFrequency,
        { mandatoryIf: [FQ] }
      ),
      minFrequency: ClusterFactory.AsConditional(
        PulseWidthModulation2.FrequencyComponent.attributes.minFrequency,
        { mandatoryIf: [FQ] }
      ),
      maxFrequency: ClusterFactory.AsConditional(
        PulseWidthModulation2.FrequencyComponent.attributes.maxFrequency,
        { mandatoryIf: [FQ] }
      ),
      startUpCurrentLevel: ClusterFactory.AsConditional(
        PulseWidthModulation2.LightingComponent.attributes.startUpCurrentLevel,
        { mandatoryIf: [LT] }
      )
    },
    commands: {
      ...PulseWidthModulation2.Cluster.commands,
      moveToClosestFrequency: ClusterFactory.AsConditional(
        PulseWidthModulation2.FrequencyComponent.commands.moveToClosestFrequency,
        { mandatoryIf: [FQ] }
      )
    }
  });
})(PulseWidthModulation || (PulseWidthModulation = {}));
const PulseWidthModulationCluster = PulseWidthModulation.Cluster;
export {
  PulseWidthModulation,
  PulseWidthModulationCluster
};
