// @project-chip/matter.js/dist/esm/cluster/definitions/ScenesCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitField, BitFlags } from "../../schema/BitmapSchema.js";
import {
  Attribute,
  OptionalAttribute,
  Command,
  AccessLevel,
  TlvNoResponse,
  OptionalCommand
} from "../../cluster/Cluster.js";
import { TlvUInt8, TlvBitmap, TlvUInt16, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvGroupId, GroupId } from "../../datatype/GroupId.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
import { TlvNodeId } from "../../datatype/NodeId.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvObject, TlvField, TlvOptionalField } from "../../tlv/TlvObject.js";
import { TlvString } from "../../tlv/TlvString.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvClusterId } from "../../datatype/ClusterId.js";
import { TlvAttributeId } from "../../datatype/AttributeId.js";
import { TlvAny } from "../../tlv/TlvAny.js";
var Scenes;
((Scenes2) => {
  Scenes2.NameSupport = {
    /**
     * The ability to store a name for a scene.
     */
    nameSupport: BitFlag(7)
  };
  Scenes2.TlvAttributeValuePair = TlvObject({
    /**
     * This field shall be present or not present, for all instances in the Scenes cluster. If this field is not
     * present, then the data type of AttributeValue shall be determined by the order and data type defined in the
     * cluster specification. Otherwise the data type of AttributeValue shall be the data type of the attribute
     * indicated by AttributeID.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.6.1.1
     */
    attributeId: TlvOptionalField(0, TlvAttributeId),
    /**
     * This is the attribute value as part of an extension field set. See AttributeID to determine the data type
     * for this field.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.6.1.2
     */
    attributeValue: TlvField(1, TlvAny)
  });
  Scenes2.TlvExtensionFieldSet = TlvObject({
    clusterId: TlvField(0, TlvClusterId),
    attributeValueList: TlvField(1, TlvArray(Scenes2.TlvAttributeValuePair))
  });
  Scenes2.TlvAddSceneRequest = TlvObject({
    groupId: TlvField(0, TlvGroupId),
    sceneId: TlvField(1, TlvUInt8),
    transitionTime: TlvField(2, TlvUInt16),
    sceneName: TlvField(3, TlvString.bound({ maxLength: 16 })),
    extensionFieldSets: TlvField(4, TlvArray(Scenes2.TlvExtensionFieldSet))
  });
  Scenes2.TlvAddSceneResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    groupId: TlvField(1, TlvGroupId),
    sceneId: TlvField(2, TlvUInt8)
  });
  Scenes2.TlvViewSceneRequest = TlvObject({ groupId: TlvField(0, TlvGroupId), sceneId: TlvField(1, TlvUInt8) });
  Scenes2.TlvViewSceneResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    groupId: TlvField(1, TlvGroupId),
    sceneId: TlvField(2, TlvUInt8),
    transitionTime: TlvOptionalField(3, TlvUInt16),
    sceneName: TlvOptionalField(4, TlvString.bound({ maxLength: 16 })),
    extensionFieldSets: TlvOptionalField(5, TlvArray(Scenes2.TlvExtensionFieldSet))
  });
  Scenes2.TlvRemoveSceneRequest = TlvObject({
    groupId: TlvField(0, TlvGroupId),
    sceneId: TlvField(1, TlvUInt8)
  });
  Scenes2.TlvRemoveSceneResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    groupId: TlvField(1, TlvGroupId),
    sceneId: TlvField(2, TlvUInt8)
  });
  Scenes2.TlvRemoveAllScenesRequest = TlvObject({ groupId: TlvField(0, TlvGroupId) });
  Scenes2.TlvRemoveAllScenesResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    groupId: TlvField(1, TlvGroupId)
  });
  Scenes2.TlvStoreSceneRequest = TlvObject({ groupId: TlvField(0, TlvGroupId), sceneId: TlvField(1, TlvUInt8) });
  Scenes2.TlvStoreSceneResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    groupId: TlvField(1, TlvGroupId),
    sceneId: TlvField(2, TlvUInt8)
  });
  Scenes2.TlvRecallSceneRequest = TlvObject({
    groupId: TlvField(0, TlvGroupId),
    sceneId: TlvField(1, TlvUInt8),
    transitionTime: TlvOptionalField(2, TlvNullable(TlvUInt16))
  });
  Scenes2.TlvGetSceneMembershipRequest = TlvObject({ groupId: TlvField(0, TlvGroupId) });
  Scenes2.TlvGetSceneMembershipResponse = TlvObject({
    status: TlvField(0, TlvEnum()),
    capacity: TlvField(1, TlvNullable(TlvUInt8)),
    groupId: TlvField(2, TlvGroupId),
    sceneList: TlvOptionalField(3, TlvArray(TlvUInt8))
  });
  Scenes2.Mode = { copyAllScenes: BitFlag(0), reserved: BitField(1, 7) };
  Scenes2.TlvCopySceneRequest = TlvObject({
    /**
     * The Mode field contains information of how the scene copy is to proceed. This field shall be formatted as
     * illustrated in Format of the Mode Field of the CopyScene Command.
     *
     * The CopyAllScenes subfield is 1-bit in length and indicates whether all scenes are to be copied. If this
     * value is set to 1, all scenes are to be copied and the SceneIdentifierFrom and SceneIdentifierTo fields
     * shall be ignored. Otherwise this field is set to 0.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11.1
     */
    mode: TlvField(0, TlvBitmap(TlvUInt8, Scenes2.Mode)),
    /**
     * The GroupIdentifierFrom field specifies the identifier of the group from which the scene is to be copied.
     * Together with the SceneIdentifierFrom field, this field uniquely identifies the scene to copy from the Scene
     * Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11.2
     */
    groupIdentifierFrom: TlvField(1, TlvGroupId),
    /**
     * The SceneIdentifierFrom field specifies the identifier of the scene from which the scene is to be copied.
     * Together with the GroupIdentifierFrom field, this field uniquely identifies the scene to copy from the Scene
     * Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11.3
     */
    sceneIdentifierFrom: TlvField(2, TlvUInt8),
    /**
     * The GroupIdentifierTo field specifies the identifier of the group to which the scene is to be copied.
     * Together with the SceneIdentifierTo field, this field uniquely identifies the scene to copy to the Scene
     * Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11.4
     */
    groupIdentifierTo: TlvField(3, TlvGroupId),
    /**
     * The SceneIdentifierTo field specifies the identifier of the scene to which the scene is to be copied.
     * Together with the GroupIdentifierTo field, this field uniquely identifies the scene to copy to the Scene
     * Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11.5
     */
    sceneIdentifierTo: TlvField(4, TlvUInt8)
  });
  Scenes2.TlvCopySceneResponse = TlvObject({
    /**
     * The Status field contains the status of the copy scene attempt. This field shall be set to one of the
     * non-reserved values listed in Values of the Status Field of the CopySceneResponse Command.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.20.1
     */
    status: TlvField(0, TlvEnum()),
    /**
     * The GroupIdentifierFrom field specifies the identifier of the group from which the scene was copied, as
     * specified in the CopyScene command. Together with the SceneIdentifierFrom field, this field uniquely
     * identifies the scene that was copied from the Scene Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.20.2
     */
    groupIdentifierFrom: TlvField(1, TlvGroupId),
    /**
     * The SceneIdentifierFrom field is specifies the identifier of the scene from which the scene was copied, as
     * specified in the CopyScene command. Together with the GroupIdentifierFrom field, this field uniquely
     * identifies the scene that was copied from the Scene Table.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.20.3
     */
    sceneIdentifierFrom: TlvField(2, TlvUInt8)
  });
  let Feature;
  ((Feature2) => {
    Feature2["SceneNames"] = "SceneNames";
  })(Feature = Scenes2.Feature || (Scenes2.Feature = {}));
  Scenes2.Base = ClusterFactory.Definition({
    id: 5,
    name: "Scenes",
    revision: 4,
    features: {
      /**
       * SceneNames
       *
       * The ability to store a name for a scene.
       */
      sceneNames: BitFlag(0)
    },
    attributes: {
      /**
       * The SceneCount attribute specifies the number of scenes currently in the server’s Scene Table.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.1
       */
      sceneCount: Attribute(0, TlvUInt8, { default: 0 }),
      /**
       * The CurrentScene attribute holds the scene identifier of the scene last invoked.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.2
       */
      currentScene: Attribute(1, TlvUInt8, { default: 0 }),
      /**
       * The CurrentGroup attribute holds the group identifier of the scene last invoked, or 0 if the scene last
       * invoked is not associated with a group.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.3
       */
      currentGroup: Attribute(2, TlvGroupId, { default: GroupId(0) }),
      /**
       * The SceneValid attribute indicates whether the state of the server corresponds to that associated with
       * the CurrentScene and CurrentGroup attributes. TRUE indicates that these attributes are valid, FALSE
       * indicates that they are not valid.
       *
       * Before a scene has been stored or recalled, this attribute is set to FALSE. After a successful
       * StoreScene or RecallScene command it is set to TRUE. If, after a scene is stored or recalled, the state
       * of the server is modified, this attribute is set to FALSE.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.4
       */
      sceneValid: Attribute(3, TlvBoolean, { default: true }),
      /**
       * This attribute provides legacy, read-only access to whether the Scene Names feature is supported. The
       * most significant bit, bit 7, shall be equal to bit 0 of the FeatureMap attribute. All other bits shall
       * be 0.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.5
       */
      nameSupport: Attribute(4, TlvBitmap(TlvUInt8, Scenes2.NameSupport)),
      /**
       * The LastConfiguredBy attribute holds the Node ID (the IEEE address in case of Zigbee) of the node that
       * last configured the Scene Table.
       *
       * The null value indicates that the server has not been configured, or that the identifier of the node
       * that last configured the Scenes cluster is not known.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.7.6
       */
      lastConfiguredBy: OptionalAttribute(5, TlvNullable(TlvNodeId), { default: null })
    },
    commands: {
      /**
       * It is not mandatory for an extension field set to be included in the command for every cluster on that
       * endpoint that has a defined extension field set. Extension field sets may be omitted, including the case
       * of no extension field sets at all.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.2
       */
      addScene: Command(0, Scenes2.TlvAddSceneRequest, 0, Scenes2.TlvAddSceneResponse, { invokeAcl: AccessLevel.Manage }),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.3
       */
      viewScene: Command(1, Scenes2.TlvViewSceneRequest, 1, Scenes2.TlvViewSceneResponse),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.4
       */
      removeScene: Command(
        2,
        Scenes2.TlvRemoveSceneRequest,
        2,
        Scenes2.TlvRemoveSceneResponse,
        { invokeAcl: AccessLevel.Manage }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.5
       */
      removeAllScenes: Command(
        3,
        Scenes2.TlvRemoveAllScenesRequest,
        3,
        Scenes2.TlvRemoveAllScenesResponse,
        { invokeAcl: AccessLevel.Manage }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.6
       */
      storeScene: Command(
        4,
        Scenes2.TlvStoreSceneRequest,
        4,
        Scenes2.TlvStoreSceneResponse,
        { invokeAcl: AccessLevel.Manage }
      ),
      /**
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.7
       */
      recallScene: Command(5, Scenes2.TlvRecallSceneRequest, 5, TlvNoResponse),
      /**
       * The GetSceneMembership command can be used to find an unused scene identifier within a certain group
       * when no commissioning tool is in the network, or for a commissioning tool to get the used scene
       * identifiers within a certain group, for the endpoint that implements this cluster.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.8
       */
      getSceneMembership: Command(6, Scenes2.TlvGetSceneMembershipRequest, 6, Scenes2.TlvGetSceneMembershipResponse),
      /**
       * The EnhancedAddScene command allows a scene to be added using a finer scene transition time than the
       * AddScene command.
       *
       * This command shall have the same data fields as the AddScene command, with the following difference:
       *
       * The TransitionTime data field shall be measured in tenths of a second rather than in seconds.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.9
       */
      enhancedAddScene: OptionalCommand(
        64,
        Scenes2.TlvAddSceneRequest,
        64,
        Scenes2.TlvAddSceneResponse,
        { invokeAcl: AccessLevel.Manage }
      ),
      /**
       * The EnhancedViewScene command allows a scene to be retrieved using a finer scene transition time than
       * the ViewScene command.
       *
       * This command shall have the same data fields as the ViewScene command.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.10
       */
      enhancedViewScene: OptionalCommand(65, Scenes2.TlvViewSceneRequest, 65, Scenes2.TlvViewSceneResponse),
      /**
       * The CopyScene command allows a client to efficiently copy scenes from one group/scene identifier pair to
       * another group/scene identifier pair.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 1.4.9.11
       */
      copyScene: OptionalCommand(
        66,
        Scenes2.TlvCopySceneRequest,
        66,
        Scenes2.TlvCopySceneResponse,
        { invokeAcl: AccessLevel.Manage }
      )
    }
  });
  Scenes2.Cluster = ClusterFactory.Extensible(
    { ...Scenes2.Base, supportedFeatures: { sceneNames: true } },
    /**
     * Use this factory method to create a Scenes cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a Scenes cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...Scenes2.Base,
        supportedFeatures: BitFlags(Scenes2.Base.features, ...features)
      });
      return cluster;
    }
  );
})(Scenes || (Scenes = {}));
const ScenesCluster = Scenes.Cluster;
export {
  Scenes,
  ScenesCluster
};
