// @project-chip/matter.js/dist/esm/cluster/definitions/TargetNavigatorCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { Attribute, OptionalAttribute, Command } from "../../cluster/Cluster.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvObject, TlvField, TlvOptionalField } from "../../tlv/TlvObject.js";
import { TlvUInt8, TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvString } from "../../tlv/TlvString.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
var TargetNavigator;
((TargetNavigator2) => {
  TargetNavigator2.TlvTargetInfoStruct = TlvObject({
    /**
     * An unique id within the TargetList.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.5.1.1
     */
    identifier: TlvField(0, TlvUInt8),
    /**
     * A name string for the TargetInfoStruct.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.5.1.2
     */
    name: TlvField(1, TlvString.bound({ maxLength: 32 }))
  });
  TargetNavigator2.TlvNavigateTargetRequest = TlvObject({
    /**
     * This shall indicate the Identifier for the target for UX navigation. The Target shall be an Identifier value
     * contained within one of the TargetInfoStruct objects in the TargetList attribute list.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.4.1.1
     */
    target: TlvField(0, TlvUInt8),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.4.1.2
     */
    data: TlvOptionalField(1, TlvString)
  });
  let Status;
  ((Status2) => {
    Status2[Status2["Success"] = 0] = "Success";
    Status2[Status2["TargetNotFound"] = 1] = "TargetNotFound";
    Status2[Status2["NotAllowed"] = 2] = "NotAllowed";
  })(Status = TargetNavigator2.Status || (TargetNavigator2.Status = {}));
  TargetNavigator2.TlvNavigateTargetResponse = TlvObject({
    /**
     * This shall indicate the of the command.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.4.2.1
     */
    status: TlvField(0, TlvEnum()),
    /**
     * This shall indicate Optional app-specific data.
     *
     * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.4.2.2
     */
    data: TlvOptionalField(1, TlvString)
  });
  TargetNavigator2.Cluster = ClusterFactory.Definition({
    id: 1285,
    name: "TargetNavigator",
    revision: 1,
    attributes: {
      /**
       * The TargetList attribute shall represent a list of targets that can be navigated to within the
       * experience presented to the user by the Endpoint (Video Player or Content App). The list shall not
       * contain any entries with the same Identifier in the TargetInfoStruct object.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.3.1
       */
      targetList: Attribute(0, TlvArray(TargetNavigator2.TlvTargetInfoStruct, { maxLength: 254 }), { default: [] }),
      /**
       * The CurrentTarget attribute shall represent the Identifier for the target which is currently in
       * foreground on the corresponding Endpoint (Video Player or Content App), or null to indicate that no
       * target is in the foreground.
       *
       * When not null, the CurrentTarget shall be an Identifier value contained within one of the
       * TargetInfoStruct objects in the TargetList attribute list.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.3.2
       */
      currentTarget: OptionalAttribute(1, TlvNullable(TlvUInt8), { default: null })
    },
    commands: {
      /**
       * Upon receipt, this shall navigation the UX to the target identified.
       *
       * @see {@link MatterApplicationClusterSpecificationV1_1} § 6.11.4.1
       */
      navigateTarget: Command(0, TargetNavigator2.TlvNavigateTargetRequest, 1, TargetNavigator2.TlvNavigateTargetResponse)
    }
  });
})(TargetNavigator || (TargetNavigator = {}));
const TargetNavigatorCluster = TargetNavigator.Cluster;
export {
  TargetNavigator,
  TargetNavigatorCluster
};
