// @project-chip/matter.js/dist/esm/cluster/definitions/TimeFormatLocalizationCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { WritableAttribute, AccessLevel, FixedAttribute } from "../../cluster/Cluster.js";
import { TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvArray } from "../../tlv/TlvArray.js";
var TimeFormatLocalization;
((TimeFormatLocalization2) => {
  let HourFormat;
  ((HourFormat2) => {
    HourFormat2[HourFormat2["12Hr"] = 0] = "12Hr";
    HourFormat2[HourFormat2["24Hr"] = 1] = "24Hr";
  })(HourFormat = TimeFormatLocalization2.HourFormat || (TimeFormatLocalization2.HourFormat = {}));
  let CalendarType;
  ((CalendarType2) => {
    CalendarType2[CalendarType2["Buddhist"] = 0] = "Buddhist";
    CalendarType2[CalendarType2["Chinese"] = 1] = "Chinese";
    CalendarType2[CalendarType2["Coptic"] = 2] = "Coptic";
    CalendarType2[CalendarType2["Ethiopian"] = 3] = "Ethiopian";
    CalendarType2[CalendarType2["Gregorian"] = 4] = "Gregorian";
    CalendarType2[CalendarType2["Hebrew"] = 5] = "Hebrew";
    CalendarType2[CalendarType2["Indian"] = 6] = "Indian";
    CalendarType2[CalendarType2["Islamic"] = 7] = "Islamic";
    CalendarType2[CalendarType2["Japanese"] = 8] = "Japanese";
    CalendarType2[CalendarType2["Korean"] = 9] = "Korean";
    CalendarType2[CalendarType2["Persian"] = 10] = "Persian";
    CalendarType2[CalendarType2["Taiwanese"] = 11] = "Taiwanese";
  })(CalendarType = TimeFormatLocalization2.CalendarType || (TimeFormatLocalization2.CalendarType = {}));
  let Feature;
  ((Feature2) => {
    Feature2["CalendarFormat"] = "CalendarFormat";
  })(Feature = TimeFormatLocalization2.Feature || (TimeFormatLocalization2.Feature = {}));
  TimeFormatLocalization2.Base = ClusterFactory.Definition({
    id: 44,
    name: "TimeFormatLocalization",
    revision: 1,
    features: {
      /**
       * CalendarFormat
       *
       * The Node can be configured to use different calendar formats when conveying values to a user.
       */
      calendarFormat: BitFlag(0)
    },
    attributes: {
      /**
       * The HourFormat attribute shall represent the format that the Node is currently configured to use when
       * conveying the hour unit of time. If provided, this value shall take priority over any unit
       *
       * implied through the ActiveLocale Attribute.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.4.6.1
       */
      hourFormat: WritableAttribute(
        0,
        TlvNullable(TlvEnum()),
        { persistent: true, default: null, writeAcl: AccessLevel.Manage }
      )
    }
  });
  TimeFormatLocalization2.CalendarFormatComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The ActiveCalendarType attribute shall represent the calendar format that the Node is currently
       * configured to use when conveying dates. If provided, this value shall take priority over any unit
       * implied through the ActiveLocale Attribute.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.4.6.2
       */
      activeCalendarType: WritableAttribute(
        1,
        TlvNullable(TlvEnum()),
        { persistent: true, default: null, writeAcl: AccessLevel.Manage }
      ),
      /**
       * The SupportedCalendarTypes attribute shall represent a list of CalendarTypeEnum values that are
       * supported by the Node. The list shall NOT contain any duplicate entries. The ordering of items within
       * the list SHOULD NOT express any meaning. The maximum length of the SupportedCalendarTypes list shall be
       * equivalent to the number of enumerations within CalendarTypeEnum.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.4.6.3
       */
      supportedCalendarTypes: FixedAttribute(2, TlvArray(TlvEnum()), { default: [] })
    }
  });
  TimeFormatLocalization2.Cluster = ClusterFactory.Extensible(
    TimeFormatLocalization2.Base,
    /**
     * Use this factory method to create a TimeFormatLocalization cluster with support for optional features.
     * Include each {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a TimeFormatLocalization cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...TimeFormatLocalization2.Base,
        supportedFeatures: BitFlags(TimeFormatLocalization2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, TimeFormatLocalization2.CalendarFormatComponent, { calendarFormat: true });
      return cluster;
    }
  );
  const CALFMT = { calendarFormat: true };
  TimeFormatLocalization2.Complete = ClusterFactory.Definition({
    id: TimeFormatLocalization2.Cluster.id,
    name: TimeFormatLocalization2.Cluster.name,
    revision: TimeFormatLocalization2.Cluster.revision,
    features: TimeFormatLocalization2.Cluster.features,
    attributes: {
      ...TimeFormatLocalization2.Cluster.attributes,
      activeCalendarType: ClusterFactory.AsConditional(
        TimeFormatLocalization2.CalendarFormatComponent.attributes.activeCalendarType,
        { mandatoryIf: [CALFMT] }
      ),
      supportedCalendarTypes: ClusterFactory.AsConditional(
        TimeFormatLocalization2.CalendarFormatComponent.attributes.supportedCalendarTypes,
        { mandatoryIf: [CALFMT] }
      )
    }
  });
})(TimeFormatLocalization || (TimeFormatLocalization = {}));
const TimeFormatLocalizationCluster = TimeFormatLocalization.Cluster;
export {
  TimeFormatLocalization,
  TimeFormatLocalizationCluster
};
