// @project-chip/matter.js/dist/esm/cluster/definitions/TimeSyncCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import {
  Attribute,
  OptionalAttribute,
  WritableAttribute,
  AccessLevel,
  Command,
  TlvNoResponse,
  FixedAttribute,
  Event,
  EventPriority
} from "../../cluster/Cluster.js";
import { TlvEpochUs, TlvEnum, TlvInt32, TlvUInt16 } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
import { TlvNodeId } from "../../datatype/NodeId.js";
import { TlvObject, TlvField, TlvOptionalField } from "../../tlv/TlvObject.js";
import { TlvString } from "../../tlv/TlvString.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvBoolean } from "../../tlv/TlvBoolean.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var TimeSync;
((TimeSync2) => {
  let Granularity;
  ((Granularity2) => {
    Granularity2[Granularity2["NoTimeGranularity"] = 0] = "NoTimeGranularity";
    Granularity2[Granularity2["MinutesGranularity"] = 1] = "MinutesGranularity";
    Granularity2[Granularity2["SecondsGranularity"] = 2] = "SecondsGranularity";
    Granularity2[Granularity2["MillisecondsGranularity"] = 3] = "MillisecondsGranularity";
    Granularity2[Granularity2["MicrosecondsGranularity"] = 4] = "MicrosecondsGranularity";
  })(Granularity = TimeSync2.Granularity || (TimeSync2.Granularity = {}));
  let TimeSource;
  ((TimeSource2) => {
    TimeSource2[TimeSource2["None"] = 0] = "None";
    TimeSource2[TimeSource2["Unknown"] = 1] = "Unknown";
    TimeSource2[TimeSource2["Admin"] = 2] = "Admin";
    TimeSource2[TimeSource2["NodeTimeCluster"] = 3] = "NodeTimeCluster";
    TimeSource2[TimeSource2["NonFabricSntp"] = 4] = "NonFabricSntp";
    TimeSource2[TimeSource2["NonFabricNtp"] = 5] = "NonFabricNtp";
    TimeSource2[TimeSource2["FabricSntp"] = 6] = "FabricSntp";
    TimeSource2[TimeSource2["FabricNtp"] = 7] = "FabricNtp";
    TimeSource2[TimeSource2["MixedNtp"] = 8] = "MixedNtp";
    TimeSource2[TimeSource2["NonFabricSntpNts"] = 9] = "NonFabricSntpNts";
    TimeSource2[TimeSource2["NonFabricNtpNts"] = 10] = "NonFabricNtpNts";
    TimeSource2[TimeSource2["FabricSntpNts"] = 11] = "FabricSntpNts";
    TimeSource2[TimeSource2["FabricNtpNts"] = 12] = "FabricNtpNts";
    TimeSource2[TimeSource2["MixedNtpNts"] = 13] = "MixedNtpNts";
    TimeSource2[TimeSource2["CloudSource"] = 14] = "CloudSource";
    TimeSource2[TimeSource2["Ptp"] = 15] = "Ptp";
    TimeSource2[TimeSource2["Gnss"] = 16] = "Gnss";
  })(TimeSource = TimeSync2.TimeSource || (TimeSync2.TimeSource = {}));
  TimeSync2.TlvSetUtcTimeRequest = TlvObject({
    /**
     * This shall give the Client’s UTC Time.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.9.1.1
     */
    utcTime: TlvField(0, TlvEpochUs),
    /**
     * This shall give the Client’s Granularity, as described in Section 11.16.8.2, “Granularity Attribute”.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.9.1.2
     */
    granularity: TlvField(1, TlvEnum()),
    /**
     * This shall give the Client’s TimeSource, as described in Section 11.16.8.3, “TimeSource Attribute”.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.9.1.3
     */
    timeSource: TlvOptionalField(2, TlvEnum())
  });
  let StatusCode;
  ((StatusCode2) => {
    StatusCode2[StatusCode2["TimeNotAccepted"] = 2] = "TimeNotAccepted";
  })(StatusCode = TimeSync2.StatusCode || (TimeSync2.StatusCode = {}));
  TimeSync2.TlvTimeZoneStruct = TlvObject({
    /**
     * The time zone offset from UTC in seconds.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.6.3.1
     */
    offset: TlvField(0, TlvInt32.bound({ min: -43200, max: 50400 })),
    /**
     * The UTC time when the offset shall be applied.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.6.3.2
     */
    validAt: TlvField(1, TlvEpochUs),
    /**
     * The time zone name SHOULD provide a human-readable time zone name and it SHOULD use the country/city format
     * specified by the IANA time zone database [https://www.iana.org/time-zones].
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.6.3.3
     */
    name: TlvOptionalField(2, TlvString.bound({ minLength: 0, maxLength: 64 }))
  });
  TimeSync2.TlvDSTOffsetStruct = TlvObject({
    offset: TlvField(0, TlvInt32),
    /**
     * The UTC time when the offset shall be applied.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.6.4.1
     */
    validStarting: TlvField(1, TlvEpochUs),
    /**
     * The UTC time when the offset shall stop being applied. This value shall be larger than the ValidStarting
     * time.
     *
     * @see {@link MatterCoreSpecificationV1_1} § 11.16.6.4.2
     */
    validUntil: TlvField(2, TlvEpochUs)
  });
  TimeSync2.TlvTimeZoneStatusEvent = TlvObject({
    offset: TlvField(0, TlvInt32.bound({ min: -43200, max: 50400 })),
    name: TlvOptionalField(1, TlvString.bound({ minLength: 0, maxLength: 64 }))
  });
  let Feature;
  ((Feature2) => {
    Feature2["TimeZone"] = "TimeZone";
    Feature2["NtpClient"] = "NtpClient";
    Feature2["NtpServer"] = "NtpServer";
  })(Feature = TimeSync2.Feature || (TimeSync2.Feature = {}));
  TimeSync2.Base = ClusterFactory.Definition({
    id: 56,
    name: "TimeSync",
    revision: 1,
    features: {
      /**
       * TimeZone
       *
       * Server supports time zone.
       */
      timeZone: BitFlag(0),
      /**
       * NtpClient
       *
       * Server supports an NTP or SNTP client.
       */
      ntpClient: BitFlag(1),
      /**
       * NtpServer
       *
       * Server supports an NTP server role.
       */
      ntpServer: BitFlag(2)
    },
    attributes: {
      /**
       * If the server has achieved time synchronization, this shall indicate the current time as a UTC epoch-us
       * (Epoch Time in Microseconds).
       *
       * If the server has not achieved time synchronization, this shall be null. This attribute may be set when
       * a Section 11.16.9.1, “SetUtcTime Command” is received.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.1
       */
      utcTime: Attribute(0, TlvNullable(TlvEpochUs), { omitChanges: true, default: null }),
      /**
       * The granularity of the error that the server is willing to guarantee on the time synchronization. It is
       * of type GranularityEnum.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.2
       */
      granularity: Attribute(1, TlvEnum(), { default: 0 /* NoTimeGranularity */ }),
      /**
       * The server’s time source. This attribute indicates what method the server is using to sync, whether the
       * source uses NTS or not and whether the source is internal or external to the Fabric. This attribute may
       * be used by a client to determine its level of trust in the UTCTime. It is of type TimeSourceEnum.
       *
       * If a server is unsure if the selected NTP server is within the Fabric, it SHOULD indicate the server is
       * NonFabric.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.3
       */
      timeSource: OptionalAttribute(2, TlvEnum(), { default: 0 /* None */ }),
      /**
       * The Node ID of a trusted Time Cluster. The TrustedTimeNodeId Node is used as a check on external time
       * sync sources and may be used as the primary time source if other time sources are unavailable. See
       * Section 11.16.13, “Time source prioritization”. This attribute is writeable only by an Administrator. It
       * SHOULD be set by the Commissioner during commissioning. If no appropriate TrustedTimeNodeId is
       * available, the commissioner may set this value to null.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.5
       */
      trustedTimeNodeId: WritableAttribute(
        3,
        TlvNullable(TlvNodeId),
        { default: null, writeAcl: AccessLevel.Administer }
      )
    },
    commands: {
      /**
       * This command may be issued by Administrator to set the time. If the Commissioner does not have a valid
       * time source, it may send a Granularity of NoTimeGranularity.
       *
       * Upon receipt of this command, the server may update its UTCTime attribute to match the time specified in
       * the command, if the stated Granularity and TimeSource are acceptable. The server shall update its
       * UTCTime attribute if its current Granularity is NoTimeGranularity.
       *
       * If the time is updated, the server shall also update its Granularity attribute as appropriate
       *
       * server does not plan to maintain time). It shall also update its TimeSource attribute to Admin. It shall
       * also update its last known good UTC time.
       *
       * If the server updates its UTCTime attribute, it shall accept the command with a status code of SUCCESS.
       * If it opts to not update its time, it shall fail the command with a cluster specific Status Code of
       * TimeNotAccepted.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.9.1
       */
      setUtcTime: Command(0, TimeSync2.TlvSetUtcTimeRequest, 0, TlvNoResponse, { invokeAcl: AccessLevel.Administer })
    }
  });
  TimeSync2.NtpClientComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The default NTP server the server’s Node may use if other time sources are unavailable. This attribute
       * may contain a domain name or a static IPv6 address in text format as specified in RFC 5952
       * [https://tools.ietf.org/html/rfc5952]. See Section 11.16.13, “Time source prioritization”. This
       * attribute is writeable only by an Administrator. It SHOULD be set by the Commissioner during
       * commissioning. If no default NTP is available, the Commissioner may set this value to null.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.4
       */
      defaultNtp: WritableAttribute(
        4,
        TlvNullable(TlvString.bound({ maxLength: 128 })),
        { default: null, writeAcl: AccessLevel.Administer }
      )
    }
  });
  TimeSync2.TimeZoneComponent = ClusterFactory.Component({
    attributes: {
      /**
       * A list of time zone offsets from UTC and when they shall take effect. This attribute uses a list of time
       * offset configurations to allow Nodes to handle scheduled regulatory time zone changes. This attribute
       * shall NOT be used to indicate daylight savings time changes (see Section 11.16.8.7, “DSTOffset
       * Attribute” for daylight savings time).
       *
       * The first entry shall have a ValidAt entry of 0. If there is a second entry, it shall have a non-zero
       * ValidAt time.
       *
       * If a server supports a TimeZoneDatabase, the server may update its own DSTOffset list (Section
       * 11.16.8.7, “DSTOffset Attribute”) to add new DST change times as required, based on the Name fields of
       * the TimeZoneStruct. Administrators may add additional entries to the DSTOffset of other Nodes with the
       * same time zone, if required.
       *
       * If a server does not support a TimeZoneDatabase, the Name field of the TimeZoneStruct is only applicable
       * for client-side localization. In particular:
       *
       *   • If the server does not support a TimeZoneDatabase, the Name field shall NOT be used to calculate the
       *     local time.
       *
       *   • If the server does not support a TimeZoneDatabase, the Name field shall NOT be used to calculate DST
       *     start or end dates.
       *
       * Upon writing this attribute, the server shall recompute its LocalTime, taking into account the Offset of
       * the currently used TimeZoneStruct.
       *
       * When time passes, the server SHOULD remove any entries which are no longer active and change the ValidAt
       * time for the currently used TimeZoneStruct list item to zero.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.6
       */
      timeZone: WritableAttribute(
        5,
        TlvArray(TimeSync2.TlvTimeZoneStruct, { minLength: 1, maxLength: 2 }),
        { writeAcl: AccessLevel.Manage }
      ),
      /**
       * A list of offsets to apply for daylight savings time, and their validity period. List entries shall be
       * sorted by ValidStarting time.
       *
       * A list entry shall NOT have a ValidStarting time that is smaller than the ValidUntil time of the
       * previous entry.
       *
       * Upon writing this attribute, the server shall recompute its LocalTime.
       *
       * This list may hold up to 20 entries. If a server does not have sufficient storage for 20 entries, it may
       * truncate the list by removing entries with the largest ValidStarting times. The server shall reserve
       * sufficient storage for at least one entry.
       *
       * Over time, the server SHOULD remove any entries which are no longer active from the list.
       *
       * Over time, if the server supports a TimeZoneDatabase, it may update its own list to add additional
       * entries.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.7
       */
      dstOffset: WritableAttribute(
        6,
        TlvArray(TimeSync2.TlvDSTOffsetStruct, { maxLength: 20 }),
        { default: [], writeAcl: AccessLevel.Manage }
      ),
      /**
       * The computed current local time of the server as a epoch-us (Epoch Time in Microseconds). The local time
       * offset of the value is the sum of the currently used TimeZoneEntry’s offset and the currently used DST
       * offset, if any.
       *
       * If the server has not achieved time synchronization, this shall be null.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.8
       */
      localTime: Attribute(7, TlvNullable(TlvEpochUs), { omitChanges: true, default: 0 }),
      /**
       * Indicates whether the server has access to a time zone database. Nodes with a time zone database may
       * update their own DSTOffset attribute to add new entries and may push DSTOffset updates to other Nodes in
       * the same time zone as required.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.9
       */
      timeZoneDatabase: FixedAttribute(8, TlvBoolean, { default: true })
    },
    events: {
      /**
       * This event shall be generated when the server stops applying the current DSTOffset and there are no
       * entries in the list with a larger ValidStarting time, indicating the need to possibly get new DST data.
       *
       * There is no data for this event.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.10.1
       */
      dstTableEmpty: Event(0, EventPriority.Info, TlvNoArguments),
      /**
       * This event shall be generated when the server starts or stops applying a DST offset.
       *
       * This event contains a boolean predicate that indicates whether the server is applying the DST offset.
       * When the value is "true", the current DST offset is being applied (i.e, daylight savings time is
       * applied, as opposed to standard time).
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.10.2
       */
      dstStatus: Event(1, EventPriority.Info, TlvNoArguments),
      /**
       * This event shall be generated when the server changes its time zone offset or name. It shall NOT be sent
       * for DST changes that are not accompanied by a time zone change.
       *
       * This event returns a structure as follows:
       *
       * Current time zone offset from UTC in seconds.
       *
       * Current time zone name. This name SHOULD use the country/city format specified by the IANA time zone
       * database [https://www.iana.org/time-zones].
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.10.3
       */
      timeZoneStatus: Event(2, EventPriority.Info, TimeSync2.TlvTimeZoneStatusEvent)
    }
  });
  TimeSync2.NtpServerComponent = ClusterFactory.Component({
    attributes: {
      /**
       * If the server is running an NTP server, this value shall be the port number for the service. If the
       * server is not currently running an NTP server, this value shall be null.
       *
       * This attribute shall be present if this server is capable of providing an NTP server instance. See
       * Section 11.16.15, “Acting as an NTP Server” for more information.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.16.8.10
       */
      ntpServerPort: Attribute(9, TlvNullable(TlvUInt16), { default: null })
    }
  });
  TimeSync2.Cluster = ClusterFactory.Extensible(
    TimeSync2.Base,
    /**
     * Use this factory method to create a TimeSync cluster with support for optional features. Include each
     * {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns a TimeSync cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...TimeSync2.Base,
        supportedFeatures: BitFlags(TimeSync2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, TimeSync2.NtpClientComponent, { ntpClient: true });
      ClusterFactory.extend(cluster, TimeSync2.TimeZoneComponent, { timeZone: true });
      ClusterFactory.extend(cluster, TimeSync2.NtpServerComponent, { ntpServer: true });
      return cluster;
    }
  );
  const NTPC = { ntpClient: true };
  const TZ = { timeZone: true };
  const NTPS = { ntpServer: true };
  TimeSync2.Complete = ClusterFactory.Definition({
    id: TimeSync2.Cluster.id,
    name: TimeSync2.Cluster.name,
    revision: TimeSync2.Cluster.revision,
    features: TimeSync2.Cluster.features,
    attributes: {
      ...TimeSync2.Cluster.attributes,
      defaultNtp: ClusterFactory.AsConditional(TimeSync2.NtpClientComponent.attributes.defaultNtp, { mandatoryIf: [NTPC] }),
      timeZone: ClusterFactory.AsConditional(TimeSync2.TimeZoneComponent.attributes.timeZone, { mandatoryIf: [TZ] }),
      dstOffset: ClusterFactory.AsConditional(TimeSync2.TimeZoneComponent.attributes.dstOffset, { mandatoryIf: [TZ] }),
      localTime: ClusterFactory.AsConditional(TimeSync2.TimeZoneComponent.attributes.localTime, { mandatoryIf: [TZ] }),
      timeZoneDatabase: ClusterFactory.AsConditional(
        TimeSync2.TimeZoneComponent.attributes.timeZoneDatabase,
        { mandatoryIf: [TZ] }
      ),
      ntpServerPort: ClusterFactory.AsConditional(
        TimeSync2.NtpServerComponent.attributes.ntpServerPort,
        { mandatoryIf: [NTPS] }
      )
    },
    commands: TimeSync2.Cluster.commands,
    events: {
      dstTableEmpty: ClusterFactory.AsConditional(TimeSync2.TimeZoneComponent.events.dstTableEmpty, { mandatoryIf: [TZ] }),
      dstStatus: ClusterFactory.AsConditional(TimeSync2.TimeZoneComponent.events.dstStatus, { mandatoryIf: [TZ] }),
      timeZoneStatus: ClusterFactory.AsConditional(
        TimeSync2.TimeZoneComponent.events.timeZoneStatus,
        { mandatoryIf: [TZ] }
      )
    }
  });
})(TimeSync || (TimeSync = {}));
const TimeSyncCluster = TimeSync.Cluster;
export {
  TimeSync,
  TimeSyncCluster
};
