// @project-chip/matter.js/dist/esm/cluster/definitions/UnitLocalizationCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { BitFlag, BitFlags } from "../../schema/BitmapSchema.js";
import { WritableAttribute, AccessLevel } from "../../cluster/Cluster.js";
import { TlvEnum } from "../../tlv/TlvNumber.js";
import { TlvNullable } from "../../tlv/TlvNullable.js";
var UnitLocalization;
((UnitLocalization2) => {
  let TempUnit;
  ((TempUnit2) => {
    TempUnit2[TempUnit2["Fahrenheit"] = 0] = "Fahrenheit";
    TempUnit2[TempUnit2["Celsius"] = 1] = "Celsius";
    TempUnit2[TempUnit2["Kelvin"] = 2] = "Kelvin";
  })(TempUnit = UnitLocalization2.TempUnit || (UnitLocalization2.TempUnit = {}));
  let Feature;
  ((Feature2) => {
    Feature2["TemperatureUnit"] = "TemperatureUnit";
  })(Feature = UnitLocalization2.Feature || (UnitLocalization2.Feature = {}));
  UnitLocalization2.Base = ClusterFactory.Definition({
    id: 45,
    name: "UnitLocalization",
    revision: 1,
    features: {
      /**
       * TemperatureUnit
       *
       * The Node can be configured to use different units of temperature when conveying values to a user.
       */
      temperatureUnit: BitFlag(0)
    }
  });
  UnitLocalization2.TemperatureUnitComponent = ClusterFactory.Component({
    attributes: {
      /**
       * The TemperatureUnit attribute shall indicate the unit for the Node to use only when conveying
       * temperature in communication to the user. If provided, this value shall take priority over any unit
       * implied through the ActiveLocale Attribute.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 11.5.6.1
       */
      temperatureUnit: WritableAttribute(
        0,
        TlvNullable(TlvEnum()),
        { persistent: true, default: null, writeAcl: AccessLevel.Manage }
      )
    }
  });
  UnitLocalization2.Cluster = ClusterFactory.Extensible(
    UnitLocalization2.Base,
    /**
     * Use this factory method to create an UnitLocalization cluster with support for optional features. Include
     * each {@link Feature} you wish to support.
     *
     * @param features the optional features to support
     * @returns an UnitLocalization cluster with specified features enabled
     * @throws {IllegalClusterError} if the feature combination is disallowed by the Matter specification
     */
    (...features) => {
      ClusterFactory.validateFeatureSelection(features, Feature);
      const cluster = ClusterFactory.Definition({
        ...UnitLocalization2.Base,
        supportedFeatures: BitFlags(UnitLocalization2.Base.features, ...features)
      });
      ClusterFactory.extend(cluster, UnitLocalization2.TemperatureUnitComponent, { temperatureUnit: true });
      return cluster;
    }
  );
  const TEMP = { temperatureUnit: true };
  UnitLocalization2.Complete = ClusterFactory.Definition({
    id: UnitLocalization2.Cluster.id,
    name: UnitLocalization2.Cluster.name,
    revision: UnitLocalization2.Cluster.revision,
    features: UnitLocalization2.Cluster.features,
    attributes: {
      temperatureUnit: ClusterFactory.AsConditional(
        UnitLocalization2.TemperatureUnitComponent.attributes.temperatureUnit,
        { mandatoryIf: [TEMP] }
      )
    }
  });
})(UnitLocalization || (UnitLocalization = {}));
const UnitLocalizationCluster = UnitLocalization.Cluster;
export {
  UnitLocalization,
  UnitLocalizationCluster
};
