// @project-chip/matter.js/dist/esm/cluster/definitions/UserLabelCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { WritableAttribute, AccessLevel } from "../../cluster/Cluster.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { Label } from "../../cluster/definitions/LabelCluster.js";
var UserLabel;
((UserLabel2) => {
  UserLabel2.Cluster = ClusterFactory.Definition({
    id: 65,
    name: "UserLabel",
    revision: 1,
    attributes: {
      /**
       * An implementation shall support at least 4 list entries per node for all User Label cluster instances on
       * the node.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.9.4.1
       */
      labelList: WritableAttribute(
        0,
        TlvArray(Label.TlvLabelStruct, { minLength: 0 }),
        { persistent: true, default: [], writeAcl: AccessLevel.Manage }
      )
    }
  });
})(UserLabel || (UserLabel = {}));
const UserLabelCluster = UserLabel.Cluster;
export {
  UserLabel,
  UserLabelCluster
};
