// @project-chip/matter.js/dist/esm/cluster/definitions/ValidProxiesCluster.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterFactory } from "../../cluster/ClusterFactory.js";
import { WritableFixedAttribute, Command } from "../../cluster/Cluster.js";
import { TlvArray } from "../../tlv/TlvArray.js";
import { TlvObject, TlvField } from "../../tlv/TlvObject.js";
import { TlvNodeId } from "../../datatype/NodeId.js";
import { TlvNoArguments } from "../../tlv/TlvNoArguments.js";
var ValidProxies;
((ValidProxies2) => {
  ValidProxies2.TlvValidProxyStruct = TlvObject({ nodeId: TlvField(1, TlvNodeId) });
  ValidProxies2.Cluster = ClusterFactory.Definition({
    id: 68,
    name: "ValidProxies",
    revision: 1,
    attributes: {
      /**
       * List of valid proxies that can proxy this Node. Each entry in this list is fabric-scoped.
       *
       * @see {@link MatterCoreSpecificationV1_1} § 9.15.15.5.1
       */
      validProxyList: WritableFixedAttribute(
        0,
        TlvArray(ValidProxies2.TlvValidProxyStruct),
        { persistent: true, default: [] }
      )
    },
    commands: {
      /**
       * @see {@link MatterCoreSpecificationV1_1} § 9.15.15.6
       */
      getValidProxiesRequest: Command(0, TlvNoArguments, 1, TlvNoArguments)
    }
  });
})(ValidProxies || (ValidProxies = {}));
const ValidProxiesCluster = ValidProxies.Cluster;
export {
  ValidProxies,
  ValidProxiesCluster
};
