// @project-chip/matter.js/dist/esm/CommissioningController.js
/**
 * @license
 * Copyright 2022 The matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterController } from "./MatterController.js";
import { MatterNode } from "./MatterNode.js";
import { GlobalAttributes } from "./cluster/Cluster.js";
import { SupportedAttributeClient } from "./cluster/client/AttributeClient.js";
import { BasicInformation } from "./cluster/definitions/BasicInformationCluster.js";
import { ImplementationError } from "./common/MatterError.js";
import { EndpointNumber } from "./datatype/EndpointNumber.js";
import { PairedNode } from "./device/PairedNode.js";
import { Logger } from "./log/Logger.js";
import { UdpInterface } from "./net/UdpInterface.js";
import { ControllerDiscovery } from "./protocol/ControllerDiscovery.js";
const logger = new Logger("CommissioningController");
class CommissioningController extends MatterNode {
  /**
   * Creates a new CommissioningController instance
   *
   * @param options The options for the CommissioningController
   */
  constructor(options) {
    super();
    this.options = options;
    this.started = false;
    this.connectedNodes = /* @__PURE__ */ new Map();
    this.sessionDisconnectedHandler = /* @__PURE__ */ new Map();
  }
  get nodeId() {
    return this.controllerInstance?.nodeId;
  }
  assertIsAddedToMatterServer() {
    if (this.mdnsScanner === void 0 || this.storage === void 0) {
      throw new ImplementationError("Add the node to the Matter instance before.");
    }
    if (!this.started) {
      throw new ImplementationError("The node needs to be started before interacting with the controller.");
    }
    return { mdnsScanner: this.mdnsScanner, storage: this.storage };
  }
  assertControllerIsStarted(errorText) {
    if (this.controllerInstance === void 0) {
      throw new ImplementationError(
        errorText ?? "Controller instance not yet started. Please call start() first."
      );
    }
    return this.controllerInstance;
  }
  /** Internal method to initialize a MatterController instance. */
  async initializeController() {
    const { mdnsScanner, storage } = this.assertIsAddedToMatterServer();
    if (this.controllerInstance !== void 0) {
      return this.controllerInstance;
    }
    const { localPort, adminFabricId, adminVendorId, adminFabricIndex, caseAuthenticatedTags } = this.options;
    return await MatterController.create(
      mdnsScanner,
      this.ipv4Disabled ? void 0 : await UdpInterface.create("udp4", localPort, this.listeningAddressIpv4),
      await UdpInterface.create("udp6", localPort, this.listeningAddressIpv6),
      storage,
      (peerNodeId) => {
        logger.info(`Session for peer node ${peerNodeId} disconnected ...`);
        const handler = this.sessionDisconnectedHandler.get(peerNodeId);
        if (handler !== void 0) {
          handler().catch((error) => logger.warn(`Error while handling session disconnect: ${error}`));
        }
      },
      adminVendorId,
      adminFabricId,
      adminFabricIndex,
      caseAuthenticatedTags
    );
  }
  /**
   * Commissions/Pairs a new device into the controller fabric. The method returns a PairedNode instance of the
   * paired node on success.
   */
  async commissionNode(nodeOptions) {
    this.assertIsAddedToMatterServer();
    const controller = this.assertControllerIsStarted();
    const nodeId = await controller.commission(nodeOptions);
    return this.connectNode(nodeId, {
      ...nodeOptions,
      autoSubscribe: nodeOptions.autoSubscribe ?? this.options.autoSubscribe,
      subscribeMinIntervalFloorSeconds: nodeOptions.subscribeMinIntervalFloorSeconds ?? this.options.subscribeMinIntervalFloorSeconds,
      subscribeMaxIntervalCeilingSeconds: nodeOptions.subscribeMaxIntervalCeilingSeconds ?? this.options.subscribeMaxIntervalCeilingSeconds
    });
  }
  /** Check if a given node id is commissioned on this controller. */
  isNodeCommissioned(nodeId) {
    const controller = this.assertControllerIsStarted();
    return controller.getCommissionedNodes().includes(nodeId) ?? false;
  }
  /**
   * Remove a Node id from the controller. This method should only be used if the decommission method on the
   * PairedNode instance returns an error. By default it tries to decommission the node from the controller but will
   * remove it also in case of an error during decommissioning. Ideally try to decommission the node before and only
   * use this in case of an error.
   */
  async removeNode(nodeId, tryDecommissioning = true) {
    const controller = this.assertControllerIsStarted();
    const node = this.connectedNodes.get(nodeId);
    if (tryDecommissioning) {
      try {
        if (node == void 0) {
          throw new ImplementationError(`Node ${nodeId} is not connected.`);
        }
        await node.decommission();
      } catch (error) {
        logger.warn(`Decommissioning node ${nodeId} failed with error, remove node anyway: ${error}`);
      }
    }
    if (node !== void 0) {
      node.close();
    }
    await controller.removeNode(nodeId);
    this.connectedNodes.delete(nodeId);
  }
  async disconnectNode(nodeId) {
    const node = this.connectedNodes.get(nodeId);
    if (node === void 0) {
      throw new ImplementationError(`Node ${nodeId} is not connected!`);
    }
    await this.controllerInstance?.disconnect(nodeId);
  }
  /**
   * Connect to an already paired Node.
   * After connection the endpoint data of the device is analyzed and an object structure is created.
   */
  async connectNode(nodeId, connectOptions) {
    const controller = this.assertControllerIsStarted();
    if (!controller.getCommissionedNodes().includes(nodeId)) {
      throw new ImplementationError(`Node ${nodeId} is not commissioned!`);
    }
    const existingNode = this.connectedNodes.get(nodeId);
    if (existingNode !== void 0) {
      if (!existingNode.isConnected) {
        await existingNode.reconnect();
      }
      return existingNode;
    }
    const pairedNode = await PairedNode.create(
      nodeId,
      this,
      connectOptions,
      async () => this.createInteractionClient(nodeId),
      (handler) => this.sessionDisconnectedHandler.set(nodeId, handler)
    );
    this.connectedNodes.set(nodeId, pairedNode);
    if (connectOptions?.autoSubscribe !== false) {
      await this.enhanceDeviceDetailsFromCache(nodeId, pairedNode);
    } else {
      await this.enhanceDeviceDetailsFromRemote(nodeId, pairedNode);
    }
    return pairedNode;
  }
  async enhanceDeviceDetailsFromCache(nodeId, pairedNode) {
    const controller = this.assertControllerIsStarted();
    const globalAttributeKeys = Object.keys(GlobalAttributes({}));
    const basicInformationClient = pairedNode.getRootClusterClient(BasicInformation.Cluster);
    if (basicInformationClient === void 0) {
      logger.info(`No basic information cluster found for node ${nodeId}`);
      return;
    }
    const basicInformationData = {};
    for (const attributeName of Object.keys(basicInformationClient.attributes)) {
      if (globalAttributeKeys.includes(attributeName)) {
        continue;
      }
      const attribute = basicInformationClient.attributes[attributeName];
      if (attribute instanceof SupportedAttributeClient) {
        try {
          basicInformationData[attributeName] = await attribute.get();
        } catch (error) {
          logger.info(`Error while getting attribute ${attributeName} for node ${nodeId}: ${error}`);
        }
      }
    }
    controller.enhanceCommissionedNodeDetails(nodeId, { basicInformationData });
  }
  async enhanceDeviceDetailsFromRemote(nodeId, pairedNode) {
    const controller = this.assertControllerIsStarted();
    const globalAttributeKeys = Object.keys(GlobalAttributes({}));
    try {
      const interactionClient = await pairedNode.getInteractionClient();
      const basicInformationAttributes = await interactionClient.getMultipleAttributes({
        attributes: [{ endpointId: EndpointNumber(0), clusterId: BasicInformation.Cluster.id }]
      });
      const basicInformationData = {};
      for (const {
        path: { attributeName },
        value
      } of basicInformationAttributes) {
        if (!globalAttributeKeys.includes(attributeName)) {
          basicInformationData[attributeName] = value;
        }
      }
      controller.enhanceCommissionedNodeDetails(nodeId, { basicInformationData });
    } catch (error) {
      logger.info(`Error while enhancing basic information for node ${nodeId}: ${error}`);
    }
  }
  /**
   * Connects to all paired nodes.
   * After connection the endpoint data of the device is analyzed and an object structure is created.
   */
  async connect(connectOptions) {
    const controller = this.assertControllerIsStarted();
    if (!controller.isCommissioned()) {
      throw new ImplementationError(
        "Controller instance not yet paired with any device, so nothing to connect to."
      );
    }
    for (const nodeId of controller.getCommissionedNodes()) {
      await this.connectNode(nodeId, connectOptions);
    }
    return Array.from(this.connectedNodes.values());
  }
  /**
   * Set the MDNS Scanner instance. Should be only used internally
   *
   * @param mdnsScanner MdnsScanner instance
   */
  setMdnsScanner(mdnsScanner) {
    this.mdnsScanner = mdnsScanner;
  }
  /**
   * Set the MDNS Broadcaster instance. Should be only used internally
   *
   * @param _mdnsBroadcaster MdnsBroadcaster instance
   */
  setMdnsBroadcaster(_mdnsBroadcaster) {
  }
  /**
   * Set the Storage instance. Should be only used internally
   *
   * @param storage storage context to use
   */
  setStorage(storage) {
    this.storage = storage;
  }
  /** Returns true if t least one node is commissioned/paired with this controller instance. */
  isCommissioned() {
    const controller = this.assertControllerIsStarted();
    return controller.isCommissioned();
  }
  /**
   * Creates and Return a new InteractionClient to communicate with a node. This is mainly used internally and should
   * not be used directly. See the PairedNode class for the public API.
   */
  async createInteractionClient(nodeId) {
    const controller = this.assertControllerIsStarted();
    return controller.connect(nodeId);
  }
  /** Returns the PairedNode instance for a given node id, if this node is connected. */
  getConnectedNode(nodeId) {
    return this.connectedNodes.get(nodeId);
  }
  /** Returns an array with the Node Ids for all commissioned nodes. */
  getCommissionedNodes() {
    const controller = this.assertControllerIsStarted();
    return controller.getCommissionedNodes() ?? [];
  }
  getCommissionedNodesDetails() {
    const controller = this.assertControllerIsStarted();
    return controller.getCommissionedNodesDetails() ?? [];
  }
  /** Disconnects all connected nodes and Closes the network connections and other resources of the controller. */
  async close() {
    for (const node of this.connectedNodes.values()) {
      node.close();
    }
    await this.controllerInstance?.close();
    this.controllerInstance = void 0;
    this.connectedNodes.clear();
    this.ipv4Disabled = void 0;
    this.started = false;
  }
  getPort() {
    return this.options.localPort;
  }
  initialize(ipv4Disabled) {
    if (this.started) {
      throw new ImplementationError("Controller instance already started.");
    }
    if (this.ipv4Disabled !== void 0 && this.ipv4Disabled !== ipv4Disabled) {
      throw new ImplementationError(
        "Changing the IPv4 disabled flag after starting the controller is not supported."
      );
    }
    this.ipv4Disabled = ipv4Disabled;
  }
  /** Initialize the controller and connect to all commissioned nodes if autoConnect is not set to false. */
  async start() {
    if (this.ipv4Disabled === void 0) {
      throw new ImplementationError("Initialization not done. Add the controller to the MatterServer first.");
    }
    this.started = true;
    if (this.controllerInstance === void 0) {
      this.controllerInstance = await this.initializeController();
    }
    if (this.options.autoConnect !== false && this.controllerInstance.isCommissioned()) {
      await this.connect();
    }
  }
  cancelCommissionableDeviceDiscovery(identifierData, discoveryCapabilities) {
    this.assertIsAddedToMatterServer();
    const controller = this.assertControllerIsStarted();
    controller.collectScanners(discoveryCapabilities).forEach((scanner) => ControllerDiscovery.cancelCommissionableDeviceDiscovery(scanner, identifierData));
  }
  async discoverCommissionableDevices(identifierData, discoveryCapabilities, discoveredCallback, timeoutSeconds = 900) {
    this.assertIsAddedToMatterServer();
    const controller = this.assertControllerIsStarted();
    return await ControllerDiscovery.discoverCommissionableDevices(
      controller.collectScanners(discoveryCapabilities),
      timeoutSeconds,
      identifierData,
      discoveredCallback
    );
  }
  resetStorage() {
    this.assertControllerIsStarted(
      "Storage can not be reset while the controller is operating! Please close the controller first."
    );
    const { storage } = this.assertIsAddedToMatterServer();
    storage.clearAll();
  }
  /** Returns active session information for all connected nodes. */
  getActiveSessionInformation() {
    return this.controllerInstance?.getActiveSessionInformation() ?? [];
  }
}
export {
  CommissioningController
};
