// @project-chip/matter.js/dist/esm/MatterController.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Ble } from "./ble/Ble.js";
import { RootCertificateManager } from "./certificate/RootCertificateManager.js";
import { GeneralCommissioning } from "./cluster/definitions/GeneralCommissioningCluster.js";
import { NoProviderError } from "./common/MatterError.js";
import { serverAddressToString } from "./common/ServerAddress.js";
import { tryCatchAsync } from "./common/TryCatchHandler.js";
import { CRYPTO_SYMMETRIC_KEY_LENGTH, Crypto } from "./crypto/Crypto.js";
import { FabricId } from "./datatype/FabricId.js";
import { FabricIndex } from "./datatype/FabricIndex.js";
import { NodeId } from "./datatype/NodeId.js";
import { VendorId } from "./datatype/VendorId.js";
import { Fabric, FabricBuilder } from "./fabric/Fabric.js";
import { Logger } from "./log/Logger.js";
import { MdnsScanner } from "./mdns/MdnsScanner.js";
import { ChannelManager, NoChannelError } from "./protocol/ChannelManager.js";
import { ControllerCommissioner } from "./protocol/ControllerCommissioner.js";
import { ControllerDiscovery, DiscoveryError } from "./protocol/ControllerDiscovery.js";
import { ExchangeManager, ExchangeProvider, MessageChannel } from "./protocol/ExchangeManager.js";
import { RetransmissionLimitReachedError } from "./protocol/MessageExchange.js";
import { InteractionClient } from "./protocol/interaction/InteractionClient.js";
import { SECURE_CHANNEL_PROTOCOL_ID } from "./protocol/securechannel/SecureChannelMessages.js";
import { StatusReportOnlySecureChannelProtocol } from "./protocol/securechannel/SecureChannelProtocol.js";
import { SessionManager } from "./session/SessionManager.js";
import { CaseClient } from "./session/case/CaseClient.js";
import { PaseClient } from "./session/pase/PaseClient.js";
import { Time } from "./time/Time.js";
import { TlvEnum } from "./tlv/TlvNumber.js";
import { TlvField, TlvObject } from "./tlv/TlvObject.js";
import { TlvString } from "./tlv/TlvString.js";
import { isIPv6 } from "./util/Ip.js";
import { anyPromise, createPromise } from "./util/Promises.js";
const TlvCommissioningSuccessFailureResponse = TlvObject({
  /** Contain the result of the operation. */
  errorCode: TlvField(0, TlvEnum()),
  /** Should help developers in troubleshooting errors. The value MAY go into logs or crash reports, not User UIs. */
  debugText: TlvField(1, TlvString.bound({ maxLength: 128 }))
});
const DEFAULT_FABRIC_INDEX = FabricIndex(1);
const DEFAULT_FABRIC_ID = FabricId(1);
const DEFAULT_ADMIN_VENDOR_ID = VendorId(65521);
const RECONNECTION_POLLING_INTERVAL = 10 * 60 * 1e3;
const logger = Logger.get("MatterController");
class PairRetransmissionLimitReachedError extends RetransmissionLimitReachedError {
}
class MatterController {
  constructor(mdnsScanner, netInterfaceIpv4, netInterfaceIpv6, certificateManager, fabric, storage, adminVendorId, sessionClosedCallback) {
    this.mdnsScanner = mdnsScanner;
    this.netInterfaceIpv4 = netInterfaceIpv4;
    this.netInterfaceIpv6 = netInterfaceIpv6;
    this.certificateManager = certificateManager;
    this.fabric = fabric;
    this.storage = storage;
    this.adminVendorId = adminVendorId;
    this.sessionClosedCallback = sessionClosedCallback;
    this.channelManager = new ChannelManager();
    this.paseClient = new PaseClient();
    this.caseClient = new CaseClient();
    this.commissionedNodes = /* @__PURE__ */ new Map();
    this.controllerStorage = this.storage.createContext("MatterController");
    if (this.controllerStorage.has("commissionedNodes")) {
      const commissionedNodes = this.controllerStorage.get("commissionedNodes");
      this.commissionedNodes.clear();
      for (const [nodeId, details] of commissionedNodes) {
        this.commissionedNodes.set(nodeId, details);
      }
    } else if (this.controllerStorage.has("operationalServerAddress")) {
      const operationalServerAddress = this.controllerStorage.get("operationalServerAddress");
      this.setOperationalDeviceData(this.fabric.nodeId, operationalServerAddress);
      this.controllerStorage.delete("operationalServerAddress");
    } else if (this.controllerStorage.has("fabricCommissioned")) {
      this.commissionedNodes.set(this.fabric.nodeId, {});
      this.controllerStorage.set("commissionedNodes", Array.from(this.commissionedNodes.entries()));
      this.controllerStorage.delete("fabricCommissioned");
    }
    this.sessionManager = new SessionManager(this, this.storage);
    this.sessionManager.initFromStorage([this.fabric]);
    this.exchangeManager = new ExchangeManager(this.sessionManager, this.channelManager);
    this.exchangeManager.addProtocolHandler(new StatusReportOnlySecureChannelProtocol());
    if (netInterfaceIpv4 !== void 0) {
      this.addTransportInterface(netInterfaceIpv4);
    }
    this.addTransportInterface(netInterfaceIpv6);
  }
  static async create(scanner, netInterfaceIpv4, netInterfaceIpv6, storage, sessionClosedCallback, adminVendorId = VendorId(DEFAULT_ADMIN_VENDOR_ID), adminFabricId = FabricId(DEFAULT_FABRIC_ID), adminFabricIndex = FabricIndex(DEFAULT_FABRIC_INDEX), caseAuthenticatedTags) {
    const certificateManager = new RootCertificateManager(storage);
    const controllerStorage = storage.createContext("MatterController");
    if (controllerStorage.has("fabric")) {
      const storedFabric = Fabric.createFromStorageObject(controllerStorage.get("fabric"));
      return new MatterController(
        scanner,
        netInterfaceIpv4,
        netInterfaceIpv6,
        certificateManager,
        storedFabric,
        storage,
        storedFabric.rootVendorId,
        sessionClosedCallback
      );
    } else {
      const rootNodeId = NodeId.getRandomOperationalNodeId();
      const ipkValue = Crypto.getRandomData(CRYPTO_SYMMETRIC_KEY_LENGTH);
      const fabricBuilder = new FabricBuilder().setRootCert(certificateManager.getRootCert()).setRootNodeId(rootNodeId).setIdentityProtectionKey(ipkValue).setRootVendorId(adminVendorId);
      fabricBuilder.setOperationalCert(
        certificateManager.generateNoc(
          fabricBuilder.getPublicKey(),
          adminFabricId,
          rootNodeId,
          caseAuthenticatedTags
        )
      );
      return new MatterController(
        scanner,
        netInterfaceIpv4,
        netInterfaceIpv6,
        certificateManager,
        await fabricBuilder.build(adminFabricIndex),
        storage,
        adminVendorId,
        sessionClosedCallback
      );
    }
  }
  get nodeId() {
    return this.fabric.rootNodeId;
  }
  addTransportInterface(netInterface) {
    this.exchangeManager.addTransportInterface(netInterface);
  }
  collectScanners(discoveryCapabilities = { onIpNetwork: true }) {
    const scannersToUse = new Array();
    scannersToUse.push(this.mdnsScanner);
    if (discoveryCapabilities.ble) {
      if (this.bleScanner === void 0) {
        let ble;
        try {
          ble = Ble.get();
          this.netInterfaceBle = ble.getBleCentralInterface();
          this.addTransportInterface(this.netInterfaceBle);
          this.bleScanner = ble.getBleScanner();
          scannersToUse.push(this.bleScanner);
        } catch (error) {
          if (error instanceof NoProviderError) {
            logger.warn(
              "BLE is not supported on this platform. The device to commission might not be found!"
            );
          } else {
            throw error;
          }
        }
      }
    }
    if (discoveryCapabilities.softAccessPoint) {
      logger.info("SoftAP is not supported yet");
    }
    return scannersToUse;
  }
  /**
   * Commission a device by its identifier and the Passcode. If a known address is provided this is tried first
   * before discovering devices in the network. If multiple addresses or devices are found, they are tried all after
   * each other. It returns the NodeId of the commissioned device.
   * If it throws an PairRetransmissionLimitReachedError that means that no found device responded to the pairing
   * request or the passode did not match to any discovered device/address.
   */
  async commission(options) {
    const {
      commissioning: commissioningOptions = {
        regulatoryLocation: GeneralCommissioning.RegulatoryLocationType.Outdoor,
        // Set to the most restrictive if relevant
        regulatoryCountryCode: "XX"
      },
      discovery: { timeoutSeconds = 30 },
      passcode
    } = options;
    const commissionableDevice = "commissionableDevice" in options.discovery ? options.discovery.commissionableDevice : void 0;
    let {
      discovery: { discoveryCapabilities = {}, knownAddress }
    } = options;
    let identifierData = "identifierData" in options.discovery ? options.discovery.identifierData : {};
    discoveryCapabilities.onIpNetwork = true;
    if (commissionableDevice !== void 0) {
      let { addresses } = commissionableDevice;
      if (discoveryCapabilities.ble === true) {
        discoveryCapabilities = { onIpNetwork: true, ble: addresses.some((address) => address.type === "ble") };
      } else {
        addresses = addresses.filter((address) => address.type !== "ble");
      }
      addresses.sort((a) => a.type === "udp" ? -1 : 1);
      knownAddress = addresses[0];
      if ("instanceId" in commissionableDevice && commissionableDevice.instanceId !== void 0) {
        identifierData = { instanceId: commissionableDevice.instanceId };
      } else {
        identifierData = { longDiscriminator: commissionableDevice.D };
      }
    }
    const scannersToUse = this.collectScanners(discoveryCapabilities);
    logger.info(
      `Commissioning device with identifier ${Logger.toJSON(identifierData)} and ${scannersToUse.length} scanners and knownAddress ${JSON.stringify(knownAddress)}`
    );
    let paseSecureChannel;
    let discoveryData;
    if (knownAddress !== void 0) {
      try {
        paseSecureChannel = await this.initializePaseSecureChannel(knownAddress, passcode);
      } catch (error) {
        if (!(error instanceof RetransmissionLimitReachedError)) {
          throw error;
        }
      }
    }
    if (paseSecureChannel === void 0) {
      const discoveredDevices = await ControllerDiscovery.discoverDeviceAddressesByIdentifier(
        scannersToUse,
        identifierData,
        timeoutSeconds
      );
      const { result } = await ControllerDiscovery.iterateServerAddresses(
        discoveredDevices,
        RetransmissionLimitReachedError,
        async () => scannersToUse.flatMap((scanner) => scanner.getDiscoveredCommissionableDevices(identifierData)),
        async (address, device) => {
          const channel = await this.initializePaseSecureChannel(address, passcode, device);
          discoveryData = device;
          return channel;
        }
      );
      paseSecureChannel = result;
    }
    return await this.commissionDevice(paseSecureChannel, commissioningOptions, discoveryData);
  }
  async disconnect(nodeId) {
    await this.sessionManager.removeAllSessionsForNode(nodeId, true);
    await this.channelManager.removeChannel(this.fabric, nodeId);
  }
  async removeNode(nodeId) {
    logger.info(`Removing commissioned node ${nodeId} from controller.`);
    await this.sessionManager.removeAllSessionsForNode(nodeId);
    this.sessionManager.removeResumptionRecord(nodeId);
    await this.channelManager.removeChannel(this.fabric, nodeId);
    this.commissionedNodes.delete(nodeId);
    this.storeCommissionedNodes();
  }
  /**
   * Method to start commission process with a PASE pairing.
   * If this not successful and throws an RetransmissionLimitReachedError the address is invalid or the passcode
   * is wrong.
   */
  async initializePaseSecureChannel(address, passcode, device) {
    let paseChannel;
    if (device !== void 0) {
      logger.info(`Commissioning device`, MdnsScanner.discoveryDataDiagnostics(device));
    }
    if (address.type === "udp") {
      const { ip } = address;
      const isIpv6Address = isIPv6(ip);
      const paseInterface = isIpv6Address ? this.netInterfaceIpv6 : this.netInterfaceIpv4;
      if (paseInterface === void 0) {
        throw new PairRetransmissionLimitReachedError(
          `IPv${isIpv6Address ? "6" : "4"} interface not initialized. Cannot use ${ip} for commissioning.`
        );
      }
      paseChannel = await paseInterface.openChannel(address);
    } else {
      if (this.netInterfaceBle === void 0) {
        throw new PairRetransmissionLimitReachedError(
          `BLE interface not initialized. Cannot use ${address.peripheralAddress} for commissioning.`
        );
      }
      paseChannel = await this.netInterfaceBle.openChannel(address);
    }
    const unsecureSession = this.sessionManager.createUnsecureSession({
      sessionParameters: {
        idleIntervalMs: device?.SII,
        activeIntervalMs: device?.SAI,
        activeThresholdMs: device?.SAT
      },
      isInitiator: true
    });
    const paseUnsecureMessageChannel = new MessageChannel(paseChannel, unsecureSession);
    const paseExchange = this.exchangeManager.initiateExchangeWithChannel(
      paseUnsecureMessageChannel,
      SECURE_CHANNEL_PROTOCOL_ID
    );
    const paseSecureSession = await tryCatchAsync(
      async () => await this.paseClient.pair(this, paseExchange, passcode),
      Error,
      async (error) => {
        await paseExchange.close();
        throw error;
      }
    );
    await unsecureSession.destroy();
    return new MessageChannel(paseChannel, paseSecureSession);
  }
  /**
   * Method to commission a device with a PASE secure channel. It returns the NodeId of the commissioned device on
   * success.
   */
  async commissionDevice(paseSecureMessageChannel, commissioningOptions, discoveryData) {
    const peerNodeId = commissioningOptions.nodeId ?? NodeId.getRandomOperationalNodeId();
    const commissioningManager = new ControllerCommissioner(
      // Use the created secure session to do the commissioning
      new InteractionClient(new ExchangeProvider(this.exchangeManager, paseSecureMessageChannel), peerNodeId),
      this.certificateManager,
      this.fabric,
      commissioningOptions,
      peerNodeId,
      this.adminVendorId,
      async () => {
        await paseSecureMessageChannel.close();
        return await this.connect(peerNodeId, 120, discoveryData);
      }
    );
    try {
      await commissioningManager.executeCommissioning();
    } catch (error) {
      if (this.commissionedNodes.has(peerNodeId)) {
        this.commissionedNodes.delete(peerNodeId);
      }
      throw error;
    }
    this.controllerStorage.set("fabric", this.fabric.toStorageObject());
    return peerNodeId;
  }
  async reconnectLastKnownAddress(peerNodeId, operationalAddress, discoveryData) {
    const { ip, port } = operationalAddress;
    try {
      logger.debug(`Resume device connection to configured server at ${ip}:${port}`);
      const channel = await this.pair(peerNodeId, operationalAddress, discoveryData);
      this.setOperationalDeviceData(peerNodeId, operationalAddress);
      return channel;
    } catch (error) {
      if (error instanceof RetransmissionLimitReachedError || error instanceof Error && error.message.includes("EHOSTUNREACH")) {
        logger.debug(`Failed to resume device connection with ${ip}:${port}, discover the device ...`, error);
        return void 0;
      } else {
        throw error;
      }
    }
  }
  async connectOrDiscoverNode(peerNodeId, operationalAddress, timeoutSeconds, discoveryData) {
    const discoveryPromises = new Array();
    let reconnectionPollingTimer;
    if (operationalAddress !== void 0) {
      const directReconnection = await this.reconnectLastKnownAddress(
        peerNodeId,
        operationalAddress,
        discoveryData
      );
      if (directReconnection !== void 0) {
        return directReconnection;
      }
      if (timeoutSeconds === void 0) {
        const { promise, resolver, rejecter } = createPromise();
        reconnectionPollingTimer = Time.getPeriodicTimer(RECONNECTION_POLLING_INTERVAL, async () => {
          try {
            logger.debug(`Polling for device at ${serverAddressToString(operationalAddress)} ...`);
            const result = await this.reconnectLastKnownAddress(
              peerNodeId,
              operationalAddress,
              discoveryData
            );
            if (result !== void 0 && reconnectionPollingTimer?.isRunning) {
              reconnectionPollingTimer?.stop();
              resolver(result);
            }
          } catch (error) {
            if (reconnectionPollingTimer?.isRunning) {
              reconnectionPollingTimer?.stop();
              rejecter(error);
            }
          }
        }).start();
        discoveryPromises.push(() => promise);
      }
    }
    discoveryPromises.push(async () => {
      const scanResult = await ControllerDiscovery.discoverOperationalDevice(
        this.fabric,
        peerNodeId,
        this.mdnsScanner,
        timeoutSeconds,
        timeoutSeconds === void 0
      );
      if (reconnectionPollingTimer?.isRunning) {
        reconnectionPollingTimer?.stop();
      }
      const { result } = await ControllerDiscovery.iterateServerAddresses(
        [scanResult],
        PairRetransmissionLimitReachedError,
        async () => {
          const device = this.mdnsScanner.getDiscoveredOperationalDevice(this.fabric, peerNodeId);
          return device !== void 0 ? [device] : [];
        },
        async (address, device) => {
          const result2 = await this.pair(peerNodeId, address, device);
          this.setOperationalDeviceData(peerNodeId, address, {
            ...discoveryData,
            ...device
          });
          return result2;
        }
      );
      return result;
    });
    return await anyPromise(discoveryPromises);
  }
  /**
   * Resume a device connection and establish a CASE session that was previously paired with the controller. This
   * method will try to connect to the device using the previously used server address (if set). If that fails, the
   * device is discovered again using its operational instance details.
   * It returns the operational MessageChannel on success.
   */
  async resume(peerNodeId, timeoutSeconds, discoveryData) {
    const operationalAddress = this.getLastOperationalAddress(peerNodeId);
    try {
      return await this.connectOrDiscoverNode(peerNodeId, operationalAddress, timeoutSeconds, discoveryData);
    } catch (error) {
      if ((error instanceof DiscoveryError || error instanceof PairRetransmissionLimitReachedError) && this.commissionedNodes.has(peerNodeId)) {
        logger.info(`Resume failed, remove all sessions for node ${peerNodeId}`);
        await this.sessionManager.removeAllSessionsForNode(peerNodeId);
      }
      throw error;
    }
  }
  /** Pair with an operational device (already commissioned) and establish a CASE session. */
  async pair(peerNodeId, operationalServerAddress, discoveryData) {
    const { ip, port } = operationalServerAddress;
    const isIpv6Address = isIPv6(ip);
    const operationalInterface = isIpv6Address ? this.netInterfaceIpv6 : this.netInterfaceIpv4;
    if (operationalInterface === void 0) {
      throw new PairRetransmissionLimitReachedError(
        `IPv${isIpv6Address ? "6" : "4"} interface not initialized for port ${port}. Cannot use ${ip} for pairing.`
      );
    }
    const operationalChannel = await operationalInterface.openChannel(operationalServerAddress);
    const { sessionParameters } = this.findResumptionRecordByNodeId(peerNodeId) ?? {};
    const unsecureSession = this.sessionManager.createUnsecureSession({
      sessionParameters: {
        idleIntervalMs: discoveryData?.SII ?? sessionParameters?.idleIntervalMs,
        activeIntervalMs: discoveryData?.SAI ?? sessionParameters?.activeIntervalMs,
        activeThresholdMs: discoveryData?.SAT ?? sessionParameters?.activeThresholdMs
      },
      isInitiator: true
    });
    const operationalUnsecureMessageExchange = new MessageChannel(operationalChannel, unsecureSession);
    const operationalSecureSession = await tryCatchAsync(
      async () => {
        const exchange = this.exchangeManager.initiateExchangeWithChannel(
          operationalUnsecureMessageExchange,
          SECURE_CHANNEL_PROTOCOL_ID
        );
        return tryCatchAsync(
          async () => await this.caseClient.pair(this, exchange, this.fabric, peerNodeId),
          Error,
          async (error) => {
            await exchange.close();
            throw error;
          }
        );
      },
      RetransmissionLimitReachedError,
      (error) => {
        throw new PairRetransmissionLimitReachedError(error.message);
      }
      // Convert error
    );
    await unsecureSession.destroy();
    const channel = new MessageChannel(operationalChannel, operationalSecureSession);
    await this.channelManager.setChannel(this.fabric, peerNodeId, channel);
    return channel;
  }
  isCommissioned() {
    return this.commissionedNodes.size > 0;
  }
  getCommissionedNodes() {
    return Array.from(this.commissionedNodes.keys());
  }
  getCommissionedNodesDetails() {
    return Array.from(this.commissionedNodes.entries()).map(
      ([nodeId, { operationalServerAddress, discoveryData, basicInformationData }]) => ({
        nodeId,
        operationalAddress: operationalServerAddress ? serverAddressToString(operationalServerAddress) : void 0,
        advertisedName: discoveryData?.DN,
        discoveryData,
        basicInformationData
      })
    );
  }
  setOperationalDeviceData(nodeId, operationalServerAddress, discoveryData) {
    const nodeDetails = this.commissionedNodes.get(nodeId) ?? {};
    nodeDetails.operationalServerAddress = operationalServerAddress;
    if (discoveryData !== void 0) {
      nodeDetails.discoveryData = {
        ...nodeDetails.discoveryData,
        ...discoveryData
      };
    }
    this.commissionedNodes.set(nodeId, nodeDetails);
    this.storeCommissionedNodes();
  }
  enhanceCommissionedNodeDetails(nodeId, data) {
    const nodeDetails = this.commissionedNodes.get(nodeId);
    if (nodeDetails === void 0) {
      throw new Error(`Node ${nodeId} is not commissioned.`);
    }
    const { basicInformationData } = data;
    nodeDetails.basicInformationData = basicInformationData;
    this.commissionedNodes.set(nodeId, nodeDetails);
    this.storeCommissionedNodes();
  }
  getLastOperationalAddress(nodeId) {
    return this.commissionedNodes.get(nodeId)?.operationalServerAddress;
  }
  storeCommissionedNodes() {
    this.controllerStorage.set("commissionedNodes", Array.from(this.commissionedNodes.entries()));
  }
  /**
   * Connect to the device by opening a channel and creating a new CASE session if necessary.
   * Returns a InteractionClient on success.
   */
  async connect(peerNodeId, timeoutSeconds, discoveryData) {
    if (discoveryData == void 0) {
      discoveryData = this.commissionedNodes.get(peerNodeId)?.discoveryData;
    }
    let channel;
    try {
      channel = this.channelManager.getChannel(this.fabric, peerNodeId);
    } catch (error) {
      if (error instanceof NoChannelError) {
        channel = await this.resume(peerNodeId, timeoutSeconds, discoveryData);
      } else {
        throw error;
      }
    }
    return new InteractionClient(
      new ExchangeProvider(this.exchangeManager, channel, async () => {
        await this.channelManager.removeChannel(this.fabric, peerNodeId);
        await this.resume(peerNodeId, 60);
        return this.channelManager.getChannel(this.fabric, peerNodeId);
      }),
      peerNodeId
    );
  }
  async getNextAvailableSessionId() {
    return this.sessionManager.getNextAvailableSessionId();
  }
  async createSecureSession(args) {
    const session = await this.sessionManager.createSecureSession({
      ...args,
      closeCallback: async () => {
        logger.debug(`Remove ${session.isPase() ? "PASE" : "CASE"} session`, session.name);
        if (!session.closingAfterExchangeFinished) {
          await this.exchangeManager.closeSession(session);
        }
        this.sessionClosedCallback?.(args.peerNodeId);
      }
    });
    return session;
  }
  getResumptionRecord(resumptionId) {
    return this.sessionManager.findResumptionRecordById(resumptionId);
  }
  findResumptionRecordByNodeId(nodeId) {
    return this.sessionManager.findResumptionRecordByNodeId(nodeId);
  }
  saveResumptionRecord(resumptionRecord) {
    return this.sessionManager.saveResumptionRecord(resumptionRecord);
  }
  announce() {
  }
  async close() {
    await this.exchangeManager.close();
    await this.sessionManager.close();
    await this.channelManager.close();
    await this.netInterfaceBle?.close();
    await this.netInterfaceIpv4?.close();
    await this.netInterfaceIpv6.close();
  }
  getActiveSessionInformation() {
    return this.sessionManager.getActiveSessionInformation();
  }
}
export {
  MatterController,
  PairRetransmissionLimitReachedError
};
