// @project-chip/matter.js/dist/esm/MatterDevice.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { AdministratorCommissioning } from "./cluster/definitions/AdministratorCommissioningCluster.js";
import { BasicInformation } from "./cluster/definitions/BasicInformationCluster.js";
import { GeneralCommissioning } from "./cluster/definitions/GeneralCommissioningCluster.js";
import { OperationalCredentials } from "./cluster/definitions/OperationalCredentialsCluster.js";
import { MAXIMUM_COMMISSIONING_TIMEOUT_S } from "./cluster/server/AdministratorCommissioningServer.js";
import { FailSafeManager } from "./common/FailSafeManager.js";
import { InternalError, MatterFlowError } from "./common/MatterError.js";
import { Crypto } from "./crypto/Crypto.js";
import { FabricManager } from "./fabric/FabricManager.js";
import { Logger } from "./log/Logger.js";
import { isNetworkInterface } from "./net/NetInterface.js";
import { NetworkError } from "./net/Network.js";
import { ChannelManager } from "./protocol/ChannelManager.js";
import { ExchangeManager } from "./protocol/ExchangeManager.js";
import { StatusCode, StatusResponseError } from "./protocol/interaction/StatusCode.js";
import { SecureChannelProtocol } from "./protocol/securechannel/SecureChannelProtocol.js";
import { SessionManager } from "./session/SessionManager.js";
import { PaseServer } from "./session/pase/PaseServer.js";
import { Time } from "./time/Time.js";
const logger = Logger.get("MatterDevice");
const DEVICE_ANNOUNCEMENT_DURATION_MS = MAXIMUM_COMMISSIONING_TIMEOUT_S * 1e3;
const DEVICE_ANNOUNCEMENT_INTERVAL_MS = 60 * 1e3;
class MatterDevice {
  constructor(deviceName, deviceType, vendorId, productId, discriminator, initialPasscode, storage, commissioningChangedCallback, sessionChangedCallback) {
    this.deviceName = deviceName;
    this.deviceType = deviceType;
    this.vendorId = vendorId;
    this.productId = productId;
    this.discriminator = discriminator;
    this.initialPasscode = initialPasscode;
    this.storage = storage;
    this.commissioningChangedCallback = commissioningChangedCallback;
    this.sessionChangedCallback = sessionChangedCallback;
    this.scanners = new Array();
    this.broadcasters = new Array();
    this.transportInterfaces = new Array();
    this.channelManager = new ChannelManager();
    this.secureChannelProtocol = new SecureChannelProtocol(() => this.endCommissioning());
    this.activeCommissioningMode = AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen;
    this.announcementStartedTime = null;
    this.isClosing = false;
    this.fabricManager = new FabricManager(this.storage, (fabricIndex, peerNodeId) => {
      this.sessionManager.removeResumptionRecord(peerNodeId);
      this.commissioningChangedCallback(fabricIndex);
    });
    this.sessionManager = new SessionManager(this, this.storage);
    this.sessionManager.initFromStorage(this.fabricManager.getFabrics());
    this.exchangeManager = new ExchangeManager(this.sessionManager, this.channelManager);
    this.addProtocolHandler(this.secureChannelProtocol);
    this.announceInterval = Time.getPeriodicTimer(DEVICE_ANNOUNCEMENT_INTERVAL_MS, () => this.announce());
  }
  addScanner(scanner) {
    this.scanners.push(scanner);
    return this;
  }
  hasBroadcaster(broadcaster) {
    return this.broadcasters.includes(broadcaster);
  }
  addBroadcaster(broadcaster) {
    this.broadcasters.push(broadcaster);
    return this;
  }
  addTransportInterface(netInterface) {
    this.exchangeManager.addTransportInterface(netInterface);
    this.transportInterfaces.push(netInterface);
    return this;
  }
  hasProtocolHandler(protocolId) {
    return this.exchangeManager.hasProtocolHandler(protocolId);
  }
  addProtocolHandler(protocol) {
    this.exchangeManager.addProtocolHandler(protocol);
    return this;
  }
  async start() {
    await this.startAnnouncement();
  }
  async startAnnouncement() {
    if (this.isClosing)
      return;
    if (this.announceInterval.isRunning) {
      this.announceInterval.stop();
    }
    this.announcementStartedTime = Time.nowMs();
    this.announceInterval.start();
    await this.announce();
  }
  async announce(announceOnce = false) {
    if (!announceOnce) {
      if (this.announcementStartedTime !== null && Time.nowMs() - this.announcementStartedTime > DEVICE_ANNOUNCEMENT_DURATION_MS) {
        await this.endCommissioning();
        logger.debug("Announcement duration reached, stop announcing");
        return;
      }
      if (this.activeCommissioningMode !== AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen) {
        for (const broadcaster of this.broadcasters) {
          await broadcaster.announce();
        }
        return;
      }
    }
    const fabrics = this.fabricManager.getFabrics();
    if (fabrics.length) {
      let fabricsWithoutSessions = 0;
      for (const fabric of fabrics) {
        const session = this.sessionManager.getSessionForNode(fabric, fabric.rootNodeId);
        if (session === void 0 || !session.isSecure() || session.numberOfActiveSubscriptions === 0) {
          fabricsWithoutSessions++;
          logger.debug("Announcing", Logger.dict({ fabric: fabric.fabricId }));
        }
      }
      for (const broadcaster of this.broadcasters) {
        await broadcaster.setFabrics(fabrics);
        if (fabricsWithoutSessions > 0) {
          await broadcaster.announce();
        }
      }
    } else {
      for (const broadcaster of this.broadcasters) {
        await broadcaster.expireFabricAnnouncement();
      }
      await this.allowBasicCommissioning();
    }
  }
  async announceAsCommissionable(mode, activeCommissioningEndCallback, discriminator) {
    if (this.activeCommissioningMode !== AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen) {
      throw new InternalError("Commissioning window already open with different mode!");
    }
    if (this.activeCommissioningEndCallback !== void 0) {
      throw new InternalError("Commissioning window already open with different callback!");
    }
    this.activeCommissioningMode = mode;
    if (activeCommissioningEndCallback !== void 0) {
      this.activeCommissioningEndCallback = activeCommissioningEndCallback;
    }
    this.sendCommissionableAnnouncement(mode, discriminator).catch(
      (error) => logger.warn("Error sending announcement", error)
    );
  }
  async sendCommissionableAnnouncement(mode, discriminator) {
    for (const broadcaster of this.broadcasters) {
      await broadcaster.setCommissionMode(
        mode === AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen ? 2 : 1,
        {
          deviceName: this.deviceName,
          deviceType: this.deviceType,
          vendorId: this.vendorId,
          productId: this.productId,
          discriminator: discriminator ?? this.discriminator
        }
      );
    }
    await this.startAnnouncement();
  }
  async getNextAvailableSessionId() {
    return this.sessionManager.getNextAvailableSessionId();
  }
  async createSecureSession(args) {
    const { fabric } = args;
    const session = await this.sessionManager.createSecureSession({
      ...args,
      closeCallback: async () => {
        logger.debug(`Remove ${session.isPase() ? "PASE" : "CASE"} session`, session.name);
        if (!session.closingAfterExchangeFinished) {
          await this.exchangeManager.closeSession(session);
        }
        const currentFabric = session.getFabric();
        if (currentFabric !== void 0) {
          this.sessionChangedCallback(currentFabric.fabricIndex);
        }
        await this.startAnnouncement();
      },
      subscriptionChangedCallback: () => {
        const currentFabric = session.getFabric();
        if (currentFabric !== void 0) {
          this.sessionChangedCallback(currentFabric.fabricIndex);
        }
      }
    });
    if (fabric !== void 0) {
      this.sessionChangedCallback(fabric.fabricIndex);
    }
    return session;
  }
  findFabricFromDestinationId(destinationId, peerRandom) {
    return this.fabricManager.findFabricFromDestinationId(destinationId, peerRandom);
  }
  updateFabric(fabric) {
    this.fabricManager.updateFabric(fabric);
    this.sessionManager.updateFabricForResumptionRecords(fabric);
    this.commissioningChangedCallback(fabric.fabricIndex);
  }
  getNextFabricIndex() {
    return this.fabricManager.getNextFabricIndex();
  }
  async addFabric(fabric) {
    this.fabricManager.addFabric(fabric);
    if (this.failSafeContext !== void 0) {
      this.failSafeContext.associatedFabric = fabric;
    }
    this.commissioningChangedCallback(fabric.fabricIndex);
    const fabrics = this.fabricManager.getFabrics();
    this.sendFabricAnnouncements(fabrics, true).catch(
      (error) => logger.warn(`Error sending Fabric announcement for Index ${fabric.fabricIndex}`, error)
    );
    logger.info("Announce done", Logger.dict({ fabric: fabric.fabricId, fabricIndex: fabric.fabricIndex }));
    return fabric.fabricIndex;
  }
  async sendFabricAnnouncements(fabrics, expireCommissioningAnnouncement = false) {
    for (const broadcaster of this.broadcasters) {
      await broadcaster.setFabrics(fabrics, expireCommissioningAnnouncement);
      await broadcaster.announce();
    }
  }
  getFabricByIndex(fabricIndex) {
    return this.fabricManager.getFabrics().find((fabric) => fabric.fabricIndex === fabricIndex);
  }
  initiateExchange(fabric, nodeId, protocolId) {
    return this.exchangeManager.initiateExchange(fabric, nodeId, protocolId);
  }
  findResumptionRecordById(resumptionId) {
    return this.sessionManager.findResumptionRecordById(resumptionId);
  }
  saveResumptionRecord(resumptionRecord) {
    return this.sessionManager.saveResumptionRecord(resumptionRecord);
  }
  async removePaseSession() {
    const session = this.sessionManager.getPaseSession();
    if (session) {
      await session.close(true);
    }
  }
  assertFailSafeArmed(message) {
    if (this.isFailsafeArmed())
      return;
    throw new StatusResponseError(
      message ?? "Failsafe timer needs to be armed to execute this action.",
      StatusCode.FailsafeRequired
    );
  }
  async failSafeExpired() {
    if (this.failSafeContext === void 0)
      return;
    const failSafeContext = this.failSafeContext;
    this.failSafeContext = void 0;
    logger.info("Failsafe timer expired, Reset fabric builder.");
    if (failSafeContext.fabricIndex !== void 0 && !failSafeContext.forUpdateNoc) {
      logger.debug(`Revoking fabric with index ${failSafeContext.fabricIndex}`);
      await this.fabricManager.revokeFabric(failSafeContext.fabricIndex);
    }
    await this.removePaseSession();
    let fabric = void 0;
    if (failSafeContext.fabricIndex !== void 0) {
      const fabricIndex = failSafeContext.fabricIndex;
      fabric = this.fabricManager.getFabrics().find((fabric2) => fabric2.fabricIndex === fabricIndex);
      if (fabric !== void 0) {
        const session = this.sessionManager.getSessionForNode(fabric, fabric.rootNodeId);
        if (session !== void 0 && session.isSecure()) {
          await session.close(false);
        }
      }
    }
    failSafeContext.restoreEndpointState();
    if (failSafeContext.associatedFabric !== void 0) {
      if (failSafeContext.forUpdateNoc) {
        this.updateFabric(failSafeContext.associatedFabric);
      }
      const operationalCredentialsCluster = failSafeContext.rootEndpoint.getClusterServer(
        OperationalCredentials.Cluster
      );
      operationalCredentialsCluster?.attributes.nocs.updatedLocalForFabric(failSafeContext.associatedFabric);
      operationalCredentialsCluster?.attributes.fabrics.updatedLocalForFabric(failSafeContext.associatedFabric);
    }
    if (fabric !== void 0) {
      const fabricIndex = failSafeContext.fabricBuilder.getFabricIndex();
      if (fabricIndex !== void 0) {
        const fabric2 = this.fabricManager.getFabrics().find((fabric3) => fabric3.fabricIndex === fabricIndex);
        if (fabric2 !== void 0) {
          const basicInformationCluster = failSafeContext.rootEndpoint.getClusterServer(
            BasicInformation.Cluster
          );
          basicInformationCluster?.triggerLeaveEvent?.({ fabricIndex });
          await fabric2.remove();
          const operationalCredentialsCluster = failSafeContext.rootEndpoint.getClusterServer(
            OperationalCredentials.Cluster
          );
          operationalCredentialsCluster?.attributes.nocs.updatedLocalForFabric(fabric2);
          operationalCredentialsCluster?.attributes.commissionedFabrics.updatedLocal();
          operationalCredentialsCluster?.attributes.fabrics.updatedLocalForFabric(fabric2);
          operationalCredentialsCluster?.attributes.trustedRootCertificates.updatedLocal();
        }
      }
    }
    const generalCommissioningCluster = failSafeContext.rootEndpoint.getClusterServer(GeneralCommissioning.Cluster);
    generalCommissioningCluster?.setBreadcrumbAttribute(0);
  }
  async armFailSafe(expiryLengthSeconds, maxCumulativeFailsafeSeconds, associatedFabric, endpoint) {
    if (this.failSafeContext === void 0) {
      if (expiryLengthSeconds === 0)
        return;
      this.failSafeContext = new FailSafeManager(
        this,
        associatedFabric,
        expiryLengthSeconds,
        maxCumulativeFailsafeSeconds,
        () => this.failSafeExpired(),
        endpoint
      );
      logger.debug(`Arm failSafe timer for ${expiryLengthSeconds}s.`);
    } else {
      await this.failSafeContext.reArm(associatedFabric, expiryLengthSeconds);
      if (expiryLengthSeconds > 0) {
        logger.debug(`Extend failSafe timer for ${expiryLengthSeconds}s.`);
      }
    }
  }
  isFailsafeArmed() {
    return this.failSafeContext !== void 0;
  }
  getFailSafeContext() {
    if (this.failSafeContext === void 0)
      throw new MatterFlowError("armFailSafe should be called first!");
    return this.failSafeContext;
  }
  getFabrics() {
    return this.fabricManager.getFabrics();
  }
  async completeCommission() {
    if (this.failSafeContext === void 0) {
      throw new MatterFlowError("armFailSafe should be called first!");
    }
    this.failSafeContext.complete();
    if (this.failSafeContext.fabricIndex !== void 0) {
      this.fabricManager.persistFabrics();
    }
    this.failSafeContext = void 0;
    await this.endCommissioning();
    await this.removePaseSession();
  }
  isCommissioned() {
    return !!this.fabricManager.getFabrics().length;
  }
  async allowEnhancedCommissioning(discriminator, paseServer, commissioningEndCallback) {
    if (this.activeCommissioningMode === AdministratorCommissioning.CommissioningWindowStatus.BasicWindowOpen) {
      throw new MatterFlowError(
        "Basic commissioning window is already open! Can not set Enhanced commissioning mode."
      );
    }
    this.secureChannelProtocol.setPaseCommissioner(paseServer);
    await this.announceAsCommissionable(
      AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen,
      commissioningEndCallback,
      discriminator
    );
  }
  async allowBasicCommissioning(commissioningEndCallback) {
    if (this.activeCommissioningMode === AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen) {
      throw new MatterFlowError(
        "Enhanced commissioning window is already open! Can not set Basic commissioning mode."
      );
    }
    this.secureChannelProtocol.setPaseCommissioner(
      await PaseServer.fromPin(this.initialPasscode, {
        iterations: 1e3,
        salt: Crypto.get().getRandomData(32)
      })
    );
    await this.announceAsCommissionable(
      AdministratorCommissioning.CommissioningWindowStatus.BasicWindowOpen,
      commissioningEndCallback
    );
  }
  async endCommissioning() {
    logger.debug("Commissioning mode ended, stop announcements.");
    if (this.activeCommissioningMode === AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen || this.isCommissioned()) {
      this.secureChannelProtocol.removePaseCommissioner();
    }
    this.activeCommissioningMode = AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen;
    this.announceInterval.stop();
    this.announcementStartedTime = null;
    if (this.activeCommissioningEndCallback !== void 0) {
      const activeCommissioningEndCallback = this.activeCommissioningEndCallback;
      this.activeCommissioningEndCallback = void 0;
      activeCommissioningEndCallback();
    }
    for (const broadcaster of this.broadcasters) {
      await broadcaster.expireCommissioningAnnouncement();
    }
    logger.info("All Announcements expired");
  }
  existsOpenPaseSession() {
    return !!this.sessionManager.getPaseSession();
  }
  async findDevice(fabric, nodeId, timeOutSeconds = 5) {
    const device = await this.scanners[0].findOperationalDevice(fabric, nodeId, timeOutSeconds);
    if (device === void 0)
      return void 0;
    const session = this.sessionManager.getSessionForNode(fabric, nodeId);
    if (session === void 0)
      return void 0;
    const networkInterface = this.transportInterfaces.find((netInterface) => isNetworkInterface(netInterface));
    if (networkInterface === void 0 || !isNetworkInterface(networkInterface)) {
      throw new NetworkError("No network interface found");
    }
    return { session, channel: await networkInterface.openChannel(device.addresses[0]) };
  }
  async stop() {
    this.isClosing = true;
    await this.endCommissioning();
    for (const broadcaster of this.broadcasters) {
      await broadcaster.expireAllAnnouncements();
    }
    if (this.failSafeContext) {
      await this.failSafeContext.expire();
      this.failSafeContext = void 0;
    }
    await this.exchangeManager.close();
    await this.sessionManager.close();
    await this.channelManager.close();
    for (const transportInterface of this.transportInterfaces) {
      await transportInterface.close();
    }
  }
  getActiveSessionInformation() {
    return this.sessionManager.getActiveSessionInformation();
  }
}
export {
  MatterDevice
};
