// @project-chip/matter.js/dist/esm/net/fake/UdpChannelFake.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { isIPv4 } from "../../util/Ip.js";
import { NetworkError } from "../Network.js";
import { FAKE_INTERFACE_NAME, SimulatedNetwork } from "./SimulatedNetwork.js";
class UdpChannelFake {
  constructor(localAddress, listeningAddress, listeningPort) {
    this.localAddress = localAddress;
    this.listeningAddress = listeningAddress;
    this.netListeners = new Array();
    this.simulatedNetwork = SimulatedNetwork.get();
    this.listeningPort = listeningPort ?? 1024 + Math.floor(Math.random() * 64511);
  }
  static async create(network, { listeningAddress, listeningPort, netInterface, type }) {
    const { ips } = network.getIpMac(netInterface ?? FAKE_INTERFACE_NAME);
    const ipv4 = type === "udp4";
    const localAddress = ips.filter((ip) => isIPv4(ip) || !ipv4)[0];
    if (localAddress === void 0)
      throw new NetworkError("No matching IP on the specified interface");
    return new UdpChannelFake(localAddress, listeningAddress, listeningPort);
  }
  onData(listener) {
    const netListener = this.simulatedNetwork.onUdpData(this.listeningAddress, this.listeningPort, listener);
    this.netListeners.push(netListener);
    return netListener;
  }
  async send(host, port, data) {
    this.simulatedNetwork.sendUdp(this.localAddress, this.listeningPort, host, port, data);
  }
  async close() {
    for (const netListener of this.netListeners) {
      await netListener.close();
    }
    this.netListeners.length = 0;
  }
}
export {
  UdpChannelFake
};
