// @project-chip/matter.js/dist/esm/model/logic/ModelVariantTraversal.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { InternalError } from "../../common/MatterError.js";
import { Logger } from "../../log/Logger.js";
import { ElementTag } from "../definitions/index.js";
import { Model, ValueModel } from "../models/index.js";
import { ModelTraversal } from "./ModelTraversal.js";
const logger = Logger.get("ModelVariantTraversal");
class ModelVariantTraversal {
  /**
   * Create a new visitor.  Must list the valid names of sources.  The order
   * of this list implies the priority used for choosing a name when multiple
   * model variants have different names.
   */
  constructor(sourceNames) {
    this.sourceNames = sourceNames;
    this.visiting = false;
    this.modelTraversal = new ModelTraversal();
  }
  /**
   * Initiate traversal.  The class is stateful so this call should not be
   * invoked while traversal is ongoing.
   */
  traverse(variants) {
    if (this.visiting) {
      throw new InternalError("ModelVariantVisitor.visit called with active visit; reentrancy not supported");
    }
    delete this.clusterState;
    this.visiting = true;
    try {
      return this.visitVariants(
        this.createVariantDetail(
          Object.fromEntries(
            Object.entries(variants).map(([sourceName, element]) => {
              if (!(element instanceof Model)) {
                element = Model.create(element);
              }
              return [sourceName, element];
            })
          )
        )
      );
    } finally {
      this.visiting = false;
    }
  }
  /**
   * Get the canonical name for a model.  Within cluster scope alternate
   * names may be selected, otherwise the name of the model is returned.
   */
  getCanonicalName(model) {
    if (this.clusterState) {
      const name = this.clusterState.canonicalNames.get(model);
      if (name !== void 0) {
        return name;
      }
    }
    return model.name;
  }
  /**
   * Determine if we are entering a cluster and install cluster state if so.
   */
  enterCluster(variants) {
    if (variants.tag === ElementTag.Cluster) {
      this.clusterState = {
        canonicalNames: computeCanonicalNames(this.sourceNames, variants)
      };
      return true;
    }
    return false;
  }
  /**
   * This is the function that actually recurses during the visit.
   */
  visitVariants(variants) {
    const state = this.modelTraversal.operation(() => {
      return this.visit(variants, () => {
        const enteredCluster = this.enterCluster(variants);
        const mappings = this.mapChildren(variants);
        const result = Array();
        for (const mapping of Object.values(mappings)) {
          mappings:
            for (const childVariants of mapping.slots) {
              const detail = this.createVariantDetail(childVariants);
              let inherited = false;
              let overrides = false;
              if (variants.tag === ElementTag.Cluster) {
                for (const k in variants.map) {
                  if (childVariants[k] === void 0) {
                    const inheritedModel = variants.map[k].base?.member(detail.name, [detail.tag]);
                    if (inheritedModel) {
                      inherited = true;
                    }
                    continue;
                  }
                  const child = childVariants[k];
                  if (child instanceof ValueModel && child.overridesShadow) {
                    overrides = true;
                  }
                }
              }
              if (inherited && !overrides) {
                continue mappings;
              }
              result.push(this.visitVariants(detail));
            }
        }
        if (enteredCluster) {
          delete this.clusterState;
        }
        return result;
      });
    });
    return state;
  }
  mapChildren(variants) {
    const mappings = {};
    for (const [sourceName, variant] of Object.entries(variants.map)) {
      for (let i = 0; i < variant.children.length; i++) {
        const child = variant.children[i];
        if (child.global) {
          continue;
        }
        const mapping = mappings[child.tag] || (mappings[child.tag] = { slots: [], idToSlot: {}, nameToSlot: {} });
        const childId = child.key;
        const childName = this.getCanonicalName(child);
        let slot;
        if (childId !== void 0) {
          slot = mapping.idToSlot[childId];
        }
        if (slot === void 0) {
          slot = mapping.nameToSlot[childName];
        }
        if (slot === void 0) {
          slot = mapping.slots.length;
          mapping.slots.push({});
        }
        if (childId !== void 0) {
          if (mapping.idToSlot[childId] === void 0) {
            mapping.idToSlot[childId] = slot;
          }
        }
        if (mapping.nameToSlot[childName] === void 0) {
          mapping.nameToSlot[childName] = slot;
        }
        mapping.slots[slot][sourceName] = child;
      }
    }
    return mappings;
  }
  /**
   * Create a VariantDetail from a VariantMap.
   */
  createVariantDetail(map) {
    let tag;
    let id;
    let name;
    this.sourceNames.forEach((sourceName) => {
      const variant = map[sourceName];
      if (variant) {
        if (!tag) {
          tag = variant.tag;
        } else if (tag !== variant.tag) {
          throw new InternalError(
            `Variant tag mismatch; previous variant identified as ${tag} but ${sourceName} identifies as ${tag}`
          );
        }
        if (!id) {
          id = variant.id;
        }
        if (!name) {
          name = this.getCanonicalName(variant);
        }
      }
    });
    if (!tag) {
      throw new InternalError("No tag identified in variant set");
    }
    if (!name) {
      throw new InternalError("No name identified in variant set");
    }
    return { tag, id, name, map };
  }
}
function computeCanonicalNames(sourceNames, variants) {
  const datatypeNameMap = /* @__PURE__ */ new Map();
  let numberOfMappings;
  do {
    numberOfMappings = datatypeNameMap.size;
    inferEquivalentDatatypes(sourceNames, variants, datatypeNameMap);
  } while (numberOfMappings != datatypeNameMap.size);
  const canonicalNames = chooseCanonicalNames(sourceNames, variants, datatypeNameMap);
  return canonicalNames;
}
function inferEquivalentDatatypes(sourceNames, variants, datatypeNameMap) {
  const nameVariants = /* @__PURE__ */ new Map();
  const traversal = new class extends ModelVariantTraversal {
    visit(variants2, recurse) {
      let mapEntry;
      for (let priority = 0; priority < sourceNames.length; priority++) {
        const sourceName = sourceNames[priority];
        const variant = variants2.map[sourceName];
        const base = variant?.base;
        if (!base || base.parent?.tag !== ElementTag.Cluster) {
          continue;
        }
        if (!mapEntry) {
          mapEntry = {
            mapTo: base.name,
            priority
          };
        }
        const existingEntry = nameVariants.get(base);
        if (existingEntry) {
          if (existingEntry.priority > mapEntry.priority) {
            nameVariants.set(base, mapEntry);
          } else if (existingEntry.priority === mapEntry.priority && existingEntry.mapTo !== mapEntry.mapTo) {
            logger.warn(
              `Mapping ${sourceName} ${base.tag} ${base.name} to ${existingEntry.mapTo} but it also maps to ${mapEntry.mapTo}`
            );
          }
        } else {
          nameVariants.set(base, mapEntry);
        }
      }
      recurse();
    }
    enterCluster() {
      if (variants.tag === ElementTag.Cluster) {
        this.clusterState = { canonicalNames: datatypeNameMap };
        return true;
      }
      return false;
    }
  }(sourceNames);
  traversal.traverse(variants.map);
  for (const [model, mapEntry] of nameVariants) {
    if (mapEntry.mapTo && mapEntry.mapTo !== model.name) {
      datatypeNameMap.set(model, mapEntry.mapTo);
    }
  }
}
function chooseCanonicalNames(sourceNames, variants, datatypeNameMap) {
  const canonicalNames = /* @__PURE__ */ new Map();
  const traversal = new class extends ModelVariantTraversal {
    visit(variants2, recurse) {
      let canonicalName;
      for (let i = 0; i < sourceNames.length; i++) {
        const name = variants2.map[sourceNames[i]]?.name;
        if (!i && name !== void 0) {
          break;
        }
        if (!canonicalName) {
          canonicalName = name;
          continue;
        }
        if (name?.length < canonicalName?.length) {
          canonicalName = name;
          continue;
        }
        if (canonicalName?.toLowerCase() === name?.toLowerCase()) {
          if (canonicalName === name) {
            continue;
          }
          if (canonicalName.replace(/[^A-Z]/g, "").length < name.replace(/[^A-Z]/g, "").length) {
            canonicalName = name;
            continue;
          }
        }
      }
      if (!canonicalName) {
        canonicalName = variants2.name;
      }
      for (const sourceName in variants2.map) {
        const variant = variants2.map[sourceName];
        if (variant.name !== canonicalName) {
          canonicalNames.set(variant, canonicalName);
        }
      }
      recurse();
    }
    enterCluster(variants2) {
      if (variants2.tag === ElementTag.Cluster) {
        this.clusterState = { canonicalNames: datatypeNameMap };
        return true;
      }
      return false;
    }
  }(sourceNames);
  traversal.traverse(variants.map);
  return canonicalNames;
}
export {
  ModelVariantTraversal
};
