// @project-chip/matter.js/dist/esm/model/logic/ModelTraversal.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { InternalError } from "../../common/MatterError.js";
import { ElementTag, FieldValue, Metatype } from "../definitions/index.js";
import { Globals } from "../elements/index.js";
import { CommandModel } from "../models/index.js";
const OPERATION_DEPTH_LIMIT = 20;
class ModelTraversal {
  constructor() {
    this.operationDepth = 0;
  }
  /**
   * Perform an operation with iteration tracking.  If iteration depth limit
   */
  operation(operator, toDismiss) {
    if (this.operationDepth > OPERATION_DEPTH_LIMIT) {
      throw new InternalError("Likely cycle detected (or OPERATION_DEPTH_LIMIT needs to be bumped)");
    }
    if (toDismiss) {
      if (!this.dismissed) {
        this.dismissed = /* @__PURE__ */ new Set();
      }
      this.dismissed.add(toDismiss);
    }
    this.operationDepth++;
    try {
      return operator();
    } finally {
      this.operationDepth--;
      if (toDismiss) {
        this.dismissed?.delete(toDismiss);
      }
    }
  }
  /**
   * Perform an operation with a model dismissed from consideration for type
   * lookup.
   */
  operationWithDismissal(toDismiss, operator) {
    return this.operation(operator, toDismiss);
  }
  /**
   * Determine the type for a model.  This is the string name of the base
   * model.  Usually this is simply the type field but we infer the type of
   * some datatypes based on their parent's type.
   */
  getTypeName(model) {
    return this.operation(() => {
      if (!model) {
        return void 0;
      }
      if (model.type) {
        return model.type;
      }
      if (model.tag === ElementTag.Command || model.tag === ElementTag.Event) {
        return "struct";
      }
      let result;
      const name = model.name;
      this.visitInheritance(model.parent, (ancestor) => {
        if (ancestor.metatype) {
          switch (ancestor.name) {
            case Globals.enum8.name:
            case Globals.map8.name:
              result = Globals.uint8.name;
              return false;
            case Globals.enum16.name:
            case Globals.map16.name:
              result = Globals.uint16.name;
              return false;
            case Globals.map32.name:
              result = Globals.uint32.name;
              return false;
            case Globals.map64.name:
              result = Globals.uint64.name;
              return false;
          }
        }
        const overridden = this.findLocal(ancestor, name, [model.tag]);
        if (overridden?.type) {
          result = overridden.type;
          return false;
        }
      });
      return result;
    });
  }
  /**
   * Find the model in my inheritance hierarchy that has semantic meaning.
   * This will be the first inherited model with a metatype.
   */
  findMetabase(model) {
    return this.operation(() => {
      while (model && !model.metatype) {
        model = this.findBase(model);
      }
      return model;
    });
  }
  /**
   * Find the model a model derives from, if any.
   */
  findBase(model) {
    return this.operationWithDismissal(model, () => {
      if (!model) {
        return;
      }
      const type = this.getTypeName(model);
      if (type !== void 0) {
        for (const tag of model.allowedBaseTags) {
          const found = this.findType(model.parent, type, tag);
          if (found) {
            return found;
          }
        }
      }
    });
  }
  /**
   * Find the first global model this model derives from, if any.
   */
  findGlobalBase(model) {
    if (!model) {
      return;
    }
    let result;
    this.visitInheritance(model, (model2) => {
      if (model2.global) {
        result = model2;
        return false;
      }
    });
    return result;
  }
  /**
   * Determine whether this model derives from another.
   */
  instanceOf(model, other) {
    if (!model || !other) {
      return false;
    }
    if (model === other) {
      return true;
    }
    let result = false;
    this.visitInheritance(model, (model2) => {
      if (model2.name === other.name && model2.global === other.global) {
        result = true;
        return false;
      }
    });
    return result;
  }
  /**
   * Find an xref from this model or a parent.
   */
  findXref(model) {
    return this.operationWithDismissal(model, () => {
      if (!model) {
        return;
      }
      if (model.xref) {
        return model.xref;
      }
      return this.findXref(model.parent);
    });
  }
  /**
   * Find the model this model derives from that has children, if any.
   */
  findDefiningModel(model) {
    let result;
    this.visitInheritance(model, (model2) => {
      if (!model2.isType) {
        return false;
      }
      if (model2.children.length) {
        result = model2;
        return false;
      }
    });
    return result;
  }
  /**
   * Find a child in the parent's inheritance hierarchy with the same tag
   * and ID/name.
   */
  findShadow(model) {
    if (model === void 0) {
      return void 0;
    }
    let shadow;
    this.operationWithDismissal(model, () => {
      this.visitInheritance(this.findBase(model?.parent), (parent) => {
        if (model.id !== void 0) {
          shadow = this.findLocal(parent, model.id, [model.tag]);
          if (shadow) {
            return false;
          }
        }
        shadow = this.findLocal(parent, model.name, [model.tag]);
        if (shadow) {
          return false;
        }
      });
    });
    return shadow;
  }
  /**
   * Get an aspect that reflects extension of any shadowed aspects.  Note
   * that this searches parent's inheritance and the model's inheritance.
   * This is because aspects can be inherited by overriding an element in
   * the parent or by direct type inheritance.  Aspects in shadowed elements
   * take priority as they are presumably more specific.
   */
  findAspect(model, symbol) {
    if (!model) {
      return;
    }
    return this.operation(() => {
      let aspect = model[symbol];
      const shadowedAspect = this.findAspect(this.findShadow(model), symbol);
      if (shadowedAspect) {
        if (aspect) {
          aspect = shadowedAspect.extend(aspect);
        } else {
          aspect = shadowedAspect;
        }
      }
      const inheritedAspect = this.findAspect(this.findBase(model), symbol);
      if (inheritedAspect) {
        if (aspect) {
          aspect = inheritedAspect.extend(aspect);
        } else {
          aspect = inheritedAspect;
        }
      }
      return aspect;
    });
  }
  /**
   * Constraint aspects are specialized because we infer constraint fields
   * that are referenced in other models.
   */
  findConstraint(model, symbol, field) {
    return this.operation(() => {
      let constraint = this.findAspect(model, symbol);
      if (constraint === void 0) {
        return;
      }
      const bounds = {};
      const resolve = (field2) => {
        const value = constraint[field2];
        const name = FieldValue.referenced(value);
        if (name === void 0) {
          return;
        }
        const referenced = this.findMember(model.parent, name, [
          ElementTag.Attribute,
          ElementTag.Datatype
        ]);
        if (!referenced) {
          return;
        }
        const otherConstraint = this.findConstraint(referenced, symbol, field2);
        if (otherConstraint?.[field2]) {
          bounds[field2] = otherConstraint[field2];
        }
      };
      if (field) {
        resolve(field);
      } else {
        resolve("value");
        resolve("min");
        resolve("max");
      }
      if (Object.keys(bounds).length) {
        constraint = constraint.extend(bounds);
      }
      return constraint;
    });
  }
  /**
   * Search inherited scope for a named member.
   */
  findMember(scope, key, allowedTags) {
    return this.operation(() => {
      while (scope) {
        const result = this.findLocal(scope, key, allowedTags);
        if (result) {
          return result;
        }
        scope = this.findBase(scope);
      }
    });
  }
  /**
   * Retrieve all children of a specific type, inherited or otherwise.
   */
  findMembers(scope, allowedTags) {
    const members = Array();
    this.visitInheritance(scope, (model) => {
      for (const child of model.children) {
        if (allowedTags.indexOf(child.tag) !== -1) {
          members.push(child);
        }
      }
    });
    return members;
  }
  /**
   * Search inherited scope for a bit definition.
   */
  findBitDefinition(scope, bit) {
    return this.operation(() => {
      while (scope) {
        if (!scope.isType) {
          return;
        }
        if (scope.effectiveMetatype !== Metatype.bitmap) {
          scope = scope.parent;
          continue;
        }
        for (const c of scope.children) {
          if (c.constraint.test(bit)) {
            return c;
          }
        }
        scope = this.findBase(scope);
      }
    });
  }
  /**
   * Search inherited and structural type scope for a named type.
   */
  findType(scope, name, tag) {
    return this.operation(() => {
      if (!scope) {
        return;
      }
      const queue = Array(scope);
      for (scope = queue.shift(); scope; scope = queue.shift()) {
        if (scope.isTypeScope) {
          const result = this.findLocal(scope, name, [tag]);
          if (result) {
            return result;
          }
        }
        const inheritedScope = this.findBase(scope);
        if (inheritedScope) {
          queue.unshift(inheritedScope);
        }
        if (scope.parent) {
          queue.push(scope.parent);
        }
      }
    });
  }
  /**
   * Find the response model for a command.
   */
  findResponse(command) {
    if (command.response && command.response !== "status") {
      return new ModelTraversal().findType(command, command.response, ElementTag.Command);
    }
  }
  /**
   * Find all children of a node that reference a specific type.
   */
  findReferences(scope, type) {
    if (!scope || !type) {
      return [];
    }
    const references = Array();
    this.visit(scope, (model) => {
      if (this.findBase(model) === type) {
        references.push(model);
        return;
      }
      if (model instanceof CommandModel && this.findResponse(model) === type) {
        references.push(model);
        return;
      }
      if (model.isType) {
        const defaultValue = model.default;
        if (FieldValue.is(defaultValue, FieldValue.reference)) {
          if (defaultValue.name === type.name) {
            references.push(model);
          }
        }
      }
    });
    return references;
  }
  /**
   * Find an owning model of a specific type.
   */
  findOwner(constructor, model) {
    if (!model || model instanceof constructor || !model.parent) {
      return model;
    }
    return this.operation(() => {
      return this.findOwner(constructor, model.parent);
    });
  }
  /**
   * Find the root model.
   */
  findRoot(model) {
    if (!model) {
      return void 0;
    }
    if (!model.parent) {
      return model;
    }
    this.operation(() => {
      return this.findRoot(model.parent);
    });
  }
  /**
   * Visit all nodes in the model tree.
   */
  visit(model, visitor) {
    return this.operation(() => {
      if (visitor(model) === false) {
        return false;
      }
      for (const c of model.children) {
        if (this.visit(c, visitor) === false) {
          return false;
        }
      }
      return true;
    });
  }
  /**
   * Visit all nodes in the inheritance hierarchy until the visitor returns false.
   */
  visitInheritance(model, visitor) {
    return this.operation(() => {
      if (!model) {
        return;
      }
      if (visitor(model) === false) {
        return false;
      }
      const base = this.findBase(model);
      this.visitInheritance(base, visitor);
    });
  }
  /**
   * Search for a direct child by name.
   */
  findLocal(scope, key, allowedTags) {
    for (const c of scope.children) {
      if (c.is(key) && allowedTags.indexOf(c.tag) !== -1 && !this.dismissed?.has(c)) {
        return c;
      }
    }
  }
}
export {
  ModelTraversal
};
