// @project-chip/matter.js/dist/cjs/mdns/MdnsInstanceBroadcaster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MdnsInstanceBroadcaster_exports = {};
__export(MdnsInstanceBroadcaster_exports, {
  MdnsInstanceBroadcaster: () => MdnsInstanceBroadcaster
});
module.exports = __toCommonJS(MdnsInstanceBroadcaster_exports);
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class MdnsInstanceBroadcaster {
  constructor(instancePort, mdnsBroadcaster) {
    this.instancePort = instancePort;
    this.mdnsBroadcaster = mdnsBroadcaster;
  }
  async setCommissionMode(mode, deviceData) {
    await this.mdnsBroadcaster.setCommissionMode(this.instancePort, mode, deviceData);
  }
  /** Set the Broadcaster Data to announce a device for operative discovery (aka "already paired") */
  async setFabrics(fabrics, expireCommissioningAnnouncements = false) {
    if (expireCommissioningAnnouncements) {
      await this.mdnsBroadcaster.expireCommissioningAnnouncement(this.instancePort);
    }
    await this.mdnsBroadcaster.setFabrics(this.instancePort, fabrics);
  }
  /** Set the Broadcaster data to announce a Commissioner (aka Commissioner discovery) */
  async setCommissionerInfo(commissionerData) {
    await this.mdnsBroadcaster.setCommissionerInfo(this.instancePort, commissionerData);
  }
  async announce() {
    await this.mdnsBroadcaster.announce(this.instancePort);
  }
  async expireFabricAnnouncement() {
    await this.mdnsBroadcaster.expireFabricAnnouncement(this.instancePort);
  }
  async expireCommissioningAnnouncement() {
    await this.mdnsBroadcaster.expireCommissioningAnnouncement(this.instancePort);
  }
  async expireAllAnnouncements() {
    await this.mdnsBroadcaster.expireAllAnnouncements(this.instancePort);
  }
}
