// @project-chip/matter.js/dist/cjs/schema/PairingCodeSchema.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PairingCodeSchema_exports = {};
__export(PairingCodeSchema_exports, {
  CommissionningFlowType: () => CommissionningFlowType,
  DiscoveryCapabilitiesBitmap: () => DiscoveryCapabilitiesBitmap,
  DiscoveryCapabilitiesSchema: () => DiscoveryCapabilitiesSchema,
  ManualPairingCodeCodec: () => ManualPairingCodeCodec,
  QrPairingCodeCodec: () => QrPairingCodeCodec
});
module.exports = __toCommonJS(PairingCodeSchema_exports);
var import_MatterError = require("../common/MatterError.js");
var import_VendorId = require("../datatype/VendorId.js");
var import_Verhoeff = require("../math/Verhoeff.js");
var import_Base38Schema = require("./Base38Schema.js");
var import_BitmapSchema = require("./BitmapSchema.js");
var import_Schema = require("./Schema.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var CommissionningFlowType = /* @__PURE__ */ ((CommissionningFlowType2) => {
  CommissionningFlowType2[CommissionningFlowType2["Standard"] = 0] = "Standard";
  CommissionningFlowType2[CommissionningFlowType2["UserIntent"] = 1] = "UserIntent";
  CommissionningFlowType2[CommissionningFlowType2["Custom"] = 2] = "Custom";
  return CommissionningFlowType2;
})(CommissionningFlowType || {});
const DiscoveryCapabilitiesBitmap = {
  /** Device supports hosting a Soft-AP when not commissioned. */
  softAccessPoint: (0, import_BitmapSchema.BitFlag)(0),
  /** Device supports BLE for discovery when not commissioned. */
  ble: (0, import_BitmapSchema.BitFlag)(1),
  /** Device is already on the IP network. */
  onIpNetwork: (0, import_BitmapSchema.BitFlag)(2)
};
const DiscoveryCapabilitiesSchema = (0, import_BitmapSchema.BitmapSchema)(DiscoveryCapabilitiesBitmap);
const QrCodeDataSchema = (0, import_BitmapSchema.ByteArrayBitmapSchema)({
  version: (0, import_BitmapSchema.BitField)(0, 3),
  vendorId: (0, import_BitmapSchema.BitField)(3, 16),
  productId: (0, import_BitmapSchema.BitField)(19, 16),
  flowType: (0, import_BitmapSchema.BitFieldEnum)(35, 2),
  discoveryCapabilities: (0, import_BitmapSchema.BitField)(37, 8),
  discriminator: (0, import_BitmapSchema.BitField)(45, 12),
  passcode: (0, import_BitmapSchema.BitField)(57, 27)
});
const PREFIX = "MT:";
class QrPairingCodeSchema extends import_Schema.Schema {
  encodeInternal(payloadData) {
    return PREFIX + import_Base38Schema.Base38.encode(QrCodeDataSchema.encode(payloadData));
  }
  decodeInternal(encoded) {
    if (!encoded.startsWith(PREFIX))
      throw new import_MatterError.UnexpectedDataError("The pairing code should start with MT:");
    return QrCodeDataSchema.decode(import_Base38Schema.Base38.decode(encoded.slice(PREFIX.length)));
  }
}
const QrPairingCodeCodec = new QrPairingCodeSchema();
class ManualPairingCodeSchema extends import_Schema.Schema {
  encodeInternal({ discriminator, passcode, vendorId, productId }) {
    if (discriminator === void 0)
      throw new import_MatterError.UnexpectedDataError("discriminator is required");
    if (discriminator > 4095)
      throw new import_MatterError.UnexpectedDataError("discriminator value must be less than 4096");
    let result = "";
    const hasVendorProductIds = vendorId !== void 0 && productId !== void 0;
    result += discriminator >> 10 | (hasVendorProductIds ? 1 << 2 : 0);
    result += ((discriminator & 768) << 6 | passcode & 16383).toString().padStart(5, "0");
    result += (passcode >> 14).toString().padStart(4, "0");
    if (hasVendorProductIds) {
      result += vendorId.toString().padStart(5, "0");
      result += productId.toString().padStart(5, "0");
    }
    result += new import_Verhoeff.Verhoeff().computeChecksum(result);
    return result;
  }
  decodeInternal(encoded) {
    if (encoded.length !== 11 && encoded.length != 21) {
      throw new import_MatterError.UnexpectedDataError("Invalid pairing code");
    }
    if (new import_Verhoeff.Verhoeff().computeChecksum(encoded.slice(0, -1)) !== parseInt(encoded.slice(-1))) {
      throw new import_MatterError.UnexpectedDataError("Invalid checksum");
    }
    const hasVendorProductIds = !!(parseInt(encoded[0]) & 1 << 2);
    const shortDiscriminator = (parseInt(encoded[0]) & 3) << 2 | parseInt(encoded.slice(1, 6)) >> 14 & 3;
    const passcode = parseInt(encoded.slice(1, 6)) & 16383 | parseInt(encoded.slice(6, 10)) << 14;
    let vendorId;
    let productId;
    if (hasVendorProductIds) {
      vendorId = (0, import_VendorId.VendorId)(parseInt(encoded.slice(10, 15)));
      productId = parseInt(encoded.slice(15, 20));
    }
    return { shortDiscriminator, passcode, vendorId, productId };
  }
}
const ManualPairingCodeCodec = new ManualPairingCodeSchema();
