// @project-chip/matter.js/dist/esm/cluster/server/AttributeServer.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ImplementationError, InternalError, MatterError } from "../../common/MatterError.js";
import { tryCatch } from "../../common/TryCatchHandler.js";
import { ValidationError } from "../../common/ValidationError.js";
import { Globals } from "../../model/index.js";
import { StatusCode, StatusResponseError } from "../../protocol/interaction/StatusCode.js";
import { NoAssociatedFabricError, assertSecureSession } from "../../session/SecureSession.js";
import { isDeepEqual } from "../../util/DeepEqual.js";
class FabricScopeError extends MatterError {
}
function createAttributeServer(clusterDef, attributeDef, attributeName, defaultValue, getClusterDataVersion, increaseClusterDataVersion, getter, setter, validator) {
  const { id, schema, writable, fabricScoped, fixed, omitChanges, timed } = attributeDef;
  if (fixed) {
    return new FixedAttributeServer(
      id,
      attributeName,
      schema,
      writable,
      false,
      timed,
      defaultValue,
      getClusterDataVersion,
      getter
    );
  }
  if (fabricScoped) {
    return new FabricScopedAttributeServer(
      id,
      attributeName,
      schema,
      writable,
      !omitChanges,
      timed,
      defaultValue,
      clusterDef,
      getClusterDataVersion,
      increaseClusterDataVersion,
      getter,
      setter,
      validator
    );
  }
  return new AttributeServer(
    id,
    attributeName,
    schema,
    writable,
    !omitChanges,
    timed,
    defaultValue,
    getClusterDataVersion,
    increaseClusterDataVersion,
    getter,
    setter,
    validator
  );
}
class BaseAttributeServer {
  constructor(id, name, schema, isWritable, isSubscribable, requiresTimedInteraction, defaultValue) {
    this.id = id;
    this.name = name;
    this.schema = schema;
    this.isWritable = isWritable;
    this.isSubscribable = isSubscribable;
    this.requiresTimedInteraction = requiresTimedInteraction;
    this.defaultValue = defaultValue;
    /**
     * The value is undefined when getter/setter are used. But we still handle the version number here.
     */
    this.value = void 0;
    this.validateWithSchema(defaultValue);
    this.value = defaultValue;
  }
  validateWithSchema(value) {
    try {
      this.schema.validate(value);
    } catch (error) {
      if (error instanceof ValidationError) {
        error.message = `Validation error for attribute "${this.name}"${error.fieldName !== void 0 ? `in field ${error.fieldName}` : ""}: ${error.message}`;
      }
      throw error;
    }
  }
  assignToEndpoint(endpoint) {
    this.endpoint = endpoint;
  }
}
class FixedAttributeServer extends BaseAttributeServer {
  constructor(id, name, schema, isWritable, isSubscribable, requiresTimedInteraction, defaultValue, getClusterDataVersion, getter) {
    super(id, name, schema, isWritable, isSubscribable, requiresTimedInteraction, defaultValue);
    this.getClusterDataVersion = getClusterDataVersion;
    this.isFixed = true;
    if (getter === void 0) {
      this.getter = () => {
        if (this.value === void 0) {
          throw new InternalError(`Attribute value for attribute "${name}" is not initialized.`);
        }
        return this.value;
      };
    } else {
      this.getter = getter;
    }
  }
  /**
   * Get the value of the attribute. This method is used by the Interaction model to read the value of the attribute
   * and includes the ACL check. It should not be used locally in the code!
   * If a getter is defined the value is determined by that getter method.
   */
  get(session, isFabricFiltered) {
    return this.getter(session, this.endpoint, isFabricFiltered);
  }
  /**
   * Get the value of the attribute including the version number. This method is used by the Interaction model to read
   * the value of the attribute and includes the ACL check. It should not be used locally in the code!
   * If a getter is defined the value is determined by that getter method. The version number is always 0 for fixed
   * attributes.
   */
  getWithVersion(session, isFabricFiltered) {
    return { version: this.getClusterDataVersion(), value: this.get(session, isFabricFiltered) };
  }
  /**
   * Get the value of the attribute locally. This method should be used locally in the code and does not include the
   * ACL check.
   * If a getter is defined the value is determined by that getter method.
   */
  getLocal() {
    return this.getter(void 0, this.endpoint);
  }
  /**
   * Initialize the value of the attribute, used when a persisted value is set initially or when values needs to be
   * adjusted before the Device gets announced. Do not use this method to change values when the device is in use!
   * If a getter or setter is defined the value must be undefined The version number must also be undefined.
   */
  init(value) {
    if (value === void 0) {
      throw new InternalError(`Can not initialize fixed attribute "${this.name}" with undefined value.`);
    }
    this.validateWithSchema(value);
    this.value = value;
  }
  /**
   * Add an internal listener that is called when the value of the attribute changes. The listener is called with the
   * new value and the version number.
   */
  addValueChangeListener(_listener) {
  }
  /**
   * Remove an internal listener.
   */
  removeValueChangeListener(_listener) {
  }
  /**
   * Add an external listener that is called when the value of the attribute changes. The listener is called with the
   * new value and the old value.
   */
  addValueSetListener(_listener) {
  }
  /**
   * Add an external listener that is called when the value of the attribute changes. The listener is called with the
   * new value and the old value. This method is a convenient alias for addValueSetListener.
   */
  subscribe(_listener) {
  }
  /**
   * Remove an external listener.
   */
  removeValueSetListener(_listener) {
  }
}
class AttributeServer extends FixedAttributeServer {
  constructor(id, name, schema, isWritable, isSubscribable, requiresTimedInteraction, defaultValue, getClusterDataVersion, increaseClusterDataVersion, getter, setter, validator) {
    if (isWritable && (getter === void 0 || setter === void 0) && !(getter === void 0 && setter === void 0)) {
      throw new ImplementationError(
        `Getter and setter must be implemented together for writeable attribute "${name}".`
      );
    }
    super(
      id,
      name,
      schema,
      isWritable,
      isSubscribable,
      requiresTimedInteraction,
      defaultValue,
      getClusterDataVersion,
      getter
    );
    this.increaseClusterDataVersion = increaseClusterDataVersion;
    this.isFixed = false;
    this.valueChangeListeners = new Array();
    this.valueSetListeners = new Array();
    if (setter === void 0) {
      this.setter = (value) => {
        const oldValue = this.value;
        this.value = value;
        return !isDeepEqual(value, oldValue);
      };
    } else {
      this.setter = setter;
    }
    this.validator = (value, session, endpoint) => {
      this.validateWithSchema(value);
      if (validator !== void 0) {
        validator(value, session, endpoint);
      }
    };
  }
  /**
   * Initialize the value of the attribute, used when a persisted value is set initially or when values needs to be
   * adjusted before the Device gets announced. Do not use this method to change values when the device is in use!
   */
  init(value) {
    if (value === void 0) {
      value = this.getter(void 0, this.endpoint);
    }
    if (value === void 0) {
      throw new InternalError(`Can not initialize attribute "${this.name}" with undefined value.`);
    }
    this.validator(value, void 0, this.endpoint);
    this.value = value;
  }
  /**
   * Set the value of the attribute. This method is used by the Interaction model to write the value of the attribute
   * and includes the ACL check. It should not be used locally in the code!
   * If a setter is defined this setter method is called to store the value.
   * Listeners are called when the value changes (internal listeners) or in any case (external listeners).
   */
  set(value, session) {
    if (!this.isWritable) {
      throw new StatusResponseError(`Attribute "${this.name}" is not writable.`, StatusCode.UnsupportedWrite);
    }
    this.setRemote(value, session);
  }
  /**
   * Method that contains the logic to set a value "from remote" (e.g. from a client).
   */
  setRemote(value, session) {
    this.processSet(value, session);
    this.value = value;
  }
  /**
   * Set the value of the attribute locally. This method should be used locally in the code and does not include the
   * ACL check.
   * If a setter is defined this setter method is called to validate and store the value.
   * Else if a validator is defined the value is validated before it is stored.
   * Listeners are called when the value changes (internal listeners) or in any case (external listeners).
   */
  setLocal(value) {
    this.processSet(value, void 0);
    this.value = value;
  }
  /**
   * Helper Method to process the set of a value in a generic way. This method is used internally.
   */
  processSet(value, session) {
    this.validator(value, session, this.endpoint);
    const oldValue = this.getter(session, this.endpoint);
    const valueChanged = this.setter(value, session, this.endpoint);
    this.handleVersionAndTriggerListeners(value, oldValue, valueChanged);
  }
  /**
   * Helper Method to handle needed version increases and trigger the relevant listeners. This method is used
   * internally.
   */
  handleVersionAndTriggerListeners(value, oldValue, considerVersionChanged) {
    if (considerVersionChanged) {
      const version = this.increaseClusterDataVersion();
      this.valueChangeListeners.forEach((listener) => listener(value, version));
    }
    if (oldValue !== void 0) {
      this.valueSetListeners.forEach((listener) => listener(value, oldValue));
    }
  }
  /**
   * When the value is handled by getter or setter methods and is changed by other processes this method can be used
   * to notify the attribute server that the value has changed. This will increase the version number and trigger the
   * listeners.
   * ACL checks needs to be performed before calling this method.
   */
  updated(session) {
    const oldValue = this.value ?? this.defaultValue;
    this.value = tryCatch(
      () => this.get(session, false),
      NoAssociatedFabricError,
      // Handle potential error cases where the session does not have a fabric assigned.
      this.value ?? this.defaultValue
    );
    this.handleVersionAndTriggerListeners(this.value, oldValue, true);
  }
  /**
   * When the value is handled by getter or setter methods and is changed by other processes and no session from the
   * originating process is known this method can be used to notify the attribute server that the value has changed.
   * This will increase the version number and trigger the listeners.
   * ACL checks needs to be performed before calling this method.
   */
  updatedLocal() {
    const oldValue = this.value ?? this.defaultValue;
    this.value = this.getLocal();
    this.handleVersionAndTriggerListeners(this.value, oldValue, true);
  }
  /**
   * Add an internal listener that is called when the value of the attribute changes. The listener is called with the
   * new value and the version number.
   */
  addValueChangeListener(listener) {
    this.valueChangeListeners.push(listener);
  }
  /**
   * Remove an internal listener.
   */
  removeValueChangeListener(listener) {
    const entryIndex = this.valueChangeListeners.indexOf(listener);
    if (entryIndex !== -1) {
      this.valueChangeListeners.splice(entryIndex, 1);
    }
  }
  /**
   * Add an external listener that is called when the value of the attribute changes. The listener is called with the
   * new value and the old value.
   */
  addValueSetListener(listener) {
    this.valueSetListeners.push(listener);
  }
  /**
   * Add an external listener that is called when the value of the attribute changes. The listener is called with the
   * new value and the old value. This method is a convenient alias for addValueSetListener.
   */
  subscribe(listener) {
    this.addValueSetListener(listener);
  }
  /**
   * Remove an external listener.
   */
  removeValueSetListener(listener) {
    const entryIndex = this.valueSetListeners.indexOf(listener);
    if (entryIndex !== -1) {
      this.valueSetListeners.splice(entryIndex, 1);
    }
  }
}
class FabricScopedAttributeServer extends AttributeServer {
  constructor(id, name, schema, isWritable, isSubscribable, requiresTimedInteraction, defaultValue, cluster, getClusterDataVersion, increaseClusterDataVersion, getter, setter, validator) {
    if (isWritable && (getter === void 0 || setter === void 0) && !(getter === void 0 && setter === void 0)) {
      throw new ImplementationError(
        `Getter and setter must be implemented together writeable fabric scoped attribute "${name}".`
      );
    }
    let isCustomGetter = false;
    if (getter === void 0) {
      getter = (session, _endpoint, isFabricFiltered) => {
        if (session === void 0)
          throw new FabricScopeError(`Session is required for fabric scoped attribute ${name}`);
        if (isFabricFiltered === true) {
          assertSecureSession(session);
          return this.getLocalForFabric(session.getAssociatedFabric());
        } else {
          const fabrics = session.getContext().getFabrics();
          const values = new Array();
          for (const fabric of fabrics) {
            const value = this.getLocalForFabric(fabric);
            if (!Array.isArray(value)) {
              throw new FabricScopeError(
                `Fabric scoped attribute "${name}" can only be read for all fabrics if they are arrays.`
              );
            }
            values.push(...value);
          }
          return values;
        }
      };
    } else {
      isCustomGetter = true;
    }
    let isCustomSetter = false;
    if (setter === void 0) {
      setter = (value, session) => {
        if (session === void 0)
          throw new FabricScopeError(`Session is required for fabric scoped attribute "${name}".`);
        assertSecureSession(session);
        const fabric = session.getAssociatedFabric();
        const oldData = fabric.getScopedClusterDataValue(this.cluster, this.name);
        const oldValue = oldData?.value ?? this.defaultValue;
        if (!isDeepEqual(value, oldValue)) {
          fabric.setScopedClusterDataValue(this.cluster, this.name, { value });
          return true;
        }
        return false;
      };
    } else {
      isCustomSetter = true;
    }
    super(
      id,
      name,
      schema,
      isWritable,
      isSubscribable,
      requiresTimedInteraction,
      defaultValue,
      getClusterDataVersion,
      increaseClusterDataVersion,
      getter,
      setter,
      validator
    );
    this.cluster = cluster;
    this.isCustomGetter = isCustomGetter;
    this.isCustomSetter = isCustomSetter;
  }
  /**
   * Initialize the attribute with a value. Because the value is stored on fabric level this method only initializes
   * the version number.
   */
  init(value) {
    if (value !== void 0) {
      throw new InternalError(`Can not initialize fabric scoped attribute "${this.name}" with a value.`);
    }
  }
  /**
   * Method that contains the logic to set a value "from remote" (e.g. from a client). For Fabric scoped attributes
   * we need to inject the fabric index into the value.
   */
  setRemote(value, session) {
    value = this.schema.injectField(
      value,
      Globals.FabricIndex.id,
      session.getAssociatedFabric().fabricIndex,
      (existingFieldIndex) => existingFieldIndex === void 0
    );
    super.setRemote(value, session);
  }
  /**
   * Set Local is not allowed for fabric scoped attributes. Use setLocalForFabric instead.
   */
  setLocal(_value) {
    throw new FabricScopeError(
      `Fabric scoped attribute "${this.name}" can only be set locally by providing a Fabric. Use setLocalForFabric instead.`
    );
  }
  /**
   * Set the value of the attribute locally for a fabric. This method should be used locally in the code and does not
   * include the ACL check.
   * If a setter is defined this method can not be used!
   * If a validator is defined the value is validated before it is stored.
   * Listeners are called when the value changes (internal listeners) or in any case (external listeners).
   */
  setLocalForFabric(value, fabric) {
    if (this.isCustomSetter) {
      throw new FabricScopeError(
        `Fabric scoped attribute "${this.name}" can not be set locally when a custom setter is defined.`
      );
    }
    this.validator(value, void 0, this.endpoint);
    const oldValue = this.getLocalForFabric(fabric);
    const valueChanged = !isDeepEqual(value, oldValue);
    if (valueChanged) {
      fabric.setScopedClusterDataValue(this.cluster, this.name, { value });
    }
    this.handleVersionAndTriggerListeners(value, oldValue, valueChanged);
  }
  /**
   * When the value is handled by getter or setter methods and is changed by other processes and no session from the
   * originating process is known this method can be used to notify the attribute server that the value has changed.
   * This will increase the version number and trigger the listeners.
   * ACL checks needs to be performed before calling this method.
   */
  updatedLocalForFabric(fabric) {
    const oldValue = this.value ?? this.defaultValue;
    this.value = tryCatch(
      () => this.getLocalForFabric(fabric),
      FabricScopeError,
      // Handle potential error cases where a custom getter is used.
      this.value ?? this.defaultValue
    );
    this.handleVersionAndTriggerListeners(this.value, oldValue, true);
  }
  /**
   * Get the value of the attribute locally for a special Fabric. This method should be used locally in the code and
   * does not include the ACL check.
   * If a getter is defined this method returns an error and the value should be retrieved directly internally.
   */
  getLocalForFabric(fabric) {
    if (this.isCustomGetter) {
      throw new FabricScopeError(
        `Fabric scoped attribute "${this.name}" can not be read locally when a custom getter is defined.`
      );
    }
    const data = fabric.getScopedClusterDataValue(this.cluster, this.name);
    return data?.value ?? this.defaultValue;
  }
}
export {
  AttributeServer,
  BaseAttributeServer,
  FabricScopeError,
  FabricScopedAttributeServer,
  FixedAttributeServer,
  createAttributeServer
};
