// @project-chip/matter.js/dist/esm/cluster/server/CommandServer.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger } from "../../log/Logger.js";
import { Globals } from "../../model/index.js";
import { StatusCode } from "../../protocol/interaction/StatusCode.js";
const logger = Logger.get("CommandServer");
class CommandServer {
  constructor(invokeId, responseId, name, requestSchema, responseSchema, requiresTimedInteraction, handler) {
    this.invokeId = invokeId;
    this.responseId = responseId;
    this.name = name;
    this.requestSchema = requestSchema;
    this.responseSchema = responseSchema;
    this.requiresTimedInteraction = requiresTimedInteraction;
    this.handler = handler;
  }
  async invoke(session, args, message, endpoint) {
    let request = this.requestSchema.decodeTlv(args);
    if (session.isSecure()) {
      const fabric = session.getFabric();
      if (fabric) {
        request = this.requestSchema.injectField(
          request,
          Globals.FabricIndex.id,
          session.getAssociatedFabric().fabricIndex,
          (existingFieldIndex) => existingFieldIndex === void 0
        );
      }
    }
    this.requestSchema.validate(request);
    logger.debug(`Invoke ${this.name} with data ${Logger.toJSON(request)}`);
    const response = await this.handler(request, session, message, endpoint);
    logger.debug(`Invoke ${this.name} response : ${Logger.toJSON(response)}`);
    return {
      code: StatusCode.Success,
      responseId: this.responseId,
      response: this.responseSchema.encodeTlv(response)
    };
  }
}
export {
  CommandServer
};
