// @project-chip/matter.js/dist/esm/cluster/server/EventServer.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { InternalError } from "../../common/MatterError.js";
import { Time } from "../../time/Time.js";
class EventServer {
  constructor(id, clusterId, name, schema, priority) {
    this.id = id;
    this.clusterId = clusterId;
    this.name = name;
    this.schema = schema;
    this.priority = priority;
    this.eventList = new Array();
    this.listeners = new Array();
  }
  assignToEndpoint(endpoint) {
    this.endpoint = endpoint;
  }
  // TODO Try to get rid of that late binding and simply things again
  //      potentially with refactoring out MatterDevice and MatterController
  bindToEventHandler(eventHandler) {
    this.eventHandler = eventHandler;
    for (const event of this.eventList) {
      const finalEvent = this.eventHandler.pushEvent(event);
      this.listeners.forEach((listener) => listener(finalEvent));
    }
    this.eventList = [];
  }
  triggerEvent(data) {
    if (this.endpoint === void 0 || this.endpoint.id === void 0) {
      throw new InternalError("Endpoint not assigned");
    }
    const event = {
      eventId: this.id,
      clusterId: this.clusterId,
      endpointId: this.endpoint.id,
      epochTimestamp: Time.nowMs(),
      priority: this.priority,
      data
    };
    if (this.eventHandler === void 0) {
      this.eventList.push(event);
    } else {
      const finalEvent = this.eventHandler.pushEvent(event);
      this.listeners.forEach((listener) => listener(finalEvent));
    }
  }
  addListener(listener) {
    this.listeners.push(listener);
  }
  removeListener(listener) {
    const entryIndex = this.listeners.indexOf(listener);
    if (entryIndex !== -1) {
      this.listeners.splice(entryIndex, 1);
    }
  }
}
export {
  EventServer
};
