// @project-chip/matter.js/dist/esm/cluster/server/GroupsServer.js
/**
 * @license
 * Copyright 2022 The node-matter Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { StatusCode } from "../../protocol/interaction/StatusCode.js";
import { assertSecureSession } from "../../session/SecureSession.js";
import { GroupsCluster } from "../definitions/GroupsCluster.js";
import { IdentifyCluster } from "../definitions/IdentifyCluster.js";
import { ClusterServer } from "./ClusterServer.js";
import { ScenesManager } from "./ScenesServer.js";
class GroupsManager {
  static setGroup(fabric, endpointId, groupId, groupName) {
    let endpointGroups = fabric.getScopedClusterDataValue(
      GroupsCluster,
      endpointId.toString()
    );
    if (endpointGroups === void 0) {
      endpointGroups = /* @__PURE__ */ new Map();
    }
    endpointGroups.set(groupId, groupName || "");
    fabric.setScopedClusterDataValue(GroupsCluster, endpointId.toString(), endpointGroups);
  }
  static getGroupName(fabric, endpointId, groupId) {
    const endpointGroups = fabric.getScopedClusterDataValue(
      GroupsCluster,
      endpointId.toString()
    );
    return endpointGroups?.get(groupId);
  }
  static hasGroup(fabric, endpointId, groupId) {
    const endpointGroups = fabric.getScopedClusterDataValue(
      GroupsCluster,
      endpointId.toString()
    );
    return endpointGroups?.has(groupId) ?? false;
  }
  static getGroups(fabric, endpointId) {
    return fabric.getScopedClusterDataValue(GroupsCluster, endpointId.toString()) ?? /* @__PURE__ */ new Map();
  }
  static removeGroup(fabric, endpointId, groupId) {
    const endpointGroups = fabric.getScopedClusterDataValue(
      GroupsCluster,
      endpointId.toString()
    );
    if (endpointGroups !== void 0) {
      if (endpointGroups.delete(groupId)) {
        fabric.persist();
        return true;
      }
    }
    return false;
  }
  static removeAllGroups(fabric, endpointId) {
    fabric.deleteScopedClusterDataValue(GroupsCluster, endpointId.toString());
  }
}
const GroupsClusterHandler = () => {
  const addGroupLogic = (groupId, groupName, fabric, endpointId) => {
    if (groupId < 1) {
      return { status: StatusCode.ConstraintError, groupId };
    }
    if (groupName.length > 16) {
      return { status: StatusCode.ConstraintError, groupId };
    }
    GroupsManager.setGroup(fabric, endpointId, groupId, groupName);
    return { status: StatusCode.Success, groupId };
  };
  return {
    addGroup: async ({ request: { groupId, groupName }, session, endpoint }) => {
      assertSecureSession(session);
      return addGroupLogic(groupId, groupName, session.getAssociatedFabric(), endpoint.getId());
    },
    viewGroup: async ({ request: { groupId }, session, endpoint }) => {
      if (groupId < 1) {
        return { status: StatusCode.ConstraintError, groupId, groupName: "" };
      }
      assertSecureSession(session);
      const groupName = GroupsManager.getGroupName(session.getAssociatedFabric(), endpoint.getId(), groupId);
      if (groupName !== void 0) {
        return { status: StatusCode.Success, groupId, groupName };
      }
      return { status: StatusCode.NotFound, groupId, groupName: "" };
    },
    getGroupMembership: async ({ request: { groupList }, session, endpoint }) => {
      assertSecureSession(session);
      const endpointGroups = GroupsManager.getGroups(session.getAssociatedFabric(), endpoint.getId());
      const fabricGroupsList = Array.from(endpointGroups.keys());
      const capacity = fabricGroupsList.length < 255 ? 254 - fabricGroupsList.length : 0;
      if (groupList.length === 0) {
        return { capacity, groupList: fabricGroupsList };
      }
      const filteredGroupsList = groupList.filter((groupId) => endpointGroups.get(groupId));
      if (filteredGroupsList.length === 0) {
        return { capacity, groupList: [] };
      }
      return { capacity, groupList: filteredGroupsList };
    },
    removeGroup: async ({ request: { groupId }, session, endpoint }) => {
      if (groupId < 1) {
        return { status: StatusCode.ConstraintError, groupId };
      }
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      if (GroupsManager.removeGroup(session.getAssociatedFabric(), endpoint.getId(), groupId)) {
        ScenesManager.removeAllScenesForGroup(fabric, endpoint.getId(), groupId);
        return { status: StatusCode.Success, groupId };
      }
      return { status: StatusCode.NotFound, groupId };
    },
    removeAllGroups: async ({ session, endpoint }) => {
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      GroupsManager.removeAllGroups(fabric, endpoint.getId());
      ScenesManager.removeAllNonGlobalScenesForEndpoint(fabric, endpoint.getId());
      return;
    },
    addGroupIfIdentifying: async ({ request: { groupId, groupName }, session, endpoint }) => {
      const identifyCluster = endpoint.getClusterServer(IdentifyCluster);
      if (identifyCluster) {
        if (identifyCluster.attributes.identifyTime.getLocal() > 0) {
          assertSecureSession(session);
          addGroupLogic(groupId, groupName, session.getAssociatedFabric(), endpoint.getId());
        }
      }
    }
  };
};
const createDefaultGroupsClusterServer = () => ClusterServer(
  GroupsCluster,
  {
    nameSupport: {
      nameSupport: true
    }
  },
  GroupsClusterHandler()
);
export {
  GroupsClusterHandler,
  GroupsManager,
  createDefaultGroupsClusterServer
};
