// @project-chip/matter.js/dist/esm/cluster/server/OperationalCredentialsServer.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterFabricConflictError } from "../../common/FailSafeManager.js";
import { MatterFlowError } from "../../common/MatterError.js";
import { tryCatch } from "../../common/TryCatchHandler.js";
import { Crypto } from "../../crypto/Crypto.js";
import { PrivateKey } from "../../crypto/Key.js";
import { FabricIndex } from "../../datatype/FabricIndex.js";
import { FabricTableFullError } from "../../fabric/FabricManager.js";
import { Logger } from "../../log/Logger.js";
import { StatusCode, StatusResponseError } from "../../protocol/interaction/StatusCode.js";
import { NoAssociatedFabricError, assertSecureSession } from "../../session/SecureSession.js";
import { TlvUInt32 } from "../../tlv/TlvNumber.js";
import { TlvField, TlvObject, TlvOptionalField } from "../../tlv/TlvObject.js";
import { TlvByteString } from "../../tlv/TlvString.js";
import { BasicInformation } from "../definitions/BasicInformationCluster.js";
import { OperationalCredentials } from "../definitions/OperationalCredentialsCluster.js";
const logger = Logger.get("OperationalCredentialsServer");
const TlvAttestation = TlvObject({
  declaration: TlvField(1, TlvByteString),
  attestationNonce: TlvField(2, TlvByteString.bound({ length: 32 })),
  timestamp: TlvField(3, TlvUInt32),
  // TODO: check actual max length in specs
  firmwareInfo: TlvOptionalField(4, TlvByteString)
});
const TlvCertSigningRequest = TlvObject({
  certSigningRequest: TlvField(1, TlvByteString),
  csrNonce: TlvField(2, TlvByteString.bound({ length: 32 })),
  vendorReserved1: TlvOptionalField(3, TlvByteString),
  vendorReserved2: TlvOptionalField(4, TlvByteString),
  vendorReserved3: TlvOptionalField(5, TlvByteString)
});
function signWithDeviceKey(conf, session, data) {
  return Crypto.sign(PrivateKey(conf.devicePrivateKey), [data, session.getAttestationChallengeKey()]);
}
const OperationalCredentialsClusterHandler = (conf) => ({
  attestationRequest: async ({ request: { attestationNonce }, session }) => {
    assertSecureSession(session);
    const elements = TlvAttestation.encode({
      declaration: conf.certificationDeclaration,
      attestationNonce,
      timestamp: 0
    });
    return { attestationElements: elements, attestationSignature: signWithDeviceKey(conf, session, elements) };
  },
  csrRequest: async ({ request: { csrNonce, isForUpdateNoc }, session }) => {
    assertSecureSession(session);
    if (isForUpdateNoc && session.isPase()) {
      throw new StatusResponseError(
        "csrRequest for UpdateNoc received on a PASE session.",
        StatusCode.InvalidCommand
      );
    }
    const device = session.getContext();
    device.assertFailSafeArmed("csrRequest received while failsafe is not armed.");
    const failSafeContext = device.getFailSafeContext();
    if (failSafeContext.fabricIndex !== void 0) {
      throw new StatusResponseError(
        `csrRequest received after ${failSafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"} already invoked.`,
        StatusCode.ConstraintError
      );
    }
    const certSigningRequest = failSafeContext.createCertificateSigningRequest(
      isForUpdateNoc ?? false,
      session.getId()
    );
    const nocsrElements = TlvCertSigningRequest.encode({ certSigningRequest, csrNonce });
    return { nocsrElements, attestationSignature: signWithDeviceKey(conf, session, nocsrElements) };
  },
  certificateChainRequest: async ({ request: { certificateType } }) => {
    switch (certificateType) {
      case OperationalCredentials.CertificateChainType.DacCertificate:
        return { certificate: conf.deviceCertificate };
      case OperationalCredentials.CertificateChainType.PaiCertificate:
        return { certificate: conf.deviceIntermediateCertificate };
      default:
        throw new MatterFlowError(`Unsupported certificate type: ${certificateType}`);
    }
  },
  addNoc: async ({
    request: { nocValue, icacValue, ipkValue, caseAdminSubject, adminVendorId },
    attributes: { nocs, commissionedFabrics, fabrics, trustedRootCertificates, supportedFabrics },
    session
  }) => {
    if (!session.isSecure())
      throw new MatterFlowError("addOperationalCert should be called on a secure session.");
    const device = session.getContext();
    device.assertFailSafeArmed("addNoc received while failsafe is not armed.");
    const failSafeContext = device.getFailSafeContext();
    if (failSafeContext.fabricIndex !== void 0) {
      throw new StatusResponseError(
        `addNoc received after ${failSafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"} already invoked.`,
        StatusCode.ConstraintError
      );
    }
    if (!failSafeContext.fabricBuilder.hasRootCert()) {
      return {
        statusCode: OperationalCredentials.NodeOperationalCertStatus.InvalidNoc,
        debugText: "Root certificate not found."
      };
    }
    if (failSafeContext.csrSessionId !== session.getId()) {
      return {
        statusCode: OperationalCredentials.NodeOperationalCertStatus.MissingCsr,
        debugText: "CSR not found in failsafe context."
      };
    }
    if (failSafeContext.forUpdateNoc) {
      throw new StatusResponseError(
        `addNoc received after csr request was invoked for UpdateNOC.`,
        StatusCode.ConstraintError
      );
    }
    if (device.getFabrics().length === supportedFabrics.getLocal()) {
      return {
        statusCode: OperationalCredentials.NodeOperationalCertStatus.TableFull,
        debugText: `No more fabrics can be added because limit ${supportedFabrics.getLocal()} reached.`
      };
    }
    let fabric;
    try {
      fabric = await failSafeContext.buildFabric({
        nocValue,
        icacValue,
        adminVendorId,
        ipkValue,
        caseAdminSubject
      });
    } catch (error) {
      if (error instanceof MatterFabricConflictError) {
        return {
          statusCode: OperationalCredentials.NodeOperationalCertStatus.FabricConflict,
          debugText: error.message
        };
      } else if (error instanceof FabricTableFullError) {
        return {
          statusCode: OperationalCredentials.NodeOperationalCertStatus.TableFull,
          debugText: error.message
        };
      } else {
        throw error;
      }
    }
    await device.addFabric(fabric);
    assertSecureSession(session);
    if (session.isPase()) {
      logger.debug(`Add Fabric with index ${fabric.fabricIndex} to PASE session ${session.name}.`);
      session.addAssociatedFabric(fabric);
    }
    nocs.updated(session);
    commissionedFabrics.updated(session);
    fabrics.updated(session);
    trustedRootCertificates.updated(session);
    logger.info(`addNoc success, adminVendorId ${adminVendorId}, caseAdminSubject ${caseAdminSubject}`);
    return { statusCode: OperationalCredentials.NodeOperationalCertStatus.Ok, fabricIndex: fabric.fabricIndex };
  },
  fabricsAttributeGetter: ({ session, isFabricFiltered }) => {
    if (session === void 0 || !session.isSecure())
      return [];
    const fabrics = isFabricFiltered ? [session.getAssociatedFabric()] : session.getContext().getFabrics();
    return fabrics.map((fabric) => ({
      fabricId: fabric.fabricId,
      label: fabric.label,
      nodeId: fabric.nodeId,
      rootPublicKey: fabric.rootPublicKey,
      vendorId: fabric.rootVendorId,
      fabricIndex: fabric.fabricIndex
    }));
  },
  nocsAttributeGetter: ({ session, isFabricFiltered }) => {
    if (session === void 0 || !session.isSecure())
      return [];
    const fabrics = isFabricFiltered ? [session.getAssociatedFabric()] : session.getContext().getFabrics();
    return fabrics.map((fabric) => ({
      noc: fabric.operationalCert,
      icac: fabric.intermediateCACert ?? null,
      fabricIndex: fabric.fabricIndex
    }));
  },
  commissionedFabricsAttributeGetter: ({ session }) => {
    if (session === void 0 || !session.isSecure())
      return 0;
    return session.getContext().getFabrics().length;
  },
  trustedRootCertificatesAttributeGetter: ({ session, isFabricFiltered }) => {
    if (session === void 0 || !session.isSecure())
      return [];
    const fabrics = isFabricFiltered ? [session.getAssociatedFabric()] : session.getContext().getFabrics();
    return fabrics.map((fabric) => fabric.rootCert);
  },
  currentFabricIndexAttributeGetter: ({ session }) => {
    if (session === void 0 || !session.isSecure())
      return FabricIndex.NO_FABRIC;
    assertSecureSession(session);
    return session.getFabric()?.fabricIndex ?? FabricIndex.NO_FABRIC;
  },
  updateNoc: async ({ request: { nocValue, icacValue }, attributes: { nocs, fabrics }, session }) => {
    assertSecureSession(session);
    tryCatch(
      () => session.getAssociatedFabric(),
      NoAssociatedFabricError,
      () => {
        throw new StatusResponseError("No associated fabric existing", StatusCode.UnsupportedAccess);
      }
    );
    const device = session.getContext();
    device.assertFailSafeArmed("updateNoc received while failsafe is not armed.");
    const failSafeContext = device.getFailSafeContext();
    if (failSafeContext.fabricIndex !== void 0) {
      throw new StatusResponseError(
        `updateNoc received after ${failSafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"} already invoked.`,
        StatusCode.ConstraintError
      );
    }
    if (failSafeContext.forUpdateNoc) {
      throw new StatusResponseError(
        `addNoc received after csr request was invoked for UpdateNOC.`,
        StatusCode.ConstraintError
      );
    }
    if (failSafeContext.fabricBuilder.hasRootCert()) {
      throw new StatusResponseError(
        "Trusted root certificate added in this session which is now allowed for UpdateNOC.",
        StatusCode.ConstraintError
      );
    }
    if (!failSafeContext.forUpdateNoc) {
      throw new StatusResponseError("csrRequest not invoked for UpdateNOC.", StatusCode.ConstraintError);
    }
    if (session.getAssociatedFabric().fabricIndex !== failSafeContext.associatedFabric?.fabricIndex) {
      throw new StatusResponseError(
        "Fabric of this session and the failsafe context do not match.",
        StatusCode.ConstraintError
      );
    }
    const updateFabric = await failSafeContext.buildUpdatedFabric(nocValue, icacValue);
    device.updateFabric(updateFabric);
    nocs.updated(session);
    fabrics.updated(session);
    return {
      statusCode: OperationalCredentials.NodeOperationalCertStatus.Ok,
      fabricIndex: updateFabric.fabricIndex
    };
  },
  updateFabricLabel: async ({ request: { label }, attributes: { fabrics }, session }) => {
    assertSecureSession(session, "updateOperationalCert should be called on a secure session.");
    const fabric = tryCatch(
      () => session.getAssociatedFabric(),
      NoAssociatedFabricError,
      () => {
        throw new StatusResponseError("No associated fabric existing", StatusCode.UnsupportedAccess);
      }
    );
    const currentFabricIndex = fabric.fabricIndex;
    const device = session.getContext();
    const conflictingLabelFabric = device.getFabrics().find((f) => f.label === label && f.fabricIndex !== currentFabricIndex);
    if (conflictingLabelFabric !== void 0) {
      return {
        statusCode: OperationalCredentials.NodeOperationalCertStatus.LabelConflict,
        debugText: `Label ${label} already used by fabric ${conflictingLabelFabric.fabricIndex}`
      };
    }
    fabric.setLabel(label);
    fabrics.updated(session);
    return { statusCode: OperationalCredentials.NodeOperationalCertStatus.Ok, fabricIndex: fabric.fabricIndex };
  },
  removeFabric: async ({
    request: { fabricIndex },
    attributes: { nocs, commissionedFabrics, fabrics, trustedRootCertificates },
    session,
    endpoint
  }) => {
    const device = session.getContext();
    const fabric = device.getFabricByIndex(fabricIndex);
    if (fabric === void 0) {
      return {
        statusCode: OperationalCredentials.NodeOperationalCertStatus.InvalidFabricIndex,
        debugText: `Fabric ${fabricIndex} not found`
      };
    }
    const basicInformationCluster = endpoint.getClusterServer(BasicInformation.Cluster);
    basicInformationCluster?.triggerLeaveEvent?.({ fabricIndex });
    assertSecureSession(session);
    await fabric.remove(session.getId());
    nocs.updated(session);
    commissionedFabrics.updated(session);
    fabrics.updated(session);
    trustedRootCertificates.updated(session);
    return {
      statusCode: OperationalCredentials.NodeOperationalCertStatus.Ok,
      fabricIndex
    };
  },
  addTrustedRootCertificate: async ({ request: { rootCaCertificate }, session }) => {
    const device = session.getContext();
    device.assertFailSafeArmed("addTrustedRootCertificate received while failsafe is not armed.");
    const failSafeContext = device.getFailSafeContext();
    if (failSafeContext.fabricBuilder.hasRootCert()) {
      throw new StatusResponseError(
        "Trusted root certificate already added in this FailSafe context.",
        StatusCode.ConstraintError
      );
    }
    if (failSafeContext.fabricIndex !== void 0) {
      throw new StatusResponseError(
        `Can not add trusted root certificates after ${failSafeContext.forUpdateNoc ? "UpdateNOC" : "AddNOC"}.`,
        StatusCode.ConstraintError
      );
    }
    session.getContext().getFailSafeContext().setRootCert(rootCaCertificate);
  }
});
export {
  OperationalCredentialsClusterHandler,
  TlvAttestation,
  TlvCertSigningRequest
};
