// @project-chip/matter.js/dist/esm/cluster/server/ScenesServer.js
/**
 * @license
 * Copyright 2022 The node-matter Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { InternalError } from "../../common/MatterError.js";
import { GroupId } from "../../datatype/GroupId.js";
import { StatusCode, StatusResponseError } from "../../protocol/interaction/StatusCode.js";
import { assertSecureSession } from "../../session/SecureSession.js";
import { ScenesCluster } from "../definitions/ScenesCluster.js";
import { ClusterServer } from "./ClusterServer.js";
import { asClusterServerInternal } from "./ClusterServerTypes.js";
import { GroupsManager } from "./GroupsServer.js";
class ScenesManager {
  static getEndpointScenes(fabric, endpointId) {
    return fabric.getScopedClusterDataValue(
      ScenesCluster,
      endpointId.toString()
    );
  }
  static setEndpointScenes(fabric, endpointId, endpointScenes) {
    fabric.setScopedClusterDataValue(ScenesCluster, endpointId.toString(), endpointScenes);
  }
  static setScenes(fabric, endpointId, sceneEntries) {
    let endpointScenes = ScenesManager.getEndpointScenes(fabric, endpointId);
    if (endpointScenes === void 0) {
      endpointScenes = /* @__PURE__ */ new Map();
    }
    for (const sceneEntry of sceneEntries) {
      const { scenesGroupId, sceneId } = sceneEntry;
      let scenesGroupIdMap = endpointScenes.get(scenesGroupId);
      if (scenesGroupIdMap === void 0) {
        scenesGroupIdMap = /* @__PURE__ */ new Map();
        endpointScenes.set(scenesGroupId, scenesGroupIdMap);
      }
      scenesGroupIdMap.set(sceneId, sceneEntry);
    }
    ScenesManager.setEndpointScenes(fabric, endpointId, endpointScenes);
  }
  static getSceneEntry(fabric, endpointId, groupId, sceneId) {
    return ScenesManager.getEndpointScenes(fabric, endpointId)?.get(groupId)?.get(sceneId);
  }
  static getAllScenes(fabric, endpointId, groupId) {
    const endpointScenes = ScenesManager.getEndpointScenes(fabric, endpointId);
    if (endpointScenes === void 0)
      return [];
    return Array.from(endpointScenes.get(groupId)?.values() ?? []);
  }
  static removeScene(fabric, endpointId, groupId, sceneId) {
    const endpointScenes = ScenesManager.getEndpointScenes(fabric, endpointId);
    if (endpointScenes !== void 0) {
      const groupScenes = endpointScenes.get(groupId);
      if (groupScenes !== void 0) {
        if (groupScenes.delete(sceneId)) {
          fabric.persist();
          return true;
        }
      }
    }
    return false;
  }
  static removeAllScenesForGroup(fabric, endpointId, groupId) {
    const endpointScenes = ScenesManager.getEndpointScenes(fabric, endpointId);
    if (endpointScenes !== void 0) {
      if (endpointScenes.delete(groupId)) {
        fabric.persist();
      }
    }
  }
  static removeAllNonGlobalScenesForEndpoint(fabric, endpointId) {
    const endpointScenes = ScenesManager.getEndpointScenes(fabric, endpointId);
    if (endpointScenes !== void 0) {
      endpointScenes.forEach((_groupScenes, groupId) => {
        if (groupId !== 0) {
          endpointScenes.delete(groupId);
        }
      });
      fabric.persist();
    }
  }
}
const ScenesClusterHandler = () => {
  const addSceneLogic = (endpointId, groupId, sceneId, sceneTransitionTime, sceneName, extensionFieldSets, transitionTime100ms, fabric) => {
    if (groupId !== 0 && !GroupsManager.hasGroup(fabric, endpointId, groupId)) {
      return { status: StatusCode.InvalidCommand, groupId, sceneId };
    }
    if (groupId < 1 || sceneId < 1) {
      return { status: StatusCode.ConstraintError, groupId, sceneId };
    }
    if (groupId > 65527) {
      return { status: StatusCode.ConstraintError, groupId, sceneId };
    }
    if (sceneName.length > 16) {
      return { status: StatusCode.ConstraintError, groupId, sceneId };
    }
    ScenesManager.setScenes(fabric, endpointId, [
      {
        scenesGroupId: groupId,
        sceneId,
        sceneName,
        sceneTransitionTime,
        extensionFieldSets,
        transitionTime100ms
      }
    ]);
    return { status: StatusCode.Success, groupId, sceneId };
  };
  return {
    addScene: async ({
      request: { groupId, sceneId, transitionTime, sceneName, extensionFieldSets },
      attributes: { sceneCount },
      session,
      endpoint
    }) => {
      assertSecureSession(session);
      const result = addSceneLogic(
        endpoint.getId(),
        groupId,
        sceneId,
        transitionTime,
        sceneName,
        extensionFieldSets,
        0,
        session.getAssociatedFabric()
      );
      if (result.status === StatusCode.Success) {
        sceneCount.updated(session);
      }
      return result;
    },
    viewScene: async ({ request: { groupId, sceneId }, session, endpoint }) => {
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      if (groupId !== 0 && !GroupsManager.hasGroup(fabric, endpoint.getId(), groupId)) {
        return { status: StatusCode.InvalidCommand, groupId, sceneId };
      }
      const sceneEntry = ScenesManager.getSceneEntry(fabric, endpoint.getId(), groupId, sceneId);
      if (sceneEntry === void 0) {
        return { status: StatusCode.NotFound, groupId, sceneId };
      }
      const { sceneName, sceneTransitionTime, extensionFieldSets } = sceneEntry;
      return {
        status: StatusCode.Success,
        groupId,
        sceneId,
        sceneName,
        transitionTime: sceneTransitionTime,
        extensionFieldSets
      };
    },
    removeScene: async ({ request: { groupId, sceneId }, attributes: { sceneCount }, session, endpoint }) => {
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      if (groupId !== 0 && !GroupsManager.hasGroup(fabric, endpoint.getId(), groupId)) {
        return { status: StatusCode.InvalidCommand, groupId, sceneId };
      }
      if (ScenesManager.removeScene(fabric, endpoint.getId(), groupId, sceneId)) {
        sceneCount.updated(session);
        return { status: StatusCode.Success, groupId, sceneId };
      }
      return { status: StatusCode.NotFound, groupId, sceneId };
    },
    removeAllScenes: async ({ request: { groupId }, session, endpoint }) => {
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      if (groupId !== 0 && !GroupsManager.hasGroup(fabric, endpoint.getId(), groupId)) {
        return { status: StatusCode.InvalidCommand, groupId };
      }
      ScenesManager.removeAllScenesForGroup(fabric, endpoint.getId(), groupId);
      return { status: StatusCode.Success, groupId };
    },
    storeScene: async ({
      request: { groupId, sceneId },
      session,
      attributes: { currentScene, currentGroup, sceneValid },
      endpoint
    }) => {
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      if (groupId !== 0 && !GroupsManager.hasGroup(fabric, endpoint.getId(), groupId)) {
        return { status: StatusCode.InvalidCommand, groupId, sceneId };
      }
      const extensionFieldSets = new Array();
      endpoint.getAllClusterServers().forEach((cluster) => {
        const attributeValueList = asClusterServerInternal(cluster)._getSceneExtensionFieldSets();
        if (attributeValueList.length) {
          extensionFieldSets.push({ clusterId: cluster.id, attributeValueList });
        }
      });
      const newSceneEntry = {
        scenesGroupId: groupId,
        sceneId,
        sceneName: "",
        sceneTransitionTime: 0,
        extensionFieldSets,
        transitionTime100ms: 0
      };
      const existingSceneEntry = ScenesManager.getSceneEntry(fabric, endpoint.getId(), groupId, sceneId);
      if (existingSceneEntry !== void 0) {
        newSceneEntry.sceneName = existingSceneEntry.sceneName ?? "";
        newSceneEntry.sceneTransitionTime = existingSceneEntry.sceneTransitionTime ?? 0;
      }
      ScenesManager.setScenes(fabric, endpoint.getId(), [newSceneEntry]);
      currentScene.setLocal(sceneId);
      currentGroup.setLocal(groupId);
      sceneValid.updated(session);
      return { status: StatusCode.Success, groupId, sceneId };
    },
    recallScene: async ({
      request: { groupId, sceneId, transitionTime },
      attributes: { currentScene, currentGroup, sceneValid },
      session,
      endpoint
    }) => {
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      if (groupId !== 0 && !GroupsManager.hasGroup(fabric, endpoint.getId(), groupId)) {
        throw new StatusResponseError(
          `Group ${groupId} does not exist on this endpoint`,
          StatusCode.InvalidCommand
        );
      }
      const existingSceneEntry = ScenesManager.getSceneEntry(fabric, endpoint.getId(), groupId, sceneId);
      if (existingSceneEntry === void 0) {
        throw new StatusResponseError(
          `Scene ${sceneId} does not exist for group ${groupId}`,
          StatusCode.NotFound
        );
      }
      const usedTransitionTime = transitionTime ?? existingSceneEntry.sceneTransitionTime + existingSceneEntry.transitionTime100ms / 10;
      existingSceneEntry.extensionFieldSets.forEach((clusterData) => {
        const { clusterId, attributeValueList } = clusterData;
        const cluster = endpoint.getClusterServerById(clusterId);
        if (cluster !== void 0) {
          asClusterServerInternal(cluster)._setSceneExtensionFieldSets(
            attributeValueList,
            usedTransitionTime
          );
        }
      });
      currentScene.setLocal(sceneId);
      currentGroup.setLocal(groupId);
      sceneValid.updated(session);
    },
    getSceneMembership: async ({ request: { groupId }, session, endpoint }) => {
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      const endpointScenes = ScenesManager.getAllScenes(fabric, endpoint.getId(), groupId);
      const capacity = endpointScenes.length < 255 ? 254 - endpointScenes.length : 0;
      if (groupId !== 0 && !GroupsManager.hasGroup(fabric, endpoint.getId(), groupId)) {
        return { status: StatusCode.InvalidCommand, groupId, capacity };
      }
      const sceneList = endpointScenes.map(({ sceneId }) => sceneId);
      return { status: StatusCode.Success, groupId, capacity, sceneList };
    },
    enhancedAddScene: async ({
      request: { groupId, sceneId, transitionTime, sceneName, extensionFieldSets },
      attributes: { sceneCount },
      session,
      endpoint
    }) => {
      assertSecureSession(session);
      const result = addSceneLogic(
        endpoint.getId(),
        groupId,
        sceneId,
        Math.floor(transitionTime / 10),
        sceneName,
        extensionFieldSets,
        transitionTime % 10,
        session.getAssociatedFabric()
      );
      if (result.status === StatusCode.Success) {
        sceneCount.updated(session);
      }
      return result;
    },
    enhancedViewScene: async ({ request: { groupId, sceneId }, session, endpoint }) => {
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      if (groupId !== 0 && !GroupsManager.hasGroup(fabric, endpoint.getId(), groupId)) {
        return { status: StatusCode.InvalidCommand, groupId, sceneId };
      }
      const sceneEntry = ScenesManager.getSceneEntry(fabric, endpoint.getId(), groupId, sceneId);
      if (sceneEntry === void 0) {
        return { status: StatusCode.NotFound, groupId, sceneId };
      }
      const { sceneName, sceneTransitionTime, transitionTime100ms, extensionFieldSets } = sceneEntry;
      return {
        status: StatusCode.Success,
        groupId,
        sceneId,
        sceneName,
        transitionTime: sceneTransitionTime * 10 + transitionTime100ms,
        extensionFieldSets
      };
    },
    copyScene: async ({
      request: { mode, groupIdentifierFrom, sceneIdentifierFrom, groupIdentifierTo, sceneIdentifierTo },
      attributes: { sceneCount },
      session,
      endpoint
    }) => {
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      if (groupIdentifierFrom !== 0 && !GroupsManager.hasGroup(fabric, endpoint.getId(), groupIdentifierFrom)) {
        return { status: StatusCode.InvalidCommand, groupIdentifierFrom, sceneIdentifierFrom };
      }
      if (groupIdentifierTo !== 0 && !GroupsManager.hasGroup(fabric, endpoint.getId(), groupIdentifierTo)) {
        return { status: StatusCode.InvalidCommand, groupIdentifierFrom, sceneIdentifierFrom };
      }
      if (mode.copyAllScenes) {
        const sceneEntries = ScenesManager.getAllScenes(fabric, endpoint.getId(), groupIdentifierFrom);
        const newSceneEntries = sceneEntries.map(
          ({ sceneId, sceneName, sceneTransitionTime, extensionFieldSets, transitionTime100ms }) => ({
            scenesGroupId: groupIdentifierTo,
            sceneId,
            sceneName,
            sceneTransitionTime,
            extensionFieldSets,
            transitionTime100ms
          })
        );
        ScenesManager.setScenes(fabric, endpoint.getId(), newSceneEntries);
      } else {
        const sceneEntryFrom = ScenesManager.getSceneEntry(
          fabric,
          endpoint.getId(),
          groupIdentifierFrom,
          sceneIdentifierFrom
        );
        if (sceneEntryFrom === void 0) {
          return { status: StatusCode.NotFound, groupIdentifierFrom, sceneIdentifierFrom };
        }
        const { sceneName, sceneTransitionTime, transitionTime100ms, extensionFieldSets } = sceneEntryFrom;
        ScenesManager.setScenes(fabric, endpoint.getId(), [
          {
            scenesGroupId: groupIdentifierTo,
            sceneId: sceneIdentifierTo,
            sceneName,
            sceneTransitionTime,
            extensionFieldSets,
            transitionTime100ms
          }
        ]);
      }
      sceneCount.updated(session);
      return { status: StatusCode.Success, groupIdentifierFrom, sceneIdentifierFrom };
    },
    sceneValidAttributeGetter: ({ session, attributes: { currentScene, currentGroup }, endpoint }) => {
      if (session === void 0 || endpoint === void 0) {
        return false;
      }
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      const existingSceneEntry = ScenesManager.getSceneEntry(
        fabric,
        endpoint.getId(),
        currentGroup.getLocal(),
        currentScene.getLocal()
      );
      if (existingSceneEntry === void 0) {
        return false;
      }
      for (const clusterData of existingSceneEntry.extensionFieldSets) {
        const { clusterId, attributeValueList } = clusterData;
        const cluster = endpoint.getClusterServerById(clusterId);
        if (cluster !== void 0) {
          if (!asClusterServerInternal(cluster)._verifySceneExtensionFieldSets(attributeValueList)) {
            return false;
          }
        }
      }
      return true;
    },
    sceneCountAttributeGetter: ({ session, endpoint }) => {
      if (session === void 0 || endpoint === void 0) {
        throw new InternalError("getSceneCount: session or endpoint undefined");
      }
      assertSecureSession(session);
      const fabric = session.getAssociatedFabric();
      const endpointScenes = ScenesManager.getEndpointScenes(fabric, endpoint.getId());
      if (endpointScenes === void 0)
        return 0;
      let sceneCount = 0;
      for (const scenes of endpointScenes.values()) {
        sceneCount += scenes.size;
      }
      return sceneCount;
    }
  };
};
const createDefaultScenesClusterServer = () => ClusterServer(
  ScenesCluster,
  {
    sceneCount: 0,
    currentScene: 0,
    currentGroup: GroupId(0),
    sceneValid: false,
    nameSupport: {
      nameSupport: true
    },
    lastConfiguredBy: null
  },
  ScenesClusterHandler()
);
export {
  ScenesClusterHandler,
  ScenesManager,
  createDefaultScenesClusterServer
};
