// @project-chip/matter.js/dist/cjs/crypto/Crypto.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Crypto_exports = {};
__export(Crypto_exports, {
  CRYPTO_AUTH_TAG_LENGTH: () => CRYPTO_AUTH_TAG_LENGTH,
  CRYPTO_EC_CURVE: () => CRYPTO_EC_CURVE,
  CRYPTO_EC_KEY_BYTES: () => CRYPTO_EC_KEY_BYTES,
  CRYPTO_ENCRYPT_ALGORITHM: () => CRYPTO_ENCRYPT_ALGORITHM,
  CRYPTO_HASH_ALGORITHM: () => CRYPTO_HASH_ALGORITHM,
  CRYPTO_RANDOM_LENGTH: () => CRYPTO_RANDOM_LENGTH,
  CRYPTO_SYMMETRIC_KEY_LENGTH: () => CRYPTO_SYMMETRIC_KEY_LENGTH,
  Crypto: () => Crypto,
  CryptoError: () => CryptoError,
  ec: () => ec
});
module.exports = __toCommonJS(Crypto_exports);
var import_bn = __toESM(require("bn.js"), 1);
var import_MatterError = require("../common/MatterError.js");
var import_ByteArray = require("../util/ByteArray.js");
var import_DataReader = require("../util/DataReader.js");
var import_elliptic = __toESM(require("elliptic"), 1);
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const CRYPTO_RANDOM_LENGTH = 32;
const CRYPTO_ENCRYPT_ALGORITHM = "aes-128-ccm";
const CRYPTO_HASH_ALGORITHM = "sha256";
const CRYPTO_EC_CURVE = "prime256v1";
const CRYPTO_EC_KEY_BYTES = 32;
const CRYPTO_AUTH_TAG_LENGTH = 16;
const CRYPTO_SYMMETRIC_KEY_LENGTH = 16;
class CryptoError extends import_MatterError.MatterError {
}
class Crypto {
  static {
    this.get = () => {
      throw new import_MatterError.NoProviderError("No provider configured");
    };
  }
  static {
    this.encrypt = (key, data, nonce, aad) => Crypto.get().encrypt(key, data, nonce, aad);
  }
  static {
    this.decrypt = (key, data, nonce, aad) => Crypto.get().decrypt(key, data, nonce, aad);
  }
  static {
    this.getRandomData = (length) => Crypto.get().getRandomData(length);
  }
  static {
    this.getRandom = () => Crypto.get().getRandomData(CRYPTO_RANDOM_LENGTH);
  }
  static {
    this.getRandomUInt16 = () => new import_DataReader.DataReader(Crypto.get().getRandomData(2), import_ByteArray.Endian.Little).readUInt16();
  }
  static {
    this.getRandomUInt32 = () => new import_DataReader.DataReader(Crypto.get().getRandomData(4), import_ByteArray.Endian.Little).readUInt32();
  }
  static {
    this.getRandomBigUInt64 = () => new import_DataReader.DataReader(Crypto.get().getRandomData(8), import_ByteArray.Endian.Little).readUInt64();
  }
  static {
    this.getRandomBN = (size, maxValue) => {
      while (true) {
        const random = new import_bn.default(Crypto.getRandomData(size));
        if (random < maxValue)
          return random;
      }
    };
  }
  static {
    this.ecdhGeneratePublicKey = () => Crypto.get().ecdhGeneratePublicKey();
  }
  static {
    this.ecdhGeneratePublicKeyAndSecret = (peerPublicKey) => Crypto.get().ecdhGeneratePublicKeyAndSecret(peerPublicKey);
  }
  static {
    this.ecdhGenerateSecret = (peerPublicKey, ecdh) => Crypto.get().ecdhGenerateSecret(peerPublicKey, ecdh);
  }
  static {
    this.hash = (data) => Crypto.get().hash(data);
  }
  static {
    this.pbkdf2 = (secret, salt, iteration, keyLength) => Crypto.get().pbkdf2(secret, salt, iteration, keyLength);
  }
  static {
    this.hkdf = (secret, salt, info, length) => Crypto.get().hkdf(secret, salt, info, length);
  }
  static {
    this.hmac = (key, data) => Crypto.get().hmac(key, data);
  }
  static {
    this.sign = (privateKey, data, dsaEncoding) => Crypto.get().sign(privateKey, data, dsaEncoding);
  }
  static {
    this.verify = (publicKey, data, signature, dsaEncoding) => Crypto.get().verify(publicKey, data, signature, dsaEncoding);
  }
  static {
    this.createKeyPair = () => Crypto.get().createKeyPair();
  }
}
const { ec } = import_elliptic.default;
if (typeof MatterHooks !== "undefined") {
  MatterHooks.cryptoSetup?.(Crypto);
}
