// @project-chip/matter.js/dist/cjs/crypto/Spake2p.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Spake2p_exports = {};
__export(Spake2p_exports, {
  Spake2p: () => Spake2p
});
module.exports = __toCommonJS(Spake2p_exports);
var import_bn = __toESM(require("bn.js"), 1);
var import_MatterError = require("../common/MatterError.js");
var import_ByteArray = require("../util/ByteArray.js");
var import_DataWriter = require("../util/DataWriter.js");
var import_Crypto = require("./Crypto.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const P256_CURVE = new import_Crypto.ec("p256").curve;
const M = P256_CURVE.decodePoint("02886e2f97ace46e55ba9dd7242579f2993b64e16ef3dcab95afd497333d8fa12f", "hex");
const N = P256_CURVE.decodePoint("03d8bbd6c639c62937b04d997f38c3770719c629d7014d49a24b4f98baa1292b49", "hex");
class Spake2p {
  constructor(context, random, w0) {
    this.context = context;
    this.random = random;
    this.w0 = w0;
  }
  static async computeW0W1({ iterations, salt }, pin) {
    const pinWriter = new import_DataWriter.DataWriter(import_ByteArray.Endian.Little);
    pinWriter.writeUInt32(pin);
    const ws = await import_Crypto.Crypto.pbkdf2(pinWriter.toByteArray(), salt, iterations, 80);
    const w0 = new import_bn.default(ws.slice(0, 40)).mod(P256_CURVE.n);
    const w1 = new import_bn.default(ws.slice(40, 80)).mod(P256_CURVE.n);
    return { w0, w1 };
  }
  static async computeW0L(pbkdfParameters, pin) {
    const { w0, w1 } = await this.computeW0W1(pbkdfParameters, pin);
    const L = P256_CURVE.g.mul(w1).encode();
    return { w0, L };
  }
  static create(context, w0) {
    const random = import_Crypto.Crypto.getRandomBN(32, P256_CURVE.p);
    return new Spake2p(context, random, w0);
  }
  computeX() {
    const X = P256_CURVE.g.mul(this.random).add(M.mul(this.w0));
    return import_ByteArray.ByteArray.from(X.encode());
  }
  computeY() {
    const Y = P256_CURVE.g.mul(this.random).add(N.mul(this.w0));
    return import_ByteArray.ByteArray.from(Y.encode());
  }
  async computeSecretAndVerifiersFromY(w1, X, Y) {
    const YPoint = P256_CURVE.decodePoint(Y);
    if (!YPoint.validate())
      throw new import_MatterError.InternalError("Y is not on the curve");
    const yNwo = YPoint.add(N.mul(this.w0).neg());
    const Z = yNwo.mul(this.random);
    const V = yNwo.mul(w1);
    return this.computeSecretAndVerifiers(X, Y, import_ByteArray.ByteArray.from(Z.encode()), import_ByteArray.ByteArray.from(V.encode()));
  }
  async computeSecretAndVerifiersFromX(L, X, Y) {
    const XPoint = P256_CURVE.decodePoint(X);
    const LPoint = P256_CURVE.decodePoint(L);
    if (!XPoint.validate())
      throw new import_MatterError.InternalError("X is not on the curve");
    const Z = XPoint.add(M.mul(this.w0).neg()).mul(this.random);
    const V = LPoint.mul(this.random);
    return this.computeSecretAndVerifiers(X, Y, import_ByteArray.ByteArray.from(Z.encode()), import_ByteArray.ByteArray.from(V.encode()));
  }
  async computeSecretAndVerifiers(X, Y, Z, V) {
    const TT_HASH = this.computeTranscriptHash(X, Y, Z, V);
    const Ka = TT_HASH.slice(0, 16);
    const Ke = TT_HASH.slice(16, 32);
    const KcAB = await import_Crypto.Crypto.hkdf(Ka, new import_ByteArray.ByteArray(0), import_ByteArray.ByteArray.fromString("ConfirmationKeys"), 32);
    const KcA = KcAB.slice(0, 16);
    const KcB = KcAB.slice(16, 32);
    const hAY = import_Crypto.Crypto.hmac(KcA, Y);
    const hBX = import_Crypto.Crypto.hmac(KcB, X);
    return { Ke, hAY, hBX };
  }
  computeTranscriptHash(X, Y, Z, V) {
    const TTwriter = new import_DataWriter.DataWriter(import_ByteArray.Endian.Little);
    this.addToContext(TTwriter, this.context);
    this.addToContext(TTwriter, import_ByteArray.ByteArray.fromString(""));
    this.addToContext(TTwriter, import_ByteArray.ByteArray.fromString(""));
    this.addToContext(TTwriter, import_ByteArray.ByteArray.from(M.encode()));
    this.addToContext(TTwriter, import_ByteArray.ByteArray.from(N.encode()));
    this.addToContext(TTwriter, X);
    this.addToContext(TTwriter, Y);
    this.addToContext(TTwriter, Z);
    this.addToContext(TTwriter, V);
    this.addToContext(TTwriter, import_ByteArray.ByteArray.from(this.w0.toArray()));
    return import_Crypto.Crypto.hash(TTwriter.toByteArray());
  }
  addToContext(TTwriter, data) {
    TTwriter.writeUInt64(data.length);
    TTwriter.writeByteArray(data);
  }
}
