// @project-chip/matter.js/dist/esm/model/logic/cluster-variance/IllegalFeatureCombinations.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { InternalError } from "../../../common/MatterError.js";
import { isDeepEqual } from "../../../util/DeepEqual.js";
import { Conformance } from "../../aspects/index.js";
import { FeatureBitmap } from "./FeatureBitmap.js";
function IllegalFeatureCombinations(cluster) {
  const illegal = [];
  const choices = {};
  for (const f of cluster.features) {
    addFeatureNode(f, f.conformance.ast, illegal, choices);
  }
  function add(flags) {
    if (!illegal.some((e) => isDeepEqual(e, flags))) {
      illegal.push(flags);
    }
  }
  for (const choice of Object.values(choices)) {
    if (choice.exclusive) {
      for (const f1 of choice.features) {
        for (const f2 of choice.features) {
          if (f1 !== f2) {
            add({ [f1]: true, [f2]: true });
          }
        }
      }
    }
    const flags = FeatureBitmap();
    for (const f of choice.features) {
      flags[f] = false;
    }
    add(flags);
  }
  return { illegal, requiresFeatures: !!choices.length };
}
function addFeatureNode(feature, node, illegal, choices) {
  function unsupported() {
    throw new InternalError(`New rule required to support ${feature.path} conformance "${feature.conformance}"`);
  }
  function addDisjunctRequirement(flags, node2) {
    switch (node2.type) {
      case Conformance.Special.Name:
        flags[node2.param] = false;
        break;
      case Conformance.OR:
        addDisjunctRequirement(flags, node2.param.lhs);
        addDisjunctRequirement(flags, node2.param.rhs);
        break;
      default:
        unsupported();
    }
  }
  function addConjunctRequirement(feature2, node2) {
    switch (node2.type) {
      case Conformance.Special.Name:
        illegal.push({ [feature2]: true, [node2.param]: false });
        break;
      case Conformance.AND:
        addConjunctRequirement(feature2, node2.param.lhs);
        addConjunctRequirement(feature2, node2.param.rhs);
        break;
    }
  }
  switch (node.type) {
    case Conformance.Special.Empty:
    case Conformance.Flag.Optional:
    case Conformance.Flag.Provisional:
      break;
    case Conformance.Special.Group:
      node.param.forEach((ast) => addFeatureNode(feature, ast, illegal, choices));
      break;
    case Conformance.Special.Choice:
      if (node.param.num > 1) {
        unsupported();
      }
      let choice = choices[node.param.name];
      if (choice) {
        choice.features.push(feature.name);
      } else {
        choice = choices[node.param.name] = {
          exclusive: !node.param.orMore,
          features: [feature.name]
        };
      }
      break;
    case Conformance.Special.Name:
      illegal.push({ [node.param]: true, [feature.name]: false });
      break;
    case Conformance.Special.OptionalIf:
      if (node.param.type == Conformance.AND) {
        addConjunctRequirement(feature.name, node.param);
      } else {
        const flags = FeatureBitmap({ [feature.name]: true });
        addDisjunctRequirement(flags, node.param);
        illegal.push(flags);
      }
      break;
    default:
      unsupported();
      break;
  }
}
export {
  IllegalFeatureCombinations
};
