// @project-chip/matter.js/dist/cjs/time/Time.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Time_exports = {};
__export(Time_exports, {
  Time: () => Time
});
module.exports = __toCommonJS(Time_exports);
var import_MatterError = require("../common/MatterError.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class Time {
  static {
    this.get = () => DefaultTime;
  }
  static {
    this.now = () => Time.get().now();
  }
  static {
    this.nowMs = () => Time.get().nowMs();
  }
  static {
    this.getTimer = (durationMs, callback) => Time.get().getTimer(durationMs, callback);
  }
  static {
    this.getPeriodicTimer = (intervalMs, callback) => Time.get().getPeriodicTimer(intervalMs, callback);
  }
  static {
    this.sleep = async (durationMs) => new Promise((resolve) => Time.get().getTimer(durationMs, resolve).start());
  }
}
const DefaultTime = new class extends Time {
  now() {
    return /* @__PURE__ */ new Date();
  }
  nowMs() {
    return this.now().getTime();
  }
  getTimer() {
    throw new import_MatterError.NoProviderError("Default time provider does not implement timers");
  }
  getPeriodicTimer() {
    return this.getTimer();
  }
}();
if (typeof MatterHooks !== "undefined") {
  MatterHooks.timeSetup?.(Time);
}
