// @project-chip/matter.js/dist/esm/model/aspects/Conformance.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { FieldValue } from "../definitions/index.js";
import { Aspect } from "./Aspect.js";
class Conformance extends Aspect {
  get type() {
    return this.ast.type;
  }
  get empty() {
    return this.type === Conformance.Special.Empty;
  }
  /**
   * Initialize from a Conformance.Definition or the conformance DSL defined
   * by the Matter Specification.
   */
  constructor(definition) {
    super(definition);
    let ast;
    if (definition === void 0) {
      this.ast = { type: Conformance.Special.Empty };
      return;
    } else if (typeof definition === "string") {
      if (definition.toLowerCase() === "desc") {
        this.ast = { type: Conformance.Special.Desc };
        return;
      }
      ast = new Parser(this, definition).ast;
    } else if (Array.isArray(definition)) {
      const asts = definition.map((def) => new Parser(this, def).ast);
      if (asts.length === 1) {
        ast = asts[0];
      } else {
        ast = {
          type: Conformance.Special.Group,
          param: asts
        };
      }
    } else {
      ast = definition.ast;
    }
    this.ast = ast;
  }
  validateReferences(lookup) {
    return Conformance.validateReferences(this, this.ast, lookup);
  }
  toString() {
    return Conformance.serialize(this.ast);
  }
}
((Conformance2) => {
  let Special;
  ((Special2) => {
    Special2["Empty"] = "empty";
    Special2["Desc"] = "desc";
    Special2["Name"] = "name";
    Special2["Value"] = "value";
    Special2["Choice"] = "choice";
    Special2["Group"] = "group";
    Special2["OptionalIf"] = "optionalIf";
  })(Special = Conformance2.Special || (Conformance2.Special = {}));
  let Flag;
  ((Flag2) => {
    Flag2["Mandatory"] = "M";
    Flag2["Optional"] = "O";
    Flag2["Provisional"] = "P";
    Flag2["Deprecated"] = "D";
    Flag2["Disallowed"] = "X";
  })(Flag = Conformance2.Flag || (Conformance2.Flag = {}));
  let Operator;
  ((Operator2) => {
    Operator2["NOT"] = "!";
    Operator2["EQ"] = "==";
    Operator2["NE"] = "!=";
    Operator2["OR"] = "|";
    Operator2["XOR"] = "^";
    Operator2["AND"] = "&";
    Operator2["DOT"] = ".";
    Operator2["GT"] = ">";
    Operator2["LT"] = "<";
    Operator2["GTE"] = ">=";
    Operator2["LTE"] = "<=";
  })(Operator = Conformance2.Operator || (Conformance2.Operator = {}));
  Conformance2.M = "M" /* Mandatory */;
  Conformance2.O = "O" /* Optional */;
  Conformance2.P = "P" /* Provisional */;
  Conformance2.D = "D" /* Deprecated */;
  Conformance2.X = "X" /* Disallowed */;
  Conformance2.EQ = "==" /* EQ */;
  Conformance2.NE = "!=" /* NE */;
  Conformance2.OR = "|" /* OR */;
  Conformance2.XOR = "^" /* XOR */;
  Conformance2.AND = "&" /* AND */;
  Conformance2.DOT = "." /* DOT */;
  Conformance2.GT = ">" /* GT */;
  Conformance2.LT = "<" /* LT */;
  Conformance2.GTE = ">=" /* GTE */;
  Conformance2.LTE = "<=" /* LTE */;
  function serializeAtomic(ast, otherOperator) {
    const serialized = serialize(ast);
    if (ast.type === "group" /* Group */ || otherOperator !== void 0 && isHigherPrecedence(otherOperator, ast.type)) {
      return `(${serialized})`;
    }
    return serialized;
  }
  function validateReferences(conformance, ast, resolver) {
    switch (ast.type) {
      case "|" /* OR */:
      case "^" /* XOR */:
      case "&" /* AND */:
      case "==" /* EQ */:
      case "!=" /* NE */:
      case ">" /* GT */:
      case "<" /* LT */:
      case ">=" /* GTE */:
      case "<=" /* LTE */:
        validateReferences(conformance, ast.param.lhs, resolver);
        validateReferences(conformance, ast.param.rhs, resolver);
        break;
      case "!" /* NOT */:
        validateReferences(conformance, ast.param, resolver);
        break;
      case "group" /* Group */:
        for (const a of ast.param) {
          validateReferences(conformance, a, resolver);
        }
        break;
      case "name" /* Name */:
        if (!resolver(ast.param)) {
          conformance.error(
            `UNRESOLVED_CONFORMANCE_${ast.type.toUpperCase()}`,
            `Conformance ${ast.type} reference "${ast.param}" does not resolve`
          );
        }
        break;
    }
  }
  Conformance2.validateReferences = validateReferences;
  function serialize(ast) {
    switch (ast.type) {
      case "|" /* OR */:
      case "^" /* XOR */:
      case "&" /* AND */:
      case "==" /* EQ */:
      case "!=" /* NE */:
      case ">" /* GT */:
      case "<" /* LT */:
      case ">=" /* GTE */:
      case "<=" /* LTE */:
        const lhs = serializeAtomic(ast.param.lhs, ast.type);
        const rhs = serializeAtomic(ast.param.rhs, ast.type);
        return `${lhs} ${ast.type} ${rhs}`;
      case "!" /* NOT */:
        return `!${serializeAtomic(ast.param)}`;
      case "empty" /* Empty */:
        return "";
      case "desc" /* Desc */:
        return "desc";
      case "choice" /* Choice */:
        let result = `${serializeAtomic(ast.param.expr)}.${ast.param.name}`;
        if (ast.param.num > 1) {
          result = `${result}${ast.param.num}`;
        }
        if (ast.param.orMore) {
          result = `${result}+`;
        }
        return result;
      case "group" /* Group */:
        return ast.param.map((d) => serialize(d)).join(", ");
      case "optionalIf" /* OptionalIf */:
        return `[${serialize(ast.param)}]`;
      case "name" /* Name */:
      case "value" /* Value */:
        return FieldValue.serialize(ast.param);
      default:
        return ast.type;
    }
  }
  Conformance2.serialize = serialize;
  function isBinaryOperator(type) {
    return Parser.BinaryOperators.has(type);
  }
  Conformance2.isBinaryOperator = isBinaryOperator;
  function precedenceOf(operator) {
    const index = Parser.BinaryOperatorPrecedence.findIndex((ops) => ops.indexOf(operator) != -1);
    return index === -1 ? void 0 : index;
  }
  Conformance2.precedenceOf = precedenceOf;
  function isHigherPrecedence(operator, other) {
    const precedence1 = precedenceOf(operator);
    if (precedence1 === void 0) {
      return false;
    }
    const precedence2 = precedenceOf(other);
    if (precedence2 === void 0) {
      return false;
    }
    return precedence1 < precedence2;
  }
  Conformance2.isHigherPrecedence = isHigherPrecedence;
})(Conformance || (Conformance = {}));
function isNameChar(c) {
  return c >= "A" && c <= "Z" || c >= "a" && c <= "z" || c >= "0" && c <= "9" || c === "_";
}
var Tokenizer;
((Tokenizer2) => {
  let Special;
  ((Special2) => {
    Special2["Not"] = "!";
    Special2["Equal"] = "==";
    Special2["NotEqual"] = "!=";
    Special2["Or"] = "|";
    Special2["Xor"] = "^";
    Special2["And"] = "&";
    Special2["Dot"] = ".";
    Special2["Comma"] = ",";
    Special2["OptionalBegin"] = "[";
    Special2["OptionalEnd"] = "]";
    Special2["GroupBegin"] = "(";
    Special2["GroupEnd"] = ")";
    Special2["Plus"] = "+";
    Special2["GreaterThan"] = ">";
    Special2["LessThan"] = "<";
    Special2["GreaterThanOrEqual"] = ">=";
    Special2["LessThanOrEqual"] = "<=";
  })(Special = Tokenizer2.Special || (Tokenizer2.Special = {}));
  let TokenType;
  ((TokenType2) => {
    TokenType2[TokenType2["Flag"] = 0] = "Flag";
    TokenType2[TokenType2["Special"] = 1] = "Special";
    TokenType2[TokenType2["Name"] = 2] = "Name";
    TokenType2[TokenType2["Choice"] = 3] = "Choice";
    TokenType2[TokenType2["Number"] = 4] = "Number";
  })(TokenType = Tokenizer2.TokenType || (Tokenizer2.TokenType = {}));
  function* tokenize(conformance, definition) {
    const i = definition[Symbol.iterator]();
    let current = i.next();
    if (current.done) {
      return;
    }
    let peeked = i.next();
    function next() {
      current = peeked;
      if (!current.done) {
        peeked = i.next();
      }
    }
    function tokenizeName() {
      const name = [current.value];
      while (isNameChar(peeked.value)) {
        next();
        name.push(current.value);
      }
      return { type: 2 /* Name */, value: name.join("") };
    }
    while (!current.done) {
      switch (current.value) {
        case "M" /* Mandatory */:
        case "O" /* Optional */:
        case "P" /* Provisional */:
        case "D" /* Deprecated */:
        case "X" /* Disallowed */:
          if (isNameChar(peeked.value)) {
            yield tokenizeName();
          } else {
            yield { type: 0 /* Flag */, value: current.value };
          }
          break;
        case "|" /* Or */:
        case "^" /* Xor */:
        case "&" /* And */:
        case "." /* Dot */:
        case "," /* Comma */:
        case "[" /* OptionalBegin */:
        case "]" /* OptionalEnd */:
        case "(" /* GroupBegin */:
        case ")" /* GroupEnd */:
        case "+" /* Plus */:
          yield { type: 1 /* Special */, value: current.value };
          break;
        case "0":
        case "1":
        case "2":
        case "3":
        case "4":
        case "5":
        case "6":
        case "7":
        case "8":
        case "9":
          let num = 0;
          while (true) {
            num = num * 10 + current.value.charCodeAt(0) - "0".charCodeAt(0);
            if (peeked.done || peeked.value < "0" || peeked.value > "9") {
              break;
            }
            next();
          }
          if (peeked.value === "%") {
            next();
            num = FieldValue.Percent(num);
          } else if (peeked.value === "\xB0") {
            next();
            if (peeked.value?.toLowerCase() === "C") {
              next();
            }
            num = FieldValue.Celsius(num);
          }
          yield { type: 4 /* Number */, value: num };
          break;
        case "!":
        case ">":
        case "<":
          {
            const base = current.value;
            if (peeked.value === "=") {
              next();
              yield { type: 1 /* Special */, value: `${base}${peeked.value}` };
            } else {
              yield { type: 1 /* Special */, value: base };
            }
          }
          break;
        case "=":
          if (peeked.value === "=") {
            next();
          } else {
            conformance.error("BAD_EQUAL", `"=" must be followed by another "="`);
          }
          yield { type: 1 /* Special */, value: "==" /* Equal */ };
          break;
        case " ":
        case "	":
        case "\r":
        case "\n":
        case "\v":
        case "\f":
          break;
        default:
          if (current.value >= "a" && current.value <= "z") {
            yield { type: 3 /* Choice */, value: current.value };
          } else if (current.value >= "A" && current.value <= "Z") {
            yield tokenizeName();
          } else {
            conformance.error("GARBAGE_CHARACTER", `Unexpected character "${current.value}"`);
          }
          break;
      }
      next();
    }
  }
  Tokenizer2.tokenize = tokenize;
})(Tokenizer || (Tokenizer = {}));
class Parser {
  constructor(conformance, definition) {
    this.conformance = conformance;
    definition = definition.replace(" or ", " | ");
    this.tokens = Tokenizer.tokenize(conformance, definition);
    const next = this.tokens.next();
    if (!next.done) {
      this.peeked = next.value;
      this.next();
    }
    this.ast = this.parse();
  }
  next() {
    this.token = this.peeked;
    const next = this.tokens.next();
    if (next.done) {
      this.peeked = void 0;
    } else {
      this.peeked = next.value;
    }
  }
  // Note that Conformance.Definition effectively serves as our AST.  Its
  // design is slightly suboptimal for this purpose because it also attempts
  // to serve as a DSL for manual expression of conformance
  parse() {
    return this.parseGroup();
  }
  parseGroup(end) {
    const group = [];
    function groupAsAst() {
      if (group.length === 1) {
        return group[0];
      }
      return {
        type: "group" /* Group */,
        param: group
      };
    }
    while (true) {
      if (!this.token) {
        if (end) {
          this.conformance.error("UNTERMINATED_CONFORMANCE_GROUPING", "Unterminated conformance grouping");
        }
        return groupAsAst();
      }
      const optional = !end && this.atSpecial("[" /* OptionalBegin */);
      if (optional) {
        this.next();
        let expr = {
          type: "optionalIf" /* OptionalIf */,
          param: this.parseGroup("]" /* OptionalEnd */)
        };
        expr = this.parseChoice(expr);
        group.push(expr);
      } else {
        const expr = this.parseExpression();
        if (expr) {
          group.push(expr);
        }
      }
      if (this.atSpecial("," /* Comma */)) {
        this.next();
      } else if (end && this.atSpecial(end)) {
        this.next();
        return groupAsAst();
      }
    }
  }
  atSpecial(special) {
    return this.token && this.token.type === 1 /* Special */ && this.token.value === special;
  }
  parseExpression() {
    const elements = [];
    let expr = this.parseAtomicExpression();
    if (expr) {
      elements.push(expr);
    }
    while (this.token && this.token.type === 1 /* Special */ && Parser.BinaryOperators.has(this.token.value)) {
      elements.push(this.token.value);
      this.next();
      expr = this.parseAtomicExpression();
      if (expr) {
        elements.push(expr);
      }
    }
    Parser.BinaryOperatorPrecedence.forEach((operators) => {
      for (let i = 0; i < elements.length; i++) {
        if (operators.indexOf(elements[i + 1]) !== -1) {
          const [lhs, op, rhs] = elements.splice(i, 3);
          elements.splice(i, 0, {
            type: op,
            param: { lhs, rhs }
          });
          i--;
        }
      }
    });
    return elements[0];
  }
  parseChoice(expr) {
    if (!this.atSpecial("." /* Dot */)) {
      return expr;
    }
    this.next();
    if (this.token?.type !== 3 /* Choice */) {
      this.conformance.error(
        "INVALID_CHOICE",
        'Choice indicator (".") must be followed by a single lowercase letter'
      );
    }
    const choice = {
      name: this.token?.value ?? "?",
      expr,
      num: 1
    };
    this.next();
    if (this.token?.type === 4 /* Number */) {
      choice.num = this.token?.value;
      this.next();
    }
    if (this.atSpecial("+" /* Plus */)) {
      choice.orMore = true;
      this.next();
    }
    return {
      type: "choice" /* Choice */,
      param: choice
    };
  }
  parseAtomicExpression() {
    const expr = this.parseAtomicExpressionWithoutChoice();
    if (!expr) {
      return;
    }
    return this.parseChoice(expr);
  }
  parseAtomicExpressionWithoutChoice() {
    if (!this.token) {
      this.conformance.error("PREMATURE_CONFORMANCE_TERMINATION", "Terminated with expression expected");
      return;
    }
    if (this.token.type === 0 /* Flag */) {
      const value = this.token.value;
      this.next();
      return { type: value };
    }
    if (this.token.type === 2 /* Name */) {
      const name = this.token.value;
      this.next();
      return { type: "name" /* Name */, param: name };
    }
    if (this.token.type === 4 /* Number */) {
      const value = this.token.value;
      this.next();
      return { type: "value" /* Value */, param: value };
    }
    if (this.atSpecial("!" /* Not */)) {
      this.next();
      return { type: "!" /* NOT */, param: this.parseAtomicExpression() };
    }
    if (this.atSpecial("(" /* GroupBegin */)) {
      this.next();
      return this.parseGroup(")" /* GroupEnd */);
    }
    this.conformance.error("UNEXPECTED_CONFORMANCE_TOKEN", `Unexpected "${this.token.value}"`);
    this.next();
  }
}
((Parser2) => {
  Parser2.BinaryOperatorPrecedence = [
    ["&" /* And */],
    ["|" /* Or */, "^" /* Xor */],
    [
      ">" /* GreaterThan */,
      "<" /* LessThan */,
      ">=" /* GreaterThanOrEqual */,
      "<=" /* LessThanOrEqual */
    ],
    ["==" /* Equal */, "!=" /* NotEqual */]
  ];
  Parser2.BinaryOperators = new Set(Parser2.BinaryOperatorPrecedence.flat());
})(Parser || (Parser = {}));
export {
  Conformance
};
