// @project-chip/matter.js/dist/esm/model/aspects/Constraint.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { camelize } from "../../util/String.js";
import { FieldValue } from "../definitions/index.js";
import { Aspect } from "./Aspect.js";
class Constraint extends Aspect {
  /**
   * Initialize from a Constraint.Definition or the constraint DSL defined
   * by the Matter Specification.
   */
  constructor(definition) {
    super(definition);
    let ast;
    switch (typeof definition) {
      case "string":
        ast = Constraint.parse(this, definition);
        break;
      case "number":
        ast = { value: definition };
        break;
      default:
        ast = definition;
    }
    Object.assign(this, ast);
  }
  /**
   * Test a value against a constraint.
   */
  test(value) {
    if (value === void 0) {
      return false;
    }
    if (this.value === value) {
      return true;
    }
    if (this.value !== void 0 || value === null) {
      return false;
    }
    if (this.min !== void 0 && this.min !== null) {
      if (typeof this.min !== typeof value || this.min > value) {
        return false;
      }
    }
    if (this.max !== void 0 && this.max !== null) {
      if (typeof this.max !== typeof value || this.max <= value) {
        return false;
      }
    }
    return true;
  }
  toString() {
    return Constraint.serialize(this);
  }
}
((Constraint2) => {
  function parseValue(numOrName) {
    let value;
    if (numOrName.startsWith("0x") || numOrName.startsWith("0b")) {
      value = Number.parseInt(numOrName);
    } else {
      value = Number.parseFloat(numOrName);
    }
    if (Number.isNaN(value)) {
      return FieldValue.Reference(camelize(numOrName));
    }
    if (numOrName.endsWith("%")) {
      return FieldValue.Percent(value);
    }
    if (numOrName.endsWith("\xB0C")) {
      return FieldValue.Celsius(value);
    }
    return value;
  }
  function parseAtom(constraint, words) {
    switch (words.length) {
      case 0:
        return void 0;
      case 1:
        switch (words[0].toLowerCase()) {
          case "desc":
            return { desc: true };
          case "all":
          case "any":
            return {};
        }
        const value = parseValue(words[0]);
        if (value === void 0 || value === null) {
          return;
        }
        return { value };
      case 2:
        switch (words[0].toLowerCase()) {
          case "min":
            const min = parseValue(words[1]);
            if (min === void 0 || min === null) {
              return;
            }
            return { min };
          case "max":
            const max = parseValue(words[1]);
            if (max === void 0 || max === null) {
              return;
            }
            return { max };
          default:
            constraint.error(
              "INVALID_CONSTRAINT",
              `Two word constraint "${words.join(" ")}" does not start with "min" or "max"`
            );
        }
        return;
      case 3:
        if (words[1].toLowerCase() === "to") {
          let parseBound2 = function(name, pos) {
            if (words[pos].toLowerCase() === name) {
              return void 0;
            }
            return parseValue(words[pos]);
          };
          var parseBound = parseBound2;
          const ast = {};
          const min = parseBound2("min", 0);
          if (min !== void 0 && min !== null) {
            ast.min = min;
          }
          const max = parseBound2("max", 2);
          if (max !== void 0 && max !== null) {
            ast.max = max;
          }
          if (ast.min !== void 0 && ast.min !== null || ast.max !== void 0 && ast.max !== null) {
            return ast;
          }
        }
        return;
    }
    constraint.error("INVALID_CONSTRAINT", `Unrecognized value constraint "${words.join(" ")}"`);
  }
  function parse(constraint, definition) {
    let pos = 2;
    let current = definition[0];
    let peeked = definition[1];
    function next() {
      current = peeked;
      if (pos === definition.length) {
        peeked = void 0;
      } else {
        peeked = definition[pos];
        pos++;
      }
    }
    function scan(depth) {
      const parts = Array();
      let words = Array();
      let word = "";
      function parseWords() {
        if (word) {
          words.push(word);
          word = "";
        }
        const atom = parseAtom(constraint, words);
        words = Array();
        return atom;
      }
      function emit() {
        const atom = parseWords();
        if (atom !== void 0) {
          parts.push(atom);
        }
      }
      while (current !== void 0) {
        switch (current) {
          case " ":
          case "	":
          case "\r":
          case "\n":
          case "\v":
          case "\f":
            if (word) {
              words.push(word);
              word = "";
            }
            break;
          case "[":
            next();
            let ast = parseWords();
            const entry = scan(depth + 1);
            if (entry) {
              if (!ast) {
                ast = {};
              }
              ast.entry = entry;
            }
            if (ast) {
              parts.push(ast);
            }
            break;
          case "]":
            if (!depth) {
              constraint.error("INVALID_CONSTRAINT", 'Unexpected "]"');
              break;
            }
            emit();
            if (parts.length > 1) {
              return { parts };
            }
            return parts[0];
          case ",":
            emit();
            break;
          default:
            word += current;
            break;
        }
        next();
      }
      if (depth) {
        constraint.error("INVALID_CONSTRAINT", "Unterminated sub-constraint");
      }
      emit();
      if (parts.length < 2) {
        return parts[0];
      }
      return { parts };
    }
    return scan(0);
  }
  Constraint2.parse = parse;
  function serializeAtom(ast) {
    if (ast.desc) {
      return "desc";
    }
    if (ast.value !== void 0 && ast.value !== null) {
      return `${FieldValue.serialize(ast.value)}`;
    } else if (ast.min !== void 0 && ast.min !== null) {
      if (ast.max === void 0 || ast.max === null) {
        return `min ${FieldValue.serialize(ast.min)}`;
      }
      return `${FieldValue.serialize(ast.min)} to ${FieldValue.serialize(ast.max)}`;
    } else if (ast.max !== void 0 && ast.max !== null) {
      return `max ${FieldValue.serialize(ast.max)}`;
    }
    return "all";
  }
  function serialize(ast) {
    if (ast.parts) {
      return ast.parts.map(serialize).join(", ");
    }
    if (ast.entry) {
      return `${serializeAtom(ast)}[${serialize(ast.entry)}]`;
    }
    return serializeAtom(ast);
  }
  Constraint2.serialize = serialize;
})(Constraint || (Constraint = {}));
export {
  Constraint
};
