// @project-chip/matter.js/dist/esm/certificate/AttestationCertificateManager.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Crypto } from "../crypto/Crypto.js";
import { PrivateKey } from "../crypto/Key.js";
import { Time } from "../time/Time.js";
import { ByteArray } from "../util/ByteArray.js";
import { CertificateManager, jsToMatterDate } from "./CertificateManager.js";
import {
  TestCert_PAA_NoVID_PrivateKey,
  TestCert_PAA_NoVID_PublicKey,
  TestCert_PAA_NoVID_SKID
} from "./ChipPAAuthorities.js";
function getPaiCommonName(vendorId, productId) {
  return `node-matter Dev PAI 0x${vendorId.toString(16).toUpperCase()} ${productId === void 0 ? "no PID" : `0x${productId.toString(16).toUpperCase()}`}`;
}
function getDacCommonName(vendorId, productId) {
  return `node-matter Dev DAC 0x${vendorId.toString(16).toUpperCase()}/0x${productId.toString(16).toUpperCase()}`;
}
function getPaaCommonName() {
  return "Matter Test PAA";
}
class AttestationCertificateManager {
  constructor(vendorId) {
    this.vendorId = vendorId;
    this.paaCertId = BigInt(0);
    // We use the official PAA cert for now because else pairing with Chip tool do not work because
    // only this one is the Certificate store
    this.paaKeyPair = PrivateKey(TestCert_PAA_NoVID_PrivateKey, {
      publicKey: TestCert_PAA_NoVID_PublicKey
    });
    this.paaKeyIdentifier = TestCert_PAA_NoVID_SKID;
    this.paiCertId = BigInt(1);
    this.paiKeyPair = Crypto.createKeyPair();
    this.paiKeyIdentifier = Crypto.hash(this.paiKeyPair.publicKey).slice(0, 20);
    this.nextCertificateId = 2;
    this.paiCertBytes = this.generatePAICert(vendorId);
  }
  getPAICert() {
    return this.paiCertBytes;
  }
  getDACert(productId) {
    const dacKeyPair = Crypto.createKeyPair();
    return {
      keyPair: dacKeyPair,
      dac: this.generateDaCert(dacKeyPair.publicKey, this.vendorId, productId)
    };
  }
  // Method unused for now because we use the official Matter Test PAA, but is functional
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  generatePAACert(vendorId) {
    const now = Time.get().now();
    const unsignedCertificate = {
      serialNumber: ByteArray.of(Number(this.paaCertId)),
      signatureAlgorithm: 1,
      publicKeyAlgorithm: 1,
      ellipticCurveIdentifier: 1,
      issuer: {
        commonName: getPaaCommonName(),
        vendorId
      },
      notBefore: jsToMatterDate(now, -1),
      notAfter: jsToMatterDate(now, 10),
      subject: {
        commonName: getPaaCommonName(),
        vendorId
      },
      ellipticCurvePublicKey: this.paaKeyPair.publicKey,
      extensions: {
        basicConstraints: {
          isCa: true,
          pathLen: 1
        },
        keyUsage: 96,
        subjectKeyIdentifier: this.paaKeyIdentifier,
        authorityKeyIdentifier: this.paaKeyIdentifier
      }
    };
    return CertificateManager.paaCertToAsn1(unsignedCertificate, this.paaKeyPair);
  }
  generatePAICert(vendorId, productId) {
    const now = Time.get().now();
    const unsignedCertificate = {
      serialNumber: ByteArray.of(Number(this.paiCertId)),
      signatureAlgorithm: 1,
      publicKeyAlgorithm: 1,
      ellipticCurveIdentifier: 1,
      issuer: {
        commonName: getPaaCommonName()
      },
      notBefore: jsToMatterDate(now, -1),
      notAfter: jsToMatterDate(now, 10),
      subject: {
        commonName: getPaiCommonName(vendorId, productId),
        vendorId,
        productId
      },
      ellipticCurvePublicKey: this.paiKeyPair.publicKey,
      extensions: {
        basicConstraints: {
          isCa: true,
          pathLen: 0
        },
        keyUsage: 96,
        subjectKeyIdentifier: this.paiKeyIdentifier,
        authorityKeyIdentifier: this.paaKeyIdentifier
      }
    };
    return CertificateManager.paiCertToAsn1(unsignedCertificate, this.paaKeyPair);
  }
  generateDaCert(publicKey, vendorId, productId) {
    const now = Time.get().now();
    const certId = this.nextCertificateId++;
    const unsignedCertificate = {
      serialNumber: ByteArray.of(certId),
      // TODO: figure out what should happen if certId > 255
      signatureAlgorithm: 1,
      publicKeyAlgorithm: 1,
      ellipticCurveIdentifier: 1,
      notBefore: jsToMatterDate(now, -1),
      notAfter: jsToMatterDate(now, 10),
      issuer: {
        commonName: getPaiCommonName(vendorId),
        vendorId
      },
      subject: {
        commonName: getDacCommonName(vendorId, productId),
        vendorId,
        productId
      },
      ellipticCurvePublicKey: publicKey,
      extensions: {
        basicConstraints: {
          isCa: false
        },
        keyUsage: 1,
        subjectKeyIdentifier: Crypto.hash(publicKey).slice(0, 20),
        authorityKeyIdentifier: this.paiKeyIdentifier
      }
    };
    return CertificateManager.daCertToAsn1(unsignedCertificate, this.paiKeyPair);
  }
}
export {
  AttestationCertificateManager
};
