// @project-chip/matter.js/dist/esm/certificate/CertificateManager.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  AuthorityKeyIdentifier_X509,
  BasicConstraints_X509,
  BitByteArray,
  BYTES_KEY,
  ContextTagged,
  ContextTaggedBytes,
  DerCodec,
  DerObject,
  EcdsaWithSHA256_X962,
  ELEMENTS_KEY,
  ExtendedKeyUsage_X509,
  KeyUsage_Signature_ContentCommited_X509,
  KeyUsage_Signature_X509,
  OBJECT_ID_KEY,
  OrganisationName_X520,
  Pkcs7Data,
  Pkcs7SignedData,
  PublicKeyEcPrime256v1_X962,
  SHA256_CMS,
  SubjectKeyIdentifier_X509
} from "../codec/DerCodec.js";
import { MatterError } from "../common/MatterError.js";
import { Crypto } from "../crypto/Crypto.js";
import { PublicKey } from "../crypto/Key.js";
import { CaseAuthenticatedTag, TlvCaseAuthenticatedTag } from "../datatype/CaseAuthenticatedTag.js";
import { TlvFabricId } from "../datatype/FabricId.js";
import { NodeId, TlvNodeId } from "../datatype/NodeId.js";
import { TlvVendorId } from "../datatype/VendorId.js";
import { TlvArray } from "../tlv/TlvArray.js";
import { TlvBoolean } from "../tlv/TlvBoolean.js";
import { TlvUInt16, TlvUInt32, TlvUInt64, TlvUInt8 } from "../tlv/TlvNumber.js";
import { TlvField, TlvObject, TlvOptionalField, TlvOptionalRepeatedField, TlvTaggedList } from "../tlv/TlvObject.js";
import { TlvByteString, TlvString } from "../tlv/TlvString.js";
import { ByteArray } from "../util/ByteArray.js";
class CertificateError extends MatterError {
}
const YEAR_S = 365 * 24 * 60 * 60;
const EPOCH_OFFSET_S = 10957 * 24 * 60 * 60;
function matterToJsDate(date) {
  return new Date((date + EPOCH_OFFSET_S) * 1e3);
}
function jsToMatterDate(date, addYears = 0) {
  return Math.floor(date.getTime() / 1e3) - EPOCH_OFFSET_S + addYears * YEAR_S;
}
function intTo16Chars(value) {
  const byteArray = new ByteArray(8);
  const dataView = byteArray.getDataView();
  dataView.setBigUint64(0, typeof value === "bigint" ? value : BigInt(value));
  return byteArray.toHex().toUpperCase();
}
function uInt16To8Chars(value) {
  const byteArray = new ByteArray(4);
  const dataView = byteArray.getDataView();
  dataView.setUint32(0, value);
  return byteArray.toHex().toUpperCase();
}
function uInt16To4Chars(value) {
  const byteArray = new ByteArray(2);
  const dataView = byteArray.getDataView();
  dataView.setUint16(0, value);
  return byteArray.toHex().toUpperCase();
}
const CommonName_X520 = (name) => [DerObject("550403", { value: name })];
const NodeId_Matter = (nodeId) => [DerObject("2b0601040182a27c0101", { value: intTo16Chars(nodeId) })];
const FirmwareSigningId_Matter = (signingId) => [
  DerObject("2b0601040182a27c0102", { value: intTo16Chars(signingId) })
];
const IcacId_Matter = (id) => [DerObject("2b0601040182a27c0103", { value: intTo16Chars(id) })];
const RcacId_Matter = (id) => [DerObject("2b0601040182a27c0104", { value: intTo16Chars(id) })];
const FabricId_Matter = (id) => [DerObject("2b0601040182a27c0105", { value: intTo16Chars(id) })];
const NocCat_Matter = (cat) => [DerObject("2b0601040182a27c0106", { value: uInt16To8Chars(cat) })];
const VendorId_Matter = (vendorId) => [
  DerObject("2b0601040182a27c0201", { value: uInt16To4Chars(vendorId) })
];
const ProductId_Matter = (id) => [DerObject("2b0601040182a27c0202", { value: uInt16To4Chars(id) })];
const TlvRootCertificate = TlvObject({
  serialNumber: TlvField(1, TlvByteString.bound({ maxLength: 20 })),
  signatureAlgorithm: TlvField(2, TlvUInt8),
  issuer: TlvField(
    3,
    TlvTaggedList({
      issuerRcacId: TlvOptionalField(20, TlvUInt64)
    })
  ),
  notBefore: TlvField(4, TlvUInt32),
  notAfter: TlvField(5, TlvUInt32),
  subject: TlvField(
    6,
    TlvTaggedList({
      rcacId: TlvField(20, TlvUInt64)
    })
  ),
  publicKeyAlgorithm: TlvField(7, TlvUInt8),
  ellipticCurveIdentifier: TlvField(8, TlvUInt8),
  ellipticCurvePublicKey: TlvField(9, TlvByteString),
  extensions: TlvField(
    10,
    TlvTaggedList({
      basicConstraints: TlvField(
        1,
        TlvObject({
          isCa: TlvField(1, TlvBoolean),
          pathLen: TlvOptionalField(2, TlvUInt8)
        })
      ),
      keyUsage: TlvField(2, TlvUInt16),
      extendedKeyUsage: TlvOptionalField(3, TlvArray(TlvUInt8)),
      subjectKeyIdentifier: TlvField(4, TlvByteString.bound({ length: 20 })),
      authorityKeyIdentifier: TlvField(5, TlvByteString.bound({ length: 20 })),
      futureExtension: TlvOptionalField(6, TlvByteString)
    })
  ),
  signature: TlvField(11, TlvByteString)
});
const TlvOperationalCertificate = TlvObject({
  serialNumber: TlvField(1, TlvByteString.bound({ maxLength: 20 })),
  signatureAlgorithm: TlvField(2, TlvUInt8),
  issuer: TlvField(
    3,
    TlvTaggedList({
      issuerRcacId: TlvOptionalField(20, TlvUInt64)
    })
  ),
  notBefore: TlvField(4, TlvUInt32),
  notAfter: TlvField(5, TlvUInt32),
  subject: TlvField(
    6,
    TlvTaggedList({
      fabricId: TlvField(21, TlvFabricId),
      nodeId: TlvField(17, TlvNodeId),
      caseAuthenticatedTags: TlvOptionalRepeatedField(22, TlvCaseAuthenticatedTag, { maxLength: 3 })
    })
  ),
  publicKeyAlgorithm: TlvField(7, TlvUInt8),
  ellipticCurveIdentifier: TlvField(8, TlvUInt8),
  ellipticCurvePublicKey: TlvField(9, TlvByteString),
  extensions: TlvField(
    10,
    TlvTaggedList({
      basicConstraints: TlvField(
        1,
        TlvObject({
          isCa: TlvField(1, TlvBoolean),
          pathLen: TlvOptionalField(2, TlvUInt8)
        })
      ),
      keyUsage: TlvField(2, TlvUInt16),
      extendedKeyUsage: TlvOptionalField(3, TlvArray(TlvUInt8)),
      subjectKeyIdentifier: TlvField(4, TlvByteString.bound({ length: 20 })),
      authorityKeyIdentifier: TlvField(5, TlvByteString.bound({ length: 20 })),
      futureExtension: TlvOptionalField(6, TlvByteString)
    })
  ),
  signature: TlvField(11, TlvByteString)
});
const TlvIntermediateCertificate = TlvObject({
  serialNumber: TlvField(1, TlvByteString.bound({ maxLength: 20 })),
  signatureAlgorithm: TlvField(2, TlvUInt8),
  issuer: TlvField(
    3,
    TlvTaggedList({
      issuerRcacId: TlvOptionalField(20, TlvUInt64)
    })
  ),
  notBefore: TlvField(4, TlvUInt32),
  notAfter: TlvField(5, TlvUInt32),
  subject: TlvField(
    6,
    TlvTaggedList({
      fabricId: TlvOptionalField(21, TlvFabricId),
      icacId: TlvField(19, TlvUInt64)
    })
  ),
  publicKeyAlgorithm: TlvField(7, TlvUInt8),
  ellipticCurveIdentifier: TlvField(8, TlvUInt8),
  ellipticCurvePublicKey: TlvField(9, TlvByteString),
  extensions: TlvField(
    10,
    TlvTaggedList({
      basicConstraints: TlvField(
        1,
        TlvObject({
          isCa: TlvField(1, TlvBoolean)
        })
      ),
      keyUsage: TlvField(2, TlvUInt16),
      subjectKeyIdentifier: TlvField(4, TlvByteString.bound({ length: 20 })),
      authorityKeyIdentifier: TlvField(5, TlvByteString.bound({ length: 20 }))
    })
  ),
  signature: TlvField(11, TlvByteString)
});
const TlvCertificationDeclaration = TlvObject({
  formatVersion: TlvField(0, TlvUInt16),
  vendorId: TlvField(1, TlvVendorId),
  produceIdArray: TlvField(2, TlvArray(TlvUInt16, { minLength: 1, maxLength: 100 })),
  deviceTypeId: TlvField(3, TlvUInt32),
  certificateId: TlvField(4, TlvString.bound({ length: 19 })),
  securityLevel: TlvField(5, TlvUInt8),
  securityInformation: TlvField(6, TlvUInt16),
  versionNumber: TlvField(7, TlvUInt16),
  certificationType: TlvField(8, TlvUInt8),
  dacOriginVendorId: TlvOptionalField(9, TlvVendorId),
  dacOriginProductId: TlvOptionalField(10, TlvUInt16),
  authorizedPaaList: TlvOptionalField(
    11,
    TlvArray(TlvByteString.bound({ length: 20 }), { minLength: 1, maxLength: 10 })
  )
});
class CertificateManager {
  static rootCertToAsn1({
    serialNumber,
    notBefore,
    notAfter,
    issuer: { issuerRcacId },
    subject: { rcacId },
    ellipticCurvePublicKey,
    extensions: { subjectKeyIdentifier, authorityKeyIdentifier }
  }) {
    return DerCodec.encode({
      version: ContextTagged(0, 2),
      serialNumber: serialNumber[0],
      signatureAlgorithm: EcdsaWithSHA256_X962,
      issuer: {
        issuerRcacId: issuerRcacId === void 0 ? void 0 : RcacId_Matter(issuerRcacId)
      },
      validity: {
        notBefore: matterToJsDate(notBefore),
        notAfter: matterToJsDate(notAfter)
      },
      subject: {
        rcacId: RcacId_Matter(rcacId)
      },
      publicKey: PublicKeyEcPrime256v1_X962(ellipticCurvePublicKey),
      extensions: ContextTagged(3, {
        basicConstraints: BasicConstraints_X509({ isCa: true }),
        keyUsage: KeyUsage_Signature_ContentCommited_X509,
        subjectKeyIdentifier: SubjectKeyIdentifier_X509(subjectKeyIdentifier),
        authorityKeyIdentifier: AuthorityKeyIdentifier_X509(authorityKeyIdentifier)
      })
    });
  }
  static nocCertToAsn1({
    serialNumber,
    notBefore,
    notAfter,
    issuer: { issuerRcacId },
    subject: { fabricId, nodeId, caseAuthenticatedTags },
    ellipticCurvePublicKey,
    extensions: { subjectKeyIdentifier, authorityKeyIdentifier }
  }) {
    if (caseAuthenticatedTags !== void 0) {
      CaseAuthenticatedTag.validateNocTagList(caseAuthenticatedTags);
    }
    const cat0 = caseAuthenticatedTags?.[0];
    const cat1 = caseAuthenticatedTags?.[1];
    const cat2 = caseAuthenticatedTags?.[2];
    return DerCodec.encode({
      version: ContextTagged(0, 2),
      serialNumber: serialNumber[0],
      signatureAlgorithm: EcdsaWithSHA256_X962,
      issuer: {
        issuerRcacId: issuerRcacId === void 0 ? void 0 : RcacId_Matter(issuerRcacId)
      },
      validity: {
        notBefore: matterToJsDate(notBefore),
        notAfter: matterToJsDate(notAfter)
      },
      subject: {
        fabricId: FabricId_Matter(fabricId),
        nodeId: NodeId_Matter(NodeId(nodeId)),
        cat0: cat0 !== void 0 ? NocCat_Matter(cat0) : void 0,
        cat1: cat1 !== void 0 ? NocCat_Matter(cat1) : void 0,
        cat2: cat2 !== void 0 ? NocCat_Matter(cat2) : void 0
      },
      publicKey: PublicKeyEcPrime256v1_X962(ellipticCurvePublicKey),
      extensions: ContextTagged(3, {
        basicConstraints: BasicConstraints_X509({}),
        keyUsage: KeyUsage_Signature_X509,
        extendedKeyUsage: ExtendedKeyUsage_X509({ serverAuth: true, clientAuth: true }),
        subjectKeyIdentifier: SubjectKeyIdentifier_X509(subjectKeyIdentifier),
        authorityKeyIdentifier: AuthorityKeyIdentifier_X509(authorityKeyIdentifier)
      })
    });
  }
  static daCertToAsn1({
    serialNumber,
    notBefore,
    notAfter,
    issuer: { commonName: issuerCommonName, vendorId: issuerVendorId },
    subject: { commonName: subjectCommonName, vendorId: subjectVendorId, productId: subjectProductId },
    ellipticCurvePublicKey,
    extensions: { subjectKeyIdentifier, authorityKeyIdentifier }
  }, key) {
    const certificate = {
      version: ContextTagged(0, 2),
      serialNumber: serialNumber[0],
      signatureAlgorithm: EcdsaWithSHA256_X962,
      issuer: {
        commonName: CommonName_X520(issuerCommonName),
        vendorId: VendorId_Matter(issuerVendorId)
      },
      validity: {
        notBefore: matterToJsDate(notBefore),
        notAfter: matterToJsDate(notAfter)
      },
      subject: {
        commonName: CommonName_X520(subjectCommonName),
        vendorId: VendorId_Matter(subjectVendorId),
        productId: ProductId_Matter(subjectProductId)
      },
      publicKey: PublicKeyEcPrime256v1_X962(ellipticCurvePublicKey),
      extensions: ContextTagged(3, {
        basicConstraints: BasicConstraints_X509({
          isCa: false
        }),
        keyUsage: KeyUsage_Signature_X509,
        subjectKeyIdentifier: SubjectKeyIdentifier_X509(subjectKeyIdentifier),
        authorityKeyIdentifier: AuthorityKeyIdentifier_X509(authorityKeyIdentifier)
      })
    };
    return DerCodec.encode({
      certificate,
      signAlgorithm: EcdsaWithSHA256_X962,
      signature: BitByteArray(Crypto.sign(key, DerCodec.encode(certificate), "der"))
    });
  }
  static paiCertToAsn1({
    serialNumber,
    notBefore,
    notAfter,
    issuer: { commonName: issuerCommonName, vendorId: issuerVendorId },
    subject: { commonName, vendorId, productId },
    ellipticCurvePublicKey,
    extensions: { subjectKeyIdentifier, authorityKeyIdentifier }
  }, key) {
    const certificate = {
      version: ContextTagged(0, 2),
      serialNumber: serialNumber[0],
      signatureAlgorithm: EcdsaWithSHA256_X962,
      issuer: {
        commonName: CommonName_X520(issuerCommonName),
        vendorId: issuerVendorId === void 0 ? void 0 : VendorId_Matter(issuerVendorId)
      },
      validity: {
        notBefore: matterToJsDate(notBefore),
        notAfter: matterToJsDate(notAfter)
      },
      subject: {
        commonName: CommonName_X520(commonName),
        vendorId: VendorId_Matter(vendorId),
        productId: productId === void 0 ? void 0 : ProductId_Matter(productId)
      },
      publicKey: PublicKeyEcPrime256v1_X962(ellipticCurvePublicKey),
      extensions: ContextTagged(3, {
        basicConstraints: BasicConstraints_X509({
          isCa: true,
          pathLen: 0
        }),
        keyUsage: KeyUsage_Signature_ContentCommited_X509,
        subjectKeyIdentifier: SubjectKeyIdentifier_X509(subjectKeyIdentifier),
        authorityKeyIdentifier: AuthorityKeyIdentifier_X509(authorityKeyIdentifier)
      })
    };
    return DerCodec.encode({
      certificate,
      signAlgorithm: EcdsaWithSHA256_X962,
      signature: BitByteArray(Crypto.sign(key, DerCodec.encode(certificate), "der"))
    });
  }
  static paaCertToAsn1({
    serialNumber,
    notBefore,
    notAfter,
    issuer: { commonName: issuerCommonName, vendorId: issuerVendorId },
    subject: { commonName, vendorId },
    ellipticCurvePublicKey,
    extensions: { subjectKeyIdentifier, authorityKeyIdentifier }
  }, key) {
    const certificate = {
      version: ContextTagged(0, 2),
      serialNumber: serialNumber[0],
      signatureAlgorithm: EcdsaWithSHA256_X962,
      issuer: {
        commonName: CommonName_X520(issuerCommonName),
        vendorId: issuerVendorId === void 0 ? void 0 : VendorId_Matter(issuerVendorId)
      },
      validity: {
        notBefore: matterToJsDate(notBefore),
        notAfter: matterToJsDate(notAfter)
      },
      subject: {
        commonName: CommonName_X520(commonName),
        vendorId: vendorId === void 0 ? void 0 : VendorId_Matter(vendorId)
      },
      publicKey: PublicKeyEcPrime256v1_X962(ellipticCurvePublicKey),
      extensions: ContextTagged(3, {
        basicConstraints: BasicConstraints_X509({
          isCa: false
        }),
        keyUsage: KeyUsage_Signature_ContentCommited_X509,
        subjectKeyIdentifier: SubjectKeyIdentifier_X509(subjectKeyIdentifier),
        authorityKeyIdentifier: authorityKeyIdentifier === void 0 ? void 0 : AuthorityKeyIdentifier_X509(authorityKeyIdentifier)
      })
    };
    return DerCodec.encode({
      certificate,
      signAlgorithm: EcdsaWithSHA256_X962,
      signature: BitByteArray(Crypto.sign(key, DerCodec.encode(certificate), "der"))
    });
  }
  static CertificationDeclarationToAsn1(eContent, subjectKeyIdentifier, privateKey) {
    const certificate = {
      version: 3,
      digestAlgorithm: [SHA256_CMS],
      encapContentInfo: Pkcs7Data(eContent),
      signerInfo: [
        {
          version: 3,
          subjectKeyIdentifier: ContextTaggedBytes(0, subjectKeyIdentifier),
          digestAlgorithm: SHA256_CMS,
          signatureAlgorithm: EcdsaWithSHA256_X962,
          signature: Crypto.sign(privateKey, eContent, "der")
        }
      ]
    };
    return DerCodec.encode(Pkcs7SignedData(certificate));
  }
  static validateRootCertificate(rootCert) {
    Crypto.verify(PublicKey(rootCert.ellipticCurvePublicKey), this.rootCertToAsn1(rootCert), rootCert.signature);
  }
  static validateNocCertificate(rootCert, nocCert) {
    Crypto.verify(PublicKey(rootCert.ellipticCurvePublicKey), this.nocCertToAsn1(nocCert), nocCert.signature);
  }
  static createCertificateSigningRequest(key) {
    const request = {
      version: 0,
      subject: { organization: OrganisationName_X520("CSR") },
      publicKey: PublicKeyEcPrime256v1_X962(key.publicKey),
      endSignedBytes: ContextTagged(0)
    };
    return DerCodec.encode({
      request,
      signAlgorithm: EcdsaWithSHA256_X962,
      signature: BitByteArray(Crypto.sign(key, DerCodec.encode(request), "der"))
    });
  }
  static getPublicKeyFromCsr(csr) {
    const { [ELEMENTS_KEY]: rootElements } = DerCodec.decode(csr);
    if (rootElements?.length !== 3)
      throw new CertificateError("Invalid CSR data");
    const [requestNode, signAlgorithmNode, signatureNode] = rootElements;
    const { [ELEMENTS_KEY]: requestElements } = requestNode;
    if (requestElements?.length !== 4)
      throw new CertificateError("Invalid CSR data");
    const [versionNode, _subjectNode, publicKeyNode] = requestElements;
    const requestVersion = versionNode[BYTES_KEY][0];
    if (requestVersion !== 0)
      throw new CertificateError(`Unsupported request version${requestVersion}`);
    const { [ELEMENTS_KEY]: publicKeyElements } = publicKeyNode;
    if (publicKeyElements?.length !== 2)
      throw new CertificateError("Invalid CSR data");
    const [_publicKeyTypeNode, publicKeyBytesNode] = publicKeyElements;
    const publicKey = publicKeyBytesNode[BYTES_KEY];
    if (!EcdsaWithSHA256_X962[OBJECT_ID_KEY][BYTES_KEY].equals(signAlgorithmNode[ELEMENTS_KEY]?.[0]?.[BYTES_KEY]))
      throw new CertificateError("Unsupported signature type");
    Crypto.verify(PublicKey(publicKey), DerCodec.encode(requestNode), signatureNode[BYTES_KEY], "der");
    return publicKey;
  }
}
export {
  CertificateError,
  CertificateManager,
  CommonName_X520,
  FabricId_Matter,
  FirmwareSigningId_Matter,
  IcacId_Matter,
  NocCat_Matter,
  NodeId_Matter,
  ProductId_Matter,
  RcacId_Matter,
  TlvCertificationDeclaration,
  TlvIntermediateCertificate,
  TlvOperationalCertificate,
  TlvRootCertificate,
  VendorId_Matter,
  jsToMatterDate,
  matterToJsDate
};
