// @project-chip/matter.js/dist/esm/certificate/RootCertificateManager.js
/**
 * @license
 * Copyright 2022 The node-matter Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Crypto } from "../crypto/Crypto.js";
import { PrivateKey } from "../crypto/Key.js";
import { Time } from "../time/Time.js";
import { ByteArray } from "../util/ByteArray.js";
import {
  CertificateManager,
  TlvOperationalCertificate,
  TlvRootCertificate,
  jsToMatterDate
} from "./CertificateManager.js";
class RootCertificateManager {
  constructor(storageContext) {
    this.rootCertId = BigInt(0);
    this.rootKeyPair = Crypto.createKeyPair();
    this.rootKeyIdentifier = Crypto.hash(this.rootKeyPair.publicKey).slice(0, 20);
    this.rootCertBytes = this.generateRootCert();
    this.nextCertificateId = 1;
    const storage = storageContext.createContext("RootCertificateManager");
    if (storage.has("rootCertId")) {
      try {
        const rootCertId = storage.get("rootCertId");
        const rootKeyPair = storage.get("rootKeyPair");
        const rootKeyIdentifier = storage.get("rootKeyIdentifier");
        const rootCertBytes = storage.get("rootCertBytes");
        const nextCertificateId = storage.get("nextCertificateId");
        this.rootCertId = rootCertId;
        this.rootKeyPair = PrivateKey(rootKeyPair);
        this.rootKeyIdentifier = rootKeyIdentifier;
        this.rootCertBytes = rootCertBytes;
        this.nextCertificateId = nextCertificateId;
        return;
      } catch (error) {
        console.error("Failed to load root certificate from storage, generating new one", error);
      }
    }
    storage.set("rootCertId", this.rootCertId);
    storage.set("rootKeyPair", this.rootKeyPair.keyPair);
    storage.set("rootKeyIdentifier", this.rootKeyIdentifier);
    storage.set("rootCertBytes", this.rootCertBytes);
    storage.set("nextCertificateId", this.nextCertificateId);
  }
  getRootCert() {
    return this.rootCertBytes;
  }
  generateRootCert() {
    const now = Time.get().now();
    const unsignedCertificate = {
      serialNumber: ByteArray.of(Number(this.rootCertId)),
      signatureAlgorithm: 1,
      publicKeyAlgorithm: 1,
      ellipticCurveIdentifier: 1,
      issuer: { issuerRcacId: this.rootCertId },
      notBefore: jsToMatterDate(now, -1),
      notAfter: jsToMatterDate(now, 10),
      subject: { rcacId: this.rootCertId },
      ellipticCurvePublicKey: this.rootKeyPair.publicKey,
      extensions: {
        basicConstraints: { isCa: true },
        keyUsage: 96,
        subjectKeyIdentifier: this.rootKeyIdentifier,
        authorityKeyIdentifier: this.rootKeyIdentifier
      }
    };
    const signature = Crypto.sign(this.rootKeyPair, CertificateManager.rootCertToAsn1(unsignedCertificate));
    return TlvRootCertificate.encode({ ...unsignedCertificate, signature });
  }
  generateNoc(publicKey, fabricId, nodeId, caseAuthenticatedTags) {
    const now = Time.get().now();
    const certId = this.nextCertificateId++;
    const unsignedCertificate = {
      serialNumber: ByteArray.of(certId),
      // TODO: figure out what should happen if certId > 255
      signatureAlgorithm: 1,
      publicKeyAlgorithm: 1,
      ellipticCurveIdentifier: 1,
      issuer: { issuerRcacId: this.rootCertId },
      notBefore: jsToMatterDate(now, -1),
      notAfter: jsToMatterDate(now, 10),
      subject: { fabricId, nodeId, caseAuthenticatedTags },
      ellipticCurvePublicKey: publicKey,
      extensions: {
        basicConstraints: { isCa: false },
        keyUsage: 1,
        extendedKeyUsage: [2, 1],
        subjectKeyIdentifier: Crypto.hash(publicKey).slice(0, 20),
        authorityKeyIdentifier: this.rootKeyIdentifier
      }
    };
    const signature = Crypto.sign(this.rootKeyPair, CertificateManager.nocCertToAsn1(unsignedCertificate));
    return TlvOperationalCertificate.encode({ ...unsignedCertificate, signature });
  }
}
export {
  RootCertificateManager
};
