// @project-chip/matter.js/dist/esm/protocol/ChannelManager.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterError } from "../common/MatterError.js";
import { Logger } from "../log/Logger.js";
import { MessageChannel } from "./ExchangeManager.js";
const logger = Logger.get("ChannelManager");
class NoChannelError extends MatterError {
}
class ChannelManager {
  constructor() {
    this.channels = /* @__PURE__ */ new Map();
    this.paseChannels = /* @__PURE__ */ new Map();
  }
  getChannelKey(fabric, nodeId) {
    return `${fabric.fabricId}/${nodeId}`;
  }
  async setChannel(fabric, nodeId, channel) {
    const channelKey = this.getChannelKey(fabric, nodeId);
    const currentChannel = this.channels.get(channelKey);
    if (currentChannel !== void 0) {
      const { session } = currentChannel;
      if (session.getId() !== channel.session.getId()) {
        logger.debug(
          `Existing channel for fabricIndex ${fabric.fabricIndex} and node ${nodeId} with session ${session.name} gets replaced. Consider former session already inactive and so close it.`
        );
        await session.destroy(false, false);
      }
      await currentChannel.close();
    }
    this.channels.set(channelKey, channel);
  }
  getChannel(fabric, nodeId) {
    const result = this.channels.get(this.getChannelKey(fabric, nodeId));
    if (result === void 0)
      throw new NoChannelError(`Can't find a channel to node ${nodeId}`);
    return result;
  }
  getChannelForSession(session) {
    if (session.isSecure() && !session.isPase()) {
      const secureSession = session;
      const fabric = secureSession.getFabric();
      const nodeId = secureSession.getPeerNodeId();
      if (fabric === void 0) {
        return this.paseChannels.get(session);
      }
      const channel = this.getChannel(fabric, nodeId);
      if (channel.session.getId() !== session.getId()) {
        return void 0;
      }
      return channel;
    }
    return this.paseChannels.get(session);
  }
  async removeChannel(fabric, nodeId) {
    const channelKey = this.getChannelKey(fabric, nodeId);
    const channel = this.channels.get(channelKey);
    if (channel !== void 0) {
      await channel.close();
      this.channels.delete(channelKey);
    }
  }
  getOrCreateAsPaseChannel(byteArrayChannel, session) {
    const msgChannel = new MessageChannel(
      byteArrayChannel,
      session,
      async () => void this.paseChannels.delete(session)
    );
    this.paseChannels.set(session, msgChannel);
    return msgChannel;
  }
  async getOrCreateChannel(byteArrayChannel, session) {
    if (!session.isSecure()) {
      return this.getOrCreateAsPaseChannel(byteArrayChannel, session);
    }
    const secureSession = session;
    const fabric = secureSession.getFabric();
    const nodeId = secureSession.getPeerNodeId();
    if (fabric === void 0) {
      return this.getOrCreateAsPaseChannel(byteArrayChannel, session);
    }
    let result = this.channels.get(this.getChannelKey(fabric, nodeId));
    if (result === void 0 || result.session.getId() !== session.getId()) {
      result = new MessageChannel(
        byteArrayChannel,
        session,
        async () => await this.removeChannel(fabric, nodeId)
      );
      await this.setChannel(fabric, nodeId, result);
    }
    return result;
  }
  async close() {
    for (const channel of this.paseChannels.values()) {
      await channel.close();
    }
    this.paseChannels.clear();
    for (const channel of this.channels.values()) {
      await channel.close();
    }
    this.channels.clear();
  }
}
export {
  ChannelManager,
  NoChannelError
};
