// @project-chip/matter.js/dist/esm/protocol/MessageExchange.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MessageCodec, SessionType } from "../codec/MessageCodec.js";
import { MatterError, MatterFlowError } from "../common/MatterError.js";
import { Logger } from "../log/Logger.js";
import {
  MRP_MAX_TRANSMISSIONS,
  SESSION_ACTIVE_INTERVAL_MS,
  SESSION_ACTIVE_THRESHOLD_MS,
  SESSION_IDLE_INTERVAL_MS
} from "../session/Session.js";
import { Time } from "../time/Time.js";
import { ByteArray } from "../util/ByteArray.js";
import { createPromise } from "../util/Promises.js";
import { Queue } from "../util/Queue.js";
import { StatusCode, StatusResponseError } from "./interaction/StatusCode.js";
import { MessageType, SECURE_CHANNEL_PROTOCOL_ID } from "./securechannel/SecureChannelMessages.js";
import { SecureChannelProtocol } from "./securechannel/SecureChannelProtocol.js";
const logger = Logger.get("MessageExchange");
class RetransmissionLimitReachedError extends MatterError {
}
class UnexpectedMessageError extends MatterError {
  constructor(message, receivedMessage) {
    super(`(${MessageCodec.messageDiagnostics(receivedMessage)}) ${message}`);
    this.receivedMessage = receivedMessage;
  }
}
const MRP_BACKOFF_BASE = 1.6;
const MRP_BACKOFF_JITTER = 0.25;
const MRP_BACKOFF_MARGIN = 1.1;
const MRP_BACKOFF_THRESHOLD = 1;
const MAXIMUM_TRANSMISSION_TIME_MS = 9495;
const MRP_STANDALONE_ACK_TIMEOUT_MS = 200;
class MessageExchange {
  constructor(session, channel, isInitiator, peerSessionId, nodeId, peerNodeId, exchangeId, protocolId, closeCallback) {
    this.session = session;
    this.channel = channel;
    this.isInitiator = isInitiator;
    this.peerSessionId = peerSessionId;
    this.nodeId = nodeId;
    this.peerNodeId = peerNodeId;
    this.exchangeId = exchangeId;
    this.protocolId = protocolId;
    this.closeCallback = closeCallback;
    this.messagesQueue = new Queue();
    this.receivedMessageAckTimer = Time.getTimer(MRP_STANDALONE_ACK_TIMEOUT_MS, () => {
      if (this.receivedMessageToAck !== void 0) {
        const messageToAck = this.receivedMessageToAck;
        this.receivedMessageToAck = void 0;
        this.sendStandaloneAckForMessage(messageToAck).catch(
          (error) => logger.error("An error happened when sending a standalone ack", error)
        );
      }
    });
    const { activeIntervalMs, idleIntervalMs, activeThresholdMs } = session.getSessionParameters();
    this.activeIntervalMs = activeIntervalMs ?? SESSION_ACTIVE_INTERVAL_MS;
    this.idleIntervalMs = idleIntervalMs ?? SESSION_IDLE_INTERVAL_MS;
    this.activeThresholdMs = activeThresholdMs ?? SESSION_ACTIVE_THRESHOLD_MS;
    this.retransmissionRetries = MRP_MAX_TRANSMISSIONS;
    logger.debug(
      "New exchange",
      Logger.dict({
        protocol: this.protocolId,
        id: this.exchangeId,
        session: session.name,
        peerSessionId: this.peerSessionId,
        "active threshold ms": this.activeThresholdMs,
        "active interval ms": this.activeIntervalMs,
        "idle interval ms": this.idleIntervalMs,
        retries: this.retransmissionRetries
      })
    );
  }
  static fromInitialMessage(channel, initialMessage, closeCallback) {
    const { session } = channel;
    return new MessageExchange(
      session,
      channel,
      false,
      session.getId(),
      initialMessage.packetHeader.destNodeId,
      initialMessage.packetHeader.sourceNodeId,
      initialMessage.payloadHeader.exchangeId,
      initialMessage.payloadHeader.protocolId,
      closeCallback
    );
  }
  static initiate(channel, exchangeId, protocolId, closeCallback) {
    const { session } = channel;
    return new MessageExchange(
      session,
      channel,
      true,
      session.getPeerSessionId(),
      session.getNodeId(),
      session.getPeerNodeId(),
      exchangeId,
      protocolId,
      closeCallback
    );
  }
  async sendStandaloneAckForMessage(message) {
    const {
      packetHeader: { messageId },
      payloadHeader: { requiresAck }
    } = message;
    if (!requiresAck)
      return;
    await this.send(MessageType.StandaloneAck, new ByteArray(0), { includeAcknowledgeMessageId: messageId });
  }
  async onMessageReceived(message, isDuplicate = false) {
    const {
      packetHeader: { messageId },
      payloadHeader: { requiresAck, ackedMessageId, protocolId, messageType }
    } = message;
    logger.debug("Message \xAB", MessageCodec.messageDiagnostics(message, isDuplicate));
    this.session.notifyActivity(true);
    if (isDuplicate) {
      if (requiresAck) {
        await this.sendStandaloneAckForMessage(message);
      }
      return;
    }
    if (messageId === this.sentMessageToAck?.payloadHeader.ackedMessageId) {
      await this.channel.send(this.sentMessageToAck);
      return;
    }
    const sentMessageIdToAck = this.sentMessageToAck?.packetHeader.messageId;
    if (sentMessageIdToAck !== void 0) {
      if (ackedMessageId === void 0) {
        throw new MatterFlowError("Previous message ack is missing");
      } else if (ackedMessageId !== sentMessageIdToAck) {
        if (SecureChannelProtocol.isStandaloneAck(protocolId, messageType)) {
        } else {
          throw new MatterFlowError(
            `Incorrect ack received. Expected ${sentMessageIdToAck}, received: ${ackedMessageId}`
          );
        }
      } else {
        this.retransmissionTimer?.stop();
        this.sentMessageAckSuccess?.(message);
        this.sentMessageAckSuccess = void 0;
        this.sentMessageAckFailure = void 0;
        this.sentMessageToAck = void 0;
      }
    }
    if (SecureChannelProtocol.isStandaloneAck(protocolId, messageType)) {
      return;
    }
    if (protocolId !== this.protocolId) {
      throw new MatterFlowError(
        `Received a message for an unexpected protocol. Expected: ${this.protocolId}, received: ${protocolId}`
      );
    }
    if (requiresAck) {
      if (this.receivedMessageToAck !== void 0) {
        this.receivedMessageAckTimer.stop();
        await this.sendStandaloneAckForMessage(this.receivedMessageToAck);
        return;
      }
      this.receivedMessageToAck = message;
      this.receivedMessageAckTimer.start();
    }
    await this.messagesQueue.write(message);
  }
  async send(messageType, payload, options) {
    const {
      expectAckOnly = false,
      minimumResponseTimeoutMs,
      requiresAck,
      includeAcknowledgeMessageId
    } = options ?? {};
    if (messageType === MessageType.StandaloneAck && requiresAck) {
      throw new MatterFlowError("A standalone ack may not require acknowledgement.");
    }
    if (this.sentMessageToAck !== void 0 && messageType !== MessageType.StandaloneAck)
      throw new MatterFlowError("The previous message has not been acked yet, cannot send a new message.");
    this.session.notifyActivity(false);
    let ackedMessageId = includeAcknowledgeMessageId;
    if (ackedMessageId === void 0) {
      ackedMessageId = this.receivedMessageToAck?.packetHeader.messageId;
      if (ackedMessageId !== void 0 && messageType !== MessageType.StandaloneAck) {
        this.receivedMessageAckTimer.stop();
        this.receivedMessageToAck = void 0;
      }
    }
    const message = {
      packetHeader: {
        sessionId: this.peerSessionId,
        sessionType: SessionType.Unicast,
        // TODO: support multicast/groups
        messageId: this.session.getIncrementedMessageCounter(),
        destNodeId: this.peerNodeId,
        sourceNodeId: this.nodeId,
        hasPrivacyEnhancements: false,
        isControlMessage: false,
        hasMessageExtensions: false
      },
      payloadHeader: {
        exchangeId: this.exchangeId,
        protocolId: messageType === MessageType.StandaloneAck ? SECURE_CHANNEL_PROTOCOL_ID : this.protocolId,
        messageType,
        isInitiatorMessage: this.isInitiator,
        requiresAck: requiresAck ?? messageType !== MessageType.StandaloneAck,
        ackedMessageId,
        hasSecuredExtension: false
      },
      payload
    };
    let ackPromise;
    if (message.payloadHeader.requiresAck) {
      this.sentMessageToAck = message;
      this.retransmissionTimer = Time.getTimer(
        this.getResubmissionBackOffTime(0),
        () => this.retransmitMessage(
          message,
          0,
          minimumResponseTimeoutMs !== void 0 ? Time.nowMs() + minimumResponseTimeoutMs : void 0
        )
      );
      const { promise, resolver, rejecter } = createPromise();
      ackPromise = promise;
      this.sentMessageAckSuccess = resolver;
      this.sentMessageAckFailure = rejecter;
    }
    await this.channel.send(message);
    if (ackPromise !== void 0) {
      this.retransmissionTimer?.start();
      const responseMessage = await ackPromise;
      this.sentMessageAckSuccess = void 0;
      this.sentMessageAckFailure = void 0;
      const {
        payloadHeader: { protocolId, messageType: messageType2 }
      } = responseMessage;
      if (expectAckOnly && !SecureChannelProtocol.isStandaloneAck(protocolId, messageType2)) {
        throw new UnexpectedMessageError("Expected ack only", responseMessage);
      }
    }
  }
  nextMessage() {
    return this.messagesQueue.read();
  }
  async waitFor(messageType, timeoutMs = 18e4) {
    const message = await this.messagesQueue.read(timeoutMs);
    const {
      payloadHeader: { messageType: receivedMessageType }
    } = message;
    if (receivedMessageType !== messageType)
      throw new MatterFlowError(
        `Received unexpected message type ${receivedMessageType.toString(16)}. Expected ${messageType.toString(
          16
        )}`
      );
    return message;
  }
  /** @see {@link MatterCoreSpecificationV1_0}, section 4.11.2.1 */
  getResubmissionBackOffTime(retransmissionCount) {
    const baseInterval = this.session.isPeerActive() ? this.activeIntervalMs : this.idleIntervalMs;
    return Math.floor(
      MRP_BACKOFF_MARGIN * baseInterval * Math.pow(MRP_BACKOFF_BASE, Math.max(0, retransmissionCount - MRP_BACKOFF_THRESHOLD)) * (1 + Math.random() * MRP_BACKOFF_JITTER)
    );
  }
  retransmitMessage(message, retransmissionCount, notTimeoutBeforeTimeMs) {
    retransmissionCount++;
    if (retransmissionCount >= this.retransmissionRetries && (notTimeoutBeforeTimeMs === void 0 || Time.nowMs() > notTimeoutBeforeTimeMs)) {
      if (this.sentMessageToAck !== void 0 && this.sentMessageAckFailure !== void 0) {
        this.receivedMessageToAck = void 0;
        this.sentMessageAckFailure(new RetransmissionLimitReachedError());
        this.sentMessageAckFailure = void 0;
        this.sentMessageAckSuccess = void 0;
      }
      if (this.closeTimer !== void 0) {
        this.closeInternal().catch((error) => logger.error("An error happened when closing the exchange", error));
      }
      return;
    }
    this.session.notifyActivity(false);
    if (retransmissionCount === 1) {
    }
    const resubmissionBackoffTime = this.getResubmissionBackOffTime(retransmissionCount);
    logger.debug(
      `Resubmit message ${message.packetHeader.messageId} (retransmission attempt ${retransmissionCount}, next backoff time ${resubmissionBackoffTime}ms))`
    );
    this.channel.send(message).then(() => {
      this.retransmissionTimer = Time.getTimer(
        resubmissionBackoffTime,
        () => this.retransmitMessage(message, retransmissionCount, notTimeoutBeforeTimeMs)
      ).start();
    }).catch((error) => logger.error("An error happened when retransmitting a message", error));
  }
  async destroy() {
    if (this.closeTimer === void 0 && this.receivedMessageToAck !== void 0) {
      this.receivedMessageAckTimer.stop();
      const messageToAck = this.receivedMessageToAck;
      this.receivedMessageToAck = void 0;
      try {
        await this.sendStandaloneAckForMessage(messageToAck);
      } catch (error) {
        logger.error("An error happened when closing the exchange", error);
      }
    }
    await this.closeInternal();
  }
  startTimedInteraction(timeoutMs) {
    if (this.timedInteractionTimer !== void 0 && this.timedInteractionTimer.isRunning) {
      this.timedInteractionTimer.stop();
      throw new StatusResponseError(
        "Timed interaction already running for this exchange",
        StatusCode.InvalidAction
      );
    }
    logger.debug(
      `Starting timed interaction with Transaction ID ${this.exchangeId} for ${timeoutMs}ms from ${this.channel.name}`
    );
    this.timedInteractionTimer = Time.getTimer(timeoutMs, () => {
      logger.debug(
        `Timed interaction with Transaction ID ${this.exchangeId} from ${this.channel.name} timed out`
      );
    }).start();
  }
  clearTimedInteraction() {
    if (this.timedInteractionTimer !== void 0) {
      logger.debug(`Clearing timed interaction with Transaction ID ${this.exchangeId} from ${this.channel.name}`);
      this.timedInteractionTimer.stop();
      this.timedInteractionTimer = void 0;
    }
  }
  hasTimedInteraction() {
    return this.timedInteractionTimer !== void 0;
  }
  hasActiveTimedInteraction() {
    return this.timedInteractionTimer !== void 0 && this.timedInteractionTimer.isRunning;
  }
  hasExpiredTimedInteraction() {
    return this.timedInteractionTimer !== void 0 && !this.timedInteractionTimer.isRunning;
  }
  async close() {
    if (this.closeTimer !== void 0)
      return;
    this.closeTimer = Time.getTimer(MAXIMUM_TRANSMISSION_TIME_MS, async () => await this.closeInternal()).start();
    if (this.receivedMessageToAck !== void 0) {
      this.receivedMessageAckTimer.stop();
      const messageToAck = this.receivedMessageToAck;
      this.receivedMessageToAck = void 0;
      try {
        await this.sendStandaloneAckForMessage(messageToAck);
      } catch (error) {
        logger.error("An error happened when closing the exchange", error);
      }
    }
  }
  async closeInternal() {
    this.retransmissionTimer?.stop();
    this.closeTimer?.stop();
    this.timedInteractionTimer?.stop();
    this.messagesQueue.close();
    await this.closeCallback();
  }
}
export {
  MessageExchange,
  RetransmissionLimitReachedError,
  UnexpectedMessageError
};
