// @project-chip/matter.js/dist/cjs/fabric/Fabric.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Fabric_exports = {};
__export(Fabric_exports, {
  Fabric: () => Fabric,
  FabricBuilder: () => FabricBuilder
});
module.exports = __toCommonJS(Fabric_exports);
var import_CertificateManager = require("../certificate/CertificateManager.js");
var import_MatterError = require("../common/MatterError.js");
var import_Crypto = require("../crypto/Crypto.js");
var import_Key = require("../crypto/Key.js");
var import_CaseAuthenticatedTag = require("../datatype/CaseAuthenticatedTag.js");
var import_FabricId = require("../datatype/FabricId.js");
var import_Logger = require("../log/Logger.js");
var import_ByteArray = require("../util/ByteArray.js");
var import_DataWriter = require("../util/DataWriter.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("Fabric");
const COMPRESSED_FABRIC_ID_INFO = import_ByteArray.ByteArray.fromString("CompressedFabric");
const GROUP_SECURITY_INFO = import_ByteArray.ByteArray.fromString("GroupKey v1.0");
class Fabric {
  constructor(fabricIndex, fabricId, nodeId, rootNodeId, operationalId, rootPublicKey, keyPair, rootVendorId, rootCert, identityProtectionKey, operationalIdentityProtectionKey, intermediateCACert, operationalCert, label, caseAuthenticatedTags = new Array(), scopedClusterData) {
    this.fabricIndex = fabricIndex;
    this.fabricId = fabricId;
    this.nodeId = nodeId;
    this.rootNodeId = rootNodeId;
    this.operationalId = operationalId;
    this.rootPublicKey = rootPublicKey;
    this.keyPair = keyPair;
    this.rootVendorId = rootVendorId;
    this.rootCert = rootCert;
    this.identityProtectionKey = identityProtectionKey;
    this.operationalIdentityProtectionKey = operationalIdentityProtectionKey;
    this.intermediateCACert = intermediateCACert;
    this.operationalCert = operationalCert;
    this.label = label;
    this.caseAuthenticatedTags = caseAuthenticatedTags;
    this.sessions = new Array();
    this.removeCallbacks = new Array();
    this.scopedClusterData = scopedClusterData ?? /* @__PURE__ */ new Map();
  }
  toStorageObject() {
    return {
      fabricIndex: this.fabricIndex,
      fabricId: this.fabricId,
      nodeId: this.nodeId,
      rootNodeId: this.rootNodeId,
      operationalId: this.operationalId,
      rootPublicKey: this.rootPublicKey,
      keyPair: this.keyPair.keyPair,
      rootVendorId: this.rootVendorId,
      rootCert: this.rootCert,
      identityProtectionKey: this.identityProtectionKey,
      operationalIdentityProtectionKey: this.operationalIdentityProtectionKey,
      intermediateCACert: this.intermediateCACert,
      operationalCert: this.operationalCert,
      label: this.label,
      caseAuthenticatedTags: this.caseAuthenticatedTags,
      scopedClusterData: this.scopedClusterData
    };
  }
  static createFromStorageObject(fabricObject) {
    return new Fabric(
      fabricObject.fabricIndex,
      fabricObject.fabricId,
      fabricObject.nodeId,
      fabricObject.rootNodeId,
      fabricObject.operationalId,
      fabricObject.rootPublicKey,
      (0, import_Key.PrivateKey)(fabricObject.keyPair),
      fabricObject.rootVendorId,
      fabricObject.rootCert,
      fabricObject.identityProtectionKey,
      fabricObject.operationalIdentityProtectionKey,
      fabricObject.intermediateCACert,
      fabricObject.operationalCert,
      fabricObject.label,
      fabricObject.caseAuthenticatedTags,
      fabricObject.scopedClusterData
    );
  }
  setLabel(label) {
    this.label = label;
    this.persist();
  }
  getPublicKey() {
    return this.keyPair.publicKey;
  }
  sign(data) {
    return import_Crypto.Crypto.sign(this.keyPair, data);
  }
  verifyCredentials(_operationalCert, _intermediateCACert) {
    return;
  }
  matchesFabricIdAndRootPublicKey(fabricId, rootPublicKey) {
    return this.fabricId === fabricId && this.rootPublicKey?.equals(rootPublicKey);
  }
  getDestinationId(nodeId, random) {
    const writer = new import_DataWriter.DataWriter(import_ByteArray.Endian.Little);
    writer.writeByteArray(random);
    writer.writeByteArray(this.rootPublicKey);
    writer.writeUInt64(this.fabricId);
    writer.writeUInt64(nodeId);
    return import_Crypto.Crypto.hmac(this.operationalIdentityProtectionKey, writer.toByteArray());
  }
  addSession(session) {
    this.sessions.push(session);
  }
  removeSession(session) {
    const index = this.sessions.indexOf(session);
    if (index >= 0) {
      this.sessions.splice(index, 1);
    }
  }
  addRemoveCallback(callback) {
    this.removeCallbacks.push(callback);
  }
  deleteRemoveCallback(callback) {
    const index = this.removeCallbacks.indexOf(callback);
    if (index >= 0) {
      this.removeCallbacks.splice(index, 1);
    }
  }
  setPersistCallback(callback) {
    this.persistCallback = callback;
  }
  async remove(currentSessionId) {
    this.removeCallbacks.forEach((callback) => callback());
    for (const session of [...this.sessions]) {
      await session.destroy(false, session.getId() === currentSessionId);
    }
  }
  persist() {
    this.persistCallback?.();
  }
  getScopedClusterDataValue(cluster, clusterDataKey) {
    const dataMap = this.scopedClusterData.get(cluster.id);
    if (dataMap === void 0) {
      return void 0;
    }
    return dataMap.get(clusterDataKey);
  }
  setScopedClusterDataValue(cluster, clusterDataKey, value) {
    if (!this.scopedClusterData.has(cluster.id)) {
      this.scopedClusterData.set(cluster.id, /* @__PURE__ */ new Map());
    }
    this.scopedClusterData.get(cluster.id).set(clusterDataKey, value);
    this.persist();
  }
  deleteScopedClusterDataValue(cluster, clusterDataKey) {
    if (!this.scopedClusterData.has(cluster.id)) {
      return;
    }
    this.scopedClusterData.get(cluster.id).delete(clusterDataKey);
    this.persist();
  }
  hasScopedClusterDataValue(cluster, clusterDataKey) {
    return this.scopedClusterData.has(cluster.id) && this.scopedClusterData.get(cluster.id).has(clusterDataKey);
  }
  deleteScopedClusterData(cluster) {
    this.scopedClusterData.delete(cluster.id);
    this.persist();
  }
  getScopedClusterDataKeys(cluster) {
    if (!this.scopedClusterData.has(cluster.id)) {
      return [];
    }
    return Array.from(this.scopedClusterData.get(cluster.id).keys());
  }
  getExternalInformation() {
    return {
      fabricIndex: this.fabricIndex,
      fabricId: this.fabricId,
      nodeId: this.nodeId,
      rootNodeId: this.rootNodeId,
      rootVendorId: this.rootVendorId,
      label: this.label
    };
  }
}
class FabricBuilder {
  constructor() {
    this.keyPair = import_Crypto.Crypto.createKeyPair();
    this.label = "";
    this.caseAuthenticatedTags = new Array();
  }
  getPublicKey() {
    return this.keyPair.publicKey;
  }
  getFabricIndex() {
    return this.fabricIndex;
  }
  createCertificateSigningRequest() {
    return import_CertificateManager.CertificateManager.createCertificateSigningRequest(this.keyPair);
  }
  setRootCert(rootCert) {
    this.rootCert = rootCert;
    this.rootPublicKey = import_CertificateManager.TlvRootCertificate.decode(rootCert).ellipticCurvePublicKey;
    return this;
  }
  hasRootCert() {
    return this.rootCert !== void 0;
  }
  setOperationalCert(operationalCert) {
    this.operationalCert = operationalCert;
    const {
      subject: { nodeId, fabricId, caseAuthenticatedTags }
    } = import_CertificateManager.TlvOperationalCertificate.decode(operationalCert);
    logger.debug(`FabricBuilder setOperationalCert: nodeId=${nodeId}, fabricId=${fabricId}`);
    this.fabricId = (0, import_FabricId.FabricId)(fabricId);
    this.nodeId = nodeId;
    if (caseAuthenticatedTags !== void 0) {
      import_CaseAuthenticatedTag.CaseAuthenticatedTag.validateNocTagList(caseAuthenticatedTags);
      this.caseAuthenticatedTags = caseAuthenticatedTags;
    }
    return this;
  }
  setIntermediateCACert(certificate) {
    this.intermediateCACert = certificate;
    return this;
  }
  setRootVendorId(rootVendorId) {
    this.rootVendorId = rootVendorId;
    return this;
  }
  setRootNodeId(rootNodeId) {
    this.rootNodeId = rootNodeId;
    return this;
  }
  setIdentityProtectionKey(key) {
    this.identityProtectionKey = key;
    return this;
  }
  initializeFromFabricForUpdate(fabric) {
    this.rootVendorId = fabric.rootVendorId;
    this.rootNodeId = fabric.rootNodeId;
    this.identityProtectionKey = fabric.identityProtectionKey;
    this.rootCert = fabric.rootCert;
    this.rootPublicKey = fabric.rootPublicKey;
    this.label = fabric.label;
  }
  matchesToFabric(fabric) {
    if (this.fabricId === void 0 || this.rootPublicKey === void 0) {
      throw new import_MatterError.MatterFlowError("Node Operational Data needs to be set first.");
    }
    return fabric.matchesFabricIdAndRootPublicKey(this.fabricId, this.rootPublicKey);
  }
  getNodeId() {
    return this.nodeId;
  }
  getFabricId() {
    return this.fabricId;
  }
  async build(fabricIndex) {
    if (this.fabricIndex !== void 0)
      throw new import_MatterError.InternalError("FabricBuilder can only be built once");
    if (this.rootNodeId === void 0)
      throw new import_MatterError.InternalError("rootNodeId needs to be set");
    if (this.rootVendorId === void 0)
      throw new import_MatterError.InternalError("vendorId needs to be set");
    if (this.rootCert === void 0 || this.rootPublicKey === void 0)
      throw new import_MatterError.InternalError("rootCert needs to be set");
    if (this.identityProtectionKey === void 0)
      throw new import_MatterError.InternalError("identityProtectionKey needs to be set");
    if (this.operationalCert === void 0 || this.fabricId === void 0 || this.nodeId === void 0)
      throw new import_MatterError.InternalError("operationalCert needs to be set");
    this.fabricIndex = fabricIndex;
    const saltWriter = new import_DataWriter.DataWriter(import_ByteArray.Endian.Big);
    saltWriter.writeUInt64(this.fabricId);
    const operationalId = await import_Crypto.Crypto.hkdf(
      this.rootPublicKey.slice(1),
      saltWriter.toByteArray(),
      COMPRESSED_FABRIC_ID_INFO,
      8
    );
    return new Fabric(
      this.fabricIndex,
      this.fabricId,
      this.nodeId,
      this.rootNodeId,
      operationalId,
      this.rootPublicKey,
      this.keyPair,
      this.rootVendorId,
      this.rootCert,
      this.identityProtectionKey,
      await import_Crypto.Crypto.hkdf(this.identityProtectionKey, operationalId, GROUP_SECURITY_INFO, 16),
      this.intermediateCACert,
      this.operationalCert,
      this.label,
      this.caseAuthenticatedTags
    );
  }
}
