// @project-chip/matter.js/dist/cjs/fabric/FabricManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FabricManager_exports = {};
__export(FabricManager_exports, {
  FabricManager: () => FabricManager,
  FabricNotFoundError: () => FabricNotFoundError,
  FabricTableFullError: () => FabricTableFullError
});
module.exports = __toCommonJS(FabricManager_exports);
var import_MatterError = require("../common/MatterError.js");
var import_FabricIndex = require("../datatype/FabricIndex.js");
var import_Fabric = require("./Fabric.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class FabricNotFoundError extends import_MatterError.MatterError {
}
class FabricTableFullError extends import_MatterError.MatterError {
}
class FabricManager {
  constructor(storage, fabricRemoveCallback) {
    this.fabricRemoveCallback = fabricRemoveCallback;
    this.nextFabricIndex = 1;
    this.fabrics = /* @__PURE__ */ new Map();
    this.fabricStorage = storage.createContext("FabricManager");
    const fabrics = this.fabricStorage.get("fabrics", []);
    fabrics.forEach((fabric) => this.addFabric(import_Fabric.Fabric.createFromStorageObject(fabric)));
    this.nextFabricIndex = this.fabricStorage.get("nextFabricIndex", this.nextFabricIndex);
  }
  getNextFabricIndex() {
    for (let i = 0; i < 254; i++) {
      const fabricIndex = this.nextFabricIndex++;
      if (this.nextFabricIndex > 254)
        this.nextFabricIndex = 1;
      if (!this.fabrics.has((0, import_FabricIndex.FabricIndex)(fabricIndex))) {
        return (0, import_FabricIndex.FabricIndex)(fabricIndex);
      }
    }
    throw new FabricTableFullError("No free fabric index available.");
  }
  persistFabrics() {
    this.fabricStorage.set(
      "fabrics",
      Array.from(this.fabrics.values()).map((fabric) => fabric.toStorageObject())
    );
    this.fabricStorage.set("nextFabricIndex", this.nextFabricIndex);
  }
  addFabric(fabric) {
    const { fabricIndex } = fabric;
    if (this.fabrics.has(fabricIndex)) {
      throw new import_MatterError.MatterFlowError(`Fabric with index ${fabricIndex} already exists.`);
    }
    this.fabrics.set(fabricIndex, fabric);
    fabric.addRemoveCallback(() => this.removeFabric(fabricIndex));
    fabric.setPersistCallback(() => this.persistFabrics());
  }
  removeFabric(fabricIndex) {
    const fabric = this.fabrics.get(fabricIndex);
    if (fabric === void 0)
      throw new FabricNotFoundError(
        `Fabric with index ${fabricIndex} cannot be removed because it does not exist.`
      );
    this.fabrics.delete(fabricIndex);
    this.persistFabrics();
    this.fabricRemoveCallback?.(fabricIndex, fabric.rootNodeId);
  }
  getFabrics() {
    return Array.from(this.fabrics.values());
  }
  findFabricFromDestinationId(destinationId, initiatorRandom) {
    for (const fabric of this.fabrics.values()) {
      const candidateDestinationId = fabric.getDestinationId(fabric.nodeId, initiatorRandom);
      if (!candidateDestinationId.equals(destinationId))
        continue;
      return fabric;
    }
    throw new import_MatterError.InternalError("Fabric cannot be found from destinationId");
  }
  updateFabric(fabric) {
    const { fabricIndex } = fabric;
    if (!this.fabrics.has(fabricIndex)) {
      throw new FabricNotFoundError(
        `Fabric with index ${fabricIndex} cannot be updated because it does not exist.`
      );
    }
    this.fabrics.set(fabricIndex, fabric);
  }
  async revokeFabric(fabricIndex) {
    const fabric = this.fabrics.get(fabricIndex);
    if (fabric === void 0) {
      throw new import_MatterError.MatterFlowError(`Fabric with index ${fabricIndex} does not exist to revoke.`);
    }
    await fabric.remove();
  }
}
