// @project-chip/matter.js/dist/cjs/cluster/client/AttributeClient.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AttributeClient_exports = {};
__export(AttributeClient_exports, {
  AttributeClient: () => AttributeClient,
  SupportedAttributeClient: () => SupportedAttributeClient,
  UnknownSupportedAttributeClient: () => UnknownSupportedAttributeClient,
  createAttributeClient: () => createAttributeClient
});
module.exports = __toCommonJS(AttributeClient_exports);
var import_TryCatchHandler = require("../../common/TryCatchHandler.js");
var import_FabricIndex = require("../../datatype/FabricIndex.js");
var import_model = require("../../model/index.js");
var import_SecureSession = require("../../session/SecureSession.js");
var import_Cluster = require("../Cluster.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function createAttributeClient(attribute, name, endpointId, clusterId, interactionClient, present = false) {
  if (attribute.unknown) {
    return new UnknownSupportedAttributeClient(attribute, name, endpointId, clusterId, interactionClient);
  }
  if (present) {
    return new SupportedAttributeClient(attribute, name, endpointId, clusterId, interactionClient);
  }
  return new AttributeClient(attribute, name, endpointId, clusterId, interactionClient);
}
class AttributeClient {
  constructor(attribute, name, endpointId, clusterId, interactionClient) {
    this.attribute = attribute;
    this.name = name;
    this.endpointId = endpointId;
    this.clusterId = clusterId;
    this.interactionClient = interactionClient;
    this.listeners = new Array();
    const { schema, writable, fabricScoped, id } = attribute;
    this.schema = schema;
    this.isWritable = writable;
    this.isFabricScoped = fabricScoped;
    this.id = id;
  }
  /**
   * Set the value of the attribute. When dataVersion parameter is provided the value is only set when the
   * cluster dataVersion of the server matches. If it does not match it is rejected with an Error.
   */
  async set(value, dataVersion) {
    if (!this.isWritable)
      throw new import_Cluster.AttributeError(`Attribute ${this.name} is not writable`);
    value = this.schema.injectField(
      value,
      import_model.Globals.FabricIndex.id,
      import_FabricIndex.FabricIndex.OMIT_FABRIC,
      (existingFieldIndex) => existingFieldIndex === void 0
    );
    this.schema.validate(value);
    if (this.isFabricScoped) {
      value = this.schema.removeField(
        value,
        import_model.Globals.FabricIndex.id,
        (existingFieldIndex) => existingFieldIndex === import_FabricIndex.FabricIndex.OMIT_FABRIC
      );
      value = (0, import_TryCatchHandler.tryCatch)(
        () => {
          const sessionFabric = this.interactionClient.session.getAssociatedFabric();
          return this.schema.removeField(
            value,
            import_model.Globals.FabricIndex.id,
            (existingFieldIndex) => existingFieldIndex.index === sessionFabric.fabricIndex
          );
        },
        import_SecureSession.NoAssociatedFabricError,
        value
      );
    }
    return await this.interactionClient.setAttribute({
      attributeData: {
        endpointId: this.endpointId,
        clusterId: this.clusterId,
        attribute: this.attribute,
        value,
        dataVersion
      }
    });
  }
  /** Get the value of the attribute. Fabric scoped reads are always done with the remote. */
  async get(alwaysRequestFromRemote, isFabricFiltered = true) {
    if (alwaysRequestFromRemote === void 0) {
      alwaysRequestFromRemote = this.isFabricScoped;
    } else if (!alwaysRequestFromRemote && this.isFabricScoped) {
      alwaysRequestFromRemote = true;
    }
    return await this.interactionClient.getAttribute({
      endpointId: this.endpointId,
      clusterId: this.clusterId,
      attribute: this.attribute,
      isFabricFiltered,
      alwaysRequestFromRemote
    });
  }
  /**
   * Get the value with version of the attribute. Fabric scoped reads are always done with the remote.
   * */
  async getWithVersion(alwaysRequestFromRemote, isFabricFiltered = true) {
    if (alwaysRequestFromRemote === void 0) {
      alwaysRequestFromRemote = this.isFabricScoped;
    } else if (!alwaysRequestFromRemote && this.isFabricScoped) {
      alwaysRequestFromRemote = true;
    }
    return await this.interactionClient.getAttributeWithVersion({
      endpointId: this.endpointId,
      clusterId: this.clusterId,
      attribute: this.attribute,
      isFabricFiltered,
      alwaysRequestFromRemote
    });
  }
  /** Subscribe to the attribute. */
  async subscribe(minIntervalFloorSeconds, maxIntervalCeilingSeconds, knownDataVersion, isFabricFiltered = true) {
    return await this.interactionClient.subscribeAttribute({
      endpointId: this.endpointId,
      clusterId: this.clusterId,
      attribute: this.attribute,
      minIntervalFloorSeconds,
      maxIntervalCeilingSeconds,
      isFabricFiltered,
      knownDataVersion,
      listener: this.update.bind(this)
    });
  }
  /**
   * Update the value of the attribute. Just internally used!
   * @private
   */
  update(value) {
    this.listeners.forEach((listener) => listener(value));
  }
  /** Add a listener to the attribute. */
  addListener(listener) {
    this.listeners.push(listener);
  }
  /** Remove a listener from the attribute. */
  removeListener(listener) {
    const entryIndex = this.listeners.indexOf(listener);
    if (entryIndex !== -1) {
      this.listeners.splice(entryIndex, 1);
    }
  }
}
class SupportedAttributeClient extends AttributeClient {
}
class UnknownSupportedAttributeClient extends SupportedAttributeClient {
}
