// @project-chip/matter.js/dist/cjs/model/logic/record-validation/ValidatorImplementation.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ValidatorImplementation_exports = {};
__export(ValidatorImplementation_exports, {
  ValidatorImplementation: () => ValidatorImplementation
});
module.exports = __toCommonJS(ValidatorImplementation_exports);
var import_ValidatorBuilder = require("./ValidatorBuilder.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class ValidatorImplementation {
  constructor(fields, definedFeatures, enabledFeatures) {
    this.fields = fields;
    this.enabledFeatures = enabledFeatures;
    const builder = new import_ValidatorBuilder.ValidatorBuilder(fields, definedFeatures, enabledFeatures);
    this.validate = builder.compile();
    this.logFailure = () => builder.logFailure();
  }
  // Compiled validator invokes to test group values
  testGroup(...args) {
    for (const a of args) {
      if (a === true) {
        return true;
      }
      if (a === null) {
        return false;
      }
    }
    return false;
  }
  // Compiled validator invokes to record active choice.  A counter is kept
  // using a key of the form ".a" in values.  After conformance is tested on
  // each field, choices must be reexamined to ensure choices are correct
  testChoice(name, result, hasValue, num, orMore) {
    if (!this.choices) {
      this.choices = {};
    }
    let choice = this.choices[name];
    if (!choice) {
      choice = this.choices[name] = {
        count: 0,
        num,
        orMore
      };
    }
    if (hasValue) {
      choice.count++;
    }
    return result;
  }
  // Compiled validator invokes to check choices
  checkChoices() {
    for (const name in this.choices) {
      const choice = this.choices[name];
      if (choice.count < choice.num) {
        this.error("CHOICE_THRESHOLD_NOT_MET", name, `Choice minimum of ${choice.num} not met`);
      } else if (choice.count > choice.num && !choice.orMore) {
        this.error("CHOICE_THRESHOLD_EXCEEDED", name, `Choice maximum of ${choice.num} exceeded`);
      }
    }
  }
  // Compiled validator calls to add a validation error
  error(code, source, message) {
    const result = this.result;
    if (!result) {
      return;
    }
    if (!result.errors) {
      result.valid = false;
      result.errors = [];
    }
    result.errors.push({ code, source, message });
  }
}
