// @project-chip/matter.js/dist/cjs/model/logic/record-validation/ValidatorBuilder.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ValidatorBuilder_exports = {};
__export(ValidatorBuilder_exports, {
  ValidatorBuilder: () => ValidatorBuilder
});
module.exports = __toCommonJS(ValidatorBuilder_exports);
var import_Logger = require("../../../log/Logger.js");
var import_String = require("../../../util/String.js");
var import_aspects = require("../../aspects/index.js");
var import_models = require("../../models/index.js");
var import_conformance = require("./conformance.js");
var import_constraint = require("./constraint.js");
var import_quality = require("./quality.js");
var import_type = require("./type.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = new import_Logger.Logger("ValidatorBuilder");
class ValidatorBuilder extends Array {
  constructor(fields, definedFeatures, enabledFeatures) {
    super("this.result = { valid: true }", "let v");
    this.definedFeatures = definedFeatures;
    this.enabledFeatures = enabledFeatures;
    this.hasChoices = false;
    for (const child of fields) {
      if (!(child instanceof import_models.ValueModel)) {
        continue;
      }
      const aspects = child.validationAspects;
      if (!aspects.length) {
        continue;
      }
      this.push(`v = record[${JSON.stringify((0, import_String.camelize)(child.name, false))}]`);
      for (const aspect of aspects) {
        if (aspect instanceof import_aspects.Constraint) {
          (0, import_constraint.addConstraint)(this, child, aspect);
        } else if (aspect instanceof import_aspects.Conformance) {
          (0, import_conformance.addConformance)(this, child, aspect);
        } else if (aspect instanceof import_aspects.Quality) {
          (0, import_quality.addQuality)(this, child, aspect);
        }
        (0, import_type.addType)(this, child);
      }
    }
    if (this.hasChoices) {
      this.unshift("delete this.choices");
      this.push("this.checkChoices()");
    }
    this.push("return this.result");
  }
  addTest(test, code, source, message) {
    if (test == "true") {
      return;
    }
    this.push(
      `if (!(${test})) { this.error(${JSON.stringify(code)}, ${JSON.stringify(source.path)}, ${JSON.stringify(
        message
      )}) }`
    );
  }
  logInternalEvaluationError(message, e) {
    logger.error(`${message}: ${e}`);
    this.logFailure();
  }
  logFailure() {
    logger.debug("Failing record validator implementation:");
    import_Logger.Logger.nest(() => {
      const lineNoWidth = (this.length - 1).toString().length;
      for (let i = 0; i < this.length; i++) {
        logger.debug(`${i.toString().padStart(lineNoWidth, " ")} ${this[i]}`);
      }
    });
  }
  compile() {
    const logInternalEvaluationError = this.logInternalEvaluationError.bind(this);
    try {
      const fn = new Function("record", this.join("\n"));
      return function(record) {
        try {
          return fn.call(this, record);
        } catch (e) {
          logInternalEvaluationError(`Record validator evaluation failed`, e);
          throw e;
        }
      };
    } catch (e) {
      logInternalEvaluationError("Record validator compilation failed", e);
      throw e;
    }
  }
}
