// @project-chip/matter.js/dist/esm/codec/DerCodec.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { UnexpectedDataError } from "../common/MatterError.js";
import { ByteArray, Endian } from "../util/ByteArray.js";
import { DataReader } from "../util/DataReader.js";
const OBJECT_ID_KEY = "_objectId";
const TAG_ID_KEY = "_tag";
const BYTES_KEY = "_bytes";
const ELEMENTS_KEY = "_elements";
const BITS_PADDING = "_padding";
var DerType = /* @__PURE__ */ ((DerType2) => {
  DerType2[DerType2["Boolean"] = 1] = "Boolean";
  DerType2[DerType2["UnsignedInt"] = 2] = "UnsignedInt";
  DerType2[DerType2["BitString"] = 3] = "BitString";
  DerType2[DerType2["OctetString"] = 4] = "OctetString";
  DerType2[DerType2["ObjectIdentifier"] = 6] = "ObjectIdentifier";
  DerType2[DerType2["UTF8String"] = 12] = "UTF8String";
  DerType2[DerType2["Sequence"] = 16] = "Sequence";
  DerType2[DerType2["Set"] = 17] = "Set";
  DerType2[DerType2["UtcDate"] = 23] = "UtcDate";
  return DerType2;
})(DerType || {});
const CONSTRUCTED = 32;
var DerClass = /* @__PURE__ */ ((DerClass2) => {
  DerClass2[DerClass2["Universal"] = 0] = "Universal";
  DerClass2[DerClass2["Application"] = 64] = "Application";
  DerClass2[DerClass2["ContextSpecific"] = 128] = "ContextSpecific";
  DerClass2[DerClass2["Private"] = 192] = "Private";
  return DerClass2;
})(DerClass || {});
const ObjectId = (objectId) => ({
  [TAG_ID_KEY]: 6 /* ObjectIdentifier */,
  [BYTES_KEY]: ByteArray.fromHex(objectId)
});
const DerObject = (objectId, content = {}) => ({ [OBJECT_ID_KEY]: ObjectId(objectId), ...content });
const BitByteArray = (data, padding = 0) => ({
  [TAG_ID_KEY]: 3 /* BitString */,
  [BYTES_KEY]: data,
  [BITS_PADDING]: padding
});
const ContextTagged = (tagId, value) => ({
  [TAG_ID_KEY]: tagId | 128 /* ContextSpecific */ | CONSTRUCTED,
  [BYTES_KEY]: value === void 0 ? new ByteArray(0) : DerCodec.encode(value)
});
const ContextTaggedBytes = (tagId, value) => ({
  [TAG_ID_KEY]: tagId | 128 /* ContextSpecific */,
  [BYTES_KEY]: value
});
class DerCodec {
  static encode(value) {
    if (Array.isArray(value)) {
      return this.encodeArray(value);
    } else if (value instanceof ByteArray) {
      return this.encodeOctetString(value);
    } else if (value instanceof Date) {
      return this.encodeDate(value);
    } else if (typeof value === "object" && value[TAG_ID_KEY] !== void 0) {
      return this.encodeAnsi1(
        value[TAG_ID_KEY],
        value[BITS_PADDING] === void 0 ? value[BYTES_KEY] : ByteArray.concat(ByteArray.of(value[BITS_PADDING]), value[BYTES_KEY])
      );
    } else if (typeof value === "object") {
      return this.encodeObject(value);
    } else if (typeof value === "string") {
      return this.encodeString(value);
    } else if (typeof value === "number") {
      return this.encodeUnsignedInt(value);
    } else if (typeof value === "boolean") {
      return this.encodeBoolean(value);
    } else if (value === void 0) {
      return new ByteArray(0);
    } else {
      throw new UnexpectedDataError(`Unsupported type ${typeof value}`);
    }
  }
  static encodeDate(date) {
    return this.encodeAnsi1(
      23 /* UtcDate */,
      ByteArray.fromString(
        date.toISOString().replace(/[-:.T]/g, "").slice(2, 14) + "Z"
      )
    );
  }
  static encodeBoolean(bool) {
    return this.encodeAnsi1(1 /* Boolean */, ByteArray.of(bool ? 255 : 0));
  }
  static encodeArray(array) {
    return this.encodeAnsi1(
      17 /* Set */ | CONSTRUCTED,
      ByteArray.concat(...array.map((element) => this.encode(element)))
    );
  }
  static encodeOctetString(value) {
    return this.encodeAnsi1(4 /* OctetString */, value);
  }
  static encodeObject(object) {
    const attributes = new Array();
    for (const key in object) {
      attributes.push(this.encode(object[key]));
    }
    return this.encodeAnsi1(16 /* Sequence */ | CONSTRUCTED, ByteArray.concat(...attributes));
  }
  static encodeString(value) {
    return this.encodeAnsi1(12 /* UTF8String */, ByteArray.fromString(value));
  }
  static encodeUnsignedInt(value) {
    const byteArray = new ByteArray(5);
    const dataView = byteArray.getDataView();
    dataView.setUint32(1, value);
    let start = 0;
    while (true) {
      if (dataView.getUint8(start) !== 0)
        break;
      if (dataView.getUint8(start + 1) >= 128)
        break;
      start++;
      if (start === 4)
        break;
    }
    return this.encodeAnsi1(2 /* UnsignedInt */, byteArray.slice(start));
  }
  static encodeLengthBytes(value) {
    const byteArray = new ByteArray(5);
    const dataView = byteArray.getDataView();
    dataView.setUint32(1, value);
    let start = 0;
    while (true) {
      if (dataView.getUint8(start) !== 0)
        break;
      start++;
      if (start === 4)
        break;
    }
    const lengthLength = byteArray.length - start;
    if (lengthLength > 1 || dataView.getUint8(start) >= 128) {
      start--;
      dataView.setUint8(start, 128 + lengthLength);
    }
    return byteArray.slice(start);
  }
  static encodeAnsi1(tag, data) {
    return ByteArray.concat(ByteArray.of(tag), this.encodeLengthBytes(data.length), data);
  }
  static decode(data) {
    return this.decodeRec(new DataReader(data, Endian.Big));
  }
  static decodeRec(reader) {
    const { tag, bytes } = this.decodeAnsi1(reader);
    if (tag === 3 /* BitString */)
      return { [TAG_ID_KEY]: tag, [BYTES_KEY]: bytes.slice(1), [BITS_PADDING]: bytes[0] };
    if ((tag & CONSTRUCTED) === 0)
      return { [TAG_ID_KEY]: tag, [BYTES_KEY]: bytes };
    const elementsReader = new DataReader(bytes, Endian.Big);
    const elements = [];
    while (elementsReader.getRemainingBytesCount() > 0) {
      elements.push(this.decodeRec(elementsReader));
    }
    return { [TAG_ID_KEY]: tag, [BYTES_KEY]: bytes, [ELEMENTS_KEY]: elements };
  }
  static decodeAnsi1(reader) {
    const tag = reader.readUInt8();
    let length = reader.readUInt8();
    if ((length & 128) !== 0) {
      let lengthLength = length & 127;
      length = 0;
      while (lengthLength > 0) {
        length = (length << 8) + reader.readUInt8();
        lengthLength--;
      }
    }
    const bytes = reader.readByteArray(length);
    return { tag, bytes };
  }
}
const PublicKeyEcPrime256v1_X962 = (key) => ({
  type: {
    algorithm: ObjectId("2A8648CE3D0201"),
    curve: ObjectId("2A8648CE3D030107")
  },
  bytes: BitByteArray(key)
});
const EcdsaWithSHA256_X962 = DerObject("2A8648CE3D040302");
const SHA256_CMS = DerObject("608648016503040201");
const OrganisationName_X520 = (name) => [DerObject("55040A", { name })];
const SubjectKeyIdentifier_X509 = (identifier) => DerObject("551d0e", { value: DerCodec.encode(identifier) });
const AuthorityKeyIdentifier_X509 = (identifier) => DerObject("551d23", { value: DerCodec.encode({ id: ContextTaggedBytes(0, identifier) }) });
const BasicConstraints_X509 = (constraints) => DerObject("551d13", { critical: true, value: DerCodec.encode(constraints) });
const ExtendedKeyUsage_X509 = ({ clientAuth, serverAuth }) => DerObject("551d25", {
  critical: true,
  value: DerCodec.encode({
    client: clientAuth ? ObjectId("2b06010505070302") : void 0,
    server: serverAuth ? ObjectId("2b06010505070301") : void 0
  })
});
const KeyUsage_Signature_X509 = DerObject("551d0f", {
  critical: true,
  value: DerCodec.encode(BitByteArray(ByteArray.of(1 << 7), 7))
});
const KeyUsage_Signature_ContentCommited_X509 = DerObject("551d0f", {
  critical: true,
  value: DerCodec.encode(BitByteArray(ByteArray.of(3 << 1), 1))
});
const Pkcs7Data = (data) => DerObject("2A864886F70D010701", { value: ContextTagged(0, data) });
const Pkcs7SignedData = (data) => DerObject("2a864886f70d010702", { value: ContextTagged(0, data) });
export {
  AuthorityKeyIdentifier_X509,
  BITS_PADDING,
  BYTES_KEY,
  BasicConstraints_X509,
  BitByteArray,
  ContextTagged,
  ContextTaggedBytes,
  DerCodec,
  DerObject,
  DerType,
  ELEMENTS_KEY,
  EcdsaWithSHA256_X962,
  ExtendedKeyUsage_X509,
  KeyUsage_Signature_ContentCommited_X509,
  KeyUsage_Signature_X509,
  OBJECT_ID_KEY,
  ObjectId,
  OrganisationName_X520,
  Pkcs7Data,
  Pkcs7SignedData,
  PublicKeyEcPrime256v1_X962,
  SHA256_CMS,
  SubjectKeyIdentifier_X509,
  TAG_ID_KEY
};
