// @project-chip/matter.js/dist/cjs/tlv/TlvNullable.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TlvNullable_exports = {};
__export(TlvNullable_exports, {
  NullableSchema: () => NullableSchema,
  TlvNullable: () => TlvNullable
});
module.exports = __toCommonJS(TlvNullable_exports);
var import_TlvCodec = require("./TlvCodec.js");
var import_TlvSchema = require("./TlvSchema.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class NullableSchema extends import_TlvSchema.TlvSchema {
  constructor(schema) {
    super();
    this.schema = schema;
  }
  encodeTlvInternal(writer, value, tag, forWriteInteraction) {
    if (value === null) {
      writer.writeTag({ type: import_TlvCodec.TlvType.Null }, tag);
    } else {
      this.schema.encodeTlvInternal(writer, value, tag, forWriteInteraction);
    }
  }
  decodeTlvInternalValue(reader, typeLength) {
    if (typeLength.type === import_TlvCodec.TlvType.Null)
      return null;
    return this.schema.decodeTlvInternalValue(reader, typeLength);
  }
  validate(value) {
    if (value !== null)
      this.schema.validate(value);
  }
  injectField(value, fieldId, fieldValue, injectChecker) {
    if (value !== null) {
      return this.schema.injectField(value, fieldId, fieldValue, injectChecker);
    }
    return value;
  }
  removeField(value, fieldId, removeChecker) {
    if (value !== null) {
      return this.schema.removeField(value, fieldId, removeChecker);
    }
    return value;
  }
}
const TlvNullable = (schema) => new NullableSchema(schema);
