// @project-chip/matter.js/dist/cjs/tlv/TlvObject.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TlvObject_exports = {};
__export(TlvObject_exports, {
  ObjectSchema: () => ObjectSchema,
  TlvField: () => TlvField,
  TlvObject: () => TlvObject,
  TlvOptionalField: () => TlvOptionalField,
  TlvOptionalRepeatedField: () => TlvOptionalRepeatedField,
  TlvRepeatedField: () => TlvRepeatedField,
  TlvTaggedList: () => TlvTaggedList
});
module.exports = __toCommonJS(TlvObject_exports);
var import_MatterError = require("../common/MatterError.js");
var import_TryCatchHandler = require("../common/TryCatchHandler.js");
var import_ValidationError = require("../common/ValidationError.js");
var import_model = require("../model/index.js");
var import_TlvAny = require("./TlvAny.js");
var import_TlvCodec = require("./TlvCodec.js");
var import_TlvSchema = require("./TlvSchema.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class ObjectSchema extends import_TlvSchema.TlvSchema {
  constructor(fieldDefinitions, type = import_TlvCodec.TlvType.Structure) {
    super();
    this.fieldDefinitions = fieldDefinitions;
    this.type = type;
    this.fieldById = new Array();
    for (const name in this.fieldDefinitions) {
      const field = this.fieldDefinitions[name];
      if (field.repeated && type !== import_TlvCodec.TlvType.List) {
        throw new Error("Repeated fields are only allowed in TLV List.");
      }
      this.fieldById[field.id] = { name, field };
    }
  }
  encodeTlvInternal(writer, value, tag, forWriteInteraction) {
    writer.writeTag({ type: this.type }, tag);
    for (const name in this.fieldDefinitions) {
      const { id, schema, optional: isOptional, repeated: isRepeated } = this.fieldDefinitions[name];
      const fieldValue = value[name];
      if (fieldValue === void 0) {
        if (!isOptional) {
          if (forWriteInteraction && id === import_model.Globals.FabricIndex.id) {
            continue;
          }
          throw new import_ValidationError.ValidationError(`Missing mandatory field ${name}`, name);
        }
        continue;
      }
      if (isRepeated) {
        if (!Array.isArray(fieldValue)) {
          throw new import_ValidationError.ValidationError(`Repeated field ${name} should be an array.`, name);
        }
        for (const element of fieldValue) {
          schema.encodeTlvInternal(writer, element, { id }, forWriteInteraction);
        }
      } else {
        schema.encodeTlvInternal(writer, fieldValue, { id }, forWriteInteraction);
      }
    }
    writer.writeTag({ type: import_TlvCodec.TlvType.EndOfContainer });
  }
  decodeTlvInternalValue(reader, typeLength) {
    if (typeLength.type !== this.type)
      throw new import_MatterError.UnexpectedDataError(`Unexpected type ${typeLength.type} (expected ${this.type}).`);
    const result = {};
    while (true) {
      const { tag: { profile, id } = {}, typeLength: elementTypeLength } = reader.readTagType();
      if (elementTypeLength.type === import_TlvCodec.TlvType.EndOfContainer)
        break;
      if (profile !== void 0)
        throw new import_MatterError.UnexpectedDataError("Structure element tags should be context-specific.");
      if (id === void 0)
        throw new import_MatterError.UnexpectedDataError("Structure element tags should have an id.");
      const fieldName = this.fieldById[id];
      if (fieldName === void 0) {
        import_TlvAny.TlvAny.decodeTlvInternalValue(reader, elementTypeLength);
        continue;
      }
      const { field, name } = fieldName;
      const decoded = field.schema.decodeTlvInternalValue(reader, elementTypeLength);
      if (field.repeated) {
        if (result[name] === void 0) {
          result[name] = [decoded];
        } else {
          result[name].push(decoded);
        }
      } else {
        result[name] = decoded;
      }
    }
    for (const name in this.fieldDefinitions) {
      const { optional, fallback, repeated } = this.fieldDefinitions[name];
      if (optional)
        continue;
      const value = result[name];
      if (value !== void 0)
        continue;
      if (fallback !== void 0) {
        if (repeated) {
          result[name] = [fallback];
        } else {
          result[name] = fallback;
        }
      }
    }
    return result;
  }
  validate(value) {
    for (const name in this.fieldDefinitions) {
      const { optional, schema, repeated: isRepeated } = this.fieldDefinitions[name];
      const data = value[name];
      if (data === void 0) {
        if (optional) {
          continue;
        }
        throw new import_ValidationError.ValidationError(`Missing mandatory field ${name}`, name);
      }
      if (isRepeated) {
        const { minLength = 2, maxLength = 65535 } = this.fieldDefinitions[name];
        if (!Array.isArray(data)) {
          throw new import_ValidationError.ValidationError(`Repeated field ${name} should be an array.`, name);
        }
        if (data.length > maxLength)
          throw new import_ValidationError.ValidationError(
            `Repeated field list for ${name} is too long: ${data.length}, max ${maxLength}.`,
            name
          );
        if (data.length < minLength)
          throw new import_ValidationError.ValidationError(
            `Repeated field list for ${name} is too short: ${data.length}, min ${minLength}.`,
            name
          );
        for (const element of data) {
          (0, import_TryCatchHandler.tryCatch)(
            () => schema.validate(element),
            import_ValidationError.ValidationError,
            (error) => {
              error.fieldName = `${name}${error.fieldName !== void 0 ? `.${error.fieldName}` : ""}`;
              throw error;
            }
          );
        }
      } else {
        (0, import_TryCatchHandler.tryCatch)(
          () => schema.validate(data),
          import_ValidationError.ValidationError,
          (error) => {
            error.fieldName = `${name}${error.fieldName !== void 0 ? `.${error.fieldName}` : ""}`;
            throw error;
          }
        );
      }
    }
  }
  injectField(value, fieldId, fieldValue, injectChecker) {
    for (const k in this.fieldDefinitions) {
      const field = this.fieldDefinitions[k];
      if (field.id === fieldId) {
        if (injectChecker(value[k])) {
          field.schema.validate(fieldValue);
          value[k] = fieldValue;
        }
      } else {
        value[k] = field.schema.injectField(value[k], fieldId, fieldValue, injectChecker);
      }
    }
    return value;
  }
  removeField(value, fieldId, removeChecker) {
    for (const k in this.fieldDefinitions) {
      const field = this.fieldDefinitions[k];
      if (field.id === fieldId) {
        if (value[k] !== void 0 && removeChecker(value[k])) {
          delete value[k];
        }
      } else {
        value[k] = field.schema.removeField(value[k], fieldId, removeChecker);
      }
    }
    return value;
  }
}
const TlvObject = (fields) => new ObjectSchema(fields, import_TlvCodec.TlvType.Structure);
const TlvTaggedList = (fields) => new ObjectSchema(fields, import_TlvCodec.TlvType.List);
const TlvField = (id, schema, fallback) => ({ id, schema, fallback, optional: false });
const TlvOptionalField = (id, schema) => ({ id, schema, optional: true });
const TlvRepeatedField = (id, schema, lengthOptions) => {
  const { minLength, maxLength, length } = lengthOptions ?? {};
  return {
    id,
    schema,
    optional: false,
    repeated: true,
    minLength: length ?? minLength,
    maxLength: length ?? maxLength
  };
};
const TlvOptionalRepeatedField = (id, schema, lengthOptions) => {
  const { maxLength } = lengthOptions ?? {};
  return {
    id,
    schema,
    optional: true,
    repeated: true,
    minLength: 0,
    maxLength
  };
};
