// @project-chip/matter.js/dist/esm/tlv/TlvObject.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { UnexpectedDataError } from "../common/MatterError.js";
import { tryCatch } from "../common/TryCatchHandler.js";
import { ValidationError } from "../common/ValidationError.js";
import { Globals } from "../model/index.js";
import { TlvAny } from "./TlvAny.js";
import { TlvType } from "./TlvCodec.js";
import { TlvSchema } from "./TlvSchema.js";
class ObjectSchema extends TlvSchema {
  constructor(fieldDefinitions, type = TlvType.Structure) {
    super();
    this.fieldDefinitions = fieldDefinitions;
    this.type = type;
    this.fieldById = new Array();
    for (const name in this.fieldDefinitions) {
      const field = this.fieldDefinitions[name];
      if (field.repeated && type !== TlvType.List) {
        throw new Error("Repeated fields are only allowed in TLV List.");
      }
      this.fieldById[field.id] = { name, field };
    }
  }
  encodeTlvInternal(writer, value, tag, forWriteInteraction) {
    writer.writeTag({ type: this.type }, tag);
    for (const name in this.fieldDefinitions) {
      const { id, schema, optional: isOptional, repeated: isRepeated } = this.fieldDefinitions[name];
      const fieldValue = value[name];
      if (fieldValue === void 0) {
        if (!isOptional) {
          if (forWriteInteraction && id === Globals.FabricIndex.id) {
            continue;
          }
          throw new ValidationError(`Missing mandatory field ${name}`, name);
        }
        continue;
      }
      if (isRepeated) {
        if (!Array.isArray(fieldValue)) {
          throw new ValidationError(`Repeated field ${name} should be an array.`, name);
        }
        for (const element of fieldValue) {
          schema.encodeTlvInternal(writer, element, { id }, forWriteInteraction);
        }
      } else {
        schema.encodeTlvInternal(writer, fieldValue, { id }, forWriteInteraction);
      }
    }
    writer.writeTag({ type: TlvType.EndOfContainer });
  }
  decodeTlvInternalValue(reader, typeLength) {
    if (typeLength.type !== this.type)
      throw new UnexpectedDataError(`Unexpected type ${typeLength.type} (expected ${this.type}).`);
    const result = {};
    while (true) {
      const { tag: { profile, id } = {}, typeLength: elementTypeLength } = reader.readTagType();
      if (elementTypeLength.type === TlvType.EndOfContainer)
        break;
      if (profile !== void 0)
        throw new UnexpectedDataError("Structure element tags should be context-specific.");
      if (id === void 0)
        throw new UnexpectedDataError("Structure element tags should have an id.");
      const fieldName = this.fieldById[id];
      if (fieldName === void 0) {
        TlvAny.decodeTlvInternalValue(reader, elementTypeLength);
        continue;
      }
      const { field, name } = fieldName;
      const decoded = field.schema.decodeTlvInternalValue(reader, elementTypeLength);
      if (field.repeated) {
        if (result[name] === void 0) {
          result[name] = [decoded];
        } else {
          result[name].push(decoded);
        }
      } else {
        result[name] = decoded;
      }
    }
    for (const name in this.fieldDefinitions) {
      const { optional, fallback, repeated } = this.fieldDefinitions[name];
      if (optional)
        continue;
      const value = result[name];
      if (value !== void 0)
        continue;
      if (fallback !== void 0) {
        if (repeated) {
          result[name] = [fallback];
        } else {
          result[name] = fallback;
        }
      }
    }
    return result;
  }
  validate(value) {
    for (const name in this.fieldDefinitions) {
      const { optional, schema, repeated: isRepeated } = this.fieldDefinitions[name];
      const data = value[name];
      if (data === void 0) {
        if (optional) {
          continue;
        }
        throw new ValidationError(`Missing mandatory field ${name}`, name);
      }
      if (isRepeated) {
        const { minLength = 2, maxLength = 65535 } = this.fieldDefinitions[name];
        if (!Array.isArray(data)) {
          throw new ValidationError(`Repeated field ${name} should be an array.`, name);
        }
        if (data.length > maxLength)
          throw new ValidationError(
            `Repeated field list for ${name} is too long: ${data.length}, max ${maxLength}.`,
            name
          );
        if (data.length < minLength)
          throw new ValidationError(
            `Repeated field list for ${name} is too short: ${data.length}, min ${minLength}.`,
            name
          );
        for (const element of data) {
          tryCatch(
            () => schema.validate(element),
            ValidationError,
            (error) => {
              error.fieldName = `${name}${error.fieldName !== void 0 ? `.${error.fieldName}` : ""}`;
              throw error;
            }
          );
        }
      } else {
        tryCatch(
          () => schema.validate(data),
          ValidationError,
          (error) => {
            error.fieldName = `${name}${error.fieldName !== void 0 ? `.${error.fieldName}` : ""}`;
            throw error;
          }
        );
      }
    }
  }
  injectField(value, fieldId, fieldValue, injectChecker) {
    for (const k in this.fieldDefinitions) {
      const field = this.fieldDefinitions[k];
      if (field.id === fieldId) {
        if (injectChecker(value[k])) {
          field.schema.validate(fieldValue);
          value[k] = fieldValue;
        }
      } else {
        value[k] = field.schema.injectField(value[k], fieldId, fieldValue, injectChecker);
      }
    }
    return value;
  }
  removeField(value, fieldId, removeChecker) {
    for (const k in this.fieldDefinitions) {
      const field = this.fieldDefinitions[k];
      if (field.id === fieldId) {
        if (value[k] !== void 0 && removeChecker(value[k])) {
          delete value[k];
        }
      } else {
        value[k] = field.schema.removeField(value[k], fieldId, removeChecker);
      }
    }
    return value;
  }
}
const TlvObject = (fields) => new ObjectSchema(fields, TlvType.Structure);
const TlvTaggedList = (fields) => new ObjectSchema(fields, TlvType.List);
const TlvField = (id, schema, fallback) => ({ id, schema, fallback, optional: false });
const TlvOptionalField = (id, schema) => ({ id, schema, optional: true });
const TlvRepeatedField = (id, schema, lengthOptions) => {
  const { minLength, maxLength, length } = lengthOptions ?? {};
  return {
    id,
    schema,
    optional: false,
    repeated: true,
    minLength: length ?? minLength,
    maxLength: length ?? maxLength
  };
};
const TlvOptionalRepeatedField = (id, schema, lengthOptions) => {
  const { maxLength } = lengthOptions ?? {};
  return {
    id,
    schema,
    optional: true,
    repeated: true,
    minLength: 0,
    maxLength
  };
};
export {
  ObjectSchema,
  TlvField,
  TlvObject,
  TlvOptionalField,
  TlvOptionalRepeatedField,
  TlvRepeatedField,
  TlvTaggedList
};
