// @project-chip/matter.js/dist/cjs/common/FailSafeManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FailSafeManager_exports = {};
__export(FailSafeManager_exports, {
  FailSafeManager: () => FailSafeManager,
  MatterFabricConflictError: () => MatterFabricConflictError
});
module.exports = __toCommonJS(FailSafeManager_exports);
var import_definitions = require("../cluster/definitions/index.js");
var import_Fabric = require("../fabric/Fabric.js");
var import_Time = require("../time/Time.js");
var import_MatterError = require("./MatterError.js");
class MatterFabricConflictError extends import_MatterError.MatterFlowError {
}
class FailSafeManager {
  constructor(device, associatedFabric, expiryLengthSeconds, maxCumulativeFailsafeSeconds, expiryCallback, rootEndpoint) {
    this.device = device;
    this.associatedFabric = associatedFabric;
    this.expiryCallback = expiryCallback;
    this.rootEndpoint = rootEndpoint;
    this.fabricBuilder = new import_Fabric.FabricBuilder();
    this.storedNetworkClusterState = /* @__PURE__ */ new Map();
    this.storeEndpointState();
    this.failSafeTimer = import_Time.Time.getTimer(expiryLengthSeconds * 1e3, () => this.expire()).start();
    this.maxCumulativeFailsafeTimer = import_Time.Time.getTimer(
      maxCumulativeFailsafeSeconds * 1e3,
      () => this.expire()
    ).start();
  }
  /** Store required CLuster data when opening the FailSafe context to allow to restore them on expiry. */
  storeEndpointState(endpoint = this.rootEndpoint) {
    const networkCluster = endpoint.getClusterServer(import_definitions.NetworkCommissioning.Cluster);
    if (networkCluster !== void 0) {
      this.storedNetworkClusterState.set(endpoint.getId(), networkCluster.getNetworksAttribute());
    }
    for (const childEndpoint of endpoint.getChildEndpoints()) {
      this.storeEndpointState(childEndpoint);
    }
  }
  /** Restore Cluster data when the FailSafe context expired. */
  restoreEndpointState(endpoint = this.rootEndpoint) {
    const endpointId = endpoint.getId();
    const networkState = this.storedNetworkClusterState.get(endpointId);
    if (networkState !== void 0) {
      const networkCluster = endpoint.getClusterServer(import_definitions.NetworkCommissioning.Cluster);
      if (networkCluster !== void 0) {
        networkCluster.setNetworksAttribute(networkState);
      }
      this.storedNetworkClusterState.delete(endpointId);
    }
    for (const childEndpoint of endpoint.getChildEndpoints()) {
      this.restoreEndpointState(childEndpoint);
    }
  }
  get fabricIndex() {
    return this.fabricBuilder.getFabricIndex();
  }
  /** Handle "Re-Arming" an existing FailSafe context to extend the timer, expire or fail if not allowed. */
  async reArm(associatedFabric, expiryLengthSeconds) {
    if (!this.failSafeTimer.isRunning) {
      throw new import_MatterError.MatterFlowError("FailSafe already expired.");
    }
    if (this.associatedFabric?.fabricIndex !== associatedFabric?.fabricIndex) {
      throw new import_MatterError.MatterFlowError(
        `FailSafe already armed (index=${this.associatedFabric?.fabricIndex}) with different fabric (index=${associatedFabric?.fabricIndex}).`
      );
    }
    this.failSafeTimer.stop();
    if (expiryLengthSeconds === 0) {
      await this.expire();
    } else {
      this.failSafeTimer = import_Time.Time.getTimer(expiryLengthSeconds * 1e3, () => this.expire()).start();
    }
  }
  /** Expire the FailSafe context. This is called by the timer and can also be called manually if needed. */
  async expire() {
    this.complete();
    await this.expiryCallback();
  }
  /** Complete the FailSafe context. This is called when the commissioning is completed. */
  complete() {
    this.failSafeTimer.stop();
    this.maxCumulativeFailsafeTimer.stop();
  }
  /**
   * Handles a CSR from OperationalCredentials cluster and stores additional internal information for further
   * validity checks.
   */
  createCertificateSigningRequest(isForUpdateNoc, sessionId) {
    const result = this.fabricBuilder.createCertificateSigningRequest();
    this.csrSessionId = sessionId;
    this.forUpdateNoc = isForUpdateNoc;
    return result;
  }
  /** Handles adding a trusted root certificate from Operational Credentials cluster. */
  setRootCert(rootCert) {
    this.fabricBuilder.setRootCert(rootCert);
  }
  /**
   * Build a new Fabric object based on an existing fabric for the "UpdateNoc" case of the Operational Credentials
   * cluster.
   */
  async buildUpdatedFabric(nocValue, icacValue) {
    if (this.associatedFabric === void 0) {
      throw new import_MatterError.MatterFlowError("No fabric associated with failsafe context, but we prepare an Fabric update.");
    }
    this.fabricBuilder.initializeFromFabricForUpdate(this.associatedFabric);
    this.fabricBuilder.setOperationalCert(nocValue);
    if (icacValue && icacValue.length > 0)
      this.fabricBuilder.setIntermediateCACert(icacValue);
    return await this.fabricBuilder.build(this.associatedFabric.fabricIndex);
  }
  /** Build a new Fabric object for a new fabric for the "AddNoc" case of the Operational Credentials cluster. */
  async buildFabric(nocData) {
    const { nocValue, icacValue, adminVendorId, ipkValue, caseAdminSubject } = nocData;
    this.fabricBuilder.setOperationalCert(nocValue);
    const fabricAlreadyExisting = this.device.getFabrics().find((fabric) => this.fabricBuilder.matchesToFabric(fabric));
    if (fabricAlreadyExisting) {
      throw new MatterFabricConflictError(
        `Fabric with Id ${this.fabricBuilder.getFabricId()} and Node Id ${this.fabricBuilder.getNodeId()} already exists.`
      );
    }
    if (icacValue && icacValue.length > 0)
      this.fabricBuilder.setIntermediateCACert(icacValue);
    this.fabricBuilder.setRootVendorId(adminVendorId);
    this.fabricBuilder.setIdentityProtectionKey(ipkValue);
    this.fabricBuilder.setRootNodeId(caseAdminSubject);
    return await this.fabricBuilder.build(this.device.getNextFabricIndex());
  }
}
