// @project-chip/matter.js/dist/cjs/common/MatterError.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MatterError_exports = {};
__export(MatterError_exports, {
  ImplementationError: () => ImplementationError,
  InternalError: () => InternalError,
  MatterError: () => MatterError,
  MatterFlowError: () => MatterFlowError,
  NoProviderError: () => NoProviderError,
  NotImplementedError: () => NotImplementedError,
  UnexpectedDataError: () => UnexpectedDataError
});
module.exports = __toCommonJS(MatterError_exports);
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class MatterError extends Error {
}
class NoProviderError extends MatterError {
}
class InternalError extends MatterError {
}
class NotImplementedError extends InternalError {
}
class MatterFlowError extends MatterError {
}
class UnexpectedDataError extends MatterError {
}
class ImplementationError extends MatterError {
}
