// @project-chip/matter.js/dist/cjs/certificate/CertificateManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CertificateManager_exports = {};
__export(CertificateManager_exports, {
  CertificateError: () => CertificateError,
  CertificateManager: () => CertificateManager,
  CommonName_X520: () => CommonName_X520,
  FabricId_Matter: () => FabricId_Matter,
  FirmwareSigningId_Matter: () => FirmwareSigningId_Matter,
  IcacId_Matter: () => IcacId_Matter,
  NocCat_Matter: () => NocCat_Matter,
  NodeId_Matter: () => NodeId_Matter,
  ProductId_Matter: () => ProductId_Matter,
  RcacId_Matter: () => RcacId_Matter,
  TlvCertificationDeclaration: () => TlvCertificationDeclaration,
  TlvIntermediateCertificate: () => TlvIntermediateCertificate,
  TlvOperationalCertificate: () => TlvOperationalCertificate,
  TlvRootCertificate: () => TlvRootCertificate,
  VendorId_Matter: () => VendorId_Matter,
  jsToMatterDate: () => jsToMatterDate,
  matterToJsDate: () => matterToJsDate
});
module.exports = __toCommonJS(CertificateManager_exports);
var import_DerCodec = require("../codec/DerCodec.js");
var import_MatterError = require("../common/MatterError.js");
var import_Crypto = require("../crypto/Crypto.js");
var import_Key = require("../crypto/Key.js");
var import_CaseAuthenticatedTag = require("../datatype/CaseAuthenticatedTag.js");
var import_FabricId = require("../datatype/FabricId.js");
var import_NodeId = require("../datatype/NodeId.js");
var import_VendorId = require("../datatype/VendorId.js");
var import_TlvArray = require("../tlv/TlvArray.js");
var import_TlvBoolean = require("../tlv/TlvBoolean.js");
var import_TlvNumber = require("../tlv/TlvNumber.js");
var import_TlvObject = require("../tlv/TlvObject.js");
var import_TlvString = require("../tlv/TlvString.js");
var import_ByteArray = require("../util/ByteArray.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class CertificateError extends import_MatterError.MatterError {
}
const YEAR_S = 365 * 24 * 60 * 60;
const EPOCH_OFFSET_S = 10957 * 24 * 60 * 60;
function matterToJsDate(date) {
  return new Date((date + EPOCH_OFFSET_S) * 1e3);
}
function jsToMatterDate(date, addYears = 0) {
  return Math.floor(date.getTime() / 1e3) - EPOCH_OFFSET_S + addYears * YEAR_S;
}
function intTo16Chars(value) {
  const byteArray = new import_ByteArray.ByteArray(8);
  const dataView = byteArray.getDataView();
  dataView.setBigUint64(0, typeof value === "bigint" ? value : BigInt(value));
  return byteArray.toHex().toUpperCase();
}
function uInt16To8Chars(value) {
  const byteArray = new import_ByteArray.ByteArray(4);
  const dataView = byteArray.getDataView();
  dataView.setUint32(0, value);
  return byteArray.toHex().toUpperCase();
}
function uInt16To4Chars(value) {
  const byteArray = new import_ByteArray.ByteArray(2);
  const dataView = byteArray.getDataView();
  dataView.setUint16(0, value);
  return byteArray.toHex().toUpperCase();
}
const CommonName_X520 = (name) => [(0, import_DerCodec.DerObject)("550403", { value: name })];
const NodeId_Matter = (nodeId) => [(0, import_DerCodec.DerObject)("2b0601040182a27c0101", { value: intTo16Chars(nodeId) })];
const FirmwareSigningId_Matter = (signingId) => [
  (0, import_DerCodec.DerObject)("2b0601040182a27c0102", { value: intTo16Chars(signingId) })
];
const IcacId_Matter = (id) => [(0, import_DerCodec.DerObject)("2b0601040182a27c0103", { value: intTo16Chars(id) })];
const RcacId_Matter = (id) => [(0, import_DerCodec.DerObject)("2b0601040182a27c0104", { value: intTo16Chars(id) })];
const FabricId_Matter = (id) => [(0, import_DerCodec.DerObject)("2b0601040182a27c0105", { value: intTo16Chars(id) })];
const NocCat_Matter = (cat) => [(0, import_DerCodec.DerObject)("2b0601040182a27c0106", { value: uInt16To8Chars(cat) })];
const VendorId_Matter = (vendorId) => [
  (0, import_DerCodec.DerObject)("2b0601040182a27c0201", { value: uInt16To4Chars(vendorId) })
];
const ProductId_Matter = (id) => [(0, import_DerCodec.DerObject)("2b0601040182a27c0202", { value: uInt16To4Chars(id) })];
const TlvRootCertificate = (0, import_TlvObject.TlvObject)({
  serialNumber: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvByteString.bound({ maxLength: 20 })),
  signatureAlgorithm: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8),
  issuer: (0, import_TlvObject.TlvField)(
    3,
    (0, import_TlvObject.TlvTaggedList)({
      issuerRcacId: (0, import_TlvObject.TlvOptionalField)(20, import_TlvNumber.TlvUInt64)
    })
  ),
  notBefore: (0, import_TlvObject.TlvField)(4, import_TlvNumber.TlvUInt32),
  notAfter: (0, import_TlvObject.TlvField)(5, import_TlvNumber.TlvUInt32),
  subject: (0, import_TlvObject.TlvField)(
    6,
    (0, import_TlvObject.TlvTaggedList)({
      rcacId: (0, import_TlvObject.TlvField)(20, import_TlvNumber.TlvUInt64)
    })
  ),
  publicKeyAlgorithm: (0, import_TlvObject.TlvField)(7, import_TlvNumber.TlvUInt8),
  ellipticCurveIdentifier: (0, import_TlvObject.TlvField)(8, import_TlvNumber.TlvUInt8),
  ellipticCurvePublicKey: (0, import_TlvObject.TlvField)(9, import_TlvString.TlvByteString),
  extensions: (0, import_TlvObject.TlvField)(
    10,
    (0, import_TlvObject.TlvTaggedList)({
      basicConstraints: (0, import_TlvObject.TlvField)(
        1,
        (0, import_TlvObject.TlvObject)({
          isCa: (0, import_TlvObject.TlvField)(1, import_TlvBoolean.TlvBoolean),
          pathLen: (0, import_TlvObject.TlvOptionalField)(2, import_TlvNumber.TlvUInt8)
        })
      ),
      keyUsage: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt16),
      extendedKeyUsage: (0, import_TlvObject.TlvOptionalField)(3, (0, import_TlvArray.TlvArray)(import_TlvNumber.TlvUInt8)),
      subjectKeyIdentifier: (0, import_TlvObject.TlvField)(4, import_TlvString.TlvByteString.bound({ length: 20 })),
      authorityKeyIdentifier: (0, import_TlvObject.TlvField)(5, import_TlvString.TlvByteString.bound({ length: 20 })),
      futureExtension: (0, import_TlvObject.TlvOptionalField)(6, import_TlvString.TlvByteString)
    })
  ),
  signature: (0, import_TlvObject.TlvField)(11, import_TlvString.TlvByteString)
});
const TlvOperationalCertificate = (0, import_TlvObject.TlvObject)({
  serialNumber: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvByteString.bound({ maxLength: 20 })),
  signatureAlgorithm: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8),
  issuer: (0, import_TlvObject.TlvField)(
    3,
    (0, import_TlvObject.TlvTaggedList)({
      issuerRcacId: (0, import_TlvObject.TlvOptionalField)(20, import_TlvNumber.TlvUInt64)
    })
  ),
  notBefore: (0, import_TlvObject.TlvField)(4, import_TlvNumber.TlvUInt32),
  notAfter: (0, import_TlvObject.TlvField)(5, import_TlvNumber.TlvUInt32),
  subject: (0, import_TlvObject.TlvField)(
    6,
    (0, import_TlvObject.TlvTaggedList)({
      fabricId: (0, import_TlvObject.TlvField)(21, import_FabricId.TlvFabricId),
      nodeId: (0, import_TlvObject.TlvField)(17, import_NodeId.TlvNodeId),
      caseAuthenticatedTags: (0, import_TlvObject.TlvOptionalRepeatedField)(22, import_CaseAuthenticatedTag.TlvCaseAuthenticatedTag, { maxLength: 3 })
    })
  ),
  publicKeyAlgorithm: (0, import_TlvObject.TlvField)(7, import_TlvNumber.TlvUInt8),
  ellipticCurveIdentifier: (0, import_TlvObject.TlvField)(8, import_TlvNumber.TlvUInt8),
  ellipticCurvePublicKey: (0, import_TlvObject.TlvField)(9, import_TlvString.TlvByteString),
  extensions: (0, import_TlvObject.TlvField)(
    10,
    (0, import_TlvObject.TlvTaggedList)({
      basicConstraints: (0, import_TlvObject.TlvField)(
        1,
        (0, import_TlvObject.TlvObject)({
          isCa: (0, import_TlvObject.TlvField)(1, import_TlvBoolean.TlvBoolean),
          pathLen: (0, import_TlvObject.TlvOptionalField)(2, import_TlvNumber.TlvUInt8)
        })
      ),
      keyUsage: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt16),
      extendedKeyUsage: (0, import_TlvObject.TlvOptionalField)(3, (0, import_TlvArray.TlvArray)(import_TlvNumber.TlvUInt8)),
      subjectKeyIdentifier: (0, import_TlvObject.TlvField)(4, import_TlvString.TlvByteString.bound({ length: 20 })),
      authorityKeyIdentifier: (0, import_TlvObject.TlvField)(5, import_TlvString.TlvByteString.bound({ length: 20 })),
      futureExtension: (0, import_TlvObject.TlvOptionalField)(6, import_TlvString.TlvByteString)
    })
  ),
  signature: (0, import_TlvObject.TlvField)(11, import_TlvString.TlvByteString)
});
const TlvIntermediateCertificate = (0, import_TlvObject.TlvObject)({
  serialNumber: (0, import_TlvObject.TlvField)(1, import_TlvString.TlvByteString.bound({ maxLength: 20 })),
  signatureAlgorithm: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt8),
  issuer: (0, import_TlvObject.TlvField)(
    3,
    (0, import_TlvObject.TlvTaggedList)({
      issuerRcacId: (0, import_TlvObject.TlvOptionalField)(20, import_TlvNumber.TlvUInt64)
    })
  ),
  notBefore: (0, import_TlvObject.TlvField)(4, import_TlvNumber.TlvUInt32),
  notAfter: (0, import_TlvObject.TlvField)(5, import_TlvNumber.TlvUInt32),
  subject: (0, import_TlvObject.TlvField)(
    6,
    (0, import_TlvObject.TlvTaggedList)({
      fabricId: (0, import_TlvObject.TlvOptionalField)(21, import_FabricId.TlvFabricId),
      icacId: (0, import_TlvObject.TlvField)(19, import_TlvNumber.TlvUInt64)
    })
  ),
  publicKeyAlgorithm: (0, import_TlvObject.TlvField)(7, import_TlvNumber.TlvUInt8),
  ellipticCurveIdentifier: (0, import_TlvObject.TlvField)(8, import_TlvNumber.TlvUInt8),
  ellipticCurvePublicKey: (0, import_TlvObject.TlvField)(9, import_TlvString.TlvByteString),
  extensions: (0, import_TlvObject.TlvField)(
    10,
    (0, import_TlvObject.TlvTaggedList)({
      basicConstraints: (0, import_TlvObject.TlvField)(
        1,
        (0, import_TlvObject.TlvObject)({
          isCa: (0, import_TlvObject.TlvField)(1, import_TlvBoolean.TlvBoolean)
        })
      ),
      keyUsage: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt16),
      subjectKeyIdentifier: (0, import_TlvObject.TlvField)(4, import_TlvString.TlvByteString.bound({ length: 20 })),
      authorityKeyIdentifier: (0, import_TlvObject.TlvField)(5, import_TlvString.TlvByteString.bound({ length: 20 }))
    })
  ),
  signature: (0, import_TlvObject.TlvField)(11, import_TlvString.TlvByteString)
});
const TlvCertificationDeclaration = (0, import_TlvObject.TlvObject)({
  formatVersion: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvUInt16),
  vendorId: (0, import_TlvObject.TlvField)(1, import_VendorId.TlvVendorId),
  produceIdArray: (0, import_TlvObject.TlvField)(2, (0, import_TlvArray.TlvArray)(import_TlvNumber.TlvUInt16, { minLength: 1, maxLength: 100 })),
  deviceTypeId: (0, import_TlvObject.TlvField)(3, import_TlvNumber.TlvUInt32),
  certificateId: (0, import_TlvObject.TlvField)(4, import_TlvString.TlvString.bound({ length: 19 })),
  securityLevel: (0, import_TlvObject.TlvField)(5, import_TlvNumber.TlvUInt8),
  securityInformation: (0, import_TlvObject.TlvField)(6, import_TlvNumber.TlvUInt16),
  versionNumber: (0, import_TlvObject.TlvField)(7, import_TlvNumber.TlvUInt16),
  certificationType: (0, import_TlvObject.TlvField)(8, import_TlvNumber.TlvUInt8),
  dacOriginVendorId: (0, import_TlvObject.TlvOptionalField)(9, import_VendorId.TlvVendorId),
  dacOriginProductId: (0, import_TlvObject.TlvOptionalField)(10, import_TlvNumber.TlvUInt16),
  authorizedPaaList: (0, import_TlvObject.TlvOptionalField)(
    11,
    (0, import_TlvArray.TlvArray)(import_TlvString.TlvByteString.bound({ length: 20 }), { minLength: 1, maxLength: 10 })
  )
});
class CertificateManager {
  static rootCertToAsn1({
    serialNumber,
    notBefore,
    notAfter,
    issuer: { issuerRcacId },
    subject: { rcacId },
    ellipticCurvePublicKey,
    extensions: { subjectKeyIdentifier, authorityKeyIdentifier }
  }) {
    return import_DerCodec.DerCodec.encode({
      version: (0, import_DerCodec.ContextTagged)(0, 2),
      serialNumber: serialNumber[0],
      signatureAlgorithm: import_DerCodec.EcdsaWithSHA256_X962,
      issuer: {
        issuerRcacId: issuerRcacId === void 0 ? void 0 : RcacId_Matter(issuerRcacId)
      },
      validity: {
        notBefore: matterToJsDate(notBefore),
        notAfter: matterToJsDate(notAfter)
      },
      subject: {
        rcacId: RcacId_Matter(rcacId)
      },
      publicKey: (0, import_DerCodec.PublicKeyEcPrime256v1_X962)(ellipticCurvePublicKey),
      extensions: (0, import_DerCodec.ContextTagged)(3, {
        basicConstraints: (0, import_DerCodec.BasicConstraints_X509)({ isCa: true }),
        keyUsage: import_DerCodec.KeyUsage_Signature_ContentCommited_X509,
        subjectKeyIdentifier: (0, import_DerCodec.SubjectKeyIdentifier_X509)(subjectKeyIdentifier),
        authorityKeyIdentifier: (0, import_DerCodec.AuthorityKeyIdentifier_X509)(authorityKeyIdentifier)
      })
    });
  }
  static nocCertToAsn1({
    serialNumber,
    notBefore,
    notAfter,
    issuer: { issuerRcacId },
    subject: { fabricId, nodeId, caseAuthenticatedTags },
    ellipticCurvePublicKey,
    extensions: { subjectKeyIdentifier, authorityKeyIdentifier }
  }) {
    if (caseAuthenticatedTags !== void 0) {
      import_CaseAuthenticatedTag.CaseAuthenticatedTag.validateNocTagList(caseAuthenticatedTags);
    }
    const cat0 = caseAuthenticatedTags?.[0];
    const cat1 = caseAuthenticatedTags?.[1];
    const cat2 = caseAuthenticatedTags?.[2];
    return import_DerCodec.DerCodec.encode({
      version: (0, import_DerCodec.ContextTagged)(0, 2),
      serialNumber: serialNumber[0],
      signatureAlgorithm: import_DerCodec.EcdsaWithSHA256_X962,
      issuer: {
        issuerRcacId: issuerRcacId === void 0 ? void 0 : RcacId_Matter(issuerRcacId)
      },
      validity: {
        notBefore: matterToJsDate(notBefore),
        notAfter: matterToJsDate(notAfter)
      },
      subject: {
        fabricId: FabricId_Matter(fabricId),
        nodeId: NodeId_Matter((0, import_NodeId.NodeId)(nodeId)),
        cat0: cat0 !== void 0 ? NocCat_Matter(cat0) : void 0,
        cat1: cat1 !== void 0 ? NocCat_Matter(cat1) : void 0,
        cat2: cat2 !== void 0 ? NocCat_Matter(cat2) : void 0
      },
      publicKey: (0, import_DerCodec.PublicKeyEcPrime256v1_X962)(ellipticCurvePublicKey),
      extensions: (0, import_DerCodec.ContextTagged)(3, {
        basicConstraints: (0, import_DerCodec.BasicConstraints_X509)({}),
        keyUsage: import_DerCodec.KeyUsage_Signature_X509,
        extendedKeyUsage: (0, import_DerCodec.ExtendedKeyUsage_X509)({ serverAuth: true, clientAuth: true }),
        subjectKeyIdentifier: (0, import_DerCodec.SubjectKeyIdentifier_X509)(subjectKeyIdentifier),
        authorityKeyIdentifier: (0, import_DerCodec.AuthorityKeyIdentifier_X509)(authorityKeyIdentifier)
      })
    });
  }
  static daCertToAsn1({
    serialNumber,
    notBefore,
    notAfter,
    issuer: { commonName: issuerCommonName, vendorId: issuerVendorId },
    subject: { commonName: subjectCommonName, vendorId: subjectVendorId, productId: subjectProductId },
    ellipticCurvePublicKey,
    extensions: { subjectKeyIdentifier, authorityKeyIdentifier }
  }, key) {
    const certificate = {
      version: (0, import_DerCodec.ContextTagged)(0, 2),
      serialNumber: serialNumber[0],
      signatureAlgorithm: import_DerCodec.EcdsaWithSHA256_X962,
      issuer: {
        commonName: CommonName_X520(issuerCommonName),
        vendorId: VendorId_Matter(issuerVendorId)
      },
      validity: {
        notBefore: matterToJsDate(notBefore),
        notAfter: matterToJsDate(notAfter)
      },
      subject: {
        commonName: CommonName_X520(subjectCommonName),
        vendorId: VendorId_Matter(subjectVendorId),
        productId: ProductId_Matter(subjectProductId)
      },
      publicKey: (0, import_DerCodec.PublicKeyEcPrime256v1_X962)(ellipticCurvePublicKey),
      extensions: (0, import_DerCodec.ContextTagged)(3, {
        basicConstraints: (0, import_DerCodec.BasicConstraints_X509)({
          isCa: false
        }),
        keyUsage: import_DerCodec.KeyUsage_Signature_X509,
        subjectKeyIdentifier: (0, import_DerCodec.SubjectKeyIdentifier_X509)(subjectKeyIdentifier),
        authorityKeyIdentifier: (0, import_DerCodec.AuthorityKeyIdentifier_X509)(authorityKeyIdentifier)
      })
    };
    return import_DerCodec.DerCodec.encode({
      certificate,
      signAlgorithm: import_DerCodec.EcdsaWithSHA256_X962,
      signature: (0, import_DerCodec.BitByteArray)(import_Crypto.Crypto.sign(key, import_DerCodec.DerCodec.encode(certificate), "der"))
    });
  }
  static paiCertToAsn1({
    serialNumber,
    notBefore,
    notAfter,
    issuer: { commonName: issuerCommonName, vendorId: issuerVendorId },
    subject: { commonName, vendorId, productId },
    ellipticCurvePublicKey,
    extensions: { subjectKeyIdentifier, authorityKeyIdentifier }
  }, key) {
    const certificate = {
      version: (0, import_DerCodec.ContextTagged)(0, 2),
      serialNumber: serialNumber[0],
      signatureAlgorithm: import_DerCodec.EcdsaWithSHA256_X962,
      issuer: {
        commonName: CommonName_X520(issuerCommonName),
        vendorId: issuerVendorId === void 0 ? void 0 : VendorId_Matter(issuerVendorId)
      },
      validity: {
        notBefore: matterToJsDate(notBefore),
        notAfter: matterToJsDate(notAfter)
      },
      subject: {
        commonName: CommonName_X520(commonName),
        vendorId: VendorId_Matter(vendorId),
        productId: productId === void 0 ? void 0 : ProductId_Matter(productId)
      },
      publicKey: (0, import_DerCodec.PublicKeyEcPrime256v1_X962)(ellipticCurvePublicKey),
      extensions: (0, import_DerCodec.ContextTagged)(3, {
        basicConstraints: (0, import_DerCodec.BasicConstraints_X509)({
          isCa: true,
          pathLen: 0
        }),
        keyUsage: import_DerCodec.KeyUsage_Signature_ContentCommited_X509,
        subjectKeyIdentifier: (0, import_DerCodec.SubjectKeyIdentifier_X509)(subjectKeyIdentifier),
        authorityKeyIdentifier: (0, import_DerCodec.AuthorityKeyIdentifier_X509)(authorityKeyIdentifier)
      })
    };
    return import_DerCodec.DerCodec.encode({
      certificate,
      signAlgorithm: import_DerCodec.EcdsaWithSHA256_X962,
      signature: (0, import_DerCodec.BitByteArray)(import_Crypto.Crypto.sign(key, import_DerCodec.DerCodec.encode(certificate), "der"))
    });
  }
  static paaCertToAsn1({
    serialNumber,
    notBefore,
    notAfter,
    issuer: { commonName: issuerCommonName, vendorId: issuerVendorId },
    subject: { commonName, vendorId },
    ellipticCurvePublicKey,
    extensions: { subjectKeyIdentifier, authorityKeyIdentifier }
  }, key) {
    const certificate = {
      version: (0, import_DerCodec.ContextTagged)(0, 2),
      serialNumber: serialNumber[0],
      signatureAlgorithm: import_DerCodec.EcdsaWithSHA256_X962,
      issuer: {
        commonName: CommonName_X520(issuerCommonName),
        vendorId: issuerVendorId === void 0 ? void 0 : VendorId_Matter(issuerVendorId)
      },
      validity: {
        notBefore: matterToJsDate(notBefore),
        notAfter: matterToJsDate(notAfter)
      },
      subject: {
        commonName: CommonName_X520(commonName),
        vendorId: vendorId === void 0 ? void 0 : VendorId_Matter(vendorId)
      },
      publicKey: (0, import_DerCodec.PublicKeyEcPrime256v1_X962)(ellipticCurvePublicKey),
      extensions: (0, import_DerCodec.ContextTagged)(3, {
        basicConstraints: (0, import_DerCodec.BasicConstraints_X509)({
          isCa: false
        }),
        keyUsage: import_DerCodec.KeyUsage_Signature_ContentCommited_X509,
        subjectKeyIdentifier: (0, import_DerCodec.SubjectKeyIdentifier_X509)(subjectKeyIdentifier),
        authorityKeyIdentifier: authorityKeyIdentifier === void 0 ? void 0 : (0, import_DerCodec.AuthorityKeyIdentifier_X509)(authorityKeyIdentifier)
      })
    };
    return import_DerCodec.DerCodec.encode({
      certificate,
      signAlgorithm: import_DerCodec.EcdsaWithSHA256_X962,
      signature: (0, import_DerCodec.BitByteArray)(import_Crypto.Crypto.sign(key, import_DerCodec.DerCodec.encode(certificate), "der"))
    });
  }
  static CertificationDeclarationToAsn1(eContent, subjectKeyIdentifier, privateKey) {
    const certificate = {
      version: 3,
      digestAlgorithm: [import_DerCodec.SHA256_CMS],
      encapContentInfo: (0, import_DerCodec.Pkcs7Data)(eContent),
      signerInfo: [
        {
          version: 3,
          subjectKeyIdentifier: (0, import_DerCodec.ContextTaggedBytes)(0, subjectKeyIdentifier),
          digestAlgorithm: import_DerCodec.SHA256_CMS,
          signatureAlgorithm: import_DerCodec.EcdsaWithSHA256_X962,
          signature: import_Crypto.Crypto.sign(privateKey, eContent, "der")
        }
      ]
    };
    return import_DerCodec.DerCodec.encode((0, import_DerCodec.Pkcs7SignedData)(certificate));
  }
  static validateRootCertificate(rootCert) {
    import_Crypto.Crypto.verify((0, import_Key.PublicKey)(rootCert.ellipticCurvePublicKey), this.rootCertToAsn1(rootCert), rootCert.signature);
  }
  static validateNocCertificate(rootCert, nocCert) {
    import_Crypto.Crypto.verify((0, import_Key.PublicKey)(rootCert.ellipticCurvePublicKey), this.nocCertToAsn1(nocCert), nocCert.signature);
  }
  static createCertificateSigningRequest(key) {
    const request = {
      version: 0,
      subject: { organization: (0, import_DerCodec.OrganisationName_X520)("CSR") },
      publicKey: (0, import_DerCodec.PublicKeyEcPrime256v1_X962)(key.publicKey),
      endSignedBytes: (0, import_DerCodec.ContextTagged)(0)
    };
    return import_DerCodec.DerCodec.encode({
      request,
      signAlgorithm: import_DerCodec.EcdsaWithSHA256_X962,
      signature: (0, import_DerCodec.BitByteArray)(import_Crypto.Crypto.sign(key, import_DerCodec.DerCodec.encode(request), "der"))
    });
  }
  static getPublicKeyFromCsr(csr) {
    const { [import_DerCodec.ELEMENTS_KEY]: rootElements } = import_DerCodec.DerCodec.decode(csr);
    if (rootElements?.length !== 3)
      throw new CertificateError("Invalid CSR data");
    const [requestNode, signAlgorithmNode, signatureNode] = rootElements;
    const { [import_DerCodec.ELEMENTS_KEY]: requestElements } = requestNode;
    if (requestElements?.length !== 4)
      throw new CertificateError("Invalid CSR data");
    const [versionNode, _subjectNode, publicKeyNode] = requestElements;
    const requestVersion = versionNode[import_DerCodec.BYTES_KEY][0];
    if (requestVersion !== 0)
      throw new CertificateError(`Unsupported request version${requestVersion}`);
    const { [import_DerCodec.ELEMENTS_KEY]: publicKeyElements } = publicKeyNode;
    if (publicKeyElements?.length !== 2)
      throw new CertificateError("Invalid CSR data");
    const [_publicKeyTypeNode, publicKeyBytesNode] = publicKeyElements;
    const publicKey = publicKeyBytesNode[import_DerCodec.BYTES_KEY];
    if (!import_DerCodec.EcdsaWithSHA256_X962[import_DerCodec.OBJECT_ID_KEY][import_DerCodec.BYTES_KEY].equals(signAlgorithmNode[import_DerCodec.ELEMENTS_KEY]?.[0]?.[import_DerCodec.BYTES_KEY]))
      throw new CertificateError("Unsupported signature type");
    import_Crypto.Crypto.verify((0, import_Key.PublicKey)(publicKey), import_DerCodec.DerCodec.encode(requestNode), signatureNode[import_DerCodec.BYTES_KEY], "der");
    return publicKey;
  }
}
