// @project-chip/matter.js/dist/cjs/certificate/RootCertificateManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RootCertificateManager_exports = {};
__export(RootCertificateManager_exports, {
  RootCertificateManager: () => RootCertificateManager
});
module.exports = __toCommonJS(RootCertificateManager_exports);
var import_Crypto = require("../crypto/Crypto.js");
var import_Key = require("../crypto/Key.js");
var import_Time = require("../time/Time.js");
var import_ByteArray = require("../util/ByteArray.js");
var import_CertificateManager = require("./CertificateManager.js");
/**
 * @license
 * Copyright 2022 The node-matter Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class RootCertificateManager {
  constructor(storageContext) {
    this.rootCertId = BigInt(0);
    this.rootKeyPair = import_Crypto.Crypto.createKeyPair();
    this.rootKeyIdentifier = import_Crypto.Crypto.hash(this.rootKeyPair.publicKey).slice(0, 20);
    this.rootCertBytes = this.generateRootCert();
    this.nextCertificateId = 1;
    const storage = storageContext.createContext("RootCertificateManager");
    if (storage.has("rootCertId")) {
      try {
        const rootCertId = storage.get("rootCertId");
        const rootKeyPair = storage.get("rootKeyPair");
        const rootKeyIdentifier = storage.get("rootKeyIdentifier");
        const rootCertBytes = storage.get("rootCertBytes");
        const nextCertificateId = storage.get("nextCertificateId");
        this.rootCertId = rootCertId;
        this.rootKeyPair = (0, import_Key.PrivateKey)(rootKeyPair);
        this.rootKeyIdentifier = rootKeyIdentifier;
        this.rootCertBytes = rootCertBytes;
        this.nextCertificateId = nextCertificateId;
        return;
      } catch (error) {
        console.error("Failed to load root certificate from storage, generating new one", error);
      }
    }
    storage.set("rootCertId", this.rootCertId);
    storage.set("rootKeyPair", this.rootKeyPair.keyPair);
    storage.set("rootKeyIdentifier", this.rootKeyIdentifier);
    storage.set("rootCertBytes", this.rootCertBytes);
    storage.set("nextCertificateId", this.nextCertificateId);
  }
  getRootCert() {
    return this.rootCertBytes;
  }
  generateRootCert() {
    const now = import_Time.Time.get().now();
    const unsignedCertificate = {
      serialNumber: import_ByteArray.ByteArray.of(Number(this.rootCertId)),
      signatureAlgorithm: 1,
      publicKeyAlgorithm: 1,
      ellipticCurveIdentifier: 1,
      issuer: { issuerRcacId: this.rootCertId },
      notBefore: (0, import_CertificateManager.jsToMatterDate)(now, -1),
      notAfter: (0, import_CertificateManager.jsToMatterDate)(now, 10),
      subject: { rcacId: this.rootCertId },
      ellipticCurvePublicKey: this.rootKeyPair.publicKey,
      extensions: {
        basicConstraints: { isCa: true },
        keyUsage: 96,
        subjectKeyIdentifier: this.rootKeyIdentifier,
        authorityKeyIdentifier: this.rootKeyIdentifier
      }
    };
    const signature = import_Crypto.Crypto.sign(this.rootKeyPair, import_CertificateManager.CertificateManager.rootCertToAsn1(unsignedCertificate));
    return import_CertificateManager.TlvRootCertificate.encode({ ...unsignedCertificate, signature });
  }
  generateNoc(publicKey, fabricId, nodeId, caseAuthenticatedTags) {
    const now = import_Time.Time.get().now();
    const certId = this.nextCertificateId++;
    const unsignedCertificate = {
      serialNumber: import_ByteArray.ByteArray.of(certId),
      // TODO: figure out what should happen if certId > 255
      signatureAlgorithm: 1,
      publicKeyAlgorithm: 1,
      ellipticCurveIdentifier: 1,
      issuer: { issuerRcacId: this.rootCertId },
      notBefore: (0, import_CertificateManager.jsToMatterDate)(now, -1),
      notAfter: (0, import_CertificateManager.jsToMatterDate)(now, 10),
      subject: { fabricId, nodeId, caseAuthenticatedTags },
      ellipticCurvePublicKey: publicKey,
      extensions: {
        basicConstraints: { isCa: false },
        keyUsage: 1,
        extendedKeyUsage: [2, 1],
        subjectKeyIdentifier: import_Crypto.Crypto.hash(publicKey).slice(0, 20),
        authorityKeyIdentifier: this.rootKeyIdentifier
      }
    };
    const signature = import_Crypto.Crypto.sign(this.rootKeyPair, import_CertificateManager.CertificateManager.nocCertToAsn1(unsignedCertificate));
    return import_CertificateManager.TlvOperationalCertificate.encode({ ...unsignedCertificate, signature });
  }
}
