// @project-chip/matter.js/dist/esm/model/definitions/FieldValue.js
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { UnexpectedDataError } from "../../common/MatterError.js";
import { serialize as stringSerialize } from "../../util/String.js";
var FieldValue;
((FieldValue2) => {
  FieldValue2.percent = "percent";
  FieldValue2.celsius = "celsius";
  FieldValue2.reference = "reference";
  FieldValue2.properties = "properties";
  function is(value, type) {
    return value && value.type === type;
  }
  FieldValue2.is = is;
  FieldValue2.Invalid = Symbol("invalid");
  function Reference(name) {
    return { type: FieldValue2.reference, name };
  }
  FieldValue2.Reference = Reference;
  function Celsius(value) {
    return { type: FieldValue2.celsius, value };
  }
  FieldValue2.Celsius = Celsius;
  function Percent(value) {
    return { type: FieldValue2.percent, value };
  }
  FieldValue2.Percent = Percent;
  function serialize(value) {
    if (value === null) {
      return "null";
    }
    if (is(value, FieldValue2.reference)) {
      return value.name;
    }
    if (is(value, FieldValue2.celsius)) {
      return `${value.value}\xB0C`;
    }
    if (is(value, FieldValue2.percent)) {
      return `${value.value}%';`;
    }
    if (is(value, FieldValue2.properties)) {
      return stringSerialize(value.properties);
    }
    return value.toString();
  }
  FieldValue2.serialize = serialize;
  function numericValue(value, typeName) {
    if (typeof value === "boolean") {
      return value ? 1 : 0;
    }
    if (typeof value === "number") {
      return value;
    }
    if (is(value, FieldValue2.celsius)) {
      switch (typeName) {
        case "temperature":
        case "temp-diff":
          return value.value * 100;
        case "temp-u8":
        case "temp-s8":
          return value.value * 10;
      }
      return;
    }
    if (is(value, FieldValue2.percent)) {
      switch (typeName) {
        case "percent100ths":
          return value.value * 100;
        default:
          return value.value;
      }
    }
  }
  FieldValue2.numericValue = numericValue;
  function objectValue(value) {
    if (is(value, FieldValue2.properties)) {
      return value.properties;
    }
  }
  FieldValue2.objectValue = objectValue;
  function unwrap(value, typeName) {
    if (value === null || typeof value !== "object" || Array.isArray(value) || value instanceof Date) {
      return value;
    }
    const type = value.type;
    switch (type) {
      case "properties":
        return objectValue(value);
      case "reference":
        return;
      case "percent":
      case "celsius":
        return numericValue(value, typeName);
      case void 0:
        throw new UnexpectedDataError(`Field value objects must be wrappers with "type" field`);
      default:
        throw new UnexpectedDataError(`Unsupported wrapped object type "${type}"`);
    }
  }
  FieldValue2.unwrap = unwrap;
  function referenced(value) {
    if (is(value, FieldValue2.reference)) {
      return value.name;
    }
  }
  FieldValue2.referenced = referenced;
})(FieldValue || (FieldValue = {}));
export {
  FieldValue
};
