// @project-chip/matter.js/dist/cjs/model/logic/cluster-variance/IllegalFeatureCombinations.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var IllegalFeatureCombinations_exports = {};
__export(IllegalFeatureCombinations_exports, {
  IllegalFeatureCombinations: () => IllegalFeatureCombinations
});
module.exports = __toCommonJS(IllegalFeatureCombinations_exports);
var import_MatterError = require("../../../common/MatterError.js");
var import_DeepEqual = require("../../../util/DeepEqual.js");
var import_aspects = require("../../aspects/index.js");
var import_FeatureBitmap = require("./FeatureBitmap.js");
/**
 * @license
 * Copyright 2022-2023 Project CHIP Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function IllegalFeatureCombinations(cluster) {
  const illegal = [];
  const choices = {};
  for (const f of cluster.features) {
    addFeatureNode(f, f.conformance.ast, illegal, choices);
  }
  function add(flags) {
    if (!illegal.some((e) => (0, import_DeepEqual.isDeepEqual)(e, flags))) {
      illegal.push(flags);
    }
  }
  for (const choice of Object.values(choices)) {
    if (choice.exclusive) {
      for (const f1 of choice.features) {
        for (const f2 of choice.features) {
          if (f1 !== f2) {
            add({ [f1]: true, [f2]: true });
          }
        }
      }
    }
    const flags = (0, import_FeatureBitmap.FeatureBitmap)();
    for (const f of choice.features) {
      flags[f] = false;
    }
    add(flags);
  }
  return { illegal, requiresFeatures: !!choices.length };
}
function addFeatureNode(feature, node, illegal, choices) {
  function unsupported() {
    throw new import_MatterError.InternalError(`New rule required to support ${feature.path} conformance "${feature.conformance}"`);
  }
  function addDisjunctRequirement(flags, node2) {
    switch (node2.type) {
      case import_aspects.Conformance.Special.Name:
        flags[node2.param] = false;
        break;
      case import_aspects.Conformance.OR:
        addDisjunctRequirement(flags, node2.param.lhs);
        addDisjunctRequirement(flags, node2.param.rhs);
        break;
      default:
        unsupported();
    }
  }
  function addConjunctRequirement(feature2, node2) {
    switch (node2.type) {
      case import_aspects.Conformance.Special.Name:
        illegal.push({ [feature2]: true, [node2.param]: false });
        break;
      case import_aspects.Conformance.AND:
        addConjunctRequirement(feature2, node2.param.lhs);
        addConjunctRequirement(feature2, node2.param.rhs);
        break;
    }
  }
  switch (node.type) {
    case import_aspects.Conformance.Special.Empty:
    case import_aspects.Conformance.Flag.Optional:
    case import_aspects.Conformance.Flag.Provisional:
      break;
    case import_aspects.Conformance.Special.Group:
      node.param.forEach((ast) => addFeatureNode(feature, ast, illegal, choices));
      break;
    case import_aspects.Conformance.Special.Choice:
      if (node.param.num > 1) {
        unsupported();
      }
      let choice = choices[node.param.name];
      if (choice) {
        choice.features.push(feature.name);
      } else {
        choice = choices[node.param.name] = {
          exclusive: !node.param.orMore,
          features: [feature.name]
        };
      }
      break;
    case import_aspects.Conformance.Special.Name:
      illegal.push({ [node.param]: true, [feature.name]: false });
      break;
    case import_aspects.Conformance.Special.OptionalIf:
      if (node.param.type == import_aspects.Conformance.AND) {
        addConjunctRequirement(feature.name, node.param);
      } else {
        const flags = (0, import_FeatureBitmap.FeatureBitmap)({ [feature.name]: true });
        addDisjunctRequirement(flags, node.param);
        illegal.push(flags);
      }
      break;
    default:
      unsupported();
      break;
  }
}
